// @taiga-ui/kit/esm5/components/pdf-viewer/pdf-viewer.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TUI_DIALOGS } from '@taiga-ui/cdk';
import { TuiButtonModule } from '@taiga-ui/core';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiPdfViewerComponent } from './pdf-viewer.component';
import { TuiPdfViewerDirective } from './pdf-viewer.directive';
import { TuiPdfViewerService } from './pdf-viewer.service';
var TuiPdfViewerModule = /** @class */ (function () {
    function TuiPdfViewerModule() {
    }
    TuiPdfViewerModule = __decorate([
        NgModule({
            imports: [CommonModule, TuiButtonModule, PolymorpheusModule],
            declarations: [TuiPdfViewerComponent, TuiPdfViewerDirective],
            exports: [TuiPdfViewerComponent, TuiPdfViewerDirective],
            providers: [
                {
                    provide: TUI_DIALOGS,
                    useExisting: TuiPdfViewerService,
                    multi: true,
                },
            ],
            entryComponents: [TuiPdfViewerComponent],
        })
    ], TuiPdfViewerModule);
    return TuiPdfViewerModule;
}());
export { TuiPdfViewerModule };
