// @taiga-ui/kit/esm5/components/pdf-viewer/pdf-viewer.directive.js
import { __decorate, __extends } from "tslib";
import { Directive } from '@angular/core';
import { AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import { TuiPdfViewerService } from './pdf-viewer.service';
var TuiPdfViewerDirective = /** @class */ (function (_super) {
    __extends(TuiPdfViewerDirective, _super);
    function TuiPdfViewerDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiPdfViewerDirective = __decorate([
        Directive({
            selector: 'ng-template[tuiPdfViewer]',
            providers: [
                {
                    provide: AbstractTuiDialogService,
                    useExisting: TuiPdfViewerService,
                },
            ],
            inputs: ['options: tuiPdfViewerOptions', 'open: tuiPdfViewer'],
            outputs: ['openChange: tuiPdfViewerChange'],
        })
    ], TuiPdfViewerDirective);
    return TuiPdfViewerDirective;
}(AbstractTuiDialogDirective));
export { TuiPdfViewerDirective };
