// @taiga-ui/core/esm2015/components/group/group.directive.js
import { __decorate, __param } from "tslib";
import { Directive, HostBinding, Inject, Input } from '@angular/core';
import { tuiDefaultProp, TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiGroupStylesComponent } from './group-styles.component';
let TuiGroupDirective = class TuiGroupDirective {
    constructor(directiveStyles) {
        this.orientation = 'horizontal';
        this.adaptive = false;
        this.collapsed = false;
        this.rounded = true;
        this.size = 'm';
        directiveStyles.addComponent(TuiGroupStylesComponent);
    }
    get orientationHorizontal() {
        return this.orientation === 'horizontal';
    }
    get orientationVertical() {
        return this.orientation === 'vertical';
    }
    get sizeLarge() {
        return this.size === 'l';
    }
};
TuiGroupDirective.ctorParameters = () => [
    { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "orientation", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_adaptive'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "adaptive", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_collapsed'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "collapsed", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_rounded'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "rounded", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "size", void 0);
__decorate([
    HostBinding('class.tui-group_orientation_horizontal')
], TuiGroupDirective.prototype, "orientationHorizontal", null);
__decorate([
    HostBinding('class.tui-group_orientation_vertical')
], TuiGroupDirective.prototype, "orientationVertical", null);
__decorate([
    HostBinding('class.tui-group_radius_large')
], TuiGroupDirective.prototype, "sizeLarge", null);
TuiGroupDirective = __decorate([
    Directive({
        selector: '[tuiGroup]:not(ng-container)',
        host: {
            class: 'tui-group',
            role: 'group',
        },
    }),
    __param(0, Inject(TuiDirectiveStylesService))
], TuiGroupDirective);
export { TuiGroupDirective };
