// @taiga-ui/kit/fesm2015/taiga-ui-kit-pipes-to-country-code.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { CHAR_PLUS } from '@taiga-ui/cdk';
import { TuiCountryIsoCode } from '@taiga-ui/i18n';
import { TUI_COUNTRIES_MASKS } from '@taiga-ui/kit/tokens';
import { tuiGetMaxAllowedPhoneLength, tuiIsoToCountryCode, tuiNotKzRegion } from '@taiga-ui/kit/utils';

// @dynamic
let TuiToCountryCodePipe = class TuiToCountryCodePipe {
    constructor(countriesMasks) {
        this.countriesMasks = countriesMasks;
    }
    transform(value, countries) {
        return countries.find(countryIsoCode => {
            const ruCodeTest = countryIsoCode === TuiCountryIsoCode.RU &&
                /^[78]/.test(value) &&
                /^(?!880[1-9 ])/.test(value) &&
                value.length + 1 ===
                    tuiGetMaxAllowedPhoneLength(this.countriesMasks, TuiCountryIsoCode.RU);
            const matched = ruCodeTest ||
                (value.startsWith(tuiIsoToCountryCode(this.countriesMasks, countryIsoCode).replace(CHAR_PLUS, ``)) &&
                    value.length + 1 ===
                        tuiGetMaxAllowedPhoneLength(this.countriesMasks, countryIsoCode));
            if (matched) {
                switch (countryIsoCode) {
                    case TuiCountryIsoCode.RU:
                        return tuiNotKzRegion(value);
                    case TuiCountryIsoCode.KZ:
                        return !tuiNotKzRegion(value);
                    default:
                        return true;
                }
            }
            return false;
        });
    }
};
TuiToCountryCodePipe.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_COUNTRIES_MASKS,] }] }
];
TuiToCountryCodePipe = __decorate([
    Pipe({
        name: `tuiToCountryCode`,
    }),
    __param(0, Inject(TUI_COUNTRIES_MASKS))
], TuiToCountryCodePipe);

let TuiExtractCountryCodeModule = class TuiExtractCountryCodeModule {
};
TuiExtractCountryCodeModule = __decorate([
    NgModule({
        declarations: [TuiToCountryCodePipe],
        exports: [TuiToCountryCodePipe],
    })
], TuiExtractCountryCodeModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiExtractCountryCodeModule, TuiToCountryCodePipe };
