// @taiga-ui/kit/fesm2015/taiga-ui-kit-components-accordion.js
import { __decorate, __param } from 'tslib';
import { TemplateRef, Inject, Self, ChangeDetectorRef, Directive, EventEmitter, ViewChild, Input, HostBinding, Output, ContentChild, Component, ChangeDetectionStrategy, forwardRef, ContentChildren, NgModule } from '@angular/core';
import { AbstractTuiInteractive, isNativeFocused, tuiDefaultProp, TUI_FOCUSABLE_ITEM_ACCESSOR, EMPTY_QUERY, itemsQueryListObservable, isPresent, TuiDestroyService, TuiFocusedModule, TuiFocusVisibleModule, TuiFocusableModule, TuiHoveredModule } from '@taiga-ui/cdk';
import { Observable, merge, identity } from 'rxjs';
import { pairwise, map, filter, switchMap, mapTo, takeUntil } from 'rxjs/operators';
import { TUI_MODE, MODE_PROVIDER, TuiSvgModule, TuiGroupModule, TuiExpandModule } from '@taiga-ui/core';
import { PolymorpheusTemplate, PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { CommonModule } from '@angular/common';

let TuiAccordionItemContentDirective = class TuiAccordionItemContentDirective extends PolymorpheusTemplate {
    constructor(templateRef, changeDetectorRef) {
        super(templateRef, changeDetectorRef);
    }
};
TuiAccordionItemContentDirective.ctorParameters = () => [
    { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }, { type: Self }] },
    { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] }
];
TuiAccordionItemContentDirective = __decorate([
    Directive({
        selector: 'ng-template[tuiAccordionItemContent]',
    }),
    __param(0, Inject(TemplateRef)),
    __param(0, Self()),
    __param(1, Inject(ChangeDetectorRef))
], TuiAccordionItemContentDirective);

let TuiAccordionItemEagerContentDirective = class TuiAccordionItemEagerContentDirective {
};
TuiAccordionItemEagerContentDirective = __decorate([
    Directive({
        selector: '[tuiAccordionItemContent]:not(ng-template)',
    })
], TuiAccordionItemEagerContentDirective);

var TuiAccordionItemComponent_1;
let TuiAccordionItemComponent = TuiAccordionItemComponent_1 = class TuiAccordionItemComponent extends AbstractTuiInteractive {
    constructor(changeDetectorRef, mode$) {
        super();
        this.changeDetectorRef = changeDetectorRef;
        this.mode$ = mode$;
        this.noPadding = false;
        this.showArrow = true;
        this.borders = 'all';
        this.size = 'm';
        this.disabled = false;
        this.disableHover = false;
        this.open = false;
        this.async = false;
        this.openChange = new EventEmitter();
    }
    get nativeFocusableElement() {
        return this.disabled || !this.focusableElement
            ? null
            : this.focusableElement.nativeElement;
    }
    get focused() {
        return isNativeFocused(this.nativeFocusableElement);
    }
    onHovered(hovered) {
        if (!this.disableHover) {
            this.updateHovered(hovered);
        }
    }
    onFocused(focused) {
        this.updateFocused(focused);
    }
    onFocusVisible(focusVisible) {
        this.updateFocusVisible(focusVisible);
    }
    onRowToggle() {
        if (!this.disabled) {
            this.updateOpen(!this.open);
        }
    }
    onItemKeyDownEsc(event) {
        if (!this.focused || !this.open) {
            return;
        }
        event.stopPropagation();
        this.updateOpen(false);
    }
    onItemKeyDownSpace(event) {
        if (!this.focused) {
            return;
        }
        event.preventDefault();
        this.onRowToggle();
    }
    close() {
        this.updateOpen(false);
        this.changeDetectorRef.markForCheck();
    }
    updateOpen(open) {
        if (this.open === open) {
            return;
        }
        this.open = open;
        this.openChange.emit(open);
    }
};
TuiAccordionItemComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] }
];
__decorate([
    ViewChild('focusableElement')
], TuiAccordionItemComponent.prototype, "focusableElement", void 0);
__decorate([
    Input(),
    HostBinding('class._no-padding'),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "noPadding", void 0);
__decorate([
    Input(),
    HostBinding('class._has-arrow'),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "showArrow", void 0);
__decorate([
    Input(),
    HostBinding('attr.data-tui-host-borders'),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "borders", void 0);
__decorate([
    Input(),
    HostBinding('attr.data-size'),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "size", void 0);
__decorate([
    Input(),
    HostBinding('class._disabled'),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "disabled", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "disableHover", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "open", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiAccordionItemComponent.prototype, "async", void 0);
__decorate([
    Output()
], TuiAccordionItemComponent.prototype, "openChange", void 0);
__decorate([
    ContentChild(TuiAccordionItemEagerContentDirective)
], TuiAccordionItemComponent.prototype, "eagerContent", void 0);
__decorate([
    ContentChild(TuiAccordionItemContentDirective)
], TuiAccordionItemComponent.prototype, "lazyContent", void 0);
TuiAccordionItemComponent = TuiAccordionItemComponent_1 = __decorate([
    Component({
        selector: 'tui-accordion-item',
        template: "<div\n    #focusableElement\n    automation-id=\"tui-accordion__item-wrapper\"\n    class=\"t-wrapper\"\n    [tuiFocusable]=\"computedFocusable\"\n    (tuiFocusVisibleChange)=\"onFocusVisible($event)\"\n    (tuiFocusedChange)=\"onFocused($event)\"\n    (keydown.space)=\"onItemKeyDownSpace($event)\"\n    (keydown.enter)=\"onItemKeyDownSpace($event)\"\n    (keydown.esc)=\"onItemKeyDownEsc($event)\"\n>\n    <div\n        automation-id=\"tui-accordion__item-header\"\n        class=\"t-header\"\n        [class.t-header_open]=\"open\"\n        [class.t-header_hoverable]=\"!disableHover\"\n        (tuiHoveredChange)=\"onHovered($event)\"\n        (click)=\"onRowToggle()\"\n    >\n        <div\n            automation-id=\"tui-accordion__item-title\"\n            class=\"t-title\"\n        >\n            <ng-content></ng-content>\n        </div>\n        <tui-svg\n            *ngIf=\"showArrow\"\n            automation-id=\"tui-accordion__item-arrow\"\n            src=\"tuiIconChevronDownLarge\"\n            class=\"t-icon\"\n            [class.t-icon_open]=\"open\"\n        ></tui-svg>\n    </div>\n    <tui-expand\n        [async]=\"async\"\n        [expanded]=\"open\"\n    >\n        <div\n            *ngIf=\"eagerContent\"\n            class=\"t-content\"\n        >\n            <ng-content select=\"[tuiAccordionItemContent]\"></ng-content>\n        </div>\n        <ng-template tuiExpandContent>\n            <div\n                *ngIf=\"lazyContent\"\n                polymorpheus-outlet\n                automation-id=\"tui-accordion__item-content\"\n                class=\"t-content\"\n                [content]=\"lazyContent\"\n            ></div>\n        </ng-template>\n    </tui-expand>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [
            {
                provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
                useExisting: forwardRef(() => TuiAccordionItemComponent_1),
            },
            MODE_PROVIDER,
        ],
        host: {
            '($.data-mode.attr)': 'mode$',
        },
        styles: [":host{position:relative;display:block;overflow:hidden;border-radius:var(--tui-radius-l)}:host[data-tui-host-borders=top-bottom]{border-radius:0!important}.t-wrapper{position:relative;border-radius:inherit}.t-wrapper:focus{outline:0}.t-wrapper:after{position:absolute;top:0;left:0;width:100%;height:100%;content:'';box-sizing:border-box;border-radius:inherit;border:1px solid var(--tui-base-04);pointer-events:none}:host:not([data-tui-host-borders]) .t-wrapper:after{border-width:0}:host[data-tui-host-borders=all] .t-wrapper:after{border-width:1px}:host[data-tui-host-borders=top-bottom] .t-wrapper:after{border-left-width:0;border-right-width:0}:host[data-tui-host-borders=top] .t-wrapper:after{border-left-width:0;border-right-width:0;border-bottom-width:0}:host[data-tui-host-borders=bottom] .t-wrapper:after{border-left-width:0;border-right-width:0;border-top-width:0}:host[data-mode=onDark] .t-wrapper:after{border-color:var(--tui-base-03)}:host[data-mode=onLight] .t-wrapper:after{border-color:var(--tui-text-01)}:host._focus-visible .t-wrapper:after{border:2px solid var(--tui-focus)}:host:not([data-mode]) .t-wrapper{background:var(--tui-base-01)}.t-header{transition-property:background;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;font:var(--tui-font-text-l);align-items:center;box-sizing:border-box;border-bottom:1px solid var(--tui-base-04);min-height:var(--tui-height-l);padding:.75rem 1.25rem;color:var(--tui-text-01);cursor:pointer}:host[data-mode=onDark] .t-header{color:var(--tui-text-01-night);border-color:var(--tui-base-03)}:host[data-mode=onLight] .t-header{border-color:var(--tui-text-01)}:host:not([data-tui-host-borders]) .t-header{border-bottom-width:0;box-shadow:none}:host._has-arrow .t-header{padding-right:.75rem}:host[data-mode=onDark] .t-header_open{background:var(--tui-clear-inverse)}:host[data-mode=onLight] .t-header_open{background:var(--tui-clear)}:host[data-size='s'] .t-header{font:var(--tui-font-text-m);min-height:var(--tui-height-m);padding:.625rem .75rem .625rem 1rem}:host._no-padding .t-header{padding-left:0;padding-right:0}:host._disabled .t-header{cursor:default}:host:hover:not([data-mode]) .t-header_hoverable{background:var(--tui-base-02)}:host:hover[data-mode=onDark] .t-header_hoverable{background:var(--tui-clear-inverse)}:host:hover[data-mode=onLight] .t-header_hoverable{background:var(--tui-clear)}.t-title{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;margin-right:0;flex-grow:1}:host._has-arrow .t-title{margin-right:.5rem}.t-icon{transition-property:transform,color;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;width:1.5rem;height:1.5rem;margin-left:auto;color:var(--tui-base-06)}.t-icon_open{transform:rotate(180deg)}:host[data-mode] .t-icon{opacity:var(--tui-disabled-opacity)}:host:hover[data-mode] .t-icon{opacity:.8}:host[data-mode=onDark] .t-icon{color:var(--tui-text-01-night)}:host[data-mode=onLight] .t-icon{color:var(--tui-text-01)}:host:hover:not([data-mode]) .t-icon{color:var(--tui-base-07)}.t-content{font:var(--tui-font-text-m);padding:1.25rem;word-wrap:break-word}:host[data-size='s'] .t-content{font:var(--tui-font-text-s);padding:1rem}"]
    }),
    __param(0, Inject(ChangeDetectorRef)),
    __param(1, Inject(TUI_MODE))
], TuiAccordionItemComponent);

let TuiAccordionComponent = class TuiAccordionComponent {
    constructor(destroy$) {
        this.destroy$ = destroy$;
        this.closeOthers = true;
        this.rounded = true;
        this.accordionItems = EMPTY_QUERY;
    }
    ngAfterContentInit() {
        const { accordionItems } = this;
        const rows$ = itemsQueryListObservable(accordionItems);
        const newOpenRow$ = rows$.pipe(pairwise(), map(([previous, current]) => current.find(item => !previous.includes(item) && item.open)), filter(isPresent));
        const rowsOpen$ = merge(rows$.pipe(switchMap(rows => merge(...rows.map(row => row.openChange.pipe(filter(identity), mapTo(row)))))), newOpenRow$).pipe(filter(() => this.closeOthers), takeUntil(this.destroy$));
        rowsOpen$.subscribe(currentRow => {
            accordionItems.forEach(row => {
                if (currentRow !== row) {
                    row.close();
                }
            });
        });
    }
};
TuiAccordionComponent.ctorParameters = () => [
    { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiAccordionComponent.prototype, "closeOthers", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiAccordionComponent.prototype, "rounded", void 0);
__decorate([
    ContentChildren(TuiAccordionItemComponent)
], TuiAccordionComponent.prototype, "accordionItems", void 0);
TuiAccordionComponent = __decorate([
    Component({
        selector: 'tui-accordion',
        template: "<!--TODO: Unnecessary nesting, probably accordion can be a directive-->\n<div\n    tuiGroup\n    automation-id=\"tui-accordion__group\"\n    orientation=\"vertical\"\n    size=\"l\"\n    class=\"t-group\"\n    [collapsed]=\"true\"\n    [rounded]=\"rounded\"\n>\n    <ng-content select=\"tui-accordion-item\"></ng-content>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [TuiDestroyService],
        styles: [":host{display:block}.t-group{display:flex}"]
    }),
    __param(0, Inject(TuiDestroyService))
], TuiAccordionComponent);

let TuiAccordionModule = class TuiAccordionModule {
};
TuiAccordionModule = __decorate([
    NgModule({
        imports: [
            CommonModule,
            TuiFocusedModule,
            TuiFocusVisibleModule,
            TuiFocusableModule,
            TuiHoveredModule,
            PolymorpheusModule,
            TuiSvgModule,
            TuiGroupModule,
            TuiExpandModule,
        ],
        declarations: [
            TuiAccordionComponent,
            TuiAccordionItemComponent,
            TuiAccordionItemContentDirective,
            TuiAccordionItemEagerContentDirective,
        ],
        exports: [
            TuiAccordionComponent,
            TuiAccordionItemComponent,
            TuiAccordionItemContentDirective,
            TuiAccordionItemEagerContentDirective,
        ],
    })
], TuiAccordionModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAccordionComponent, TuiAccordionItemComponent, TuiAccordionItemContentDirective, TuiAccordionItemEagerContentDirective, TuiAccordionModule };
