// @taiga-ui/kit/esm5/components/input-password/input-password.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputPasswordDirective = /** @class */ (function (_super) {
    __extends(TuiInputPasswordDirective, _super);
    function TuiInputPasswordDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputPasswordDirective_1 = TuiInputPasswordDirective;
    TuiInputPasswordDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    TuiInputPasswordDirective.prototype.process = function (input) {
        this.input = input;
    };
    TuiInputPasswordDirective.prototype.ngDoCheck = function () {
        if (this.input) {
            this.input.type = this.host.isPasswordHidden ? 'password' : 'text';
        }
    };
    var TuiInputPasswordDirective_1;
    TuiInputPasswordDirective = TuiInputPasswordDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-password',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputPasswordDirective_1; }),
                },
            ],
        })
    ], TuiInputPasswordDirective);
    return TuiInputPasswordDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputPasswordDirective };
