// @taiga-ui/kit/esm5/components/input-password/input-password-options.js
import { __assign } from "tslib";
import { InjectionToken } from '@angular/core';
// TODO: 3.0 remove in ivy compilation
export var PASSWORD_ICON_HIDE = function (_a) {
    var $implicit = _a.$implicit;
    return ($implicit === "s" ? "tuiIconEyeClosed" : "tuiIconHideLarge");
};
export var PASSWORD_ICON_SHOW = function (_a) {
    var $implicit = _a.$implicit;
    return $implicit === "s" ? "tuiIconEyeOpen" : "tuiIconShowLarge";
};
/** Default values for the input password options. */
export var TUI_INPUT_PASSWORD_DEFAULT_OPTIONS = {
    icons: {
        hide: PASSWORD_ICON_HIDE,
        show: PASSWORD_ICON_SHOW,
    },
};
export var TUI_INPUT_PASSWORD_OPTIONS = new InjectionToken("Default parameters for input password component", {
    factory: function () { return TUI_INPUT_PASSWORD_DEFAULT_OPTIONS; },
});
export var tuiInputPasswordOptionsProvider = function (options) { return ({
    provide: TUI_INPUT_PASSWORD_OPTIONS,
    useValue: __assign(__assign({}, TUI_INPUT_PASSWORD_DEFAULT_OPTIONS), options),
}); };
