// @taiga-ui/core/esm5/components/primitive-textfield/primitive-textfield.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, ElementRef, EventEmitter, HostBinding, HostListener, Inject, Input, Output, QueryList, ViewChild, } from '@angular/core';
import { AbstractTuiInteractive, isNativeFocusedIn, setNativeFocused, tuiDefaultProp, tuiPure, } from '@taiga-ui/cdk';
import { TUI_HINT_WATCHED_CONTROLLER, TuiHintControllerDirective, } from '@taiga-ui/core/directives/hint-controller';
import { TUI_TEXTFIELD_WATCHED_CONTROLLER, TuiTextfieldController, } from '@taiga-ui/core/directives/textfield-controller';
import { TUI_MODE, TUI_TEXTFIELD_APPEARANCE } from '@taiga-ui/core/tokens';
import { getBorder } from '@taiga-ui/core/utils/miscellaneous';
import { PolymorpheusOutletComponent } from '@tinkoff/ng-polymorpheus';
import { fromEvent, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { TUI_PRIMITIVE_TEXTFIELD_PROVIDERS } from './primitive-textfield.providers';
import { TUI_PRIMITIVE_TEXTFIELD_OPTIONS, } from './primitive-textfield-options';
var ICON_PADDING = 1.75;
var ICON_PADDING_S = 1.5;
var TuiPrimitiveTextfieldComponent = /** @class */ (function (_super) {
    __extends(TuiPrimitiveTextfieldComponent, _super);
    function TuiPrimitiveTextfieldComponent(mode$, appearance, controller, hintController, options, elementRef) {
        var _this = _super.call(this) || this;
        _this.mode$ = mode$;
        _this.appearance = appearance;
        _this.controller = controller;
        _this.hintController = hintController;
        _this.options = options;
        _this.elementRef = elementRef;
        _this.editable = true;
        _this.filler = '';
        _this.iconAlign = _this.options.iconAlign;
        // TODO: 3.0 Remove null
        _this.iconContent = null;
        _this.iconCleaner = _this.options.iconCleaner;
        _this.readOnly = false;
        _this.invalid = false;
        _this.disabled = false;
        _this.prefix = '';
        _this.postfix = '';
        _this.value = '';
        _this.valueChange = new EventEmitter();
        /**
         * @deprecated TODO: 3.0 remove
         */
        _this.autofilledChange = new EventEmitter();
        _this.autofilled = false;
        return _this;
    }
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "nativeFocusableElement", {
        get: function () {
            if (this.computedDisabled || !this.focusableElement) {
                return null;
            }
            // TODO: 3.0 Refactor this after we drop built-in input element
            return (this.focusableElement.nativeElement
                .previousElementSibling ||
                this.focusableElement.nativeElement);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "focused", {
        get: function () {
            return isNativeFocusedIn(this.elementRef.nativeElement);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "size", {
        get: function () {
            return this.controller.size;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "computedInvalid", {
        get: function () {
            return !this.readOnly && !this.disabled && this.invalid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "inputHidden", {
        get: function () {
            var _a;
            return !!((_a = this.content) === null || _a === void 0 ? void 0 : _a.length);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasValue", {
        get: function () {
            return !!this.value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasCleaner", {
        get: function () {
            return (this.controller.cleaner && this.hasValue && !this.disabled && !this.readOnly);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasTooltip", {
        get: function () {
            var _a;
            return !!((_a = this.hintController) === null || _a === void 0 ? void 0 : _a.content) && !this.disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasCustomContent", {
        get: function () {
            return !!this.controller.customContent;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "showOnlyPlaceholder", {
        get: function () {
            return (this.focused &&
                this.placeholderVisible &&
                (this.size === 's' || (this.size === 'm' && !this.placeholderRaisable)));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "placeholderVisible", {
        get: function () {
            var _a;
            var hasDecor = this.controller.exampleText ||
                this.prefix ||
                this.postfix || ((_a = this.nativeFocusableElement) === null || _a === void 0 ? void 0 : _a.placeholder);
            var showDecor = hasDecor && !this.readOnly && this.computedFocused;
            return !this.hasValue && !showDecor;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasPlaceholder", {
        get: function () {
            return (!this.showOnlyPlaceholder &&
                (this.placeholderRaisable || this.placeholderVisible));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "placeholderRaised", {
        get: function () {
            return (this.placeholderRaisable &&
                ((this.computedFocused && !this.readOnly) || this.hasValue || this.autofilled));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "borderStart", {
        get: function () {
            return this.iconAlignLeft ? this.iconPaddingLeft : 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "borderEnd", {
        get: function () {
            return getBorder(this.iconAlignRight, this.hasCleaner, this.hasTooltip, this.hasCustomContent);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "iconAlignLeft", {
        get: function () {
            return this.hasIcon && this.iconAlign === 'left';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "iconAlignRight", {
        get: function () {
            return this.hasIcon && this.iconAlign === 'right';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "name", {
        // Safari expiration date autofill workaround
        get: function () {
            return this.controller.autocomplete === "cc-exp" /* CcExp */
                ? 'ccexpiryyear'
                : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "computedId", {
        get: function () {
            var _a;
            return ((_a = this.nativeFocusableElement) === null || _a === void 0 ? void 0 : _a.id) || '';
        },
        enumerable: true,
        configurable: true
    });
    TuiPrimitiveTextfieldComponent.prototype.onFocused = function (focused) {
        this.updateFocused(focused);
    };
    TuiPrimitiveTextfieldComponent.prototype.getIndent$ = function (element) {
        return fromEvent(element, 'scroll').pipe(map(function () { return -1 * element.scrollLeft; }));
    };
    TuiPrimitiveTextfieldComponent.prototype.clear = function () {
        if (this.nativeFocusableElement) {
            this.nativeFocusableElement.value = '';
        }
        this.updateValue('');
    };
    TuiPrimitiveTextfieldComponent.prototype.onMouseDown = function (event) {
        var nativeFocusableElement = this.nativeFocusableElement;
        if (!nativeFocusableElement || event.target === nativeFocusableElement) {
            return;
        }
        event.preventDefault();
        setNativeFocused(nativeFocusableElement);
    };
    TuiPrimitiveTextfieldComponent.prototype.onModelChange = function (value) {
        this.updateValue(value);
    };
    TuiPrimitiveTextfieldComponent.prototype.onHovered = function (hovered) {
        this.updateHovered(hovered);
    };
    TuiPrimitiveTextfieldComponent.prototype.onAutofilled = function (autofilled) {
        this.updateAutofilled(autofilled);
    };
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "iconPaddingLeft", {
        get: function () {
            return this.size === 's' ? ICON_PADDING_S : ICON_PADDING;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "placeholderRaisable", {
        get: function () {
            return this.size !== 's' && !this.controller.labelOutside;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveTextfieldComponent.prototype, "hasIcon", {
        get: function () {
            return !!this.iconContent;
        },
        enumerable: true,
        configurable: true
    });
    TuiPrimitiveTextfieldComponent.prototype.updateAutofilled = function (autofilled) {
        if (this.autofilled === autofilled) {
            return;
        }
        this.autofilled = autofilled;
        this.autofilledChange.emit(autofilled);
    };
    TuiPrimitiveTextfieldComponent.prototype.updateValue = function (value) {
        this.value = value;
        this.valueChange.emit(value);
    };
    TuiPrimitiveTextfieldComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
        { type: String, decorators: [{ type: Inject, args: [TUI_TEXTFIELD_APPEARANCE,] }] },
        { type: TuiTextfieldController, decorators: [{ type: Inject, args: [TUI_TEXTFIELD_WATCHED_CONTROLLER,] }] },
        { type: TuiHintControllerDirective, decorators: [{ type: Inject, args: [TUI_HINT_WATCHED_CONTROLLER,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_PRIMITIVE_TEXTFIELD_OPTIONS,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    __decorate([
        ViewChild('focusableElement')
    ], TuiPrimitiveTextfieldComponent.prototype, "focusableElement", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "editable", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "filler", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "iconAlign", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "iconContent", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "iconCleaner", void 0);
    __decorate([
        Input(),
        HostBinding('class._readonly'),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "readOnly", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "invalid", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "disabled", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "prefix", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "postfix", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveTextfieldComponent.prototype, "value", void 0);
    __decorate([
        Output()
    ], TuiPrimitiveTextfieldComponent.prototype, "valueChange", void 0);
    __decorate([
        Output()
    ], TuiPrimitiveTextfieldComponent.prototype, "autofilledChange", void 0);
    __decorate([
        ContentChildren(PolymorpheusOutletComponent)
    ], TuiPrimitiveTextfieldComponent.prototype, "content", void 0);
    __decorate([
        HostBinding('attr.data-size')
    ], TuiPrimitiveTextfieldComponent.prototype, "size", null);
    __decorate([
        HostBinding('class._invalid')
    ], TuiPrimitiveTextfieldComponent.prototype, "computedInvalid", null);
    __decorate([
        HostBinding('class._hidden')
    ], TuiPrimitiveTextfieldComponent.prototype, "inputHidden", null);
    __decorate([
        HostBinding('style.--border-start.rem')
    ], TuiPrimitiveTextfieldComponent.prototype, "borderStart", null);
    __decorate([
        HostBinding('style.--border-end.rem')
    ], TuiPrimitiveTextfieldComponent.prototype, "borderEnd", null);
    __decorate([
        HostListener('focusin', ['true']),
        HostListener('focusout', ['false'])
    ], TuiPrimitiveTextfieldComponent.prototype, "onFocused", null);
    __decorate([
        tuiPure
    ], TuiPrimitiveTextfieldComponent.prototype, "getIndent$", null);
    TuiPrimitiveTextfieldComponent = __decorate([
        Component({
            selector: 'tui-primitive-textfield',
            template: "<ng-container *ngIf=\"content?.changes | async\"></ng-container>\n<tui-wrapper\n    automation-id=\"tui-primitive-textfield__wrapper\"\n    [appearance]=\"appearance\"\n    [readOnly]=\"readOnly\"\n    [disabled]=\"disabled\"\n    [focused]=\"computedFocused\"\n    [hovered]=\"computedHovered\"\n    [pressed]=\"computedPressed\"\n    [invalid]=\"computedInvalid\"\n    [style.--text-indent.px]=\"decor.pre$ | async\"\n    (mousedown)=\"onMouseDown($event)\"\n    (click.prevent.silent)=\"(0)\"\n    (tuiHoveredChange)=\"onHovered($event)\"\n    (tuiAutofilledChange)=\"onAutofilled($event)\"\n>\n    <ng-content select=\"input\"></ng-content>\n    <input\n        #focusableElement\n        tuiMaskAccessor\n        automation-id=\"tui-primitive-textfield__native-input\"\n        class=\"t-input\"\n        [attr.disabled]=\"computedDisabled || null\"\n        [attr.maxLength]=\"controller.maxLength\"\n        [attr.name]=\"name\"\n        [attr.aria-placeholder]=\"controller.exampleText\"\n        [attr.aria-invalid]=\"computedInvalid\"\n        [autocomplete]=\"controller.autocomplete\"\n        [type]=\"controller.type\"\n        [id]=\"id\"\n        [readOnly]=\"readOnly || !editable\"\n        [tuiInputMode]=\"controller.inputMode\"\n        [tuiFocusable]=\"computedFocusable\"\n        [tuiDescribedBy]=\"id\"\n        [ngModel]=\"value\"\n        (ngModelChange)=\"onModelChange($event)\"\n    />\n    <div\n        *ngIf=\"inputHidden\"\n        automation-id=\"tui-primitive-textfield__value\"\n        class=\"t-input t-input_template\"\n    >\n        <ng-content select=\"[polymorpheus-outlet]\"></ng-content>\n    </div>\n    <div class=\"t-content\">\n        <div\n            *ngIf=\"iconAlignLeft\"\n            polymorpheus-outlet\n            class=\"t-icon t-icon_left\"\n            [content]=\"iconContent\"\n            [context]=\"{$implicit: size}\"\n        >\n            <ng-template let-iconSrc>\n                <tui-svg [src]=\"iconSrc\"></tui-svg>\n            </ng-template>\n        </div>\n        <div class=\"t-wrapper\">\n            <label\n                *ngIf=\"hasPlaceholder\"\n                automation-id=\"tui-primitive-textfield__placeholder\"\n                class=\"t-placeholder\"\n                [class.t-placeholder_raised]=\"placeholderRaised\"\n                [for]=\"computedId\"\n            >\n                <ng-content></ng-content>\n            </label>\n            <div class=\"t-wrapper-value-decoration\">\n                <tui-value-decoration\n                    #decor\n                    automation-id=\"tui-primitive-textfield__value-decoration\"\n                    aria-hidden=\"true\"\n                    class=\"t-value-decoration\"\n                    [class.t-has-value]=\"value\"\n                    [style.textIndent.px]=\"getIndent$(focusableElement) | async\"\n                ></tui-value-decoration>\n            </div>\n        </div>\n        <div\n            *ngIf=\"hasCustomContent\"\n            polymorpheus-outlet\n            automation-id=\"tui-primitive-textfield__custom-content\"\n            class=\"t-custom-content\"\n            [content]=\"controller.customContent\"\n        >\n            <ng-template let-iconSrc>\n                <tui-svg\n                    class=\"t-custom-icon\"\n                    [src]=\"iconSrc\"\n                ></tui-svg>\n            </ng-template>\n        </div>\n        <span\n            *ngIf=\"hasCleaner\"\n            polymorpheus-outlet\n            automation-id=\"tui-primitive-textfield__cleaner\"\n            class=\"t-cleaner\"\n            [content]=\"iconCleaner\"\n            (click.stop)=\"clear()\"\n        >\n            <ng-template let-icon>\n                <tui-svg [src]=\"icon\"></tui-svg>\n            </ng-template>\n        </span>\n        <tui-tooltip\n            *ngIf=\"hasTooltip\"\n            automation-id=\"tui-primitive-textfield__tooltip\"\n            [describeId]=\"computedId\"\n            [content]=\"hintController.content || ''\"\n            [direction]=\"hintController.direction\"\n            [mode]=\"hintController.mode\"\n            [showDelay]=\"hintController.showDelay\"\n            [hideDelay]=\"hintController.hideDelay\"\n        ></tui-tooltip>\n        <div\n            *ngIf=\"iconAlignRight\"\n            polymorpheus-outlet\n            class=\"t-icon\"\n            [content]=\"iconContent\"\n            [context]=\"{$implicit: size}\"\n        >\n            <ng-template let-iconSrc>\n                <tui-svg [src]=\"iconSrc\"></tui-svg>\n            </ng-template>\n        </div>\n    </div>\n</tui-wrapper>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: TUI_PRIMITIVE_TEXTFIELD_PROVIDERS,
            host: {
                '($.data-mode.attr)': 'mode$',
                '[class._autofilled]': 'autofilled',
                '[class._label-outside]': 'controller.labelOutside',
            },
            styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);position:relative;display:block;border-radius:var(--tui-radius-m);text-align:left}:host[data-size='s']{height:var(--tui-height-s);min-height:var(--tui-height-s);max-height:var(--tui-height-s)}:host[data-size='m']{height:var(--tui-height-m);min-height:var(--tui-height-m);max-height:var(--tui-height-m)}:host[data-size='l']{height:var(--tui-height-l);min-height:var(--tui-height-l);max-height:var(--tui-height-l);font:var(--tui-font-text-m);line-height:1.25rem}.t-input{font:var(--tui-font-text-s);color:var(--tui-text-01);padding:0;margin:0;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;padding:0 var(--tui-padding-m);border:solid transparent;border-width:0 var(--border-end,0) 0 var(--border-start,0);border-inline-start-width:var(--border-start,0);border-inline-end-width:var(--border-end,0);text-indent:var(--text-indent);text-align:inherit;box-sizing:border-box;white-space:nowrap;overflow:hidden;text-transform:inherit;resize:none}.t-input:-webkit-autofill,.t-input:-webkit-autofill:focus,.t-input:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill,.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill:focus,.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill:hover,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill:focus,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill:hover,:host[data-mode=onDark] .t-input:-webkit-autofill,:host[data-mode=onDark] .t-input:-webkit-autofill:focus,:host[data-mode=onDark] .t-input:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01-night)!important;border-color:var(--tui-autofill-night);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill-night) inset!important}.t-input :host-context(tui-primitive-textfield[data-size='s']):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size='s']):not(tui-text-area),:host[data-size='s'] .t-input{padding:0 var(--tui-padding-s)}.t-input :host-context(tui-primitive-textfield[data-size='l']):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size='l']):not(tui-text-area),:host[data-size='l'] .t-input{padding:0 var(--tui-padding-l)}.t-input :host-context(tui-primitive-textfield._disabled),.t-input :host-context(tui-text-area._disabled),:host._disabled .t-input{pointer-events:none}.t-input :host-context(tui-primitive-textfield[data-size='l']:not(._label-outside)):not(tui-primitive-textfield),:host[data-size='l']:not(._label-outside) .t-input{padding-top:1.25rem}.t-input :host-context(tui-primitive-textfield[data-size='l']:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder,:host[data-size='l']:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder{font-size:.8156rem;transform:translateY(-.625rem)}.t-input :host-context(tui-primitive-textfield[data-size='m']:not(._label-outside)):not(tui-primitive-textfield),:host[data-size='m']:not(._label-outside) .t-input{padding-top:1.125rem}.t-input :host-context(tui-primitive-textfield[data-size='m']:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder,:host[data-size='m']:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder{font-size:.69rem;transform:translateY(-.5rem)}.t-input :host-context(tui-primitive-textfield._hidden),:host._hidden input.t-input{opacity:0;text-indent:-10em;-webkit-user-select:none}.t-content{display:flex;height:100%;width:100%;padding:0 var(--tui-padding-m);box-sizing:border-box;align-items:center;overflow:hidden}:host[data-size='s'] .t-content{padding:0 var(--tui-padding-s)}:host[data-size='l'] .t-content{padding:0 var(--tui-padding-l)}.t-content:after{content:'';margin-right:-.25rem}:host[data-size='m'] .t-content:after{display:none}.t-wrapper{flex:1;min-width:0;padding-right:.25rem;-webkit-padding-end:.25rem;padding-inline-end:.25rem;-webkit-padding-start:0;padding-inline-start:0}.t-placeholder{transition-property:transform,font-size,color,letter-spacing;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;width:100%;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;color:var(--tui-text-02);pointer-events:none;will-change:transform;transform:translateY(0)}.t-placeholder_raised{transform:translateY(-.625rem)}:host[data-size='m'] .t-placeholder_raised{font:var(--tui-font-text-xs);transform:translateY(-.5rem);letter-spacing:.025rem}:host._invalid:not(._focused) .t-placeholder_raised,:host._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill)}:host[data-mode=onDark]._invalid:not(._focused) .t-placeholder_raised,:host[data-mode=onDark]._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill-night)}:host._focused .t-placeholder,:host[data-size='l']._focused._label-outside .t-placeholder,:host[data-size='m']._focused._label-outside .t-placeholder{color:var(--tui-text-03)}:host[data-size='l'] .t-placeholder{font-size:.9375rem}:host[data-size='l'] .t-placeholder_raised{font-size:.8156rem}:host[data-size='l']._focused:not(._label-outside) .t-placeholder,:host[data-size='m']._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01)}:host[data-mode=onDark] .t-placeholder{color:var(--tui-text-02-night)}:host[data-size='l'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01-night)}:host[data-mode=onDark]._focused .t-placeholder,:host[data-size='l'][data-mode=onDark]._focused._label-outside .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused._label-outside .t-placeholder{color:var(--tui-text-02-night)}@supports (-webkit-hyphens:none){.t-placeholder{will-change:unset;transition-property:transform,color,letter-spacing}}.t-cleaner{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03);position:relative;box-sizing:border-box;cursor:pointer;transition-property:color,transform}.t-cleaner:hover{color:var(--tui-text-02)}:host._disabled .t-cleaner,:host._readonly .t-cleaner{pointer-events:none}:host[data-mode=onDark] .t-cleaner{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-cleaner:hover{color:var(--tui-text-01-night)}.t-icon{display:flex;align-items:center;justify-content:center;color:var(--tui-text-03)}.t-icon_left{margin:0 .5rem 0 -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem}:host[data-size='s'] .t-icon_left{margin-right:.25rem;-webkit-margin-end:.25rem;margin-inline-end:.25rem}:host._autofilled .t-value-decoration:not(.t-has-value){visibility:hidden}:host._autofilled[data-size='m']:not(._focused) .t-placeholder,:host._autofilled[data-size='s']:not(._focused) .t-placeholder{visibility:hidden}.t-custom-content{position:relative;display:flex;align-items:center;justify-content:center;min-width:2rem;height:2rem;margin-right:.25rem;pointer-events:none}.t-custom-icon{width:2rem;height:100%}.t-icon{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03);box-sizing:border-box;cursor:pointer;transition-property:color,transform;position:relative;pointer-events:none}.t-icon:hover{color:var(--tui-text-02)}:host._disabled .t-icon,:host._readonly .t-icon{pointer-events:none}:host[data-mode=onDark] .t-icon{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-icon:hover{color:var(--tui-text-01-night)}:host:hover:not(._readonly) .t-icon{color:var(--tui-text-02)}:host[data-mode=onDark]:hover:not(._readonly) .t-icon{color:var(--tui-text-01-night)}:host[data-mode=onLight]:hover:not(._readonly) .t-icon{color:var(--tui-text-01)}.t-input:not(:first-child){display:none}:host[data-size] .t-input_template{display:flex;width:calc(100% - 1rem);max-width:calc(100% - 1rem);align-items:center;pointer-events:none;-webkit-padding-end:0;padding-inline-end:0}.t-input:-webkit-autofill,.t-input:-webkit-autofill::first-line{font-size:inherit;line-height:inherit}.t-input::-webkit-caps-lock-indicator,.t-input::-webkit-contacts-auto-fill-button,.t-input::-webkit-credentials-auto-fill-button,.t-input::-webkit-credit-card-auto-fill-button,.t-input::-webkit-strong-password-auto-fill-button{content:none!important;position:absolute;left:-62.4375rem;top:-62.4375rem;z-index:-999;display:none!important;background:0 0!important;pointer-events:none!important}.t-text-template{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.t-wrapper-value-decoration{position:relative;pointer-events:none}"]
        }),
        __param(0, Inject(TUI_MODE)),
        __param(1, Inject(TUI_TEXTFIELD_APPEARANCE)),
        __param(2, Inject(TUI_TEXTFIELD_WATCHED_CONTROLLER)),
        __param(3, Inject(TUI_HINT_WATCHED_CONTROLLER)),
        __param(4, Inject(TUI_PRIMITIVE_TEXTFIELD_OPTIONS)),
        __param(5, Inject(ElementRef))
    ], TuiPrimitiveTextfieldComponent);
    return TuiPrimitiveTextfieldComponent;
}(AbstractTuiInteractive));
export { TuiPrimitiveTextfieldComponent };
