// @taiga-ui/core/esm5/components/primitive-textfield/primitive-textfield.directive.js
import { __decorate, __extends, __param } from "tslib";
import { Directive, forwardRef, Inject } from '@angular/core';
import { TuiAbstractTextfieldHost } from '@taiga-ui/core/abstract';
import { TUI_TEXTFIELD_HOST } from '@taiga-ui/core/tokens';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
var TuiPrimitiveTextfieldDirective = /** @class */ (function (_super) {
    __extends(TuiPrimitiveTextfieldDirective, _super);
    function TuiPrimitiveTextfieldDirective(textfield) {
        var _this = 
        /**
         * TuiPrimitiveTextfieldComponent satisfies all required parts of
         * AbstractTuiControl interface. A new interface is not introduced
         * so AbstractTuiControl is automatically injected for all other controls.
         */
        _super.call(this, textfield) || this;
        _this.textfield = textfield;
        return _this;
    }
    TuiPrimitiveTextfieldDirective_1 = TuiPrimitiveTextfieldDirective;
    Object.defineProperty(TuiPrimitiveTextfieldDirective.prototype, "readOnly", {
        get: function () {
            return this.textfield.readOnly || !this.textfield.editable;
        },
        enumerable: true,
        configurable: true
    });
    TuiPrimitiveTextfieldDirective.prototype.onValueChange = function (value) {
        this.textfield.onModelChange(value);
    };
    var TuiPrimitiveTextfieldDirective_1;
    TuiPrimitiveTextfieldDirective.ctorParameters = function () { return [
        { type: TuiPrimitiveTextfieldComponent, decorators: [{ type: Inject, args: [TuiPrimitiveTextfieldComponent,] }] }
    ]; };
    TuiPrimitiveTextfieldDirective = TuiPrimitiveTextfieldDirective_1 = __decorate([
        Directive({
            selector: 'tui-primitive-textfield',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiPrimitiveTextfieldDirective_1; }),
                },
            ],
        }),
        __param(0, Inject(TuiPrimitiveTextfieldComponent))
    ], TuiPrimitiveTextfieldDirective);
    return TuiPrimitiveTextfieldDirective;
}(TuiAbstractTextfieldHost));
export { TuiPrimitiveTextfieldDirective };
