// @taiga-ui/core/esm2015/directives/hint-controller/hint-controller.directive.js
var TuiHintControllerDirective_1;
import { __decorate, __param } from "tslib";
import { Directive, forwardRef, Inject, Input } from '@angular/core';
import { TuiController, tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_HINT_OPTIONS } from '@taiga-ui/core/directives/hint';
import { TUI_HINT_CONTROLLER } from './hint-controller.token';
// TODO: v2.0 use in Charts
let TuiHintControllerDirective = TuiHintControllerDirective_1 = class TuiHintControllerDirective extends TuiController {
    constructor(options) {
        super();
        this.options = options;
        // TODO: 3.0 Remove null
        this.content = null;
        this.direction = this.options.direction;
        this.mode = this.options.mode;
        this.showDelay = this.options.tuiHintShowDelay;
        this.hideDelay = this.options.tuiHintHideDelay;
    }
};
TuiHintControllerDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_HINT_OPTIONS,] }] }
];
__decorate([
    Input('tuiHintContent'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "content", void 0);
__decorate([
    Input('tuiHintDirection'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "direction", void 0);
__decorate([
    Input('tuiHintMode'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "mode", void 0);
__decorate([
    Input('tuiHintShowDelay'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "showDelay", void 0);
__decorate([
    Input('tuiHintHideDelay'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "hideDelay", void 0);
TuiHintControllerDirective = TuiHintControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiHintContent]',
        providers: [
            {
                provide: TUI_HINT_CONTROLLER,
                useExisting: forwardRef(() => TuiHintControllerDirective_1),
            },
        ],
    }),
    __param(0, Inject(TUI_HINT_OPTIONS))
], TuiHintControllerDirective);
export { TuiHintControllerDirective };
