// @taiga-ui/core/esm5/components/label/label.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, HostBinding, Inject, Input, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_MODE } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
var TuiLabelComponent = /** @class */ (function () {
    function TuiLabelComponent(mode$) {
        this.mode$ = mode$;
        // TODO: 3.0 Rename to tuiLabel
        this.label = '';
        this.context = {
            $implicit: null,
        };
    }
    TuiLabelComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiLabelComponent.prototype, "label", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiLabelComponent.prototype, "context", void 0);
    __decorate([
        ContentChild(NgControl),
        HostBinding('class._control')
    ], TuiLabelComponent.prototype, "control", void 0);
    TuiLabelComponent = __decorate([
        Component({
            selector: 'label[tuiLabel]',
            template: "<span\n    polymorpheus-outlet\n    class=\"t-label\"\n    [content]=\"label\"\n    [context]=\"context\"\n>\n    <ng-template let-text>\n        <span class=\"t-text\">{{ text }}</span>\n    </ng-template>\n</span>\n<span class=\"t-content\">\n    <ng-content></ng-content>\n</span>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [MODE_PROVIDER],
            host: {
                '($.data-mode.attr)': 'mode$',
            },
            styles: [":host{display:flex;min-height:2.75rem;justify-content:space-between;flex-direction:column;pointer-events:none}.t-label{font:var(--tui-font-text-s);max-width:100%;align-self:inherit;overflow:inherit;pointer-events:auto}.t-text{display:block;overflow:inherit;text-overflow:ellipsis}:host:not(._control) .t-text{color:var(--tui-text-02)}:host([data-mode=onDark]) .t-text{color:var(--tui-text-02-night)}.t-content{line-height:1.5rem;overflow:inherit;text-overflow:ellipsis;pointer-events:auto}"]
        }),
        __param(0, Inject(TUI_MODE))
    ], TuiLabelComponent);
    return TuiLabelComponent;
}());
export { TuiLabelComponent };
