// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-validator.js
import { __decorate } from 'tslib';
import { Input, Directive, forwardRef, NgModule } from '@angular/core';
import { Validators, NG_VALIDATORS } from '@angular/forms';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiDefaultProp } from '@taiga-ui/cdk/decorators';

var TuiValidatorDirective_1;
let TuiValidatorDirective = TuiValidatorDirective_1 = class TuiValidatorDirective {
    constructor() {
        this.onChange = EMPTY_FUNCTION;
        this.tuiValidator = Validators.nullValidator;
    }
    validate(control) {
        return this.tuiValidator(control);
    }
    registerOnValidatorChange(onChange) {
        this.onChange = onChange;
    }
    ngOnChanges() {
        this.onChange();
    }
    ngOnDestroy() {
        this.tuiValidator = Validators.nullValidator;
        this.onChange();
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], TuiValidatorDirective.prototype, "tuiValidator", void 0);
TuiValidatorDirective = TuiValidatorDirective_1 = __decorate([
    Directive({
        selector: '[tuiValidator]',
        providers: [
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => TuiValidatorDirective_1),
                multi: true,
            },
        ],
    })
], TuiValidatorDirective);

let TuiValidatorModule = class TuiValidatorModule {
};
TuiValidatorModule = __decorate([
    NgModule({
        declarations: [TuiValidatorDirective],
        exports: [TuiValidatorDirective],
    })
], TuiValidatorModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiValidatorDirective, TuiValidatorModule };
