// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-autofilled.js
import { __decorate, __param } from 'tslib';
import { Component, ViewEncapsulation, EventEmitter, Inject, HostBinding, Output, HostListener, Directive, NgModule } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk/services';

let TuiAutofilledStyleComponent = class TuiAutofilledStyleComponent {
};
TuiAutofilledStyleComponent = __decorate([
    Component({
        template: '',
        encapsulation: ViewEncapsulation.None,
        styles: [".tui-autofill input,input.tui-autofill{transition:box-shadow 10ms}"]
    })
], TuiAutofilledStyleComponent);

let TuiAutofilledDirective = class TuiAutofilledDirective {
    constructor(directiveStyles) {
        this.autofilled = false;
        this.tuiAutofilledChange = new EventEmitter();
        directiveStyles.addComponent(TuiAutofilledStyleComponent);
    }
    transitionStartHandler({ propertyName, target }) {
        var _a;
        const matchedAutofill = propertyName.includes('box-shadow') && ((_a = target) === null || _a === void 0 ? void 0 : _a.matches('input'));
        if (matchedAutofill) {
            this.autofilled = !this.autofilled;
            this.tuiAutofilledChange.emit(this.autofilled);
        }
    }
};
TuiAutofilledDirective.ctorParameters = () => [
    { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
];
__decorate([
    HostBinding('class._autofilled')
], TuiAutofilledDirective.prototype, "autofilled", void 0);
__decorate([
    Output()
], TuiAutofilledDirective.prototype, "tuiAutofilledChange", void 0);
__decorate([
    HostListener('transitionstart', ['$event'])
], TuiAutofilledDirective.prototype, "transitionStartHandler", null);
TuiAutofilledDirective = __decorate([
    Directive({
        selector: '[tuiAutofilledChange]',
        host: { class: 'tui-autofill' },
    }),
    __param(0, Inject(TuiDirectiveStylesService))
], TuiAutofilledDirective);

let TuiAutofilledModule = class TuiAutofilledModule {
};
TuiAutofilledModule = __decorate([
    NgModule({
        exports: [TuiAutofilledDirective],
        entryComponents: [TuiAutofilledStyleComponent],
        declarations: [TuiAutofilledDirective, TuiAutofilledStyleComponent],
    })
], TuiAutofilledModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAutofilledDirective, TuiAutofilledModule, TuiAutofilledStyleComponent };
