// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-classes.js
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';

const tuiAssert = {
    enabled: false,
    get assert() {
        return this.enabled
            ? Function.prototype.bind.call(console.assert, console)
            : EMPTY_FUNCTION;
    },
};
// TODO: replace `HTMLElement` to `Element` in v3.0
// TS 3.8 Argument of type Argument of type 'TuiNativeFocusableElement | HTMLElement' is not assignable to parameter of type 'HTMLElement'.
function tuiAssertIsHTMLElement(node) {
    var _a, _b;
    const defaultView = (_b = (_a = node) === null || _a === void 0 ? void 0 : _a.ownerDocument) === null || _b === void 0 ? void 0 : _b.defaultView;
    const isElement = !!defaultView &&
        (node instanceof defaultView.Element || node instanceof defaultView.HTMLDocument);
    tuiAssert.assert(isElement, `Node is not an Element`);
}

class TuiValidationError {
    constructor(message, context = {}) {
        this.message = message;
        this.context = context;
    }
}

/**
 * Generated bundle index. Do not edit.
 */

export { TuiValidationError, tuiAssert, tuiAssertIsHTMLElement };
