// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-components-dropdown-host.js
import { __decorate } from 'tslib';
import { ViewChild, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { AbstractTuiPortalHostComponent, AbstractTuiPortalService } from '@taiga-ui/cdk/abstract';
import { TuiPortalService } from '@taiga-ui/cdk/components/portal-host';
import { EMPTY_CLIENT_RECT } from '@taiga-ui/cdk/constants';

/**
 * Host element for dynamically created portals, for example using {@link TuiDropdownDirective}.
 */
let TuiDropdownHostComponent = class TuiDropdownHostComponent extends AbstractTuiPortalHostComponent {
    fixedPositionOffset() {
        var _a;
        return (((_a = this.positionFixedOffsetRef) === null || _a === void 0 ? void 0 : _a.nativeElement.getBoundingClientRect()) ||
            EMPTY_CLIENT_RECT);
    }
};
__decorate([
    ViewChild('positionFixedOffset')
], TuiDropdownHostComponent.prototype, "positionFixedOffsetRef", void 0);
TuiDropdownHostComponent = __decorate([
    Component({
        selector: 'tui-dropdown-host',
        template: "<div\n    #positionFixedOffset\n    class=\"t-position-fixed-offset\"\n></div>\n<ng-content></ng-content>\n<ng-container #viewContainer></ng-container>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [{ provide: AbstractTuiPortalService, useExisting: TuiPortalService }],
        styles: [":host{position:relative;z-index:0;display:block;height:100%}:host:before{content:'';display:block;overflow:hidden}.t-position-fixed-offset{position:fixed;left:0;top:0;pointer-events:none;visibility:hidden;width:100%;height:100%}"]
    })
], TuiDropdownHostComponent);

let TuiDropdownHostModule = class TuiDropdownHostModule {
};
TuiDropdownHostModule = __decorate([
    NgModule({
        declarations: [TuiDropdownHostComponent],
        exports: [TuiDropdownHostComponent],
    })
], TuiDropdownHostModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiDropdownHostComponent, TuiDropdownHostModule };
