// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-enums.js
/* eslint-disable @typescript-eslint/naming-convention */
var TuiCreditCardAutofillName;
(function (TuiCreditCardAutofillName) {
    TuiCreditCardAutofillName["Off"] = "off";
    TuiCreditCardAutofillName["CcName"] = "cc-name";
    TuiCreditCardAutofillName["CcNumber"] = "cc-number";
    TuiCreditCardAutofillName["CcCsc"] = "cc-csc";
    TuiCreditCardAutofillName["CcExpMonth"] = "cc-exp-month";
    /**
     * @deprecated
     * TODO: 3.0 replace with {@link CcExpMonth}
     */
    TuiCreditCardAutofillName["CcExp_mounth"] = "cc-exp-month";
    TuiCreditCardAutofillName["CcExpYear"] = "cc-exp-year";
    /**
     * @deprecated
     * TODO: 3.0 replace with {@link CcExpYear}
     */
    TuiCreditCardAutofillName["CcExp_year"] = "cc-exp-year";
    TuiCreditCardAutofillName["CcExp"] = "cc-exp";
    TuiCreditCardAutofillName["CcType"] = "cc-type";
})(TuiCreditCardAutofillName || (TuiCreditCardAutofillName = {}));
var TuiNameAutofillName;
(function (TuiNameAutofillName) {
    TuiNameAutofillName["Off"] = "off";
    TuiNameAutofillName["Name"] = "name";
    TuiNameAutofillName["GivenName"] = "given-name";
    TuiNameAutofillName["AdditionalName"] = "additional-name";
    TuiNameAutofillName["FamilyName"] = "family-name";
})(TuiNameAutofillName || (TuiNameAutofillName = {}));
var TuiAccountAutofillName;
(function (TuiAccountAutofillName) {
    TuiAccountAutofillName["Off"] = "off";
    TuiAccountAutofillName["Username"] = "username";
    TuiAccountAutofillName["NewPassword"] = "new-password";
    TuiAccountAutofillName["CurrentPassword"] = "current-password";
})(TuiAccountAutofillName || (TuiAccountAutofillName = {}));
var TuiEmailAutofillName;
(function (TuiEmailAutofillName) {
    TuiEmailAutofillName["Off"] = "off";
    TuiEmailAutofillName["Email"] = "email";
})(TuiEmailAutofillName || (TuiEmailAutofillName = {}));
var TuiAddressAutofillName;
(function (TuiAddressAutofillName) {
    TuiAddressAutofillName["Off"] = "off";
    TuiAddressAutofillName["StreetAddress"] = "street-address";
    TuiAddressAutofillName["PostalCode"] = "postal-code";
    TuiAddressAutofillName["CountryName"] = "country-name";
})(TuiAddressAutofillName || (TuiAddressAutofillName = {}));
var TuiPhoneAutofillName;
(function (TuiPhoneAutofillName) {
    TuiPhoneAutofillName["Off"] = "off";
    TuiPhoneAutofillName["Tel"] = "tel";
})(TuiPhoneAutofillName || (TuiPhoneAutofillName = {}));
var TuiDateAutofillName;
(function (TuiDateAutofillName) {
    TuiDateAutofillName["Off"] = "off";
    TuiDateAutofillName["Bday"] = "bday";
})(TuiDateAutofillName || (TuiDateAutofillName = {}));
var TuiTransactionAutofillName;
(function (TuiTransactionAutofillName) {
    TuiTransactionAutofillName["Off"] = "off";
    TuiTransactionAutofillName["TransactionCurrency"] = "transaction-currency";
    TuiTransactionAutofillName["TransactionAmount"] = "transaction-amount";
})(TuiTransactionAutofillName || (TuiTransactionAutofillName = {}));

var TuiDayOfWeek;
(function (TuiDayOfWeek) {
    TuiDayOfWeek[TuiDayOfWeek["Sunday"] = 0] = "Sunday";
    TuiDayOfWeek[TuiDayOfWeek["Monday"] = 1] = "Monday";
    TuiDayOfWeek[TuiDayOfWeek["Tuesday"] = 2] = "Tuesday";
    TuiDayOfWeek[TuiDayOfWeek["Wednesday"] = 3] = "Wednesday";
    TuiDayOfWeek[TuiDayOfWeek["Thursday"] = 4] = "Thursday";
    TuiDayOfWeek[TuiDayOfWeek["Friday"] = 5] = "Friday";
    TuiDayOfWeek[TuiDayOfWeek["Saturday"] = 6] = "Saturday";
})(TuiDayOfWeek || (TuiDayOfWeek = {}));

/**
 * @deprecated use join type {@link TuiInputModeT}
 */
var TuiInputMode;
(function (TuiInputMode) {
    TuiInputMode["None"] = "none";
    TuiInputMode["Text"] = "text";
    TuiInputMode["Decimal"] = "decimal";
    TuiInputMode["Numeric"] = "numeric";
    TuiInputMode["Tel"] = "tel";
    TuiInputMode["Email"] = "email";
    TuiInputMode["Url"] = "url";
    TuiInputMode["Search"] = "search";
})(TuiInputMode || (TuiInputMode = {}));

/**
 * @deprecated use join type {@link TuiInputTypeT}
 */
var TuiInputType;
(function (TuiInputType) {
    TuiInputType["Text"] = "text";
    TuiInputType["Tel"] = "tel";
    TuiInputType["Email"] = "email";
    TuiInputType["Url"] = "url";
    TuiInputType["Password"] = "password";
})(TuiInputType || (TuiInputType = {}));

var TuiMonthNumber;
(function (TuiMonthNumber) {
    TuiMonthNumber[TuiMonthNumber["January"] = 0] = "January";
    TuiMonthNumber[TuiMonthNumber["February"] = 1] = "February";
    TuiMonthNumber[TuiMonthNumber["March"] = 2] = "March";
    TuiMonthNumber[TuiMonthNumber["April"] = 3] = "April";
    TuiMonthNumber[TuiMonthNumber["May"] = 4] = "May";
    TuiMonthNumber[TuiMonthNumber["June"] = 5] = "June";
    TuiMonthNumber[TuiMonthNumber["July"] = 6] = "July";
    TuiMonthNumber[TuiMonthNumber["August"] = 7] = "August";
    TuiMonthNumber[TuiMonthNumber["September"] = 8] = "September";
    TuiMonthNumber[TuiMonthNumber["October"] = 9] = "October";
    TuiMonthNumber[TuiMonthNumber["November"] = 10] = "November";
    TuiMonthNumber[TuiMonthNumber["December"] = 11] = "December";
})(TuiMonthNumber || (TuiMonthNumber = {}));

/**
 * Modes for {@link TuiOverscrollDirective}:
 *
 * All — blocks all scroll actions
 * Scroll — blocks scroll actions only if content is scrollable (equivalent to overscroll-behavior: contain)
 * None — blocking is disabled
 * @deprecated use join type {@link TuiOverscrollModeT}
 */
var TuiOverscrollMode;
(function (TuiOverscrollMode) {
    TuiOverscrollMode["All"] = "all";
    TuiOverscrollMode["Scroll"] = "scroll";
    TuiOverscrollMode["None"] = "none";
})(TuiOverscrollMode || (TuiOverscrollMode = {}));

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAccountAutofillName, TuiAddressAutofillName, TuiCreditCardAutofillName, TuiDateAutofillName, TuiDayOfWeek, TuiEmailAutofillName, TuiInputMode, TuiInputType, TuiMonthNumber, TuiNameAutofillName, TuiOverscrollMode, TuiPhoneAutofillName, TuiTransactionAutofillName };
