// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-utils-miscellaneous.js
import { tuiAssert } from '@taiga-ui/cdk/classes';
import { FormArray, FormGroup } from '@angular/forms';

/**
 * @deprecated: use {@link tuiDistanceBetweenTouches} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function distanceBetweenTouches({ touches }) {
    return Math.hypot(touches[0].clientX - touches[1].clientX, touches[0].clientY - touches[1].clientY);
}
const tuiDistanceBetweenTouches = distanceBetweenTouches;

/**
 * @deprecated: use {@link tuiEaseInOutQuad} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function easeInOutQuad(t) {
    tuiAssert.assert(t >= 0 && t <= 1, `Input must be between 0 and 1 inclusive but received `, t);
    return t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;
}
const tuiEaseInOutQuad = easeInOutQuad;

/**
 * @deprecated: use {@link tuiIsPresent} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isPresent(value) {
    return value !== null && value !== undefined;
}
const tuiIsPresent = isPresent;

/**
 * @deprecated use nullish coalescing operator
 * If the value is null or undefined, returns default value
 * @param value
 * @param fallback alternative fallback
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function fallbackValue(value, fallback) {
    return isPresent(value) ? value : fallback;
}

/**
 * @deprecated: use {@link tuiFlatLength} instead
 * Flattens two-dimensional array and calculates resulting length
 *
 * @param array twi dimensional array
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function flatLength(array) {
    return array.reduce((count, section) => count + section.length, 0);
}
const tuiFlatLength = flatLength;

/**
 * @deprecated: use {@link tuiGetOriginalArrayFromQueryList} instead
 * Extracts original array from {@link QueryList} rather than
 * creating a copy like {@link QueryList.toArray} does.
 * @param queryList
 * @returns original array from {@link QueryList}.
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getOriginalArrayFromQueryList(queryList) {
    let array = [];
    queryList.find((_item, _index, originalArray) => {
        array = originalArray;
        return true;
    });
    return array;
}
const tuiGetOriginalArrayFromQueryList = getOriginalArrayFromQueryList;

/**
 * @deprecated: use {@link tuiGetSwipeDirection} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getSwipeDirection(deltaX, deltaY) {
    if (Math.abs(deltaY) > Math.abs(deltaX)) {
        return deltaY > 0 ? `top` : `bottom`;
    }
    return deltaX > 0 ? `left` : `right`;
}
const tuiGetSwipeDirection = getSwipeDirection;

/**
 * @deprecated: use identity from rxjs
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function identity(item) {
    return item;
}

/**
 * @deprecated: use {@link tuiIsElementEditable} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isElementEditable(element) {
    // TODO: iframe warning
    return ((element instanceof HTMLInputElement && !element.readOnly) ||
        (element instanceof HTMLTextAreaElement && !element.readOnly) ||
        element.isContentEditable);
}
const tuiIsElementEditable = isElementEditable;

/**
 * @deprecated: use {@link tuiIsNumber} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isNumber(value) {
    return typeof value === `number`;
}
const tuiIsNumber = isNumber;

/**
 * @deprecated: use {@link tuiMarkControlAsTouchedAndValidate} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function markControlAsTouchedAndValidate(control) {
    if (control instanceof FormArray) {
        control.controls.forEach(nestedControl => {
            markControlAsTouchedAndValidate(nestedControl);
        });
    }
    if (control instanceof FormGroup) {
        Object.values(control.controls).forEach(nestedControl => {
            markControlAsTouchedAndValidate(nestedControl);
        });
    }
    control.markAsTouched();
    control.updateValueAndValidity();
}
const tuiMarkControlAsTouchedAndValidate = markControlAsTouchedAndValidate;

/**
 * @deprecated: use {@link tuiNullableSame} instead
 * Checks identity for nullable elements.
 *
 * @param a element a
 * @param b element b
 * @param handler called if both elements are not null
 * @return true if either both are null or they pass identity handler
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function nullableSame(a, b, handler) {
    if (a === null) {
        return b === null;
    }
    if (b === null) {
        return false;
    }
    return handler(a, b);
}
const tuiNullableSame = nullableSame;

/**
 * @deprecated: use {@link tuiUniqBy} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function uniqBy(array, key) {
    return Array.from(array
        .reduce((map, item) => (map.has(item[key]) ? map : map.set(item[key], item)), new Map())
        .values());
}
const tuiUniqBy = uniqBy;

/**
 * Generated bundle index. Do not edit.
 */

export { distanceBetweenTouches, easeInOutQuad, fallbackValue, flatLength, getOriginalArrayFromQueryList, getSwipeDirection, identity, isElementEditable, isNumber, isPresent, markControlAsTouchedAndValidate, nullableSame, tuiDistanceBetweenTouches, tuiEaseInOutQuad, tuiFlatLength, tuiGetOriginalArrayFromQueryList, tuiGetSwipeDirection, tuiIsElementEditable, tuiIsNumber, tuiIsPresent, tuiMarkControlAsTouchedAndValidate, tuiNullableSame, tuiUniqBy, uniqBy };
