// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-components-alert-host.js
import { __decorate, __param } from 'tslib';
import { Injector, Inject, INJECTOR, Component, ChangeDetectionStrategy, ViewEncapsulation, NgModule } from '@angular/core';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk/constants';
import { TUI_ALERTS } from '@taiga-ui/cdk/tokens';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import { CommonModule } from '@angular/common';
import { TuiMapperPipeModule } from '@taiga-ui/cdk/pipes';

let TuiAlertHostComponent = class TuiAlertHostComponent {
    constructor(alerts, injector) {
        this.alerts = alerts;
        this.injector = injector;
        this.mapper = useValue => Injector.create({
            providers: [
                {
                    provide: POLYMORPHEUS_CONTEXT,
                    useValue,
                },
            ],
            parent: this.injector,
        });
    }
};
TuiAlertHostComponent.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [TUI_ALERTS,] }] },
    { type: Injector, decorators: [{ type: Inject, args: [INJECTOR,] }] }
];
TuiAlertHostComponent = __decorate([
    Component({
        selector: 'tui-alert-host',
        template: "<div\n    *ngFor=\"let alert$ of alerts\"\n    class=\"t-notifications-wrapper\"\n    @tuiParentAnimation\n>\n    <ng-container\n        *ngFor=\"let item of alert$ | async\"\n        [ngComponentOutlet]=\"item.component.component\"\n        [ngComponentOutletInjector]=\"item | tuiMapper: mapper\"\n    ></ng-container>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        animations: [TUI_PARENT_ANIMATION],
        encapsulation: ViewEncapsulation.None,
        styles: ["tui-alert-host>.t-notifications-wrapper{position:fixed;top:0;left:0;display:flex;width:100%;height:100%;flex-direction:column;pointer-events:none}tui-alert-host>.t-notifications-wrapper>*{pointer-events:auto}"]
    }),
    __param(0, Inject(TUI_ALERTS)),
    __param(1, Inject(INJECTOR))
], TuiAlertHostComponent);

let TuiAlertHostModule = class TuiAlertHostModule {
};
TuiAlertHostModule = __decorate([
    NgModule({
        imports: [CommonModule, TuiMapperPipeModule],
        declarations: [TuiAlertHostComponent],
        exports: [TuiAlertHostComponent],
    })
], TuiAlertHostModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAlertHostComponent, TuiAlertHostModule };
