// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-zoom.js
import { __decorate, __param } from 'tslib';
import { Inject, Directive, NgModule } from '@angular/core';
import { TuiZoomService } from '@taiga-ui/cdk/services';
import { Observable } from 'rxjs';

// @dynamic
let TuiZoomDirective = class TuiZoomDirective {
    constructor(tuiZoom) {
        this.tuiZoom = tuiZoom;
    }
};
TuiZoomDirective.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [TuiZoomService,] }] }
];
TuiZoomDirective = __decorate([
    Directive({
        selector: '[tuiZoom]',
        outputs: ['tuiZoom'],
        providers: [TuiZoomService],
        host: {
            '[style.touch-action]': '"none"',
        },
    }),
    __param(0, Inject(TuiZoomService))
], TuiZoomDirective);

/**
 * @experimental
 */
let TuiZoomModule = class TuiZoomModule {
};
TuiZoomModule = __decorate([
    NgModule({
        declarations: [TuiZoomDirective],
        exports: [TuiZoomDirective],
    })
], TuiZoomModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiZoomDirective, TuiZoomModule };
