// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-utils-dom.js
import { clamp } from '@taiga-ui/cdk/utils/math';
import { isPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { tuiAssert } from '@taiga-ui/cdk/classes';

/**
 * @deprecated: use {@link tuiCanScroll} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function canScroll(element, rootElement, vertical, scrollEnd) {
    return vertical
        ? canScrollVertical(element, rootElement, scrollEnd)
        : canScrollHorizontal(element, rootElement, scrollEnd);
}
const tuiCanScroll = canScroll;
function canScrollVertical(element, rootElement, scrollEnd) {
    let currentElement = element;
    while (currentElement !== rootElement.parentElement) {
        if ((Math.floor(currentElement.scrollTop) > 0 && !scrollEnd) ||
            (Math.ceil(currentElement.scrollTop + currentElement.clientHeight) <
                currentElement.scrollHeight &&
                scrollEnd)) {
            return true;
        }
        if (currentElement.parentElement) {
            currentElement = currentElement.parentElement;
        }
        else {
            return false;
        }
    }
    return false;
}
function canScrollHorizontal(element, rootElement, scrollEnd) {
    let currentElement = element;
    while (currentElement !== rootElement.parentElement) {
        if ((Math.floor(currentElement.scrollLeft) > 0 && !scrollEnd) ||
            (Math.ceil(currentElement.scrollLeft + currentElement.clientWidth) <
                currentElement.scrollWidth &&
                scrollEnd)) {
            return true;
        }
        if (currentElement.parentElement) {
            currentElement = currentElement.parentElement;
        }
        else {
            return false;
        }
    }
    return false;
}

/**
 * @deprecated: use {@link tuiContainsOrAfter} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function containsOrAfter(current, node) {
    return (current.contains(node) ||
        !!(node.compareDocumentPosition(current) & Node.DOCUMENT_POSITION_PRECEDING));
}
const tuiContainsOrAfter = containsOrAfter;

/** @deprecated only needed for IE, remove in 3.0 */
function tuiCustomEvent(name, { bubbles = false, cancelable = false, detail = null } = {}, documentRef) {
    if (typeof CustomEvent === `function`) {
        return new CustomEvent(name, { bubbles, cancelable, detail });
    }
    const customEvent = documentRef.createEvent(`CustomEvent`);
    customEvent.initCustomEvent(name, bubbles, cancelable, detail);
    return customEvent;
}

/**
 * @deprecated: use {@link tuiGetActualTarget} instead
 * Gets actual target from open Shadow DOM if event happened within it
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getActualTarget(event) {
    if (`composedPath` in event) {
        return event.composedPath()[0];
    }
    return event.target;
}
const tuiGetActualTarget = getActualTarget;

const DEFAULT_FORMAT = `text/plain`;
/**
 * @deprecated: use {@link tuiGetClipboardDataText} instead
 * Gets text from data of clipboardEvent, it also works in IE and Edge browsers
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getClipboardDataText(event, format = DEFAULT_FORMAT) {
    return `clipboardData` in event && event.clipboardData !== null
        ? event.clipboardData.getData(format) ||
            event.clipboardData.getData(DEFAULT_FORMAT)
        : event.target.ownerDocument.defaultView.clipboardData.getData(`text`);
}
const tuiGetClipboardDataText = getClipboardDataText;

/**
 * @deprecated only needed for IE, remove in 3.0
 * Gets the closest element by selector i.e. {@link Element.closest}
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getClosestElement(element, selector) {
    const closest = Element.prototype.closest;
    if (closest) {
        return closest.call(element, selector);
    }
    const matchesSelector = Element.prototype.matches || Element.prototype.msMatchesSelector;
    do {
        if (matchesSelector.call(element, selector)) {
            return element;
        }
        element = element.parentElement;
    } while (element !== null);
    return null;
}

/**
 * @deprecated: use {@link tuiGetDocumentOrShadowRoot} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getDocumentOrShadowRoot(node) {
    return `getRootNode` in node && node.isConnected
        ? node.getRootNode()
        : node.ownerDocument;
}
const tuiGetDocumentOrShadowRoot = getDocumentOrShadowRoot;

/**
 * @deprecated: use {@link tuiGetElementObscures} instead
 * Returns array of Elements covering edges of given element or null if at least one edge middle point is visible
 *
 * CAUTION: Empty array means element if offscreen i.e. covered by no elements, rather than not covered
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getElementObscurers(element) {
    const { ownerDocument } = element;
    if (!(ownerDocument === null || ownerDocument === void 0 ? void 0 : ownerDocument.defaultView)) {
        return null;
    }
    const { innerWidth, innerHeight } = ownerDocument.defaultView;
    const documentRef = getDocumentOrShadowRoot(element);
    const rect = element.getBoundingClientRect();
    const left = clamp(Math.round(rect.left) + 2, 0, innerWidth);
    const top = clamp(Math.round(rect.top) + 2, 0, innerHeight);
    const right = clamp(Math.round(rect.right) - 2, 0, innerWidth);
    const bottom = clamp(Math.round(rect.bottom) - 2, 0, innerHeight);
    const horizontalMiddle = clamp(Math.round(rect.left + rect.width / 2), 0, innerWidth);
    const verticalMiddle = clamp(Math.round(rect.top + rect.height / 2), 0, innerHeight);
    const elements = [
        documentRef.elementFromPoint(horizontalMiddle, top),
        documentRef.elementFromPoint(horizontalMiddle, bottom),
        documentRef.elementFromPoint(left, verticalMiddle),
        documentRef.elementFromPoint(right, verticalMiddle),
    ];
    const nonNull = elements.filter(isPresent);
    if (!nonNull.length) {
        return nonNull;
    }
    const filtered = nonNull.filter(el => !element.contains(el));
    return filtered.length === 4 ? filtered : null;
}
const tuiGetElementObscures = getElementObscurers;

/**
 * @deprecated: use {@link tuiGetElementOffset} instead
 * Calculates offset for an element relative to it's parent several levels above
 *
 * @param host parent element
 * @param element
 * @return object with offsetTop and offsetLeft number properties
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getElementOffset(host, element) {
    tuiAssert.assert(host.contains(element), `Host must contain element`);
    let { offsetTop, offsetLeft, offsetParent } = element;
    // TODO: iframe warning
    while (offsetParent && offsetParent instanceof HTMLElement && offsetParent !== host) {
        offsetTop += offsetParent.offsetTop;
        offsetLeft += offsetParent.offsetLeft;
        offsetParent = offsetParent.offsetParent;
    }
    return { offsetTop, offsetLeft };
}
const tuiGetElementOffset = getElementOffset;

/**
 * @deprecated: use {@link tuiGetScrollParent} instead
 * Finds the nearest parent with scroll in it
 *
 * @param element initial element
 * @param vertical flag for orientation of scroll
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function getScrollParent(element, vertical = true) {
    if (element === null) {
        return null;
    }
    if (vertical && element.scrollHeight > element.clientHeight) {
        return element;
    }
    if (!vertical && element.scrollWidth > element.clientWidth) {
        return element;
    }
    return getScrollParent(element.parentElement, vertical);
}
const tuiGetScrollParent = getScrollParent;

function tuiIsFormFieldElement(element) {
    const textfield = element;
    return (!!textfield &&
        `value` in textfield &&
        `selectionStart` in textfield &&
        `selectionEnd` in textfield);
}

/**
 * @description:
 * cross browser way to get selected text
 *
 * History:
 * BUG - window.getSelection() fails when text selected in a form field
 * https://bugzilla.mozilla.org/show_bug.cgi?id=85686
 */
function tuiGetSelectedText({ getSelection, document }) {
    var _a;
    return tuiIsFormFieldElement(document.activeElement)
        ? document.activeElement.value.slice(document.activeElement.selectionStart || 0, document.activeElement.selectionEnd || 0)
        : ((_a = getSelection()) === null || _a === void 0 ? void 0 : _a.toString()) || null;
}

/**
 * @deprecated: use {@link tuiIsCurrentTarget} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isCurrentTarget({ target, currentTarget }) {
    return target === currentTarget;
}
const tuiIsCurrentTarget = isCurrentTarget;

// TODO: Refactor to use this check everywhere
function tuiIsElement(node) {
    return !!node && `nodeType` in node && node.nodeType === Node.ELEMENT_NODE;
}

/**
 * @deprecated: use {@link tuiIsInsideIframe} instead
 * Checks if an app is running inside <iframe /> tag
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isInsideIframe(windowRef) {
    return windowRef.parent !== windowRef;
}
const tuiIsInsideIframe = isInsideIframe;

/**
 * @deprecated: use {@link tuiIsNodeIn} instead
 * Checks if node is inside a specific selector
 *
 * @param node
 * @param selector
 * @return true if node is inside a particular selector
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function isNodeIn(node, selector) {
    return node.nodeType === Node.TEXT_NODE
        ? !!node.parentElement && !!getClosestElement(node.parentElement, selector)
        : node.nodeType === Node.ELEMENT_NODE &&
            !!getClosestElement(node, selector);
}
const tuiIsNodeIn = isNodeIn;

function tuiPointToClientRect(x = 0, y = 0) {
    return {
        left: x,
        right: x,
        top: y,
        bottom: y,
        width: 0,
        height: 0,
    };
}

/**
 * Generated bundle index. Do not edit.
 */

export { canScroll, containsOrAfter, getActualTarget, getClipboardDataText, getClosestElement, getDocumentOrShadowRoot, getElementObscurers, getElementOffset, getScrollParent, isCurrentTarget, isInsideIframe, isNodeIn, tuiCanScroll, tuiContainsOrAfter, tuiCustomEvent, tuiGetActualTarget, tuiGetClipboardDataText, tuiGetDocumentOrShadowRoot, tuiGetElementObscures, tuiGetElementOffset, tuiGetScrollParent, tuiGetSelectedText, tuiIsCurrentTarget, tuiIsElement, tuiIsFormFieldElement, tuiIsInsideIframe, tuiIsNodeIn, tuiPointToClientRect };
