// @taiga-ui/kit/esm5/components/data-list-wrapper/data-list-group-wrapper.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, Component, forwardRef, Inject, Input, } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk/decorators';
import { TUI_DATA_LIST_ACCESSOR } from '@taiga-ui/core/tokens';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import { AbstractTuiDataListWrapper } from './data-list-wrapper';
var TuiDataListGroupWrapperComponent = /** @class */ (function (_super) {
    __extends(TuiDataListGroupWrapperComponent, _super);
    function TuiDataListGroupWrapperComponent(itemsHandlers) {
        var _this = _super.call(this, itemsHandlers) || this;
        _this.items = [];
        _this.labels = [];
        return _this;
    }
    TuiDataListGroupWrapperComponent_1 = TuiDataListGroupWrapperComponent;
    var TuiDataListGroupWrapperComponent_1;
    TuiDataListGroupWrapperComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ITEMS_HANDLERS,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiDataListGroupWrapperComponent.prototype, "items", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiDataListGroupWrapperComponent.prototype, "labels", void 0);
    TuiDataListGroupWrapperComponent = TuiDataListGroupWrapperComponent_1 = __decorate([
        Component({
            selector: 'tui-data-list-wrapper[labels]',
            template: "<tui-data-list\n    *ngIf=\"items; else loading\"\n    [emptyContent]=\"emptyContent\"\n>\n    <tui-opt-group\n        *ngFor=\"let group of items; let index = index\"\n        [label]=\"labels[index]\"\n    >\n        <button\n            *ngFor=\"let item of group\"\n            #elementRef=\"elementRef\"\n            tuiElement\n            tuiOption\n            automation-id=\"tui-data-list-wrapper__option\"\n            [size]=\"size\"\n            [value]=\"item\"\n            [disabled]=\"disabledItemHandler(item)\"\n        >\n            <span\n                polymorpheus-outlet\n                class=\"t-content\"\n                [content]=\"itemContent\"\n                [context]=\"getContext(item, elementRef)\"\n            ></span>\n        </button>\n    </tui-opt-group>\n</tui-data-list>\n<ng-template #loading>\n    <tui-loader class=\"t-loader\"></tui-loader>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                {
                    provide: TUI_DATA_LIST_ACCESSOR,
                    useExisting: forwardRef(function () { return TuiDataListGroupWrapperComponent_1; }),
                },
            ],
            styles: [":host{display:block}.t-content{flex:1}.t-loader{margin:.75rem 0}"]
        }),
        __param(0, Inject(TUI_ITEMS_HANDLERS))
    ], TuiDataListGroupWrapperComponent);
    return TuiDataListGroupWrapperComponent;
}(AbstractTuiDataListWrapper));
export { TuiDataListGroupWrapperComponent };
