// @taiga-ui/kit/esm5/components/data-list-wrapper/data-list-wrapper.js
import { __decorate } from "tslib";
import { Directive, forwardRef, Input, ViewChildren, } from '@angular/core';
import { EMPTY_QUERY, isNativeFocused, isPresent, tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiOptionComponent, } from '@taiga-ui/core';
var AbstractTuiDataListWrapper = /** @class */ (function () {
    function AbstractTuiDataListWrapper(itemsHandlers) {
        var _this = this;
        this.itemsHandlers = itemsHandlers;
        this.optionsQuery = EMPTY_QUERY;
        this.disabledItemHandler = this.itemsHandlers.disabledItemHandler;
        this.emptyContent = "";
        this.size = "m";
        this.itemContent = function (_a) {
            var $implicit = _a.$implicit;
            return _this.itemsHandlers.stringify($implicit);
        };
    }
    AbstractTuiDataListWrapper.prototype.getContext = function ($implicit, _a) {
        var nativeElement = _a.nativeElement;
        return { $implicit: $implicit, active: isNativeFocused(nativeElement) };
    };
    AbstractTuiDataListWrapper.prototype.getOptions = function (includeDisabled) {
        if (includeDisabled === void 0) { includeDisabled = false; }
        return this.optionsQuery
            .toArray()
            .filter(function (_a) {
            var disabled = _a.disabled;
            return includeDisabled || !disabled;
        })
            .map(function (_a) {
            var value = _a.value;
            return value;
        })
            .filter(isPresent);
    };
    AbstractTuiDataListWrapper.ctorParameters = function () { return [
        { type: undefined }
    ]; };
    __decorate([
        ViewChildren(forwardRef(function () { return TuiOptionComponent; }))
    ], AbstractTuiDataListWrapper.prototype, "optionsQuery", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiDataListWrapper.prototype, "disabledItemHandler", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiDataListWrapper.prototype, "emptyContent", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiDataListWrapper.prototype, "size", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiDataListWrapper.prototype, "itemContent", void 0);
    AbstractTuiDataListWrapper = __decorate([
        Directive()
    ], AbstractTuiDataListWrapper);
    return AbstractTuiDataListWrapper;
}());
export { AbstractTuiDataListWrapper };
