// @taiga-ui/kit/esm5/components/filter/filter.component.js
import { __decorate, __extends, __param, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, HostBinding, Inject, Input, Optional, Output, Self, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiMultipleControl, ALWAYS_FALSE_HANDLER, isNativeFocusedIn, TUI_DEFAULT_IDENTITY_MATCHER, TUI_DEFAULT_STRINGIFY, tuiDefaultProp, tuiPure, } from '@taiga-ui/cdk';
// @bad TODO: Add active zone to track focus
var TuiFilterComponent = /** @class */ (function (_super) {
    __extends(TuiFilterComponent, _super);
    function TuiFilterComponent(control, changeDetectorRef, elementRef) {
        var _this = _super.call(this, control, changeDetectorRef) || this;
        _this.elementRef = elementRef;
        _this.identityMatcher = TUI_DEFAULT_IDENTITY_MATCHER;
        _this.items = [];
        _this.size = 'm';
        _this.disabledItemHandler = ALWAYS_FALSE_HANDLER;
        _this.toggledItem = new EventEmitter();
        _this.content = function (_a) {
            var $implicit = _a.$implicit;
            return TUI_DEFAULT_STRINGIFY($implicit);
        };
        _this.badgeHandler = function (item) { return Number(item); };
        return _this;
    }
    Object.defineProperty(TuiFilterComponent.prototype, "focused", {
        get: function () {
            return isNativeFocusedIn(this.elementRef.nativeElement);
        },
        enumerable: true,
        configurable: true
    });
    TuiFilterComponent.prototype.getItemContentContext = function ($implicit) {
        return {
            $implicit: $implicit,
        };
    };
    TuiFilterComponent.prototype.onCheckbox = function (value, item) {
        var _this = this;
        this.toggledItem.emit(item);
        this.updateValue(value
            ? __spread(this.value, [item]) : this.value.filter(function (arrItem) { return !_this.identityMatcher(arrItem, item); }));
    };
    TuiFilterComponent.prototype.isCheckboxEnabled = function (item) {
        var _this = this;
        return this.value.some(function (arrItem) { return _this.identityMatcher(arrItem, item); });
    };
    TuiFilterComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "identityMatcher", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "items", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size'),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "size", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "disabledItemHandler", void 0);
    __decorate([
        Output()
    ], TuiFilterComponent.prototype, "toggledItem", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "content", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilterComponent.prototype, "badgeHandler", void 0);
    __decorate([
        tuiPure
    ], TuiFilterComponent.prototype, "getItemContentContext", null);
    TuiFilterComponent = __decorate([
        Component({
            selector: 'tui-filter',
            template: "<ng-container *ngFor=\"let item of items\">\n    <tui-checkbox-block\n        automation-id=\"tui-filter__checkbox\"\n        class=\"t-item\"\n        [disabled]=\"disabledItemHandler(item)\"\n        [hideCheckbox]=\"true\"\n        [ngModel]=\"isCheckboxEnabled(item)\"\n        [size]=\"size\"\n        (ngModelChange)=\"onCheckbox($event, item)\"\n    >\n        <div\n            automation-id=\"tui-filter__content\"\n            class=\"t-content\"\n        >\n            <div\n                polymorpheus-outlet\n                [content]=\"content\"\n                [context]=\"getItemContentContext(item)\"\n            ></div>\n            <tui-badge\n                *ngIf=\"badgeHandler(item) as badgeValue\"\n                automation-id=\"tui-filter__badge\"\n                tuiMode=\"onLight\"\n                class=\"t-badge\"\n                [class.t-badge_disabled]=\"disabledItemHandler(item)\"\n                [size]=\"size\"\n                [value]=\"badgeValue\"\n            ></tui-badge>\n        </div>\n    </tui-checkbox-block>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{display:block;margin-bottom:-.25rem}.t-content{display:flex;align-items:center}.t-item{margin-right:.25rem;margin-bottom:.25rem}.t-item:last-child,:host[data-size='l'] .t-item:last-child{margin-right:0}:host[data-size='l'] .t-item{margin-right:.5rem;margin-bottom:.5rem}.t-badge{margin-left:.25rem}.t-badge_disabled{opacity:var(--tui-disabled-opacity)}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(ChangeDetectorRef)),
        __param(2, Inject(ElementRef))
    ], TuiFilterComponent);
    return TuiFilterComponent;
}(AbstractTuiMultipleControl));
export { TuiFilterComponent };
