// @taiga-ui/addon-doc/esm5/utils/coerce-value.js
/**
 * @deprecated: use {@link tuiCoerceValue} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function coerceValue(value) {
    var prepared = String(value).trim();
    if (isEmptyParamValue(prepared)) {
        return null;
    }
    if (isBooleanParamValue(prepared)) {
        return String(prepared) === "true";
    }
    if (isNumberParamValue(prepared)) {
        return Number(prepared);
    }
    var decodedValue = decodeURIComponent(prepared);
    try {
        return isPossibleArray(decodedValue) || isPossibleObject(decodedValue)
            ? JSON.parse(decodedValue)
            : decodedValue;
    }
    catch (_a) {
        return decodedValue;
    }
}
export var tuiCoerceValue = coerceValue;
function isEmptyParamValue(value) {
    return ["undefined", "null", "NaN", "Infinity"].includes(value);
}
function isBooleanParamValue(value) {
    return value === "true" || value === "false";
}
function isNumberParamValue(value) {
    // TODO: investigate to disallow potentially catastrophic exponential-time regular expressions.
    // eslint-disable-next-line unicorn/no-unsafe-regex
    return /^-?[\d.]+(?:e-?\d+)?$/.test(value);
}
function isPossibleArray(value) {
    return value.startsWith("[") && value.endsWith("]");
}
function isPossibleObject(value) {
    return value.startsWith("{") && value.endsWith("}");
}
