// @taiga-ui/kit/esm2015/components/slider/helpers/slider-readonly.directive.js
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, HostListener, Inject, Input } from '@angular/core';
import { tuiCoerceBooleanProperty, tuiDefaultProp, TuiDestroyService, typedFromEvent, } from '@taiga-ui/cdk';
import { combineLatest, merge, Observable } from 'rxjs';
import { filter, mapTo, takeUntil, tap } from 'rxjs/operators';
const SLIDER_INTERACTION_KEYS = new Set([
    'ArrowLeft',
    'ArrowRight',
    'ArrowUp',
    'ArrowDown',
    'Home',
    'End',
    'PageUp',
    'PageDown',
]);
/**
 * Native <input type='range' readonly> doesn't work.
 * This directive imitates this native behaviour.
 */
// @dynamic
let TuiSliderReadonlyDirective = class TuiSliderReadonlyDirective {
    constructor(elementRef, documentRef, destroy$) {
        this.readonly = true;
        const touchStart$ = typedFromEvent(elementRef.nativeElement, 'touchstart', {
            passive: false,
        });
        const touchMove$ = typedFromEvent(documentRef, 'touchmove', {
            passive: false,
        });
        const touchEnd$ = typedFromEvent(documentRef, 'touchend', {
            passive: true,
        });
        const shouldPreventMove$ = merge(touchStart$.pipe(tap(e => this.preventEvent(e)), mapTo(true)), touchEnd$.pipe(mapTo(false)));
        /**
         * @bad TODO think about another solution.
         * Keep in mind that preventing touch event (on slider) inside `@HostListener('touchstart')` doesn't work for mobile chrome.
         */
        combineLatest([touchMove$, shouldPreventMove$])
            .pipe(filter(([_, shouldPreventMove]) => shouldPreventMove), takeUntil(destroy$))
            .subscribe(([moveEvent]) => this.preventEvent(moveEvent));
    }
    preventEvent(event) {
        if (event.cancelable && tuiCoerceBooleanProperty(this.readonly)) {
            event.preventDefault();
        }
    }
    preventKeyboardInteraction(event) {
        if (SLIDER_INTERACTION_KEYS.has(event.key)) {
            this.preventEvent(event);
        }
    }
};
TuiSliderReadonlyDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiSliderReadonlyDirective.prototype, "readonly", void 0);
__decorate([
    HostListener('mousedown', ['$event'])
], TuiSliderReadonlyDirective.prototype, "preventEvent", null);
__decorate([
    HostListener('keydown', ['$event'])
], TuiSliderReadonlyDirective.prototype, "preventKeyboardInteraction", null);
TuiSliderReadonlyDirective = __decorate([
    Directive({
        selector: 'input[tuiSlider][readonly]',
        providers: [TuiDestroyService],
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(DOCUMENT)),
    __param(2, Inject(TuiDestroyService))
], TuiSliderReadonlyDirective);
export { TuiSliderReadonlyDirective };
