// @taiga-ui/cdk/esm5/directives/autofilled/autofilled.directive.js
import { __decorate, __param } from "tslib";
import { Directive, EventEmitter, HostBinding, HostListener, Inject, Output, } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk/services';
import { TuiAutofilledStyleComponent } from './autofilled-style.component';
var TuiAutofilledDirective = /** @class */ (function () {
    function TuiAutofilledDirective(directiveStyles) {
        this.autofilled = false;
        this.tuiAutofilledChange = new EventEmitter();
        directiveStyles.addComponent(TuiAutofilledStyleComponent);
    }
    TuiAutofilledDirective.prototype.transitionStartHandler = function (_a) {
        var propertyName = _a.propertyName, target = _a.target;
        var _b;
        var matchedAutofill = propertyName.includes('box-shadow') && ((_b = target) === null || _b === void 0 ? void 0 : _b.matches('input'));
        if (matchedAutofill) {
            this.autofilled = !this.autofilled;
            this.tuiAutofilledChange.emit(this.autofilled);
        }
    };
    TuiAutofilledDirective.ctorParameters = function () { return [
        { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
    ]; };
    __decorate([
        HostBinding('class._autofilled')
    ], TuiAutofilledDirective.prototype, "autofilled", void 0);
    __decorate([
        Output()
    ], TuiAutofilledDirective.prototype, "tuiAutofilledChange", void 0);
    __decorate([
        HostListener('transitionstart', ['$event'])
    ], TuiAutofilledDirective.prototype, "transitionStartHandler", null);
    TuiAutofilledDirective = __decorate([
        Directive({
            selector: '[tuiAutofilledChange]',
            host: { class: 'tui-autofill' },
        }),
        __param(0, Inject(TuiDirectiveStylesService))
    ], TuiAutofilledDirective);
    return TuiAutofilledDirective;
}());
export { TuiAutofilledDirective };
