// @taiga-ui/kit/esm5/components/slider/helpers/slider-thumb-label/slider-thumb-label.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { tuiAssert } from '@taiga-ui/cdk';
import { TuiSliderComponent } from '../../slider.component';
var TuiSliderThumbLabelComponent = /** @class */ (function () {
    function TuiSliderThumbLabelComponent() {
    }
    Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "size", {
        get: function () {
            var _a;
            return ((_a = this.slider) === null || _a === void 0 ? void 0 : _a.size) || 'm';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ratio", {
        get: function () {
            var _a;
            return (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.valuePercentage) || 0) / 100;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ghostLeft", {
        get: function () {
            var _a;
            return this.ratio * (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.elementRef.nativeElement.offsetWidth) || 0);
        },
        enumerable: true,
        configurable: true
    });
    TuiSliderThumbLabelComponent.prototype.ngAfterContentInit = function () {
        var _a;
        tuiAssert.assert(Boolean((_a = this.control) === null || _a === void 0 ? void 0 : _a.valueChanges), '\n[tuiSliderThumbLabel] expected <input tuiSlider type="range" /> to use Angular Forms.\n' +
            'Use [(ngModel)] or [formControl] or formControlName for correct work.');
    };
    __decorate([
        ContentChild(TuiSliderComponent)
    ], TuiSliderThumbLabelComponent.prototype, "slider", void 0);
    __decorate([
        ContentChild(NgControl)
    ], TuiSliderThumbLabelComponent.prototype, "control", void 0);
    TuiSliderThumbLabelComponent = __decorate([
        Component({
            selector: '[tuiSliderThumbLabel]',
            template: "<ng-container *ngIf=\"control?.valueChanges | async\"></ng-container>\n\n<div\n    class=\"t-ghost\"\n    [attr.data-size]=\"size\"\n    [style.left.px]=\"ghostLeft\"\n    [style.--tui-slider-thumb-ratio]=\"ratio\"\n>\n    <ng-content></ng-content>\n</div>\n\n<ng-content select=\"input[type=range]\"></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{position:relative}.t-ghost{position:absolute;top:0;bottom:0;margin:auto;border-radius:50%;pointer-events:none}.t-ghost[data-size='s']{width:.5rem;height:.5rem;transform:translateX(calc(var(--tui-slider-thumb-ratio) * -.5rem))}.t-ghost[data-size='m']{width:1rem;height:1rem;transform:translateX(calc(var(--tui-slider-thumb-ratio) * -1rem))}"]
        })
    ], TuiSliderThumbLabelComponent);
    return TuiSliderThumbLabelComponent;
}());
export { TuiSliderThumbLabelComponent };
