// @taiga-ui/addon-doc/esm5/components/copy/copy.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { TuiDestroyService, tuiPure } from '@taiga-ui/cdk';
import { TUI_COPY_TEXTS } from '@taiga-ui/kit';
import { Observable, Subject, timer } from 'rxjs';
import { mapTo, startWith, switchMapTo } from 'rxjs/operators';
var COPIED_TIMEOUT = 1500;
// @dynamic
var TuiDocCopyComponent = /** @class */ (function () {
    function TuiDocCopyComponent(texts$) {
        this.texts$ = texts$;
        this.copy$ = new Subject();
    }
    Object.defineProperty(TuiDocCopyComponent.prototype, "copied$", {
        get: function () {
            return this.copy$.pipe(switchMapTo(timer(COPIED_TIMEOUT).pipe(mapTo(false), startWith(true))));
        },
        enumerable: true,
        configurable: true
    });
    TuiDocCopyComponent.prototype.onClick = function () {
        this.copy$.next();
    };
    TuiDocCopyComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_COPY_TEXTS,] }] }
    ]; };
    __decorate([
        tuiPure
    ], TuiDocCopyComponent.prototype, "copied$", null);
    TuiDocCopyComponent = __decorate([
        Component({
            selector: 'tui-doc-copy',
            template: "<button\n    *ngIf=\"texts$ | async as texts\"\n    tuiButton\n    type=\"button\"\n    appearance=\"\"\n    size=\"s\"\n    class=\"t-copy\"\n    [title]=\"texts[0]\"\n    (click)=\"onClick()\"\n>\n    <span\n        class=\"t-content\"\n        [class.t-content_moved]=\"copied$ | async\"\n    >\n        <span\n            class=\"t-initial\"\n            [attr.data-text]=\"texts[0]\"\n        >\n            <ng-content></ng-content>\n        </span>\n        {{ texts[1] }}\n    </span>\n</button>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [TuiDestroyService],
            styles: [":host{position:relative;display:inline-block;vertical-align:middle}.t-copy{transition-property:background;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;width:100%;color:var(--tui-text-01);background:var(--tui-base-01);overflow:hidden}.t-copy:hover{background:var(--tui-base-02)}.t-initial{white-space:nowrap}.t-initial:empty:before{content:attr(data-text)}.t-content{transition-property:transform;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;flex-direction:column;transform:translateY(1rem);text-align:center;line-height:2rem}.t-content_moved{transform:translateY(-1rem)}"]
        }),
        __param(0, Inject(TUI_COPY_TEXTS))
    ], TuiDocCopyComponent);
    return TuiDocCopyComponent;
}());
export { TuiDocCopyComponent };
