// @taiga-ui/kit/esm5/components/progress/progress-bar/progress-bar.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
var TuiProgressBarComponent = /** @class */ (function () {
    function TuiProgressBarComponent() {
        this.size = 'm';
    }
    __decorate([
        Input(),
        HostBinding('style.--tui-progress-color')
    ], TuiProgressBarComponent.prototype, "color", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size'),
        tuiDefaultProp()
    ], TuiProgressBarComponent.prototype, "size", void 0);
    TuiProgressBarComponent = __decorate([
        Component({
            selector: 'progress[tuiProgressBar]',
            template: '',
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;border:none;color:var(--tui-progress-color,currentColor);background-color:var(--tui-base-03);display:block;width:100%;color:var(--tui-primary);border-radius:var(--tui-radius-xs);overflow:hidden}:host::-webkit-progress-value{-webkit-transition:width var(--tui-duration) linear;transition:width var(--tui-duration) linear;background:var(--tui-progress-color,currentColor)}:host::-moz-progress-bar{background:var(--tui-progress-color,currentColor)}:host::-webkit-progress-bar{background-color:var(--tui-base-03)}:host[data-size='m']{height:1.5rem}:host[data-size='s']{border-radius:var(--tui-radius-xs);height:.5rem}:host[data-size='s']::-webkit-progress-value{border-radius:var(--tui-radius-xs)}:host[data-size='s']::-moz-progress-bar{border-radius:var(--tui-radius-xs)}:host-context(label[tuiProgressLabel]):not(:first-child){position:absolute;top:0;left:0;width:100%;height:100%;background-color:transparent}:host-context(label[tuiProgressLabel]):not(:first-child)::-webkit-progress-bar{background-color:transparent}"]
        })
    ], TuiProgressBarComponent);
    return TuiProgressBarComponent;
}());
export { TuiProgressBarComponent };
