// @taiga-ui/kit/esm5/components/progress/progress-bar/progress-color-segments.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { USER_AGENT } from '@ng-web-apis/common';
import { CHROMIUM_EDGE_START_VERSION, isEdgeOlderThan, isIE, TuiDestroyService, tuiPure, TuiResizeService, } from '@taiga-ui/cdk';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { calculateColorSegments } from '../utils/calculate-color-segments';
var TuiProgressColorSegmentsDirective = /** @class */ (function () {
    function TuiProgressColorSegmentsDirective(elementRef, resize$, userAgent) {
        this.elementRef = elementRef;
        this.resize$ = resize$;
        this.userAgent = userAgent;
        this.isOldBrowsers = isEdgeOlderThan(CHROMIUM_EDGE_START_VERSION, this.userAgent) ||
            isIE(this.userAgent);
        this.colors = [];
    }
    Object.defineProperty(TuiProgressColorSegmentsDirective.prototype, "calcSegments$", {
        get: function () {
            var _this = this;
            return this.resize$.pipe(map(function () {
                return _this.isOldBrowsers
                    ? _this.colors[0]
                    : calculateColorSegments(_this.colors, _this.elementRef.nativeElement.offsetWidth);
            }));
        },
        enumerable: true,
        configurable: true
    });
    TuiProgressColorSegmentsDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TuiResizeService,] }] },
        { type: String, decorators: [{ type: Inject, args: [USER_AGENT,] }] }
    ]; };
    __decorate([
        Input('tuiProgressColorSegments')
    ], TuiProgressColorSegmentsDirective.prototype, "colors", void 0);
    __decorate([
        tuiPure
    ], TuiProgressColorSegmentsDirective.prototype, "calcSegments$", null);
    TuiProgressColorSegmentsDirective = __decorate([
        Directive({
            selector: 'progress[tuiProgressBar][tuiProgressColorSegments]',
            host: {
                '[$.style.--tui-progress-color]': 'calcSegments$',
                '($.style.--tui-progress-color)': '0',
            },
            providers: [TuiDestroyService, TuiResizeService],
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(TuiResizeService)),
        __param(2, Inject(USER_AGENT))
    ], TuiProgressColorSegmentsDirective);
    return TuiProgressColorSegmentsDirective;
}());
export { TuiProgressColorSegmentsDirective };
