// @taiga-ui/core/esm5/components/primitive-checkbox/checkbox-options.js
import { __assign } from "tslib";
import { InjectionToken } from '@angular/core';
/** Default values for the checkbox options. */
export var TUI_CHECKBOX_DEFAULT_OPTIONS = {
    size: "m",
    appearances: {
        unchecked: "outline" /* Outline */,
        checked: "primary" /* Primary */,
        indeterminate: "primary" /* Primary */,
    },
    icons: {
        checked: function (_a) {
            var $implicit = _a.$implicit;
            return $implicit === "m" ? "tuiIconCheck" : "tuiIconCheckLarge";
        },
        indeterminate: function (_a) {
            var $implicit = _a.$implicit;
            return $implicit === "m" ? "tuiIconMinus" : "tuiIconMinusLarge";
        },
    },
};
export var TUI_CHECKBOX_OPTIONS = new InjectionToken("Default parameters for checkbox component", {
    factory: function () { return TUI_CHECKBOX_DEFAULT_OPTIONS; },
});
export var tuiCheckboxOptionsProvider = function (options) { return ({
    provide: TUI_CHECKBOX_OPTIONS,
    useValue: __assign(__assign({}, TUI_CHECKBOX_DEFAULT_OPTIONS), options),
}); };
