// @taiga-ui/kit/esm5/components/checkbox/checkbox.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, forwardRef, HostBinding, Inject, Input, Optional, Self, ViewChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiNullableControl, isNativeFocused, TUI_FOCUSABLE_ITEM_ACCESSOR, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TUI_CHECKBOX_OPTIONS } from '@taiga-ui/core';
var TuiCheckboxComponent = /** @class */ (function (_super) {
    __extends(TuiCheckboxComponent, _super);
    function TuiCheckboxComponent(control, options, changeDetectorRef) {
        var _this = _super.call(this, control, changeDetectorRef) || this;
        _this.options = options;
        _this.size = _this.options.size;
        return _this;
    }
    TuiCheckboxComponent_1 = TuiCheckboxComponent;
    Object.defineProperty(TuiCheckboxComponent.prototype, "nativeFocusableElement", {
        get: function () {
            return !this.focusableElement || this.computedDisabled
                ? null
                : this.focusableElement.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiCheckboxComponent.prototype, "focused", {
        get: function () {
            return isNativeFocused(this.nativeFocusableElement);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiCheckboxComponent.prototype, "computedFocusable", {
        get: function () {
            return this.interactive && this.focusable;
        },
        enumerable: true,
        configurable: true
    });
    TuiCheckboxComponent.prototype.onChecked = function (checked) {
        this.updateValue(checked);
    };
    TuiCheckboxComponent.prototype.onFocused = function (focused) {
        this.updateFocused(focused);
    };
    TuiCheckboxComponent.prototype.onHovered = function (hovered) {
        this.updateHovered(hovered);
    };
    TuiCheckboxComponent.prototype.onPressed = function (pressed) {
        this.updatePressed(pressed);
    };
    TuiCheckboxComponent.prototype.onFocusVisible = function (focusVisible) {
        this.updateFocusVisible(focusVisible);
    };
    var TuiCheckboxComponent_1;
    TuiCheckboxComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_CHECKBOX_OPTIONS,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] }
    ]; };
    __decorate([
        ViewChild('focusableElement')
    ], TuiCheckboxComponent.prototype, "focusableElement", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size'),
        tuiDefaultProp()
    ], TuiCheckboxComponent.prototype, "size", void 0);
    TuiCheckboxComponent = TuiCheckboxComponent_1 = __decorate([
        Component({
            selector: 'tui-checkbox',
            template: "<tui-primitive-checkbox\n    [size]=\"size\"\n    [disabled]=\"disabled\"\n    [focused]=\"computedFocusVisible\"\n    [hovered]=\"computedHovered\"\n    [pressed]=\"computedPressed\"\n    [invalid]=\"computedInvalid\"\n    [value]=\"value\"\n>\n    <input\n        #focusableElement\n        type=\"checkbox\"\n        automation-id=\"tui-checkbox__native\"\n        class=\"t-native\"\n        [disabled]=\"disabled\"\n        [id]=\"id\"\n        [tuiChecked]=\"value\"\n        [tuiFocusable]=\"computedFocusable\"\n        (tuiFocusedChange)=\"onFocused($event)\"\n        (tuiHoveredChange)=\"onHovered($event)\"\n        (tuiPressedChange)=\"onPressed($event)\"\n        (tuiCheckedChange)=\"onChecked($event)\"\n        (tuiFocusVisibleChange)=\"onFocusVisible($event)\"\n    />\n</tui-primitive-checkbox>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                {
                    provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
                    useExisting: forwardRef(function () { return TuiCheckboxComponent_1; }),
                },
            ],
            styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);position:relative;display:block;flex-shrink:0}:host[data-size='m']{width:1rem;height:1rem}:host[data-size='l']{width:1.5rem;height:1.5rem}:host._disabled,:host._readonly{pointer-events:none}.t-native{padding:0;margin:0;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;opacity:0;cursor:pointer}.t-native:-webkit-autofill,.t-native:-webkit-autofill:focus,.t-native:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(TUI_CHECKBOX_OPTIONS)),
        __param(2, Inject(ChangeDetectorRef))
    ], TuiCheckboxComponent);
    return TuiCheckboxComponent;
}(AbstractTuiNullableControl));
export { TuiCheckboxComponent };
