// @taiga-ui/kit/esm2015/pipes/filter-by-input/filter-by-input.base.js
import { __decorate } from "tslib";
/* eslint-disable @typescript-eslint/naming-convention */
import { isPresent, tuiPure, } from '@taiga-ui/cdk';
import { isFlat } from '@taiga-ui/kit/utils';
/**
 * @deprecated
 * TODO: 3.0 replace with {@link AbstractTuiFilterByInput}
 */
export class TuiFilterByInputBase {
    get query() {
        return this.accessor.nativeFocusableElement
            ? this.accessor.nativeFocusableElement.value || ``
            : ``;
    }
    filter(items, matcher, stringify, query) {
        if (!items) {
            return null;
        }
        return isFlat(items)
            ? this.filterFlat(items, matcher, stringify, query)
            : this.filter2d(items, matcher, stringify, query);
    }
    filterFlat(items, matcher, stringify, query) {
        const match = this.getMatch(items, stringify, query);
        return isPresent(match)
            ? items
            : items.filter(item => matcher(item, query, stringify));
    }
    filter2d(items, matcher, stringify, query) {
        const match = items.find(item => isPresent(this.getMatch(item, stringify, query)));
        return isPresent(match)
            ? items
            : items.map(inner => this.filterFlat(inner, matcher, stringify, query));
    }
    getMatch(items, stringify, query) {
        return items.find(item => stringify(item).toLocaleLowerCase() === query.toLocaleLowerCase());
    }
}
__decorate([
    tuiPure
], TuiFilterByInputBase.prototype, "filter", null);
export class AbstractTuiFilterByInput extends TuiFilterByInputBase {
}
