// @taiga-ui/kit/esm5/components/text-area/text-area.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiTextAreaDirective = /** @class */ (function (_super) {
    __extends(TuiTextAreaDirective, _super);
    function TuiTextAreaDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiTextAreaDirective_1 = TuiTextAreaDirective;
    TuiTextAreaDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiTextAreaDirective_1;
    TuiTextAreaDirective = TuiTextAreaDirective_1 = __decorate([
        Directive({
            selector: 'tui-text-area',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiTextAreaDirective_1; }),
                },
            ],
        })
    ], TuiTextAreaDirective);
    return TuiTextAreaDirective;
}(TuiAbstractTextfieldHost));
export { TuiTextAreaDirective };
