// @taiga-ui/cdk/esm5/directives/auto-focus/autofocus.directive.js
import { __decorate, __param } from "tslib";
import { Directive, Inject, Input } from '@angular/core';
import { TUI_AUTOFOCUS_HANDLER, TUI_AUTOFOCUS_PROVIDERS, } from './autofocus.options';
// TODO: 3.0 change input name to tuiAutoFocus and handle empty string
var TuiAutoFocusDirective = /** @class */ (function () {
    function TuiAutoFocusDirective(handler) {
        this.handler = handler;
        this.autoFocus = true;
    }
    TuiAutoFocusDirective.prototype.ngAfterViewInit = function () {
        if (this.autoFocus) {
            this.focus();
        }
    };
    TuiAutoFocusDirective.prototype.focus = function () {
        var _this = this;
        void Promise.resolve().then(function () { return _this.handler.setFocus(); });
    };
    TuiAutoFocusDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_AUTOFOCUS_HANDLER,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiAutoFocusDirective.prototype, "autoFocus", void 0);
    TuiAutoFocusDirective = __decorate([
        Directive({
            selector: '[tuiAutoFocus]',
            providers: TUI_AUTOFOCUS_PROVIDERS,
        }),
        __param(0, Inject(TUI_AUTOFOCUS_HANDLER))
    ], TuiAutoFocusDirective);
    return TuiAutoFocusDirective;
}());
export { TuiAutoFocusDirective };
