// @taiga-ui/core/esm5/directives/mask-accessor/mask-accessor.directive.js
import { __decorate } from "tslib";
import { Directive, Optional } from '@angular/core';
import { DefaultValueAccessor, NG_VALUE_ACCESSOR, } from '@angular/forms';
import { TUI_VALUE_ACCESSOR } from '@taiga-ui/core/tokens';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function fallbackAccessorFactory(accessor, fallback) {
    return accessor ? accessor[0] : fallback;
}
var TuiMaskAccessorDirective = /** @class */ (function () {
    function TuiMaskAccessorDirective() {
    }
    TuiMaskAccessorDirective = __decorate([
        Directive({
            selector: 'input[tuiMaskAccessor]',
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    deps: [[new Optional(), TUI_VALUE_ACCESSOR], DefaultValueAccessor],
                    useFactory: fallbackAccessorFactory,
                    multi: true,
                },
            ],
        })
    ], TuiMaskAccessorDirective);
    return TuiMaskAccessorDirective;
}());
export { TuiMaskAccessorDirective };
