// @taiga-ui/kit/esm5/components/tree/components/tree-item/tree-item.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, DoCheck, ElementRef, forwardRef, HostBinding, Inject, QueryList, } from '@angular/core';
import { EMPTY_QUERY } from '@taiga-ui/cdk';
import { Subject } from 'rxjs';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
import { TUI_TREE_CONTENT, TUI_TREE_CONTROLLER, TUI_TREE_LEVEL, TUI_TREE_NODE, } from '../../misc/tree.tokens';
import { TUI_TREE_ITEM_PROVIDERS } from './tree-item.providers';
var TuiTreeItemComponent = /** @class */ (function () {
    function TuiTreeItemComponent(elementRef, controller, level, content) {
        var _this = this;
        this.elementRef = elementRef;
        this.controller = controller;
        this.level = level;
        this.content = content;
        this.nested = EMPTY_QUERY;
        this.change$ = new Subject();
        this.expanded$ = this.change$.pipe(startWith(null), map(function () { return _this.isExpanded; }), distinctUntilChanged());
        this.attached$ = this.change$.pipe(map(function () { return _this.elementRef.nativeElement.isConnected; }), distinctUntilChanged());
    }
    Object.defineProperty(TuiTreeItemComponent.prototype, "isExpandable", {
        get: function () {
            return !!this.nested.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTreeItemComponent.prototype, "isExpanded", {
        get: function () {
            return this.controller.isExpanded(this);
        },
        enumerable: true,
        configurable: true
    });
    TuiTreeItemComponent.prototype.ngDoCheck = function () {
        this.change$.next();
    };
    TuiTreeItemComponent.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef(function () { return TUI_TREE_CONTROLLER; }),] }] },
        { type: Number, decorators: [{ type: Inject, args: [forwardRef(function () { return TUI_TREE_LEVEL; }),] }] },
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef(function () { return TUI_TREE_CONTENT; }),] }] }
    ]; };
    __decorate([
        ContentChildren(TUI_TREE_NODE)
    ], TuiTreeItemComponent.prototype, "nested", void 0);
    __decorate([
        HostBinding('class._expandable')
    ], TuiTreeItemComponent.prototype, "isExpandable", null);
    TuiTreeItemComponent = __decorate([
        Component({
            selector: 'tui-tree-item',
            template: "<ng-template #template>\n    <ng-content></ng-content>\n</ng-template>\n<div\n    polymorpheus-outlet\n    [content]=\"content\"\n    [context]=\"{$implicit: this, template: template}\"\n></div>\n<tui-expand\n    *ngIf=\"isExpandable\"\n    role=\"group\"\n    class=\"t-children\"\n    [expanded]=\"expanded$ | async\"\n>\n    <div>\n        <ng-content select=\"tui-tree-item\"></ng-content>\n        <ng-content select=\"tui-tree\"></ng-content>\n    </div>\n</tui-expand>\n<ng-container *ngIf=\"attached$ | async\"></ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: TUI_TREE_ITEM_PROVIDERS,
            host: {
                role: 'treeitem',
            },
            styles: [":host{display:block}.t-children{position:relative;margin-left:var(--tui-tree-item-indent,1.5rem)}"]
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(forwardRef(function () { return TUI_TREE_CONTROLLER; }))),
        __param(2, Inject(forwardRef(function () { return TUI_TREE_LEVEL; }))),
        __param(3, Inject(forwardRef(function () { return TUI_TREE_CONTENT; })))
    ], TuiTreeItemComponent);
    return TuiTreeItemComponent;
}());
export { TuiTreeItemComponent };
