// @taiga-ui/kit/esm5/tokens/i18n.js
import { InjectionToken } from '@angular/core';
import { extractI18n } from '@taiga-ui/i18n';
export var TUI_CANCEL_WORD = new InjectionToken("i18n 'cancel' word", {
    factory: extractI18n("cancel"),
});
export var TUI_DONE_WORD = new InjectionToken("i18n 'done' word", {
    factory: extractI18n("done"),
});
export var TUI_MORE_WORD = new InjectionToken("i18n 'more' word", {
    factory: extractI18n("more"),
});
export var TUI_HIDE_TEXT = new InjectionToken("i18n 'hide' word", {
    factory: extractI18n("hide"),
});
export var TUI_SHOW_ALL_TEXT = new InjectionToken("i18n 'show all' word", {
    factory: extractI18n("showAll"),
});
export var TUI_OTHER_DATE_TEXT = new InjectionToken("i18n 'Other date' text", {
    factory: extractI18n("otherDate"),
});
export var TUI_CHOOSE_DAY_OR_RANGE_TEXTS = new InjectionToken("choose day or range i18n texts", {
    factory: extractI18n("mobileCalendarTexts"),
});
export var TUI_FROM_TO_TEXTS = new InjectionToken("from and to i18n texts", {
    factory: extractI18n("range"),
});
export var TUI_PLUS_MINUS_TEXTS = new InjectionToken("plus and minus i18n texts", {
    factory: extractI18n("countTexts"),
});
export var TUI_TIME_TEXTS = new InjectionToken("time i18n texts", {
    factory: extractI18n("time"),
});
export var TUI_DATE_TEXTS = new InjectionToken("date format i18n texts", {
    factory: extractI18n("dateTexts"),
});
export var TUI_DIGITAL_INFORMATION_UNITS = new InjectionToken("short bytes, kilobytes and megabytes i18n texts", {
    factory: extractI18n("digitalInformationUnits"),
});
export var TUI_COPY_TEXTS = new InjectionToken("copy i18n texts", {
    factory: extractI18n("copyTexts"),
});
export var TUI_PASSWORD_TEXTS = new InjectionToken("password i18n texts", {
    factory: extractI18n("passwordTexts"),
});
export var TUI_CALENDAR_MONTHS = new InjectionToken("short calendar months i18n", {
    factory: extractI18n("shortCalendarMonths"),
});
export var TUI_FILE_TEXTS = new InjectionToken("file i18n texts", {
    factory: extractI18n("fileTexts"),
});
export var TUI_PAGINATION_TEXTS = new InjectionToken("pagination i18n texts", {
    factory: extractI18n("pagination"),
});
export var TUI_INPUT_FILE_TEXTS = new InjectionToken("tui-input-file i18n texts", {
    factory: extractI18n("inputFileTexts"),
});
export var TUI_COUNTRIES = new InjectionToken("Localized countries names", {
    factory: extractI18n("countries"),
});
