// @taiga-ui/addon-doc/esm2015/taiga-ui-addon-doc.js
/**
 * Generated bundle index. Do not edit.
 */
export * from './public-api';
export { TuiDocExampleCapitalizePipe as ɵd } from './components/example/example-capitalize.pipe';
export { TuiDocExampleGetTabsPipe as ɵc } from './components/example/example-get-tabs.pipe';
export { NAVIGATION_ITEMS as ɵh, NAVIGATION_LABELS as ɵg, NAVIGATION_PROVIDERS as ɵi, NAVIGATION_TITLE as ɵf, itemsProviderFactory as ɵl, labelsProviderFactory as ɵk, titleProviderFactory as ɵj } from './components/navigation/navigation.providers';
export { PAGE_PROVIDERS as ɵo, PAGE_SEE_ALSO as ɵn, seeAlsoProviderFactory as ɵp } from './components/page/page.providers';
export { TuiDocHeaderComponent as ɵm } from './internal/header/header.component';
export { TuiDocHeaderModule as ɵe } from './internal/header/header.module';
export { TuiInputOpacityDirective as ɵb } from './internal/input-opacity/input-opacity.directive';
export { TuiInputOpacityModule as ɵa } from './internal/input-opacity/input-opacity.module';
export { TuiDocSeeAlsoComponent as ɵr } from './internal/see-also/see-also.component';
export { TuiDocSeeAlsoModule as ɵq } from './internal/see-also/see-also.module';
export { TuiDocSourceCodeComponent as ɵt } from './internal/source-code/source-code.component';
export { TuiDocSourceCodeModule as ɵs } from './internal/source-code/source-code.module';
