// @taiga-ui/kit/esm5/components/input/input.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputDirective = /** @class */ (function (_super) {
    __extends(TuiInputDirective, _super);
    function TuiInputDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputDirective_1 = TuiInputDirective;
    TuiInputDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiInputDirective_1;
    TuiInputDirective = TuiInputDirective_1 = __decorate([
        Directive({
            selector: 'tui-input',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputDirective_1; }),
                },
            ],
        })
    ], TuiInputDirective);
    return TuiInputDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputDirective };
