// @taiga-ui/cdk/esm5/services/scroll.service.js
import { __decorate, __param, __read } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { ANIMATION_FRAME, PERFORMANCE } from '@ng-web-apis/common';
import { tuiAssert } from '@taiga-ui/cdk/classes';
import { tuiClamp } from '@taiga-ui/cdk/utils/math';
import { tuiEaseInOutQuad } from '@taiga-ui/cdk/utils/miscellaneous';
import { defer, Observable, of, timer } from 'rxjs';
import { map, switchMap, takeUntil, tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@ng-web-apis/common";
var SCROLL_TIME = 300;
// @dynamic
var TuiScrollService = /** @class */ (function () {
    function TuiScrollService(performanceRef, animationFrame$) {
        this.performanceRef = performanceRef;
        this.animationFrame$ = animationFrame$;
    }
    TuiScrollService.prototype.scroll$ = function (elementOrWindow, scrollTop, scrollLeft, duration) {
        var _this = this;
        if (scrollLeft === void 0) { scrollLeft = getX(elementOrWindow); }
        if (duration === void 0) { duration = SCROLL_TIME; }
        tuiAssert.assert(duration >= 0, "Duration cannot be negative");
        tuiAssert.assert(scrollTop >= 0, "scrollTop cannot be negative");
        tuiAssert.assert(scrollLeft >= 0, "scrollLeft cannot be negative");
        var initialTop = getY(elementOrWindow);
        var initialLeft = getX(elementOrWindow);
        var deltaTop = scrollTop - initialTop;
        var deltaLeft = scrollLeft - initialLeft;
        var observable = !duration
            ? of([scrollTop, scrollLeft])
            : defer(function () { return of(_this.performanceRef.now()); }).pipe(switchMap(function (start) { return _this.animationFrame$.pipe(map(function (now) { return now - start; })); }), map(function (elapsed) { return tuiEaseInOutQuad(tuiClamp(elapsed / duration, 0, 1)); }), map(function (percent) {
                return [
                    initialTop + deltaTop * percent,
                    initialLeft + deltaLeft * percent,
                ];
            }), takeUntil(timer(duration)));
        return observable.pipe(tap(function (_a) {
            var _b = __read(_a, 2), scrollTop = _b[0], scrollLeft = _b[1];
            var _c;
            (_c = elementOrWindow.scrollTo) === null || _c === void 0 ? void 0 : _c.call(elementOrWindow, scrollLeft, scrollTop);
        }));
    };
    TuiScrollService.ctorParameters = function () { return [
        { type: Performance, decorators: [{ type: Inject, args: [PERFORMANCE,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [ANIMATION_FRAME,] }] }
    ]; };
    TuiScrollService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiScrollService_Factory() { return new TuiScrollService(i0.ɵɵinject(i1.PERFORMANCE), i0.ɵɵinject(i1.ANIMATION_FRAME)); }, token: TuiScrollService, providedIn: "root" });
    TuiScrollService = __decorate([
        Injectable({
            providedIn: "root",
        }),
        __param(0, Inject(PERFORMANCE)),
        __param(1, Inject(ANIMATION_FRAME))
    ], TuiScrollService);
    return TuiScrollService;
}());
export { TuiScrollService };
function getX(elementOrWindow) {
    return "scrollX" in elementOrWindow
        ? elementOrWindow.scrollX
        : elementOrWindow.scrollLeft;
}
function getY(elementOrWindow) {
    return "scrollY" in elementOrWindow
        ? elementOrWindow.scrollY
        : elementOrWindow.scrollTop;
}
