// @taiga-ui/cdk/esm5/services/zoom.service.js
import { __decorate, __extends, __param } from "tslib";
import { ElementRef, Inject, Injectable } from '@angular/core';
import { preventDefault, typedFromEvent } from '@taiga-ui/cdk/observables';
import { TUI_ZOOM_OPTIONS } from '@taiga-ui/cdk/tokens';
import { distanceBetweenTouches } from '@taiga-ui/cdk/utils';
import { merge, Observable } from 'rxjs';
import { filter, map, scan, switchMap, takeUntil } from 'rxjs/operators';
var TOUCH_SENSITIVITY = 0.01;
// @dynamic
var TuiZoomService = /** @class */ (function (_super) {
    __extends(TuiZoomService, _super);
    function TuiZoomService(_a, _b) {
        var nativeElement = _a.nativeElement;
        var wheelSensitivity = _b.wheelSensitivity;
        return _super.call(this, function (subscriber) {
            merge(typedFromEvent(nativeElement, "touchstart", { passive: true }).pipe(filter(function (_a) {
                var touches = _a.touches;
                return touches.length > 1;
            }), switchMap(function (startEvent) {
                return typedFromEvent(nativeElement, "touchmove", {
                    passive: true,
                }).pipe(preventDefault(), scan(function (prev, event) {
                    var distance = distanceBetweenTouches(event);
                    return {
                        event: event,
                        distance: distance,
                        delta: (distance - prev.distance) *
                            TOUCH_SENSITIVITY,
                    };
                }, {
                    event: startEvent,
                    distance: distanceBetweenTouches(startEvent),
                    delta: 0,
                }), map(function (_a) {
                    var event = _a.event, delta = _a.delta;
                    var clientX = (event.touches[0].clientX +
                        event.touches[1].clientX) /
                        2;
                    var clientY = (event.touches[0].clientY +
                        event.touches[1].clientY) /
                        2;
                    return { clientX: clientX, clientY: clientY, delta: delta, event: event };
                }), takeUntil(typedFromEvent(nativeElement, "touchend")));
            })), typedFromEvent(nativeElement, "wheel", { passive: false }).pipe(preventDefault(), map(function (wheel) { return ({
                clientX: wheel.clientX,
                clientY: wheel.clientY,
                delta: -wheel.deltaY * wheelSensitivity,
                event: wheel,
            }); }))).subscribe(subscriber);
        }) || this;
    }
    TuiZoomService.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ZOOM_OPTIONS,] }] }
    ]; };
    TuiZoomService = __decorate([
        Injectable(),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(TUI_ZOOM_OPTIONS))
    ], TuiZoomService);
    return TuiZoomService;
}(Observable));
export { TuiZoomService };
