// @taiga-ui/kit/esm5/components/stepper/step/step.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, HostListener, Inject, Input, } from '@angular/core';
import { tuiDefaultProp, TuiDestroyService, TuiFocusVisibleService } from '@taiga-ui/cdk';
import { TuiRouterLinkActiveService } from '@taiga-ui/core';
import { identity, Observable } from 'rxjs';
import { filter } from 'rxjs/operators';
import { TuiStepperComponent } from '../stepper.component';
var TuiStepComponent = /** @class */ (function () {
    function TuiStepComponent(focusVisible$, routerLinkActive$, stepper, elementRef) {
        var _this = this;
        this.stepper = stepper;
        this.elementRef = elementRef;
        this.state = 'normal';
        this.icon = '';
        this.focusVisible = false;
        routerLinkActive$.pipe(filter(identity)).subscribe(function () {
            _this.activate();
        });
        focusVisible$.subscribe(function (visible) {
            _this.focusVisible = visible;
        });
    }
    Object.defineProperty(TuiStepComponent.prototype, "isActive", {
        get: function () {
            return this.stepper.isActive(this.index);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiStepComponent.prototype, "isVertical", {
        get: function () {
            return this.stepper.orientation === 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiStepComponent.prototype, "tabIndex", {
        get: function () {
            return this.isActive ? 0 : -1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiStepComponent.prototype, "index", {
        get: function () {
            return this.stepper.indexOf(this.elementRef.nativeElement);
        },
        enumerable: true,
        configurable: true
    });
    TuiStepComponent.prototype.activate = function () {
        this.stepper.activate(this.index);
    };
    TuiStepComponent.ctorParameters = function () { return [
        { type: TuiFocusVisibleService, decorators: [{ type: Inject, args: [TuiFocusVisibleService,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TuiRouterLinkActiveService,] }] },
        { type: TuiStepperComponent, decorators: [{ type: Inject, args: [TuiStepperComponent,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    __decorate([
        Input(),
        HostBinding('attr.data-state'),
        tuiDefaultProp()
    ], TuiStepComponent.prototype, "state", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiStepComponent.prototype, "icon", void 0);
    __decorate([
        HostBinding('class._focus-visible')
    ], TuiStepComponent.prototype, "focusVisible", void 0);
    __decorate([
        HostBinding('class._active')
    ], TuiStepComponent.prototype, "isActive", null);
    __decorate([
        HostBinding('class._vertical')
    ], TuiStepComponent.prototype, "isVertical", null);
    __decorate([
        HostBinding('tabIndex')
    ], TuiStepComponent.prototype, "tabIndex", null);
    __decorate([
        HostListener('click')
    ], TuiStepComponent.prototype, "activate", null);
    TuiStepComponent = __decorate([
        Component({
            selector: 'button[tuiStep], a[tuiStep]:not([routerLink]), a[tuiStep][routerLink][routerLinkActive]',
            template: "<tui-svg\n    *ngIf=\"icon\"\n    class=\"t-marker t-marker_custom\"\n    [src]=\"icon\"\n></tui-svg>\n<tui-svg\n    src=\"tuiIconWarningLarge\"\n    class=\"t-marker t-marker_error\"\n></tui-svg>\n<tui-svg\n    src=\"tuiIconCheckLarge\"\n    class=\"t-marker t-marker_pass\"\n></tui-svg>\n<div class=\"t-marker t-marker_index\"></div>\n<ng-content></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [TuiDestroyService, TuiRouterLinkActiveService, TuiFocusVisibleService],
            host: {
                type: 'button',
            },
            styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:0 0;font-size:inherit;line-height:inherit;position:relative;display:flex;align-items:center;flex-shrink:0;color:var(--tui-link);margin-right:2rem;outline:0;cursor:pointer;text-decoration:none;counter-increment:steps}:host:disabled{pointer-events:none;color:var(--tui-base-07)}:host:hover{color:var(--tui-link-hover)}:host:not(:last-of-type)._vertical{margin-bottom:1.25rem}:host._active,:host._active:hover{color:var(--tui-text-01);cursor:default}:host._focus-visible:before{content:'';position:absolute;left:2.75rem;right:0;top:50%;height:1.5rem;margin-top:-.75rem;background:var(--tui-selection)}.t-marker{transition-property:background;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:2rem;height:2rem;border-radius:100%;margin-right:.75rem;flex-shrink:0;align-items:center;justify-content:center;background:var(--tui-secondary);color:var(--tui-link)}:host:disabled .t-marker{background:var(--tui-base-03);color:var(--tui-base-07)}.t-marker_index:before{content:counter(steps)}:host:hover .t-marker_index{color:var(--tui-link-hover);background:var(--tui-secondary-hover)}:host._active .t-marker_index{color:var(--tui-primary-text);background:var(--tui-primary)}:host:not(._active) .t-marker_custom~.t-marker_index,:host:not([data-state=normal]):not(._active) .t-marker_index{display:none}.t-marker_error{background:var(--tui-error-bg);color:var(--tui-error-fill)}:host:hover .t-marker_error{background:var(--tui-error-bg-hover);color:var(--tui-error-fill)}:host._active .t-marker_error,:host:not([data-state=error]) .t-marker_error{display:none}:host._active .t-marker_pass,:host:not([data-state=pass]) .t-marker_pass{display:none}:host._active .t-marker_custom,:host:not([data-state=normal]) .t-marker_custom{display:none}:host:hover .t-marker_custom{color:var(--tui-link-hover);background:var(--tui-secondary-hover)}"]
        }),
        __param(0, Inject(TuiFocusVisibleService)),
        __param(1, Inject(TuiRouterLinkActiveService)),
        __param(2, Inject(TuiStepperComponent)),
        __param(3, Inject(ElementRef))
    ], TuiStepComponent);
    return TuiStepComponent;
}());
export { TuiStepComponent };
