// @taiga-ui/kit/esm2015/abstract/input-slider.js
import { __decorate } from "tslib";
import { Directive, HostBinding, Input } from '@angular/core';
import { AbstractTuiControl, tuiDefaultProp, tuiPure } from '@taiga-ui/cdk';
import { tuiPluralizeToICU, } from '@taiga-ui/core';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function quantumAssertion(quantum) {
    return quantum > 0;
}
/**
 * @internal
 * @deprecated TODO: 3.0 delete me after `InputSlider` and `InputRange` stop using it
 */
let AbstractTuiInputSlider = class AbstractTuiInputSlider extends AbstractTuiControl {
    constructor() {
        super(...arguments);
        this.min = 0;
        this.max = Infinity;
        /**
         * @deprecated This input-property will be deleted in next major update.
         * Use `valueContent` for `InputSlider`.
         * Use `leftValueContent` for `InputRange`.
         * ___
         * TODO remove in v3.0.
         * Dont forget to delete backward-compatibility helpers inside `InputSlider` and `InputRange`:
         *** {@link legacyMinMaxLabel}
         */
        this.minLabel = ``;
        /**
         * @deprecated This input-property will be deleted in next major update.
         * Use `valueContent` for `InputSlider`.
         * Use `rightValueContent` for `InputRange`.
         * ___
         * TODO remove in v3.0.
         * Dont forget to delete backward-compatibility helpers inside `InputSlider` and `InputRange`:
         *** {@link legacyMinMaxLabel}
         */
        this.maxLabel = ``;
        this.segments = 0;
        this.steps = 0;
        this.quantum = 1;
        this.keySteps = null;
        /**
         * @deprecated use `tuiTextfieldSize` instead
         * TODO delete in v3.0
         */
        this.size = `l`;
        /**
         * @deprecated replace by input-property {@link pluralize} in v3.0
         */
        this.pluralizeMap = null;
        /** @deprecated TODO remove in v3.0 */
        this.segmentsPluralizeMap = null;
        /** @deprecated TODO remove in v3.0 */
        this.pluralizeMapFallback = {
            one: ``,
            few: ``,
            many: ``,
            other: ``,
        };
    }
    // TODO: remove setter in v3.0:
    set pluralize(pluralize) {
        this.pluralizeMap = Array.isArray(pluralize)
            ? tuiPluralizeToICU(pluralize)
            : pluralize;
    }
    /**
     * @deprecated This input-property will be deleted in next major update.
     * You can create ticks labels below this input by yourself. It is easy!
     * Mixin `tui-slider-ticks-labels` will help with it. See examples in demo page `InputSlider`.
     * ___
     * TODO remove in v3.0.
     * Dont forget to delete:
     *** {@link segmentsPluralizeMap}
     *** {@link getTickPrefix}
     *** TuiFormatNumberPipeModule
     *** TuiRepeatTimesModule
     */
    set segmentsPluralize(pluralize) {
        this.segmentsPluralizeMap = Array.isArray(pluralize)
            ? tuiPluralizeToICU(pluralize)
            : pluralize;
    }
    /** @deprecated TODO remove in v3.0 */
    getTickPrefix(segment, segments, texts) {
        if (segments !== 1) {
            return ``;
        }
        return segment === 0 ? `${texts[0]}` : `${texts[1]}`;
    }
    get segmented() {
        return this.segments > 0;
    }
    get length() {
        return this.max - this.min;
    }
    get computedSteps() {
        return this.steps || this.length / this.quantum;
    }
    get step() {
        return this.length / this.computedSteps;
    }
    get computedKeySteps() {
        return this.computePureKeySteps(this.keySteps, this.min, this.max);
    }
    onHovered(hovered) {
        this.updateHovered(hovered);
    }
    computePureKeySteps(keySteps, min, max) {
        return [[0, min], ...(keySteps || []), [100, max]];
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "min", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "max", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "minLabel", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "maxLabel", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "pluralize", null);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "segmentsPluralize", null);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "segments", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "steps", void 0);
__decorate([
    Input(),
    tuiDefaultProp(quantumAssertion, `Quantum must be positive`)
], AbstractTuiInputSlider.prototype, "quantum", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "keySteps", void 0);
__decorate([
    Input(),
    HostBinding(`attr.data-size`),
    tuiDefaultProp()
], AbstractTuiInputSlider.prototype, "size", void 0);
__decorate([
    tuiPure
], AbstractTuiInputSlider.prototype, "getTickPrefix", null);
__decorate([
    HostBinding(`class._segmented`)
], AbstractTuiInputSlider.prototype, "segmented", null);
__decorate([
    tuiPure
], AbstractTuiInputSlider.prototype, "computePureKeySteps", null);
AbstractTuiInputSlider = __decorate([
    Directive()
], AbstractTuiInputSlider);
export { AbstractTuiInputSlider };
