// @taiga-ui/addon-doc/esm2015/internal/source-code/source-code.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { tuiPure } from '@taiga-ui/cdk';
import { TUI_DOC_SOURCE_CODE_TEXT } from '../../tokens/i18n';
import { TUI_DOC_SOURCE_CODE } from '../../tokens/source-code';
let TuiDocSourceCodeComponent = class TuiDocSourceCodeComponent {
    constructor(sourceCode, text) {
        this.sourceCode = sourceCode;
        this.text = text;
        this.header = '';
        this.package = '';
        this.type = '';
        this.path = '';
    }
    get pathOptions() {
        return this.getPathOptions(this.header, this.package, this.type, this.path);
    }
    getPathOptions(header, packageName, type, path) {
        return {
            header,
            package: packageName,
            type,
            path,
        };
    }
};
TuiDocSourceCodeComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_DOC_SOURCE_CODE,] }] },
    { type: String, decorators: [{ type: Inject, args: [TUI_DOC_SOURCE_CODE_TEXT,] }] }
];
__decorate([
    Input()
], TuiDocSourceCodeComponent.prototype, "header", void 0);
__decorate([
    Input()
], TuiDocSourceCodeComponent.prototype, "package", void 0);
__decorate([
    Input()
], TuiDocSourceCodeComponent.prototype, "type", void 0);
__decorate([
    Input()
], TuiDocSourceCodeComponent.prototype, "path", void 0);
__decorate([
    tuiPure
], TuiDocSourceCodeComponent.prototype, "getPathOptions", null);
TuiDocSourceCodeComponent = __decorate([
    Component({
        selector: 'tui-doc-source-code',
        template: "<div\n    *ngIf=\"!!sourceCode\"\n    polymorpheus-outlet\n    [content]=\"sourceCode\"\n    [context]=\"pathOptions\"\n>\n    <ng-template let-link>\n        <a\n            *ngIf=\"link\"\n            tuiIconButton\n            type=\"button\"\n            icon=\"tuiIconCodeLarge\"\n            appearance=\"icon\"\n            target=\"_blank\"\n            size=\"s\"\n            [title]=\"text\"\n            [href]=\"link\"\n        ></a>\n    </ng-template>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    __param(0, Inject(TUI_DOC_SOURCE_CODE)),
    __param(1, Inject(TUI_DOC_SOURCE_CODE_TEXT))
], TuiDocSourceCodeComponent);
export { TuiDocSourceCodeComponent };
