// @taiga-ui/cdk/bundles/taiga-ui-cdk-enums.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/enums', ['exports'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].cdk = global['taiga-ui'].cdk || {}, global['taiga-ui'].cdk.enums = {})));
}(this, (function (exports) { 'use strict';

    /* eslint-disable @typescript-eslint/naming-convention */

    (function (TuiCreditCardAutofillName) {
        TuiCreditCardAutofillName["Off"] = "off";
        TuiCreditCardAutofillName["CcName"] = "cc-name";
        TuiCreditCardAutofillName["CcNumber"] = "cc-number";
        TuiCreditCardAutofillName["CcCsc"] = "cc-csc";
        TuiCreditCardAutofillName["CcExpMonth"] = "cc-exp-month";
        /**
         * @deprecated
         * TODO: 3.0 replace with {@link CcExpMonth}
         */
        TuiCreditCardAutofillName["CcExp_mounth"] = "cc-exp-month";
        TuiCreditCardAutofillName["CcExpYear"] = "cc-exp-year";
        /**
         * @deprecated
         * TODO: 3.0 replace with {@link CcExpYear}
         */
        TuiCreditCardAutofillName["CcExp_year"] = "cc-exp-year";
        TuiCreditCardAutofillName["CcExp"] = "cc-exp";
        TuiCreditCardAutofillName["CcType"] = "cc-type";
    })(exports.TuiCreditCardAutofillName || (exports.TuiCreditCardAutofillName = {}));

    (function (TuiNameAutofillName) {
        TuiNameAutofillName["Off"] = "off";
        TuiNameAutofillName["Name"] = "name";
        TuiNameAutofillName["GivenName"] = "given-name";
        TuiNameAutofillName["AdditionalName"] = "additional-name";
        TuiNameAutofillName["FamilyName"] = "family-name";
    })(exports.TuiNameAutofillName || (exports.TuiNameAutofillName = {}));

    (function (TuiAccountAutofillName) {
        TuiAccountAutofillName["Off"] = "off";
        TuiAccountAutofillName["Username"] = "username";
        TuiAccountAutofillName["NewPassword"] = "new-password";
        TuiAccountAutofillName["CurrentPassword"] = "current-password";
    })(exports.TuiAccountAutofillName || (exports.TuiAccountAutofillName = {}));

    (function (TuiEmailAutofillName) {
        TuiEmailAutofillName["Off"] = "off";
        TuiEmailAutofillName["Email"] = "email";
    })(exports.TuiEmailAutofillName || (exports.TuiEmailAutofillName = {}));

    (function (TuiAddressAutofillName) {
        TuiAddressAutofillName["Off"] = "off";
        TuiAddressAutofillName["StreetAddress"] = "street-address";
        TuiAddressAutofillName["PostalCode"] = "postal-code";
        TuiAddressAutofillName["CountryName"] = "country-name";
    })(exports.TuiAddressAutofillName || (exports.TuiAddressAutofillName = {}));

    (function (TuiPhoneAutofillName) {
        TuiPhoneAutofillName["Off"] = "off";
        TuiPhoneAutofillName["Tel"] = "tel";
    })(exports.TuiPhoneAutofillName || (exports.TuiPhoneAutofillName = {}));

    (function (TuiDateAutofillName) {
        TuiDateAutofillName["Off"] = "off";
        TuiDateAutofillName["Bday"] = "bday";
    })(exports.TuiDateAutofillName || (exports.TuiDateAutofillName = {}));

    (function (TuiTransactionAutofillName) {
        TuiTransactionAutofillName["Off"] = "off";
        TuiTransactionAutofillName["TransactionCurrency"] = "transaction-currency";
        TuiTransactionAutofillName["TransactionAmount"] = "transaction-amount";
    })(exports.TuiTransactionAutofillName || (exports.TuiTransactionAutofillName = {}));


    (function (TuiDayOfWeek) {
        TuiDayOfWeek[TuiDayOfWeek["Sunday"] = 0] = "Sunday";
        TuiDayOfWeek[TuiDayOfWeek["Monday"] = 1] = "Monday";
        TuiDayOfWeek[TuiDayOfWeek["Tuesday"] = 2] = "Tuesday";
        TuiDayOfWeek[TuiDayOfWeek["Wednesday"] = 3] = "Wednesday";
        TuiDayOfWeek[TuiDayOfWeek["Thursday"] = 4] = "Thursday";
        TuiDayOfWeek[TuiDayOfWeek["Friday"] = 5] = "Friday";
        TuiDayOfWeek[TuiDayOfWeek["Saturday"] = 6] = "Saturday";
    })(exports.TuiDayOfWeek || (exports.TuiDayOfWeek = {}));

    /**
     * @deprecated use join type {@link TuiInputModeT}
     */

    (function (TuiInputMode) {
        TuiInputMode["None"] = "none";
        TuiInputMode["Text"] = "text";
        TuiInputMode["Decimal"] = "decimal";
        TuiInputMode["Numeric"] = "numeric";
        TuiInputMode["Tel"] = "tel";
        TuiInputMode["Email"] = "email";
        TuiInputMode["Url"] = "url";
        TuiInputMode["Search"] = "search";
    })(exports.TuiInputMode || (exports.TuiInputMode = {}));

    /**
     * @deprecated use join type {@link TuiInputTypeT}
     */

    (function (TuiInputType) {
        TuiInputType["Text"] = "text";
        TuiInputType["Tel"] = "tel";
        TuiInputType["Email"] = "email";
        TuiInputType["Url"] = "url";
        TuiInputType["Password"] = "password";
    })(exports.TuiInputType || (exports.TuiInputType = {}));


    (function (TuiMonthNumber) {
        TuiMonthNumber[TuiMonthNumber["January"] = 0] = "January";
        TuiMonthNumber[TuiMonthNumber["February"] = 1] = "February";
        TuiMonthNumber[TuiMonthNumber["March"] = 2] = "March";
        TuiMonthNumber[TuiMonthNumber["April"] = 3] = "April";
        TuiMonthNumber[TuiMonthNumber["May"] = 4] = "May";
        TuiMonthNumber[TuiMonthNumber["June"] = 5] = "June";
        TuiMonthNumber[TuiMonthNumber["July"] = 6] = "July";
        TuiMonthNumber[TuiMonthNumber["August"] = 7] = "August";
        TuiMonthNumber[TuiMonthNumber["September"] = 8] = "September";
        TuiMonthNumber[TuiMonthNumber["October"] = 9] = "October";
        TuiMonthNumber[TuiMonthNumber["November"] = 10] = "November";
        TuiMonthNumber[TuiMonthNumber["December"] = 11] = "December";
    })(exports.TuiMonthNumber || (exports.TuiMonthNumber = {}));

    /**
     * Modes for {@link TuiOverscrollDirective}:
     *
     * All — blocks all scroll actions
     * Scroll — blocks scroll actions only if content is scrollable (equivalent to overscroll-behavior: contain)
     * None — blocking is disabled
     * @deprecated use join type {@link TuiOverscrollModeT}
     */

    (function (TuiOverscrollMode) {
        TuiOverscrollMode["All"] = "all";
        TuiOverscrollMode["Scroll"] = "scroll";
        TuiOverscrollMode["None"] = "none";
    })(exports.TuiOverscrollMode || (exports.TuiOverscrollMode = {}));

    Object.defineProperty(exports, '__esModule', { value: true });

})));
