// @taiga-ui/cdk/bundles/taiga-ui-cdk-services.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@angular/common'), require('@taiga-ui/cdk/tokens'), require('@taiga-ui/cdk/observables'), require('rxjs/operators'), require('@taiga-ui/cdk/utils/dom'), require('@ng-web-apis/common'), require('@taiga-ui/cdk/constants'), require('@ng-web-apis/resize-observer'), require('@taiga-ui/cdk/classes'), require('@taiga-ui/cdk/utils/math'), require('@taiga-ui/cdk/utils/miscellaneous'), require('@taiga-ui/cdk/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/services', ['exports', '@angular/core', 'rxjs', '@angular/common', '@taiga-ui/cdk/tokens', '@taiga-ui/cdk/observables', 'rxjs/operators', '@taiga-ui/cdk/utils/dom', '@ng-web-apis/common', '@taiga-ui/cdk/constants', '@ng-web-apis/resize-observer', '@taiga-ui/cdk/classes', '@taiga-ui/cdk/utils/math', '@taiga-ui/cdk/utils/miscellaneous', '@taiga-ui/cdk/utils'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].cdk = global['taiga-ui'].cdk || {}, global['taiga-ui'].cdk.services = {}), global.ng.core, global.rxjs, global.ng.common, global['taiga-ui'].cdk.tokens, global['taiga-ui'].cdk.observables, global.rxjs.operators, global['taiga-ui'].cdk.utils.dom, global.common$1, global['taiga-ui'].cdk.constants, global.resizeObserver, global['taiga-ui'].cdk.classes, global['taiga-ui'].cdk.utils.math, global['taiga-ui'].cdk.utils.miscellaneous, global['taiga-ui'].cdk.utils));
}(this, (function (exports, core, rxjs, common, tokens, observables, operators, dom, common$1, constants, resizeObserver, classes, math, miscellaneous, utils) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    /**
     * Observable abstraction over ngOnDestroy to use with takeUntil
     */
    var TuiDestroyService = /** @class */ (function (_super) {
        __extends(TuiDestroyService, _super);
        function TuiDestroyService() {
            return _super.call(this, 1) || this;
        }
        TuiDestroyService.prototype.ngOnDestroy = function () {
            this.next();
            this.complete();
        };
        TuiDestroyService = __decorate([
            core.Injectable()
        ], TuiDestroyService);
        return TuiDestroyService;
    }(rxjs.ReplaySubject));

    /**
     * Service to use styles with directives
     * @dynamic
     */
    var TuiDirectiveStylesService = /** @class */ (function () {
        function TuiDirectiveStylesService(resolver, injector, documentRef, renderer) {
            this.resolver = resolver;
            this.injector = injector;
            this.documentRef = documentRef;
            this.renderer = renderer;
            this.map = new Map();
        }
        TuiDirectiveStylesService.prototype.addComponent = function (component) {
            if (!this.map.has(component)) {
                this.map.set(component, this.resolver.resolveComponentFactory(component).create(this.injector));
            }
        };
        /** @deprecated use components approach
         * TODO: delete in v3.0
         */
        TuiDirectiveStylesService.prototype.addStyle = function (styles, attribute) {
            if (this.documentRef.head.querySelector("style[" + attribute + "]")) {
                return;
            }
            var style = this.renderer.createElement("style");
            this.renderer.setProperty(style, "textContent", styles);
            this.renderer.setAttribute(style, attribute, "");
            this.documentRef.head.appendChild(style);
        };
        TuiDirectiveStylesService.ctorParameters = function () { return [
            { type: core.ComponentFactoryResolver, decorators: [{ type: core.Inject, args: [core.ComponentFactoryResolver,] }] },
            { type: core.Injector, decorators: [{ type: core.Inject, args: [core.INJECTOR,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.Renderer2, decorators: [{ type: core.Inject, args: [tokens.TUI_DEFAULT_RENDERER,] }] }
        ]; };
        TuiDirectiveStylesService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiDirectiveStylesService_Factory() { return new TuiDirectiveStylesService(core.ɵɵinject(core.ComponentFactoryResolver), core.ɵɵinject(core.INJECTOR), core.ɵɵinject(common.DOCUMENT), core.ɵɵinject(tokens.TUI_DEFAULT_RENDERER)); }, token: TuiDirectiveStylesService, providedIn: "root" });
        TuiDirectiveStylesService = __decorate([
            core.Injectable({
                providedIn: "root",
            }),
            __param(0, core.Inject(core.ComponentFactoryResolver)),
            __param(1, core.Inject(core.INJECTOR)),
            __param(2, core.Inject(common.DOCUMENT)),
            __param(3, core.Inject(tokens.TUI_DEFAULT_RENDERER))
        ], TuiDirectiveStylesService);
        return TuiDirectiveStylesService;
    }());

    /**
     * Service to imitate :focus-visible
     * (https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible)
     * in browsers that do not support it
     * @dynamic
     */
    var TuiFocusVisibleService = /** @class */ (function (_super) {
        __extends(TuiFocusVisibleService, _super);
        function TuiFocusVisibleService(_a, changeDetectorRef, destroy$) {
            var nativeElement = _a.nativeElement;
            var _this = _super.call(this, function (subscriber) { return _this.focusVisible$.subscribe(subscriber); }) || this;
            _this.focusVisible$ = observables.focusVisibleObservable(nativeElement).pipe(observables.watch(changeDetectorRef), operators.takeUntil(destroy$));
            return _this;
        }
        TuiFocusVisibleService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [TuiDestroyService,] }] }
        ]; };
        TuiFocusVisibleService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(TuiDestroyService))
        ], TuiFocusVisibleService);
        return TuiFocusVisibleService;
    }(rxjs.Observable));

    /** @deprecated TODO remove in v3.0 */
    // @dynamic
    var TuiHoveredService = /** @class */ (function () {
        function TuiHoveredService(documentRef, ngZone) {
            this.ngZone = ngZone;
            this.documentEvents$ = rxjs.merge(observables.typedFromEvent(documentRef, "mousemove"), observables.typedFromEvent(documentRef, "touchstart", { capture: true }));
        }
        TuiHoveredService.prototype.createHovered$ = function (target, options) {
            var _this = this;
            if (options === void 0) { options = { passive: true }; }
            return rxjs.merge(observables.typedFromEvent(target, "mouseenter", options), observables.typedFromEvent(target, "touchstart", options)).pipe(operators.switchMap(function () {
                return rxjs.merge(observables.typedFromEvent(target, "mouseleave", options), _this.documentEvents$.pipe(operators.filter(function (event) { return !target.contains(dom.getActualTarget(event)); }), observables.tuiZoneOptimized(_this.ngZone), operators.take(1))).pipe(operators.mapTo(false), operators.startWith(true));
            }), operators.distinctUntilChanged());
        };
        TuiHoveredService.ctorParameters = function () { return [
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] }
        ]; };
        TuiHoveredService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiHoveredService_Factory() { return new TuiHoveredService(core.ɵɵinject(common.DOCUMENT), core.ɵɵinject(core.NgZone)); }, token: TuiHoveredService, providedIn: "root" });
        TuiHoveredService = __decorate([
            core.Injectable({
                providedIn: "root",
            }),
            __param(0, core.Inject(common.DOCUMENT)),
            __param(1, core.Inject(core.NgZone))
        ], TuiHoveredService);
        return TuiHoveredService;
    }());

    var TUI = "tui_";
    /**
     * Generates unique ids
     */
    var TuiIdService = /** @class */ (function () {
        function TuiIdService() {
        }
        TuiIdService_1 = TuiIdService;
        TuiIdService.prototype.generate = function () {
            return "" + TUI + TuiIdService_1.autoId++ + Date.now();
        };
        var TuiIdService_1;
        TuiIdService.autoId = 0;
        TuiIdService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiIdService_Factory() { return new TuiIdService(); }, token: TuiIdService, providedIn: "root" });
        TuiIdService = TuiIdService_1 = __decorate([
            core.Injectable({
                providedIn: "root",
            })
        ], TuiIdService);
        return TuiIdService;
    }());

    /**
     * Service that subscribes to scroll events of all parent elements
     */
    // Ambient type cannot be used without dynamic https://github.com/angular/angular/issues/23395
    // @dynamic
    var TuiParentsScrollService = /** @class */ (function (_super) {
        __extends(TuiParentsScrollService, _super);
        function TuiParentsScrollService(elementRef, windowRef) {
            var _this = _super.call(this, function (subscriber) { return _this.callback$.subscribe(subscriber); }) || this;
            _this.callback$ = rxjs.defer(function () {
                var nativeElement = elementRef.nativeElement;
                var eventTargets = [windowRef, nativeElement];
                while (nativeElement.parentElement) {
                    nativeElement = nativeElement.parentElement;
                    eventTargets.push(nativeElement);
                }
                return rxjs.merge.apply(void 0, __spread(eventTargets.map(function (element) {
                    return observables.typedFromEvent(element, "scroll");
                })));
            });
            return _this;
        }
        TuiParentsScrollService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common$1.WINDOW,] }] }
        ]; };
        TuiParentsScrollService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(common$1.WINDOW))
        ], TuiParentsScrollService);
        return TuiParentsScrollService;
    }(rxjs.Observable));

    // @bad TODO: Consider Intersection Observer with fallback to current implementation
    /**
     * Service that monitors element visibility by subscribing to scrolls
     * and polling with set interval, returns either null or an array
     * of elements that overlap given element edges
     */
    // Ambient type cannot be used without dynamic https://github.com/angular/angular/issues/23395
    // @dynamic
    var TuiObscuredService = /** @class */ (function (_super) {
        __extends(TuiObscuredService, _super);
        function TuiObscuredService(parentsScroll$, _a, ngZone, windowRef, destroy$, animationFrame$) {
            var nativeElement = _a.nativeElement;
            var _this = _super.call(this, function (subscriber) { return _this.obscured$.subscribe(subscriber); }) || this;
            _this.obscured$ = rxjs.merge(
            // delay is added so it will not interfere with other listeners
            rxjs.merge(parentsScroll$, rxjs.fromEvent(windowRef, "resize")).pipe(operators.delay(0)), animationFrame$.pipe(operators.throttleTime(constants.POLLING_TIME))).pipe(operators.map(function () { return dom.getElementObscurers(nativeElement); }), operators.startWith(null), operators.distinctUntilChanged(), observables.tuiZoneOptimized(ngZone), operators.takeUntil(destroy$));
            return _this;
        }
        TuiObscuredService.ctorParameters = function () { return [
            { type: TuiParentsScrollService, decorators: [{ type: core.Inject, args: [TuiParentsScrollService,] }, { type: core.Self }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common$1.WINDOW,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [TuiDestroyService,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common$1.ANIMATION_FRAME,] }] }
        ]; };
        TuiObscuredService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(TuiParentsScrollService)),
            __param(0, core.Self()),
            __param(1, core.Inject(core.ElementRef)),
            __param(2, core.Inject(core.NgZone)),
            __param(3, core.Inject(common$1.WINDOW)),
            __param(4, core.Inject(TuiDestroyService)),
            __param(5, core.Inject(common$1.ANIMATION_FRAME))
        ], TuiObscuredService);
        return TuiObscuredService;
    }(rxjs.Observable));

    /**
     * @dynamic
     */
    var TuiPanService = /** @class */ (function (_super) {
        __extends(TuiPanService, _super);
        function TuiPanService(_a, documentRef) {
            var nativeElement = _a.nativeElement;
            return _super.call(this, function (subscriber) {
                rxjs.merge(observables.typedFromEvent(nativeElement, "touchstart", { passive: true }), observables.typedFromEvent(nativeElement, "mousedown"))
                    .pipe(operators.switchMapTo(rxjs.merge(observables.typedFromEvent(documentRef, "touchmove", {
                    passive: true,
                }).pipe(operators.filter(function (_a) {
                    var touches = _a.touches;
                    return touches.length < 2;
                }), operators.map(function (_a) {
                    var touches = _a.touches;
                    return touches[0];
                })), observables.typedFromEvent(documentRef, "mousemove"))), operators.pairwise(), operators.map(function (_a) {
                    var _b = __read(_a, 2), first = _b[0], second = _b[1];
                    var deltaX = second.clientX - first.clientX;
                    var deltaY = second.clientY - first.clientY;
                    return [deltaX, deltaY];
                }), 
                // eslint-disable-next-line rxjs/no-unsafe-takeuntil
                operators.takeUntil(rxjs.merge(observables.typedFromEvent(documentRef, "touchend"), observables.typedFromEvent(documentRef, "mouseup"))), operators.repeat())
                    .subscribe(subscriber);
            }) || this;
        }
        TuiPanService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
        ]; };
        TuiPanService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(common.DOCUMENT))
        ], TuiPanService);
        return TuiPanService;
    }(rxjs.Observable));

    // @dynamic
    var TuiResizeService = /** @class */ (function (_super) {
        __extends(TuiResizeService, _super);
        function TuiResizeService(elementRef, ngZone, destroy$, support, box, animationFrame$) {
            var _this = _super.call(this, elementRef, ngZone, support, box) || this;
            return _this.pipe(operators.catchError(function () {
                return animationFrame$.pipe(operators.throttleTime(constants.POLLING_TIME), operators.map(function () {
                    return elementRef.nativeElement.clientWidth + " " + elementRef.nativeElement.clientHeight;
                }), operators.distinctUntilChanged(), operators.mapTo(constants.EMPTY_ARRAY));
            }), operators.debounceTime(0), observables.tuiZonefree(ngZone), operators.takeUntil(destroy$));
        }
        TuiResizeService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [TuiDestroyService,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [resizeObserver.RESIZE_OBSERVER_SUPPORT,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [resizeObserver.RESIZE_OPTION_BOX,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common$1.ANIMATION_FRAME,] }] }
        ]; };
        TuiResizeService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(core.NgZone)),
            __param(2, core.Inject(TuiDestroyService)),
            __param(3, core.Inject(resizeObserver.RESIZE_OBSERVER_SUPPORT)),
            __param(4, core.Inject(resizeObserver.RESIZE_OPTION_BOX)),
            __param(5, core.Inject(common$1.ANIMATION_FRAME))
        ], TuiResizeService);
        return TuiResizeService;
    }(resizeObserver.ResizeObserverService));

    var SCROLL_TIME = 300;
    // @dynamic
    var TuiScrollService = /** @class */ (function () {
        function TuiScrollService(performanceRef, animationFrame$) {
            this.performanceRef = performanceRef;
            this.animationFrame$ = animationFrame$;
        }
        TuiScrollService.prototype.scroll$ = function (elementOrWindow, scrollTop, scrollLeft, duration) {
            var _this = this;
            if (scrollLeft === void 0) { scrollLeft = getX(elementOrWindow); }
            if (duration === void 0) { duration = SCROLL_TIME; }
            classes.tuiAssert.assert(duration >= 0, "Duration cannot be negative");
            classes.tuiAssert.assert(scrollTop >= 0, "scrollTop cannot be negative");
            classes.tuiAssert.assert(scrollLeft >= 0, "scrollLeft cannot be negative");
            var initialTop = getY(elementOrWindow);
            var initialLeft = getX(elementOrWindow);
            var deltaTop = scrollTop - initialTop;
            var deltaLeft = scrollLeft - initialLeft;
            var observable = !duration
                ? rxjs.of([scrollTop, scrollLeft])
                : rxjs.defer(function () { return rxjs.of(_this.performanceRef.now()); }).pipe(operators.switchMap(function (start) { return _this.animationFrame$.pipe(operators.map(function (now) { return now - start; })); }), operators.map(function (elapsed) { return miscellaneous.tuiEaseInOutQuad(math.tuiClamp(elapsed / duration, 0, 1)); }), operators.map(function (percent) {
                    return [
                        initialTop + deltaTop * percent,
                        initialLeft + deltaLeft * percent,
                    ];
                }), operators.takeUntil(rxjs.timer(duration)));
            return observable.pipe(operators.tap(function (_a) {
                var _b = __read(_a, 2), scrollTop = _b[0], scrollLeft = _b[1];
                var _c;
                (_c = elementOrWindow.scrollTo) === null || _c === void 0 ? void 0 : _c.call(elementOrWindow, scrollLeft, scrollTop);
            }));
        };
        TuiScrollService.ctorParameters = function () { return [
            { type: Performance, decorators: [{ type: core.Inject, args: [common$1.PERFORMANCE,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common$1.ANIMATION_FRAME,] }] }
        ]; };
        TuiScrollService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiScrollService_Factory() { return new TuiScrollService(core.ɵɵinject(common$1.PERFORMANCE), core.ɵɵinject(common$1.ANIMATION_FRAME)); }, token: TuiScrollService, providedIn: "root" });
        TuiScrollService = __decorate([
            core.Injectable({
                providedIn: "root",
            }),
            __param(0, core.Inject(common$1.PERFORMANCE)),
            __param(1, core.Inject(common$1.ANIMATION_FRAME))
        ], TuiScrollService);
        return TuiScrollService;
    }());
    function getX(elementOrWindow) {
        return "scrollX" in elementOrWindow
            ? elementOrWindow.scrollX
            : elementOrWindow.scrollLeft;
    }
    function getY(elementOrWindow) {
        return "scrollY" in elementOrWindow
            ? elementOrWindow.scrollY
            : elementOrWindow.scrollTop;
    }

    var TuiStaticRequestService = /** @class */ (function () {
        function TuiStaticRequestService() {
            this.cache = new Map();
        }
        TuiStaticRequestService.prototype.request = function (url) {
            var cache = this.cache.get(url);
            if (cache) {
                return cache;
            }
            var observable = new rxjs.Observable(function (observer) {
                var xhr = new XMLHttpRequest();
                xhr.onreadystatechange = function () {
                    if (xhr.readyState === 4) {
                        var response = xhr.responseType ? xhr.response : xhr.responseText;
                        if (xhr.status === 200) {
                            observer.next(response);
                            observer.complete();
                        }
                        else {
                            observer.error(response);
                        }
                    }
                };
                xhr.open("GET", url);
                xhr.send();
                return function () {
                    xhr.abort();
                };
            });
            var piped = observable.pipe(operators.shareReplay({ bufferSize: 1, refCount: false }));
            this.cache.set(url, piped);
            return piped;
        };
        TuiStaticRequestService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiStaticRequestService_Factory() { return new TuiStaticRequestService(); }, token: TuiStaticRequestService, providedIn: "root" });
        TuiStaticRequestService = __decorate([
            core.Injectable({
                providedIn: "root",
            })
        ], TuiStaticRequestService);
        return TuiStaticRequestService;
    }());

    /**
     * @dynamic
     */
    var TuiSwipeService = /** @class */ (function (_super) {
        __extends(TuiSwipeService, _super);
        function TuiSwipeService(_a, _b, documentRef) {
            var nativeElement = _a.nativeElement;
            var timeout = _b.timeout, threshold = _b.threshold;
            return _super.call(this, function (subscriber) {
                rxjs.merge(observables.typedFromEvent(nativeElement, "touchstart", { passive: true }), observables.typedFromEvent(documentRef, "touchend"))
                    .pipe(operators.pairwise(), operators.filter(function (_a) {
                    var _b = __read(_a, 2), first = _b[0], second = _b[1];
                    return !!first.touches.length &&
                        first.touches[0].identifier ===
                            second.changedTouches[0].identifier;
                }), operators.map(function (_a) {
                    var _b = __read(_a, 2), start = _b[0], end = _b[1];
                    var startX = start.touches[0].clientX;
                    var startY = start.touches[0].clientY;
                    var endX = end.changedTouches[0].clientX;
                    var endY = end.changedTouches[0].clientY;
                    var distanceX = startX - endX;
                    var distanceY = startY - endY;
                    var duration = end.timeStamp - start.timeStamp;
                    if ((Math.abs(distanceX) > threshold ||
                        Math.abs(distanceY) > threshold) &&
                        duration < timeout) {
                        return {
                            direction: miscellaneous.getSwipeDirection(distanceX, distanceY),
                            events: [start, end],
                        };
                    }
                    return null;
                }), operators.filter(miscellaneous.isPresent))
                    .subscribe(subscriber);
            }) || this;
        }
        TuiSwipeService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_SWIPE_OPTIONS,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
        ]; };
        TuiSwipeService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(tokens.TUI_SWIPE_OPTIONS)),
            __param(2, core.Inject(common.DOCUMENT))
        ], TuiSwipeService);
        return TuiSwipeService;
    }(rxjs.Observable));

    var TOUCH_SENSITIVITY = 0.01;
    // @dynamic
    var TuiZoomService = /** @class */ (function (_super) {
        __extends(TuiZoomService, _super);
        function TuiZoomService(_a, _b) {
            var nativeElement = _a.nativeElement;
            var wheelSensitivity = _b.wheelSensitivity;
            return _super.call(this, function (subscriber) {
                rxjs.merge(observables.typedFromEvent(nativeElement, "touchstart", { passive: true }).pipe(operators.filter(function (_a) {
                    var touches = _a.touches;
                    return touches.length > 1;
                }), operators.switchMap(function (startEvent) {
                    return observables.typedFromEvent(nativeElement, "touchmove", {
                        passive: true,
                    }).pipe(observables.preventDefault(), operators.scan(function (prev, event) {
                        var distance = utils.distanceBetweenTouches(event);
                        return {
                            event: event,
                            distance: distance,
                            delta: (distance - prev.distance) *
                                TOUCH_SENSITIVITY,
                        };
                    }, {
                        event: startEvent,
                        distance: utils.distanceBetweenTouches(startEvent),
                        delta: 0,
                    }), operators.map(function (_a) {
                        var event = _a.event, delta = _a.delta;
                        var clientX = (event.touches[0].clientX +
                            event.touches[1].clientX) /
                            2;
                        var clientY = (event.touches[0].clientY +
                            event.touches[1].clientY) /
                            2;
                        return { clientX: clientX, clientY: clientY, delta: delta, event: event };
                    }), operators.takeUntil(observables.typedFromEvent(nativeElement, "touchend")));
                })), observables.typedFromEvent(nativeElement, "wheel", { passive: false }).pipe(observables.preventDefault(), operators.map(function (wheel) { return ({
                    clientX: wheel.clientX,
                    clientY: wheel.clientY,
                    delta: -wheel.deltaY * wheelSensitivity,
                    event: wheel,
                }); }))).subscribe(subscriber);
            }) || this;
        }
        TuiZoomService.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_ZOOM_OPTIONS,] }] }
        ]; };
        TuiZoomService = __decorate([
            core.Injectable(),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(tokens.TUI_ZOOM_OPTIONS))
        ], TuiZoomService);
        return TuiZoomService;
    }(rxjs.Observable));

    exports.TuiDestroyService = TuiDestroyService;
    exports.TuiDirectiveStylesService = TuiDirectiveStylesService;
    exports.TuiFocusVisibleService = TuiFocusVisibleService;
    exports.TuiHoveredService = TuiHoveredService;
    exports.TuiIdService = TuiIdService;
    exports.TuiObscuredService = TuiObscuredService;
    exports.TuiPanService = TuiPanService;
    exports.TuiParentsScrollService = TuiParentsScrollService;
    exports.TuiResizeService = TuiResizeService;
    exports.TuiScrollService = TuiScrollService;
    exports.TuiStaticRequestService = TuiStaticRequestService;
    exports.TuiSwipeService = TuiSwipeService;
    exports.TuiZoomService = TuiZoomService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
