// @taiga-ui/cdk/bundles/taiga-ui-cdk-utils-browser.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/utils/browser', ['exports'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].cdk = global['taiga-ui'].cdk || {}, global['taiga-ui'].cdk.utils = global['taiga-ui'].cdk.utils || {}, global['taiga-ui'].cdk.utils.browser = {})));
}(this, (function (exports) { 'use strict';

    /**
     * @deprecated: use {@link tuiIsEdge} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function isEdge(userAgent) {
        return userAgent.toLowerCase().includes("edge");
    }
    var tuiIsEdge = isEdge;

    var EDGE = "edge/";
    /**
     * @deprecated: use {@link tuiIsEdgeOlderThan} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function isEdgeOlderThan(version, userAgent) {
        var currentVersion = parseInt(userAgent.slice(userAgent.toLowerCase().indexOf(EDGE) + EDGE.length), 10);
        return currentVersion < version;
    }
    var tuiIsEdgeOlderThan = isEdgeOlderThan;

    /**
     * @deprecated: use {@link tuiIsFirefox} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function isFirefox(userAgent) {
        return userAgent.toLowerCase().includes("firefox");
    }
    var tuiIsFirefox = isFirefox;

    /**
     * @deprecated: use {@link tuiIsIE} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function isIE(userAgent) {
        return userAgent.toLowerCase().includes("trident");
    }
    var tuiIsIE = isIE;

    /**
     * @deprecated: use {@link tuiIsSafari} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function isSafari(_a) {
        var documentRef = _a.ownerDocument;
        var _b, _c, _d, _e, _f, _g, _h, _j;
        var windowRef = documentRef === null || documentRef === void 0 ? void 0 : documentRef.defaultView;
        var isMacOsSafari = typeof windowRef.safari !== "undefined" &&
            ((_c = (_b = windowRef.safari) === null || _b === void 0 ? void 0 : _b.pushNotification) === null || _c === void 0 ? void 0 : _c.toString()) ===
                "[object SafariRemoteNotification]";
        var isIosSafari = !!((_e = (_d = windowRef.navigator) === null || _d === void 0 ? void 0 : _d.vendor) === null || _e === void 0 ? void 0 : _e.includes("Apple")) &&
            !((_g = (_f = windowRef.navigator) === null || _f === void 0 ? void 0 : _f.userAgent) === null || _g === void 0 ? void 0 : _g.includes("CriOS")) &&
            !((_j = (_h = windowRef.navigator) === null || _h === void 0 ? void 0 : _h.userAgent) === null || _j === void 0 ? void 0 : _j.includes("FxiOS"));
        return isMacOsSafari || isIosSafari;
    }
    var tuiIsSafari = isSafari;

    exports.isEdge = isEdge;
    exports.isEdgeOlderThan = isEdgeOlderThan;
    exports.isFirefox = isFirefox;
    exports.isIE = isIE;
    exports.isSafari = isSafari;
    exports.tuiIsEdge = tuiIsEdge;
    exports.tuiIsEdgeOlderThan = tuiIsEdgeOlderThan;
    exports.tuiIsFirefox = tuiIsFirefox;
    exports.tuiIsIE = tuiIsIE;
    exports.tuiIsSafari = tuiIsSafari;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
