// @taiga-ui/kit/esm5/providers/specific-dropdown-controllers.js
import { Optional } from '@angular/core';
import { TUI_DROPDOWN_CONTROLLER, TuiDropdownControllerDirective } from '@taiga-ui/core';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function fixedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.limitWidth = "fixed";
    return directive;
}
export var FIXED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: fixedDropdownControllerFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
export function leftAlignedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.align = "left";
    return directive;
}
export var LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: leftAlignedDropdownControllerFactory,
    },
];
