// @taiga-ui/kit/esm5/components/input-date/input-date.providers.js
import { forwardRef, Optional, Self } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiControl, TUI_FOCUSABLE_ITEM_ACCESSOR, } from '@taiga-ui/cdk';
import { LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER } from '@taiga-ui/kit/providers';
import { TUI_CALENDAR_DATA_STREAM, TUI_DATE_VALUE_TRANSFORMER } from '@taiga-ui/kit/tokens';
import { TuiReplayControlValueChangesFactory } from '@taiga-ui/kit/utils/miscellaneous';
import { TuiInputDateComponent } from './input-date.component';
// TODO: 3.0 remove in ivy compilation
export var DATE_STREAM_FACTORY = function (control, valueTransformer) {
    return TuiReplayControlValueChangesFactory(control, valueTransformer);
};
export var TUI_INPUT_DATE_PROVIDERS = [
    {
        provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
        useExisting: forwardRef(function () { return TuiInputDateComponent; }),
    },
    {
        provide: TUI_CALENDAR_DATA_STREAM,
        deps: [
            [new Optional(), new Self(), NgControl],
            [new Optional(), TUI_DATE_VALUE_TRANSFORMER],
        ],
        useFactory: DATE_STREAM_FACTORY,
    },
    {
        provide: AbstractTuiControl,
        useExisting: forwardRef(function () { return TuiInputDateComponent; }),
    },
    LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER,
];
