// @taiga-ui/kit/esm5/components/input-date/input-date.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputDateDirective = /** @class */ (function (_super) {
    __extends(TuiInputDateDirective, _super);
    function TuiInputDateDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputDateDirective_1 = TuiInputDateDirective;
    Object.defineProperty(TuiInputDateDirective.prototype, "value", {
        get: function () {
            return this.host.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiInputDateDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    TuiInputDateDirective.prototype.process = function (input) {
        input.inputMode = 'numeric';
    };
    var TuiInputDateDirective_1;
    TuiInputDateDirective = TuiInputDateDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-date',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputDateDirective_1; }),
                },
            ],
        })
    ], TuiInputDateDirective);
    return TuiInputDateDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputDateDirective };
