// @taiga-ui/core/esm2015/abstract/abstract-textfield-host.js
import { __decorate, __param } from "tslib";
import { Directive, Inject } from '@angular/core';
import { AbstractTuiControl } from '@taiga-ui/cdk';
// @dynamic
let TuiAbstractTextfieldHost = 
// eslint-disable-next-line @typescript-eslint/naming-convention
class TuiAbstractTextfieldHost {
    constructor(host) {
        this.host = host;
    }
    get readOnly() {
        return this.host.readOnly;
    }
    get disabled() {
        return this.host.computedDisabled;
    }
    get invalid() {
        return this.host.computedInvalid;
    }
    get focusable() {
        return this.host.computedFocusable;
    }
    get value() {
        return String(this.host.value);
    }
    process(_input) { }
};
TuiAbstractTextfieldHost.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [AbstractTuiControl,] }] }
];
TuiAbstractTextfieldHost = __decorate([
    Directive()
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ,
    __param(0, Inject(AbstractTuiControl))
], TuiAbstractTextfieldHost);
export { TuiAbstractTextfieldHost };
