// @taiga-ui/addon-doc/esm5/components/language-switcher/language-switcher.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { FormControl } from '@angular/forms';
import { tuiCapitalizeFirstLetter } from '@taiga-ui/core';
import { TuiCountryIsoCode, TuiLanguageName, TuiLanguageSwitcher } from '@taiga-ui/i18n';
var TuiLanguageSwitcherComponent = /** @class */ (function () {
    function TuiLanguageSwitcherComponent(switcher) {
        this.switcher = switcher;
        this.language = new FormControl(tuiCapitalizeFirstLetter(this.switcher.language));
        this.flags = new Map([
            ['chinese', TuiCountryIsoCode.CN],
            ['dutch', TuiCountryIsoCode.NL],
            ['english', TuiCountryIsoCode.GB],
            ['french', TuiCountryIsoCode.FR],
            ['german', TuiCountryIsoCode.DE],
            ['italian', TuiCountryIsoCode.IT],
            ['polish', TuiCountryIsoCode.PL],
            ['portuguese', TuiCountryIsoCode.PT],
            ['russian', TuiCountryIsoCode.RU],
            ['spanish', TuiCountryIsoCode.ES],
            ['turkish', TuiCountryIsoCode.TR],
            ['ukrainian', TuiCountryIsoCode.UA],
            ['vietnamese', TuiCountryIsoCode.VN],
        ]);
        this.names = Array.from(this.flags.keys());
    }
    TuiLanguageSwitcherComponent.ctorParameters = function () { return [
        { type: TuiLanguageSwitcher, decorators: [{ type: Inject, args: [TuiLanguageSwitcher,] }] }
    ]; };
    TuiLanguageSwitcherComponent = __decorate([
        Component({
            selector: 'tui-language-switcher',
            template: "<tui-select [formControl]=\"language\">\n    <ng-content></ng-content>\n    <tui-data-list *tuiDataList>\n        <button\n            *ngFor=\"let name of names\"\n            tuiOption\n            [value]=\"name | titlecase\"\n            (click)=\"switcher.setLanguage(name)\"\n        >\n            <img\n                alt=\"\"\n                class=\"t-flag\"\n                [src]=\"flags.get(name) | tuiFlag\"\n            />\n            {{ name | titlecase }}\n        </button>\n    </tui-data-list>\n</tui-select>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".t-flag{width:1.75rem;height:1.25rem;margin-right:.5rem}"]
        }),
        __param(0, Inject(TuiLanguageSwitcher))
    ], TuiLanguageSwitcherComponent);
    return TuiLanguageSwitcherComponent;
}());
export { TuiLanguageSwitcherComponent };
