// @taiga-ui/kit/esm5/abstract/input-slider.js
import { __decorate, __extends, __read, __spread } from "tslib";
import { Directive, HostBinding, Input } from '@angular/core';
import { AbstractTuiControl, tuiDefaultProp, tuiPure } from '@taiga-ui/cdk';
import { tuiPluralizeToICU, } from '@taiga-ui/core';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function quantumAssertion(quantum) {
    return quantum > 0;
}
/**
 * @internal
 * @deprecated TODO: 3.0 delete me after `InputSlider` and `InputRange` stop using it
 */
var AbstractTuiInputSlider = /** @class */ (function (_super) {
    __extends(AbstractTuiInputSlider, _super);
    function AbstractTuiInputSlider() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.min = 0;
        _this.max = Infinity;
        /**
         * @deprecated This input-property will be deleted in next major update.
         * Use `valueContent` for `InputSlider`.
         * Use `leftValueContent` for `InputRange`.
         * ___
         * TODO remove in v3.0.
         * Dont forget to delete backward-compatibility helpers inside `InputSlider` and `InputRange`:
         *** {@link legacyMinMaxLabel}
         */
        _this.minLabel = "";
        /**
         * @deprecated This input-property will be deleted in next major update.
         * Use `valueContent` for `InputSlider`.
         * Use `rightValueContent` for `InputRange`.
         * ___
         * TODO remove in v3.0.
         * Dont forget to delete backward-compatibility helpers inside `InputSlider` and `InputRange`:
         *** {@link legacyMinMaxLabel}
         */
        _this.maxLabel = "";
        _this.segments = 0;
        _this.steps = 0;
        _this.quantum = 1;
        _this.keySteps = null;
        /**
         * @deprecated use `tuiTextfieldSize` instead
         * TODO delete in v3.0
         */
        _this.size = "l";
        /**
         * @deprecated replace by input-property {@link pluralize} in v3.0
         */
        _this.pluralizeMap = null;
        /** @deprecated TODO remove in v3.0 */
        _this.segmentsPluralizeMap = null;
        /** @deprecated TODO remove in v3.0 */
        _this.pluralizeMapFallback = {
            one: "",
            few: "",
            many: "",
            other: "",
        };
        return _this;
    }
    Object.defineProperty(AbstractTuiInputSlider.prototype, "pluralize", {
        // TODO: remove setter in v3.0:
        set: function (pluralize) {
            this.pluralizeMap = Array.isArray(pluralize)
                ? tuiPluralizeToICU(pluralize)
                : pluralize;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AbstractTuiInputSlider.prototype, "segmentsPluralize", {
        /**
         * @deprecated This input-property will be deleted in next major update.
         * You can create ticks labels below this input by yourself. It is easy!
         * Mixin `tui-slider-ticks-labels` will help with it. See examples in demo page `InputSlider`.
         * ___
         * TODO remove in v3.0.
         * Dont forget to delete:
         *** {@link segmentsPluralizeMap}
         *** {@link getTickPrefix}
         *** TuiFormatNumberPipeModule
         *** TuiRepeatTimesModule
         */
        set: function (pluralize) {
            this.segmentsPluralizeMap = Array.isArray(pluralize)
                ? tuiPluralizeToICU(pluralize)
                : pluralize;
        },
        enumerable: true,
        configurable: true
    });
    /** @deprecated TODO remove in v3.0 */
    AbstractTuiInputSlider.prototype.getTickPrefix = function (segment, segments, texts) {
        if (segments !== 1) {
            return "";
        }
        return segment === 0 ? "" + texts[0] : "" + texts[1];
    };
    Object.defineProperty(AbstractTuiInputSlider.prototype, "segmented", {
        get: function () {
            return this.segments > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AbstractTuiInputSlider.prototype, "length", {
        get: function () {
            return this.max - this.min;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AbstractTuiInputSlider.prototype, "computedSteps", {
        get: function () {
            return this.steps || this.length / this.quantum;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AbstractTuiInputSlider.prototype, "step", {
        get: function () {
            return this.length / this.computedSteps;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AbstractTuiInputSlider.prototype, "computedKeySteps", {
        get: function () {
            return this.computePureKeySteps(this.keySteps, this.min, this.max);
        },
        enumerable: true,
        configurable: true
    });
    AbstractTuiInputSlider.prototype.onHovered = function (hovered) {
        this.updateHovered(hovered);
    };
    AbstractTuiInputSlider.prototype.computePureKeySteps = function (keySteps, min, max) {
        return __spread([[0, min]], (keySteps || []), [[100, max]]);
    };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "min", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "max", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "minLabel", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "maxLabel", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "pluralize", null);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "segmentsPluralize", null);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "segments", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "steps", void 0);
    __decorate([
        Input(),
        tuiDefaultProp(quantumAssertion, "Quantum must be positive")
    ], AbstractTuiInputSlider.prototype, "quantum", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "keySteps", void 0);
    __decorate([
        Input(),
        HostBinding("attr.data-size"),
        tuiDefaultProp()
    ], AbstractTuiInputSlider.prototype, "size", void 0);
    __decorate([
        tuiPure
    ], AbstractTuiInputSlider.prototype, "getTickPrefix", null);
    __decorate([
        HostBinding("class._segmented")
    ], AbstractTuiInputSlider.prototype, "segmented", null);
    __decorate([
        tuiPure
    ], AbstractTuiInputSlider.prototype, "computePureKeySteps", null);
    AbstractTuiInputSlider = __decorate([
        Directive()
    ], AbstractTuiInputSlider);
    return AbstractTuiInputSlider;
}(AbstractTuiControl));
export { AbstractTuiInputSlider };
