// @taiga-ui/kit/esm5/components/tabs/tab/tab.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Inject, OnDestroy, Optional, } from '@angular/core';
import { RouterLinkActive } from '@angular/router';
import { isNativeFocused, setNativeFocused, TUI_IS_ANDROID, TUI_IS_IOS, TuiFocusVisibleService, } from '@taiga-ui/cdk';
import { TUI_MODE } from '@taiga-ui/core';
import { TUI_MOBILE_AWARE, TUI_TAB_MARGIN } from '@taiga-ui/kit/tokens';
import { Observable } from 'rxjs';
import { TUI_TAB_EVENT, TUI_TAB_PROVIDERS } from './tab.providers';
var TuiTabComponent = /** @class */ (function () {
    function TuiTabComponent(routerLinkActive, elementRef, mode$, event$, margin, mobileAware, isIos, isAndroid, focusVisible$) {
        var _this = this;
        this.routerLinkActive = routerLinkActive;
        this.elementRef = elementRef;
        this.mode$ = mode$;
        this.event$ = event$;
        this.margin = margin;
        this.focusVisible = false;
        this.isIos = mobileAware && isIos;
        this.isAndroid = mobileAware && isAndroid;
        focusVisible$.subscribe(function (visible) {
            _this.focusVisible = visible;
        });
    }
    Object.defineProperty(TuiTabComponent.prototype, "isActive", {
        get: function () {
            return !!this.routerLinkActive && this.routerLinkActive.isActive;
        },
        enumerable: true,
        configurable: true
    });
    TuiTabComponent.prototype.ngOnDestroy = function () {
        if (isNativeFocused(this.elementRef.nativeElement)) {
            setNativeFocused(this.elementRef.nativeElement, false);
        }
    };
    TuiTabComponent.ctorParameters = function () { return [
        { type: RouterLinkActive, decorators: [{ type: Optional }, { type: Inject, args: [RouterLinkActive,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_TAB_EVENT,] }] },
        { type: Number, decorators: [{ type: Inject, args: [TUI_TAB_MARGIN,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_MOBILE_AWARE,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_IOS,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_ANDROID,] }] },
        { type: TuiFocusVisibleService, decorators: [{ type: Inject, args: [TuiFocusVisibleService,] }] }
    ]; };
    __decorate([
        HostBinding('class._ios')
    ], TuiTabComponent.prototype, "isIos", void 0);
    __decorate([
        HostBinding('class._android')
    ], TuiTabComponent.prototype, "isAndroid", void 0);
    __decorate([
        HostBinding('class._focus-visible')
    ], TuiTabComponent.prototype, "focusVisible", void 0);
    __decorate([
        HostBinding('class._active')
    ], TuiTabComponent.prototype, "isActive", null);
    TuiTabComponent = __decorate([
        Component({
            selector: 'a[tuiTab]:not([routerLink]), a[tuiTab][routerLink][routerLinkActive], button[tuiTab]',
            template: "<ng-container *ngIf=\"event$ | async\"></ng-container>\n<span class=\"t-highlight\">\n    <ng-content></ng-content>\n</span>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: TUI_TAB_PROVIDERS,
            host: {
                '($.data-mode.attr)': 'mode$',
                '[style.--tui-tab-margin.px]': 'margin',
                type: 'button',
            },
            styles: [":host-context(._underline):hover{box-shadow:inset 0 -2px var(--tui-base-03)}:host-context([tuiTabs]>):first-child,:host-context([tuiTabs]>:first-child),:host-context(tui-tabs>):first-child,:host-context(tui-tabs>:first-child){margin-left:0}:host{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:0 0;font-size:inherit;line-height:inherit;position:relative;display:flex;flex-shrink:0;height:100%;box-sizing:border-box;justify-content:space-between;align-items:center;white-space:nowrap;cursor:pointer;outline:0;text-decoration:none;color:inherit;transition-property:color,box-shadow,opacity;margin-left:var(--tui-tab-margin,1.5rem)}:host:disabled{opacity:var(--tui-disabled-opacity);pointer-events:none}:host:not([data-mode])._active{opacity:1;color:var(--tui-text-01);box-shadow:none}:host[data-mode=onDark]{color:rgba(255,255,255,.72)}:host[data-mode=onDark]._active,:host[data-mode=onDark]:hover{color:#fff}:host[data-mode=onDark]._focus-visible .t-highlight{background:var(--tui-clear-inverse);color:var(--tui-text-01)}:host[data-mode=onLight]{color:var(--tui-text-02)}:host[data-mode=onLight]._active,:host[data-mode=onLight]:hover{color:var(--tui-text-01)}:host[data-mode=onLight]._focus-visible .t-highlight{background:var(--tui-text-01);color:#fff}:host._focus-visible .t-highlight{background:var(--tui-selection)}:host._android{transition-property:opacity;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;height:3rem;flex:1;flex-shrink:0;justify-content:center;margin:0;text-transform:uppercase;color:var(--tui-base-09);opacity:.5;font-size:.875rem;font-weight:500;letter-spacing:-.008125rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,'Helvetica Neue',sans-serif}:host._android:hover{box-shadow:none}:host._android._active{opacity:1;color:var(--tui-base-09)}:host._android .t-highlight{width:auto}:host._ios{height:1.75rem;flex:1;flex-shrink:0;justify-content:center;margin:0;color:var(--tui-base-01);opacity:1;font-size:.8125rem;font-weight:500;letter-spacing:-.005rem;font-family:-apple-system,BlinkMacSystemFont,Roboto,'Helvetica Neue',sans-serif}:host._ios:hover{box-shadow:none}:host._ios._active{font-size:.875rem;font-weight:600;color:var(--tui-link)}:host._ios .t-highlight{width:auto}@media (hover:hover){:host:hover{color:var(--tui-text-01)}}.t-highlight{display:inline-flex;width:100%;align-items:center;justify-content:inherit}.t-highlight ::ng-deep>*{flex-shrink:0}"]
        }),
        __param(0, Optional()),
        __param(0, Inject(RouterLinkActive)),
        __param(1, Inject(ElementRef)),
        __param(2, Inject(TUI_MODE)),
        __param(3, Inject(TUI_TAB_EVENT)),
        __param(4, Inject(TUI_TAB_MARGIN)),
        __param(5, Inject(TUI_MOBILE_AWARE)),
        __param(6, Inject(TUI_IS_IOS)),
        __param(7, Inject(TUI_IS_ANDROID)),
        __param(8, Inject(TuiFocusVisibleService))
    ], TuiTabComponent);
    return TuiTabComponent;
}());
export { TuiTabComponent };
