// @taiga-ui/core/esm2015/components/alert/alert.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { AbstractTuiDialogService, TuiIdService } from '@taiga-ui/cdk';
import { TUI_NOTIFICATION_OPTIONS } from '@taiga-ui/core/tokens';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiAlertComponent } from './alert.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core/tokens";
import * as i2 from "@taiga-ui/cdk/services";
let TuiAlertService = class TuiAlertService extends AbstractTuiDialogService {
    constructor(defaultOptions, idService) {
        super(idService);
        this.defaultOptions = defaultOptions;
        this.component = new PolymorpheusComponent(TuiAlertComponent);
    }
};
TuiAlertService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_NOTIFICATION_OPTIONS,] }] },
    { type: TuiIdService, decorators: [{ type: Inject, args: [TuiIdService,] }] }
];
TuiAlertService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiAlertService_Factory() { return new TuiAlertService(i0.ɵɵinject(i1.TUI_NOTIFICATION_OPTIONS), i0.ɵɵinject(i2.TuiIdService)); }, token: TuiAlertService, providedIn: "root" });
TuiAlertService = __decorate([
    Injectable({ providedIn: `root` }),
    __param(0, Inject(TUI_NOTIFICATION_OPTIONS)),
    __param(1, Inject(TuiIdService))
], TuiAlertService);
export { TuiAlertService };
/** @deprecated use {@link TuiAlertService} */
let TuiNotificationsService = class TuiNotificationsService {
    constructor(alert) {
        this.alert = alert;
    }
    show(content, options = {}) {
        return this.alert.open(content, options);
    }
};
TuiNotificationsService.ctorParameters = () => [
    { type: TuiAlertService, decorators: [{ type: Inject, args: [TuiAlertService,] }] }
];
TuiNotificationsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiNotificationsService_Factory() { return new TuiNotificationsService(i0.ɵɵinject(TuiAlertService)); }, token: TuiNotificationsService, providedIn: "root" });
TuiNotificationsService = __decorate([
    Injectable({ providedIn: `root` }),
    __param(0, Inject(TuiAlertService))
], TuiNotificationsService);
export { TuiNotificationsService };
