// @taiga-ui/core/esm5/components/dialog/dialog.providers.js
import { DOCUMENT } from '@angular/common';
import { ElementRef, InjectionToken } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { getActualTarget, isCurrentTarget, tuiContainsOrAfter, TuiDestroyService, tuiIsElement, typedFromEvent, } from '@taiga-ui/cdk';
import { tuiGetViewportWidth } from '@taiga-ui/core/utils/dom';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import { EMPTY, merge } from 'rxjs';
import { filter, switchMap, take, takeUntil } from 'rxjs/operators';
export var TUI_DIALOGS_CLOSE = new InjectionToken("A stream to close dialogs", {
    factory: function () { return EMPTY; },
});
var SCROLLBAR_PLACEHOLDER = 17;
// eslint-disable-next-line @typescript-eslint/naming-convention
export function dialogCloseStreamFactory(documentRef, windowRef, _a, close$, destroy$, _b) {
    var nativeElement = _a.nativeElement;
    var dismissible = _b.dismissible;
    return dismissible
        ? merge(typedFromEvent(nativeElement, "click").pipe(filter(isCurrentTarget)), typedFromEvent(documentRef, "keydown").pipe(filter(function (event) {
            var key = event.key;
            var target = getActualTarget(event);
            return (key === "Escape" &&
                tuiIsElement(target) &&
                (!tuiContainsOrAfter(nativeElement, target) ||
                    nativeElement.contains(target)));
        })), typedFromEvent(documentRef, "mousedown").pipe(filter(function (event) {
            var target = getActualTarget(event);
            var clientX = event.clientX;
            return (tuiIsElement(target) &&
                tuiGetViewportWidth(windowRef) - clientX >
                    SCROLLBAR_PLACEHOLDER &&
                !tuiContainsOrAfter(nativeElement, target));
        }), switchMap(function () {
            return typedFromEvent(documentRef, "mouseup").pipe(take(1), filter(function (event) {
                var target = getActualTarget(event);
                return (tuiIsElement(target) &&
                    !tuiContainsOrAfter(nativeElement, target));
            }));
        })), close$).pipe(takeUntil(destroy$))
        : close$;
}
export var TUI_DIALOG_CLOSE_STREAM = new InjectionToken("Dialogs closing stream");
export var TUI_DIALOG_PROVIDERS = [
    TuiDestroyService,
    {
        provide: TUI_DIALOG_CLOSE_STREAM,
        deps: [
            DOCUMENT,
            WINDOW,
            ElementRef,
            TUI_DIALOGS_CLOSE,
            TuiDestroyService,
            POLYMORPHEUS_CONTEXT,
        ],
        useFactory: dialogCloseStreamFactory,
    },
];
