// @taiga-ui/cdk/esm5/directives/pressed/pressed.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, Output } from '@angular/core';
import { pressedObservable } from '@taiga-ui/cdk/observables';
import { TUI_TAKE_ONLY_TRUSTED_EVENTS } from '@taiga-ui/cdk/tokens';
var TuiPressedDirective = /** @class */ (function () {
    function TuiPressedDirective(elementRef, takeOnlyTrustedEvents) {
        this.elementRef = elementRef;
        this.takeOnlyTrustedEvents = takeOnlyTrustedEvents;
        this.tuiPressedChange = pressedObservable(this.elementRef.nativeElement, {
            onlyTrusted: this.takeOnlyTrustedEvents,
        });
    }
    TuiPressedDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_TAKE_ONLY_TRUSTED_EVENTS,] }] }
    ]; };
    __decorate([
        Output()
    ], TuiPressedDirective.prototype, "tuiPressedChange", void 0);
    TuiPressedDirective = __decorate([
        Directive({
            selector: '[tuiPressedChange]',
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(TUI_TAKE_ONLY_TRUSTED_EVENTS))
    ], TuiPressedDirective);
    return TuiPressedDirective;
}());
export { TuiPressedDirective };
