// ngx-quicklink/esm2015/src/quicklink.module.js
import { NgModule } from '@angular/core';
import { LinkDirective } from './link.directive';
import { ObservableLinkHandler, PreloadLinkHandler, LinkHandler } from './link-handler.service';
import { PrefetchRegistry } from './prefetch-registry.service';
import { QuicklinkStrategy } from './quicklink-strategy.service';
export class QuicklinkModule {
}
QuicklinkModule.decorators = [
    { type: NgModule, args: [{
                declarations: [LinkDirective],
                providers: [
                    {
                        provide: LinkHandler,
                        useClass: ObservableLinkHandler,
                        multi: true
                    },
                    {
                        provide: LinkHandler,
                        useClass: PreloadLinkHandler,
                        multi: true
                    },
                    PrefetchRegistry,
                    QuicklinkStrategy
                ],
                exports: [LinkDirective]
            },] }
];
