// @taiga-ui/core/esm5/abstract/abstract-hint.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Input, OnDestroy } from '@angular/core';
import { TuiActiveZoneDirective, tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiHintService } from '@taiga-ui/core/services';
var AbstractTuiHint = /** @class */ (function () {
    function AbstractTuiHint(elementRef, hintService, activeZone, options) {
        this.elementRef = elementRef;
        this.hintService = hintService;
        this.activeZone = activeZone;
        this.options = options;
        this.mode = this.options.mode;
        this.direction = this.options.direction;
        this.content = "";
    }
    AbstractTuiHint.prototype.ngOnDestroy = function () {
        this.hideTooltip();
    };
    AbstractTuiHint.prototype.showTooltip = function () {
        this.hintService.add(this);
    };
    AbstractTuiHint.prototype.hideTooltip = function () {
        this.hintService.remove(this);
    };
    AbstractTuiHint.ctorParameters = function () { return [
        { type: ElementRef },
        { type: TuiHintService },
        { type: TuiActiveZoneDirective },
        { type: undefined }
    ]; };
    __decorate([
        Input("tuiHintMode"),
        tuiDefaultProp()
    ], AbstractTuiHint.prototype, "mode", void 0);
    __decorate([
        Input("tuiHintDirection"),
        tuiDefaultProp()
    ], AbstractTuiHint.prototype, "direction", void 0);
    AbstractTuiHint = __decorate([
        Directive()
    ], AbstractTuiHint);
    return AbstractTuiHint;
}());
export { AbstractTuiHint };
