// @taiga-ui/core/esm2015/components/hints-host/hints-host.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Inject } from '@angular/core';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk';
import { TuiHintService } from '@taiga-ui/core/services';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { Observable } from 'rxjs';
import { TuiHintBoxComponent } from './hint-box/hint-box.component';
let TuiHintsHostComponent = class TuiHintsHostComponent {
    constructor(elementRef, hints$) {
        this.elementRef = elementRef;
        this.hints$ = hints$;
        this.component = new PolymorpheusComponent(TuiHintBoxComponent);
    }
    get clientRect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
};
TuiHintsHostComponent.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TuiHintService,] }] }
];
TuiHintsHostComponent = __decorate([
    Component({
        selector: 'tui-hints-host',
        template: "<div\n    *ngFor=\"let hint of hints$ | async\"\n    role=\"tooltip\"\n    polymorpheus-outlet\n    @tuiParentAnimation\n    [attr.id]=\"hint.id\"\n    [content]=\"hint.component || component\"\n    [context]=\"{$implicit: hint}\"\n    [tuiActiveZoneParent]=\"hint.activeZone\"\n></div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        animations: [TUI_PARENT_ANIMATION],
        host: {
            'aria-live': 'polite',
        },
        styles: [":host{position:fixed;bottom:0;left:0;width:100%;height:0}"]
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(TuiHintService))
], TuiHintsHostComponent);
export { TuiHintsHostComponent };
