// @taiga-ui/addon-doc/esm5/components/code/code.component.js
import { __decorate } from "tslib";
import { Component, HostBinding, Input } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { rawLoad } from '../../utils/raw-load';
import { tryParseMarkdownCodeBlock } from './parse-code-block';
var TuiDocCodeComponent = /** @class */ (function () {
    function TuiDocCodeComponent() {
        this.rawLoader$$ = new BehaviorSubject('');
        this.filename = '';
        this.processor$ = this.rawLoader$$.pipe(switchMap(rawLoad), map(tryParseMarkdownCodeBlock));
    }
    Object.defineProperty(TuiDocCodeComponent.prototype, "code", {
        set: function (code) {
            this.rawLoader$$.next(code);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiDocCodeComponent.prototype, "hasFilename", {
        get: function () {
            return !!this.filename;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input()
    ], TuiDocCodeComponent.prototype, "filename", void 0);
    __decorate([
        Input()
    ], TuiDocCodeComponent.prototype, "code", null);
    __decorate([
        HostBinding('class._has-filename')
    ], TuiDocCodeComponent.prototype, "hasFilename", null);
    TuiDocCodeComponent = __decorate([
        Component({
            selector: 'tui-doc-code',
            template: "<p\n    *ngIf=\"filename\"\n    class=\"t-header\"\n>\n    {{ filename }}\n</p>\n<pre\n    *ngFor=\"let content of processor$ | async\"\n    class=\"t-code\"\n>\n<code [lineNumbers]=\"true\" [highlight]=\"content\"></code>\n</pre>\n",
            styles: [":host{display:block}.t-header{font:var(--tui-font-text-s);font-weight:700}.t-code{margin:0}.t-code+.t-code{margin-top:1rem}.hljs:not(:empty){font:var(--tui-font-text-m);padding:1.5rem 2rem;font-family:monospace;background:var(--tui-clear-disabled);word-wrap:break-word;white-space:pre-wrap}@media all and (-webkit-min-device-pixel-ratio:0) and (min-resolution:0.001dpcm){.hljs:not(:empty)::-webkit-scrollbar,.hljs:not(:empty)::-webkit-scrollbar-thumb{width:1rem;height:1rem;border-radius:6.25rem;background-clip:padding-box;border:2.667rem solid transparent}.hljs:not(:empty)::-webkit-scrollbar{background-color:transparent}.hljs:not(:empty)::-webkit-scrollbar-thumb{background-color:var(--tui-clear-hover)}.hljs:not(:empty)::-webkit-scrollbar-thumb:hover{background-color:var(--tui-clear-active)}.hljs:not(:empty)::-webkit-scrollbar-thumb:active{background-color:var(--tui-text-03)}}:host._has-filename .hljs:not(:empty){border-radius:.75rem}:host-context(tui-root._mobile) .hljs:not(:empty){padding:1rem}"]
        })
    ], TuiDocCodeComponent);
    return TuiDocCodeComponent;
}());
export { TuiDocCodeComponent };
