// @tinkoff/ng-event-plugins/esm5/plugins/options.plugin.js
/**
 * @fileoverview added by tsickle
 * Generated from: plugins/options.plugin.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __extends } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractEventPlugin } from './abstract.plugin';
var OptionsEventPlugin = /** @class */ (function (_super) {
    __extends(OptionsEventPlugin, _super);
    function OptionsEventPlugin() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.modifier = 'capture.once.passive';
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    OptionsEventPlugin.prototype.supports = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return event.includes('.') && !this.unwrap(event).includes('.');
    };
    /**
     * @param {?} element
     * @param {?} event
     * @param {?} handler
     * @return {?}
     */
    OptionsEventPlugin.prototype.addEventListener = /**
     * @param {?} element
     * @param {?} event
     * @param {?} handler
     * @return {?}
     */
    function (element, event, handler) {
        var _this = this;
        element.addEventListener(this.unwrap(event), handler, {
            once: event.includes('.once'),
            passive: event.includes('.passive'),
            capture: event.includes('.capture'),
        });
        return (/**
         * @return {?}
         */
        function () { return element.removeEventListener(_this.unwrap(event), handler); });
    };
    OptionsEventPlugin.decorators = [
        { type: Injectable }
    ];
    return OptionsEventPlugin;
}(AbstractEventPlugin));
export { OptionsEventPlugin };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    OptionsEventPlugin.prototype.modifier;
}
