// ngx-acuw/esm2015/lib/lightbox/lightbox-overlay.component.js
import { ChangeDetectorRef, Component, HostListener, ViewChild } from '@angular/core';
import { ImageTransitionComponent } from '../image-transition/image-transition.component';
export class LightboxOverlayComponent {
    constructor(changeRef) {
        this.changeRef = changeRef;
        this.imageUrls = new Array();
        this.forceFullscreen = false;
        this.forceShowNavButtons = false;
        this.imageSize = 'cover';
        this.autoPlay = false;
        this.autoPlayInterval = 5000;
        this.transitionDuration = 1000;
        this.transitionType = 'split';
        this.sizeX = 50.0;
        this.sizeY = 50.0;
        this.width = 0.5;
        this.intensity = 40.0;
        this.startIndex = 0;
        this.currentImageIndex = 1;
        this.swipeCoord = new Array();
        this.swipeTime = 0;
    }
    ngOnInit() {
    }
    keyEvent(event) {
        if (event.key === 'ArrowRight') {
            this.imageTransition.next();
        }
        if (event.key === 'ArrowLeft') {
            this.imageTransition.prev();
        }
        if (event.key === 'Escape') {
            this.close();
        }
    }
    /**
     * Close the lightbox
     */
    close() {
        var _a;
        (_a = this.overlayRef) === null || _a === void 0 ? void 0 : _a.detach();
    }
    /**
     * method to set the index counter
     * @param index index of the image
     */
    imageIndexChange(index) {
        this.currentImageIndex = index + 1;
        this.changeRef.detectChanges();
    }
    /**
     * Listen to touche events for gestures (mobile)
     * @param e touch event
     * @param when indicator if 'start' or 'end'
     */
    swipe(e, when) {
        const coord = [e.changedTouches[0].clientX, e.changedTouches[0].clientY];
        const time = new Date().getTime();
        if (when === 'start') {
            this.swipeCoord = coord;
            this.swipeTime = time;
        }
        else if (when === 'end') {
            const direction = [coord[0] - this.swipeCoord[0], coord[1] - this.swipeCoord[1]];
            const duration = time - this.swipeTime;
            if (duration < 1000 && Math.abs(direction[0]) > 30 // Long enough
                && Math.abs(direction[0]) > Math.abs(direction[1] * 3)) { // Horizontal enough
                const swipe = direction[0] < 0 ? 'next' : 'previous';
                if (swipe === 'next') {
                    this.imageTransition.next();
                }
                else if (swipe === 'previous') {
                    this.imageTransition.prev();
                }
            }
        }
    }
}
LightboxOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-lightbox-overlay',
                template: "<div [ngClass]=\"{'full-container': forceFullscreen === true, 'auto-container': forceFullscreen === false }\">\n    <div class=\"index-indicator\">\n        <span>{{ currentImageIndex }}</span>\n        <span>/</span>\n        <span>{{ imageUrls.length }}</span>\n    </div>\n    <svg class=\"ligthbox-button close-button\" (click)=\"close()\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"white\">\n        <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n        <path d=\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\" />\n    </svg>\n\n    <svg [ngClass]=\"{'auto-visibility': forceShowNavButtons === false}\" \n        class=\"ligthbox-button prev-button\" (click)=\"imageTransition.prev()\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"white\">\n        <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n        <path d=\"M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z\" />\n    </svg>\n\n    <svg [ngClass]=\"{'auto-visibility': forceShowNavButtons === false}\" \n        class=\"ligthbox-button next-button\" (click)=\"imageTransition.next()\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"white\">\n        <path d=\"M0 0h24v24H0z\" fill=\"none\" />\n        <path d=\"M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z\" />\n    </svg>\n    <lib-image-transition #imageTransition [imageUrls]=\"imageUrls\" [startIndex]=\"startIndex\" (imageIndexChange)=\"imageIndexChange($event)\"\n        [imageSize]=\"imageSize\" [autoPlay]=\"autoPlay\" [autoPlayInterval]=\"autoPlayInterval\" [transitionDuration]=\"transitionDuration\"\n        [transitionType]=\"transitionType\" [sizeX]=\"sizeX\" [sizeY]=\"sizeY\" [width]=\"width\" [intensity]=\"intensity\"\n        (touchstart)=\"swipe($event, 'start')\" (touchend)=\"swipe($event, 'end')\">\n    </lib-image-transition>\n</div>",
                styles: [".auto-container{width:1100px;height:800px;max-width:90vw;max-height:90vh}.auto-container,.full-container{position:relative;background-color:#000;box-shadow:0 0 15px 10px rgba(0,0,0,.5)}.full-container{width:100vw;height:100vh;max-width:100vw;max-height:100vh}.index-indicator{color:#fff;border-radius:0 0 5px 0;padding:8px}.index-indicator,.ligthbox-button{position:absolute;z-index:1;background-color:rgba(0,0,0,.4)}.ligthbox-button{cursor:pointer;width:36px;height:36px}.ligthbox-button:hover{fill:hsla(0,0%,100%,.5)}.close-button{top:0;right:0;border-radius:0 0 0 5px}.prev-button{top:calc(50% - 18px);left:0;border-radius:0 5px 5px 0}.next-button{top:calc(50% - 18px);right:0;border-radius:5px 0 0 5px}@media (max-width:600px){.auto-container{width:100vw;height:100vh;max-width:100vw;max-height:100vh}.auto-visibility{visibility:collapse}}lib-image-transition{touch-action:none}"]
            },] }
];
LightboxOverlayComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
LightboxOverlayComponent.propDecorators = {
    imageTransition: [{ type: ViewChild, args: [ImageTransitionComponent,] }],
    keyEvent: [{ type: HostListener, args: ['window:keyup', ['$event'],] }]
};
