// ngx-acuw/esm2015/lib/lightbox/lightbox.component.js
import { Overlay } from '@angular/cdk/overlay';
import { ComponentPortal } from '@angular/cdk/portal';
import { Component, Input } from '@angular/core';
import { LightboxOverlayComponent } from './lightbox-overlay.component';
export class LightboxComponent {
    constructor(overlay) {
        this.overlay = overlay;
        this.imageUrls = new Array();
        this.forceFullscreen = false;
        this.forceShowNavButtons = false;
        this.imageSize = 'cover';
        this.autoPlay = false;
        this.autoPlayInterval = 5000;
        this.transitionDuration = 1000;
        this.transitionType = 'split';
        this.sizeX = 50.0;
        this.sizeY = 50.0;
        this.width = 0.5;
        this.intensity = 40.0;
    }
    ngOnInit() {
    }
    open(index = 0) {
        // Define settings of the overlay
        this.overlayRef = this.overlay.create({
            hasBackdrop: true,
            positionStrategy: this.overlay.position().global().centerHorizontally().centerVertically(),
            scrollStrategy: this.overlay.scrollStrategies.block()
        });
        // Listen to backdrop event for detaching the overlay
        this.overlayRef.backdropClick().subscribe({
            next: () => { var _a; (_a = this.overlayRef) === null || _a === void 0 ? void 0 : _a.detach(); }
        });
        // Attach the LightboxOverlayComponent to the overlayRef instance
        const lightboxOverlayRef = this.overlayRef.attach(new ComponentPortal(LightboxOverlayComponent));
        // Pass data to the LightboxOverlayComponent
        lightboxOverlayRef.instance.overlayRef = this.overlayRef;
        lightboxOverlayRef.instance.imageUrls = this.imageUrls;
        lightboxOverlayRef.instance.forceFullscreen = this.forceFullscreen;
        lightboxOverlayRef.instance.forceShowNavButtons = this.forceShowNavButtons;
        lightboxOverlayRef.instance.imageSize = this.imageSize;
        lightboxOverlayRef.instance.autoPlay = this.autoPlay;
        lightboxOverlayRef.instance.autoPlayInterval = this.autoPlayInterval;
        lightboxOverlayRef.instance.transitionDuration = this.transitionDuration;
        lightboxOverlayRef.instance.transitionType = this.transitionType;
        lightboxOverlayRef.instance.sizeX = this.sizeX;
        lightboxOverlayRef.instance.sizeY = this.sizeY;
        lightboxOverlayRef.instance.width = this.width;
        lightboxOverlayRef.instance.intensity = this.intensity;
        lightboxOverlayRef.instance.startIndex = index;
    }
    close() {
        var _a;
        (_a = this.overlayRef) === null || _a === void 0 ? void 0 : _a.detach();
    }
}
LightboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-lightbox',
                template: ''
            },] }
];
LightboxComponent.ctorParameters = () => [
    { type: Overlay }
];
LightboxComponent.propDecorators = {
    imageUrls: [{ type: Input }],
    forceFullscreen: [{ type: Input }],
    forceShowNavButtons: [{ type: Input }],
    imageSize: [{ type: Input }],
    autoPlay: [{ type: Input }],
    autoPlayInterval: [{ type: Input }],
    transitionDuration: [{ type: Input }],
    transitionType: [{ type: Input }],
    sizeX: [{ type: Input }],
    sizeY: [{ type: Input }],
    width: [{ type: Input }],
    intensity: [{ type: Input }]
};
