// @taiga-ui/addon-doc/esm5/components/documentation/documentation-property-connector.directive.js
import { __assign, __decorate, __param } from "tslib";
import { Location } from '@angular/common';
import { Directive, EventEmitter, Inject, Input, OnChanges, OnInit, Output, TemplateRef, } from '@angular/core';
import { ActivatedRoute, Params, UrlSerializer } from '@angular/router';
import { BehaviorSubject, Subject } from 'rxjs';
import { coerceValue } from '../../utils/coerce-value';
var SERIALIZED_SUFFIX = '$';
// @bad TODO: refactor output and value sync
var TuiDocDocumentationPropertyConnectorDirective = /** @class */ (function () {
    function TuiDocDocumentationPropertyConnectorDirective(template, locationRef, activatedRoute, urlSerializer) {
        this.template = template;
        this.locationRef = locationRef;
        this.activatedRoute = activatedRoute;
        this.urlSerializer = urlSerializer;
        this.documentationPropertyName = '';
        this.documentationPropertyMode = null;
        this.documentationPropertyType = '';
        this.documentationPropertyDeprecated = false;
        this.documentationPropertyValues = null;
        this.documentationPropertyValueChange = new EventEmitter();
        this.changed$ = new Subject();
        this.emits$ = new BehaviorSubject(1);
    }
    TuiDocDocumentationPropertyConnectorDirective.prototype.ngOnInit = function () {
        this.parseParams(this.activatedRoute.snapshot.queryParams);
    };
    Object.defineProperty(TuiDocDocumentationPropertyConnectorDirective.prototype, "attrName", {
        get: function () {
            switch (this.documentationPropertyMode) {
                case 'input':
                    return "[" + this.documentationPropertyName + "]";
                case 'output':
                    return "(" + this.documentationPropertyName + ")";
                case 'input-output':
                    return "[(" + this.documentationPropertyName + ")]";
                default:
                    return this.documentationPropertyName;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiDocDocumentationPropertyConnectorDirective.prototype, "hasItems", {
        get: function () {
            return !!this.documentationPropertyValues;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiDocDocumentationPropertyConnectorDirective.prototype, "shouldShowValues", {
        get: function () {
            return this.documentationPropertyMode !== 'output';
        },
        enumerable: true,
        configurable: true
    });
    TuiDocDocumentationPropertyConnectorDirective.prototype.ngOnChanges = function () {
        this.changed$.next();
    };
    TuiDocDocumentationPropertyConnectorDirective.prototype.onValueChange = function (value) {
        this.documentationPropertyValue = value;
        this.documentationPropertyValueChange.emit(value);
        this.setQueryParam(value);
    };
    TuiDocDocumentationPropertyConnectorDirective.prototype.emitEvent = function (event) {
        // For more convenient debugging
        console.info(this.attrName, event);
        this.emits$.next(this.emits$.value + 1);
    };
    TuiDocDocumentationPropertyConnectorDirective.prototype.parseParams = function (params) {
        var propertyValue = params[this.documentationPropertyName];
        var propertyValueWithSuffix = params["" + this.documentationPropertyName + SERIALIZED_SUFFIX];
        if (!propertyValue && !propertyValueWithSuffix) {
            return;
        }
        var value = !!propertyValueWithSuffix && this.documentationPropertyValues
            ? this.documentationPropertyValues[propertyValueWithSuffix]
            : coerceValue(propertyValue);
        this.onValueChange(value);
    };
    TuiDocDocumentationPropertyConnectorDirective.prototype.setQueryParam = function (value) {
        var _a;
        var tree = this.urlSerializer.parse(this.locationRef.path());
        var isValueAvailableByKey = value instanceof Object;
        var computedValue = isValueAvailableByKey && this.documentationPropertyValues
            ? this.documentationPropertyValues.indexOf(value)
            : value;
        var suffix = isValueAvailableByKey ? SERIALIZED_SUFFIX : '';
        var propName = this.documentationPropertyName + suffix;
        tree.queryParams = __assign(__assign({}, tree.queryParams), (_a = {}, _a[propName] = computedValue, _a));
        this.locationRef.go(String(tree));
    };
    TuiDocDocumentationPropertyConnectorDirective.ctorParameters = function () { return [
        { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }] },
        { type: Location, decorators: [{ type: Inject, args: [Location,] }] },
        { type: ActivatedRoute, decorators: [{ type: Inject, args: [ActivatedRoute,] }] },
        { type: UrlSerializer, decorators: [{ type: Inject, args: [UrlSerializer,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyName", void 0);
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyMode", void 0);
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyType", void 0);
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyValue", void 0);
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyDeprecated", void 0);
    __decorate([
        Input()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyValues", void 0);
    __decorate([
        Output()
    ], TuiDocDocumentationPropertyConnectorDirective.prototype, "documentationPropertyValueChange", void 0);
    TuiDocDocumentationPropertyConnectorDirective = __decorate([
        Directive({
            selector: 'ng-template[documentationPropertyName]',
            exportAs: 'documentationProperty',
        }),
        __param(0, Inject(TemplateRef)),
        __param(1, Inject(Location)),
        __param(2, Inject(ActivatedRoute)),
        __param(3, Inject(UrlSerializer))
    ], TuiDocDocumentationPropertyConnectorDirective);
    return TuiDocDocumentationPropertyConnectorDirective;
}());
export { TuiDocDocumentationPropertyConnectorDirective };
