// @taiga-ui/kit/esm2015/components/tree/misc/tree.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { map, mapTo, startWith, switchMap, tap } from 'rxjs/operators';
import { TUI_TREE_LOADER, TUI_TREE_LOADING, TUI_TREE_START } from './tree.tokens';
// @dynamic
let TuiTreeService = class TuiTreeService {
    constructor(loading, start, loader) {
        this.loading = loading;
        this.start = start;
        this.loader = loader;
        this.map = new Map([[this.loading, []]]);
        this.load$ = new Subject();
        this.data$ = this.load$.pipe(switchMap(item => this.loader.loadChildren(item).pipe(tap(children => this.map.set(item, children)), map(children => children.filter(item => !this.loader.hasChildren(item))), tap(children => children.forEach(child => this.map.set(child, []))))), startWith(null), mapTo(this.start));
    }
    getChildren(item) {
        var _a;
        return (_a = this.map.get(item)) !== null && _a !== void 0 ? _a : [this.loading];
    }
    loadChildren(item) {
        if (this.map.get(item)) {
            return;
        }
        this.map.set(item, [this.loading]);
        this.load$.next(item);
    }
};
TuiTreeService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_LOADING,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_START,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_LOADER,] }] }
];
TuiTreeService = __decorate([
    Injectable(),
    __param(0, Inject(TUI_TREE_LOADING)),
    __param(1, Inject(TUI_TREE_START)),
    __param(2, Inject(TUI_TREE_LOADER))
], TuiTreeService);
export { TuiTreeService };
