// @taiga-ui/kit/esm2015/components/tree/misc/tree.tokens.js
import { InjectionToken } from '@angular/core';
import { TUI_DEFAULT_TREE_CONTROLLER, TUI_TREE_ITEM_CONTENT } from './tree.constants';
export const TUI_TREE_ACCESSOR = new InjectionToken(`Controller for tracking value - TuiTreeItemComponent pairs`);
export const TUI_TREE_CONTROLLER = new InjectionToken(`Controller for expanding the tree`, {
    factory: () => TUI_DEFAULT_TREE_CONTROLLER,
});
export const TUI_TREE_NODE = new InjectionToken(`A node of a tree view`);
export const TUI_TREE_LOADING = new InjectionToken(`A tree node placeholder for loading`, { factory: () => ({}) });
export const TUI_TREE_START = new InjectionToken(`A tree node starting point`);
export const TUI_TREE_LOADER = new InjectionToken(`A service to load tree progressively`);
export const TUI_TREE_CONTENT = new InjectionToken(`Content for a tree item`, {
    factory: () => TUI_TREE_ITEM_CONTENT,
});
export const TUI_TREE_LEVEL = new InjectionToken(`Nesting level of current TreeView node`, {
    factory: () => -1,
});
