// @tinkoff/ng-polymorpheus/esm5/directives/template.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, Inject, Self, TemplateRef } from '@angular/core';
/**
 * ng-template wrapper directive also stores {\@link ChangeDetectorRef} to properly handle change detection.
 * @template C
 */
var PolymorpheusTemplate = /** @class */ (function () {
    function PolymorpheusTemplate(template, changeDetectorRef) {
        this.template = template;
        this.changeDetectorRef = changeDetectorRef;
        this.polymorpheus = '';
    }
    /**
     * @return {?}
     */
    PolymorpheusTemplate.prototype.check = /**
     * @return {?}
     */
    function () {
        this.changeDetectorRef.markForCheck();
    };
    /**
     * @template T
     * @param {?} _dir
     * @param {?} _ctx
     * @return {?}
     */
    PolymorpheusTemplate.ngTemplateContextGuard = /**
     * @template T
     * @param {?} _dir
     * @param {?} _ctx
     * @return {?}
     */
    function (_dir, _ctx) {
        return true;
    };
    PolymorpheusTemplate.decorators = [
        { type: Directive, args: [{
                    selector: 'ng-template[polymorpheus]',
                    exportAs: 'polymorpheus',
                    inputs: ['polymorpheus'],
                },] }
    ];
    /** @nocollapse */
    PolymorpheusTemplate.ctorParameters = function () { return [
        { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }, { type: Self }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] }
    ]; };
    return PolymorpheusTemplate;
}());
export { PolymorpheusTemplate };
if (false) {
    /** @type {?} */
    PolymorpheusTemplate.prototype.polymorpheus;
    /** @type {?} */
    PolymorpheusTemplate.prototype.template;
    /**
     * @type {?}
     * @private
     */
    PolymorpheusTemplate.prototype.changeDetectorRef;
}
