// @taiga-ui/kit/esm5/components/tree/misc/tree.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { map, mapTo, startWith, switchMap, tap } from 'rxjs/operators';
import { TUI_TREE_LOADER, TUI_TREE_LOADING, TUI_TREE_START } from './tree.tokens';
// @dynamic
var TuiTreeService = /** @class */ (function () {
    function TuiTreeService(loading, start, loader) {
        var _this = this;
        this.loading = loading;
        this.start = start;
        this.loader = loader;
        this.map = new Map([[this.loading, []]]);
        this.load$ = new Subject();
        this.data$ = this.load$.pipe(switchMap(function (item) {
            return _this.loader.loadChildren(item).pipe(tap(function (children) { return _this.map.set(item, children); }), map(function (children) { return children.filter(function (item) { return !_this.loader.hasChildren(item); }); }), tap(function (children) { return children.forEach(function (child) { return _this.map.set(child, []); }); }));
        }), startWith(null), mapTo(this.start));
    }
    TuiTreeService.prototype.getChildren = function (item) {
        var _a;
        return (_a = this.map.get(item)) !== null && _a !== void 0 ? _a : [this.loading];
    };
    TuiTreeService.prototype.loadChildren = function (item) {
        if (this.map.get(item)) {
            return;
        }
        this.map.set(item, [this.loading]);
        this.load$.next(item);
    };
    TuiTreeService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_LOADING,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_START,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_TREE_LOADER,] }] }
    ]; };
    TuiTreeService = __decorate([
        Injectable(),
        __param(0, Inject(TUI_TREE_LOADING)),
        __param(1, Inject(TUI_TREE_START)),
        __param(2, Inject(TUI_TREE_LOADER))
    ], TuiTreeService);
    return TuiTreeService;
}());
export { TuiTreeService };
