// @taiga-ui/core/bundles/taiga-ui-core-enums.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/enums', ['exports'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.enums = {})));
}(this, (function (exports) { 'use strict';

    (function (TuiAppearance) {
        TuiAppearance["Icon"] = "icon";
        TuiAppearance["Primary"] = "primary";
        TuiAppearance["Accent"] = "accent";
        TuiAppearance["Secondary"] = "secondary";
        TuiAppearance["SecondaryDestructive"] = "secondary-destructive";
        TuiAppearance["Table"] = "table";
        TuiAppearance["Textfield"] = "textfield";
        TuiAppearance["Flat"] = "flat";
        TuiAppearance["Outline"] = "outline";
        TuiAppearance["Whiteblock"] = "whiteblock";
        TuiAppearance["WhiteblockActive"] = "whiteblock-active";
    })(exports.TuiAppearance || (exports.TuiAppearance = {}));

    /**
     * @deprecated
     */

    (function (TuiBaseColor) {
        TuiBaseColor["Primary"] = "primary";
        TuiBaseColor["Secondary"] = "secondary";
        TuiBaseColor["Success"] = "success";
        TuiBaseColor["Error"] = "error";
    })(exports.TuiBaseColor || (exports.TuiBaseColor = {}));

    /**
     * @deprecated use string values
     */

    (function (TuiButtonShape) {
        TuiButtonShape["Square"] = "square";
        TuiButtonShape["Rounded"] = "rounded";
    })(exports.TuiButtonShape || (exports.TuiButtonShape = {}));

    /**
     * @deprecated use join type {@link TuiDecimalT}
     */

    (function (TuiDecimal) {
        TuiDecimal["NotZero"] = "not-zero";
        TuiDecimal["Always"] = "always";
        TuiDecimal["Never"] = "never";
    })(exports.TuiDecimal || (exports.TuiDecimal = {}));


    (function (TuiDropdownAnimation) {
        TuiDropdownAnimation["FadeInBottom"] = "fadeInBottom";
        TuiDropdownAnimation["FadeInTop"] = "fadeInTop";
    })(exports.TuiDropdownAnimation || (exports.TuiDropdownAnimation = {}));

    /**
     * @deprecated use join type {@link TuiDropdownWidthT}
     */

    (function (TuiDropdownWidth) {
        TuiDropdownWidth["Fixed"] = "fixed";
        TuiDropdownWidth["Min"] = "min";
        TuiDropdownWidth["Auto"] = "auto";
    })(exports.TuiDropdownWidth || (exports.TuiDropdownWidth = {}));

    /**
     * @deprecated use join type {@link TuiHintModeT}
     */

    (function (TuiHintMode) {
        TuiHintMode["Error"] = "error";
        TuiHintMode["OnDark"] = "onDark";
        TuiHintMode["Overflow"] = "overflow";
    })(exports.TuiHintMode || (exports.TuiHintMode = {}));

    /**
     * @internal used in calendar, year picker and wrapper
     */

    (function (TuiInteractiveState) {
        TuiInteractiveState["Disabled"] = "disabled";
        TuiInteractiveState["Readonly"] = "readonly";
        // TODO: 3.0 Rename to `active`
        TuiInteractiveState["Pressed"] = "pressed";
        // TODO: 3.0 Rename to `hover`
        TuiInteractiveState["Hovered"] = "hovered";
    })(exports.TuiInteractiveState || (exports.TuiInteractiveState = {}));

    /**
     * @deprecated use string values
     */

    (function (TuiLinkMode) {
        TuiLinkMode["Positive"] = "positive";
        TuiLinkMode["Negative"] = "negative";
    })(exports.TuiLinkMode || (exports.TuiLinkMode = {}));


    (function (TuiNotification) {
        TuiNotification["Info"] = "info";
        TuiNotification["Warning"] = "warning";
        TuiNotification["Success"] = "success";
        TuiNotification["Error"] = "error";
    })(exports.TuiNotification || (exports.TuiNotification = {}));

    /**
     * @deprecated use join type {@link TuiOrientationT}
     */

    (function (TuiOrientation) {
        TuiOrientation["Vertical"] = "vertical";
        TuiOrientation["Horizontal"] = "horizontal";
    })(exports.TuiOrientation || (exports.TuiOrientation = {}));

    /**
     * @internal used in calendar and year picker
     */

    (function (TuiRangeState) {
        TuiRangeState["Start"] = "start";
        TuiRangeState["End"] = "end";
        TuiRangeState["Single"] = "single";
    })(exports.TuiRangeState || (exports.TuiRangeState = {}));

    /**
     * @deprecated
     */

    (function (TuiSupportColor) {
        TuiSupportColor["Mustard"] = "support-01";
        TuiSupportColor["Texas"] = "support-02";
        TuiSupportColor["Tan"] = "support-03";
        TuiSupportColor["Salmon"] = "support-04";
        TuiSupportColor["Sienna"] = "support-05";
        TuiSupportColor["Bittersweet"] = "support-06";
        TuiSupportColor["Pinkie"] = "support-07";
        TuiSupportColor["Charm"] = "support-08";
        TuiSupportColor["Amethist"] = "support-09";
        TuiSupportColor["Helio"] = "support-10";
        TuiSupportColor["Lilac"] = "support-11";
        TuiSupportColor["Malibu"] = "support-12";
        TuiSupportColor["Havelock"] = "support-13";
        TuiSupportColor["Picton"] = "support-14";
        TuiSupportColor["Mint"] = "support-15";
        TuiSupportColor["Fountain"] = "support-16";
        TuiSupportColor["Puertorico"] = "support-17";
        TuiSupportColor["Bay"] = "support-18";
        TuiSupportColor["Forest"] = "support-19";
        TuiSupportColor["York"] = "support-20";
        TuiSupportColor["Feijoa"] = "support-21";
    })(exports.TuiSupportColor || (exports.TuiSupportColor = {}));

    Object.defineProperty(exports, '__esModule', { value: true });

})));
