// @taiga-ui/core/bundles/taiga-ui-core-components-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/tokens'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('@angular/common'), require('@ng-web-apis/common'), require('@taiga-ui/core/utils/dom'), require('rxjs/operators'), require('@taiga-ui/cdk/services'), require('@taiga-ui/core/components/button')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/animations', '@taiga-ui/core/tokens', '@tinkoff/ng-polymorpheus', 'rxjs', '@angular/common', '@ng-web-apis/common', '@taiga-ui/core/utils/dom', 'rxjs/operators', '@taiga-ui/cdk/services', '@taiga-ui/core/components/button'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components.dialog = {}), global.ng.core, global.cdk, global['taiga-ui'].core.animations, global['taiga-ui'].core.tokens, global.ngPolymorpheus, global.rxjs, global.ng.common, global.common$1, global['taiga-ui'].core.utils.dom, global.rxjs.operators, global.services, global['taiga-ui'].core.components.button));
}(this, (function (exports, core, cdk, animations, tokens, ngPolymorpheus, rxjs, common, common$1, dom, operators, services, button) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_DIALOGS_CLOSE = new core.InjectionToken("A stream to close dialogs", {
        factory: function () { return rxjs.EMPTY; },
    });
    var SCROLLBAR_PLACEHOLDER = 17;
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function dialogCloseStreamFactory(documentRef, windowRef, _a, close$, destroy$, _b) {
        var nativeElement = _a.nativeElement;
        var dismissible = _b.dismissible;
        return dismissible
            ? rxjs.merge(cdk.typedFromEvent(nativeElement, "click").pipe(operators.filter(cdk.isCurrentTarget)), cdk.typedFromEvent(documentRef, "keydown").pipe(operators.filter(function (event) {
                var key = event.key;
                var target = cdk.getActualTarget(event);
                return (key === "Escape" &&
                    cdk.tuiIsElement(target) &&
                    (!cdk.tuiContainsOrAfter(nativeElement, target) ||
                        nativeElement.contains(target)));
            })), cdk.typedFromEvent(documentRef, "mousedown").pipe(operators.filter(function (event) {
                var target = cdk.getActualTarget(event);
                var clientX = event.clientX;
                return (cdk.tuiIsElement(target) &&
                    dom.tuiGetViewportWidth(windowRef) - clientX >
                        SCROLLBAR_PLACEHOLDER &&
                    !cdk.tuiContainsOrAfter(nativeElement, target));
            }), operators.switchMap(function () {
                return cdk.typedFromEvent(documentRef, "mouseup").pipe(operators.take(1), operators.filter(function (event) {
                    var target = cdk.getActualTarget(event);
                    return (cdk.tuiIsElement(target) &&
                        !cdk.tuiContainsOrAfter(nativeElement, target));
                }));
            })), close$).pipe(operators.takeUntil(destroy$))
            : close$;
    }
    var TUI_DIALOG_CLOSE_STREAM = new core.InjectionToken("Dialogs closing stream");
    var TUI_DIALOG_PROVIDERS = [
        cdk.TuiDestroyService,
        {
            provide: TUI_DIALOG_CLOSE_STREAM,
            deps: [
                common.DOCUMENT,
                common$1.WINDOW,
                core.ElementRef,
                TUI_DIALOGS_CLOSE,
                cdk.TuiDestroyService,
                ngPolymorpheus.POLYMORPHEUS_CONTEXT,
            ],
            useFactory: dialogCloseStreamFactory,
        },
    ];

    var REQUIRED_ERROR = new Error('Required dialog was dismissed');
    // @dynamic
    var TuiDialogComponent = /** @class */ (function () {
        function TuiDialogComponent(duration, isMobile, context, close$, closeWord$) {
            var _this = this;
            this.duration = duration;
            this.isMobile = isMobile;
            this.context = context;
            this.closeWord$ = closeWord$;
            this.animation = {
                value: '',
                params: {
                    start: '40px',
                    duration: this.duration,
                },
            };
            this.fullscreenAnimation = {
                value: '',
                params: {
                    start: '100vh',
                    duration: this.duration,
                },
            };
            close$.subscribe(function () {
                _this.close();
            });
        }
        Object.defineProperty(TuiDialogComponent.prototype, "size", {
            get: function () {
                return this.context.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "header", {
            get: function () {
                return this.context.header;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "slideInTop", {
            get: function () {
                return this.size === 'fullscreen' || this.size === 'page' || this.isMobile
                    ? this.fullscreenAnimation
                    : this.animation;
            },
            enumerable: true,
            configurable: true
        });
        TuiDialogComponent.prototype.close = function () {
            if (this.context.required) {
                this.context.$implicit.error(REQUIRED_ERROR);
            }
            else {
                this.context.$implicit.complete();
            }
        };
        TuiDialogComponent.ctorParameters = function () { return [
            { type: Number, decorators: [{ type: core.Inject, args: [tokens.TUI_ANIMATIONS_DURATION,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [ngPolymorpheus.POLYMORPHEUS_CONTEXT,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [TUI_DIALOG_CLOSE_STREAM,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_CLOSE_WORD,] }] }
        ]; };
        __decorate([
            core.HostBinding('attr.data-size')
        ], TuiDialogComponent.prototype, "size", null);
        __decorate([
            core.HostBinding('class._centered')
        ], TuiDialogComponent.prototype, "header", null);
        __decorate([
            core.HostBinding('@tuiSlideInTop'),
            core.HostBinding('@tuiFadeIn')
        ], TuiDialogComponent.prototype, "slideInTop", null);
        TuiDialogComponent = __decorate([
            core.Component({
                selector: 'tui-dialog',
                template: "<header\n    *ngIf=\"header\"\n    polymorpheus-outlet\n    class=\"t-header\"\n    [content]=\"header\"\n    [context]=\"context\"\n></header>\n<div class=\"t-content\">\n    <h2\n        class=\"t-heading\"\n        [id]=\"context.id\"\n        [textContent]=\"context.label\"\n    ></h2>\n    <section\n        polymorpheus-outlet\n        [context]=\"context\"\n        [content]=\"context.content\"\n    >\n        <ng-template let-text>\n            {{ text }}\n            <div class=\"t-buttons\">\n                <button\n                    tuiButton\n                    size=\"m\"\n                    (click)=\"context.$implicit.complete()\"\n                >\n                    {{ $any(context.data)?.button || 'OK' }}\n                </button>\n            </div>\n        </ng-template>\n    </section>\n</div>\n<div\n    *ngIf=\"context.closeable\"\n    class=\"t-wrapper\"\n>\n    <button\n        tuiIconButton\n        tuiPreventDefault=\"mousedown\"\n        automation-id=\"tui-dialog__close\"\n        size=\"s\"\n        appearance=\"\"\n        icon=\"tuiIconCloseLarge\"\n        shape=\"rounded\"\n        class=\"t-close\"\n        [title]=\"closeWord$ | async\"\n        (click)=\"close()\"\n    ></button>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: TUI_DIALOG_PROVIDERS,
                animations: [animations.tuiSlideInTop, animations.tuiFadeIn],
                styles: [":host{position:relative;display:flex;font:var(--tui-font-text-m);flex-direction:column;box-sizing:border-box;margin:auto;border-radius:1rem;border:2.5rem solid transparent}:host:after{position:absolute;top:0;left:0;width:100%;height:100%;box-shadow:0 1.125rem 1.875rem rgba(0,0,0,.48);content:'';border-radius:inherit;pointer-events:none}:host[data-size=auto]{width:auto}:host[data-size='s']{width:30rem}:host[data-size='s'] .t-content{padding:1.5rem}:host[data-size='s'] .t-heading{margin-bottom:.5rem;font:var(--tui-font-heading-5)}:host[data-size='m']{width:42.5rem}:host[data-size='m'] .t-heading{margin-bottom:.75rem;font:var(--tui-font-heading-4)}:host[data-size='l']{width:55rem}:host[data-size='l'] .t-heading{margin-bottom:1rem}:host[data-size=fullscreen],:host[data-size=page]{width:100%;min-height:100%;border-radius:0;border:none;background:var(--tui-elevation-01);box-shadow:0 4rem var(--tui-base-01)}:host[data-size=fullscreen] .t-content,:host[data-size=page] .t-content{padding:3rem calc(50vw - (45rem / 2))}:host[data-size=fullscreen] .t-heading,:host[data-size=page] .t-heading{margin-bottom:1rem}:host[data-size=page] .t-content{padding:0}:host._centered{text-align:center}:host :host-context(tui-root._mobile)[data-size]{min-width:100%;width:100%;max-width:100%;border-radius:0;border:none;margin:auto 0 0}:host :host-context(tui-root._mobile)[data-size] .t-content{padding:1.5rem}:host :host-context(tui-root._mobile)[data-size] .t-heading{margin-bottom:.5rem;font:var(--tui-font-heading-5)}.t-heading{margin:0;word-wrap:break-word;font:var(--tui-font-heading-3)}.t-heading:empty{display:none}.t-header{display:flex;border-top-left-radius:inherit;border-top-right-radius:inherit;overflow:hidden}:host[data-size=fullscreen] :host-context(tui-root._mobile) .t-header{flex:1}.t-content{border-radius:inherit;padding:2rem;background:var(--tui-base-01)}.t-content:not(:first-child){border-top-left-radius:0;border-top-right-radius:0}.t-wrapper{position:-webkit-sticky;position:sticky;top:0;order:-1}.t-close{transition-property:background;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;position:fixed;top:1.5rem;right:1.5rem;color:var(--tui-base-01);background:rgba(104,104,104,.96)}.t-close:hover{background:rgba(159,159,159,.86)}:host:not([data-size=fullscreen]) .t-close{animation:tuiFadeIn var(--tui-duration)}:host:not([data-size=fullscreen]).ng-animating .t-close{display:none}:host-context(tui-root._mobile) .t-close{position:absolute;top:0;right:0;background:0 0;color:var(--tui-base-06)}:host-context(tui-root._mobile) .t-close:hover{color:var(--tui-base-07)}.t-buttons{margin-top:1.25rem;text-align:right}"]
            }),
            __param(0, core.Inject(tokens.TUI_ANIMATIONS_DURATION)),
            __param(1, core.Inject(cdk.TUI_IS_MOBILE)),
            __param(2, core.Inject(ngPolymorpheus.POLYMORPHEUS_CONTEXT)),
            __param(3, core.Inject(TUI_DIALOG_CLOSE_STREAM)),
            __param(4, core.Inject(tokens.TUI_CLOSE_WORD))
        ], TuiDialogComponent);
        return TuiDialogComponent;
    }());

    var DIALOG = new ngPolymorpheus.PolymorpheusComponent(TuiDialogComponent);
    var DEFAULT_OPTIONS = {
        size: "m",
        required: false,
        closeable: true,
        dismissible: true,
        label: "",
        header: "",
    };
    var TuiDialogService = /** @class */ (function (_super) {
        __extends(TuiDialogService, _super);
        function TuiDialogService() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.component = DIALOG;
            _this.defaultOptions = DEFAULT_OPTIONS;
            return _this;
        }
        TuiDialogService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiDialogService_Factory() { return new TuiDialogService(core.ɵɵinject(services.TuiIdService)); }, token: TuiDialogService, providedIn: "root" });
        TuiDialogService = __decorate([
            core.Injectable({
                providedIn: "root",
            })
        ], TuiDialogService);
        return TuiDialogService;
    }(cdk.AbstractTuiDialogService));

    var TuiDialogDirective = /** @class */ (function (_super) {
        __extends(TuiDialogDirective, _super);
        function TuiDialogDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TuiDialogDirective = __decorate([
            core.Directive({
                selector: 'ng-template[tuiDialog]',
                providers: [
                    {
                        provide: cdk.AbstractTuiDialogService,
                        useExisting: TuiDialogService,
                    },
                ],
                inputs: ['options: tuiDialogOptions', 'open: tuiDialog'],
                outputs: ['openChange: tuiDialogChange'],
            })
        ], TuiDialogDirective);
        return TuiDialogDirective;
    }(cdk.AbstractTuiDialogDirective));

    var TuiDialogModule = /** @class */ (function () {
        function TuiDialogModule() {
        }
        TuiDialogModule = __decorate([
            core.NgModule({
                imports: [ngPolymorpheus.PolymorpheusModule, button.TuiButtonModule, common.CommonModule, cdk.TuiPreventDefaultModule],
                declarations: [TuiDialogComponent, TuiDialogDirective],
                exports: [TuiDialogComponent, TuiDialogDirective],
                entryComponents: [TuiDialogComponent],
                providers: [
                    {
                        provide: cdk.TUI_DIALOGS,
                        useExisting: TuiDialogService,
                        multi: true,
                    },
                ],
            })
        ], TuiDialogModule);
        return TuiDialogModule;
    }());

    exports.TUI_DIALOGS_CLOSE = TUI_DIALOGS_CLOSE;
    exports.TUI_DIALOG_CLOSE_STREAM = TUI_DIALOG_CLOSE_STREAM;
    exports.TUI_DIALOG_PROVIDERS = TUI_DIALOG_PROVIDERS;
    exports.TuiDialogComponent = TuiDialogComponent;
    exports.TuiDialogDirective = TuiDialogDirective;
    exports.TuiDialogModule = TuiDialogModule;
    exports.TuiDialogService = TuiDialogService;
    exports.dialogCloseStreamFactory = dialogCloseStreamFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
