// @taiga-ui/core/bundles/taiga-ui-core-tokens.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/i18n'), require('@taiga-ui/core/utils'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/cdk/utils/svg'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/tokens', ['exports', '@angular/core', '@taiga-ui/i18n', '@taiga-ui/core/utils', '@ng-web-apis/common', '@taiga-ui/cdk', 'rxjs/operators', '@angular/common', '@taiga-ui/cdk/utils/svg', 'rxjs'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.tokens = {}), global.ng.core, global.i18n, global['taiga-ui'].core.utils, global.common, global.cdk, global.rxjs.operators, global.ng.common, global.svg, global.rxjs));
}(this, (function (exports, core, i18n, utils, common, cdk, operators, common$1, svg, rxjs) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_ANIMATIONS_DURATION = new core.InjectionToken("Duration of all Taiga UI animations in ms", {
        factory: function () { return 300; },
    });

    var TUI_ANIMATION_OPTIONS = new core.InjectionToken("Options for Taiga UI animations", {
        factory: function () { return ({
            params: {
                duration: core.inject(TUI_ANIMATIONS_DURATION),
            },
        }); },
    });

    var TUI_ASSERT_ENABLED = new core.InjectionToken("Flag to enable assertions across Taiga UI", {
        factory: function () { return core.isDevMode(); },
    });

    var TUI_DATA_LIST_ACCESSOR = new core.InjectionToken("Accessor for options");

    var TUI_DATA_LIST_HOST = new core.InjectionToken("DataList controller");

    var TUI_DOCUMENT_OR_SHADOW_ROOT = new core.InjectionToken("Document or ShadowRoot");

    var TUI_DROPDOWN_DIRECTIVE = new core.InjectionToken("Directive controlling TuiDropdownBoxComponent");

    var TUI_ELEMENT_REF = new core.InjectionToken("ElementRef when you cannot use @Input for single time injection");

    var TUI_FIRST_DAY_OF_WEEK = new core.InjectionToken("The first day of the week index", {
        factory: function () { return 1 /* Monday */; },
    });

    var TUI_MONTHS = new core.InjectionToken("Localized months names", {
        factory: i18n.extractI18n("months"),
    });
    var TUI_CLOSE_WORD = new core.InjectionToken("i18n 'close' word", {
        factory: i18n.extractI18n("close"),
    });
    var TUI_NOTHING_FOUND_MESSAGE = new core.InjectionToken("i18n 'Nothing found' message", {
        factory: i18n.extractI18n("nothingFoundMessage"),
    });
    var TUI_DEFAULT_ERROR_MESSAGE = new core.InjectionToken("i18n of error message", {
        factory: i18n.extractI18n("defaultErrorMessage"),
    });
    var TUI_SPIN_TEXTS = new core.InjectionToken("spin i18n texts", {
        factory: i18n.extractI18n("spinTexts"),
    });
    var TUI_SHORT_WEEK_DAYS = new core.InjectionToken("calendars i18n texts", {
        factory: i18n.extractI18n("shortWeekDays"),
    });

    var TUI_ICONS_PLACE = new core.InjectionToken("Recommended paths to icons from documentation guide", {
        factory: function () { return "assets/taiga-ui/icons"; },
    });

    var TUI_ICONS = new core.InjectionToken("A key/value dictionary of icon names and src to be defined with TuiSvgService", {
        factory: function () { return ({}); },
    });

    var TUI_ICONS_PATH = new core.InjectionToken("A handler to retrieve USE id for icon by name", { factory: function () { return utils.iconsPathFactory(core.inject(TUI_ICONS_PLACE)); } });

    var TUI_MEDIA = new core.InjectionToken("Token for media constant", {
        factory: function () { return ({
            mobile: 768,
            desktopSmall: 1024,
            desktopLarge: 1280,
        }); },
    });

    var TUI_IS_MOBILE_RES = new core.InjectionToken("Mobile resolution stream for private providers", {
        factory: function () {
            var windowRef = core.inject(common.WINDOW);
            var media = core.inject(TUI_MEDIA);
            return cdk.typedFromEvent(windowRef, "resize").pipe(operators.share(), operators.startWith(null), operators.map(function () { return utils.tuiIsMobile(windowRef, media); }), operators.distinctUntilChanged(), cdk.tuiZoneOptimized(core.inject(core.NgZone)));
        },
    });

    var TUI_MODE = new core.InjectionToken("Mode stream for private providers");

    /** Default values for the notification options. */
    var TUI_NOTIFICATION_DEFAULT_OPTIONS = {
        autoClose: true,
        label: "",
        status: "info" /* Info */,
        hasIcon: true,
        hasCloseButton: true,
        defaultAutoCloseTime: 3000,
    };
    var TUI_NOTIFICATION_OPTIONS = new core.InjectionToken("Default parameters for notification alert component", {
        factory: function () { return TUI_NOTIFICATION_DEFAULT_OPTIONS; },
    });
    function tuiNotificationOptionsProvider(options) {
        return {
            provide: TUI_NOTIFICATION_OPTIONS,
            useValue: __assign(__assign({}, TUI_NOTIFICATION_DEFAULT_OPTIONS), options),
        };
    }

    var TUI_NUMBER_FORMAT = new core.InjectionToken("Formatting configuration for displayed numbers", {
        factory: function () { return ({
            decimalSeparator: ",",
            thousandSeparator: cdk.CHAR_NO_BREAK_SPACE,
            zeroPadding: true,
        }); },
    });

    var TUI_OPTION_CONTENT = new core.InjectionToken("Content for tuiOption component");

    // TODO: 3.0 Remove export in ivy compilation
    var convertToSundayFirstWeekFormat = function (weekDaysNames) {
        var sundayIndex = weekDaysNames.length - 1;
        return __spread([
            weekDaysNames[sundayIndex]
        ], weekDaysNames.slice(0, sundayIndex));
    };
    var TUI_ORDERED_SHORT_WEEK_DAYS = new core.InjectionToken("Ordered calendars i18n texts", {
        factory: function () {
            var firstDayOfWeekIndex = core.inject(TUI_FIRST_DAY_OF_WEEK);
            return core.inject(TUI_SHORT_WEEK_DAYS).pipe(operators.map(convertToSundayFirstWeekFormat), operators.map(function (weekDays) {
                return __spread(weekDays.slice(firstDayOfWeekIndex), weekDays.slice(0, firstDayOfWeekIndex));
            }));
        },
    });

    var TUI_SANITIZER = cdk.TUI_SANITIZER;

    /**
     * Scrollable container
     */
    var TUI_SCROLL_REF = new core.InjectionToken("[TUI_SCROLL_REF]", {
        factory: function () { return new core.ElementRef(core.inject(common$1.DOCUMENT).documentElement); },
    });

    var TUI_SVG_CONTENT_PROCESSOR = new core.InjectionToken("Transform function the contents of the loaded svg file", { factory: function () { return svg.svgLinearGradientProcessor; } });

    var TUI_SVG_SRC_PROCESSOR = new core.InjectionToken("Source path processor for svg", {
        factory: function () { return rxjs.identity; },
    });

    var TUI_TEXTFIELD_APPEARANCE = new core.InjectionToken("Appearance for inputs", {
        factory: function () { return "textfield" /* Textfield */; },
    });

    var TUI_TEXTFIELD_HOST = new core.InjectionToken("An interface to communicate with textfield based controls");

    var TUI_THEME = new core.InjectionToken("Theme name", {
        factory: function () { return "Taiga"; },
    });

    var TUI_VALUE_ACCESSOR = new core.InjectionToken("Buffer token to pass NG_VALUE_ACCESSOR to a different Injector");

    exports.TUI_ANIMATIONS_DURATION = TUI_ANIMATIONS_DURATION;
    exports.TUI_ANIMATION_OPTIONS = TUI_ANIMATION_OPTIONS;
    exports.TUI_ASSERT_ENABLED = TUI_ASSERT_ENABLED;
    exports.TUI_CLOSE_WORD = TUI_CLOSE_WORD;
    exports.TUI_DATA_LIST_ACCESSOR = TUI_DATA_LIST_ACCESSOR;
    exports.TUI_DATA_LIST_HOST = TUI_DATA_LIST_HOST;
    exports.TUI_DEFAULT_ERROR_MESSAGE = TUI_DEFAULT_ERROR_MESSAGE;
    exports.TUI_DOCUMENT_OR_SHADOW_ROOT = TUI_DOCUMENT_OR_SHADOW_ROOT;
    exports.TUI_DROPDOWN_DIRECTIVE = TUI_DROPDOWN_DIRECTIVE;
    exports.TUI_ELEMENT_REF = TUI_ELEMENT_REF;
    exports.TUI_FIRST_DAY_OF_WEEK = TUI_FIRST_DAY_OF_WEEK;
    exports.TUI_ICONS = TUI_ICONS;
    exports.TUI_ICONS_PATH = TUI_ICONS_PATH;
    exports.TUI_ICONS_PLACE = TUI_ICONS_PLACE;
    exports.TUI_IS_MOBILE_RES = TUI_IS_MOBILE_RES;
    exports.TUI_MEDIA = TUI_MEDIA;
    exports.TUI_MODE = TUI_MODE;
    exports.TUI_MONTHS = TUI_MONTHS;
    exports.TUI_NOTHING_FOUND_MESSAGE = TUI_NOTHING_FOUND_MESSAGE;
    exports.TUI_NOTIFICATION_DEFAULT_OPTIONS = TUI_NOTIFICATION_DEFAULT_OPTIONS;
    exports.TUI_NOTIFICATION_OPTIONS = TUI_NOTIFICATION_OPTIONS;
    exports.TUI_NUMBER_FORMAT = TUI_NUMBER_FORMAT;
    exports.TUI_OPTION_CONTENT = TUI_OPTION_CONTENT;
    exports.TUI_ORDERED_SHORT_WEEK_DAYS = TUI_ORDERED_SHORT_WEEK_DAYS;
    exports.TUI_SANITIZER = TUI_SANITIZER;
    exports.TUI_SCROLL_REF = TUI_SCROLL_REF;
    exports.TUI_SHORT_WEEK_DAYS = TUI_SHORT_WEEK_DAYS;
    exports.TUI_SPIN_TEXTS = TUI_SPIN_TEXTS;
    exports.TUI_SVG_CONTENT_PROCESSOR = TUI_SVG_CONTENT_PROCESSOR;
    exports.TUI_SVG_SRC_PROCESSOR = TUI_SVG_SRC_PROCESSOR;
    exports.TUI_TEXTFIELD_APPEARANCE = TUI_TEXTFIELD_APPEARANCE;
    exports.TUI_TEXTFIELD_HOST = TUI_TEXTFIELD_HOST;
    exports.TUI_THEME = TUI_THEME;
    exports.TUI_VALUE_ACCESSOR = TUI_VALUE_ACCESSOR;
    exports.convertToSundayFirstWeekFormat = convertToSundayFirstWeekFormat;
    exports.tuiNotificationOptionsProvider = tuiNotificationOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
