// @taiga-ui/core/bundles/taiga-ui-core-directives-textfield-controller.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/directives/textfield-controller', ['exports', '@angular/core', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.directives = global['taiga-ui'].core.directives || {}, global['taiga-ui'].core.directives['textfield-controller'] = {}), global.ng.core, global.cdk, global.rxjs, global.rxjs.operators));
}(this, (function (exports, core, cdk, rxjs, operators) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TuiTextfieldController = /** @class */ (function () {
        function TuiTextfieldController(change$, autocompleteDirective, cleanerDirective, customContentDirective, exampleTextDirective, inputModeDirective, labelOutsideDirective, maxLengthDirective, sizeDirective, typeDirective) {
            this.change$ = change$;
            this.autocompleteDirective = autocompleteDirective;
            this.cleanerDirective = cleanerDirective;
            this.customContentDirective = customContentDirective;
            this.exampleTextDirective = exampleTextDirective;
            this.inputModeDirective = inputModeDirective;
            this.labelOutsideDirective = labelOutsideDirective;
            this.maxLengthDirective = maxLengthDirective;
            this.sizeDirective = sizeDirective;
            this.typeDirective = typeDirective;
        }
        Object.defineProperty(TuiTextfieldController.prototype, "autocomplete", {
            get: function () {
                return this.autocompleteDirective.autocomplete;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "cleaner", {
            get: function () {
                return this.cleanerDirective.cleaner;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "customContent", {
            get: function () {
                return this.customContentDirective.customContent || "";
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "exampleText", {
            get: function () {
                return this.exampleTextDirective.exampleText;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "inputMode", {
            get: function () {
                return this.inputModeDirective.inputMode;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "labelOutside", {
            get: function () {
                return this.labelOutsideDirective.labelOutside;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "maxLength", {
            get: function () {
                return this.maxLengthDirective.maxLength;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "size", {
            get: function () {
                return this.sizeDirective.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextfieldController.prototype, "type", {
            get: function () {
                return this.typeDirective.type;
            },
            enumerable: true,
            configurable: true
        });
        return TuiTextfieldController;
    }());

    var TUI_TEXTFIELD_AUTOCOMPLETE = new core.InjectionToken('tuiTextfieldAutocomplete', {
        factory: autocompleteDirectiveFactory,
    });
    var TuiTextfieldAutocompleteDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldAutocompleteDirective, _super);
        function TuiTextfieldAutocompleteDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.autocomplete = '';
            return _this;
        }
        TuiTextfieldAutocompleteDirective_1 = TuiTextfieldAutocompleteDirective;
        var TuiTextfieldAutocompleteDirective_1;
        __decorate([
            core.Input('tuiTextfieldAutocomplete')
        ], TuiTextfieldAutocompleteDirective.prototype, "autocomplete", void 0);
        TuiTextfieldAutocompleteDirective = TuiTextfieldAutocompleteDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldAutocomplete]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_AUTOCOMPLETE,
                        useExisting: core.forwardRef(function () { return TuiTextfieldAutocompleteDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldAutocompleteDirective);
        return TuiTextfieldAutocompleteDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function autocompleteDirectiveFactory() {
        return new TuiTextfieldAutocompleteDirective();
    }

    var TUI_TEXTFIELD_CLEANER = new core.InjectionToken('tuiTextfieldCleaner', {
        factory: cleanerDirectiveFactory,
    });
    var TuiTextfieldCleanerDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldCleanerDirective, _super);
        function TuiTextfieldCleanerDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.cleaner = false;
            return _this;
        }
        TuiTextfieldCleanerDirective_1 = TuiTextfieldCleanerDirective;
        var TuiTextfieldCleanerDirective_1;
        __decorate([
            core.Input('tuiTextfieldCleaner')
        ], TuiTextfieldCleanerDirective.prototype, "cleaner", void 0);
        TuiTextfieldCleanerDirective = TuiTextfieldCleanerDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldCleaner]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_CLEANER,
                        useExisting: core.forwardRef(function () { return TuiTextfieldCleanerDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldCleanerDirective);
        return TuiTextfieldCleanerDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function cleanerDirectiveFactory() {
        return new TuiTextfieldCleanerDirective();
    }

    var TUI_TEXTFIELD_CUSTOM_CONTENT = new core.InjectionToken('tuiTextfieldCustomContent', {
        factory: customContentDirectiveFactory,
    });
    var TuiTextfieldCustomContentDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldCustomContentDirective, _super);
        function TuiTextfieldCustomContentDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            // TODO: 3.0 Remove null
            _this.customContent = '';
            return _this;
        }
        TuiTextfieldCustomContentDirective_1 = TuiTextfieldCustomContentDirective;
        var TuiTextfieldCustomContentDirective_1;
        __decorate([
            core.Input('tuiTextfieldCustomContent')
        ], TuiTextfieldCustomContentDirective.prototype, "customContent", void 0);
        TuiTextfieldCustomContentDirective = TuiTextfieldCustomContentDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldCustomContent]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_CUSTOM_CONTENT,
                        useExisting: core.forwardRef(function () { return TuiTextfieldCustomContentDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldCustomContentDirective);
        return TuiTextfieldCustomContentDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function customContentDirectiveFactory() {
        return new TuiTextfieldCustomContentDirective();
    }

    var TUI_TEXTFIELD_EXAMPLE_TEXT = new core.InjectionToken('tuiTextfieldExampleText', {
        factory: exampleTextDirectiveFactory,
    });
    var TuiTextfieldExampleTextDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldExampleTextDirective, _super);
        function TuiTextfieldExampleTextDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.exampleText = '';
            return _this;
        }
        TuiTextfieldExampleTextDirective_1 = TuiTextfieldExampleTextDirective;
        var TuiTextfieldExampleTextDirective_1;
        __decorate([
            core.Input('tuiTextfieldExampleText')
        ], TuiTextfieldExampleTextDirective.prototype, "exampleText", void 0);
        TuiTextfieldExampleTextDirective = TuiTextfieldExampleTextDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldExampleText]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_EXAMPLE_TEXT,
                        useExisting: core.forwardRef(function () { return TuiTextfieldExampleTextDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldExampleTextDirective);
        return TuiTextfieldExampleTextDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function exampleTextDirectiveFactory() {
        return new TuiTextfieldExampleTextDirective();
    }

    var TUI_TEXTFIELD_INPUT_MODE = new core.InjectionToken('tuiTextfieldInputMode', {
        factory: inputModeDirectiveFactory,
    });
    var TuiTextfieldInputModeDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldInputModeDirective, _super);
        function TuiTextfieldInputModeDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.inputMode = 'text';
            return _this;
        }
        TuiTextfieldInputModeDirective_1 = TuiTextfieldInputModeDirective;
        var TuiTextfieldInputModeDirective_1;
        __decorate([
            core.Input('tuiTextfieldInputMode')
        ], TuiTextfieldInputModeDirective.prototype, "inputMode", void 0);
        TuiTextfieldInputModeDirective = TuiTextfieldInputModeDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldInputMode]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_INPUT_MODE,
                        useExisting: core.forwardRef(function () { return TuiTextfieldInputModeDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldInputModeDirective);
        return TuiTextfieldInputModeDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function inputModeDirectiveFactory() {
        return new TuiTextfieldInputModeDirective();
    }

    var TUI_TEXTFIELD_LABEL_OUTSIDE = new core.InjectionToken('tuiTextfieldLabelOutside', {
        factory: labelOutsideDirectiveFactory,
    });
    var TuiTextfieldLabelOutsideDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldLabelOutsideDirective, _super);
        function TuiTextfieldLabelOutsideDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.labelOutside = false;
            return _this;
        }
        TuiTextfieldLabelOutsideDirective_1 = TuiTextfieldLabelOutsideDirective;
        var TuiTextfieldLabelOutsideDirective_1;
        __decorate([
            core.Input('tuiTextfieldLabelOutside')
        ], TuiTextfieldLabelOutsideDirective.prototype, "labelOutside", void 0);
        TuiTextfieldLabelOutsideDirective = TuiTextfieldLabelOutsideDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldLabelOutside]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_LABEL_OUTSIDE,
                        useExisting: core.forwardRef(function () { return TuiTextfieldLabelOutsideDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldLabelOutsideDirective);
        return TuiTextfieldLabelOutsideDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function labelOutsideDirectiveFactory() {
        return new TuiTextfieldLabelOutsideDirective();
    }

    var TUI_TEXTFIELD_MAX_LENGTH = new core.InjectionToken('tuiTextfieldMaxLength', {
        factory: maxLengthDirectiveFactory,
    });
    var TuiTextfieldMaxLengthDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldMaxLengthDirective, _super);
        function TuiTextfieldMaxLengthDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.maxLength = null;
            return _this;
        }
        TuiTextfieldMaxLengthDirective_1 = TuiTextfieldMaxLengthDirective;
        var TuiTextfieldMaxLengthDirective_1;
        __decorate([
            core.Input('tuiTextfieldMaxLength')
        ], TuiTextfieldMaxLengthDirective.prototype, "maxLength", void 0);
        TuiTextfieldMaxLengthDirective = TuiTextfieldMaxLengthDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldMaxLength]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_MAX_LENGTH,
                        useExisting: core.forwardRef(function () { return TuiTextfieldMaxLengthDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldMaxLengthDirective);
        return TuiTextfieldMaxLengthDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function maxLengthDirectiveFactory() {
        return new TuiTextfieldMaxLengthDirective();
    }

    var TUI_TEXTFIELD_SIZE = new core.InjectionToken('tuiTextfieldSize', {
        factory: sizeDirectiveFactory,
    });
    var TuiTextfieldSizeDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldSizeDirective, _super);
        function TuiTextfieldSizeDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.size = 'l';
            return _this;
        }
        TuiTextfieldSizeDirective_1 = TuiTextfieldSizeDirective;
        var TuiTextfieldSizeDirective_1;
        __decorate([
            core.Input('tuiTextfieldSize')
        ], TuiTextfieldSizeDirective.prototype, "size", void 0);
        TuiTextfieldSizeDirective = TuiTextfieldSizeDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldSize]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_SIZE,
                        useExisting: core.forwardRef(function () { return TuiTextfieldSizeDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldSizeDirective);
        return TuiTextfieldSizeDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function sizeDirectiveFactory() {
        return new TuiTextfieldSizeDirective();
    }

    var TUI_TEXTFIELD_TYPE = new core.InjectionToken('tuiTextfieldType', {
        factory: typeDirectiveFactory,
    });
    var TuiTextfieldTypeDirective = /** @class */ (function (_super) {
        __extends(TuiTextfieldTypeDirective, _super);
        function TuiTextfieldTypeDirective() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.type = 'text';
            return _this;
        }
        TuiTextfieldTypeDirective_1 = TuiTextfieldTypeDirective;
        var TuiTextfieldTypeDirective_1;
        __decorate([
            core.Input('tuiTextfieldType')
        ], TuiTextfieldTypeDirective.prototype, "type", void 0);
        TuiTextfieldTypeDirective = TuiTextfieldTypeDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiTextfieldType]',
                providers: [
                    {
                        provide: TUI_TEXTFIELD_TYPE,
                        useExisting: core.forwardRef(function () { return TuiTextfieldTypeDirective_1; }),
                    },
                ],
            })
        ], TuiTextfieldTypeDirective);
        return TuiTextfieldTypeDirective;
    }(cdk.TuiController));
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function typeDirectiveFactory() {
        return new TuiTextfieldTypeDirective();
    }

    var TuiTextfieldControllerModule = /** @class */ (function () {
        function TuiTextfieldControllerModule() {
        }
        TuiTextfieldControllerModule = __decorate([
            core.NgModule({
                declarations: [
                    TuiTextfieldAutocompleteDirective,
                    TuiTextfieldCleanerDirective,
                    TuiTextfieldCustomContentDirective,
                    TuiTextfieldExampleTextDirective,
                    TuiTextfieldInputModeDirective,
                    TuiTextfieldLabelOutsideDirective,
                    TuiTextfieldMaxLengthDirective,
                    TuiTextfieldSizeDirective,
                    TuiTextfieldTypeDirective,
                ],
                exports: [
                    TuiTextfieldAutocompleteDirective,
                    TuiTextfieldCleanerDirective,
                    TuiTextfieldCustomContentDirective,
                    TuiTextfieldExampleTextDirective,
                    TuiTextfieldInputModeDirective,
                    TuiTextfieldLabelOutsideDirective,
                    TuiTextfieldMaxLengthDirective,
                    TuiTextfieldSizeDirective,
                    TuiTextfieldTypeDirective,
                ],
            })
        ], TuiTextfieldControllerModule);
        return TuiTextfieldControllerModule;
    }());

    var TUI_TEXTFIELD_WATCHED_CONTROLLER = new core.InjectionToken("watched textfield controller");
    /**
     * @deprecated: use {@link tuiTextfieldWatchedControllerFactory} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function textfieldWatchedControllerFactory(changeDetectorRef, destroy$) {
        var controllers = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            controllers[_i - 2] = arguments[_i];
        }
        var change$ = rxjs.merge.apply(void 0, __spread(controllers.map(function (_a) {
            var change$ = _a.change$;
            return change$ || rxjs.NEVER;
        }))).pipe(cdk.watch(changeDetectorRef), operators.takeUntil(destroy$));
        change$.subscribe();
        return new (TuiTextfieldController.bind.apply(TuiTextfieldController, __spread([void 0, change$], controllers)))();
    }
    var TEXTFIELD_CONTROLLER_PROVIDER = [
        cdk.TuiDestroyService,
        {
            provide: TUI_TEXTFIELD_WATCHED_CONTROLLER,
            deps: [
                core.ChangeDetectorRef,
                cdk.TuiDestroyService,
                TUI_TEXTFIELD_AUTOCOMPLETE,
                TUI_TEXTFIELD_CLEANER,
                TUI_TEXTFIELD_CUSTOM_CONTENT,
                TUI_TEXTFIELD_EXAMPLE_TEXT,
                TUI_TEXTFIELD_INPUT_MODE,
                TUI_TEXTFIELD_LABEL_OUTSIDE,
                TUI_TEXTFIELD_MAX_LENGTH,
                TUI_TEXTFIELD_SIZE,
                TUI_TEXTFIELD_TYPE,
            ],
            useFactory: textfieldWatchedControllerFactory,
        },
    ];
    var tuiTextfieldWatchedControllerFactory = textfieldWatchedControllerFactory;

    exports.TEXTFIELD_CONTROLLER_PROVIDER = TEXTFIELD_CONTROLLER_PROVIDER;
    exports.TUI_TEXTFIELD_AUTOCOMPLETE = TUI_TEXTFIELD_AUTOCOMPLETE;
    exports.TUI_TEXTFIELD_CLEANER = TUI_TEXTFIELD_CLEANER;
    exports.TUI_TEXTFIELD_CUSTOM_CONTENT = TUI_TEXTFIELD_CUSTOM_CONTENT;
    exports.TUI_TEXTFIELD_EXAMPLE_TEXT = TUI_TEXTFIELD_EXAMPLE_TEXT;
    exports.TUI_TEXTFIELD_INPUT_MODE = TUI_TEXTFIELD_INPUT_MODE;
    exports.TUI_TEXTFIELD_LABEL_OUTSIDE = TUI_TEXTFIELD_LABEL_OUTSIDE;
    exports.TUI_TEXTFIELD_MAX_LENGTH = TUI_TEXTFIELD_MAX_LENGTH;
    exports.TUI_TEXTFIELD_SIZE = TUI_TEXTFIELD_SIZE;
    exports.TUI_TEXTFIELD_TYPE = TUI_TEXTFIELD_TYPE;
    exports.TUI_TEXTFIELD_WATCHED_CONTROLLER = TUI_TEXTFIELD_WATCHED_CONTROLLER;
    exports.TuiTextfieldAutocompleteDirective = TuiTextfieldAutocompleteDirective;
    exports.TuiTextfieldCleanerDirective = TuiTextfieldCleanerDirective;
    exports.TuiTextfieldController = TuiTextfieldController;
    exports.TuiTextfieldControllerModule = TuiTextfieldControllerModule;
    exports.TuiTextfieldCustomContentDirective = TuiTextfieldCustomContentDirective;
    exports.TuiTextfieldExampleTextDirective = TuiTextfieldExampleTextDirective;
    exports.TuiTextfieldInputModeDirective = TuiTextfieldInputModeDirective;
    exports.TuiTextfieldLabelOutsideDirective = TuiTextfieldLabelOutsideDirective;
    exports.TuiTextfieldMaxLengthDirective = TuiTextfieldMaxLengthDirective;
    exports.TuiTextfieldSizeDirective = TuiTextfieldSizeDirective;
    exports.TuiTextfieldTypeDirective = TuiTextfieldTypeDirective;
    exports.autocompleteDirectiveFactory = autocompleteDirectiveFactory;
    exports.cleanerDirectiveFactory = cleanerDirectiveFactory;
    exports.customContentDirectiveFactory = customContentDirectiveFactory;
    exports.exampleTextDirectiveFactory = exampleTextDirectiveFactory;
    exports.inputModeDirectiveFactory = inputModeDirectiveFactory;
    exports.labelOutsideDirectiveFactory = labelOutsideDirectiveFactory;
    exports.maxLengthDirectiveFactory = maxLengthDirectiveFactory;
    exports.sizeDirectiveFactory = sizeDirectiveFactory;
    exports.textfieldWatchedControllerFactory = textfieldWatchedControllerFactory;
    exports.tuiTextfieldWatchedControllerFactory = tuiTextfieldWatchedControllerFactory;
    exports.typeDirectiveFactory = typeDirectiveFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
