// @taiga-ui/core/bundles/taiga-ui-core-components-data-list.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/tokens'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/directives/dropdown'), require('@tinkoff/ng-event-plugins'), require('@angular/common'), require('@taiga-ui/core/components/svg'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/data-list', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/tokens', 'rxjs', 'rxjs/operators', '@taiga-ui/core/directives/dropdown', '@tinkoff/ng-event-plugins', '@angular/common', '@taiga-ui/core/components/svg', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components['data-list'] = {}), global.ng.core, global.cdk, global['taiga-ui'].core.tokens, global.rxjs, global.rxjs.operators, global['taiga-ui'].core.directives.dropdown, global.ngEventPlugins, global.ng.common, global['taiga-ui'].core.components.svg, global.ngPolymorpheus));
}(this, (function (exports, core, cdk, tokens, rxjs, operators, dropdown, ngEventPlugins, common, svg, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    // eslint-disable-next-line @typescript-eslint/naming-convention
    function shouldFocus(_a) {
        var currentTarget = _a.currentTarget;
        return !cdk.isNativeFocused(currentTarget);
    }
    // TODO: Consider all use cases for aria roles
    var TuiOptionComponent = /** @class */ (function () {
        // TODO: 3.0 Fix dataList type when updated to Ivy and compilation drops metadata
        function TuiOptionComponent(content, dataList, elementRef, host, dropdown) {
            this.content = content;
            this.dataList = dataList;
            this.elementRef = elementRef;
            this.host = host;
            this.dropdown = dropdown;
            this.size = 'm';
            this.role = 'option';
            this.disabled = false;
        }
        Object.defineProperty(TuiOptionComponent.prototype, "active", {
            get: function () {
                return !!this.dropdown && !!this.dropdown.dropdownBoxRef;
            },
            enumerable: true,
            configurable: true
        });
        TuiOptionComponent.prototype.onClick = function () {
            if (this.host && this.value !== undefined) {
                this.host.handleOption(this.value);
            }
        };
        // @bad TODO: Consider aria-activedescendant for proper accessibility implementation
        TuiOptionComponent.prototype.onMouseMove = function (_a) {
            var currentTarget = _a.currentTarget;
            cdk.setNativeFocused(currentTarget, true, true);
        };
        // Preventing focus loss upon focused option removal
        TuiOptionComponent.prototype.ngOnDestroy = function () {
            this.dataList.handleFocusLossIfNecessary(this.elementRef.nativeElement);
        };
        TuiOptionComponent.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_OPTION_CONTENT,] }] },
            { type: TuiDataListComponent, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return TuiDataListComponent; }),] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_DATA_LIST_HOST,] }] },
            { type: dropdown.TuiDropdownDirective, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [dropdown.TuiDropdownDirective,] }] }
        ]; };
        __decorate([
            core.Input(),
            core.HostBinding('attr.data-size'),
            cdk.tuiDefaultProp()
        ], TuiOptionComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('attr.role'),
            cdk.tuiDefaultProp()
        ], TuiOptionComponent.prototype, "role", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiOptionComponent.prototype, "disabled", void 0);
        __decorate([
            core.Input()
        ], TuiOptionComponent.prototype, "value", void 0);
        __decorate([
            core.HostBinding('class._with-dropdown')
        ], TuiOptionComponent.prototype, "active", null);
        __decorate([
            core.HostListener('click')
        ], TuiOptionComponent.prototype, "onClick", null);
        __decorate([
            ngEventPlugins.shouldCall(shouldFocus),
            core.HostListener('mousemove.init', ['$event']),
            core.HostListener('mousemove.silent', ['$event'])
        ], TuiOptionComponent.prototype, "onMouseMove", null);
        TuiOptionComponent = __decorate([
            core.Component({
                selector: 'button[tuiOption], a[tuiOption]',
                template: "<span\n    *ngIf=\"content; else base\"\n    polymorpheus-outlet\n    class=\"t-content\"\n    [content]=\"content\"\n    [context]=\"{$implicit: base}\"\n></span>\n<ng-template #base>\n    <ng-content></ng-content>\n    <tui-svg\n        *ngIf=\"dropdown\"\n        src=\"tuiIconChevronRight\"\n        class=\"t-arrow\"\n    ></tui-svg>\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                host: {
                    tabIndex: '-1',
                    type: 'button',
                    '[attr.disabled]': 'disabled || null',
                },
                styles: [":host{-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:padding-box;font-size:inherit;line-height:inherit;transition-property:background;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;font:var(--tui-font-text-m);width:100%;align-items:center;justify-content:space-between;text-align:left;box-sizing:border-box;color:var(--tui-text-01);outline:0;text-decoration:none;cursor:pointer}:host:disabled{opacity:var(--tui-disabled-opacity);cursor:default}:host._with-dropdown,:host:focus{background-color:var(--tui-base-02)}:host[data-size=xs]{font:var(--tui-font-text-s);min-height:2rem;padding:.3125rem .75rem}:host[data-size='s']{min-height:2rem;padding:.3125rem .75rem}:host[data-size='m']{min-height:2.75rem;padding:.375rem 1rem}:host[data-size='l']{min-height:3.5rem;padding:.75rem 1rem}.t-content{display:flex;flex:1;max-width:100%}.t-arrow{margin:0 -.5rem 0 .75rem}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Inject(tokens.TUI_OPTION_CONTENT)),
            __param(1, core.Inject(core.forwardRef(function () { return TuiDataListComponent; }))),
            __param(2, core.Inject(core.ElementRef)),
            __param(3, core.Optional()),
            __param(3, core.Inject(tokens.TUI_DATA_LIST_HOST)),
            __param(4, core.Optional()),
            __param(4, core.Self()),
            __param(4, core.Inject(dropdown.TuiDropdownDirective))
        ], TuiOptionComponent);
        return TuiOptionComponent;
    }());

    // TODO: Consider aria-activedescendant for proper accessibility implementation
    // @dynamic
    var TuiDataListComponent = /** @class */ (function () {
        function TuiDataListComponent(elementRef, defaultEmptyContent$) {
            this.elementRef = elementRef;
            this.defaultEmptyContent$ = defaultEmptyContent$;
            this.options = cdk.EMPTY_QUERY;
            this.role = 'listbox';
            this.emptyContent = '';
        }
        TuiDataListComponent_1 = TuiDataListComponent;
        Object.defineProperty(TuiDataListComponent.prototype, "empty$", {
            get: function () {
                return cdk.itemsQueryListObservable(this.options).pipe(operators.map(function (_a) {
                    var length = _a.length;
                    return !length;
                }));
            },
            enumerable: true,
            configurable: true
        });
        TuiDataListComponent.prototype.onFocusIn = function (relatedTarget, currentTarget) {
            if (!currentTarget.contains(relatedTarget) && !this.origin) {
                this.origin = relatedTarget;
            }
        };
        TuiDataListComponent.prototype.noop = function () { };
        TuiDataListComponent.prototype.onKeyDownArrow = function (current, step) {
            var elements = this.elements;
            cdk.moveFocus(elements.indexOf(current), elements, step);
        };
        // TODO: Consider aria-activedescendant for proper accessibility implementation
        TuiDataListComponent.prototype.handleFocusLossIfNecessary = function (element) {
            if (element === void 0) { element = this.elementRef.nativeElement; }
            if (this.origin && cdk.isNativeFocusedIn(element)) {
                cdk.setNativeMouseFocused(this.origin, true, true);
            }
        };
        TuiDataListComponent.prototype.getOptions = function (includeDisabled) {
            if (includeDisabled === void 0) { includeDisabled = false; }
            return this.options
                .toArray()
                .filter(function (_a) {
                var disabled = _a.disabled;
                return includeDisabled || !disabled;
            })
                .map(function (_a) {
                var value = _a.value;
                return value;
            })
                .filter(cdk.isPresent);
        };
        TuiDataListComponent.prototype.onFocus = function (_a, top) {
            var target = _a.target;
            cdk.tuiAssertIsHTMLElement(target);
            var elements = this.elements;
            cdk.moveFocus(top ? -1 : elements.length, elements, top ? 1 : -1);
            this.handleFocusLossIfNecessary(target);
        };
        Object.defineProperty(TuiDataListComponent.prototype, "elements", {
            get: function () {
                return Array.from(this.elementRef.nativeElement.querySelectorAll('[tuiOption]'));
            },
            enumerable: true,
            configurable: true
        });
        var TuiDataListComponent_1;
        TuiDataListComponent.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_NOTHING_FOUND_MESSAGE,] }] }
        ]; };
        __decorate([
            core.ContentChildren(core.forwardRef(function () { return TuiOptionComponent; }), { descendants: true })
        ], TuiDataListComponent.prototype, "options", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('attr.role'),
            cdk.tuiDefaultProp()
        ], TuiDataListComponent.prototype, "role", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiDataListComponent.prototype, "emptyContent", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiDataListComponent.prototype, "empty$", null);
        __decorate([
            core.HostListener('focusin', ['$event.relatedTarget', '$event.currentTarget'])
        ], TuiDataListComponent.prototype, "onFocusIn", null);
        __decorate([
            core.HostListener('mousedown.prevent')
        ], TuiDataListComponent.prototype, "noop", null);
        __decorate([
            core.HostListener('keydown.arrowDown.prevent', ['$event.target', '1']),
            core.HostListener('keydown.arrowUp.prevent', ['$event.target', '-1'])
        ], TuiDataListComponent.prototype, "onKeyDownArrow", null);
        __decorate([
            core.HostListener('wheel.silent.passive'),
            core.HostListener('mouseleave', ['$event.target'])
        ], TuiDataListComponent.prototype, "handleFocusLossIfNecessary", null);
        TuiDataListComponent = TuiDataListComponent_1 = __decorate([
            core.Component({
                selector: 'tui-data-list',
                template: "<div\n    tabindex=\"0\"\n    class=\"t-trap\"\n    (focusin)=\"onFocus($event, true)\"\n></div>\n<ng-content></ng-content>\n<div\n    *ngIf=\"empty$ | async\"\n    polymorpheus-outlet\n    class=\"t-empty\"\n    [content]=\"emptyContent || (defaultEmptyContent$ | async)\"\n></div>\n<div\n    tabindex=\"0\"\n    class=\"t-trap\"\n    (focusin)=\"onFocus($event, false)\"\n></div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                encapsulation: core.ViewEncapsulation.None,
                providers: [
                    {
                        provide: tokens.TUI_DATA_LIST_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiDataListComponent_1; }),
                    },
                ],
                styles: ["tui-data-list{display:flex;font:var(--tui-font-text-m);flex-direction:column;padding:.5rem 0;color:var(--tui-text-03)}tui-data-list:focus-within .t-trap{display:none}tui-data-list:focus-within [tuiOption]._with-dropdown:not(:focus){background-color:transparent}tui-data-list>.t-empty{margin:.75rem 1rem}tui-opt-group{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;position:relative;display:flex;font:var(--tui-font-text-xs);flex-direction:column;line-height:2rem}tui-opt-group:empty:after,tui-opt-group:empty:before{display:none}tui-opt-group:before{content:attr(data-label);padding:.5rem 1rem;color:var(--tui-text-02);line-height:1rem;white-space:normal}tui-opt-group:not(:empty)~tui-opt-group:before{padding:1.5rem 1rem .5rem}tui-opt-group:not(:empty)~tui-opt-group[data-label='']:before{padding:1rem 0 0}tui-opt-group:not(:empty)~tui-opt-group:after{content:'';position:absolute;left:0;right:0;top:.5rem;height:1px;background:var(--tui-base-03)}tui-opt-group[data-label='']:before{padding:0 1rem}"]
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(tokens.TUI_NOTHING_FOUND_MESSAGE))
        ], TuiDataListComponent);
        return TuiDataListComponent;
    }());

    var TuiDataListDirective = /** @class */ (function () {
        function TuiDataListDirective() {
        }
        TuiDataListDirective = __decorate([
            core.Directive({
                selector: 'ng-template[tuiDataList]',
            })
        ], TuiDataListDirective);
        return TuiDataListDirective;
    }());

    // TODO: 3.0 Move into separate module
    var TuiDataListDropdownManagerDirective = /** @class */ (function () {
        function TuiDataListDropdownManagerDirective() {
            this.dropdowns = cdk.EMPTY_QUERY;
            this.elements = cdk.EMPTY_QUERY;
        }
        TuiDataListDropdownManagerDirective.prototype.ngAfterViewInit = function () {
            var _this = this;
            this.right$.subscribe(function (index) {
                _this.tryToFocus(index);
            });
            rxjs.merge(this.immediate$, this.debounce$)
                .pipe(operators.switchMap(function (active) {
                _this.dropdowns.forEach(function (dropdown, index) {
                    dropdown.open = index === active;
                });
                var element = _this.elements.toArray()[active];
                var dropdown = _this.dropdowns.toArray()[active];
                if (!element || !dropdown || !dropdown.dropdownBoxRef) {
                    return rxjs.EMPTY;
                }
                var nativeElement = dropdown.dropdownBoxRef.location.nativeElement;
                var mouseEnter$ = cdk.typedFromEvent(nativeElement, 'mouseenter').pipe(operators.take(1));
                var esc$ = rxjs.merge(cdk.typedFromEvent(element.nativeElement, 'keydown'), cdk.typedFromEvent(nativeElement, 'keydown')).pipe(operators.filter(function (_a) {
                    var keyCode = _a.keyCode;
                    return keyCode === 27;
                }));
                return rxjs.merge(mouseEnter$, esc$).pipe(operators.tap(function (event) {
                    if (dropdown.dropdownBoxRef) {
                        event.stopPropagation();
                    }
                    cdk.setNativeFocused(element.nativeElement);
                    // TODO: iframe warning
                    dropdown.open = event instanceof MouseEvent;
                }));
            }))
                .subscribe();
        };
        Object.defineProperty(TuiDataListDropdownManagerDirective.prototype, "elements$", {
            get: function () {
                return cdk.itemsQueryListObservable(this.elements).pipe(operators.map(function (array) { return array.map(function (_a) {
                    var nativeElement = _a.nativeElement;
                    return nativeElement;
                }); }), operators.shareReplay({ bufferSize: 1, refCount: true }));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDataListDropdownManagerDirective.prototype, "right$", {
            get: function () {
                return this.elements$.pipe(operators.switchMap(function (elements) {
                    return rxjs.merge.apply(void 0, __spread(elements.map(function (element, index) {
                        return cdk.typedFromEvent(element, 'keydown').pipe(operators.filter(function (_a) {
                            var keyCode = _a.keyCode;
                            return keyCode === 39;
                        }), cdk.preventDefault(), operators.mapTo(index));
                    })));
                }));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDataListDropdownManagerDirective.prototype, "immediate$", {
            get: function () {
                return this.elements$.pipe(operators.switchMap(function (elements) {
                    return rxjs.merge.apply(void 0, __spread(elements.map(function (element, index) {
                        return cdk.typedFromEvent(element, 'click').pipe(operators.mapTo(index));
                    })));
                }));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDataListDropdownManagerDirective.prototype, "debounce$", {
            get: function () {
                var _this = this;
                return this.elements$.pipe(operators.switchMap(function (elements) {
                    return rxjs.merge.apply(void 0, __spread(elements.map(function (element, index) {
                        return rxjs.merge(cdk.typedFromEvent(element, 'focus'), cdk.typedFromEvent(element, 'blur')).pipe(operators.filter(function (_a) {
                            var relatedTarget = _a.relatedTarget;
                            return _this.notInDropdown(relatedTarget, index);
                        }), operators.map(function (_a) {
                            var type = _a.type;
                            return (type === 'focus' ? index : NaN);
                        }));
                    })));
                }), operators.debounceTime(300));
            },
            enumerable: true,
            configurable: true
        });
        TuiDataListDropdownManagerDirective.prototype.notInDropdown = function (element, index) {
            var _a;
            var dropdown = this.dropdowns.toArray()[index];
            return !((_a = dropdown === null || dropdown === void 0 ? void 0 : dropdown.dropdownBoxRef) === null || _a === void 0 ? void 0 : _a.location.nativeElement.contains(element));
        };
        TuiDataListDropdownManagerDirective.prototype.tryToFocus = function (index) {
            var _a;
            var dropdown = this.dropdowns.toArray()[index];
            var content = (_a = dropdown === null || dropdown === void 0 ? void 0 : dropdown.dropdownBoxRef) === null || _a === void 0 ? void 0 : _a.instance.contentElementRef;
            if (!content) {
                return;
            }
            var item = cdk.getClosestKeyboardFocusable(content.nativeElement, false, content.nativeElement);
            if (item) {
                cdk.setNativeFocused(item);
            }
        };
        __decorate([
            core.ContentChildren(dropdown.TuiDropdownDirective, { descendants: true })
        ], TuiDataListDropdownManagerDirective.prototype, "dropdowns", void 0);
        __decorate([
            core.ContentChildren(dropdown.TuiDropdownDirective, { read: core.ElementRef, descendants: true })
        ], TuiDataListDropdownManagerDirective.prototype, "elements", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiDataListDropdownManagerDirective.prototype, "elements$", null);
        __decorate([
            cdk.tuiPure
        ], TuiDataListDropdownManagerDirective.prototype, "right$", null);
        __decorate([
            cdk.tuiPure
        ], TuiDataListDropdownManagerDirective.prototype, "immediate$", null);
        __decorate([
            cdk.tuiPure
        ], TuiDataListDropdownManagerDirective.prototype, "debounce$", null);
        TuiDataListDropdownManagerDirective = __decorate([
            core.Directive({
                selector: 'tui-data-list[tuiDataListDropdownManager]',
            })
        ], TuiDataListDropdownManagerDirective);
        return TuiDataListDropdownManagerDirective;
    }());

    var TuiOptGroupDirective = /** @class */ (function () {
        function TuiOptGroupDirective() {
            this.label = '';
        }
        __decorate([
            core.Input(),
            core.HostBinding('attr.data-label'),
            cdk.tuiDefaultProp()
        ], TuiOptGroupDirective.prototype, "label", void 0);
        TuiOptGroupDirective = __decorate([
            core.Directive({
                selector: 'tui-opt-group',
                host: {
                    role: 'group',
                },
            })
        ], TuiOptGroupDirective);
        return TuiOptGroupDirective;
    }());

    var TuiDataListModule = /** @class */ (function () {
        function TuiDataListModule() {
        }
        TuiDataListModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, svg.TuiSvgModule, ngPolymorpheus.PolymorpheusModule],
                declarations: [
                    TuiDataListComponent,
                    TuiOptionComponent,
                    TuiOptGroupDirective,
                    TuiDataListDirective,
                    TuiDataListDropdownManagerDirective,
                ],
                exports: [
                    TuiDataListComponent,
                    TuiOptionComponent,
                    TuiOptGroupDirective,
                    TuiDataListDirective,
                    TuiDataListDropdownManagerDirective,
                ],
            })
        ], TuiDataListModule);
        return TuiDataListModule;
    }());

    exports.TuiDataListComponent = TuiDataListComponent;
    exports.TuiDataListDirective = TuiDataListDirective;
    exports.TuiDataListDropdownManagerDirective = TuiDataListDropdownManagerDirective;
    exports.TuiDataListModule = TuiDataListModule;
    exports.TuiOptGroupDirective = TuiOptGroupDirective;
    exports.TuiOptionComponent = TuiOptionComponent;
    exports.shouldFocus = shouldFocus;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
