// @taiga-ui/core/bundles/taiga-ui-core-components-hints-host.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/directives'), require('@taiga-ui/core/directives/pointer-hint'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/services'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/hints-host', ['exports', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/animations', '@taiga-ui/core/directives', '@taiga-ui/core/directives/pointer-hint', '@taiga-ui/core/tokens', '@taiga-ui/core/utils', '@tinkoff/ng-polymorpheus', 'rxjs', 'rxjs/operators', '@taiga-ui/core/services', '@angular/common'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components['hints-host'] = {}), global.ng.core, global.common, global.cdk, global['taiga-ui'].core.animations, global['taiga-ui'].core.directives, global['taiga-ui'].core.directives['pointer-hint'], global['taiga-ui'].core.tokens, global['taiga-ui'].core.utils, global.ngPolymorpheus, global.rxjs, global.rxjs.operators, global['taiga-ui'].core.services, global.ng.common));
}(this, (function (exports, core, common, cdk, animations, directives, pointerHint, tokens, utils, ngPolymorpheus, rxjs, operators, services, common$1) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TuiHintsHostComponent = /** @class */ (function () {
        function TuiHintsHostComponent(elementRef, hints$) {
            this.elementRef = elementRef;
            this.hints$ = hints$;
            this.component = new ngPolymorpheus.PolymorpheusComponent(TuiHintBoxComponent);
        }
        Object.defineProperty(TuiHintsHostComponent.prototype, "clientRect", {
            get: function () {
                return this.elementRef.nativeElement.getBoundingClientRect();
            },
            enumerable: true,
            configurable: true
        });
        TuiHintsHostComponent.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [services.TuiHintService,] }] }
        ]; };
        TuiHintsHostComponent = __decorate([
            core.Component({
                selector: 'tui-hints-host',
                template: "<div\n    *ngFor=\"let hint of hints$ | async\"\n    role=\"tooltip\"\n    polymorpheus-outlet\n    @tuiParentAnimation\n    [attr.id]=\"hint.id\"\n    [content]=\"hint.component || component\"\n    [context]=\"{$implicit: hint}\"\n    [tuiActiveZoneParent]=\"hint.activeZone\"\n></div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                animations: [cdk.TUI_PARENT_ANIMATION],
                host: {
                    'aria-live': 'polite',
                },
                styles: [":host{position:fixed;bottom:0;left:0;width:100%;height:0}"]
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(services.TuiHintService))
        ], TuiHintsHostComponent);
        return TuiHintsHostComponent;
    }());

    var SPACE = 8;
    var BORDER_WIDTH = 1;
    var LEFT_PADDING = 16;
    var TOP_PADDING = 12;
    var ARROW_SIZE = 8;
    var ARROW_OFFSET = 16;
    var ARROWHEAD_OFFSET = ARROW_OFFSET + (ARROW_SIZE * Math.sqrt(2)) / 2;
    var reverseDirectionsVertical = {
        'top-left': 'bottom-left',
        'top-right': 'bottom-right',
        'bottom-left': 'top-left',
        'bottom-right': 'top-right',
        left: 'right',
        right: 'left',
        'top-middle': 'bottom-middle',
        'bottom-middle': 'top-middle',
    };
    var reverseDirectionsHorizontal = {
        'top-left': 'top-right',
        'top-right': 'top-left',
        'bottom-left': 'bottom-right',
        'bottom-right': 'bottom-left',
        left: 'right',
        right: 'left',
        'top-middle': 'top-middle',
        'bottom-middle': 'bottom-middle',
    };
    // TODO: consider abstracting UI and move to CDK, split hint and overflow
    // Ambient type cannot be used without dynamic https://github.com/angular/angular/issues/23395
    // @dynamic
    var TuiHintBoxComponent = /** @class */ (function () {
        function TuiHintBoxComponent(animationFrame$, destroy$, ngZone, options, elementRef, windowRef, media, hintsHost, context) {
            var _this = this;
            this.options = options;
            this.elementRef = elementRef;
            this.windowRef = windowRef;
            this.media = media;
            this.hintsHost = hintsHost;
            this.context = context;
            this.animation = __assign({ value: '' }, this.options);
            animationFrame$
                .pipe(cdk.tuiZonefree(ngZone), operators.takeUntil(destroy$))
                .subscribe(function () { return _this.calculatePosition(); });
        }
        Object.defineProperty(TuiHintBoxComponent.prototype, "hint", {
            get: function () {
                return this.context.$implicit;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiHintBoxComponent.prototype, "isUntouchable", {
            get: function () {
                return this.hint instanceof pointerHint.TuiPointerHintDirective;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiHintBoxComponent.prototype, "mode", {
            get: function () {
                return this.hint.mode;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiHintBoxComponent.prototype, "isMobile", {
            get: function () {
                return utils.tuiIsMobile(this.windowRef, this.media);
            },
            enumerable: true,
            configurable: true
        });
        TuiHintBoxComponent.prototype.onHovered = function (hovered) {
            if (this.hint instanceof directives.TuiHintDirective) {
                this.hint.componentHovered$.next(hovered);
            }
        };
        /**
         * Calculates wrapper position.
         * Styles are set directly to avoid visual shake of element
         */
        TuiHintBoxComponent.prototype.calculatePosition = function () {
            if (this.mode !== 'overflow') {
                this.calculateCoordinates();
            }
            else {
                this.setOverflowStyles();
            }
        };
        TuiHintBoxComponent.prototype.calculateCoordinates = function () {
            if (this.isMobile) {
                this.calculateMobileCoordinates();
                return;
            }
            var hostRect = this.hint.getElementClientRect();
            var portalRect = this.hintsHost.clientRect;
            var tooltip = this.elementRef.nativeElement;
            var style = tooltip.style;
            var tooltipRect = tooltip.getBoundingClientRect();
            var isHostLong = hostRect.width > ARROWHEAD_OFFSET * 2;
            var directions = [
                'left',
                'right',
                'bottom-left',
                'bottom-right',
                'bottom-middle',
                'top-left',
                'top-right',
                'top-middle',
            ];
            var top = 0;
            var left = 0;
            var direction = this.hint.direction;
            var horizontalTop = hostRect.top + hostRect.height / 2 - tooltipRect.height / 2 - portalRect.top;
            var horizontalLeft = hostRect.left - tooltipRect.width - SPACE - portalRect.left;
            var horizontalRight = hostRect.left + hostRect.width + SPACE - portalRect.left;
            var verticalBottom = hostRect.bottom + SPACE - portalRect.top;
            var verticalTop = hostRect.top - tooltipRect.height - SPACE - portalRect.top;
            var verticalRight = isHostLong
                ? hostRect.left - portalRect.left
                : hostRect.left + hostRect.width / 2 - ARROWHEAD_OFFSET - portalRect.left;
            var verticalLeft = isHostLong
                ? hostRect.left - tooltipRect.width + hostRect.width - portalRect.left
                : hostRect.left -
                    tooltipRect.width +
                    hostRect.width / 2 +
                    ARROWHEAD_OFFSET -
                    portalRect.left;
            var verticalMiddle = hostRect.left - tooltipRect.width / 2 + hostRect.width / 2 - portalRect.left;
            directions.splice(directions.indexOf(direction), 1);
            // eslint-disable-next-line no-constant-condition
            while (true) {
                switch (direction) {
                    case 'left':
                        top = horizontalTop;
                        left = horizontalLeft;
                        break;
                    case 'right':
                        top = horizontalTop;
                        left = horizontalRight;
                        break;
                    case 'top-right':
                        top = verticalTop;
                        left = verticalRight;
                        break;
                    case 'top-left':
                        top = verticalTop;
                        left = verticalLeft;
                        break;
                    case 'bottom-right':
                        top = verticalBottom;
                        left = verticalRight;
                        break;
                    case 'bottom-left':
                        top = verticalBottom;
                        left = verticalLeft;
                        break;
                    case 'bottom-middle':
                        top = verticalBottom;
                        left = verticalMiddle;
                        break;
                    case 'top-middle':
                        top = verticalTop;
                        left = verticalMiddle;
                        break;
                }
                var verticalFit = top + portalRect.top > SPACE &&
                    top + tooltipRect.height + SPACE + portalRect.top <
                        this.windowRef.innerHeight;
                var horizontalFit = left > SPACE &&
                    left + tooltipRect.width + SPACE + portalRect.left < portalRect.width;
                if (directions.length === 0 || (verticalFit && horizontalFit)) {
                    break;
                }
                direction = verticalFit
                    ? reverseDirectionsHorizontal[direction]
                    : reverseDirectionsVertical[direction];
                direction =
                    directions.splice(directions.indexOf(direction), 1)[0] || direction;
            }
            style.top = cdk.px(top);
            style.left = cdk.px(left);
            tooltip.setAttribute('data-tui-host-direction', direction);
        };
        TuiHintBoxComponent.prototype.calculateMobileCoordinates = function () {
            var hostRect = this.hint.getElementClientRect();
            var portalRect = this.hintsHost.clientRect;
            var tooltip = this.elementRef.nativeElement;
            var style = tooltip.style;
            var tooltipRect = tooltip.getBoundingClientRect();
            var verticalTop = hostRect.top - tooltipRect.height - SPACE - portalRect.top;
            var verticalBottom = hostRect.bottom + SPACE - portalRect.top;
            var verticalTopFit = verticalTop + portalRect.top > SPACE &&
                hostRect.top < this.windowRef.innerHeight;
            var verticalBottomFit = hostRect.bottom > 0 &&
                hostRect.bottom + 2 * SPACE + tooltipRect.height < this.windowRef.innerHeight;
            var direction = (this.hint.direction.includes('top') && verticalTopFit) || !verticalBottomFit
                ? 'top'
                : 'bottom';
            var attemptedLeft = portalRect.left + hostRect.left + hostRect.width / 2 - tooltipRect.width / 2;
            var left = Math.max(attemptedLeft + tooltipRect.width + SPACE > portalRect.right
                ? portalRect.right - SPACE - tooltipRect.width
                : attemptedLeft, SPACE * 2);
            style.left = cdk.px(left);
            style.top = direction === 'top' ? cdk.px(verticalTop) : cdk.px(verticalBottom);
            if (this.arrow) {
                this.arrow.nativeElement.style.left = cdk.px(hostRect.left <= SPACE * 2 && hostRect.width > ARROW_OFFSET * 2
                    ? ARROW_OFFSET
                    : hostRect.left + hostRect.width / 2 - left - ARROW_SIZE / 2);
            }
            tooltip.setAttribute('data-tui-host-direction', direction);
        };
        TuiHintBoxComponent.prototype.setOverflowStyles = function () {
            var hostRect = this.hint.getElementClientRect();
            var style = this.elementRef.nativeElement.style;
            style.top = cdk.px(hostRect.top - window.innerHeight - TOP_PADDING - BORDER_WIDTH);
            style.left = cdk.px(hostRect.left - LEFT_PADDING - BORDER_WIDTH);
            style.width = cdk.px(hostRect.width + LEFT_PADDING * 2 + BORDER_WIDTH * 2);
        };
        TuiHintBoxComponent.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common.ANIMATION_FRAME,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_ANIMATION_OPTIONS,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common.WINDOW,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_MEDIA,] }] },
            { type: TuiHintsHostComponent, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return TuiHintsHostComponent; }),] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [ngPolymorpheus.POLYMORPHEUS_CONTEXT,] }] }
        ]; };
        __decorate([
            core.ViewChild('arrow')
        ], TuiHintBoxComponent.prototype, "arrow", void 0);
        __decorate([
            core.HostBinding('@tuiFadeIn')
        ], TuiHintBoxComponent.prototype, "animation", void 0);
        __decorate([
            cdk.tuiPure,
            core.HostBinding('class._untouchable')
        ], TuiHintBoxComponent.prototype, "isUntouchable", null);
        __decorate([
            core.HostBinding('attr.data-mode')
        ], TuiHintBoxComponent.prototype, "mode", null);
        TuiHintBoxComponent = __decorate([
            core.Component({
                selector: 'tui-hint-box[hint]',
                template: "<!-- @bad TODO: Refactor to use this arrow in all cases -->\n<div\n    *ngIf=\"isMobile && mode !== 'overflow'\"\n    #arrow\n    class=\"t-arrow\"\n></div>\n<div\n    automation-id=\"tui-hint-box__tooltip\"\n    polymorpheus-outlet\n    class=\"t-tooltip\"\n    [content]=\"hint.content\"\n    (tuiHoveredChange)=\"onHovered($event)\"\n>\n    <ng-template let-text>\n        <span class=\"t-text\">{{ text }}</span>\n    </ng-template>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [cdk.TuiDestroyService],
                animations: [animations.tuiFadeIn],
                styles: [":host{position:absolute;top:0;left:0;max-width:18rem;min-height:var(--tui-height-m);padding:.75rem 1rem;background:var(--tui-primary);border-radius:var(--tui-radius-l);color:var(--tui-primary-text);box-sizing:border-box}:host .t-arrow,:host:not([data-mode=overflow]):before{content:'';position:absolute;width:.5rem;height:.5rem;border-radius:.125rem;box-sizing:border-box;background-color:inherit;transform:rotate(45deg)}:host[data-tui-host-direction=top] .t-arrow{bottom:-.25rem}:host[data-tui-host-direction=bottom] .t-arrow{top:-.25rem}:host[data-tui-host-direction=bottom]:before,:host[data-tui-host-direction=top]:before{display:none}:host[data-tui-host-direction=top-left]:before,:host[data-tui-host-direction=top-middle]:before,:host[data-tui-host-direction=top-right]:before{bottom:-.25rem}:host[data-tui-host-direction=bottom-left]:before,:host[data-tui-host-direction=bottom-middle]:before,:host[data-tui-host-direction=bottom-right]:before{top:-.25rem}:host[data-tui-host-direction=bottom-middle]:before,:host[data-tui-host-direction=top-middle]:before{left:calc(50% - .2rem)}:host[data-tui-host-direction=bottom-left]:before,:host[data-tui-host-direction=top-left]:before{right:1.075rem}:host[data-tui-host-direction=bottom-right]:before,:host[data-tui-host-direction=top-right]:before{left:1.075rem}:host[data-tui-host-direction=left]:before,:host[data-tui-host-direction=right]:before{top:50%;margin-top:-.25rem}:host[data-tui-host-direction=left]:before{right:-.25rem}:host[data-tui-host-direction=right]:before{left:-.25rem}:host[data-mode=error]{background-color:var(--tui-error-fill)}:host[data-mode=onDark],:host[data-mode=overflow]{box-shadow:0 .5rem 1rem rgba(0,0,0,.16);border:1px solid var(--tui-base-03);background-color:var(--tui-base-01);color:var(--tui-text-01)}:host[data-mode=onDark]:before,:host[data-mode=overflow]:before{border:1px solid var(--tui-base-03)}:host[data-mode=onDark][data-tui-host-direction=left]:before,:host[data-mode=overflow][data-tui-host-direction=left]:before{border-left-color:transparent;border-bottom-color:transparent}:host[data-mode=onDark][data-tui-host-direction=right]:before,:host[data-mode=overflow][data-tui-host-direction=right]:before{border-right-color:transparent;border-top-color:transparent}:host[data-mode=onDark][data-tui-host-direction=top-left]:before,:host[data-mode=onDark][data-tui-host-direction=top-middle]:before,:host[data-mode=onDark][data-tui-host-direction=top-right]:before,:host[data-mode=overflow][data-tui-host-direction=top-left]:before,:host[data-mode=overflow][data-tui-host-direction=top-middle]:before,:host[data-mode=overflow][data-tui-host-direction=top-right]:before{border-left-color:transparent;border-top-color:transparent}:host[data-mode=onDark][data-tui-host-direction=bottom-left]:before,:host[data-mode=onDark][data-tui-host-direction=bottom-middle]:before,:host[data-mode=onDark][data-tui-host-direction=bottom-right]:before,:host[data-mode=overflow][data-tui-host-direction=bottom-left]:before,:host[data-mode=overflow][data-tui-host-direction=bottom-middle]:before,:host[data-mode=overflow][data-tui-host-direction=bottom-right]:before{border-right-color:transparent;border-bottom-color:transparent}:host[data-mode=overflow]{max-width:none}:host._untouchable{pointer-events:none}.t-tooltip{font:var(--tui-font-text-s);word-wrap:break-word}:host[data-mode=overflow] .t-tooltip{font:inherit}.t-text{white-space:pre-wrap}"]
            }),
            __param(0, core.Inject(common.ANIMATION_FRAME)),
            __param(1, core.Inject(cdk.TuiDestroyService)),
            __param(2, core.Inject(core.NgZone)),
            __param(3, core.Inject(tokens.TUI_ANIMATION_OPTIONS)),
            __param(4, core.Inject(core.ElementRef)),
            __param(5, core.Inject(common.WINDOW)),
            __param(6, core.Inject(tokens.TUI_MEDIA)),
            __param(7, core.Inject(core.forwardRef(function () { return TuiHintsHostComponent; }))),
            __param(8, core.Inject(ngPolymorpheus.POLYMORPHEUS_CONTEXT))
        ], TuiHintBoxComponent);
        return TuiHintBoxComponent;
    }());

    var TuiHintBoxModule = /** @class */ (function () {
        function TuiHintBoxModule() {
        }
        TuiHintBoxModule = __decorate([
            core.NgModule({
                imports: [common$1.CommonModule, ngPolymorpheus.PolymorpheusModule, cdk.TuiHoveredModule],
                declarations: [TuiHintBoxComponent],
                exports: [TuiHintBoxComponent],
                entryComponents: [TuiHintBoxComponent],
            })
        ], TuiHintBoxModule);
        return TuiHintBoxModule;
    }());

    var TuiHintsHostModule = /** @class */ (function () {
        function TuiHintsHostModule() {
        }
        TuiHintsHostModule = __decorate([
            core.NgModule({
                imports: [common$1.CommonModule, ngPolymorpheus.PolymorpheusModule, TuiHintBoxModule, cdk.TuiActiveZoneModule],
                declarations: [TuiHintsHostComponent],
                exports: [TuiHintsHostComponent],
            })
        ], TuiHintsHostModule);
        return TuiHintsHostModule;
    }());

    exports.TuiHintBoxComponent = TuiHintBoxComponent;
    exports.TuiHintBoxModule = TuiHintBoxModule;
    exports.TuiHintsHostComponent = TuiHintsHostComponent;
    exports.TuiHintsHostModule = TuiHintsHostModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
