// @taiga-ui/core/bundles/taiga-ui-core-components-dropdown-box.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/constants'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils/dom'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/components/scrollbar'), require('@taiga-ui/core/directives/mode'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/dropdown-box', ['exports', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/animations', '@taiga-ui/core/constants', '@taiga-ui/core/tokens', '@taiga-ui/core/utils/dom', 'rxjs', 'rxjs/operators', '@taiga-ui/core/components/scrollbar', '@taiga-ui/core/directives/mode', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components['dropdown-box'] = {}), global.ng.core, global.common, global.cdk, global['taiga-ui'].core.animations, global['taiga-ui'].core.constants, global['taiga-ui'].core.tokens, global['taiga-ui'].core.utils.dom, global.rxjs, global.rxjs.operators, global['taiga-ui'].core.components.scrollbar, global['taiga-ui'].core.directives.mode, global.ngPolymorpheus));
}(this, (function (exports, core, common, cdk, animations, constants, tokens, dom, rxjs, operators, scrollbar, mode, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    /**
     *  This component is used to show template in a portal using default style of white rounded box with a shadow
     */
    // @bad TODO: OnPush
    // Ambient type cannot be used without dynamic https://github.com/angular/angular/issues/23395
    // @dynamic
    var TuiDropdownBoxComponent = /** @class */ (function () {
        function TuiDropdownBoxComponent(destroy$, ngZone, directive, windowRef, elementRef, portalHost, options, animationFrame$) {
            var _this = this;
            this.directive = directive;
            this.windowRef = windowRef;
            this.elementRef = elementRef;
            this.portalHost = portalHost;
            this.options = options;
            this.animationTop = __assign({ value: "fadeInTop" /* FadeInTop */ }, this.options);
            this.animationBottom = __assign({ value: "fadeInBottom" /* FadeInBottom */ }, this.options);
            /**
             * Is previous position on top (to prevent jumping up and down on scroll)
             */
            this.prevDirectionIsTop = false;
            rxjs.merge(animationFrame$.pipe(operators.throttleTime(cdk.POLLING_TIME)), this.directive.refresh$, rxjs.fromEvent(this.windowRef, 'resize'))
                .pipe(cdk.tuiZonefree(ngZone), operators.takeUntil(destroy$))
                .subscribe(function () {
                _this.calculatePositionAndSize();
            });
        }
        Object.defineProperty(TuiDropdownBoxComponent.prototype, "overscroll", {
            get: function () {
                return this.inModal ? 'all' : 'scroll';
            },
            enumerable: true,
            configurable: true
        });
        TuiDropdownBoxComponent.prototype.getContext = function (context, activeZone) {
            return __assign(__assign({}, context), { activeZone: activeZone });
        };
        TuiDropdownBoxComponent.prototype.ngAfterViewChecked = function () {
            this.calculatePositionAndSize();
        };
        TuiDropdownBoxComponent.prototype.onTopFocus = function () {
            this.moveFocusOutside(true);
        };
        TuiDropdownBoxComponent.prototype.onBottomFocus = function () {
            this.moveFocusOutside(false);
        };
        Object.defineProperty(TuiDropdownBoxComponent.prototype, "inModal", {
            get: function () {
                // @awful TODO: get rid of component tag name dependency
                return !!this.directive.host.closest('tui-dialog-host');
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDropdownBoxComponent.prototype, "inOption", {
            get: function () {
                // @awful TODO: get rid of component tag name dependency
                return !!this.directive.host.closest('[tuiOption]');
            },
            enumerable: true,
            configurable: true
        });
        TuiDropdownBoxComponent.prototype.calculatePositionAndSize = function () {
            var clientRect = this.directive.clientRect;
            var style = this.elementRef.nativeElement.style;
            var hostRect = this.directive.fixed
                ? this.portalHost.fixedPositionOffset()
                : this.portalHost.clientRect;
            style.position = this.directive.fixed ? 'fixed' : 'absolute';
            this.calculateVerticalPosition(style, clientRect, hostRect);
            this.calculateHorizontalPosition(style, clientRect, hostRect);
            this.calculateWidth(style, clientRect);
        };
        TuiDropdownBoxComponent.prototype.getFinalAlign = function (style, directiveRect) {
            var dropdownRect = this.elementRef.nativeElement.getBoundingClientRect();
            var dropdownWidth = this.elementRef.nativeElement.offsetWidth;
            var screenWidth = dom.getScreenWidth(this.windowRef.document);
            var isDropdownSizeHypotheticallyFitsViewport = directiveRect.left + dropdownWidth < screenWidth ||
                directiveRect.right - dropdownWidth > 0;
            var isDropdownSizeActuallyFitsViewport = dropdownRect.right <= screenWidth && dropdownRect.left >= 0;
            var finalAlign = this.directive.align;
            switch (this.directive.align) {
                case 'left':
                    if (isDropdownSizeHypotheticallyFitsViewport &&
                        dropdownRect.right > screenWidth) {
                        finalAlign = 'right';
                    }
                    break;
                case 'right':
                    if (isDropdownSizeHypotheticallyFitsViewport && dropdownRect.left < 0) {
                        finalAlign = 'left';
                    }
                    break;
            }
            if (style.right === 'auto' && isDropdownSizeActuallyFitsViewport) {
                finalAlign = 'left';
            }
            if (style.left === 'auto' && isDropdownSizeActuallyFitsViewport) {
                finalAlign = 'right';
            }
            return finalAlign;
        };
        /**
         * Calculates horizontal position
         *
         * @param style dropdownBox elementRef styles object
         * @param directiveRect ClientRect of hosting directive
         * @param hostRect ClientRect of  portal host
         */
        TuiDropdownBoxComponent.prototype.calculateHorizontalPosition = function (style, directiveRect, hostRect) {
            var offset = this.directive.sided
                ? this.elementRef.nativeElement.getBoundingClientRect().width + constants.DEFAULT_MARGIN
                : 0;
            var left = Math.ceil(directiveRect.left - hostRect.left - offset);
            var right = Math.floor(hostRect.right - directiveRect.right - offset);
            var viewportWidth = dom.tuiGetViewportWidth(this.windowRef);
            switch (this.getFinalAlign(style, directiveRect)) {
                case 'left':
                    if (right + constants.DEFAULT_MARGIN > viewportWidth ||
                        cdk.inRange(left + constants.DEFAULT_MARGIN, 0, viewportWidth)) {
                        style.left = cdk.px(left);
                        style.right = 'auto';
                    }
                    else {
                        style.left = 'auto';
                        style.right = cdk.px(right);
                    }
                    break;
                case 'right':
                    if (cdk.inRange(right + constants.DEFAULT_MARGIN, 0, viewportWidth) ||
                        left + constants.DEFAULT_MARGIN > viewportWidth) {
                        style.left = 'auto';
                        style.right = cdk.px(right);
                    }
                    else {
                        style.left = cdk.px(left);
                        style.right = 'auto';
                    }
                    break;
            }
        };
        /**
         * Calculates vertical position and height
         *
         * @param style dropdownBox elementRef styles object
         * @param directiveRect ClientRect of hosting directive
         * @param hostRect ClientRect of  portal host
         */
        TuiDropdownBoxComponent.prototype.calculateVerticalPosition = function (style, directiveRect, hostRect) {
            var windowHeight = this.windowRef.innerHeight;
            // Maximum height of the box
            var boxHeightLimit = Math.min(this.directive.maxHeight, windowHeight - constants.DEFAULT_MARGIN * 2);
            var offset = this.directive.sided
                ? constants.DEFAULT_MARGIN - directiveRect.height
                : constants.DEFAULT_MARGIN * 2;
            var topAvailableHeight = directiveRect.top - offset;
            var bottomAvailableHeight = windowHeight - directiveRect.bottom - offset;
            var finalDirection = this.getFinalDirection(directiveRect);
            var optionOffset = this.inOption ? constants.DEFAULT_MARGIN * 2 : 0;
            this.prevDirectionIsTop = finalDirection === 'top';
            if (finalDirection === 'top') {
                this.dropdownAnimation = this.animationBottom;
                style.maxHeight = cdk.px(Math.min(boxHeightLimit, topAvailableHeight));
                style.top = 'auto';
                style.bottom = cdk.px(hostRect.bottom -
                    directiveRect.top -
                    constants.DEFAULT_MARGIN +
                    offset -
                    optionOffset);
            }
            else {
                this.dropdownAnimation = this.animationTop;
                style.maxHeight = cdk.px(Math.min(boxHeightLimit, bottomAvailableHeight));
                style.top = cdk.px(directiveRect.bottom -
                    hostRect.top -
                    constants.DEFAULT_MARGIN +
                    offset -
                    optionOffset);
                style.bottom = 'auto';
            }
        };
        TuiDropdownBoxComponent.prototype.getFinalDirection = function (directiveRect) {
            var windowHeight = this.windowRef.innerHeight;
            var offset = this.directive.sided
                ? constants.DEFAULT_MARGIN - directiveRect.height
                : constants.DEFAULT_MARGIN * 2;
            // Maximum space available on top and on the bottom in the viewport
            var topAvailableHeight = directiveRect.top - offset;
            var bottomAvailableHeight = windowHeight - directiveRect.bottom - offset;
            var finalDirection = null;
            // Given direction is applied if we can fit the box in the limits that way
            switch (this.directive.direction) {
                case 'top':
                    if (topAvailableHeight >= this.directive.minHeight) {
                        finalDirection = 'top';
                    }
                    break;
                case 'bottom':
                    if (bottomAvailableHeight >= this.directive.minHeight) {
                        finalDirection = 'bottom';
                    }
                    break;
            }
            // Maximum height of the box
            var boxHeightLimit = Math.min(this.directive.maxHeight, windowHeight - constants.DEFAULT_MARGIN * 2);
            // Choose direction if given direction did not fit
            if (finalDirection === null && this.contentElementRef) {
                // Box height if it fits without scroll
                var visualHeight = Math.min(this.contentElementRef.nativeElement.getBoundingClientRect().height +
                    (this.elementRef.nativeElement.offsetHeight -
                        this.elementRef.nativeElement.clientHeight), boxHeightLimit);
                // If there is enough space to fit below without scroll,
                // choose 'bottom', unless it was previously on the top
                if (this.prevDirectionIsTop && topAvailableHeight >= visualHeight) {
                    finalDirection = 'top';
                }
                else if (bottomAvailableHeight >= visualHeight) {
                    finalDirection = 'bottom';
                }
                else {
                    // Corner case — select direction with more space
                    finalDirection =
                        bottomAvailableHeight >= topAvailableHeight ? 'bottom' : 'top';
                }
            }
            return finalDirection;
        };
        /**
         * Calculates width
         *
         * @param style dropdownBox elementRef styles object
         * @param directiveRect ClientRect of hosting directive
         */
        TuiDropdownBoxComponent.prototype.calculateWidth = function (style, directiveRect) {
            style.width =
                this.directive.limitMinWidth === 'fixed' && !this.directive.sided
                    ? cdk.px(directiveRect.width)
                    : '';
            if (this.directive.limitMinWidth === 'min' && !this.directive.sided) {
                style.minWidth = cdk.px(directiveRect.width);
                style.maxWidth = cdk.px(constants.DEFAULT_MAX_WIDTH);
                return;
            }
            style.minWidth = '';
            style.maxWidth = '';
        };
        TuiDropdownBoxComponent.prototype.moveFocusOutside = function (previous) {
            var host = this.directive.host;
            var ownerDocument = host.ownerDocument;
            var root = ownerDocument ? ownerDocument.body : host;
            cdk.tuiAssertIsHTMLElement(host);
            var focusable = cdk.getClosestFocusable(host, previous, root);
            while (focusable !== null && host.contains(focusable)) {
                focusable = cdk.getClosestFocusable(focusable, previous, root);
            }
            focusable === null || focusable === void 0 ? void 0 : focusable.focus();
        };
        TuiDropdownBoxComponent.ctorParameters = function () { return [
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_DROPDOWN_DIRECTIVE,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common.WINDOW,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: cdk.TuiDropdownHostComponent, decorators: [{ type: core.Inject, args: [cdk.AbstractTuiPortalHostComponent,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_ANIMATION_OPTIONS,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common.ANIMATION_FRAME,] }] }
        ]; };
        __decorate([
            core.HostBinding('@tuiDropdownAnimation')
        ], TuiDropdownBoxComponent.prototype, "dropdownAnimation", void 0);
        __decorate([
            core.ViewChild('content', { read: core.ElementRef })
        ], TuiDropdownBoxComponent.prototype, "contentElementRef", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiDropdownBoxComponent.prototype, "getContext", null);
        __decorate([
            cdk.tuiPure
        ], TuiDropdownBoxComponent.prototype, "inModal", null);
        __decorate([
            cdk.tuiPure
        ], TuiDropdownBoxComponent.prototype, "inOption", null);
        TuiDropdownBoxComponent = __decorate([
            core.Component({
                selector: 'tui-dropdown-box',
                template: "<div\n    #activeZone=\"tuiActiveZone\"\n    class=\"t-wrapper\"\n    [tuiOverscroll]=\"overscroll\"\n    [tuiMode]=\"null\"\n    [tuiActiveZoneParent]=\"directive.activeZone || null\"\n>\n    <tui-scrollbar class=\"t-scroll\">\n        <div\n            tabindex=\"0\"\n            (focus)=\"onTopFocus()\"\n        ></div>\n        <div\n            #content\n            polymorpheus-outlet\n            class=\"t-content\"\n            [content]=\"directive.content\"\n            [context]=\"getContext(directive.context, activeZone)\"\n        ></div>\n        <div\n            tabindex=\"0\"\n            (focus)=\"onBottomFocus()\"\n        ></div>\n    </tui-scrollbar>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.Default,
                providers: [cdk.TuiDestroyService],
                animations: [animations.tuiDropdownAnimation],
                styles: [":host{z-index:0;box-shadow:0 .5rem 1rem rgba(0,0,0,.16);position:absolute;top:0;left:0;display:flex;background-color:#fff;background-color:var(--tui-elevation-01);border-radius:var(--tui-radius-m);overflow:hidden;border:1px solid var(--tui-base-03);box-sizing:border-box}:host.ng-animating{pointer-events:none}.t-content{display:flex;flex-direction:column;max-height:100%}.t-wrapper{flex-grow:1;max-width:100%;max-height:inherit;overflow:visible}.t-scroll{height:100%}"]
            }),
            __param(0, core.Inject(cdk.TuiDestroyService)),
            __param(1, core.Inject(core.NgZone)),
            __param(2, core.Inject(tokens.TUI_DROPDOWN_DIRECTIVE)),
            __param(3, core.Inject(common.WINDOW)),
            __param(4, core.Inject(core.ElementRef)),
            __param(5, core.Inject(cdk.AbstractTuiPortalHostComponent)),
            __param(6, core.Inject(tokens.TUI_ANIMATION_OPTIONS)),
            __param(7, core.Inject(common.ANIMATION_FRAME))
        ], TuiDropdownBoxComponent);
        return TuiDropdownBoxComponent;
    }());

    var TuiDropdownBoxModule = /** @class */ (function () {
        function TuiDropdownBoxModule() {
        }
        TuiDropdownBoxModule = __decorate([
            core.NgModule({
                imports: [
                    cdk.TuiActiveZoneModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiOverscrollModule,
                    scrollbar.TuiScrollbarModule,
                    mode.TuiModeModule,
                ],
                entryComponents: [TuiDropdownBoxComponent],
                declarations: [TuiDropdownBoxComponent],
                exports: [TuiDropdownBoxComponent],
            })
        ], TuiDropdownBoxModule);
        return TuiDropdownBoxModule;
    }());

    exports.TuiDropdownBoxComponent = TuiDropdownBoxComponent;
    exports.TuiDropdownBoxModule = TuiDropdownBoxModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
