// @taiga-ui/core/esm5/components/error/error.component.js
import { __assign, __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { tuiDefaultProp, TuiValidationError } from '@taiga-ui/cdk';
import { tuiFadeIn, tuiHeightCollapse } from '@taiga-ui/core/animations';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_ANIMATION_OPTIONS, TUI_DEFAULT_ERROR_MESSAGE, TUI_MODE, } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
var TuiErrorComponent = /** @class */ (function () {
    function TuiErrorComponent(options, mode$, defaultErrorMessage$) {
        this.options = options;
        this.mode$ = mode$;
        this.defaultErrorMessage$ = defaultErrorMessage$;
        this.error = null;
        this.animation = __assign({ value: '' }, this.options);
    }
    Object.defineProperty(TuiErrorComponent.prototype, "errorSetter", {
        set: function (error) {
            this.error = typeof error === 'string' ? new TuiValidationError(error) : error;
        },
        enumerable: true,
        configurable: true
    });
    TuiErrorComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ANIMATION_OPTIONS,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_DEFAULT_ERROR_MESSAGE,] }] }
    ]; };
    __decorate([
        Input('error'),
        tuiDefaultProp()
    ], TuiErrorComponent.prototype, "errorSetter", null);
    TuiErrorComponent = __decorate([
        Component({
            selector: 'tui-error',
            template: "<div\n    *ngIf=\"error\"\n    polymorpheus-outlet\n    automation-id=\"tui-error__text\"\n    class=\"t-message-text\"\n    [@tuiHeightCollapse]=\"animation\"\n    [@tuiFadeIn]=\"animation\"\n    [class.t-message-text_light]=\"(mode$ | async) === 'onDark'\"\n    [content]=\"error.message || (defaultErrorMessage$ | async)\"\n    [context]=\"error.context || {}\"\n></div>\n",
            providers: [MODE_PROVIDER],
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiHeightCollapse, tuiFadeIn],
            styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);display:block;color:var(--tui-negative);word-wrap:break-word}.t-message-text{margin-top:.25rem;white-space:pre-wrap}.t-message-text_light{color:var(--tui-negative-night)}"]
        }),
        __param(0, Inject(TUI_ANIMATION_OPTIONS)),
        __param(1, Inject(TUI_MODE)),
        __param(2, Inject(TUI_DEFAULT_ERROR_MESSAGE))
    ], TuiErrorComponent);
    return TuiErrorComponent;
}());
export { TuiErrorComponent };
