// @taiga-ui/kit/esm5/components/rating/rating.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, forwardRef, HostListener, Inject, Input, Optional, Self, ViewChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiControl, isNativeFocused, TUI_FOCUSABLE_ITEM_ACCESSOR, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TUI_RATING_OPTIONS } from './rating-options';
var TuiRatingComponent = /** @class */ (function (_super) {
    __extends(TuiRatingComponent, _super);
    function TuiRatingComponent(ngControl, changeDetectorRef, options) {
        var _this = _super.call(this, ngControl, changeDetectorRef) || this;
        _this.options = options;
        _this.min = _this.options.min;
        _this.max = _this.options.max;
        _this.iconNormal = _this.options.iconNormal;
        _this.iconFilled = _this.options.iconFilled;
        return _this;
    }
    TuiRatingComponent_1 = TuiRatingComponent;
    Object.defineProperty(TuiRatingComponent.prototype, "nativeFocusableElement", {
        get: function () {
            return this.computedDisabled || !this.focusableElement
                ? null
                : this.focusableElement.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRatingComponent.prototype, "focused", {
        get: function () {
            return isNativeFocused(this.nativeFocusableElement);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRatingComponent.prototype, "isFocusable", {
        get: function () {
            return !(this.readOnly || this.disabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRatingComponent.prototype, "percent", {
        get: function () {
            return (100 * this.value) / this.max;
        },
        enumerable: true,
        configurable: true
    });
    TuiRatingComponent.prototype.onFocused = function (focused) {
        this.updateFocused(focused);
    };
    TuiRatingComponent.prototype.setRateByReverseIndex = function (index) {
        this.updateValue(this.max - index);
    };
    TuiRatingComponent.prototype.setRate = function (value) {
        this.updateValue(value);
    };
    TuiRatingComponent.prototype.getFallbackValue = function () {
        return 0;
    };
    var TuiRatingComponent_1;
    TuiRatingComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_RATING_OPTIONS,] }] }
    ]; };
    __decorate([
        ViewChild('focusableElement')
    ], TuiRatingComponent.prototype, "focusableElement", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRatingComponent.prototype, "min", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRatingComponent.prototype, "max", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRatingComponent.prototype, "iconNormal", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRatingComponent.prototype, "iconFilled", void 0);
    __decorate([
        HostListener('focusin', ['true']),
        HostListener('focusout', ['false'])
    ], TuiRatingComponent.prototype, "onFocused", null);
    TuiRatingComponent = TuiRatingComponent_1 = __decorate([
        Component({
            selector: 'tui-rating',
            template: "<label>\n    <ng-content></ng-content>\n\n    <div class=\"t-host\">\n        <div class=\"t-interactive\">\n            <tui-svg\n                *tuiRepeatTimes=\"let index of max\"\n                class=\"t-icon-filled\"\n                [src]=\"iconFilled\"\n                (click)=\"setRateByReverseIndex(index)\"\n            ></tui-svg>\n        </div>\n\n        <div\n            class=\"t-wrapper\"\n            [style.width.%]=\"percent\"\n        >\n            <tui-svg\n                *tuiRepeatTimes=\"let index of max\"\n                class=\"t-icon-filled\"\n                [src]=\"iconFilled\"\n            ></tui-svg>\n        </div>\n\n        <tui-svg\n            *tuiRepeatTimes=\"let item of max\"\n            [src]=\"iconNormal\"\n        ></tui-svg>\n\n        <input\n            #focusableElement\n            type=\"range\"\n            step=\"1\"\n            class=\"t-input\"\n            [min]=\"min\"\n            [max]=\"max\"\n            [ngModel]=\"value\"\n            [tuiFocusable]=\"isFocusable\"\n            (ngModelChange)=\"setRate($event)\"\n        />\n    </div>\n</label>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                {
                    provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
                    useExisting: forwardRef(function () { return TuiRatingComponent_1; }),
                },
            ],
            styles: [":host._disabled,:host._readonly{pointer-events:none}:host._disabled{opacity:var(--tui-disabled-opacity)}.t-host{position:relative;display:inline-flex}.t-input{pointer-events:none;-webkit-appearance:none;-moz-appearance:none;appearance:none;background:0 0}.t-input::-webkit-slider-thumb{display:none}.t-input::-moz-range-thumb{opacity:0}.t-input,.t-interactive,.t-wrapper{position:absolute;left:0;bottom:0;width:100%;height:100%}.t-interactive,.t-wrapper{display:inline-flex;overflow:hidden;flex-direction:row-reverse;justify-content:flex-end;white-space:nowrap}.t-wrapper{pointer-events:none}.t-interactive{opacity:0}@media (hover:hover) and (pointer:fine){.t-interactive:hover{opacity:1}.t-interactive:hover~.t-wrapper{visibility:hidden}.t-interactive:hover .t-icon-filled{opacity:0}.t-interactive:hover .t-icon-filled:hover,.t-interactive:hover .t-icon-filled:hover~.t-icon-filled{opacity:1}}.t-icon-filled{position:relative;cursor:pointer}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(ChangeDetectorRef)),
        __param(2, Inject(TUI_RATING_OPTIONS))
    ], TuiRatingComponent);
    return TuiRatingComponent;
}(AbstractTuiControl));
export { TuiRatingComponent };
