// @taiga-ui/kit/esm5/components/multi-select/hide-selected.pipe.js
import { __decorate, __param } from "tslib";
import { Inject, Pipe } from '@angular/core';
import { tuiPure } from '@taiga-ui/cdk';
import { isFlat } from '@taiga-ui/kit/utils';
import { TuiMultiSelectComponent } from './multi-select.component';
var TuiHideSelectedPipe = /** @class */ (function () {
    function TuiHideSelectedPipe(component) {
        this.component = component;
    }
    TuiHideSelectedPipe.prototype.transform = function (items, component) {
        var _a = component !== null && component !== void 0 ? component : this.component, value = _a.value, identityMatcher = _a.identityMatcher;
        if (!items) {
            return null;
        }
        return isFlat(items)
            ? this.filter(items, value, identityMatcher)
            : this.filter2d(items, value, identityMatcher);
    };
    TuiHideSelectedPipe.prototype.filter2d = function (items, value, matcher) {
        var _this = this;
        return items.map(function (subItems) { return _this.filter(subItems, value, matcher); });
    };
    TuiHideSelectedPipe.prototype.filter = function (items, value, matcher) {
        return items.filter(function (item) { return value.every(function (selected) { return !matcher(selected, item); }); });
    };
    TuiHideSelectedPipe.ctorParameters = function () { return [
        { type: TuiMultiSelectComponent, decorators: [{ type: Inject, args: [TuiMultiSelectComponent,] }] }
    ]; };
    __decorate([
        tuiPure
    ], TuiHideSelectedPipe.prototype, "filter2d", null);
    __decorate([
        tuiPure
    ], TuiHideSelectedPipe.prototype, "filter", null);
    TuiHideSelectedPipe = __decorate([
        Pipe({
            name: "tuiHideSelected",
            pure: false,
        }),
        __param(0, Inject(TuiMultiSelectComponent))
    ], TuiHideSelectedPipe);
    return TuiHideSelectedPipe;
}());
export { TuiHideSelectedPipe };
