// @taiga-ui/kit/esm5/components/input-month-range/input-month-range.directive.js
import { __decorate, __extends, __param, __read, __spread } from "tslib";
import { Directive, forwardRef, Inject, Self } from '@angular/core';
import { AbstractTuiControl, TuiDestroyService, } from '@taiga-ui/cdk';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
import { TUI_MONTH_FORMATTER } from '@taiga-ui/kit/tokens';
import { combineLatest, Observable, Subject } from 'rxjs';
import { distinctUntilChanged, switchMap, takeUntil } from 'rxjs/operators';
import { TuiInputMonthRangeComponent } from './input-month-range.component';
var TuiInputMonthRangeDirective = /** @class */ (function (_super) {
    __extends(TuiInputMonthRangeDirective, _super);
    function TuiInputMonthRangeDirective(host, formatter, destroy$) {
        var _this = _super.call(this, host) || this;
        _this.value$ = new Subject();
        _this.localizedValue = ['', ''];
        _this.value$
            .pipe(distinctUntilChanged(), switchMap(function (value) {
            return combineLatest([
                formatter((value === null || value === void 0 ? void 0 : value.from) || null),
                formatter((value === null || value === void 0 ? void 0 : value.to) || null),
            ]);
        }), takeUntil(destroy$))
            .subscribe(function (localizedValue) {
            _this.localizedValue = localizedValue;
        });
        return _this;
    }
    TuiInputMonthRangeDirective_1 = TuiInputMonthRangeDirective;
    Object.defineProperty(TuiInputMonthRangeDirective.prototype, "readOnly", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiInputMonthRangeDirective.prototype, "value", {
        get: function () {
            var _a;
            return this.localizedValue[0]
                ? (_a = this.host).computeValue.apply(_a, __spread(this.localizedValue)) : '';
        },
        enumerable: true,
        configurable: true
    });
    TuiInputMonthRangeDirective.prototype.ngDoCheck = function () {
        this.value$.next(this.host.value);
    };
    TuiInputMonthRangeDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiInputMonthRangeDirective_1;
    TuiInputMonthRangeDirective.ctorParameters = function () { return [
        { type: TuiInputMonthRangeComponent, decorators: [{ type: Inject, args: [AbstractTuiControl,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_MONTH_FORMATTER,] }] },
        { type: Observable, decorators: [{ type: Self }, { type: Inject, args: [TuiDestroyService,] }] }
    ]; };
    TuiInputMonthRangeDirective = TuiInputMonthRangeDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-month-range',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputMonthRangeDirective_1; }),
                },
                TuiDestroyService,
            ],
        }),
        __param(0, Inject(AbstractTuiControl)),
        __param(1, Inject(TUI_MONTH_FORMATTER)),
        __param(2, Self()), __param(2, Inject(TuiDestroyService))
    ], TuiInputMonthRangeDirective);
    return TuiInputMonthRangeDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputMonthRangeDirective };
