// @taiga-ui/kit/esm5/components/arrow/arrow-options.js
import { __assign } from "tslib";
import { InjectionToken } from '@angular/core';
import { TUI_ARROW } from './arrow.component';
// TODO: 3.0 remove in ivy compilation
export var TUI_ARROW_ICON_SMALL = "tuiIconChevronDown";
export var TUI_ARROW_ICON_LARGE = "tuiIconChevronDownLarge";
/** Default values for arrow options */
export var TUI_ARROW_DEFAULT_OPTIONS = {
    iconSmall: TUI_ARROW_ICON_SMALL,
    iconLarge: TUI_ARROW_ICON_LARGE,
};
export var TUI_ARROW_OPTIONS = new InjectionToken("Default parameters for arrow component", {
    factory: function () { return TUI_ARROW_DEFAULT_OPTIONS; },
});
export var tuiArrowOptionsProvider = function (options) { return ({
    provide: TUI_ARROW_OPTIONS,
    useValue: __assign(__assign({}, TUI_ARROW_DEFAULT_OPTIONS), options),
}); };
export var TUI_ARROW_MODE = new InjectionToken("Type of icon in dropdowns for interactive or disable mode", {
    factory: function () { return ({
        interactive: TUI_ARROW,
        disabled: TUI_ARROW,
    }); },
});
export var tuiArrowModeProvider = function (options) { return ({
    provide: TUI_ARROW_MODE,
    useValue: __assign({ interactive: TUI_ARROW, disabled: TUI_ARROW }, options),
}); };
