// @taiga-ui/cdk/esm2015/abstract/dialog.directive.js
import { __decorate, __param } from "tslib";
import { ChangeDetectorRef, Directive, Inject, TemplateRef } from '@angular/core';
import { PolymorpheusTemplate } from '@tinkoff/ng-polymorpheus';
import { EMPTY, Subject } from 'rxjs';
import { endWith, ignoreElements, switchMap } from 'rxjs/operators';
import { AbstractTuiDialogService } from './dialog.service';
let AbstractTuiDialogDirective = class AbstractTuiDialogDirective extends PolymorpheusTemplate {
    constructor(templateRef, changeDetectorRef, service) {
        super(templateRef, changeDetectorRef);
        this.service = service;
        this.open$ = new Subject();
        this.options = {};
        this.openChange = this.open$.pipe(switchMap(open => open
            ? this.service
                .open(this, this.options)
                .pipe(ignoreElements(), endWith(false))
            : EMPTY));
    }
    set open(open) {
        this.open$.next(open);
    }
};
AbstractTuiDialogDirective.ctorParameters = () => [
    { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }] },
    { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
    { type: AbstractTuiDialogService, decorators: [{ type: Inject, args: [AbstractTuiDialogService,] }] }
];
AbstractTuiDialogDirective = __decorate([
    Directive(),
    __param(0, Inject(TemplateRef)),
    __param(1, Inject(ChangeDetectorRef)),
    __param(2, Inject(AbstractTuiDialogService))
], AbstractTuiDialogDirective);
export { AbstractTuiDialogDirective };
