// @taiga-ui/cdk/esm2015/abstract/portal-host.js
var AbstractTuiPortalHostComponent_1;
import { __decorate, __param } from "tslib";
import { ComponentFactory, ComponentRef, Directive, ElementRef, EmbeddedViewRef, Inject, Injector, TemplateRef, ViewChild, ViewContainerRef, } from '@angular/core';
import { AbstractTuiPortalService } from './portal-service';
/**
 * Abstract class for host element for dynamically created portals.
 */
let AbstractTuiPortalHostComponent = AbstractTuiPortalHostComponent_1 = class AbstractTuiPortalHostComponent {
    constructor(elementRef, portalService) {
        this.elementRef = elementRef;
        portalService.attach(this);
    }
    get clientRect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    addComponentChild(componentFactory, injector) {
        return this.viewContainerRef.createComponent(componentFactory, undefined, Injector.create({
            parent: injector,
            providers: [
                {
                    provide: AbstractTuiPortalHostComponent_1,
                    useValue: this,
                },
            ],
        }));
    }
    addTemplateChild(templateRef, context) {
        return this.viewContainerRef.createEmbeddedView(templateRef, context);
    }
};
AbstractTuiPortalHostComponent.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: AbstractTuiPortalService, decorators: [{ type: Inject, args: [AbstractTuiPortalService,] }] }
];
__decorate([
    ViewChild(`viewContainer`, { read: ViewContainerRef })
], AbstractTuiPortalHostComponent.prototype, "viewContainerRef", void 0);
AbstractTuiPortalHostComponent = AbstractTuiPortalHostComponent_1 = __decorate([
    Directive(),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(AbstractTuiPortalService))
], AbstractTuiPortalHostComponent);
export { AbstractTuiPortalHostComponent };
