// @taiga-ui/cdk/esm2015/abstract/theme-switcher.js
var AbstractTuiThemeSwitcher_1;
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, Inject } from '@angular/core';
/**
 * Use this abstract class to create your own toggleable themes.
 * A component extending this class must have CSS variables definitions
 * and have ViewEncapsulation set to NONE. A boolean input allows to
 * switch theme on or off.
 * @dynamic
 */
let AbstractTuiThemeSwitcher = AbstractTuiThemeSwitcher_1 = class AbstractTuiThemeSwitcher {
    constructor(documentRef) {
        this.documentRef = documentRef;
        if (this.style !== null) {
            this.addTheme();
            return;
        }
        const styles = this.documentRef.head.querySelectorAll(`style`);
        this.constructor.style =
            styles[styles.length - 1];
    }
    get style() {
        return this.constructor.style;
    }
    ngOnDestroy() {
        this.removeTheme();
    }
    addTheme() {
        if (this.style) {
            this.documentRef.head.appendChild(this.style);
        }
    }
    removeTheme() {
        if (this.style && this.documentRef.head.contains(this.style)) {
            this.documentRef.head.removeChild(this.style);
        }
    }
};
AbstractTuiThemeSwitcher.style = null;
AbstractTuiThemeSwitcher.ctorParameters = () => [
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
AbstractTuiThemeSwitcher = AbstractTuiThemeSwitcher_1 = __decorate([
    Directive(),
    __param(0, Inject(DOCUMENT))
], AbstractTuiThemeSwitcher);
export { AbstractTuiThemeSwitcher };
