// @taiga-ui/addon-doc/esm5/languages/russian.js
import { TUI_DOC_DEMO_TEXTS, TUI_DOC_DOCUMENTATION_TEXTS, TUI_DOC_EXAMPLE_TEXTS, TUI_DOC_MENU_TEXT, TUI_DOC_SEARCH_TEXT, TUI_DOC_SEE_ALSO_TEXT, TUI_DOC_SOURCE_CODE_TEXT, } from '../tokens/i18n';
// TODO: 3.0 remove in ivy compilation
export var DEMO_TEXTS = ["\u0421\u0434\u0435\u043B\u0430\u043D\u043E \u0441 \u043F\u043E\u043C\u043E\u0449\u044C\u044E \u0434\u0438\u0440\u0435\u043A\u0442\u0438\u0432\u044B: ", "\u0424\u043E\u043D", "\u0414\u0435\u0442\u0430\u043B\u0438 \u0444\u043E\u0440\u043C\u044B"];
export var DOC_TEXTS = [
    "\u0410\u0440\u0433\u0443\u043C\u0435\u043D\u0442",
    "\u0422\u0438\u043F",
    "\u0418\u043C\u044F \u0438 \u043E\u043F\u0438\u0441\u0430\u043D\u0438\u0435",
    "\u0417\u043D\u0430\u0447\u0435\u043D\u0438\u0435",
    "\u0414\u043B\u044F \u0440\u0430\u0431\u043E\u0442\u044B \u0441 \u0434\u0438\u043D\u0430\u043C\u0438\u0447\u0435\u0441\u043A\u0438\u043C\u0438 \u0448\u0430\u0431\u043B\u043E\u043D\u0430\u043C\u0438 \u0438\u0441\u043F\u043E\u043B\u044C\u0437\u0443\u0435\u0442\u0441\u044F",
];
export var EXAMPLE_TEXTS = ["\u041F\u0440\u0435\u0432\u044C\u044E", "\u0421\u0441\u044B\u043B\u043A\u0430 \u043D\u0430 \u043F\u0440\u0438\u043C\u0435\u0440 \u0441\u043A\u043E\u043F\u0438\u0440\u043E\u0432\u0430\u043D\u0430", "\u0413\u043E\u0442\u043E\u0432\u043E"];
export var TUI_DOC_RUSSIAN = [
    {
        provide: TUI_DOC_DEMO_TEXTS,
        useValue: DEMO_TEXTS,
    },
    {
        provide: TUI_DOC_DOCUMENTATION_TEXTS,
        useValue: DOC_TEXTS,
    },
    {
        provide: TUI_DOC_EXAMPLE_TEXTS,
        useValue: EXAMPLE_TEXTS,
    },
    {
        provide: TUI_DOC_MENU_TEXT,
        useValue: "\u041C\u0435\u043D\u044E",
    },
    {
        provide: TUI_DOC_SEARCH_TEXT,
        useValue: "\u041F\u043E\u0438\u0441\u043A",
    },
    {
        provide: TUI_DOC_SEE_ALSO_TEXT,
        useValue: "\u0421\u043C\u043E\u0442\u0440\u0438\u0442\u0435 \u0442\u0430\u043A\u0436\u0435",
    },
    {
        provide: TUI_DOC_SOURCE_CODE_TEXT,
        useValue: "\u0418\u0441\u0445\u043E\u0434\u043D\u044B\u0439 \u043A\u043E\u0434",
    },
];
