// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-files.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/files'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/kit/components/files'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/kit/utils')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-files', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/files', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/kit/components/files', '@tinkoff/ng-polymorpheus', '@taiga-ui/kit/utils'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-files'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.utils.files, global.rxjs, global.rxjs.operators, global.ng.common, global['taiga-ui'].kit.components.files, global.ngPolymorpheus, global['taiga-ui'].kit.utils));
}(this, (function (exports, core, forms, cdk, core$1, tokens, files, rxjs, operators, common, files$1, ngPolymorpheus, utils) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var DEFAULT_MAX_SIZE = 30 * 1024 * 1024; // 30 MiB
    // @dynamic
    var TuiInputFilesComponent = /** @class */ (function (_super) {
        __extends(TuiInputFilesComponent, _super);
        function TuiInputFilesComponent(control, changeDetectorRef, isMobile, inputFileTexts$) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.isMobile = isMobile;
            _this.inputFileTexts$ = inputFileTexts$;
            _this.dataTransfer = null;
            _this.link = '';
            _this.label = '';
            _this.accept = '';
            _this.multiple = false;
            _this.size = 'm';
            _this.maxFileSize = DEFAULT_MAX_SIZE;
            _this.reject = new core.EventEmitter();
            return _this;
        }
        TuiInputFilesComponent_1 = TuiInputFilesComponent;
        Object.defineProperty(TuiInputFilesComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.input) === null || _a === void 0 ? void 0 : _a.nativeElement) || null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "computedPseudoHovered", {
            get: function () {
                var _a;
                return (_a = this.pseudoHovered) !== null && _a !== void 0 ? _a : (this.fileDragged || null);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "computedLink$", {
            get: function () {
                return this.computeLink$(this.fileDragged, this.multiple, this.link);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "computedLabel$", {
            get: function () {
                return this.computeLabel$(this.isMobile, this.fileDragged, this.multiple, this.label);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "fileDragged", {
            get: function () {
                var _a;
                return !!((_a = this.dataTransfer) === null || _a === void 0 ? void 0 : _a.types.includes('Files'));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputFilesComponent.prototype, "arrayValue", {
            get: function () {
                return this.getValueArray(this.value);
            },
            enumerable: true,
            configurable: true
        });
        TuiInputFilesComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputFilesComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiInputFilesComponent.prototype.onPressed = function (pressed) {
            this.updatePressed(pressed);
        };
        TuiInputFilesComponent.prototype.onFilesSelected = function (input, formatRejection, maxSizeRejection) {
            this.processSelectedFiles(input.files, { formatRejection: formatRejection, maxSizeRejection: maxSizeRejection });
            input.value = '';
        };
        TuiInputFilesComponent.prototype.onDropped = function (event, formatRejection, maxSizeRejection) {
            this.processSelectedFiles(event.files, { formatRejection: formatRejection, maxSizeRejection: maxSizeRejection });
        };
        TuiInputFilesComponent.prototype.onDragOver = function (dataTransfer) {
            this.dataTransfer = dataTransfer;
        };
        TuiInputFilesComponent.prototype.removeFile = function (removedFile) {
            this.updateValue(this.multiple ? this.arrayValue.filter(function (file) { return file !== removedFile; }) : null);
        };
        TuiInputFilesComponent.prototype.computeLink$ = function (fileDragged, multiple, link) {
            return fileDragged
                ? rxjs.of('')
                : this.inputFileTexts$.pipe(operators.map(function (texts) {
                    return multiple && link === ''
                        ? texts.defaultLinkMultiple
                        : link || texts.defaultLinkSingle;
                }));
        };
        TuiInputFilesComponent.prototype.computeLabel$ = function (isMobile, fileDragged, multiple, label) {
            if (isMobile) {
                return rxjs.of('');
            }
            if (fileDragged) {
                return this.inputFileTexts$.pipe(operators.map(function (texts) { return (multiple ? texts.dropMultiple : texts.drop); }));
            }
            return this.inputFileTexts$.pipe(operators.map(function (texts) {
                return multiple && label === ''
                    ? texts.defaultLabelMultiple
                    : label || texts.defaultLabelSingle;
            }));
        };
        TuiInputFilesComponent.prototype.getValueArray = function (value) {
            if (!value) {
                return cdk.EMPTY_ARRAY;
            }
            return Array.isArray(value) ? value : [value];
        };
        TuiInputFilesComponent.prototype.processSelectedFiles = function (files, errors) {
            var _this = this;
            // IE11 after selecting a file through the open dialog generates a second event passing an empty FileList.
            if (!(files === null || files === void 0 ? void 0 : files.length)) {
                return;
            }
            var newFiles = this.multiple ? Array.from(files) : [files[0]];
            var tooBigFiles = newFiles.filter(function (file) { return file.size > _this.maxFileSize; });
            var wrongFormatFiles = newFiles.filter(function (file) { return !_this.isFormatAcceptable(file) && !tooBigFiles.includes(file); });
            var acceptedFiles = newFiles.filter(function (file) { return !tooBigFiles.includes(file) && !wrongFormatFiles.includes(file); });
            if (tooBigFiles.length || wrongFormatFiles.length) {
                this.rejectFiles(__spread(tooBigFiles.map(function (file) { return ({
                    name: file.name,
                    type: file.type,
                    size: file.size,
                    content: errors.maxSizeRejection,
                }); }), wrongFormatFiles.map(function (file) { return ({
                    name: file.name,
                    type: file.type,
                    size: file.size,
                    content: errors.formatRejection,
                }); })));
            }
            this.updateValue(this.multiple
                ? __spread(this.arrayValue, acceptedFiles) : acceptedFiles[0] || null);
        };
        TuiInputFilesComponent.prototype.isFormatAcceptable = function (file) {
            if (!this.accept) {
                return true;
            }
            var extension = "." + (file.name.split('.').pop() || '').toLowerCase();
            return files.getAcceptArray(this.accept).some(function (format) {
                return format === extension ||
                    format === file.type ||
                    (format.split('/')[1] === '*' &&
                        file.type.split('/')[0] === format.split('/')[0]);
            });
        };
        TuiInputFilesComponent.prototype.rejectFiles = function (rejectedFiles) {
            this.reject.emit(this.multiple ? rejectedFiles : rejectedFiles[0]);
        };
        var TuiInputFilesComponent_1;
        TuiInputFilesComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_INPUT_FILE_TEXTS,] }] }
        ]; };
        __decorate([
            core.ViewChild('input')
        ], TuiInputFilesComponent.prototype, "input", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "link", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "label", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "accept", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "multiple", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputFilesComponent.prototype, "maxFileSize", void 0);
        __decorate([
            core.Output()
        ], TuiInputFilesComponent.prototype, "reject", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiInputFilesComponent.prototype, "computeLink$", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputFilesComponent.prototype, "computeLabel$", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputFilesComponent.prototype, "getValueArray", null);
        TuiInputFilesComponent = TuiInputFilesComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-files',
                template: "<tui-wrapper\n    appearance=\"input-file\"\n    class=\"t-wrapper\"\n    [class._mobile]=\"isMobile\"\n    [focused]=\"computedFocused\"\n    [hovered]=\"computedHovered || fileDragged\"\n    [pressed]=\"computedPressed\"\n    [disabled]=\"computedDisabled\"\n>\n    <label\n        automation-id=\"tui-input-file__label\"\n        class=\"t-label\"\n    >\n        <a tuiLink>\n            <span\n                polymorpheus-outlet\n                class=\"t-inline\"\n                [content]=\"computedLink$ | async\"\n            ></span>\n        </a>\n        <ng-container *ngIf=\"computedLabel$ | async as computedLabel\">\n            <span>&nbsp;</span>\n            <span\n                polymorpheus-outlet\n                class=\"t-inline\"\n                [content]=\"computedLabel\"\n            ></span>\n        </ng-container>\n        <ng-container *ngIf=\"!readOnly && !computedDisabled\">\n            <input\n                #input\n                type=\"file\"\n                class=\"t-native\"\n                [id]=\"id\"\n                [accept]=\"accept\"\n                [multiple]=\"multiple\"\n                [tuiFocusable]=\"focusable\"\n                (change)=\"onFilesSelected(input, formatRejection, maxSizeRejection)\"\n                (tuiHoveredChange)=\"onHovered($event)\"\n                (tuiFocusedChange)=\"onFocused($event)\"\n                (tuiPressedChange)=\"onPressed($event)\"\n                (tuiDroppableDropped)=\"onDropped($event, formatRejection, maxSizeRejection)\"\n                (tuiDroppableDragOverChange)=\"onDragOver($event)\"\n                (mousedown.prevent.silent)=\"(0)\"\n            />\n        </ng-container>\n    </label>\n</tui-wrapper>\n\n<ng-template #formatRejection>\n    {{ (inputFileTexts$ | async)?.formatRejectionReason || '' }}\n</ng-template>\n\n<ng-template #maxSizeRejection>\n    {{ maxFileSize | tuiMaxSizeRejectionError | async }}\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                encapsulation: core.ViewEncapsulation.None,
                providers: [
                    core$1.MODE_PROVIDER,
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputFilesComponent_1; }),
                    },
                ],
                styles: ["tui-input-files{display:block}tui-input-files .t-native{position:absolute;top:0;left:0;width:100%;height:100%;opacity:0;cursor:pointer}tui-input-files .t-native::-webkit-file-upload-button{display:none}tui-input-files .t-inline{display:inline}tui-input-files .t-label{text-align:center}tui-wrapper[data-appearance=input-file]{display:flex;background:0 0;font:var(--tui-font-text-m);word-wrap:break-word;color:var(--tui-text-02);flex:1;justify-content:center;align-items:center;min-height:var(--tui-height-l);border-radius:var(--tui-radius-m);padding:1rem .5rem;box-sizing:border-box}tui-wrapper[data-appearance=input-file]:after{border:1px dashed;color:var(--tui-link)}tui-wrapper[data-appearance=input-file]._mobile:after{border:1px solid}tui-wrapper[data-appearance=input-file][data-state=hovered]{background:var(--tui-secondary)}tui-wrapper[data-appearance=input-file][data-state=hovered]:after{color:var(--tui-link-hover)}tui-wrapper[data-appearance=input-file][data-state=pressed]{background:var(--tui-secondary-hover)}tui-wrapper[data-appearance=input-file][data-state=disabled]{opacity:var(--tui-disabled-opacity);pointer-events:none}tui-wrapper[data-appearance=input-file][data-state=disabled]:after{color:var(--tui-text-03)}tui-wrapper[data-appearance=input-file]._focused:after{border-style:solid;border-width:2px;color:var(--tui-focus)}tui-wrapper[data-appearance=input-file][data-mode=onDark],tui-wrapper[data-appearance=input-file][data-mode=onDark]._focused:after,tui-wrapper[data-appearance=input-file][data-mode=onDark]:after{color:var(--tui-text-01-night)}tui-wrapper[data-appearance=input-file][data-mode=onDark][data-state=hovered]{background:var(--tui-clear-inverse-hover)}tui-wrapper[data-appearance=input-file][data-mode=onDark][data-state=hovered]:after,tui-wrapper[data-appearance=input-file][data-mode=onDark][data-state=pressed]:after{color:var(--tui-text-03-night)}tui-wrapper[data-appearance=input-file][data-mode=onDark][data-state=pressed]{background:var(--tui-clear-inverse-active)}tui-wrapper[data-appearance=input-file][data-mode=onLight],tui-wrapper[data-appearance=input-file][data-mode=onLight]:after{color:var(--tui-text-01)}tui-wrapper[data-appearance=input-file][data-mode=onLight][data-state=hovered]{background:var(--tui-clear-hover)}tui-wrapper[data-appearance=input-file][data-mode=onLight][data-state=hovered]:after,tui-wrapper[data-appearance=input-file][data-mode=onLight][data-state=pressed]:after{color:var(--tui-text-03)}tui-wrapper[data-appearance=input-file][data-mode=onLight][data-state=pressed]{background:var(--tui-clear-active)}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(cdk.TUI_IS_MOBILE)),
            __param(3, core.Inject(tokens.TUI_INPUT_FILE_TEXTS))
        ], TuiInputFilesComponent);
        return TuiInputFilesComponent;
    }(cdk.AbstractTuiNullableControl));

    var TuiMaxSizeRejectionErrorPipe = /** @class */ (function () {
        function TuiMaxSizeRejectionErrorPipe(inputFileTexts$, units$) {
            this.inputFileTexts$ = inputFileTexts$;
            this.units$ = units$;
        }
        TuiMaxSizeRejectionErrorPipe.prototype.transform = function (maxFileSize) {
            return rxjs.combineLatest([this.inputFileTexts$, this.units$]).pipe(operators.map(function (_a) {
                var _b = __read(_a, 2), maxSizeRejectionReason = _b[0].maxSizeRejectionReason, units = _b[1];
                return maxSizeRejectionReason + utils.tuiFormatSize(units, maxFileSize);
            }));
        };
        TuiMaxSizeRejectionErrorPipe.ctorParameters = function () { return [
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_INPUT_FILE_TEXTS,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_DIGITAL_INFORMATION_UNITS,] }] }
        ]; };
        TuiMaxSizeRejectionErrorPipe = __decorate([
            core.Pipe({ name: "tuiMaxSizeRejectionError" }),
            __param(0, core.Inject(tokens.TUI_INPUT_FILE_TEXTS)),
            __param(1, core.Inject(tokens.TUI_DIGITAL_INFORMATION_UNITS))
        ], TuiMaxSizeRejectionErrorPipe);
        return TuiMaxSizeRejectionErrorPipe;
    }());

    var TuiInputFilesModule = /** @class */ (function () {
        function TuiInputFilesModule() {
        }
        TuiInputFilesModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiLetModule,
                    cdk.TuiFocusedModule,
                    cdk.TuiFocusVisibleModule,
                    cdk.TuiPressedModule,
                    cdk.TuiHoveredModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiDroppableModule,
                    core$1.TuiWrapperModule,
                    core$1.TuiSvgModule,
                    core$1.TuiLinkModule,
                    core$1.TuiLoaderModule,
                    core$1.TuiButtonModule,
                    core$1.TuiGroupModule,
                    files$1.TuiFilesModule,
                ],
                declarations: [TuiInputFilesComponent, TuiMaxSizeRejectionErrorPipe],
                exports: [
                    TuiInputFilesComponent,
                    files$1.TuiFilesComponent,
                    files$1.TuiFileComponent,
                    cdk.TuiItemDirective,
                ],
            })
        ], TuiInputFilesModule);
        return TuiInputFilesModule;
    }());

    exports.TuiInputFilesComponent = TuiInputFilesComponent;
    exports.TuiInputFilesModule = TuiInputFilesModule;
    exports.ɵa = TuiMaxSizeRejectionErrorPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
