// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-slider.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/abstract'), require('@taiga-ui/kit/components/input-number'), require('@taiga-ui/kit/components/slider'), require('@taiga-ui/kit/constants'), require('@taiga-ui/kit/tokens'), require('rxjs'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-slider', ['exports', '@angular/common', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/abstract', '@taiga-ui/kit/components/input-number', '@taiga-ui/kit/components/slider', '@taiga-ui/kit/constants', '@taiga-ui/kit/tokens', 'rxjs', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-slider'] = {}), global.ng.common, global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.abstract, global['taiga-ui'].kit.components['input-number'], global['taiga-ui'].kit.components.slider, global['taiga-ui'].kit.constants, global['taiga-ui'].kit.tokens, global.rxjs, global.ngPolymorpheus));
}(this, (function (exports, common, core, forms, cdk, core$1, abstract, inputNumber, slider, constants, tokens, rxjs, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    /**
     * Turn on new `InputSlider`'s version.
     * The new version will behave almost the same as `InputSlider` from the next major release.
     * @deprecated TODO remove me in v3.0 and make `InputSlider` always "new".
     */
    var TuiNewInputSliderDirective = /** @class */ (function () {
        function TuiNewInputSliderDirective() {
        }
        TuiNewInputSliderDirective = __decorate([
            core.Directive({
                selector: 'tui-input-slider[new]',
            })
        ], TuiNewInputSliderDirective);
        return TuiNewInputSliderDirective;
    }());
    // @dynamic
    var TuiInputSliderComponent = /** @class */ (function (_super) {
        __extends(TuiInputSliderComponent, _super);
        function TuiInputSliderComponent(control, changeDetectorRef, controller, fromToTexts$, isNew, i18nPlural) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.controller = controller;
            _this.fromToTexts$ = fromToTexts$;
            _this.isNew = isNew;
            _this.i18nPlural = i18nPlural;
            _this.min = 0;
            _this.max = Infinity;
            _this.quantum = 1;
            _this.steps = 0;
            _this.segments = 0;
            _this.keySteps = null;
            _this.valueContent = '';
            _this.prefix = '';
            _this.postfix = '';
            /**
             * @deprecated use `tuiTextfieldCustomContent` instead
             * TODO delete in v3.0
             */
            _this.secondary = '';
            return _this;
        }
        TuiInputSliderComponent_1 = TuiInputSliderComponent;
        Object.defineProperty(TuiInputSliderComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a;
                return !((_a = this.inputNumberRef) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || this.computedDisabled
                    ? null
                    : this.inputNumberRef.nativeFocusableElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "focused", {
            get: function () {
                var _a;
                return (cdk.isNativeFocused(this.nativeFocusableElement) ||
                    cdk.isNativeFocused(((_a = this.sliderRef) === null || _a === void 0 ? void 0 : _a.nativeElement) || null));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "computedSteps", {
            get: function () {
                return this.steps || (this.max - this.min) / this.quantum;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "precision", {
            get: function () {
                return core$1.getFractionPartPadded(this.quantum).length;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "decimal", {
            get: function () {
                return this.precision ? 'not-zero' : 'never';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "showValueContent", {
            get: function () {
                return Boolean(this.computedValueContent && !this.focused);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "computedSize", {
            /**
             * TODO remove old property `size` in v3.0
             */
            get: function () {
                if (this.isNew) {
                    cdk.tuiAssert.assert(this.controller.size !== 's', 
                    // eslint-disable-next-line @typescript-eslint/quotes
                    "Size 's' is not supported by this input.");
                    return this.controller.size === 'l' ? 'l' : 'm';
                }
                return this.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputSliderComponent.prototype, "computedValueContent", {
            /**
             * @deprecated for backward compatibility
             * TODO replace by just `this.valueContent` in v3.0
             */
            get: function () {
                return this.minLabel || this.maxLabel
                    ? legacyMinMaxLabel(this)
                    : this.valueContent;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputSliderComponent.prototype.focusTextInput = function () {
            var _a;
            var focusableElement = (_a = this.inputNumberRef) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement;
            if (focusableElement) {
                cdk.setNativeFocused(focusableElement);
            }
        };
        TuiInputSliderComponent.prototype.safelyUpdateValue = function (value) {
            this.updateValue(this.valueGuard(value !== null && value !== void 0 ? value : this.safeCurrentValue));
        };
        TuiInputSliderComponent.prototype.onVerticalArrowKeyDown = function (coefficient) {
            if (this.readOnly || !this.step) {
                return;
            }
            var value = this.value + coefficient * this.step;
            if (value !== this.value) {
                this.safelyUpdateValue(value);
            }
            this.updateTextInputValue(this.valueGuard(value));
        };
        TuiInputSliderComponent.prototype.onSliderChange = function (newValue) {
            this.safelyUpdateValue(newValue);
            this.updateTextInputValue(this.value);
        };
        TuiInputSliderComponent.prototype.onFocused = function (focused) {
            var _a = this, value = _a.value, textInputValue = _a.textInputValue, safeCurrentValue = _a.safeCurrentValue, inputNumberRef = _a.inputNumberRef;
            if (!focused && textInputValue !== (inputNumberRef === null || inputNumberRef === void 0 ? void 0 : inputNumberRef.getFormattedValue(value))) {
                this.updateTextInputValue(value !== null && value !== void 0 ? value : safeCurrentValue);
            }
            this.updateFocused(focused);
        };
        TuiInputSliderComponent.prototype.onPressed = function (pressed) {
            this.updatePressed(pressed);
        };
        TuiInputSliderComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        Object.defineProperty(TuiInputSliderComponent.prototype, "textInputValue", {
            get: function () {
                var _a;
                return ((_a = this.inputNumberRef) === null || _a === void 0 ? void 0 : _a.nativeValue) || '';
            },
            enumerable: true,
            configurable: true
        });
        TuiInputSliderComponent.prototype.getFallbackValue = function () {
            return 0;
        };
        TuiInputSliderComponent.prototype.valueGuard = function (value) {
            var roundedValue = cdk.round(Math.round(value / this.quantum) * this.quantum, constants.TUI_FLOATING_PRECISION);
            return cdk.clamp(roundedValue, this.min, this.max);
        };
        TuiInputSliderComponent.prototype.updateTextInputValue = function (value) {
            if (this.inputNumberRef) {
                this.inputNumberRef.nativeValue =
                    this.inputNumberRef.getFormattedValue(value);
            }
        };
        var TuiInputSliderComponent_1;
        TuiInputSliderComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core$1.TuiTextfieldController, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_FROM_TO_TEXTS,] }] },
            { type: TuiNewInputSliderDirective, decorators: [{ type: core.Optional }, { type: core.Inject, args: [TuiNewInputSliderDirective,] }] },
            { type: common.I18nPluralPipe, decorators: [{ type: core.Inject, args: [common.I18nPluralPipe,] }] }
        ]; };
        __decorate([
            core.ViewChild(inputNumber.TuiInputNumberComponent)
        ], TuiInputSliderComponent.prototype, "inputNumberRef", void 0);
        __decorate([
            core.ViewChild(slider.TuiSliderComponent, { read: core.ElementRef })
        ], TuiInputSliderComponent.prototype, "sliderRef", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp(quantumAssertion, 'Quantum must be positive')
        ], TuiInputSliderComponent.prototype, "quantum", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "steps", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "segments", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "keySteps", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "valueContent", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "prefix", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "postfix", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputSliderComponent.prototype, "secondary", void 0);
        TuiInputSliderComponent = TuiInputSliderComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-slider',
                template: "<tui-input-number\n    [min]=\"min\"\n    [max]=\"max\"\n    [precision]=\"precision\"\n    [decimal]=\"decimal\"\n    [prefix]=\"showValueContent ? '' : prefix\"\n    [postfix]=\"showValueContent ? '' : postfix || (value | i18nPlural: pluralizeMap || pluralizeMapFallback)\"\n    [tuiTextfieldCustomContent]=\"controller.customContent || deprecatedSecondary\"\n    [tuiTextfieldSize]=\"computedSize\"\n    [tuiTextfieldLabelOutside]=\"computedSize !== 'l'\"\n    [disabled]=\"computedDisabled\"\n    [readOnly]=\"readOnly\"\n    [pseudoFocused]=\"computedFocused\"\n    [pseudoHovered]=\"computedHovered\"\n    [pseudoPressed]=\"computedPressed\"\n    [pseudoInvalid]=\"pseudoInvalid\"\n    [ngModel]=\"value\"\n    (ngModelChange)=\"safelyUpdateValue($event)\"\n    (focusedChange)=\"onFocused($event)\"\n    (hoveredChange)=\"onHovered($event)\"\n    (pressedChange)=\"onPressed($event)\"\n    (keydown.arrowUp.prevent)=\"onVerticalArrowKeyDown(1)\"\n    (keydown.arrowDown.prevent)=\"onVerticalArrowKeyDown(-1)\"\n>\n    <ng-content></ng-content>\n    <div\n        *ngIf=\"showValueContent\"\n        polymorpheus-outlet\n        automation-id=\"tui-input-slider__value-content\"\n        [content]=\"computedValueContent\"\n        [context]=\"{$implicit: value}\"\n    ></div>\n</tui-input-number>\n\n<!--TODO: remove in v3.0-->\n<ng-template #deprecatedSecondary>{{ secondary }}</ng-template>\n\n<!--TODO: remove wrapper + ticks labels (leave only slider) in v3.0-->\n<div class=\"t-slider-wrapper\">\n    <input\n        tuiSlider\n        type=\"range\"\n        [tuiFocusable]=\"false\"\n        [max]=\"computedSteps\"\n        [segments]=\"segments\"\n        [keySteps]=\"computedKeySteps\"\n        [attr.disabled]=\"readOnly || computedDisabled || null\"\n        [ngModel]=\"value\"\n        (keyStepsInput)=\"onSliderChange($event)\"\n        (click)=\"focusTextInput()\"\n    />\n\n    <!--TODO delete ticks labels in v3.0-->\n    <div\n        *ngIf=\"segments && !isNew\"\n        class=\"t-ticks-labels\"\n    >\n        <span\n            *tuiRepeatTimes=\"let tickIndex of segments + 1\"\n            automation-id=\"tui-slider__segment\"\n        >\n            <span *ngIf=\"fromToTexts$ | async as fromToText\">\n                {{ getTickPrefix(tickIndex, segments, fromToText) }}\n                {{ tickIndex | tuiSliderTickLabel: segments:computedKeySteps | tuiFormatNumber }}\n                <span *ngIf=\"segmentsPluralizeMap\">\n                    {{ tickIndex | tuiSliderTickLabel: segments:computedKeySteps | i18nPlural: segmentsPluralizeMap }}\n                </span>\n            </span>\n        </span>\n    </div>\n</div>\n",
                host: {
                    /**
                     * TODO delete it in v3.0
                     * Dont forget to clear html-tags
                     */
                    '[class._show-ticks-labels]': '!isNew',
                },
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    common.I18nPluralPipe,
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputSliderComponent_1; }),
                    },
                    slider.tuiSliderOptionsProvider({ trackColor: 'transparent' }),
                    core$1.HINT_CONTROLLER_PROVIDER,
                    core$1.TEXTFIELD_CONTROLLER_PROVIDER,
                ],
                styles: [":host{position:relative;z-index:0;display:block;height:-webkit-max-content;height:-moz-max-content;height:max-content}:host._segmented._show-ticks-labels{border-bottom:26px solid transparent}.t-slider-wrapper{position:absolute;top:100%;left:calc(var(--tui-radius-m)/ 2);right:0;margin:-.5625rem 0 0}input[type=range]{border-top-left-radius:0;border-bottom-left-radius:calc(var(--tui-radius-m) * 5) calc(var(--tui-radius-m) * 1.5)}.t-ticks-labels{display:flex;margin:0 .5rem;font:var(--tui-font-text-s);color:var(--tui-text-02)}.t-ticks-labels>*{position:relative;flex:2;text-align:center}.t-ticks-labels>:first-child{left:-.5rem;flex:1;text-align:left}.t-ticks-labels>:last-child{right:-.5rem;flex:1;text-align:right}tui-input-slider+.t-ticks-labels{margin-left:calc(var(--tui-radius-m)/ 2 + .5rem)}tui-input-range+.t-ticks-labels,tui-range+.t-ticks-labels{margin-left:1rem;margin-right:1rem}tui-input-range+.t-ticks-labels>:first-child,tui-range+.t-ticks-labels>:first-child{left:-1rem}tui-input-range+.t-ticks-labels>:last-child,tui-range+.t-ticks-labels>:last-child{right:-1rem}"]
            })
            /**
             * TODO replace `extends AbstractTuiInputSlider<number>` by `extends AbstractTuiControl<number> implements TuiWithOptionalMinMax<number>`
             * in v3.0
             */
            ,
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER)),
            __param(3, core.Inject(tokens.TUI_FROM_TO_TEXTS)),
            __param(4, core.Optional()),
            __param(4, core.Inject(TuiNewInputSliderDirective)),
            __param(5, core.Inject(common.I18nPluralPipe))
        ], TuiInputSliderComponent);
        return TuiInputSliderComponent;
    }(abstract.AbstractTuiInputSlider));
    function quantumAssertion(quantum) {
        return quantum > 0;
    }
    /**
     * @deprecated helper for backward compatibility.
     * TODO remove in v3.0
     */
    function legacyMinMaxLabel(_a) {
        var min = _a.min, max = _a.max, minLabel = _a.minLabel, maxLabel = _a.maxLabel, i18nPlural = _a.i18nPlural, pluralizeMap = _a.pluralizeMap;
        return function (_a) {
            var value = _a.$implicit;
            var valueWithPlural = value + " " + (pluralizeMap ? i18nPlural.transform(value, pluralizeMap) : '');
            switch (value) {
                case min:
                    return minLabel || valueWithPlural;
                case max:
                    return maxLabel || valueWithPlural;
                default:
                    return valueWithPlural;
            }
        };
    }

    var TuiInputSliderModule = /** @class */ (function () {
        function TuiInputSliderModule() {
        }
        TuiInputSliderModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    forms.ReactiveFormsModule,
                    cdk.TuiFocusableModule,
                    slider.TuiSliderModule,
                    core$1.TuiFormatNumberPipeModule,
                    cdk.TuiRepeatTimesModule,
                    ngPolymorpheus.PolymorpheusModule,
                    inputNumber.TuiInputNumberModule,
                    core$1.TuiTextfieldControllerModule,
                ],
                declarations: [TuiInputSliderComponent, TuiNewInputSliderDirective],
                exports: [TuiInputSliderComponent, TuiNewInputSliderDirective],
            })
        ], TuiInputSliderModule);
        return TuiInputSliderModule;
    }());

    exports.TuiInputSliderComponent = TuiInputSliderComponent;
    exports.TuiInputSliderModule = TuiInputSliderModule;
    exports.TuiNewInputSliderDirective = TuiNewInputSliderDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
