// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-phone-international.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/i18n'), require('@taiga-ui/kit/components/arrow'), require('@taiga-ui/kit/components/input-phone'), require('@taiga-ui/kit/pipes'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils'), require('rxjs'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-phone-international', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/i18n', '@taiga-ui/kit/components/arrow', '@taiga-ui/kit/components/input-phone', '@taiga-ui/kit/pipes', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils', 'rxjs', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-phone-international'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global.i18n, global['taiga-ui'].kit.components.arrow, global['taiga-ui'].kit.components['input-phone'], global['taiga-ui'].kit.pipes, global['taiga-ui'].kit.providers, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.utils, global.rxjs, global.ng.common, global.ngPolymorpheus));
}(this, (function (exports, core, forms, cdk, core$1, i18n, arrow, inputPhone, pipes, providers, tokens, utils, rxjs, common, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS = {
        countries: [],
        countryIsoCode: i18n.TuiCountryIsoCode.RU,
    };
    var TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS = new core.InjectionToken("Default parameters for input phone international component", {
        factory: function () { return TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS; },
    });
    var tuiInputPhoneInternationalOptionsProvider = function (options) { return ({
        provide: TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS,
        useValue: __assign(__assign({}, TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS), options),
    }); };

    /**
     * @deprecated: use {@link tuiExtractValueFromEvent} instead
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function extractValueFromEvent(event) {
        var _a;
        // TODO: iframe warning
        return event instanceof DragEvent
            ? ((_a = event.dataTransfer) === null || _a === void 0 ? void 0 : _a.getData("text/plain")) || ""
            : cdk.getClipboardDataText(event);
    }
    var tuiExtractValueFromEvent = extractValueFromEvent;

    // @dynamic
    var TuiInputPhoneInternationalComponent = /** @class */ (function (_super) {
        __extends(TuiInputPhoneInternationalComponent, _super);
        function TuiInputPhoneInternationalComponent(control, changeDetectorRef, countriesNames$, countriesMasks, options, flagPipe, extractCountryCodePipe) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.countriesNames$ = countriesNames$;
            _this.countriesMasks = countriesMasks;
            _this.options = options;
            _this.flagPipe = flagPipe;
            _this.extractCountryCodePipe = extractCountryCodePipe;
            _this.countries = _this.options.countries;
            _this.countryIsoCodeChange = new core.EventEmitter();
            _this.countryIsoCode = _this.options.countryIsoCode;
            _this.open = false;
            _this.arrow = arrow.TUI_ARROW;
            _this.isoToCountryCodeMapper = function (item) {
                return utils.tuiIsoToCountryCode(_this.countriesMasks, item);
            };
            return _this;
        }
        TuiInputPhoneInternationalComponent_1 = TuiInputPhoneInternationalComponent;
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "isoCode", {
            set: function (code) {
                var _a;
                (_a = this.inputPhoneComponent) === null || _a === void 0 ? void 0 : _a.writeValue(this.value);
                this.countryIsoCode = code;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.inputPhoneComponent && !this.computedDisabled
                    ? this.inputPhoneComponent.nativeFocusableElement
                    : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "focused", {
            get: function () {
                return ((!!this.primitiveTextfield && this.primitiveTextfield.focused) ||
                    (!!this.inputPhoneComponent && this.inputPhoneComponent.focused));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "inputPhoneCountryCode", {
            get: function () {
                return utils.tuiIsoToCountryCode(this.countriesMasks, this.countryIsoCode);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "phoneMaskAfterCountryCode", {
            get: function () {
                var countryCode = this.inputPhoneCountryCode;
                return this.calculateMaskAfterCountryCode(this.countriesMasks[this.countryIsoCode], countryCode);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneInternationalComponent.prototype, "countryFlagPath", {
            /**
             * @deprecated use `<img [src]="countryIsoCode | tuiFlagPipe" />`
             * TODO drop in v4.0
             */
            get: function () {
                return this.getFlagPath(this.countryIsoCode);
            },
            enumerable: true,
            configurable: true
        });
        TuiInputPhoneInternationalComponent.prototype.onPaste = function (event) {
            var value = tuiExtractValueFromEvent(event).replace(core$1.TUI_NON_DIGITS_REGEXP, '');
            var countryIsoCode = this.extractCountryCodePipe.transform(value, this.countries);
            if (!countryIsoCode) {
                this.updateValue(("" + this.inputPhoneCountryCode + value)
                    .replace(core$1.TUI_MASK_SYMBOLS_REGEXP, '')
                    .slice(0, utils.tuiGetMaxAllowedPhoneLength(this.countriesMasks, this.countryIsoCode)));
                return;
            }
            if (countryIsoCode === i18n.TuiCountryIsoCode.RU) {
                value = value.replace(/^8/, '7');
            }
            this.updateCountryIsoCode(countryIsoCode);
            this.updateValue("" + cdk.CHAR_PLUS + value);
        };
        /**
         * @deprecated use `<img [src]="countryIsoCode | tuiFlagPipe" />`
         * TODO drop in v4.0
         */
        TuiInputPhoneInternationalComponent.prototype.getFlagPath = function (code) {
            return this.flagPipe.transform(code);
        };
        TuiInputPhoneInternationalComponent.prototype.onItemClick = function (isoCode) {
            this.open = false;
            this.updateCountryIsoCode(isoCode);
            // recalculates mask inside inputPhone to prevent isoCode conflict
            this.changeDetectorRef.detectChanges();
            var maxLength = utils.tuiGetMaxAllowedPhoneLength(this.countriesMasks, isoCode);
            if (this.value.length > maxLength) {
                this.updateValue(this.value.slice(0, maxLength));
            }
            if (this.nativeFocusableElement) {
                cdk.setNativeFocused(this.nativeFocusableElement);
            }
        };
        TuiInputPhoneInternationalComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.close();
        };
        /**
         * @deprecated use `{{ countryIsoCode | tuiIsoToCountryCode }}`
         * TODO drop in v4.0
         */
        TuiInputPhoneInternationalComponent.prototype.isoToCountryCode = function (isoCode) {
            return utils.tuiIsoToCountryCode(this.countriesMasks, isoCode);
        };
        TuiInputPhoneInternationalComponent.prototype.onModelChange = function (value) {
            this.updateValue(value);
        };
        TuiInputPhoneInternationalComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiInputPhoneInternationalComponent.prototype.getFallbackValue = function () {
            return '';
        };
        TuiInputPhoneInternationalComponent.prototype.calculateMaskAfterCountryCode = function (mask, countryCode) {
            return mask.replace(countryCode, '').trim();
        };
        TuiInputPhoneInternationalComponent.prototype.close = function () {
            this.open = false;
        };
        TuiInputPhoneInternationalComponent.prototype.updateCountryIsoCode = function (code) {
            this.countryIsoCode = code;
            this.countryIsoCodeChange.emit(code);
        };
        var TuiInputPhoneInternationalComponent_1;
        TuiInputPhoneInternationalComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_COUNTRIES,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_COUNTRIES_MASKS,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS,] }] },
            { type: core$1.TuiFlagPipe, decorators: [{ type: core.Inject, args: [core$1.TuiFlagPipe,] }] },
            { type: pipes.TuiToCountryCodePipe, decorators: [{ type: core.Inject, args: [pipes.TuiToCountryCodePipe,] }] }
        ]; };
        __decorate([
            core.ViewChild(inputPhone.TuiInputPhoneComponent)
        ], TuiInputPhoneInternationalComponent.prototype, "inputPhoneComponent", void 0);
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputPhoneInternationalComponent.prototype, "primitiveTextfield", void 0);
        __decorate([
            core.Input('countryIsoCode'),
            cdk.tuiDefaultProp()
        ], TuiInputPhoneInternationalComponent.prototype, "isoCode", null);
        __decorate([
            core.Input()
        ], TuiInputPhoneInternationalComponent.prototype, "countries", void 0);
        __decorate([
            core.Output()
        ], TuiInputPhoneInternationalComponent.prototype, "countryIsoCodeChange", void 0);
        __decorate([
            core.HostListener('paste.capture.prevent.stop', ['$event']),
            core.HostListener('drop.capture.prevent.stop', ['$event'])
        ], TuiInputPhoneInternationalComponent.prototype, "onPaste", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputPhoneInternationalComponent.prototype, "calculateMaskAfterCountryCode", null);
        TuiInputPhoneInternationalComponent = TuiInputPhoneInternationalComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-phone-international',
                template: "<tui-hosted-dropdown\n    *ngIf=\"countriesNames$ | async as countriesNames\"\n    class=\"t-hosted-dropdown\"\n    [content]=\"dropdown\"\n    [canOpen]=\"!readOnly\"\n    [(open)]=\"open\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <div tuiGroup>\n        <tui-primitive-textfield\n            tuiHostedDropdownHost\n            class=\"t-country-select tui-group__auto-width-item\"\n            [tuiTextfieldCustomContent]=\"countryValueContent\"\n            [tuiTextfieldLabelOutside]=\"true\"\n            [disabled]=\"disabled\"\n            [focusable]=\"focusable\"\n            [editable]=\"false\"\n            [iconContent]=\"arrow\"\n            [pseudoFocused]=\"open || null\"\n            [readOnly]=\"readOnly\"\n            [tuiHintContent]=\"null\"\n        ></tui-primitive-textfield>\n        <tui-input-phone\n            tuiTextfieldAutocomplete=\"off\"\n            class=\"t-input-phone tui-group__auto-width-item\"\n            [ngModel]=\"value\"\n            [disabled]=\"disabled\"\n            [focusable]=\"focusable\"\n            [countryCode]=\"inputPhoneCountryCode\"\n            [phoneMaskAfterCountryCode]=\"phoneMaskAfterCountryCode\"\n            [readOnly]=\"readOnly\"\n            [pseudoInvalid]=\"computedInvalid\"\n            [pseudoFocused]=\"pseudoFocused\"\n            [pseudoHovered]=\"pseudoHovered\"\n            (ngModelChange)=\"onModelChange($event)\"\n        >\n            <ng-content></ng-content>\n        </tui-input-phone>\n    </div>\n\n    <ng-template #dropdown>\n        <tui-data-list>\n            <button\n                *ngFor=\"let isoCode of countries\"\n                tuiOption\n                (click)=\"onItemClick(isoCode)\"\n            >\n                <img\n                    alt=\"\"\n                    class=\"t-country-item-flag\"\n                    [src]=\"getFlagPath(isoCode)\"\n                />\n                <span class=\"t-country-item-name\">\n                    {{ countriesNames[isoCode] }}\n                </span>\n                <span class=\"t-country-item-code\">\n                    {{ isoCode | tuiMapper: isoToCountryCodeMapper }}\n                </span>\n            </button>\n        </tui-data-list>\n    </ng-template>\n\n    <ng-template #countryValueContent>\n        <img\n            class=\"t-flag\"\n            [alt]=\"countriesNames[countryIsoCode]\"\n            [src]=\"countryFlagPath\"\n        />\n    </ng-template>\n</tui-hosted-dropdown>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputPhoneInternationalComponent_1; }),
                    },
                    providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER,
                    // TODO: for backward compatibility only. Drop in v4.0
                    core$1.TuiFlagPipe,
                    pipes.TuiToCountryCodePipe,
                ],
                styles: [":host{display:block}:host._disabled{pointer-events:none}.t-hosted-dropdown{display:block}.t-country-select{width:5.625rem}.t-country-select:not(._readonly) ::ng-deep input:not(:disabled){cursor:pointer}.t-country-select._readonly ::ng-deep input{cursor:default}.t-arrow-icon{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03);position:relative;box-sizing:border-box;cursor:pointer;transition-property:color,transform}.t-arrow-icon:hover{color:var(--tui-text-02)}:host._disabled .t-arrow-icon,:host._readonly .t-arrow-icon{pointer-events:none}:host[data-mode=onDark] .t-arrow-icon{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-arrow-icon:hover{color:var(--tui-text-01-night)}.t-arrow-icon_open{transform:rotate(180deg)}.t-input-phone{flex:1}.t-flag{width:1.75rem;height:1.25rem;margin-left:-.5rem}.t-country-item-flag{width:1.75rem;height:1.25rem}.t-country-item-name{margin-left:.75rem;margin-right:auto}.t-country-item-code{color:var(--tui-text-02);margin-right:.25rem}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(tokens.TUI_COUNTRIES)),
            __param(3, core.Inject(tokens.TUI_COUNTRIES_MASKS)),
            __param(4, core.Inject(TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS)),
            __param(5, core.Inject(core$1.TuiFlagPipe)),
            __param(6, core.Inject(pipes.TuiToCountryCodePipe))
        ], TuiInputPhoneInternationalComponent);
        return TuiInputPhoneInternationalComponent;
    }(cdk.AbstractTuiControl));

    var TuiInputPhoneInternationalModule = /** @class */ (function () {
        function TuiInputPhoneInternationalModule() {
        }
        TuiInputPhoneInternationalModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    ngPolymorpheus.PolymorpheusModule,
                    core$1.TuiSvgModule,
                    inputPhone.TuiInputPhoneModule,
                    core$1.TuiGroupModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    core$1.TuiTextfieldControllerModule,
                    core$1.TuiHintControllerModule,
                    core$1.TuiHostedDropdownModule,
                    core$1.TuiDataListModule,
                    arrow.TuiArrowModule,
                    core$1.TuiFlagPipeModule,
                    cdk.TuiMapperPipeModule,
                    cdk.TuiActiveZoneModule,
                ],
                declarations: [TuiInputPhoneInternationalComponent],
                exports: [TuiInputPhoneInternationalComponent],
            })
        ], TuiInputPhoneInternationalModule);
        return TuiInputPhoneInternationalModule;
    }());

    exports.TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS = TUI_INPUT_PHONE_INTERNATIONAL_DEFAULT_OPTIONS;
    exports.TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS = TUI_INPUT_PHONE_INTERNATIONAL_OPTIONS;
    exports.TuiInputPhoneInternationalComponent = TuiInputPhoneInternationalComponent;
    exports.TuiInputPhoneInternationalModule = TuiInputPhoneInternationalModule;
    exports.extractValueFromEvent = extractValueFromEvent;
    exports.tuiExtractValueFromEvent = tuiExtractValueFromEvent;
    exports.tuiInputPhoneInternationalOptionsProvider = tuiInputPhoneInternationalOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
