// @taiga-ui/kit/bundles/taiga-ui-kit-enums.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/enums', ['exports'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.enums = {})));
}(this, (function (exports) { 'use strict';

    /**
     * @deprecated use string values
     */

    (function (TuiBorders) {
        TuiBorders["All"] = "all";
        TuiBorders["TopBottom"] = "top-bottom";
    })(exports.TuiBorders || (exports.TuiBorders = {}));

    /**
     * @deprecated use {@link TuiDropdownPositionT}
     */

    (function (TuiDropdownPosition) {
        TuiDropdownPosition["Selection"] = "selection";
        TuiDropdownPosition["Word"] = "word";
        TuiDropdownPosition["Tag"] = "tag";
    })(exports.TuiDropdownPosition || (exports.TuiDropdownPosition = {}));

    /**
     * @internal
     */

    (function (TuiFileState) {
        TuiFileState["Error"] = "error";
        TuiFileState["Deleted"] = "deleted";
        TuiFileState["Loading"] = "loading";
        TuiFileState["Normal"] = "normal";
    })(exports.TuiFileState || (exports.TuiFileState = {}));

    /**
     * @deprecated use join type {@link TuiMarkerIconModeT}
     */

    (function (TuiMarkerIconMode) {
        TuiMarkerIconMode["Link"] = "link";
        TuiMarkerIconMode["Primary"] = "primary";
        TuiMarkerIconMode["Warning"] = "warning";
        TuiMarkerIconMode["White"] = "white";
        TuiMarkerIconMode["Secondary"] = "secondary";
        TuiMarkerIconMode["Success"] = "success";
        TuiMarkerIconMode["Error"] = "error";
        TuiMarkerIconMode["OnDark"] = "onDark";
    })(exports.TuiMarkerIconMode || (exports.TuiMarkerIconMode = {}));

    /**
     * @deprecated use join type {@link TuiStatusT}
     */

    (function (TuiStatus) {
        TuiStatus["Default"] = "default";
        TuiStatus["Primary"] = "primary";
        TuiStatus["Custom"] = "custom";
        TuiStatus["Success"] = "success";
        TuiStatus["Error"] = "error";
        TuiStatus["Warning"] = "warning";
    })(exports.TuiStatus || (exports.TuiStatus = {}));

    /**
     * @deprecated use {@link TuiStepStateT}
     */

    (function (TuiStepState) {
        TuiStepState["Normal"] = "normal";
        TuiStepState["Error"] = "error";
        TuiStepState["Pass"] = "pass";
    })(exports.TuiStepState || (exports.TuiStepState = {}));

    /**
     * @deprecated use string values
     */

    (function (TuiTextAlign) {
        TuiTextAlign["Left"] = "left";
        TuiTextAlign["Center"] = "center";
        TuiTextAlign["Right"] = "right";
    })(exports.TuiTextAlign || (exports.TuiTextAlign = {}));

    Object.defineProperty(exports, '__esModule', { value: true });

})));
