// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-month-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/providers'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/kit/components/calendar-month')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-month-range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/tokens', '@taiga-ui/kit/providers', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/kit/components/calendar-month'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-month-range'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.providers, global.rxjs, global.rxjs.operators, global.ng.common, global['taiga-ui'].kit.components['calendar-month']));
}(this, (function (exports, core, forms, cdk, core$1, tokens, providers, rxjs, operators, common, calendarMonth) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_INPUT_MONTH_RANGE_PROVIDERS = [
        {
            provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
            useExisting: core.forwardRef(function () { return TuiInputMonthRangeComponent; }),
        },
        {
            provide: cdk.AbstractTuiControl,
            useExisting: core.forwardRef(function () { return TuiInputMonthRangeComponent; }),
        },
        providers.LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER,
        providers.TUI_MONTH_FORMATTER_PROVIDER,
        core$1.TuiMonthPipe,
    ];

    // @dynamic
    var TuiInputMonthRangeComponent = /** @class */ (function (_super) {
        __extends(TuiInputMonthRangeComponent, _super);
        function TuiInputMonthRangeComponent(control, changeDetectorRef, formatter, textfieldSize) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.formatter = formatter;
            _this.textfieldSize = textfieldSize;
            _this.min = cdk.TUI_FIRST_DAY;
            _this.max = cdk.TUI_LAST_DAY;
            _this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            _this.open = false;
            return _this;
        }
        Object.defineProperty(TuiInputMonthRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputMonthRangeComponent.prototype, "focused", {
            get: function () {
                return !!this.textfield && this.textfield.focused;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputMonthRangeComponent.prototype, "calendarIcon", {
            get: function () {
                return core$1.sizeBigger(this.textfieldSize.size)
                    ? 'tuiIconCalendarLarge'
                    : 'tuiIconCalendar';
            },
            enumerable: true,
            configurable: true
        });
        TuiInputMonthRangeComponent.prototype.computeValue = function (from, to) {
            var formattedTo = from === to && this.focused && !this.readOnly ? '' : to;
            return from + " " + cdk.CHAR_EN_DASH + " " + formattedTo;
        };
        TuiInputMonthRangeComponent.prototype.onValueChange = function (value) {
            if (value) {
                return;
            }
            this.updateValue(null);
            this.onOpenChange(true);
        };
        TuiInputMonthRangeComponent.prototype.onMonthClick = function (month) {
            if (this.value === null || !this.value.isSingleMonth) {
                this.writeValue(new cdk.TuiMonthRange(month, month));
                return;
            }
            this.updateValue(cdk.TuiMonthRange.sort(this.value.from, month));
            this.close();
        };
        TuiInputMonthRangeComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputMonthRangeComponent.prototype.onOpenChange = function (open) {
            this.open = open;
        };
        TuiInputMonthRangeComponent.prototype.onActiveZone = function (focused) {
            var _a;
            this.updateFocused(focused);
            if (focused) {
                return;
            }
            if ((_a = this.value) === null || _a === void 0 ? void 0 : _a.isSingleMonth) {
                this.updateValue(new cdk.TuiMonthRange(this.value.from, this.value.from));
            }
        };
        TuiInputMonthRangeComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.close();
        };
        TuiInputMonthRangeComponent.prototype.close = function () {
            this.open = false;
        };
        TuiInputMonthRangeComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_MONTH_FORMATTER,] }] },
            { type: core$1.TuiTextfieldSizeDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_SIZE,] }] }
        ]; };
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputMonthRangeComponent.prototype, "textfield", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputMonthRangeComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputMonthRangeComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputMonthRangeComponent.prototype, "disabledItemHandler", void 0);
        TuiInputMonthRangeComponent = __decorate([
            core.Component({
                selector: 'tui-input-month-range',
                template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"interactive\"\n    [content]=\"dropdown\"\n    [open]=\"open && interactive\"\n    (openChange)=\"onOpenChange($event)\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-primitive-textfield\n        class=\"t-textfield\"\n        [editable]=\"false\"\n        [disabled]=\"computedDisabled\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [iconContent]=\"calendarIcon\"\n        [pseudoFocused]=\"pseudoFocused\"\n        [pseudoHovered]=\"pseudoHovered\"\n        [invalid]=\"computedInvalid\"\n        [value]=\"\n            value\n                ? computeValue(value.from | tuiMapper: formatter | async, value.to | tuiMapper: formatter | async)\n                : ''\n        \"\n        (valueChange)=\"onValueChange($event)\"\n        (hoveredChange)=\"onHovered($event)\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n    </tui-primitive-textfield>\n\n    <ng-template #dropdown>\n        <tui-calendar-month\n            tuiPreventDefault=\"mousedown\"\n            [disabledItemHandler]=\"disabledItemHandler\"\n            [min]=\"min\"\n            [max]=\"max\"\n            [value]=\"value\"\n            (monthClick)=\"onMonthClick($event)\"\n        ></tui-calendar-month>\n    </ng-template>\n</tui-hosted-dropdown>\n",
                providers: TUI_INPUT_MONTH_RANGE_PROVIDERS,
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(tokens.TUI_MONTH_FORMATTER)),
            __param(3, core.Inject(core$1.TUI_TEXTFIELD_SIZE))
        ], TuiInputMonthRangeComponent);
        return TuiInputMonthRangeComponent;
    }(cdk.AbstractTuiNullableControl));

    var TuiInputMonthRangeDirective = /** @class */ (function (_super) {
        __extends(TuiInputMonthRangeDirective, _super);
        function TuiInputMonthRangeDirective(host, formatter, destroy$) {
            var _this = _super.call(this, host) || this;
            _this.value$ = new rxjs.Subject();
            _this.localizedValue = ['', ''];
            _this.value$
                .pipe(operators.distinctUntilChanged(), operators.switchMap(function (value) {
                return rxjs.combineLatest([
                    formatter((value === null || value === void 0 ? void 0 : value.from) || null),
                    formatter((value === null || value === void 0 ? void 0 : value.to) || null),
                ]);
            }), operators.takeUntil(destroy$))
                .subscribe(function (localizedValue) {
                _this.localizedValue = localizedValue;
            });
            return _this;
        }
        TuiInputMonthRangeDirective_1 = TuiInputMonthRangeDirective;
        Object.defineProperty(TuiInputMonthRangeDirective.prototype, "readOnly", {
            get: function () {
                return true;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputMonthRangeDirective.prototype, "value", {
            get: function () {
                var _a;
                return this.localizedValue[0]
                    ? (_a = this.host).computeValue.apply(_a, __spread(this.localizedValue)) : '';
            },
            enumerable: true,
            configurable: true
        });
        TuiInputMonthRangeDirective.prototype.ngDoCheck = function () {
            this.value$.next(this.host.value);
        };
        TuiInputMonthRangeDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        var TuiInputMonthRangeDirective_1;
        TuiInputMonthRangeDirective.ctorParameters = function () { return [
            { type: TuiInputMonthRangeComponent, decorators: [{ type: core.Inject, args: [cdk.AbstractTuiControl,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_MONTH_FORMATTER,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Self }, { type: core.Inject, args: [cdk.TuiDestroyService,] }] }
        ]; };
        TuiInputMonthRangeDirective = TuiInputMonthRangeDirective_1 = __decorate([
            core.Directive({
                selector: 'tui-input-month-range',
                providers: [
                    {
                        provide: core$1.TUI_TEXTFIELD_HOST,
                        useExisting: core.forwardRef(function () { return TuiInputMonthRangeDirective_1; }),
                    },
                    cdk.TuiDestroyService,
                ],
            }),
            __param(0, core.Inject(cdk.AbstractTuiControl)),
            __param(1, core.Inject(tokens.TUI_MONTH_FORMATTER)),
            __param(2, core.Self()), __param(2, core.Inject(cdk.TuiDestroyService))
        ], TuiInputMonthRangeDirective);
        return TuiInputMonthRangeDirective;
    }(core$1.TuiAbstractTextfieldHost));

    var TuiInputMonthRangeModule = /** @class */ (function () {
        function TuiInputMonthRangeModule() {
        }
        TuiInputMonthRangeModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    calendarMonth.TuiCalendarMonthModule,
                    core$1.TuiHostedDropdownModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    core$1.TuiSvgModule,
                    cdk.TuiPreventDefaultModule,
                    cdk.TuiActiveZoneModule,
                    cdk.TuiMapperPipeModule,
                ],
                declarations: [TuiInputMonthRangeComponent, TuiInputMonthRangeDirective],
                exports: [
                    TuiInputMonthRangeComponent,
                    TuiInputMonthRangeDirective,
                    core$1.TuiTextfieldComponent,
                ],
            })
        ], TuiInputMonthRangeModule);
        return TuiInputMonthRangeModule;
    }());

    exports.TUI_INPUT_MONTH_RANGE_PROVIDERS = TUI_INPUT_MONTH_RANGE_PROVIDERS;
    exports.TuiInputMonthRangeComponent = TuiInputMonthRangeComponent;
    exports.TuiInputMonthRangeDirective = TuiInputMonthRangeDirective;
    exports.TuiInputMonthRangeModule = TuiInputMonthRangeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
