// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-date-time.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/constants'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/mask'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/utils/miscellaneous'), require('@angular/common'), require('@taiga-ui/kit/directives'), require('@tinkoff/ng-polymorpheus'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-date-time', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/constants', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/mask', 'rxjs', 'rxjs/operators', '@taiga-ui/kit/providers', '@taiga-ui/kit/utils/miscellaneous', '@angular/common', '@taiga-ui/kit/directives', '@tinkoff/ng-polymorpheus', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-date-time'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.constants, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.utils.mask, global.rxjs, global.rxjs.operators, global['taiga-ui'].kit.providers, global['taiga-ui'].kit.utils.miscellaneous, global.ng.common, global['taiga-ui'].kit.directives, global.ngPolymorpheus, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, constants, tokens, mask, rxjs, operators, providers, miscellaneous, common, directives, ngPolymorpheus, angular2TextMask) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    // TODO: 3.0 remove in ivy compilation
    var TIME_STREAM_FACTORY = function (control, valueTransformer) {
        return miscellaneous.TuiReplayControlValueChangesFactory(control, valueTransformer);
    };
    var TUI_INPUT_DATE_TIME_PROVIDERS = [
        {
            provide: cdk.AbstractTuiControl,
            useExisting: core.forwardRef(function () { return TuiInputDateTimeComponent; }),
        },
        {
            provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
            useExisting: core.forwardRef(function () { return TuiInputDateTimeComponent; }),
        },
        {
            provide: tokens.TUI_CALENDAR_DATA_STREAM,
            deps: [
                [new core.Optional(), new core.Self(), forms.NgControl],
                [new core.Optional(), core.forwardRef(function () { return tokens.TUI_DATE_TIME_VALUE_TRANSFORMER; })],
            ],
            useFactory: TIME_STREAM_FACTORY,
        },
        providers.LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER,
    ];

    // @dynamic
    var TuiInputDateTimeComponent = /** @class */ (function (_super) {
        __extends(TuiInputDateTimeComponent, _super);
        function TuiInputDateTimeComponent(control, changeDetectorRef, textfieldSize, dateFormat, dateSeparator, timeTexts$, dateTexts$, valueTransformer) {
            var _this = _super.call(this, control, changeDetectorRef, valueTransformer) || this;
            _this.textfieldSize = textfieldSize;
            _this.dateFormat = dateFormat;
            _this.dateSeparator = dateSeparator;
            _this.timeTexts$ = timeTexts$;
            _this.dateTexts$ = dateTexts$;
            _this.valueTransformer = valueTransformer;
            _this.month = null;
            _this.min = cdk.TUI_FIRST_DAY;
            _this.max = cdk.TUI_LAST_DAY;
            _this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            _this.defaultActiveYearMonth = cdk.TuiMonth.currentLocal();
            _this.timeMode = 'HH:MM';
            _this.open = false;
            _this.filler$ = rxjs.combineLatest([
                _this.dateTexts$.pipe(operators.map(function (dateTexts) {
                    return cdk.changeDateSeparator(dateTexts[_this.dateFormat], _this.dateSeparator);
                })),
                _this.timeTexts$.pipe(operators.pluck(_this.timeMode)),
            ]).pipe(operators.map(function (fillers) { return _this.getDateTimeString.apply(_this, __spread(fillers)); }));
            return _this;
        }
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "fillerLength", {
            get: function () {
                return cdk.DATE_FILLER_LENGTH + constants.DATE_TIME_SEPARATOR.length + this.timeMode.length;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "textMaskOptions", {
            get: function () {
                return this.calculateMask(this.value[0], this.calendarMinDay, this.calendarMaxDay, this.timeMode, this.dateFormat, this.dateSeparator);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "focused", {
            get: function () {
                return !!this.textfield && this.textfield.focused;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarIcon", {
            get: function () {
                return core$1.sizeBigger(this.textfieldSize.size)
                    ? 'tuiIconCalendarLarge'
                    : 'tuiIconCalendar';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "computedValue", {
            get: function () {
                var _a = this, value = _a.value, nativeValue = _a.nativeValue, timeMode = _a.timeMode;
                var _b = __read(value, 2), date = _b[0], time = _b[1];
                var hasTimeInputChars = nativeValue.length > cdk.DATE_FILLER_LENGTH;
                if (!date || (!time && hasTimeInputChars)) {
                    return nativeValue;
                }
                return this.getDateTimeString(date, time, timeMode);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarValue", {
            get: function () {
                return this.value[0];
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarMinDay", {
            get: function () {
                return Array.isArray(this.min) ? this.min[0] : this.min;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "calendarMaxDay", {
            get: function () {
                return Array.isArray(this.max) ? this.max[0] : this.max;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "computedActiveYearMonth", {
            get: function () {
                return this.month || this.value[0] || this.defaultActiveYearMonth;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputDateTimeComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputDateTimeComponent.prototype.onClick = function () {
            this.open = !this.open;
        };
        TuiInputDateTimeComponent.prototype.onValueChange = function (value) {
            if (!value) {
                this.onOpenChange(true);
            }
            if (value.length < cdk.DATE_FILLER_LENGTH) {
                this.updateValue([null, null]);
                return;
            }
            var _a = __read(value.split(constants.DATE_TIME_SEPARATOR), 2), date = _a[0], time = _a[1];
            var parsedDate = cdk.TuiDay.normalizeParse(date, this.dateFormat);
            var parsedTime = time && time.length === this.timeMode.length
                ? this.clampTime(cdk.TuiTime.fromString(time), parsedDate)
                : null;
            this.open = false;
            this.updateValue([parsedDate, parsedTime]);
        };
        TuiInputDateTimeComponent.prototype.onDayClick = function (day) {
            var modifiedTime = this.value[1] && this.clampTime(this.value[1], day);
            this.updateValue([day, modifiedTime]);
            this.updateNativeValue(day);
            this.open = false;
        };
        TuiInputDateTimeComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputDateTimeComponent.prototype.onMonthChange = function (month) {
            this.month = month;
        };
        TuiInputDateTimeComponent.prototype.onOpenChange = function (open) {
            this.open = open;
        };
        TuiInputDateTimeComponent.prototype.onFocused = function (focused) {
            var _this = this;
            this.updateFocused(focused);
            if (focused ||
                this.value[0] === null ||
                this.value[1] !== null ||
                this.nativeValue.length <= this.fillerLength + constants.DATE_TIME_SEPARATOR.length ||
                this.timeMode === 'HH:MM') {
                return;
            }
            var _a = __read(this.nativeValue.split(constants.DATE_TIME_SEPARATOR), 2), time = _a[1];
            if (!time) {
                return;
            }
            var parsedTime = cdk.TuiTime.fromString(time);
            this.updateValue([this.value[0], parsedTime]);
            setTimeout(function () {
                if (_this.nativeValue.endsWith('.') || _this.nativeValue.endsWith(':')) {
                    _this.nativeValue = _this.nativeValue.slice(0, -1);
                }
            });
        };
        TuiInputDateTimeComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        TuiInputDateTimeComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.nativeValue =
                this.value && (this.value[0] || this.value[1]) ? this.computedValue : '';
        };
        TuiInputDateTimeComponent.prototype.getFallbackValue = function () {
            return [null, null];
        };
        TuiInputDateTimeComponent.prototype.valueIdenticalComparator = function (oldValue, newValue) {
            return (cdk.nullableSame(oldValue[0], newValue[0], function (a, b) { return a.daySame(b); }) &&
                cdk.nullableSame(oldValue[1], newValue[1], function (a, b) { return String(a) === String(b); }));
        };
        TuiInputDateTimeComponent.prototype.calculateMask = function (day, min, max, timeMode, dateFormat, dateSeparator) {
            return {
                mask: __spread(mask.tuiCreateDateMask(dateFormat, dateSeparator), [
                    ',',
                    ' '
                ], mask.tuiCreateTimeMask(timeMode)),
                pipe: mask.tuiCreateAutoCorrectedDateTimePipe({
                    value: day,
                    min: min,
                    max: max,
                    dateFormat: dateFormat,
                    dateSeparator: dateSeparator,
                    timeMode: timeMode,
                }),
                guide: false,
            };
        };
        TuiInputDateTimeComponent.prototype.getDateTimeString = function (date, time, timeMode) {
            if (timeMode === void 0) { timeMode = 'HH:MM'; }
            var dateString = date instanceof cdk.TuiDay
                ? date.toString(this.dateFormat, this.dateSeparator)
                : date;
            var timeString = time instanceof cdk.TuiTime ? time.toString(timeMode) : time || '';
            return "" + dateString + constants.DATE_TIME_SEPARATOR + timeString;
        };
        TuiInputDateTimeComponent.prototype.updateNativeValue = function (day) {
            var time = this.nativeValue.split(constants.DATE_TIME_SEPARATOR)[1] || '';
            this.nativeValue = this.getDateTimeString(day, time);
        };
        TuiInputDateTimeComponent.prototype.clampTime = function (time, day) {
            var ms = time.toAbsoluteMilliseconds();
            var min = Array.isArray(this.min) && day.daySame(this.calendarMinDay)
                ? this.min[1].toAbsoluteMilliseconds()
                : -Infinity;
            var max = Array.isArray(this.max) && day.daySame(this.calendarMaxDay)
                ? this.max[1].toAbsoluteMilliseconds()
                : Infinity;
            return cdk.TuiTime.fromAbsoluteMilliseconds(cdk.clamp(ms, min, max));
        };
        TuiInputDateTimeComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core$1.TuiTextfieldSizeDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_SIZE,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [cdk.TUI_DATE_FORMAT,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [cdk.TUI_DATE_SEPARATOR,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_TIME_TEXTS,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_DATE_TEXTS,] }] },
            { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_DATE_TIME_VALUE_TRANSFORMER,] }] }
        ]; };
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputDateTimeComponent.prototype, "textfield", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputDateTimeComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputDateTimeComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputDateTimeComponent.prototype, "disabledItemHandler", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputDateTimeComponent.prototype, "defaultActiveYearMonth", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputDateTimeComponent.prototype, "timeMode", void 0);
        __decorate([
            core.HostListener('click')
        ], TuiInputDateTimeComponent.prototype, "onClick", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputDateTimeComponent.prototype, "calculateMask", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputDateTimeComponent.prototype, "getDateTimeString", null);
        TuiInputDateTimeComponent = __decorate([
            core.Component({
                selector: 'tui-input-date-time',
                template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"interactive\"\n    [content]=\"dropdown\"\n    [open]=\"open && interactive\"\n    (openChange)=\"onOpenChange($event)\"\n>\n    <tui-primitive-textfield\n        automation-id=\"tui-input-date-time__textfield\"\n        tuiValueAccessor\n        tuiTextfieldInputMode=\"numeric\"\n        class=\"t-textfield\"\n        [pseudoFocused]=\"pseudoFocused\"\n        [pseudoHovered]=\"pseudoHovered\"\n        [invalid]=\"computedInvalid\"\n        [filler]=\"(filler$ | async) || ''\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [iconContent]=\"iconContent\"\n        [disabled]=\"computedDisabled\"\n        [textMask]=\"textMaskOptions\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($event)\"\n        (hoveredChange)=\"onHovered($event)\"\n        (focusedChange)=\"onFocused($event)\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n    </tui-primitive-textfield>\n\n    <ng-template #iconContent>\n        <tui-svg\n            [class.t-icon]=\"!computedDisabled\"\n            [src]=\"calendarIcon\"\n        ></tui-svg>\n    </ng-template>\n\n    <ng-template\n        #dropdown=\"polymorpheus\"\n        [polymorpheus]=\"type\"\n    >\n        <tui-calendar\n            tuiPreventDefault=\"mousedown\"\n            automation-id=\"tui-input-date-time__calendar\"\n            [min]=\"calendarMinDay\"\n            [max]=\"calendarMaxDay\"\n            [disabledItemHandler]=\"disabledItemHandler\"\n            [month]=\"computedActiveYearMonth\"\n            [value]=\"calendarValue\"\n            (dayClick)=\"onDayClick($event)\"\n            (monthChange)=\"onMonthChange($event)\"\n        ></tui-calendar>\n    </ng-template>\n</tui-hosted-dropdown>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: TUI_INPUT_DATE_TIME_PROVIDERS,
                styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-icon{pointer-events:auto}.t-button{display:flex;height:2.75rem;justify-content:center;box-shadow:inset 0 1px var(--tui-base-03)}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core$1.TUI_TEXTFIELD_SIZE)),
            __param(3, core.Inject(cdk.TUI_DATE_FORMAT)),
            __param(4, core.Inject(cdk.TUI_DATE_SEPARATOR)),
            __param(5, core.Inject(tokens.TUI_TIME_TEXTS)),
            __param(6, core.Inject(tokens.TUI_DATE_TEXTS)),
            __param(7, core.Optional()),
            __param(7, core.Inject(tokens.TUI_DATE_TIME_VALUE_TRANSFORMER))
        ], TuiInputDateTimeComponent);
        return TuiInputDateTimeComponent;
    }(cdk.AbstractTuiControl));

    var TuiInputDateTimeDirective = /** @class */ (function (_super) {
        __extends(TuiInputDateTimeDirective, _super);
        function TuiInputDateTimeDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TuiInputDateTimeDirective_1 = TuiInputDateTimeDirective;
        Object.defineProperty(TuiInputDateTimeDirective.prototype, "value", {
            get: function () {
                return this.host.computedValue;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputDateTimeDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        TuiInputDateTimeDirective.prototype.process = function (input) {
            input.inputMode = 'numeric';
        };
        var TuiInputDateTimeDirective_1;
        TuiInputDateTimeDirective = TuiInputDateTimeDirective_1 = __decorate([
            core.Directive({
                selector: 'tui-input-date-time',
                providers: [
                    {
                        provide: core$1.TUI_TEXTFIELD_HOST,
                        useExisting: core.forwardRef(function () { return TuiInputDateTimeDirective_1; }),
                    },
                ],
            })
        ], TuiInputDateTimeDirective);
        return TuiInputDateTimeDirective;
    }(core$1.TuiAbstractTextfieldHost));

    var TuiInputDateTimeModule = /** @class */ (function () {
        function TuiInputDateTimeModule() {
        }
        TuiInputDateTimeModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    angular2TextMask.TextMaskModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiPreventDefaultModule,
                    core$1.TuiCalendarModule,
                    core$1.TuiSvgModule,
                    core$1.TuiLinkModule,
                    core$1.TuiHostedDropdownModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    directives.TuiValueAccessorModule,
                    core$1.TuiTextfieldControllerModule,
                ],
                declarations: [TuiInputDateTimeComponent, TuiInputDateTimeDirective],
                exports: [
                    TuiInputDateTimeComponent,
                    TuiInputDateTimeDirective,
                    core$1.TuiTextfieldComponent,
                ],
            })
        ], TuiInputDateTimeModule);
        return TuiInputDateTimeModule;
    }());

    exports.TIME_STREAM_FACTORY = TIME_STREAM_FACTORY;
    exports.TUI_INPUT_DATE_TIME_PROVIDERS = TUI_INPUT_DATE_TIME_PROVIDERS;
    exports.TuiInputDateTimeComponent = TuiInputDateTimeComponent;
    exports.TuiInputDateTimeDirective = TuiInputDateTimeDirective;
    exports.TuiInputDateTimeModule = TuiInputDateTimeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
