// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-tag.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/components/tag'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-tag', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/components/tag', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', 'rxjs', 'rxjs/operators', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-tag'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.components.tag, global['taiga-ui'].kit.providers, global['taiga-ui'].kit.tokens, global.rxjs, global.rxjs.operators, global.ng.common, global.ngPolymorpheus));
}(this, (function (exports, core, forms, cdk, core$1, tag, providers, tokens, rxjs, operators, common, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_INPUT_TAG_DEFAULT_OPTIONS = {
        separator: ",",
        uniqueTags: true,
    };
    var TUI_INPUT_TAG_OPTIONS = new core.InjectionToken("Default parameters for input-tag", {
        factory: function () { return TUI_INPUT_TAG_DEFAULT_OPTIONS; },
    });

    var EVENT_Y_TO_X_COEFFICIENT = 3;
    var TuiInputTagComponent = /** @class */ (function (_super) {
        __extends(TuiInputTagComponent, _super);
        function TuiInputTagComponent(control, changeDetectorRef, tuiScrollService, elementRef, appearance, modeDirective, mode$, tagStatus, hintController, controller, options, parentHostedDropdown) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.tuiScrollService = tuiScrollService;
            _this.elementRef = elementRef;
            _this.appearance = appearance;
            _this.modeDirective = modeDirective;
            _this.mode$ = mode$;
            _this.tagStatus = tagStatus;
            _this.hintController = hintController;
            _this.controller = controller;
            _this.options = options;
            _this.parentHostedDropdown = parentHostedDropdown;
            _this.tags = cdk.EMPTY_QUERY;
            _this.scrollToStart$ = new rxjs.Subject();
            _this.scrollToEnd$ = new rxjs.Subject();
            // TODO: 3.0 Remove
            _this.allowSpaces = true;
            _this.separator = _this.options.separator;
            _this.icon = '';
            _this.iconAlign = 'right';
            _this.search = '';
            _this.editable = true;
            _this.tagValidator = cdk.ALWAYS_TRUE_HANDLER;
            _this.expandable = true;
            _this.inputHidden = false;
            _this.uniqueTags = _this.options.uniqueTags;
            _this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            _this.searchChange = new core.EventEmitter();
            _this.status$ = _this.mode$.pipe(operators.map(function () { return _this.status; }));
            _this.open = false;
            return _this;
        }
        TuiInputTagComponent_1 = TuiInputTagComponent;
        Object.defineProperty(TuiInputTagComponent.prototype, "pseudoFocusedSetter", {
            set: function (value) {
                if (!value && !this.focused) {
                    this.scrollToStart$.next();
                }
                this.pseudoFocused = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "scrollerSetter", {
            set: function (scroller) {
                this.initScrollerSubscription(scroller);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.focusableElement || this.computedDisabled
                    ? null
                    : this.focusableElement.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "focused", {
            get: function () {
                var _a;
                return (cdk.isNativeFocusedIn(this.elementRef.nativeElement) ||
                    !!((_a = this.hostedDropdown) === null || _a === void 0 ? void 0 : _a.focused));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "size", {
            get: function () {
                return this.controller.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "labelOutside", {
            get: function () {
                var _a = this.controller, size = _a.size, labelOutside = _a.labelOutside;
                return size === 's' || labelOutside;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasCleaner", {
            get: function () {
                return this.controller.cleaner && this.hasValue && this.interactive;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasNativeValue", {
            get: function () {
                return !!this.search;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasValue", {
            get: function () {
                return !!this.value.length || this.hasNativeValue;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasPlaceholder", {
            get: function () {
                return (!this.labelOutside ||
                    (!this.hasValue && (!this.hasExampleText || this.inputHidden)));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "placeholderRaised", {
            get: function () {
                return (!this.labelOutside &&
                    ((this.computedFocused && !this.readOnly) || this.hasValue));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasExampleText", {
            get: function () {
                return (!!this.controller.exampleText &&
                    this.computedFocused &&
                    !this.hasValue &&
                    !this.readOnly);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasTooltip", {
            get: function () {
                return !!this.hintController.content && !this.disabled;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "iconAlignLeft", {
            get: function () {
                return !!this.icon && this.iconAlign === 'left';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "iconAlignRight", {
            get: function () {
                return !!this.icon && this.iconAlign === 'right';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "hasRightIcons", {
            get: function () {
                return this.hasCleaner || this.hasTooltip || this.iconAlignRight;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "status", {
            get: function () {
                var _a;
                return ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.mode) ? 'default' : this.tagStatus;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTagComponent.prototype, "canOpen", {
            get: function () {
                return this.interactive && !!this.datalist;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputTagComponent.prototype.getLeftContent = function (tag) {
            return !this.tagValidator(tag) && this.errorIconTemplate
                ? this.errorIconTemplate
                : '';
        };
        TuiInputTagComponent.prototype.onCleanerClick = function () {
            var _a;
            this.updateSearch('');
            this.clear();
            this.focusInput();
            (_a = this.parentHostedDropdown) === null || _a === void 0 ? void 0 : _a.updateOpen(true);
        };
        TuiInputTagComponent.prototype.onActiveZone = function (active) {
            this.open = false;
            this.addTag();
            this.updateFocused(active);
            if (!this.computedFocused) {
                this.scrollToStart$.next();
            }
        };
        TuiInputTagComponent.prototype.onMouseDown = function (event) {
            var _a;
            var actualTarget = cdk.getActualTarget(event);
            if (!this.focusableElement ||
                actualTarget === this.focusableElement.nativeElement ||
                // TODO: iframe warning
                !(event.target instanceof Element) || ((_a = this.cleanerSvg) === null || _a === void 0 ? void 0 : _a.nativeElement.contains(event.target)) ||
                (this.tagsContainer &&
                    actualTarget !== this.tagsContainer.nativeElement &&
                    this.tagsContainer.nativeElement.contains(actualTarget))) {
                return;
            }
            event.preventDefault();
            this.focusInput();
        };
        TuiInputTagComponent.prototype.onFieldKeyDownBackspace = function (event) {
            if (!this.labelOutside && !this.hasNativeValue && this.value.length) {
                this.deleteLastEnabledItem();
            }
            else {
                this.onFieldKeyDownArrowLeft(event);
            }
        };
        TuiInputTagComponent.prototype.onFieldKeyDownArrowLeft = function (event) {
            if (!this.labelOutside || this.hasNativeValue || !this.value.length) {
                return;
            }
            event.preventDefault();
            cdk.setNativeFocused(this.tags.last.nativeElement);
        };
        TuiInputTagComponent.prototype.onFieldKeyDownEnter = function () {
            this.addTag();
            this.scrollToEnd$.next();
        };
        TuiInputTagComponent.prototype.onTagKeyDownArrowLeft = function (currentIndex) {
            if (currentIndex > 0) {
                this.onScrollKeyDown(currentIndex, -1);
            }
        };
        TuiInputTagComponent.prototype.onTagKeyDownArrowRight = function (currentIndex) {
            if (currentIndex === this.value.length - 1) {
                this.focusInput();
            }
            else {
                this.onScrollKeyDown(currentIndex, 1);
            }
        };
        TuiInputTagComponent.prototype.onTagEdited = function (value, index) {
            var _this = this;
            this.focusInput();
            this.updateValue(this.value
                .map(function (tag, tagIndex) {
                return tagIndex !== index
                    ? tag
                    : value
                        .split(_this.separator)
                        .map(function (tag) { return tag.trim(); })
                        .filter(Boolean);
            })
                .reduce(function (result, item) { return result.concat(item); }, []));
        };
        TuiInputTagComponent.prototype.handleOption = function (item) {
            this.focusInput();
            this.updateSearch('');
            this.updateValue(this.value.concat(item));
            this.open = false;
            this.scrollToEnd$.next();
        };
        TuiInputTagComponent.prototype.onInput = function (value) {
            var _this = this;
            var array = this.allowSpaces
                ? value.split(this.separator)
                : value.split(tag.ALLOWED_SPACE_REGEXP);
            var tags = array
                .map(function (item) { return item.trim(); })
                .filter(function (item, index, _a) {
                var length = _a.length;
                return item.length > 0 && index !== length - 1;
            });
            var validated = tags.filter(function (tag) { return !_this.disabledItemHandler(tag); });
            if (array.length > 1) {
                this.updateSearch(array[array.length - 1].trim());
                this.updateValue(__spread(this.value, validated));
            }
            else {
                this.updateSearch(value);
            }
            this.open = this.hasNativeValue;
        };
        TuiInputTagComponent.prototype.onHoveredChange = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputTagComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        TuiInputTagComponent.prototype.trackByFn = function (_, tag) {
            // Actually tag has TuiStringifiableItem type not string
            return tag.toString();
        };
        TuiInputTagComponent.prototype.updateValue = function (value) {
            var _this = this;
            var seen = new Set();
            _super.prototype.updateValue.call(this, value
                .reverse()
                .filter(function (item) {
                return !_this.uniqueTags || (!!item && !seen.has(item) && seen.add(item));
            })
                .reverse());
        };
        TuiInputTagComponent.prototype.onScrollKeyDown = function (currentIndex, flag) {
            var tag = this.tags.find(function (_item, index) { return index === currentIndex + flag; });
            if (!tag || !this.scrollBar) {
                return;
            }
            cdk.setNativeFocused(tag.nativeElement);
            if (flag * this.scrollBar.nativeElement.clientWidth -
                flag * tag.nativeElement.offsetLeft -
                tag.nativeElement.clientWidth <
                0) {
                this.scrollBar.nativeElement.scrollLeft +=
                    flag * tag.nativeElement.clientWidth;
            }
        };
        TuiInputTagComponent.prototype.initScrollerSubscription = function (scroller) {
            var _this = this;
            if (!(scroller === null || scroller === void 0 ? void 0 : scroller.browserScrollRef)) {
                return;
            }
            var nativeElement = scroller.browserScrollRef.nativeElement;
            var wheel$ = cdk.typedFromEvent(nativeElement, 'wheel', { passive: false }).pipe(operators.filter(function (event) { return event.deltaX === 0 && _this.shouldScroll(nativeElement); }), cdk.preventDefault(), operators.map(function (_a) {
                var deltaY = _a.deltaY;
                return Math.max(nativeElement.scrollLeft + deltaY * EVENT_Y_TO_X_COEFFICIENT, 0);
            }));
            var start$ = this.scrollToStart$.pipe(operators.mapTo(0));
            var end$ = this.scrollToEnd$.pipe(operators.map(function () { return nativeElement.scrollWidth; }));
            rxjs.merge(wheel$, start$, end$)
                .pipe(operators.switchMap(function (left) { return _this.tuiScrollService.scroll$(nativeElement, 0, left); }), operators.takeUntil(this.destroy$))
                .subscribe();
        };
        TuiInputTagComponent.prototype.updateSearch = function (value) {
            if (this.focusableElement) {
                this.focusableElement.nativeElement.value = value;
            }
            this.search = value;
            this.searchChange.emit(value);
        };
        TuiInputTagComponent.prototype.shouldScroll = function (_a) {
            var scrollWidth = _a.scrollWidth, offsetWidth = _a.offsetWidth;
            return scrollWidth > offsetWidth;
        };
        TuiInputTagComponent.prototype.addTag = function () {
            var inputValue = this.search.trim();
            if (!inputValue || this.disabledItemHandler(inputValue)) {
                return;
            }
            this.updateSearch('');
            this.updateValue(this.value.concat(inputValue));
        };
        TuiInputTagComponent.prototype.deleteLastEnabledItem = function () {
            for (var index = this.value.length - 1; index >= 0; index--) {
                if (!this.disabledItemHandler(this.value[index])) {
                    this.updateValue(__spread(this.value.slice(0, index), this.value.slice(index + 1, this.value.length)));
                    break;
                }
            }
        };
        TuiInputTagComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                cdk.setNativeFocused(this.nativeFocusableElement, true, preventScroll);
            }
        };
        var TuiInputTagComponent_1;
        TuiInputTagComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: cdk.TuiScrollService, decorators: [{ type: core.Inject, args: [cdk.TuiScrollService,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_APPEARANCE,] }] },
            { type: core$1.TuiModeDirective, decorators: [{ type: core.Optional }, { type: core.Inject, args: [core$1.TuiModeDirective,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [core$1.TUI_MODE,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_TAG_STATUS,] }] },
            { type: core$1.TuiHintControllerDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_HINT_WATCHED_CONTROLLER,] }] },
            { type: core$1.TuiTextfieldController, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [TUI_INPUT_TAG_OPTIONS,] }] },
            { type: core$1.TuiHostedDropdownComponent, decorators: [{ type: core.Optional }, { type: core.Inject, args: [core$1.TuiHostedDropdownComponent,] }] }
        ]; };
        __decorate([
            core.ViewChild(core$1.TuiHostedDropdownComponent)
        ], TuiInputTagComponent.prototype, "hostedDropdown", void 0);
        __decorate([
            core.ViewChild('focusableElement')
        ], TuiInputTagComponent.prototype, "focusableElement", void 0);
        __decorate([
            core.ViewChild('tagsContainer')
        ], TuiInputTagComponent.prototype, "tagsContainer", void 0);
        __decorate([
            core.ViewChildren('tag', { read: core.ElementRef })
        ], TuiInputTagComponent.prototype, "tags", void 0);
        __decorate([
            core.ViewChild('cleaner', { read: core.ElementRef })
        ], TuiInputTagComponent.prototype, "cleanerSvg", void 0);
        __decorate([
            core.ViewChild(core$1.TuiScrollbarComponent, { read: core.ElementRef })
        ], TuiInputTagComponent.prototype, "scrollBar", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "allowSpaces", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "separator", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "icon", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "iconAlign", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "search", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "editable", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "tagValidator", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class._expandable'),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "expandable", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "inputHidden", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "uniqueTags", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTagComponent.prototype, "disabledItemHandler", void 0);
        __decorate([
            core.Input('pseudoFocused')
        ], TuiInputTagComponent.prototype, "pseudoFocusedSetter", null);
        __decorate([
            core.Output()
        ], TuiInputTagComponent.prototype, "searchChange", void 0);
        __decorate([
            core.ContentChild(core$1.TuiDataListDirective, { read: core.TemplateRef })
        ], TuiInputTagComponent.prototype, "datalist", void 0);
        __decorate([
            core.ViewChild('errorIcon')
        ], TuiInputTagComponent.prototype, "errorIconTemplate", void 0);
        __decorate([
            core.ViewChild(core$1.TuiScrollbarComponent)
        ], TuiInputTagComponent.prototype, "scrollerSetter", null);
        __decorate([
            core.HostBinding('attr.data-size')
        ], TuiInputTagComponent.prototype, "size", null);
        __decorate([
            core.HostBinding('class._label-outside')
        ], TuiInputTagComponent.prototype, "labelOutside", null);
        TuiInputTagComponent = TuiInputTagComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-tag',
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"canOpen\"\n    [content]=\"datalist || ''\"\n    [(open)]=\"open\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-wrapper\n        [appearance]=\"appearance\"\n        [readOnly]=\"readOnly\"\n        [focused]=\"computedFocused\"\n        [hovered]=\"computedHovered\"\n        [disabled]=\"computedDisabled\"\n        [invalid]=\"computedInvalid\"\n        (tuiHoveredChange)=\"onHoveredChange($event)\"\n        (mousedown)=\"onMouseDown($event)\"\n    >\n        <div class=\"t-content\">\n            <div\n                *ngIf=\"iconAlignLeft\"\n                class=\"t-icons t-icons_left\"\n            >\n                <tui-svg\n                    class=\"t-icon\"\n                    [src]=\"icon\"\n                ></tui-svg>\n            </div>\n            <div class=\"t-wrapper\">\n                <div class=\"t-absolute-wrapper\">\n                    <div\n                        *ngIf=\"hasPlaceholder\"\n                        automation-id=\"tui-input-tag__placeholder\"\n                        class=\"t-placeholder\"\n                        [class.t-placeholder_raised]=\"placeholderRaised\"\n                    >\n                        <ng-content></ng-content>\n                    </div>\n                </div>\n                <tui-scrollbar\n                    *ngIf=\"!expandable; else items\"\n                    class=\"t-scrollbar\"\n                    [hidden]=\"true\"\n                >\n                    <ng-container [ngTemplateOutlet]=\"items\"></ng-container>\n                </tui-scrollbar>\n                <ng-template #items>\n                    <div\n                        #tagsContainer\n                        class=\"t-tags\"\n                    >\n                        <ng-container *ngIf=\"labelOutside; else text\">\n                            <tui-tag\n                                *ngFor=\"let item of value; index as index; trackBy: trackByFn\"\n                                #tag\n                                automation-id=\"tui-input-tag__tag\"\n                                class=\"t-tag\"\n                                [status]=\"(status$ | async)!\"\n                                [leftContent]=\"getLeftContent(item)\"\n                                [allowSpaces]=\"allowSpaces\"\n                                [tuiFocusable]=\"false\"\n                                [disabled]=\"computedDisabled || disabledItemHandler(item)\"\n                                [editable]=\"editable && !readOnly\"\n                                [hoverable]=\"!readOnly\"\n                                [removable]=\"!readOnly\"\n                                [separator]=\"separator\"\n                                [maxLength]=\"controller.maxLength\"\n                                [size]=\"controller.size\"\n                                [value]=\"item.toString()\"\n                                (edited)=\"onTagEdited($event, index)\"\n                                (keydown.arrowLeft.prevent)=\"onTagKeyDownArrowLeft(index)\"\n                                (keydown.arrowRight.prevent)=\"onTagKeyDownArrowRight(index)\"\n                            ></tui-tag>\n                        </ng-container>\n                        <ng-template #text>\n                            <span\n                                *ngFor=\"let item of value\"\n                                class=\"t-text\"\n                                [class.t-text_disabled]=\"disabledItemHandler(item)\"\n                                [class.t-text_error]=\"!tagValidator(item)\"\n                                [textContent]=\"item\"\n                            ></span>\n                        </ng-template>\n                        <div\n                            class=\"t-input-wrapper\"\n                            [class.t-input-wrapper_collapsed]=\"computedDisabled || readOnly || inputHidden\"\n                        >\n                            <div class=\"t-ghost\">{{ search }}</div>\n                            <input\n                                #focusableElement\n                                type=\"text\"\n                                automation-id=\"tui-input-tag__native\"\n                                class=\"t-native\"\n                                [placeholder]=\"controller.exampleText\"\n                                [class.t-native_hidden]=\"inputHidden\"\n                                [attr.maxLength]=\"controller.maxLength\"\n                                [id]=\"id\"\n                                [disabled]=\"computedDisabled\"\n                                [readOnly]=\"readOnly || inputHidden\"\n                                [tuiFocusable]=\"computedFocusable\"\n                                [ngModel]=\"search\"\n                                (ngModelChange)=\"onInput($event)\"\n                                (keydown.arrowLeft)=\"onFieldKeyDownArrowLeft($event)\"\n                                (keydown.backspace)=\"onFieldKeyDownBackspace($event)\"\n                                (keydown.enter.prevent)=\"onFieldKeyDownEnter()\"\n                            />\n                        </div>\n                    </div>\n                </ng-template>\n            </div>\n            <div\n                *ngIf=\"hasRightIcons\"\n                class=\"t-icons t-icons_right\"\n            >\n                <tui-svg\n                    *ngIf=\"hasCleaner\"\n                    #cleaner\n                    automation-id=\"tui-input-tag__cleaner\"\n                    src=\"tuiIconCloseLarge\"\n                    class=\"t-cleaner\"\n                    (click.stop)=\"onCleanerClick()\"\n                ></tui-svg>\n                <tui-tooltip\n                    *ngIf=\"hasTooltip\"\n                    automation-id=\"tui-input-tag__tooltip\"\n                    describeId=\"placeholer_until_accesibility_is_added\"\n                    class=\"t-tooltip\"\n                    [direction]=\"hintController.direction\"\n                    [mode]=\"hintController.mode\"\n                    [content]=\"hintController.content || ''\"\n                    [showDelay]=\"hintController.showDelay\"\n                    [hideDelay]=\"hintController.hideDelay\"\n                ></tui-tooltip>\n                <tui-svg\n                    *ngIf=\"iconAlignRight\"\n                    class=\"t-icon\"\n                    [src]=\"icon\"\n                ></tui-svg>\n            </div>\n        </div>\n        <ng-template #errorIcon>\n            <tui-svg\n                src=\"tuiIconAttention\"\n                class=\"t-error-icon\"\n            ></tui-svg>\n        </ng-template>\n    </tui-wrapper>\n</tui-hosted-dropdown>\n",
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputTagComponent_1; }),
                    },
                    {
                        provide: core$1.TUI_DATA_LIST_HOST,
                        useExisting: core.forwardRef(function () { return TuiInputTagComponent_1; }),
                    },
                    providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER,
                    core$1.TEXTFIELD_CONTROLLER_PROVIDER,
                    core$1.HINT_CONTROLLER_PROVIDER,
                    core$1.MODE_PROVIDER,
                ],
                styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);position:relative;display:block;border-radius:var(--tui-radius-m);text-align:left}:host[data-size='s']{height:var(--tui-height-s);min-height:var(--tui-height-s);max-height:var(--tui-height-s)}:host[data-size='m']{height:var(--tui-height-m);min-height:var(--tui-height-m);max-height:var(--tui-height-m)}:host[data-size='l']{height:var(--tui-height-l);min-height:var(--tui-height-l);max-height:var(--tui-height-l);font:var(--tui-font-text-m);line-height:1.25rem}.t-input{font:var(--tui-font-text-s);color:var(--tui-text-01);padding:0;margin:0;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;padding:0 var(--tui-padding-m);border:solid transparent;border-width:0 var(--border-end,0) 0 var(--border-start,0);border-inline-start-width:var(--border-start,0);border-inline-end-width:var(--border-end,0);text-indent:var(--text-indent);text-align:inherit;box-sizing:border-box;white-space:nowrap;overflow:hidden;text-transform:inherit;resize:none}.t-input:-webkit-autofill,.t-input:-webkit-autofill:focus,.t-input:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill,.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill:focus,.t-input :host-context(tui-primitive-textfield[data-mode=onDark]):-webkit-autofill:hover,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill:focus,.t-input :host-context(tui-text-area[data-mode=onDark]):-webkit-autofill:hover,:host[data-mode=onDark] .t-input:-webkit-autofill,:host[data-mode=onDark] .t-input:-webkit-autofill:focus,:host[data-mode=onDark] .t-input:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01-night)!important;border-color:var(--tui-autofill-night);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill-night) inset!important}.t-input :host-context(tui-primitive-textfield[data-size='s']):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size='s']):not(tui-text-area),:host[data-size='s'] .t-input{padding:0 var(--tui-padding-s)}.t-input :host-context(tui-primitive-textfield[data-size='l']):not(tui-primitive-textfield),.t-input :host-context(tui-text-area[data-size='l']):not(tui-text-area),:host[data-size='l'] .t-input{padding:0 var(--tui-padding-l)}.t-input :host-context(tui-primitive-textfield._disabled),.t-input :host-context(tui-text-area._disabled),:host._disabled .t-input{pointer-events:none}.t-input :host-context(tui-primitive-textfield[data-size='l']:not(._label-outside)):not(tui-primitive-textfield),:host[data-size='l']:not(._label-outside) .t-input{padding-top:1.25rem}.t-input :host-context(tui-primitive-textfield[data-size='l']:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder,:host[data-size='l']:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder{font-size:.8156rem;transform:translateY(-.625rem)}.t-input :host-context(tui-primitive-textfield[data-size='m']:not(._label-outside)):not(tui-primitive-textfield),:host[data-size='m']:not(._label-outside) .t-input{padding-top:1.125rem}.t-input :host-context(tui-primitive-textfield[data-size='m']:not(._label-outside)):not(tui-primitive-textfield):-webkit-autofill+.t-content .t-placeholder,:host[data-size='m']:not(._label-outside) .t-input:-webkit-autofill+.t-content .t-placeholder{font-size:.69rem;transform:translateY(-.5rem)}.t-input :host-context(tui-primitive-textfield._hidden),:host._hidden input.t-input{opacity:0;text-indent:-10em;-webkit-user-select:none}.t-content{display:flex;height:100%;width:100%;padding:0 var(--tui-padding-m);box-sizing:border-box;align-items:center;overflow:hidden}:host[data-size='s'] .t-content{padding:0 var(--tui-padding-s)}:host[data-size='l'] .t-content{padding:0 var(--tui-padding-l)}.t-content:after{content:'';margin-right:-.25rem}:host[data-size='m'] .t-content:after{display:none}.t-wrapper{flex:1;min-width:0;-webkit-padding-end:.25rem;padding-inline-end:.25rem;-webkit-padding-start:0;padding-inline-start:0}.t-placeholder{transition-property:transform,font-size,color,letter-spacing;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;width:100%;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;color:var(--tui-text-02);pointer-events:none;will-change:transform;transform:translateY(0)}.t-placeholder_raised{transform:translateY(-.625rem)}:host[data-size='m'] .t-placeholder_raised{font:var(--tui-font-text-xs);transform:translateY(-.5rem);letter-spacing:.025rem}:host._invalid:not(._focused) .t-placeholder_raised,:host._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill)}:host[data-mode=onDark]._invalid:not(._focused) .t-placeholder_raised,:host[data-mode=onDark]._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill-night)}:host._focused .t-placeholder,:host[data-size='l']._focused._label-outside .t-placeholder,:host[data-size='m']._focused._label-outside .t-placeholder{color:var(--tui-text-03)}:host[data-size='l'] .t-placeholder{font-size:.9375rem}:host[data-size='l'] .t-placeholder_raised{font-size:.8156rem}:host[data-size='l']._focused:not(._label-outside) .t-placeholder,:host[data-size='m']._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01)}:host[data-mode=onDark] .t-placeholder{color:var(--tui-text-02-night)}:host[data-size='l'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01-night)}:host[data-mode=onDark]._focused .t-placeholder,:host[data-size='l'][data-mode=onDark]._focused._label-outside .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused._label-outside .t-placeholder{color:var(--tui-text-02-night)}@supports (-webkit-hyphens:none){.t-placeholder{will-change:unset;transition-property:transform,color,letter-spacing}}.t-cleaner{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03);position:relative;box-sizing:border-box;cursor:pointer;transition-property:color,transform}.t-cleaner:hover{color:var(--tui-text-02)}:host._disabled .t-cleaner,:host._readonly .t-cleaner{pointer-events:none}:host[data-mode=onDark] .t-cleaner{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-cleaner:hover{color:var(--tui-text-01-night)}.t-icon{display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03)}:host[data-mode=onDark] .t-icon{color:var(--tui-text-03-night)}.t-icon_left{margin:0 .5rem 0 -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem}:host[data-size='s'] .t-icon_left{margin-right:.25rem;-webkit-margin-end:.25rem;margin-inline-end:.25rem}:host{cursor:text}:host._expandable{height:auto;max-height:none}:host._disabled{pointer-events:none}:host._readonly{cursor:default}.t-hosted{display:block;border-radius:inherit}.t-scrollbar{margin:0 -.25rem;padding:0 .25rem}.t-tags{display:flex;align-items:center;box-sizing:border-box}:host._expandable .t-tags{flex-wrap:wrap;white-space:normal}:host._readonly .t-tags{pointer-events:none}:host[data-size='s'] .t-tags{min-height:var(--tui-height-s);padding:.25rem .25rem .25rem 0}:host[data-size='m'] .t-tags{min-height:var(--tui-height-m);padding:.5rem .25rem .5rem 0}:host[data-size='l'] .t-tags{min-height:var(--tui-height-l);padding:.625rem .25rem .625rem 0}:host[data-size='m']:not(._label-outside) .t-tags{padding:1.1875rem 1rem 0 0}:host[data-size='l']:not(._label-outside) .t-tags{padding:1.6875rem 1rem .5625rem 0}.t-tags :host[data-size]{-webkit-padding-start:0;padding-inline-start:0;-webkit-padding-end:.25rem;padding-inline-end:.25rem}:host[data-size]:not(._label-outside) .t-tags{-webkit-padding-end:1rem;padding-inline-end:1rem}.t-content{align-items:flex-start}.t-icons{display:flex;align-items:center;padding:0 0 0 .75rem;-webkit-padding-start:.75rem;padding-inline-start:.75rem;-webkit-padding-end:0;padding-inline-end:0}.t-icons_left{padding:0;margin:0 .5rem 0 -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem}:host._expandable .t-icons_right{margin:0 0 0 -.625rem;-webkit-margin-start:-.625rem;margin-inline-start:-.625rem;-webkit-margin-end:0;margin-inline-end:0}:host[data-size='s'] .t-icons{height:var(--tui-height-s)}:host[data-size='m'] .t-icons{height:var(--tui-height-m)}:host[data-size='l'] .t-icons{height:var(--tui-height-l)}.t-absolute-wrapper{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;flex-direction:column;justify-content:center}:host[data-size='m'] .t-absolute-wrapper{height:var(--tui-height-m)}:host[data-size='l'] .t-absolute-wrapper{height:var(--tui-height-l)}.t-wrapper{position:relative;padding:0}.t-tag{margin:.125rem .5rem .125rem -.25rem;-webkit-margin-start:-.25rem;margin-inline-start:-.25rem;-webkit-margin-end:.5rem;margin-inline-end:.5rem;max-width:100%;flex-shrink:0}.t-text:after{content:',\\00a0'}.t-text_disabled{color:var(--tui-text-03)}.t-text_error{color:var(--tui-negative)}:host:not(._expandable) .t-text{white-space:nowrap}:host:not(._focused) .t-text:last-of-type:after{content:''}.t-input-wrapper{position:relative;flex:1;max-width:100%}.t-input-wrapper_collapsed{flex:0;margin:0 0 0 -.5rem;-webkit-margin-start:-.5rem;margin-inline-start:-.5rem;-webkit-margin-end:0;margin-inline-end:0}:host[data-size='s'] .t-input-wrapper{min-height:1.5rem}:host[data-size='m'] .t-input-wrapper{min-height:calc(var(--tui-height-xs) + 2 * .125rem)}:host[data-size='l'] .t-input-wrapper{min-height:calc(var(--tui-height-s) + 2 * .125rem)}:host:not(._label-outside) .t-input-wrapper{min-height:1.25rem}.t-ghost{visibility:hidden;white-space:pre;text-overflow:clip;min-width:.125rem}:host:not(._expandable) .t-ghost{min-width:2rem}.t-native{margin:0;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;outline:0;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;box-sizing:border-box;padding:0;cursor:inherit}.t-native:-webkit-autofill,.t-native:-webkit-autofill:focus,.t-native:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}.t-native_hidden{opacity:0;text-indent:-10em}.t-native::-ms-input-placeholder{color:var(--tui-text-03);opacity:0}.t-native::placeholder{color:var(--tui-text-03);opacity:0}:host[data-mode=onDark] .t-native::-ms-input-placeholder{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-native::placeholder{color:var(--tui-text-03-night)}:host._focused .t-native:not(:read-only)::-ms-input-placeholder{opacity:1}:host._focused .t-native:not(:-moz-read-only)::placeholder{opacity:1}:host._focused .t-native:not(:read-only)::placeholder{opacity:1}.t-error-icon{display:block;color:var(--tui-error-fill);width:1rem;height:1rem}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(cdk.TuiScrollService)),
            __param(3, core.Inject(core.ElementRef)),
            __param(4, core.Inject(core$1.TUI_TEXTFIELD_APPEARANCE)),
            __param(5, core.Optional()),
            __param(5, core.Inject(core$1.TuiModeDirective)),
            __param(6, core.Inject(core$1.TUI_MODE)),
            __param(7, core.Inject(tokens.TUI_TAG_STATUS)),
            __param(8, core.Inject(core$1.TUI_HINT_WATCHED_CONTROLLER)),
            __param(9, core.Inject(core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER)),
            __param(10, core.Inject(TUI_INPUT_TAG_OPTIONS)),
            __param(11, core.Optional()),
            __param(11, core.Inject(core$1.TuiHostedDropdownComponent))
        ], TuiInputTagComponent);
        return TuiInputTagComponent;
    }(cdk.AbstractTuiMultipleControl));

    var TuiInputTagModule = /** @class */ (function () {
        function TuiInputTagModule() {
        }
        TuiInputTagModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiHoveredModule,
                    cdk.TuiActiveZoneModule,
                    core$1.TuiSvgModule,
                    core$1.TuiScrollbarModule,
                    core$1.TuiTooltipModule,
                    core$1.TuiHostedDropdownModule,
                    tag.TuiTagModule,
                    core$1.TuiWrapperModule,
                ],
                providers: [cdk.TuiScrollService],
                declarations: [TuiInputTagComponent],
                exports: [TuiInputTagComponent],
            })
        ], TuiInputTagModule);
        return TuiInputTagModule;
    }());

    exports.TUI_INPUT_TAG_DEFAULT_OPTIONS = TUI_INPUT_TAG_DEFAULT_OPTIONS;
    exports.TUI_INPUT_TAG_OPTIONS = TUI_INPUT_TAG_OPTIONS;
    exports.TuiInputTagComponent = TuiInputTagComponent;
    exports.TuiInputTagModule = TuiInputTagModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
