// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/abstract'), require('@taiga-ui/kit/components/input-number'), require('@taiga-ui/kit/components/range'), require('@taiga-ui/kit/constants'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/abstract', '@taiga-ui/kit/components/input-number', '@taiga-ui/kit/components/range', '@taiga-ui/kit/constants', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-range'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.abstract, global['taiga-ui'].kit.components['input-number'], global['taiga-ui'].kit.components.range, global['taiga-ui'].kit.constants, global.ng.common, global.ngPolymorpheus));
}(this, (function (exports, core, forms, cdk, core$1, abstract, inputNumber, range, constants, common, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    // @dynamic
    var TuiInputRangeComponent = /** @class */ (function (_super) {
        __extends(TuiInputRangeComponent, _super);
        function TuiInputRangeComponent(control, changeDetectorRef, isMobile, appearance, elementRef, controller, isNew) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.isMobile = isMobile;
            _this.appearance = appearance;
            _this.elementRef = elementRef;
            _this.controller = controller;
            _this.isNew = isNew;
            _this.inputNumberRefs = cdk.EMPTY_QUERY;
            _this.rangeRef = null;
            _this.min = 0;
            /* TODO: make `100` as default value (like in native sliders) */
            _this.max = Infinity;
            _this.quantum = 1;
            _this.steps = 0;
            _this.segments = 0;
            _this.keySteps = null;
            _this.leftValueContent = '';
            _this.rightValueContent = '';
            _this.lastActiveSide = 'left';
            return _this;
        }
        TuiInputRangeComponent_1 = TuiInputRangeComponent;
        Object.defineProperty(TuiInputRangeComponent.prototype, "leftFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.inputNumberRefs.first) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "rightFocusableElement", {
            get: function () {
                var _a;
                return ((_a = this.inputNumberRefs.last) === null || _a === void 0 ? void 0 : _a.nativeFocusableElement) || null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.disabled
                    ? null
                    : this.leftFocusableElement || this.rightFocusableElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocusedIn(this.elementRef.nativeElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showLeftValueContent", {
            get: function () {
                var _a;
                return Boolean((this.minLabel || this.leftValueContent) &&
                    !cdk.isNativeFocused(this.leftFocusableElement) &&
                    !(((_a = this.rangeRef) === null || _a === void 0 ? void 0 : _a.focused) && this.lastActiveSide === 'left'));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showRightValueContent", {
            get: function () {
                var _a;
                return Boolean((this.maxLabel || this.rightValueContent) &&
                    !cdk.isNativeFocused(this.rightFocusableElement) &&
                    !(((_a = this.rangeRef) === null || _a === void 0 ? void 0 : _a.focused) && this.lastActiveSide === 'right'));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "precision", {
            get: function () {
                return core$1.getFractionPartPadded(this.quantum).length;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "decimal", {
            get: function () {
                return this.precision ? 'not-zero' : 'never';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "computedSteps", {
            get: function () {
                return this.steps || (this.max - this.min) / this.quantum;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "computedSize", {
            get: function () {
                return this.isNew && this.controller.size !== 's'
                    ? this.controller.size
                    : this.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "legacyLabelOutside", {
            get: function () {
                return this.isNew ? this.controller.labelOutside : this.computedSize === 'm';
            },
            enumerable: true,
            configurable: true
        });
        TuiInputRangeComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiInputRangeComponent.prototype.onTextInputFocused = function (focused, right) {
            if (focused) {
                return;
            }
            var _a = __read(this.inputNumberRefs, 2), leftTextInputRef = _a[0], rightTextInputRef = _a[1];
            var inputRef = right ? rightTextInputRef : leftTextInputRef;
            var valueIndex = right ? 1 : 0;
            if (!inputRef.nativeValue || inputRef.value !== this.value[valueIndex]) {
                this.updateTextInputValue(this.safeCurrentValue[valueIndex], right);
            }
        };
        TuiInputRangeComponent.prototype.changeByStep = function (event, _a) {
            var _b = __read(_a, 2), leftCoefficient = _b[0], rightCoefficient = _b[1];
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            var newValue = this.valueGuard([
                this.value[0] + leftCoefficient * this.step,
                this.value[1] + rightCoefficient * this.step,
            ]);
            var leftValueChanged = newValue[0] !== this.value[0];
            var rightValueChanged = newValue[1] !== this.value[1];
            if (leftValueChanged || rightValueChanged) {
                this.safelyUpdateValue(newValue);
                this.updateTextInputValue(newValue[rightValueChanged ? 1 : 0], rightValueChanged);
            }
        };
        TuiInputRangeComponent.prototype.onInputLeft = function (value) {
            this.safelyUpdateValue([value !== null && value !== void 0 ? value : this.safeCurrentValue[0], this.value[1]]);
        };
        TuiInputRangeComponent.prototype.onInputRight = function (value) {
            this.safelyUpdateValue([this.value[0], value !== null && value !== void 0 ? value : this.safeCurrentValue[1]]);
        };
        TuiInputRangeComponent.prototype.onRangeValue = function (value) {
            this.safelyUpdateValue(value);
            var rightValueChanged = this.lastActiveSide === 'right';
            this.updateTextInputValue(this.value[rightValueChanged ? 1 : 0], rightValueChanged);
        };
        TuiInputRangeComponent.prototype.focusToTextInput = function () {
            var element = this.lastActiveSide === 'left'
                ? this.leftFocusableElement
                : this.rightFocusableElement;
            if (!this.isMobile && element) {
                element.focus();
            }
        };
        TuiInputRangeComponent.prototype.onActiveThumbChange = function (activeThumb) {
            this.lastActiveSide = activeThumb;
        };
        TuiInputRangeComponent.prototype.getFallbackValue = function () {
            return [0, 0];
        };
        TuiInputRangeComponent.prototype.safelyUpdateValue = function (value) {
            this.updateValue(this.valueGuard(value));
        };
        TuiInputRangeComponent.prototype.valueGuard = function (_a) {
            var _b = __read(_a, 2), leftValue = _b[0], rightValue = _b[1];
            var leftCalibratedValue = this.calibrate(leftValue);
            var rightCalibratedValue = this.calibrate(rightValue);
            return [
                Math.min(leftCalibratedValue, this.value[1]),
                Math.max(rightCalibratedValue, this.value[0]),
            ];
        };
        TuiInputRangeComponent.prototype.calibrate = function (value) {
            var roundedValue = cdk.round(Math.round(value / this.quantum) * this.quantum, constants.TUI_FLOATING_PRECISION);
            return cdk.clamp(roundedValue, this.min, this.max);
        };
        TuiInputRangeComponent.prototype.updateTextInputValue = function (value, right) {
            var _a = __read(this.inputNumberRefs, 2), leftInputRef = _a[0], rightInputRef = _a[1];
            var textInputRef = right ? rightInputRef : leftInputRef;
            if (textInputRef) {
                textInputRef.nativeValue = textInputRef.getFormattedValue(value);
            }
        };
        var TuiInputRangeComponent_1;
        TuiInputRangeComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_APPEARANCE,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core$1.TuiTextfieldController, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER,] }] },
            { type: range.TuiNewRangeDirective, decorators: [{ type: core.Optional }, { type: core.Inject, args: [range.TuiNewRangeDirective,] }] }
        ]; };
        __decorate([
            core.ViewChildren(inputNumber.TuiInputNumberComponent)
        ], TuiInputRangeComponent.prototype, "inputNumberRefs", void 0);
        __decorate([
            core.ViewChild(range.TuiRangeComponent)
        ], TuiInputRangeComponent.prototype, "rangeRef", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp(abstract.quantumAssertion, 'Quantum must be positive')
        ], TuiInputRangeComponent.prototype, "quantum", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "steps", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "segments", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "keySteps", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "leftValueContent", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputRangeComponent.prototype, "rightValueContent", void 0);
        __decorate([
            core.HostBinding('class._label-outside')
        ], TuiInputRangeComponent.prototype, "legacyLabelOutside", null);
        TuiInputRangeComponent = TuiInputRangeComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-range',
                template: "<tui-wrapper\n    class=\"t-wrapper\"\n    [appearance]=\"appearance\"\n    [readOnly]=\"readOnly\"\n    [disabled]=\"disabled\"\n    [focused]=\"computedFocused\"\n    [hovered]=\"computedHovered\"\n    [invalid]=\"computedInvalid\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-input-number\n        tuiTextfieldAppearance=\"none\"\n        automation-id=\"tui-input-range__left-input\"\n        class=\"t-left t-text-input\"\n        [min]=\"min\"\n        [max]=\"value[1]\"\n        [precision]=\"precision\"\n        [decimal]=\"decimal\"\n        [postfix]=\"pluralizeMap && !showLeftValueContent ? (value[0] | i18nPlural: pluralizeMap) : ''\"\n        [disabled]=\"computedDisabled\"\n        [readOnly]=\"readOnly\"\n        [tuiTextfieldSize]=\"computedSize\"\n        [tuiTextfieldLabelOutside]=\"legacyLabelOutside\"\n        [ngModel]=\"value[0]\"\n        (ngModelChange)=\"onInputLeft($event)\"\n        (focusedChange)=\"onTextInputFocused($event, false)\"\n        (keydown.arrowUp)=\"changeByStep($event, [1, 0])\"\n        (keydown.arrowDown)=\"changeByStep($event, [-1, 0])\"\n    >\n        <ng-content></ng-content>\n        <div\n            *ngIf=\"showLeftValueContent\"\n            polymorpheus-outlet\n            [content]=\"minLabel ? legacyMinLabel : leftValueContent\"\n            [context]=\"{$implicit: value[0]}\"\n        ></div>\n    </tui-input-number>\n\n    <div class=\"t-right\">\n        <tui-input-number\n            tuiTextfieldAppearance=\"none\"\n            automation-id=\"tui-input-range__right-input\"\n            class=\"t-text-input\"\n            [min]=\"value[0]\"\n            [max]=\"max\"\n            [precision]=\"precision\"\n            [decimal]=\"decimal\"\n            [disabled]=\"computedDisabled\"\n            [readOnly]=\"readOnly\"\n            [tuiTextfieldSize]=\"computedSize\"\n            [tuiTextfieldLabelOutside]=\"legacyLabelOutside\"\n            [ngModel]=\"value[1]\"\n            (ngModelChange)=\"onInputRight($event)\"\n            (focusedChange)=\"onTextInputFocused($event, true)\"\n            (keydown.arrowUp)=\"changeByStep($event, [0, 1])\"\n            (keydown.arrowDown)=\"changeByStep($event, [0, -1])\"\n        >\n            <div\n                *ngIf=\"showRightValueContent\"\n                polymorpheus-outlet\n                [content]=\"maxLabel ? legacyMaxLabel : rightValueContent\"\n                [context]=\"{$implicit: value[1]}\"\n            ></div>\n        </tui-input-number>\n\n        <!-- TODO replace by postfix of the right InputNumber (after fix https://github.com/Tinkoff/taiga-ui/issues/1193) -->\n        <span\n            *ngIf=\"!showRightValueContent && pluralizeMap\"\n            automation-id=\"tui-input-range__pluralize-right\"\n            class=\"t-pluralize-right\"\n        >\n            &nbsp;{{ value[1] | i18nPlural: pluralizeMap }}\n        </span>\n    </div>\n\n    <tui-range\n        class=\"t-slider\"\n        [min]=\"min\"\n        [max]=\"max\"\n        [steps]=\"computedSteps\"\n        [segments]=\"segments\"\n        [keySteps]=\"keySteps\"\n        [pluralize]=\"segmentsPluralizeMap\"\n        [focusable]=\"false\"\n        [disabled]=\"readOnly || disabled\"\n        [ngModel]=\"value\"\n        (ngModelChange)=\"onRangeValue($event)\"\n        (tuiPressedChange)=\"focusToTextInput()\"\n        (activeThumbChange)=\"onActiveThumbChange($event)\"\n    ></tui-range>\n</tui-wrapper>\n\n<!-- TODO: 3.0 remove -->\n<ng-template\n    #legacyMinLabel\n    let-value\n>\n    <ng-container *ngIf=\"value === min\">{{ minLabel }}</ng-container>\n    <ng-container *ngIf=\"value !== min\">\n        {{ value }}&nbsp;{{ value | i18nPlural: pluralizeMap || pluralizeMapFallback }}\n    </ng-container>\n</ng-template>\n\n<!-- TODO: 3.0 remove -->\n<ng-template\n    #legacyMaxLabel\n    let-value\n>\n    <ng-container *ngIf=\"value === max\">{{ maxLabel }}</ng-container>\n    <ng-container *ngIf=\"value !== max\">\n        {{ value }}&nbsp;{{ value | i18nPlural: pluralizeMap || pluralizeMapFallback }}\n    </ng-container>\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputRangeComponent_1; }),
                    },
                    core$1.TEXTFIELD_CONTROLLER_PROVIDER,
                ],
                host: {
                    /**
                     * TODO: 3.0 delete it
                     * Dont forget to clear html-tags
                     */
                    '[class._show-ticks-labels]': '!isNew',
                },
                styles: [".t-wrapper{position:relative}.t-left{width:100%}.t-right{position:absolute;right:0;top:0;display:flex;width:50%;height:100%;text-align:right}.t-right .t-text-input{flex:1}.t-text-input{text-align:inherit}.t-pluralize-right{display:flex;align-items:center;padding:1.125rem var(--tui-padding-m) 0 0;margin-left:-.75rem;font:var(--tui-font-text-s)}:host[data-size='l'] .t-pluralize-right{font:var(--tui-font-text-m);padding-top:1.25rem;margin-left:-1rem;padding-right:var(--tui-padding-l)}:host._label-outside .t-pluralize-right{padding-top:0}:host._disabled .t-pluralize-right{color:var(--tui-text-03)}:host{display:block;border-radius:var(--tui-radius-m)}:host._show-ticks-labels._segmented{border-bottom:26px solid transparent}.t-slider{position:absolute;top:100%;left:0;right:0;z-index:1;margin:-.125rem 0 0;background:0 0}"]
            })
            /**
             * `AbstractTuiInputSlider` includes all legacy code (it can be deleted in v3.0)
             * TODO replace `extends AbstractTuiInputSlider<[number, number]>` by `extends AbstractTuiControl<[number, number]> implements TuiWithOptionalMinMax<number>`
             */
            ,
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(cdk.TUI_IS_MOBILE)),
            __param(3, core.Inject(core$1.TUI_TEXTFIELD_APPEARANCE)),
            __param(4, core.Inject(core.ElementRef)),
            __param(5, core.Inject(core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER)),
            __param(6, core.Optional()),
            __param(6, core.Inject(range.TuiNewRangeDirective))
        ], TuiInputRangeComponent);
        return TuiInputRangeComponent;
    }(abstract.AbstractTuiInputSlider));
    function tuiTextfieldAppearanceDirectiveFactory(_a) {
        var nativeElement = _a.nativeElement;
        return nativeElement.getAttribute('tuiTextfieldAppearance');
    }
    var TuiTextfieldAppearanceDirective = /** @class */ (function () {
        function TuiTextfieldAppearanceDirective() {
        }
        TuiTextfieldAppearanceDirective = __decorate([
            core.Directive({
                selector: '[tuiTextfieldAppearance]',
                providers: [
                    {
                        provide: core$1.TUI_TEXTFIELD_APPEARANCE,
                        deps: [core.ElementRef],
                        useFactory: tuiTextfieldAppearanceDirectiveFactory,
                    },
                ],
            })
        ], TuiTextfieldAppearanceDirective);
        return TuiTextfieldAppearanceDirective;
    }());

    var TuiInputRangeModule = /** @class */ (function () {
        function TuiInputRangeModule() {
        }
        TuiInputRangeModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    forms.ReactiveFormsModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiActiveZoneModule,
                    inputNumber.TuiInputNumberModule,
                    cdk.TuiPressedModule,
                    range.TuiRangeModule,
                    core$1.TuiTextfieldControllerModule,
                    core$1.TuiWrapperModule,
                ],
                declarations: [TuiInputRangeComponent, TuiTextfieldAppearanceDirective],
                exports: [TuiInputRangeComponent, range.TuiNewRangeDirective],
            })
        ], TuiInputRangeModule);
        return TuiInputRangeModule;
    }());

    exports.TuiInputRangeComponent = TuiInputRangeComponent;
    exports.TuiInputRangeModule = TuiInputRangeModule;
    exports.TuiTextfieldAppearanceDirective = TuiTextfieldAppearanceDirective;
    exports.tuiTextfieldAppearanceDirectiveFactory = tuiTextfieldAppearanceDirectiveFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
