// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-count.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/tokens'), require('rxjs'), require('@angular/common'), require('@taiga-ui/kit/directives'), require('@tinkoff/ng-polymorpheus'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-count', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/tokens', 'rxjs', '@angular/common', '@taiga-ui/kit/directives', '@tinkoff/ng-polymorpheus', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-count'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.tokens, global.rxjs, global.ng.common, global['taiga-ui'].kit.directives, global.ngPolymorpheus, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, tokens, rxjs, common, directives, ngPolymorpheus, angular2TextMask) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    // TODO: 3.0 remove in ivy compilation
    var PASSWORD_ICON_UP = "tuiIconPlus";
    var PASSWORD_ICON_DOWN = "tuiIconMinus";
    /** Default values for the input count options. */
    var TUI_INPUT_COUNT_DEFAULT_OPTIONS = {
        icons: {
            up: PASSWORD_ICON_UP,
            down: PASSWORD_ICON_DOWN,
        },
        appearance: "textfield",
        hideButtons: false,
        min: 0,
        max: Infinity,
        step: 1,
        postfix: "",
    };
    var TUI_INPUT_COUNT_OPTIONS = new core.InjectionToken("Default parameters for input count component", {
        factory: function () { return TUI_INPUT_COUNT_DEFAULT_OPTIONS; },
    });
    var tuiInputCountOptionsProvider = function (options) { return ({
        provide: TUI_INPUT_COUNT_OPTIONS,
        useValue: __assign(__assign({}, TUI_INPUT_COUNT_DEFAULT_OPTIONS), options),
    }); };

    // @dynamic
    var TuiInputCountComponent = /** @class */ (function (_super) {
        __extends(TuiInputCountComponent, _super);
        function TuiInputCountComponent(control, changeDetectorRef, appearance, textfieldSize, minusTexts$, isMobile, options, numberFormat) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.appearance = appearance;
            _this.textfieldSize = textfieldSize;
            _this.minusTexts$ = minusTexts$;
            _this.isMobile = isMobile;
            _this.options = options;
            _this.numberFormat = numberFormat;
            _this.unfinishedValue = '';
            _this.step = _this.options.step;
            _this.min = _this.options.min;
            _this.max = _this.options.max;
            _this.hideButtons = _this.options.hideButtons;
            _this.prefix = '';
            _this.postfix = _this.options.postfix;
            return _this;
        }
        TuiInputCountComponent_1 = TuiInputCountComponent;
        TuiInputCountComponent.prototype.getMask = function (allowNegative) {
            return {
                mask: core$1.tuiCreateNumberMask({
                    allowNegative: allowNegative,
                    decimalSymbol: this.numberFormat.decimalSeparator,
                    thousandSymbol: this.numberFormat.thousandSeparator,
                }),
                guide: false,
            };
        };
        Object.defineProperty(TuiInputCountComponent.prototype, "buttonsHidden", {
            // TODO: 3.0 Remove in v.3
            get: function () {
                return this.hideButtons || this.appearance === 'table';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "iconUp", {
            get: function () {
                return this.options.icons.up;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "iconDown", {
            get: function () {
                return this.options.icons.down;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.primitiveTextfield || this.computedDisabled
                    ? null
                    : this.primitiveTextfield.nativeFocusableElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "size", {
            get: function () {
                return this.textfieldSize.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "exampleText", {
            get: function () {
                return String(this.min);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "computedValue", {
            get: function () {
                return this.focused ? this.nativeValue : this.formatNumber(this.value);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "minusButtonDisabled", {
            get: function () {
                return !this.interactive || (cdk.isPresent(this.value) && this.value <= this.min);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "plusButtonDisabled", {
            get: function () {
                return !this.interactive || (cdk.isPresent(this.value) && this.value >= this.max);
            },
            enumerable: true,
            configurable: true
        });
        TuiInputCountComponent.prototype.onButtonMouseDown = function (event, disabled) {
            if (disabled === void 0) { disabled = false; }
            if (disabled || !this.nativeFocusableElement || this.isMobile) {
                return;
            }
            event.preventDefault();
            cdk.setNativeFocused(this.nativeFocusableElement);
        };
        TuiInputCountComponent.prototype.onFocused = function (focused) {
            if (!focused) {
                this.onBlur();
            }
            this.updateFocused(focused);
        };
        TuiInputCountComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputCountComponent.prototype.onPressed = function (pressed) {
            this.updatePressed(pressed);
        };
        TuiInputCountComponent.prototype.onValueChange = function () {
            if (this.isNativeValueNotFinished) {
                this.unfinishedValue = this.nativeValue;
                return;
            }
            this.unfinishedValue = null;
            var capped = this.capValue(this.nativeNumberValue);
            if (this.isNotNumber(capped)) {
                this.updateValue(null);
                return;
            }
            var newValue = this.formatNumber(capped);
            if (this.nativeValue !== newValue) {
                this.nativeValue = newValue;
            }
            this.updateValue(capped);
        };
        TuiInputCountComponent.prototype.decreaseValue = function () {
            if (this.readOnly) {
                return;
            }
            var newValue = (this.value || 0) - this.step;
            this.safeUpdateValue(newValue);
        };
        TuiInputCountComponent.prototype.increaseValue = function () {
            if (this.readOnly) {
                return;
            }
            var newValue = (this.value || 0) + this.step;
            this.safeUpdateValue(newValue);
        };
        TuiInputCountComponent.prototype.onKeydown = function (event) {
            switch (event.key) {
                case 'ArrowUp':
                case 'Up':
                    this.increaseValue();
                    event.preventDefault();
                    break;
                case 'ArrowDown':
                case 'Down':
                    this.decreaseValue();
                    event.preventDefault();
                    break;
                default:
                    break;
            }
        };
        Object.defineProperty(TuiInputCountComponent.prototype, "nativeNumberValue", {
            get: function () {
                return parseInt(this.nativeValue.split(this.numberFormat.thousandSeparator).join(''), 10);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputCountComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputCountComponent.prototype.safeUpdateValue = function (newValue) {
            var value = cdk.clamp(newValue, this.min, this.max);
            this.updateValue(value);
            this.nativeValue = this.formatNumber(value);
        };
        TuiInputCountComponent.prototype.capValue = function (value) {
            var capped = Math.min(value, this.max);
            return Number.isNaN(capped) || capped < this.min ? null : capped;
        };
        TuiInputCountComponent.prototype.onBlur = function () {
            var nativeNumberValue = this.unfinishedValue
                ? core$1.tuiMaskedNumberStringToNumber(this.unfinishedValue, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator)
                : this.nativeNumberValue;
            this.unfinishedValue = null;
            if (this.isNotNumber(nativeNumberValue)) {
                this.updateValue(null);
                return;
            }
            var value = Math.max(nativeNumberValue || 0, this.min);
            var formattedValue = this.formatNumber(value);
            this.nativeValue = formattedValue;
            this.updateValue(value);
            if (this.primitiveTextfield) {
                this.primitiveTextfield.value = formattedValue;
            }
        };
        TuiInputCountComponent.prototype.formatNumber = function (value) {
            return this.isNotNumber(value)
                ? ''
                : core$1.formatNumber(value, null, this.numberFormat.decimalSeparator, this.numberFormat.thousandSeparator);
        };
        TuiInputCountComponent.prototype.isNotNumber = function (value) {
            return !cdk.isPresent(value) || Number.isNaN(value);
        };
        Object.defineProperty(TuiInputCountComponent.prototype, "isNativeValueNotFinished", {
            get: function () {
                var nativeNumberValue = this.nativeNumberValue;
                return nativeNumberValue < 0
                    ? nativeNumberValue > this.max
                    : nativeNumberValue < this.min;
            },
            enumerable: true,
            configurable: true
        });
        var TuiInputCountComponent_1;
        TuiInputCountComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_APPEARANCE,] }] },
            { type: core$1.TuiTextfieldSizeDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_SIZE,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_PLUS_MINUS_TEXTS,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [TUI_INPUT_COUNT_OPTIONS,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [core$1.TUI_NUMBER_FORMAT,] }] }
        ]; };
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputCountComponent.prototype, "primitiveTextfield", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "step", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "hideButtons", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "prefix", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputCountComponent.prototype, "postfix", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiInputCountComponent.prototype, "getMask", null);
        __decorate([
            core.HostBinding('class._hide-buttons')
        ], TuiInputCountComponent.prototype, "buttonsHidden", null);
        __decorate([
            core.HostBinding('attr.data-size')
        ], TuiInputCountComponent.prototype, "size", null);
        TuiInputCountComponent = TuiInputCountComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-count',
                template: "<tui-primitive-textfield\n    tuiValueAccessor\n    tuiTextfieldAutocomplete=\"off\"\n    tuiTextfieldInputMode=\"numeric\"\n    class=\"t-textfield\"\n    [tuiTextfieldMaxLength]=\"18\"\n    [pseudoFocused]=\"pseudoFocused\"\n    [pseudoHovered]=\"pseudoHovered\"\n    [pseudoPressed]=\"pseudoPressed\"\n    [focusable]=\"focusable\"\n    [nativeId]=\"nativeId\"\n    [prefix]=\"prefix\"\n    [postfix]=\"postfix\"\n    [disabled]=\"computedDisabled\"\n    [readOnly]=\"readOnly\"\n    [textMask]=\"getMask(min < 0)\"\n    [invalid]=\"computedInvalid\"\n    [value]=\"computedValue\"\n    (valueChange)=\"onValueChange()\"\n    (keydown)=\"onKeydown($event)\"\n    (focusedChange)=\"onFocused($event)\"\n    (hoveredChange)=\"onHovered($event)\"\n    (pressedChange)=\"onPressed($event)\"\n>\n    <ng-content></ng-content>\n    <ng-content\n        select=\"input\"\n        ngProjectAs=\"input\"\n    ></ng-content>\n</tui-primitive-textfield>\n<ng-container *ngIf=\"!buttonsHidden\">\n    <ng-container *ngIf=\"minusTexts$ | async as texts\">\n        <section class=\"t-buttons\">\n            <button\n                tuiIconButton\n                type=\"button\"\n                automation-id=\"tui-input-count__plus-button\"\n                size=\"s\"\n                class=\"t-button t-button_plus\"\n                [icon]=\"iconUp\"\n                [appearance]=\"options.appearance\"\n                [title]=\"texts[0]\"\n                [disabled]=\"plusButtonDisabled\"\n                [focusable]=\"false\"\n                (mousedown)=\"onButtonMouseDown($event, plusButtonDisabled)\"\n                (click)=\"increaseValue()\"\n            ></button>\n            <button\n                tuiIconButton\n                type=\"button\"\n                automation-id=\"tui-input-count__minus-button\"\n                size=\"s\"\n                class=\"t-button t-button_minus\"\n                [icon]=\"iconDown\"\n                [appearance]=\"options.appearance\"\n                [disabled]=\"minusButtonDisabled\"\n                [focusable]=\"false\"\n                [title]=\"texts[1]\"\n                (mousedown)=\"onButtonMouseDown($event, minusButtonDisabled)\"\n                (click)=\"decreaseValue()\"\n            ></button>\n        </section>\n    </ng-container>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputCountComponent_1; }),
                    },
                    {
                        provide: cdk.AbstractTuiControl,
                        useExisting: core.forwardRef(function () { return TuiInputCountComponent_1; }),
                    },
                ],
                styles: [":host{display:flex;font:var(--tui-font-text-s);border-radius:var(--tui-radius-m);height:var(--tui-height-m);color:var(--tui-text-01);text-align:left}:host._disabled{pointer-events:none}:host[data-size='l']{font:var(--tui-font-text-m);height:var(--tui-height-l)}.t-textfield{position:relative;border-radius:inherit;text-align:inherit;flex-grow:1;min-width:0}:host:not(._hide-buttons) .t-textfield{border-top-right-radius:0;border-bottom-right-radius:0}.t-buttons{display:flex;flex-direction:column;margin-left:.125rem;height:100%}.t-buttons .t-button{display:flex;width:calc(var(--tui-height-m) * .75);height:calc(50% - .0625rem)}.t-buttons .t-button_plus{margin-bottom:.125rem;border-radius:0 var(--tui-radius-m) 0 0}.t-buttons .t-button_minus{border-radius:0 0 var(--tui-radius-m)}:host[data-size='l'] .t-buttons .t-button{width:calc(var(--tui-height-l) * .75)}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core$1.TUI_TEXTFIELD_APPEARANCE)),
            __param(3, core.Inject(core$1.TUI_TEXTFIELD_SIZE)),
            __param(4, core.Inject(tokens.TUI_PLUS_MINUS_TEXTS)),
            __param(5, core.Inject(cdk.TUI_IS_MOBILE)),
            __param(6, core.Inject(TUI_INPUT_COUNT_OPTIONS)),
            __param(7, core.Inject(core$1.TUI_NUMBER_FORMAT))
        ], TuiInputCountComponent);
        return TuiInputCountComponent;
    }(cdk.AbstractTuiNullableControl));

    var TuiInputCountDirective = /** @class */ (function (_super) {
        __extends(TuiInputCountDirective, _super);
        function TuiInputCountDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TuiInputCountDirective_1 = TuiInputCountDirective;
        TuiInputCountDirective.prototype.onValueChange = function () {
            this.host.onValueChange();
        };
        TuiInputCountDirective.prototype.process = function (input) {
            input.autocomplete = 'off';
            input.inputMode = 'numeric';
        };
        var TuiInputCountDirective_1;
        TuiInputCountDirective = TuiInputCountDirective_1 = __decorate([
            core.Directive({
                selector: 'tui-input-count',
                providers: [
                    {
                        provide: core$1.TUI_TEXTFIELD_HOST,
                        useExisting: core.forwardRef(function () { return TuiInputCountDirective_1; }),
                    },
                ],
            })
        ], TuiInputCountDirective);
        return TuiInputCountDirective;
    }(core$1.TuiAbstractTextfieldHost));

    var TuiInputCountModule = /** @class */ (function () {
        function TuiInputCountModule() {
        }
        TuiInputCountModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    angular2TextMask.TextMaskModule,
                    core$1.TuiButtonModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    core$1.TuiTextfieldControllerModule,
                    directives.TuiValueAccessorModule,
                    ngPolymorpheus.PolymorpheusModule,
                ],
                declarations: [TuiInputCountComponent, TuiInputCountDirective],
                exports: [TuiInputCountComponent, TuiInputCountDirective, core$1.TuiTextfieldComponent],
            })
        ], TuiInputCountModule);
        return TuiInputCountModule;
    }());

    exports.PASSWORD_ICON_DOWN = PASSWORD_ICON_DOWN;
    exports.PASSWORD_ICON_UP = PASSWORD_ICON_UP;
    exports.TUI_INPUT_COUNT_DEFAULT_OPTIONS = TUI_INPUT_COUNT_DEFAULT_OPTIONS;
    exports.TUI_INPUT_COUNT_OPTIONS = TUI_INPUT_COUNT_OPTIONS;
    exports.TuiInputCountComponent = TuiInputCountComponent;
    exports.TuiInputCountDirective = TuiInputCountDirective;
    exports.TuiInputCountModule = TuiInputCountModule;
    exports.tuiInputCountOptionsProvider = tuiInputCountOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
