// @taiga-ui/kit/bundles/taiga-ui-kit-constants.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@taiga-ui/i18n'), require('@taiga-ui/cdk'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/constants', ['exports', '@taiga-ui/i18n', '@taiga-ui/cdk', '@taiga-ui/core'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.constants = {}), global.i18n, global.cdk, global.core));
}(this, (function (exports, i18n, cdk, core) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var _a;
    /**
     * @deprecated Dont use it (will be deleted soon).
     * - Use token `TUI_COUNTRIES_MASKS` to get phone mask by country iso code
     * - Use token `TUI_COUNTRIES` to get localized country name by its iso code
     * @example
     * import {TUI_COUNTRIES, TUI_COUNTRIES_MASKS} from '@taiga-ui/kit';
     *
     * @todo TODO: remove it in 3.0
     */
    var COUNTRIES = (_a = {},
        _a[i18n.TuiCountryIsoCode.AD] = { mask: "+376###-###", name: "\u0410\u043D\u0434\u043E\u0440\u0440\u0430" },
        _a[i18n.TuiCountryIsoCode.AE] = {
            mask: "+971-##-###-####",
            name: "\u041E\u0431\u044A\u0435\u0434\u0438\u043D\u0435\u043D\u043D\u044B\u0435 \u0410\u0440\u0430\u0431\u0441\u043A\u0438\u0435 \u042D\u043C\u0438\u0440\u0430\u0442\u044B",
        },
        _a[i18n.TuiCountryIsoCode.AF] = { mask: "+93##-###-####", name: "\u0410\u0444\u0433\u0430\u043D\u0438\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.AG] = { mask: "+1(268) ###-####", name: "\u0410\u043D\u0442\u0438\u0433\u0443\u0430 \u0438 \u0411\u0430\u0440\u0431\u0443\u0434\u0430" },
        _a[i18n.TuiCountryIsoCode.AI] = { mask: "+1(264) ###-####", name: "\u0410\u043D\u0433\u0438\u043B\u044C\u044F" },
        _a[i18n.TuiCountryIsoCode.AL] = { mask: "+355(###) ###-###", name: "\u0410\u043B\u0431\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.AM] = { mask: "+374##-###-###", name: "\u0410\u0440\u043C\u0435\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.AO] = { mask: "+244(###) ###-###", name: "\u0410\u043D\u0433\u043E\u043B\u0430" },
        _a[i18n.TuiCountryIsoCode.AR] = { mask: "+54(###) ###-####", name: "\u0410\u0440\u0433\u0435\u043D\u0442\u0438\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.AT] = { mask: "+43(###) ###-####", name: "\u0410\u0432\u0441\u0442\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.AU] = { mask: "+61#-####-####", name: "\u0410\u0432\u0441\u0442\u0440\u0430\u043B\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.AW] = { mask: "+297###-####", name: "\u0410\u0440\u0443\u0431\u0430" },
        _a[i18n.TuiCountryIsoCode.AZ] = { mask: "+994##-###-##-##", name: "\u0410\u0437\u0435\u0440\u0431\u0430\u0439\u0434\u0436\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.BA] = { mask: "+387 ###-####-#", name: "\u0411\u043E\u0441\u043D\u0438\u044F \u0438 \u0413\u0435\u0440\u0446\u0435\u0433\u043E\u0432\u0438\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.BB] = { mask: "+1(246) ###-####", name: "\u0411\u0430\u0440\u0431\u0430\u0434\u043E\u0441" },
        _a[i18n.TuiCountryIsoCode.BD] = { mask: "+880##-###-###", name: "\u0411\u0430\u043D\u0433\u043B\u0430\u0434\u0435\u0448" },
        _a[i18n.TuiCountryIsoCode.BE] = { mask: "+32(###) ###-###", name: "\u0411\u0435\u043B\u044C\u0433\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.BF] = { mask: "+226##-##-####", name: "\u0411\u0443\u0440\u043A\u0438\u043D\u0430 \u0424\u0430\u0441\u043E" },
        _a[i18n.TuiCountryIsoCode.BG] = { mask: "+359(###) ###-###", name: "\u0411\u043E\u043B\u0433\u0430\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.BH] = { mask: "+973####-####", name: "\u0411\u0430\u0445\u0440\u0435\u0439\u043D" },
        _a[i18n.TuiCountryIsoCode.BI] = { mask: "+257##-##-####", name: "\u0411\u0443\u0440\u0443\u043D\u0434\u0438" },
        _a[i18n.TuiCountryIsoCode.BJ] = { mask: "+229##-##-####", name: "\u0411\u0435\u043D\u0438\u043D" },
        _a[i18n.TuiCountryIsoCode.BL] = { mask: "+590 ## ## ## ## ##", name: "\u0421\u0435\u043D-\u0411\u0430\u0440\u0442\u0435\u043B\u044C\u043C\u0438" },
        _a[i18n.TuiCountryIsoCode.BM] = { mask: "+1(441) ###-####", name: "\u0411\u0435\u0440\u043C\u0443\u0434\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.BN] = { mask: "+673###-####", name: "\u0411\u0440\u0443\u043D\u0435\u0439-\u0414\u0430\u0440\u0443\u0441\u0441\u0430\u043B\u0430\u043C" },
        _a[i18n.TuiCountryIsoCode.BO] = { mask: "+591#-###-####", name: "\u0411\u043E\u043B\u0438\u0432\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.BQ] = {
            mask: "+599-###-####",
            name: "\u0411\u043E\u043D\u0430\u0439\u0440\u0435, \u0421\u0438\u043D\u0442-\u042D\u0441\u0442\u0430\u0442\u0438\u0443\u0441 \u0438 \u0421\u0430\u0431\u0430",
        },
        _a[i18n.TuiCountryIsoCode.BR] = { mask: "+55(##) ####-####", name: "\u0411\u0440\u0430\u0437\u0438\u043B\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.BS] = { mask: "+1(242) ###-####", name: "\u0411\u0430\u0433\u0430\u043C\u0441\u043A\u0438\u0435 \u041E\u0441\u0442\u0440\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.BT] = { mask: "+975#-###-###", name: "\u0411\u0443\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.BW] = { mask: "+267##-###-###", name: "\u0411\u043E\u0442\u0441\u0432\u0430\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.BY] = { mask: "+375(##) ###-##-##", name: "\u0411\u0435\u043B\u0430\u0440\u0443\u0441\u044C (\u0411\u0435\u043B\u043E\u0440\u0443\u0441\u0441\u0438\u044F)" },
        _a[i18n.TuiCountryIsoCode.BZ] = { mask: "+501###-####", name: "\u0411\u0435\u043B\u0438\u0437" },
        _a[i18n.TuiCountryIsoCode.CA] = { mask: "+1(###) ###-####", name: "\u041A\u0430\u043D\u0430\u0434\u0430" },
        _a[i18n.TuiCountryIsoCode.CD] = {
            mask: "+243(###) ###-###",
            name: "\u0414\u0435\u043C. \u0420\u0435\u0441\u043F. \u041A\u043E\u043D\u0433\u043E (\u041A\u0438\u043D\u0448\u0430\u0441\u0430)",
        },
        _a[i18n.TuiCountryIsoCode.CF] = {
            mask: "+236-##-##-####",
            name: "\u0426\u0435\u043D\u0442\u0440\u0430\u043B\u044C\u043D\u043E\u0430\u0444\u0440\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430",
        },
        _a[i18n.TuiCountryIsoCode.CG] = { mask: "+242##-###-####", name: "\u041A\u043E\u043D\u0433\u043E (\u0411\u0440\u0430\u0437\u0437\u0430\u0432\u0438\u043B\u044C)" },
        _a[i18n.TuiCountryIsoCode.CH] = { mask: "+41##-###-####", name: "\u0428\u0432\u0435\u0439\u0446\u0430\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.CI] = { mask: "+225##-###-###", name: "\u041A\u043E\u0442-\u0434\u2019\u0418\u0432\u0443\u0430\u0440" },
        _a[i18n.TuiCountryIsoCode.CL] = { mask: "+56#-####-####", name: "\u0427\u0438\u043B\u0438" },
        _a[i18n.TuiCountryIsoCode.CM] = { mask: "+237####-####", name: "\u041A\u0430\u043C\u0435\u0440\u0443\u043D" },
        _a[i18n.TuiCountryIsoCode.CN] = { mask: "+86(###) ####-####", name: "\u041A\u0438\u0442\u0430\u0439\u0441\u043A\u0430\u044F \u041D.\u0420." },
        _a[i18n.TuiCountryIsoCode.CO] = { mask: "+57(###) ###-####", name: "\u041A\u043E\u043B\u0443\u043C\u0431\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.CR] = { mask: "+506####-####", name: "\u041A\u043E\u0441\u0442\u0430-\u0420\u0438\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.CU] = { mask: "+53#-###-####", name: "\u041A\u0443\u0431\u0430" },
        _a[i18n.TuiCountryIsoCode.CV] = { mask: "+238(###) ##-##", name: "\u041A\u0430\u0431\u043E-\u0412\u0435\u0440\u0434\u0435" },
        _a[i18n.TuiCountryIsoCode.CW] = { mask: "+5999-###-####", name: "\u041A\u044E\u0440\u0430\u0441\u0430\u043E" },
        _a[i18n.TuiCountryIsoCode.CY] = { mask: "+357##-###-###", name: "\u041A\u0438\u043F\u0440" },
        _a[i18n.TuiCountryIsoCode.CZ] = { mask: "+420(###) ###-###", name: "\u0427\u0435\u0445\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.DE] = { mask: "+49(###) ###-###-##", name: "\u0413\u0435\u0440\u043C\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.DJ] = { mask: "+253##-##-##-##", name: "\u0414\u0436\u0438\u0431\u0443\u0442\u0438" },
        _a[i18n.TuiCountryIsoCode.DK] = { mask: "+45##-##-##-##", name: "\u0414\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.DM] = { mask: "+1(767) ###-####", name: "\u0414\u043E\u043C\u0438\u043D\u0438\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.DO] = { mask: "+1(###) ###-####", name: "\u0414\u043E\u043C\u0438\u043D\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.DZ] = { mask: "+213##-###-####", name: "\u0410\u043B\u0436\u0438\u0440" },
        _a[i18n.TuiCountryIsoCode.EC] = { mask: "+593##-###-####", name: "\u042D\u043A\u0432\u0430\u0434\u043E\u0440 " },
        _a[i18n.TuiCountryIsoCode.EE] = { mask: "+372####-####", name: "\u042D\u0441\u0442\u043E\u043D\u0438\u044F " },
        _a[i18n.TuiCountryIsoCode.EG] = { mask: "+20(###) ###-####", name: "\u0415\u0433\u0438\u043F\u0435\u0442" },
        _a[i18n.TuiCountryIsoCode.ER] = { mask: "+291#-###-###", name: "\u042D\u0440\u0438\u0442\u0440\u0435\u044F" },
        _a[i18n.TuiCountryIsoCode.ES] = { mask: "+34(###) ###-###", name: "\u0418\u0441\u043F\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.ET] = { mask: "+251##-###-####", name: "\u042D\u0444\u0438\u043E\u043F\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.FI] = { mask: "+358(###) ###-##-##", name: "\u0424\u0438\u043D\u043B\u044F\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.FJ] = { mask: "+679##-#####", name: "\u0424\u0438\u0434\u0436\u0438" },
        _a[i18n.TuiCountryIsoCode.FK] = { mask: "+500#####", name: "\u0424\u043E\u043B\u043A\u043B\u0435\u043D\u0434\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.FM] = { mask: "+691###-####", name: "\u0424.\u0428. \u041C\u0438\u043A\u0440\u043E\u043D\u0435\u0437\u0438\u0438" },
        _a[i18n.TuiCountryIsoCode.FR] = { mask: "+33 ## ## ## ## ##", name: "\u0424\u0440\u0430\u043D\u0446\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GA] = { mask: "+241#-##-##-##", name: "\u0413\u0430\u0431\u043E\u043D" },
        _a[i18n.TuiCountryIsoCode.GB] = { mask: "+44##-####-####", name: "\u0412\u0435\u043B\u0438\u043A\u043E\u0431\u0440\u0438\u0442\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GD] = { mask: "+1(473) ###-####", name: "\u0413\u0440\u0435\u043D\u0430\u0434\u0430" },
        _a[i18n.TuiCountryIsoCode.GE] = { mask: "+995(###) ###-###", name: "\u0413\u0440\u0443\u0437\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GF] = { mask: "+594 ## ## ## ## ##", name: "\u0413\u0430\u0439\u0430\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.GH] = { mask: "+233(###) ###-###", name: "\u0413\u0430\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.GI] = { mask: "+350###-#####", name: "\u0413\u0438\u0431\u0440\u0430\u043B\u0442\u0430\u0440" },
        _a[i18n.TuiCountryIsoCode.GL] = { mask: "+299##-##-##", name: "\u0413\u0440\u0435\u043D\u043B\u0430\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GM] = { mask: "+220(###) ##-##", name: "\u0413\u0430\u043C\u0431\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GN] = { mask: "+224##-###-###", name: "\u0413\u0432\u0438\u043D\u0435\u044F" },
        _a[i18n.TuiCountryIsoCode.GP] = { mask: "+590 ## ## ## ## ##", name: "\u0413\u0432\u0430\u0434\u0435\u043B\u0443\u043F\u0430" },
        _a[i18n.TuiCountryIsoCode.GQ] = { mask: "+240##-###-####", name: "\u042D\u043A\u0432\u0430\u0442\u043E\u0440\u0438\u0430\u043B\u044C\u043D\u0430\u044F \u0413\u0432\u0438\u043D\u0435\u044F" },
        _a[i18n.TuiCountryIsoCode.GR] = { mask: "+30(###) ###-####", name: "\u0413\u0440\u0435\u0446\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.GT] = { mask: "+502#-###-####", name: "\u0413\u0432\u0430\u0442\u0435\u043C\u0430\u043B\u0430" },
        _a[i18n.TuiCountryIsoCode.GW] = { mask: "+245#-######", name: "\u0413\u0432\u0438\u043D\u0435\u044F-\u0411\u0438\u0441\u0430\u0443" },
        _a[i18n.TuiCountryIsoCode.GY] = { mask: "+592###-####", name: "\u0413\u0430\u0439\u0430\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.HK] = { mask: "+852####-####", name: "\u0413\u043E\u043D\u043A\u043E\u043D\u0433" },
        _a[i18n.TuiCountryIsoCode.HN] = { mask: "+504####-####", name: "\u0413\u043E\u043D\u0434\u0443\u0440\u0430\u0441" },
        _a[i18n.TuiCountryIsoCode.HR] = { mask: "+385##-###-###", name: "\u0425\u043E\u0440\u0432\u0430\u0442\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.HT] = { mask: "+509##-##-####", name: "\u0413\u0430\u0438\u0442\u0438" },
        _a[i18n.TuiCountryIsoCode.HU] = { mask: "+36(###) ###-###", name: "\u0412\u0435\u043D\u0433\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.ID] = { mask: "+62(###) ###-##-###", name: "\u0418\u043D\u0434\u043E\u043D\u0435\u0437\u0438\u044F " },
        _a[i18n.TuiCountryIsoCode.IE] = { mask: "+353(###) ###-###", name: "\u0418\u0440\u043B\u0430\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.IL] = { mask: "+972##-###-####", name: "\u0418\u0437\u0440\u0430\u0438\u043B\u044C" },
        _a[i18n.TuiCountryIsoCode.IN] = { mask: "+91(####) ###-###", name: "\u0418\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.IQ] = { mask: "+964(###) ###-####", name: "\u0418\u0440\u0430\u043A" },
        _a[i18n.TuiCountryIsoCode.IR] = { mask: "+98(###) ###-####", name: "\u0418\u0440\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.IS] = { mask: "+354###-####", name: "\u0418\u0441\u043B\u0430\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.IT] = { mask: "+39(###) ####-###", name: "\u0418\u0442\u0430\u043B\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.JM] = { mask: "+1(876) ###-####", name: "\u042F\u043C\u0430\u0439\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.JO] = { mask: "+962#-####-####", name: "\u0418\u043E\u0440\u0434\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.JP] = { mask: "+81-##-####-####", name: "\u042F\u043F\u043E\u043D\u0438\u044F " },
        _a[i18n.TuiCountryIsoCode.KE] = { mask: "+254###-######", name: "\u041A\u0435\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.KG] = { mask: "+996(###) ###-###", name: "\u041A\u0438\u0440\u0433\u0438\u0437\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.KH] = { mask: "+855##-###-###", name: "\u041A\u0430\u043C\u0431\u043E\u0434\u0436\u0430" },
        _a[i18n.TuiCountryIsoCode.KM] = { mask: "+269##-#####", name: "\u041A\u043E\u043C\u043E\u0440\u044B" },
        _a[i18n.TuiCountryIsoCode.KN] = { mask: "+1(869) ###-####", name: "\u0421\u0435\u043D\u0442-\u041A\u0438\u0442\u0441 \u0438 \u041D\u0435\u0432\u0438\u0441" },
        _a[i18n.TuiCountryIsoCode.KP] = { mask: "+850####-#############", name: "\u041A\u043E\u0440\u0435\u0439\u0441\u043A\u0430\u044F \u041D\u0414\u0420" },
        _a[i18n.TuiCountryIsoCode.KR] = { mask: "+82##-###-####", name: "\u0420\u0435\u0441\u043F. \u041A\u043E\u0440\u0435\u044F" },
        _a[i18n.TuiCountryIsoCode.KW] = { mask: "+965####-####", name: "\u041A\u0443\u0432\u0435\u0439\u0442" },
        _a[i18n.TuiCountryIsoCode.KY] = { mask: "+1(345) ###-####", name: "\u041A\u0430\u0439\u043C\u0430\u043D\u043E\u0432\u044B \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.KZ] = { mask: "+7(###) ###-##-##", name: "\u041A\u0430\u0437\u0430\u0445\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.LA] = { mask: "+856##-##-###-###", name: "\u041B\u0430\u043E\u0441" },
        _a[i18n.TuiCountryIsoCode.LB] = { mask: "+961##-###-###", name: "\u041B\u0438\u0432\u0430\u043D " },
        _a[i18n.TuiCountryIsoCode.LC] = { mask: "+1(758) ###-####", name: "\u0421\u0435\u043D\u0442-\u041B\u044E\u0441\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.LI] = { mask: "+423(###) ###-####", name: "\u041B\u0438\u0445\u0442\u0435\u043D\u0448\u0442\u0435\u0439\u043D" },
        _a[i18n.TuiCountryIsoCode.LK] = { mask: "+94##-###-####", name: "\u0428\u0440\u0438-\u041B\u0430\u043D\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.LR] = { mask: "+231##-###-###", name: "\u041B\u0438\u0431\u0435\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.LS] = { mask: "+266#-###-####", name: "\u041B\u0435\u0441\u043E\u0442\u043E" },
        _a[i18n.TuiCountryIsoCode.LT] = { mask: "+370(###) ##-###", name: "\u041B\u0438\u0442\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.LU] = { mask: "+352(###) ###-###", name: "\u041B\u044E\u043A\u0441\u0435\u043C\u0431\u0443\u0440\u0433" },
        _a[i18n.TuiCountryIsoCode.LV] = { mask: "+371##-###-###", name: "\u041B\u0430\u0442\u0432\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.LY] = { mask: "+218##-###-####", name: "\u041B\u0438\u0432\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.MA] = { mask: "+212##-####-###", name: "\u041C\u0430\u0440\u043E\u043A\u043A\u043E" },
        _a[i18n.TuiCountryIsoCode.MC] = { mask: "+377###-###-###", name: "\u041C\u043E\u043D\u0430\u043A\u043E" },
        _a[i18n.TuiCountryIsoCode.MD] = { mask: "+373####-####", name: "\u041C\u043E\u043B\u0434\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.ME] = { mask: "+382##-###-###", name: "\u0427\u0435\u0440\u043D\u043E\u0433\u043E\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.MF] = { mask: "+590 ## ## ## ## ##", name: "\u0421\u0435\u043D-\u041C\u0430\u0440\u0442\u0435\u043D" },
        _a[i18n.TuiCountryIsoCode.MG] = { mask: "+261##-##-#####", name: "\u041C\u0430\u0434\u0430\u0433\u0430\u0441\u043A\u0430\u0440" },
        _a[i18n.TuiCountryIsoCode.MK] = { mask: "+389##-###-###", name: "\u0420\u0435\u0441\u043F. \u041C\u0430\u043A\u0435\u0434\u043E\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.ML] = { mask: "+223##-##-####", name: "\u041C\u0430\u043B\u0438" },
        _a[i18n.TuiCountryIsoCode.MM] = { mask: "+95##-###-###", name: "\u0411\u0438\u0440\u043C\u0430 (\u041C\u044C\u044F\u043D\u043C\u0430)" },
        _a[i18n.TuiCountryIsoCode.MN] = { mask: "+976##-##-####", name: "\u041C\u043E\u043D\u0433\u043E\u043B\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.MO] = { mask: "+853####-####", name: "\u041C\u0430\u043A\u0430\u043E" },
        _a[i18n.TuiCountryIsoCode.MQ] = { mask: "+596 ## ## ## ## ##", name: "\u041C\u0430\u0440\u0442\u0438\u043D\u0438\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.MR] = { mask: "+222##-##-####", name: "\u041C\u0430\u0432\u0440\u0438\u0442\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.MS] = { mask: "+1(664) ###-####", name: "\u041C\u043E\u043D\u0442\u0441\u0435\u0440\u0440\u0430\u0442" },
        _a[i18n.TuiCountryIsoCode.MT] = { mask: "+356####-####", name: "\u041C\u0430\u043B\u044C\u0442\u0430" },
        _a[i18n.TuiCountryIsoCode.MU] = { mask: "+230###-####", name: "\u041C\u0430\u0432\u0440\u0438\u043A\u0438\u0439" },
        _a[i18n.TuiCountryIsoCode.MV] = { mask: "+960###-####", name: "\u041C\u0430\u043B\u044C\u0434\u0438\u0432\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
        _a[i18n.TuiCountryIsoCode.MW] = { mask: "+265#-####-####", name: "\u041C\u0430\u043B\u0430\u0432\u0438" },
        _a[i18n.TuiCountryIsoCode.MX] = { mask: "+52(###) ###-####", name: "\u041C\u0435\u043A\u0441\u0438\u043A\u0430" },
        _a[i18n.TuiCountryIsoCode.MY] = { mask: "+60(###) ###-###", name: "\u041C\u0430\u043B\u0430\u0439\u0437\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.MZ] = { mask: "+258##-###-###", name: "\u041C\u043E\u0437\u0430\u043C\u0431\u0438\u043A" },
        _a[i18n.TuiCountryIsoCode.NA] = { mask: "+264##-###-####", name: "\u041D\u0430\u043C\u0438\u0431\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.NC] = { mask: "+687 ### ###", name: "\u041A\u0430\u043B\u0435\u0434\u043E\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.NE] = { mask: "+227##-##-####", name: "\u041D\u0438\u0433\u0435\u0440" },
        _a[i18n.TuiCountryIsoCode.NG] = { mask: "+234(###) ###-####", name: "\u041D\u0438\u0433\u0435\u0440\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.NI] = { mask: "+505####-####", name: "\u041D\u0438\u043A\u0430\u0440\u0430\u0433\u0443\u0430" },
        _a[i18n.TuiCountryIsoCode.NL] = { mask: "+31##-###-####", name: "\u041D\u0438\u0434\u0435\u0440\u043B\u0430\u043D\u0434\u044B" },
        _a[i18n.TuiCountryIsoCode.NO] = { mask: "+47(###) ##-###", name: "\u041D\u043E\u0440\u0432\u0435\u0433\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.NP] = { mask: "+977##-###-###", name: "\u041D\u0435\u043F\u0430\u043B" },
        _a[i18n.TuiCountryIsoCode.NZ] = { mask: "+64(###) ###-####", name: "\u041D\u043E\u0432\u0430\u044F \u0417\u0435\u043B\u0430\u043D\u0434\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.OM] = { mask: "+968##-###-###", name: "\u041E\u043C\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.PA] = { mask: "+507###-####", name: "\u041F\u0430\u043D\u0430\u043C\u0430" },
        _a[i18n.TuiCountryIsoCode.PE] = { mask: "+51(###) ###-###", name: "\u041F\u0435\u0440\u0443" },
        _a[i18n.TuiCountryIsoCode.PF] = {
            mask: "+689##-##-##",
            name: "\u0424\u0440\u0430\u043D\u0446\u0443\u0437\u0441\u043A\u0430\u044F \u041F\u043E\u043B\u0438\u043D\u0435\u0437\u0438\u044F (\u0422\u0430\u0438\u0442\u0438)",
        },
        _a[i18n.TuiCountryIsoCode.PG] = { mask: "+675(###) ##-###", name: "\u041F\u0430\u043F\u0443\u0430-\u041D\u043E\u0432\u0430\u044F \u0413\u0432\u0438\u043D\u0435\u044F" },
        _a[i18n.TuiCountryIsoCode.PH] = { mask: "+63(###) ###-####", name: "\u0424\u0438\u043B\u0438\u043F\u043F\u0438\u043D\u044B" },
        _a[i18n.TuiCountryIsoCode.PK] = { mask: "+92(###) ###-####", name: "\u041F\u0430\u043A\u0438\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.PL] = { mask: "+48(###) ###-###", name: "\u041F\u043E\u043B\u044C\u0448\u0430" },
        _a[i18n.TuiCountryIsoCode.PT] = { mask: "+351##-###-####", name: "\u041F\u043E\u0440\u0442\u0443\u0433\u0430\u043B\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.PW] = { mask: "+680###-####", name: "\u041F\u0430\u043B\u0430\u0443" },
        _a[i18n.TuiCountryIsoCode.PY] = { mask: "+595(###) ###-###", name: "\u041F\u0430\u0440\u0430\u0433\u0432\u0430\u0439" },
        _a[i18n.TuiCountryIsoCode.QA] = { mask: "+974####-####", name: "\u041A\u0430\u0442\u0430\u0440" },
        _a[i18n.TuiCountryIsoCode.RE] = { mask: "+262 ## ## ## ## ##", name: "\u0420\u0435\u044E\u043D\u044C\u043E\u043D" },
        _a[i18n.TuiCountryIsoCode.RO] = { mask: "+40##-###-####", name: "\u0420\u0443\u043C\u044B\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.RS] = { mask: "+381##-###-####", name: "\u0421\u0435\u0440\u0431\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.RU] = { mask: "+7### ###-##-##", name: "\u0420\u043E\u0441\u0441\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.RW] = { mask: "+250(###) ###-###", name: "\u0420\u0443\u0430\u043D\u0434\u0430" },
        _a[i18n.TuiCountryIsoCode.SA] = { mask: "+966#-####-####", name: "\u0421\u0430\u0443\u0434\u043E\u0432\u0441\u043A\u0430\u044F \u0410\u0440\u0430\u0432\u0438\u044F " },
        _a[i18n.TuiCountryIsoCode.SB] = { mask: "+677###-####", name: "\u0421\u043E\u043B\u043E\u043C\u043E\u043D\u043E\u0432\u044B \u041E\u0441\u0442\u0440\u043E\u0432\u0430 " },
        _a[i18n.TuiCountryIsoCode.SC] = { mask: "+248#-###-###", name: "\u0421\u0435\u0439\u0448\u0435\u043B\u044B" },
        _a[i18n.TuiCountryIsoCode.SD] = { mask: "+249##-###-####", name: "\u0421\u0443\u0434\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.SE] = { mask: "+46##-###-####", name: "\u0428\u0432\u0435\u0446\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.SG] = { mask: "+65####-####", name: "\u0421\u0438\u043D\u0433\u0430\u043F\u0443\u0440" },
        _a[i18n.TuiCountryIsoCode.SH] = { mask: "+290####", name: "\u041E\u0441\u0442\u0440\u043E\u0432 \u0421\u0432\u044F\u0442\u043E\u0439 \u0415\u043B\u0435\u043D\u044B" },
        _a[i18n.TuiCountryIsoCode.SI] = { mask: "+386##-###-###", name: "\u0421\u043B\u043E\u0432\u0435\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.SK] = { mask: "+421(###) ###-###", name: "\u0421\u043B\u043E\u0432\u0430\u043A\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.SL] = { mask: "+232##-######", name: "\u0421\u044C\u0435\u0440\u0440\u0430-\u041B\u0435\u043E\u043D\u0435" },
        _a[i18n.TuiCountryIsoCode.SM] = { mask: "+378####-######", name: "\u0421\u0430\u043D-\u041C\u0430\u0440\u0438\u043D\u043E" },
        _a[i18n.TuiCountryIsoCode.SN] = { mask: "+221##-###-####", name: "\u0421\u0435\u043D\u0435\u0433\u0430\u043B" },
        _a[i18n.TuiCountryIsoCode.SO] = { mask: "+252##-###-###", name: "\u0421\u043E\u043C\u0430\u043B\u0438" },
        _a[i18n.TuiCountryIsoCode.SR] = { mask: "+597###-####", name: "\u0421\u0443\u0440\u0438\u043D\u0430\u043C " },
        _a[i18n.TuiCountryIsoCode.ST] = { mask: "+239##-#####", name: "\u0421\u0430\u043D-\u0422\u043E\u043C\u0435 \u0438 \u041F\u0440\u0438\u043D\u0441\u0438\u043F\u0438" },
        _a[i18n.TuiCountryIsoCode.SV] = { mask: "+503##-##-####", name: "\u0421\u0430\u043B\u044C\u0432\u0430\u0434\u043E\u0440" },
        _a[i18n.TuiCountryIsoCode.SX] = { mask: "+1(721) ###-####", name: "\u0421\u0438\u043D\u0442-\u041C\u0430\u0440\u0442\u0435\u043D" },
        _a[i18n.TuiCountryIsoCode.SY] = {
            mask: "+963##-####-###",
            name: "\u0421\u0438\u0440\u0438\u0439\u0441\u043A\u0430\u044F \u0430\u0440\u0430\u0431\u0441\u043A\u0430\u044F \u0440\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430",
        },
        _a[i18n.TuiCountryIsoCode.SZ] = { mask: "+268##-##-####", name: "\u0421\u0432\u0430\u0437\u0438\u043B\u0435\u043D\u0434" },
        _a[i18n.TuiCountryIsoCode.TC] = { mask: "+1(649) ###-####", name: "\u0422\u0451\u0440\u043A\u0441 \u0438 \u041A\u0430\u0439\u043A\u043E\u0441" },
        _a[i18n.TuiCountryIsoCode.TD] = { mask: "+235##-##-##-##", name: "\u0427\u0430\u0434" },
        _a[i18n.TuiCountryIsoCode.TG] = { mask: "+228##-###-###", name: "\u0422\u043E\u0433\u043E" },
        _a[i18n.TuiCountryIsoCode.TH] = { mask: "+66##-###-####", name: "\u0422\u0430\u0438\u043B\u0430\u043D\u0434 " },
        _a[i18n.TuiCountryIsoCode.TJ] = { mask: "+992##-###-####", name: "\u0422\u0430\u0434\u0436\u0438\u043A\u0438\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.TL] = { mask: "+670###-#####", name: "\u0412\u043E\u0441\u0442\u043E\u0447\u043D\u044B\u0439 \u0422\u0438\u043C\u043E\u0440" },
        _a[i18n.TuiCountryIsoCode.TM] = { mask: "+993#-###-####", name: "\u0422\u0443\u0440\u043A\u043C\u0435\u043D\u0438\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.TN] = { mask: "+216##-###-###", name: "\u0422\u0443\u043D\u0438\u0441" },
        _a[i18n.TuiCountryIsoCode.TO] = { mask: "+676#####", name: "\u0422\u043E\u043D\u0433\u0430" },
        _a[i18n.TuiCountryIsoCode.TR] = { mask: "+90(###) ###-####", name: "\u0422\u0443\u0440\u0446\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.TT] = { mask: "+1(868) ###-####", name: "\u0422\u0440\u0438\u043D\u0438\u0434\u0430\u0434 \u0438 \u0422\u043E\u0431\u0430\u0433\u043E" },
        _a[i18n.TuiCountryIsoCode.TW] = { mask: "+886#-####-####", name: "\u0422\u0430\u0439\u0432\u0430\u043D\u044C" },
        _a[i18n.TuiCountryIsoCode.TZ] = { mask: "+255##-###-####", name: "\u0422\u0430\u043D\u0437\u0430\u043D\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.UA] = { mask: "+380(##) ###-##-##", name: "\u0423\u043A\u0440\u0430\u0438\u043D\u0430" },
        _a[i18n.TuiCountryIsoCode.UG] = { mask: "+256(###) ###-###", name: "\u0423\u0433\u0430\u043D\u0434\u0430" },
        _a[i18n.TuiCountryIsoCode.US] = { mask: "+1(###) ###-####", name: "\u0421\u0428\u0410" },
        _a[i18n.TuiCountryIsoCode.UY] = { mask: "+598#-###-##-##", name: "\u0423\u0440\u0443\u0433\u0432\u0430\u0439" },
        _a[i18n.TuiCountryIsoCode.UZ] = { mask: "+998##-###-####", name: "\u0423\u0437\u0431\u0435\u043A\u0438\u0441\u0442\u0430\u043D" },
        _a[i18n.TuiCountryIsoCode.VC] = { mask: "+1(784) ###-####", name: "\u0421\u0435\u043D\u0442-\u0412\u0438\u043D\u0441\u0435\u043D\u0442 \u0438 \u0413\u0440\u0435\u043D\u0430\u0434\u0438\u043D\u044B" },
        _a[i18n.TuiCountryIsoCode.VE] = { mask: "+58(###) ###-####", name: "\u0412\u0435\u043D\u0435\u0441\u0443\u044D\u043B\u0430" },
        _a[i18n.TuiCountryIsoCode.VG] = {
            mask: "+1(284)###-####",
            name: "\u0411\u0440\u0438\u0442\u0430\u043D\u0441\u043A\u0438\u0435 \u0412\u0438\u0440\u0433\u0438\u043D\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430",
        },
        _a[i18n.TuiCountryIsoCode.VN] = { mask: "+84(###) ####-###", name: "\u0412\u044C\u0435\u0442\u043D\u0430\u043C" },
        _a[i18n.TuiCountryIsoCode.VU] = { mask: "+678##-#####", name: "\u0412\u0430\u043D\u0443\u0430\u0442\u0443 " },
        _a[i18n.TuiCountryIsoCode.WS] = { mask: "+685##-####", name: "\u0421\u0430\u043C\u043E\u0430" },
        _a[i18n.TuiCountryIsoCode.XK] = { mask: "+383##-###-###", name: "\u041A\u043E\u0441\u043E\u0432\u043E" },
        _a[i18n.TuiCountryIsoCode.YE] = { mask: "+967###-###-###", name: "\u0419\u0435\u043C\u0435\u043D " },
        _a[i18n.TuiCountryIsoCode.YT] = { mask: "+262 ## ## ## ## ##", name: "\u041C\u0430\u0439\u043E\u0442\u0442\u0430" },
        _a[i18n.TuiCountryIsoCode.ZA] = { mask: "+27##-###-####", name: "\u042E\u0436\u043D\u043E-\u0410\u0444\u0440\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F." },
        _a[i18n.TuiCountryIsoCode.ZM] = { mask: "+260##-###-####", name: "\u0417\u0430\u043C\u0431\u0438\u044F" },
        _a[i18n.TuiCountryIsoCode.ZW] = { mask: "+263#-######", name: "\u0417\u0438\u043C\u0431\u0430\u0431\u0432\u0435" },
        _a);
    /**
     * @deprecated will be deleted soon
     * TODO: remove in 3.0 (in ivy compilation)
     */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function countriesMasksReducer(acc, _a) {
        var _b = __read(_a, 2), countryIsoCode = _b[0], mask = _b[1].mask;
        acc[countryIsoCode] = mask;
        return acc;
    }
    /**
     * @deprecated will be deleted soon
     * Use token `TUI_COUNTRIES_MASKS`
     * TODO: remove in 3.0
     */
    var COUNTRIES_MASKS = Object.entries(COUNTRIES).reduce(countriesMasksReducer, {});

    var DATE_TIME_SEPARATOR = ", ";

    var EMPTY_MASK = {
        mask: function () {
            return false;
        },
    };

    var GROUP_CLASS_NAMES = [
        "_active",
        "_disabled",
        "_focused",
        "_focus-visible",
        "_hosted_dropdown_focused",
        "ng-invalid",
        "ng-touched",
        "_pressed",
        "_readonly",
    ];

    var iconBlank = "<svg></svg>";

    var MASK_AFTER_CODE_REGEXP = /\([#]+\)|[#\- ]/g;

    var TUI_PHONE_MASK = [
        cdk.CHAR_PLUS,
        "7",
        " ",
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        " ",
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        cdk.CHAR_HYPHEN,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        cdk.CHAR_HYPHEN,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
    ];
    var TUI_TIME_MASK = [
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        ":",
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
    ];
    /**
     * @deprecated use {@link tuiCreateDateMask} (from '@taiga-ui/kit') instead
     */
    var TUI_DATE_MASK = [
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        ".",
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        ".",
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
        core.TUI_DIGIT_REGEXP,
    ];
    /**
     * @deprecated use {@link tuiCreateDateRangeMask} (from '@taiga-ui/kit') instead
     */
    var TUI_DATE_RANGE_MASK = __spread(TUI_DATE_MASK, [
        cdk.CHAR_NO_BREAK_SPACE,
        cdk.CHAR_EN_DASH,
        cdk.CHAR_NO_BREAK_SPACE
    ], TUI_DATE_MASK);

    /**
     * Used as a limit for eliminating JS issues with floating point math
     */
    var TUI_FLOATING_PRECISION = 7;

    var MAX_DAY_RANGE_LENGTH_MAPPER = function (min, value, maxLength, backwards) {
        if (!(value === null || value === void 0 ? void 0 : value.isSingleDay) || !maxLength) {
            return min;
        }
        var dateShift = value.from
            .append(maxLength, backwards)
            .append({ day: 1 }, !backwards);
        if (backwards) {
            return dateShift.dayBefore(min) ? min : dateShift;
        }
        if (!min) {
            return dateShift;
        }
        return dateShift.dayAfter(min) ? min : dateShift;
    };

    var MAX_TIME_VALUES = {
        HH: 23,
        MM: 59,
        SS: 59,
        MS: 999,
    };

    var DEFAULT_ROUTER_LINK_OPTIONS = {
        fragment: "",
        preserveFragment: true,
        replaceUrl: false,
        skipLocationChange: false,
        queryParams: null,
        useHref: false,
        target: null,
    };

    exports.COUNTRIES = COUNTRIES;
    exports.COUNTRIES_MASKS = COUNTRIES_MASKS;
    exports.DATE_TIME_SEPARATOR = DATE_TIME_SEPARATOR;
    exports.DEFAULT_ROUTER_LINK_OPTIONS = DEFAULT_ROUTER_LINK_OPTIONS;
    exports.EMPTY_MASK = EMPTY_MASK;
    exports.GROUP_CLASS_NAMES = GROUP_CLASS_NAMES;
    exports.MASK_AFTER_CODE_REGEXP = MASK_AFTER_CODE_REGEXP;
    exports.MAX_DAY_RANGE_LENGTH_MAPPER = MAX_DAY_RANGE_LENGTH_MAPPER;
    exports.MAX_TIME_VALUES = MAX_TIME_VALUES;
    exports.TUI_DATE_MASK = TUI_DATE_MASK;
    exports.TUI_DATE_RANGE_MASK = TUI_DATE_RANGE_MASK;
    exports.TUI_FLOATING_PRECISION = TUI_FLOATING_PRECISION;
    exports.TUI_PHONE_MASK = TUI_PHONE_MASK;
    exports.TUI_TIME_MASK = TUI_TIME_MASK;
    exports.countriesMasksReducer = countriesMasksReducer;
    exports.iconBlank = iconBlank;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
