// @taiga-ui/kit/bundles/taiga-ui-kit-components-slider.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/utils'), require('rxjs/operators'), require('@ng-web-apis/common'), require('@angular/common'), require('rxjs'), require('@taiga-ui/core'), require('@taiga-ui/kit/abstract'), require('@taiga-ui/kit/tokens')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/slider', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/kit/utils', 'rxjs/operators', '@ng-web-apis/common', '@angular/common', 'rxjs', '@taiga-ui/core', '@taiga-ui/kit/abstract', '@taiga-ui/kit/tokens'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components.slider = {}), global.ng.core, global.ng.forms, global.cdk, global['taiga-ui'].kit.utils, global.rxjs.operators, global.common, global.ng.common, global.rxjs, global.core$1, global['taiga-ui'].kit.abstract, global['taiga-ui'].kit.tokens));
}(this, (function (exports, core, forms, cdk, utils, operators, common, common$1, rxjs, core$1, abstract, tokens) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_SLIDER_DEFAULT_OPTIONS = {
        size: "m",
        trackColor: "var(--tui-base-03)",
    };
    var TUI_SLIDER_OPTIONS = new core.InjectionToken("Default parameters for Slider component", { factory: function () { return TUI_SLIDER_DEFAULT_OPTIONS; } });
    function tuiSliderOptionsProvider(options) {
        return {
            provide: TUI_SLIDER_OPTIONS,
            useValue: __assign(__assign({}, TUI_SLIDER_DEFAULT_OPTIONS), options),
        };
    }

    var TuiSliderComponent = /** @class */ (function () {
        function TuiSliderComponent(control, changeDetectorRef, options, elementRef, userAgent, injector) {
            var _a;
            this.control = control;
            this.options = options;
            this.elementRef = elementRef;
            this.userAgent = userAgent;
            this.injector = injector;
            this.size = this.options.size;
            this.segments = 1;
            if (control instanceof forms.NgModel) {
                /**
                 * The ValueAccessor.writeValue method is called twice on any value accessor during component initialization,
                 * when a control is bound using [(ngModel)], first time with a phantom null value.
                 * With `changeDetection: ChangeDetectionStrategy.OnPush` the second call of writeValue with real value don't re-render the view.
                 * ___
                 * See this {@link https://github.com/angular/angular/issues/14988 issue}
                 */
                (_a = control.valueChanges) === null || _a === void 0 ? void 0 : _a.pipe(cdk.watch(changeDetectorRef), operators.take(1)).subscribe();
            }
        }
        Object.defineProperty(TuiSliderComponent.prototype, "min", {
            get: function () {
                return Number(this.elementRef.nativeElement.min);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "max", {
            get: function () {
                return Number(this.elementRef.nativeElement.max || 100);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "step", {
            get: function () {
                return Number(this.elementRef.nativeElement.step) || 1;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "value", {
            get: function () {
                var _a = this, elementRef = _a.elementRef, control = _a.control, hasKeySteps = _a.hasKeySteps;
                if (!hasKeySteps && control instanceof forms.NgModel) {
                    /**
                     * If developer uses `[(ngModel)]` and programmatically change value,
                     * the `elementRef.nativeElement.value` is equal to the previous value at this moment.
                     */
                    return control.viewModel;
                }
                return Number(elementRef.nativeElement.value) || 0;
            },
            set: function (newValue) {
                this.elementRef.nativeElement.value = "" + newValue;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "valuePercentage", {
            get: function () {
                return (100 * (this.value - this.min)) / (this.max - this.min) || 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "segmentWidth", {
            get: function () {
                return 100 / Math.max(1, this.segments);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "isOldEdge", {
            get: function () {
                return cdk.isEdgeOlderThan(cdk.CHROMIUM_EDGE_START_VERSION, this.userAgent);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderComponent.prototype, "hasKeySteps", {
            get: function () {
                return Boolean(this.injector.get(TuiSliderKeyStepsDirective, null));
            },
            enumerable: true,
            configurable: true
        });
        TuiSliderComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [TUI_SLIDER_OPTIONS,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [common.USER_AGENT,] }] },
            { type: core.Injector, decorators: [{ type: core.Inject, args: [core.Injector,] }] }
        ]; };
        __decorate([
            core.Input(),
            core.HostBinding('attr.data-size'),
            cdk.tuiDefaultProp()
        ], TuiSliderComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiSliderComponent.prototype, "segments", void 0);
        __decorate([
            core.HostBinding('style.--tui-slider-fill-percentage.%')
        ], TuiSliderComponent.prototype, "valuePercentage", null);
        __decorate([
            core.HostBinding('style.--tui-slider-segment-width.%')
        ], TuiSliderComponent.prototype, "segmentWidth", null);
        __decorate([
            core.HostBinding('class._old-edge')
        ], TuiSliderComponent.prototype, "isOldEdge", null);
        __decorate([
            cdk.tuiPure
        ], TuiSliderComponent.prototype, "hasKeySteps", null);
        TuiSliderComponent = __decorate([
            core.Component({
                /**
                 * We have to call our component as `<input tuiSlider type="range" ... />`
                 * because otherwise built-in angular
                 * {@link https://github.com/angular/angular/blob/master/packages/forms/src/directives/range_value_accessor.ts#L45 RangeValueAccessor}
                 * cannot be matched by its CSS selector.
                 */
                selector: 'input[type=range][tuiSlider]',
                template: '',
                host: {
                    /**
                     * For change detection.
                     * Webkit does not have built-in method for customization of filling progress (as Firefox).
                     * We draw filling of progress by `background: linear-gradient(...)` of the track.
                     * This function triggers change detection (for {@link valuePercentage} function) when we drag thumb of the input.
                     */
                    '(input)': '0',
                    '[style.--tui-slider-track-color]': 'options.trackColor',
                },
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{display:block;width:100%;color:var(--tui-primary);cursor:pointer}:host:active{cursor:ew-resize}:host:disabled{opacity:var(--tui-disabled-opacity);cursor:auto}:host:not(._old-edge){-webkit-appearance:none;-moz-appearance:none;appearance:none;height:.125rem;padding:.4375rem 0;background-color:transparent;background-clip:content-box;outline:0;border-radius:var(--tui-radius-m)}:host:not(._old-edge)::-webkit-slider-container{border-radius:inherit}:host:not(._old-edge)[data-size='m']::-webkit-slider-runnable-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0 .625rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width)),linear-gradient(to right,currentColor var(--tui-slider-fill-percentage),transparent var(--tui-slider-fill-percentage));background-position-x:0,.375rem,0;background-size:calc(100% - 1rem),calc(100% - 1rem),auto}:host:not(._old-edge)[data-size='s']::-webkit-slider-runnable-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0 .375rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width)),linear-gradient(to right,currentColor var(--tui-slider-fill-percentage),transparent var(--tui-slider-fill-percentage));background-position-x:0,.125rem,0;background-size:calc(100% - .5rem),calc(100% - .5rem),auto}:host:not(._old-edge)[data-size='m']::-moz-range-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0 .625rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width));background-position-x:0,.375rem;background-size:calc(100% - 1rem)}:host:not(._old-edge)[data-size='s']::-moz-range-track{height:.125rem;border-radius:inherit;background-repeat:no-repeat;background-color:var(--tui-slider-track-color);background-image:linear-gradient(to right,currentColor 0 .375rem,transparent .25rem),repeating-linear-gradient(to right,var(--tui-base-07) 0 .25rem,transparent 0 var(--tui-slider-segment-width));background-position-x:0,.125rem;background-size:calc(100% - .5rem)}:host:not(._old-edge)[data-size='m']::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:1rem;width:1rem;margin-top:-.4375rem}:not(:disabled):host:not(._old-edge)[data-size='m']::-webkit-slider-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size='m']::-webkit-slider-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size='m']::-webkit-slider-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size='m']::-webkit-slider-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size='s']::-webkit-slider-thumb{-webkit-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:.5rem;width:.5rem;margin-top:-.1875rem}:not(:disabled):host:not(._old-edge)[data-size='s']::-webkit-slider-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size='s']::-webkit-slider-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size='s']::-webkit-slider-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size='s']::-webkit-slider-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size='m']::-moz-range-thumb{-moz-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:1rem;width:1rem}:not(:disabled):host:not(._old-edge)[data-size='m']::-moz-range-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size='m']::-moz-range-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size='m']::-moz-range-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size='m']::-moz-range-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)[data-size='s']::-moz-range-thumb{-moz-appearance:none;appearance:none;background-color:currentColor;border:none;border-radius:50%;height:.5rem;width:.5rem}:not(:disabled):host:not(._old-edge)[data-size='s']::-moz-range-thumb{cursor:ew-resize}:not(:disabled):host:not(._old-edge)[data-size='s']::-moz-range-thumb:hover{background:var(--tui-primary-hover)}:not(:disabled):host:not(._old-edge)[data-size='s']::-moz-range-thumb:active{background:var(--tui-primary-active)}:focus-visible:host:not(._old-edge)[data-size='s']::-moz-range-thumb{box-shadow:0 0 0 2px inset var(--tui-focus)}:host:not(._old-edge)::-moz-range-progress{background:currentColor;border-radius:inherit;border-top-right-radius:0;border-bottom-right-radius:0}:host._old-edge::-ms-thumb{background:currentColor;border-radius:50%}:host._old-edge::-ms-fill-lower{background:currentColor}:host._old-edge::-ms-track{background:var(--tui-slider-track-color);border:none}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(TUI_SLIDER_OPTIONS)),
            __param(3, core.Inject(core.ElementRef)),
            __param(4, core.Inject(common.USER_AGENT)),
            __param(5, core.Inject(core.Injector))
        ], TuiSliderComponent);
        return TuiSliderComponent;
    }());

    // @dynamic
    var TuiSliderKeyStepsDirective = /** @class */ (function (_super) {
        __extends(TuiSliderKeyStepsDirective, _super);
        function TuiSliderKeyStepsDirective(control, changeDetectorRef, elementRef, slider) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.elementRef = elementRef;
            _this.slider = slider;
            _this.keySteps = [];
            _this.keyStepsInput = cdk.typedFromEvent(_this.elementRef.nativeElement, 'input').pipe(operators.map(function () { return _this.controlValue; }));
            return _this;
        }
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "nativeFocusableElement", {
            get: function () {
                return this.computedDisabled ? null : this.elementRef.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "min", {
            get: function () {
                var _a;
                return ((_a = this.keySteps[0]) === null || _a === void 0 ? void 0 : _a[1]) || 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "max", {
            get: function () {
                var _a;
                return ((_a = this.keySteps[this.keySteps.length - 1]) === null || _a === void 0 ? void 0 : _a[1]) || 100;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderKeyStepsDirective.prototype, "controlValue", {
            get: function () {
                return utils.tuiPercentageToKeyStepValue(this.slider.valuePercentage, this.keySteps);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * TODO: 3.0
         * ___
         * Also add @HostListener(`input`): to be similar to
         * {@link https://github.com/angular/angular/blob/main/packages/forms/src/directives/range_value_accessor.ts#L47-L48 RangeValueAccessor}
         * ___
         * Remove {@link keyStepsInput}
         */
        TuiSliderKeyStepsDirective.prototype.updateControlValue = function () {
            this.updateValue(this.controlValue);
        };
        TuiSliderKeyStepsDirective.prototype.writeValue = function (controlValue) {
            if (controlValue === null) {
                return;
            }
            var clampedControlValue = cdk.clamp(controlValue, this.min, this.max);
            cdk.tuiAssert.assert(controlValue === clampedControlValue, '\n[SliderKeySteps]: You cannot programmatically set value which is less/more than min/max');
            this.slider.value = this.transformToNativeValue(clampedControlValue);
        };
        TuiSliderKeyStepsDirective.prototype.getFallbackValue = function () {
            return 0;
        };
        TuiSliderKeyStepsDirective.prototype.transformToNativeValue = function (controlValue) {
            var _a = this.slider, min = _a.min, max = _a.max;
            var newValuePercentage = utils.tuiKeyStepValueToPercentage(controlValue, this.keySteps);
            return (newValuePercentage * (max - min)) / 100 + min;
        };
        TuiSliderKeyStepsDirective.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: TuiSliderComponent, decorators: [{ type: core.Inject, args: [core.forwardRef(function () { return TuiSliderComponent; }),] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp(utils.tuiCheckKeyStepsHaveMinMaxPercents, 'Should contain min and max values')
        ], TuiSliderKeyStepsDirective.prototype, "keySteps", void 0);
        __decorate([
            core.Output()
        ], TuiSliderKeyStepsDirective.prototype, "keyStepsInput", void 0);
        __decorate([
            core.HostListener('change')
        ], TuiSliderKeyStepsDirective.prototype, "updateControlValue", null);
        TuiSliderKeyStepsDirective = __decorate([
            core.Directive({
                selector: 'input[tuiSlider][keySteps]',
                host: {
                    '[attr.aria-valuenow]': 'safeCurrentValue',
                    '[attr.aria-valuemin]': 'min',
                    '[attr.aria-valuemax]': 'max',
                },
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core.ElementRef)),
            __param(3, core.Inject(core.forwardRef(function () { return TuiSliderComponent; })))
        ], TuiSliderKeyStepsDirective);
        return TuiSliderKeyStepsDirective;
    }(cdk.AbstractTuiControl));
    /**
     * @deprecated DONT USE IT! It is just temporary solution for internal purposes only. We will delete it in next major release.
     * TODO delete it in v3.0
     *
     */
    var TuiSliderTickLabelPipe = /** @class */ (function () {
        function TuiSliderTickLabelPipe() {
        }
        TuiSliderTickLabelPipe.prototype.transform = function (tickIndex, totalSegments, keySteps) {
            var percentage = (100 / totalSegments) * tickIndex;
            return utils.tuiPercentageToKeyStepValue(percentage, keySteps);
        };
        TuiSliderTickLabelPipe = __decorate([
            core.Pipe({ name: 'tuiSliderTickLabel' })
        ], TuiSliderTickLabelPipe);
        return TuiSliderTickLabelPipe;
    }());

    var SLIDER_INTERACTION_KEYS = new Set([
        'ArrowLeft',
        'ArrowRight',
        'ArrowUp',
        'ArrowDown',
        'Home',
        'End',
        'PageUp',
        'PageDown',
    ]);
    /**
     * Native <input type='range' readonly> doesn't work.
     * This directive imitates this native behaviour.
     */
    // @dynamic
    var TuiSliderReadonlyDirective = /** @class */ (function () {
        function TuiSliderReadonlyDirective(elementRef, documentRef, destroy$) {
            var _this = this;
            this.readonly = true;
            var touchStart$ = cdk.typedFromEvent(elementRef.nativeElement, 'touchstart', {
                passive: false,
            });
            var touchMove$ = cdk.typedFromEvent(documentRef, 'touchmove', {
                passive: false,
            });
            var touchEnd$ = cdk.typedFromEvent(documentRef, 'touchend', {
                passive: true,
            });
            var shouldPreventMove$ = rxjs.merge(touchStart$.pipe(operators.tap(function (e) { return _this.preventEvent(e); }), operators.mapTo(true)), touchEnd$.pipe(operators.mapTo(false)));
            /**
             * @bad TODO think about another solution.
             * Keep in mind that preventing touch event (on slider) inside `@HostListener('touchstart')` doesn't work for mobile chrome.
             */
            rxjs.combineLatest([touchMove$, shouldPreventMove$])
                .pipe(operators.filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], shouldPreventMove = _b[1];
                return shouldPreventMove;
            }), operators.takeUntil(destroy$))
                .subscribe(function (_a) {
                var _b = __read(_a, 1), moveEvent = _b[0];
                return _this.preventEvent(moveEvent);
            });
        }
        TuiSliderReadonlyDirective.prototype.preventEvent = function (event) {
            if (event.cancelable && cdk.tuiCoerceBooleanProperty(this.readonly)) {
                event.preventDefault();
            }
        };
        TuiSliderReadonlyDirective.prototype.preventKeyboardInteraction = function (event) {
            if (SLIDER_INTERACTION_KEYS.has(event.key)) {
                this.preventEvent(event);
            }
        };
        TuiSliderReadonlyDirective.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common$1.DOCUMENT,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiSliderReadonlyDirective.prototype, "readonly", void 0);
        __decorate([
            core.HostListener('mousedown', ['$event'])
        ], TuiSliderReadonlyDirective.prototype, "preventEvent", null);
        __decorate([
            core.HostListener('keydown', ['$event'])
        ], TuiSliderReadonlyDirective.prototype, "preventKeyboardInteraction", null);
        TuiSliderReadonlyDirective = __decorate([
            core.Directive({
                selector: 'input[tuiSlider][readonly]',
                providers: [cdk.TuiDestroyService],
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(common$1.DOCUMENT)),
            __param(2, core.Inject(cdk.TuiDestroyService))
        ], TuiSliderReadonlyDirective);
        return TuiSliderReadonlyDirective;
    }());

    var TuiSliderThumbLabelComponent = /** @class */ (function () {
        function TuiSliderThumbLabelComponent() {
        }
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "size", {
            get: function () {
                var _a;
                return ((_a = this.slider) === null || _a === void 0 ? void 0 : _a.size) || 'm';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ratio", {
            get: function () {
                var _a;
                return (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.valuePercentage) || 0) / 100;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderThumbLabelComponent.prototype, "ghostLeft", {
            get: function () {
                var _a;
                return this.ratio * (((_a = this.slider) === null || _a === void 0 ? void 0 : _a.elementRef.nativeElement.offsetWidth) || 0);
            },
            enumerable: true,
            configurable: true
        });
        TuiSliderThumbLabelComponent.prototype.ngAfterContentInit = function () {
            var _a;
            cdk.tuiAssert.assert(Boolean((_a = this.control) === null || _a === void 0 ? void 0 : _a.valueChanges), '\n[tuiSliderThumbLabel] expected <input tuiSlider type="range" /> to use Angular Forms.\n' +
                'Use [(ngModel)] or [formControl] or formControlName for correct work.');
        };
        __decorate([
            core.ContentChild(TuiSliderComponent)
        ], TuiSliderThumbLabelComponent.prototype, "slider", void 0);
        __decorate([
            core.ContentChild(forms.NgControl)
        ], TuiSliderThumbLabelComponent.prototype, "control", void 0);
        TuiSliderThumbLabelComponent = __decorate([
            core.Component({
                selector: '[tuiSliderThumbLabel]',
                template: "<ng-container *ngIf=\"control?.valueChanges | async\"></ng-container>\n\n<div\n    class=\"t-ghost\"\n    [attr.data-size]=\"size\"\n    [style.left.px]=\"ghostLeft\"\n    [style.--tui-slider-thumb-ratio]=\"ratio\"\n>\n    <ng-content></ng-content>\n</div>\n\n<ng-content select=\"input[type=range]\"></ng-content>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{position:relative}.t-ghost{position:absolute;top:0;bottom:0;margin:auto;border-radius:50%;pointer-events:none}.t-ghost[data-size='s']{width:.5rem;height:.5rem;transform:translateX(calc(var(--tui-slider-thumb-ratio) * -.5rem))}.t-ghost[data-size='m']{width:1rem;height:1rem;transform:translateX(calc(var(--tui-slider-thumb-ratio) * -1rem))}"]
            })
        ], TuiSliderThumbLabelComponent);
        return TuiSliderThumbLabelComponent;
    }());

    /**
     * @deprecated use {@link TuiSliderComponent} instead
     * TODO: 3.0 remove
     */
    // @dynamic
    var TuiSliderOldComponent = /** @class */ (function (_super) {
        __extends(TuiSliderOldComponent, _super);
        function TuiSliderOldComponent(control, changeDetectorRef, documentRef, fromToTexts$) {
            return _super.call(this, control, changeDetectorRef, documentRef, fromToTexts$) || this;
        }
        TuiSliderOldComponent_1 = TuiSliderOldComponent;
        Object.defineProperty(TuiSliderOldComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.dotRight ? this.dotRight.nativeElement : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderOldComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderOldComponent.prototype, "left", {
            get: function () {
                return 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSliderOldComponent.prototype, "right", {
            get: function () {
                return 100 - 100 * this.getFractionFromValue(this.value);
            },
            enumerable: true,
            configurable: true
        });
        TuiSliderOldComponent.prototype.processValue = function (value) {
            this.updateValue(this.valueGuard(value));
        };
        TuiSliderOldComponent.prototype.decrement = function () {
            this.processStep(false);
        };
        TuiSliderOldComponent.prototype.increment = function () {
            this.processStep(true);
        };
        TuiSliderOldComponent.prototype.getFallbackValue = function () {
            return 0;
        };
        TuiSliderOldComponent.prototype.processStep = function (increment) {
            var fraction = this.getFractionFromValue(this.value);
            var step = this.computedStep;
            var value = this.getValueFromFraction(increment ? fraction + step : fraction - step);
            this.processValue(value);
        };
        var TuiSliderOldComponent_1;
        TuiSliderOldComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common$1.DOCUMENT,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_FROM_TO_TEXTS,] }] }
        ]; };
        TuiSliderOldComponent = TuiSliderOldComponent_1 = __decorate([
            core.Component({
                selector: 'tui-slider',
                template: "<div\n    class=\"event-catcher\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n    (mousedown)=\"onMouseDown($event)\"\n    (touchstart)=\"onTouchStart($event)\"\n>\n    <div class=\"wrapper\">\n        <div\n            *ngIf=\"segmented\"\n            class=\"segments\"\n        >\n            <span\n                *tuiRepeatTimes=\"let segmentIndex of segments + 1\"\n                automation-id=\"tui-slider__segment\"\n                class=\"segment\"\n            >\n                <span\n                    *ngIf=\"fromToTexts$ | async as fromToText\"\n                    class=\"number\"\n                >\n                    {{ getSegmentPrefix(segmentIndex, fromToText) }}\n                    {{ getSegmentLabel(segmentIndex) | tuiFormatNumber }}\n                    <span *ngIf=\"pluralizeMap\">\n                        {{ getSegmentLabel(segmentIndex) | i18nPlural: pluralizeMap }}\n                    </span>\n                </span>\n            </span>\n        </div>\n        <div\n            automation-id=\"tui-slider__bar\"\n            class=\"bar\"\n            [style.left.%]=\"left\"\n            [style.right.%]=\"right\"\n        >\n            <div\n                #dotLeft\n                automation-id=\"tui-slider__left\"\n                class=\"dot\"\n                [class.dot_focus-visible]=\"focusVisibleLeft && computedFocused\"\n                [tuiFocusable]=\"isLeftFocusable\"\n                (tuiFocusVisibleChange)=\"onLeftFocusVisible($event)\"\n                (keydown.arrowLeft.prevent)=\"decrement()\"\n                (keydown.arrowDown.prevent)=\"decrement()\"\n                (keydown.arrowRight.prevent)=\"increment()\"\n                (keydown.arrowUp.prevent)=\"increment()\"\n            ></div>\n            <div\n                #dotRight\n                automation-id=\"tui-slider__right\"\n                class=\"dot\"\n                [class.dot_focus-visible]=\"focusVisibleRight && computedFocused\"\n                [tuiFocusable]=\"isRightFocusable\"\n                (tuiFocusVisibleChange)=\"onRightFocusVisible($event)\"\n                (keydown.arrowLeft.prevent)=\"decrement()\"\n                (keydown.arrowDown.prevent)=\"decrement()\"\n                (keydown.arrowRight.prevent)=\"increment()\"\n                (keydown.arrowUp.prevent)=\"increment()\"\n            ></div>\n        </div>\n    </div>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiSliderOldComponent_1; }),
                    },
                ],
                styles: [":host{display:block;border-radius:var(--tui-radius-m);color:var(--tui-base-03);min-height:.125rem;font-size:1rem}:host[data-size='s']{min-height:.125rem;font-size:.5rem}:host._segmented{padding-bottom:1.25rem}:host._disabled{cursor:default;pointer-events:none;opacity:var(--tui-disabled-opacity)}.event-catcher{min-height:inherit;border-radius:inherit;padding:.4375rem 0;cursor:pointer}.wrapper{position:relative;min-height:inherit;border-radius:inherit;background-color:currentColor;border:solid transparent;border-width:0 1em}.bar{position:absolute;top:0;bottom:0;margin:0 -1em;border-radius:inherit;background-color:var(--tui-primary)}:host-context(tui-input-slider) .bar{margin-left:calc(var(--tui-radius-m)/ 2);border-bottom-left-radius:calc(var(--tui-radius-m) * 5) calc(var(--tui-radius-m) * 1.5)}.dot{transition-property:background-color;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;position:absolute;top:50%;z-index:1;width:1em;height:1em;transform:translate(0,-50%);border-radius:100%;background-color:var(--tui-primary);outline:0;cursor:ew-resize}.dot:last-child{right:0;transform:translate(0,-50%)}.dot:hover{background-color:var(--tui-primary-hover)}.dot:active{background-color:var(--tui-primary-active)}.dot_focus-visible{box-shadow:inset 0 0 0 2px var(--tui-focus)}.segments{position:relative;z-index:1;display:flex;justify-content:space-between;min-height:inherit;margin:0 -1em}.segments-spacer{height:1.625rem}.segment{position:relative;height:inherit;width:.25rem;background-color:rgba(0,0,0,.36);pointer-events:none}.segment:first-of-type,.segment:last-of-type{background-color:transparent}.number{position:absolute;left:50%;transform:translate(-50%,0);top:100%;font:var(--tui-font-text-s);margin-top:.5rem;color:var(--tui-text-02);white-space:nowrap}.segment:first-of-type .number{left:-.05em;transform:none}.segment:last-of-type .number{left:auto;right:-.05em;transform:none}:host._disabled .number{color:var(--tui-text-01)}", ".dot:first-child{display:none}.wrapper{border-left:none}.bar,.segments{margin-left:0}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(common$1.DOCUMENT)),
            __param(3, core.Inject(tokens.TUI_FROM_TO_TEXTS))
        ], TuiSliderOldComponent);
        return TuiSliderOldComponent;
    }(abstract.AbstractTuiSlider));

    var TuiSliderModule = /** @class */ (function () {
        function TuiSliderModule() {
        }
        TuiSliderModule = __decorate([
            core.NgModule({
                imports: [
                    common$1.CommonModule,
                    cdk.TuiRepeatTimesModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiFocusVisibleModule,
                    cdk.TuiActiveZoneModule,
                    core$1.TuiFormatNumberPipeModule,
                ],
                declarations: [
                    TuiSliderComponent,
                    TuiSliderThumbLabelComponent,
                    TuiSliderKeyStepsDirective,
                    TuiSliderReadonlyDirective,
                    TuiSliderTickLabelPipe,
                    TuiSliderOldComponent,
                ],
                exports: [
                    TuiSliderComponent,
                    TuiSliderThumbLabelComponent,
                    TuiSliderKeyStepsDirective,
                    TuiSliderReadonlyDirective,
                    TuiSliderTickLabelPipe,
                    TuiSliderOldComponent,
                ],
            })
        ], TuiSliderModule);
        return TuiSliderModule;
    }());

    exports.TUI_SLIDER_DEFAULT_OPTIONS = TUI_SLIDER_DEFAULT_OPTIONS;
    exports.TUI_SLIDER_OPTIONS = TUI_SLIDER_OPTIONS;
    exports.TuiSliderComponent = TuiSliderComponent;
    exports.TuiSliderKeyStepsDirective = TuiSliderKeyStepsDirective;
    exports.TuiSliderModule = TuiSliderModule;
    exports.TuiSliderOldComponent = TuiSliderOldComponent;
    exports.TuiSliderReadonlyDirective = TuiSliderReadonlyDirective;
    exports.TuiSliderThumbLabelComponent = TuiSliderThumbLabelComponent;
    exports.TuiSliderTickLabelPipe = TuiSliderTickLabelPipe;
    exports.tuiSliderOptionsProvider = tuiSliderOptionsProvider;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
