// @taiga-ui/kit/bundles/taiga-ui-kit-components-text-area.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('rxjs'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/text-area', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', 'rxjs', '@angular/common'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['text-area'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global.rxjs, global.ng.common));
}(this, (function (exports, core, forms, cdk, core$1, rxjs, common) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TUI_TEXT_AREA_PROVIDERS = [
        {
            provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
            useExisting: core.forwardRef(function () { return TuiTextAreaComponent; }),
        },
        {
            provide: cdk.AbstractTuiControl,
            useExisting: core.forwardRef(function () { return TuiTextAreaComponent; }),
        },
        core$1.TEXTFIELD_CONTROLLER_PROVIDER,
        core$1.HINT_CONTROLLER_PROVIDER,
        core$1.MODE_PROVIDER,
    ];

    var DEFAULT_ROWS = 20;
    var LINE_HEIGHT_M = 20;
    var LINE_HEIGHT_L = 24;
    var TuiTextAreaComponent = /** @class */ (function (_super) {
        __extends(TuiTextAreaComponent, _super);
        function TuiTextAreaComponent(control, changeDetectorRef, appearance, isIOS, mode$, controller, hintController) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.appearance = appearance;
            _this.isIOS = isIOS;
            _this.mode$ = mode$;
            _this.controller = controller;
            _this.hintController = hintController;
            _this.rows = DEFAULT_ROWS;
            _this.expandable = false;
            return _this;
        }
        Object.defineProperty(TuiTextAreaComponent.prototype, "labelOutside", {
            get: function () {
                return this.controller.labelOutside;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "nativeFocusableElement", {
            get: function () {
                var _a, _b;
                if (this.computedDisabled) {
                    return null;
                }
                return (((_a = this.textfield) === null || _a === void 0 ? void 0 : _a.nativeElement) || ((_b = this.focusableElement) === null || _b === void 0 ? void 0 : _b.nativeElement) || null);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "size", {
            get: function () {
                return this.controller.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "border", {
            get: function () {
                return core$1.getBorder(false, this.hasCleaner, this.hasTooltip);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasCleaner", {
            get: function () {
                return this.controller.cleaner && this.hasValue && this.interactive;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasTooltip", {
            get: function () {
                return !!this.hintController.content && !this.disabled;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasValue", {
            get: function () {
                return this.value !== '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasCounter", {
            get: function () {
                return !!this.controller.maxLength && this.interactive;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasPlaceholder", {
            get: function () {
                return !this.controller.labelOutside || (!this.hasValue && !this.hasExampleText);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "hasExampleText", {
            get: function () {
                var _a;
                var text = this.controller.exampleText || ((_a = this.textfield) === null || _a === void 0 ? void 0 : _a.nativeElement.placeholder);
                return !!text && this.focused && !this.hasValue && !this.readOnly;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "computeMaxHeight", {
            get: function () {
                return this.expandable ? this.rows * this.lineHeight : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "placeholderRaised", {
            get: function () {
                return (!this.controller.labelOutside &&
                    ((this.computedFocused && !this.readOnly) || this.hasValue));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "fittedContent", {
            get: function () {
                return this.value.slice(0, this.controller.maxLength || Infinity);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiTextAreaComponent.prototype, "extraContent", {
            get: function () {
                return this.value.slice(this.controller.maxLength || Infinity);
            },
            enumerable: true,
            configurable: true
        });
        TuiTextAreaComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiTextAreaComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiTextAreaComponent.prototype.onPressed = function (pressed) {
            this.updatePressed(pressed);
        };
        TuiTextAreaComponent.prototype.onValueChange = function (value) {
            this.updateValue(value);
        };
        TuiTextAreaComponent.prototype.onMouseDown = function (event) {
            if (event.target === this.nativeFocusableElement) {
                return;
            }
            event.preventDefault();
            if (this.nativeFocusableElement) {
                cdk.setNativeFocused(this.nativeFocusableElement);
            }
        };
        TuiTextAreaComponent.prototype.getFallbackValue = function () {
            return '';
        };
        Object.defineProperty(TuiTextAreaComponent.prototype, "lineHeight", {
            get: function () {
                return this.controller.size === 'm' ? LINE_HEIGHT_M : LINE_HEIGHT_L;
            },
            enumerable: true,
            configurable: true
        });
        TuiTextAreaComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_APPEARANCE,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_IOS,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [core$1.TUI_MODE,] }] },
            { type: core$1.TuiTextfieldController, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER,] }] },
            { type: core$1.TuiHintControllerDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_HINT_WATCHED_CONTROLLER,] }] }
        ]; };
        __decorate([
            core.ViewChild('focusableElement')
        ], TuiTextAreaComponent.prototype, "focusableElement", void 0);
        __decorate([
            core.ContentChild(core$1.TuiTextfieldComponent, { read: core.ElementRef })
        ], TuiTextAreaComponent.prototype, "textfield", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiTextAreaComponent.prototype, "rows", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class._expandable'),
            cdk.tuiDefaultProp()
        ], TuiTextAreaComponent.prototype, "expandable", void 0);
        __decorate([
            core.HostBinding('class._label-outside')
        ], TuiTextAreaComponent.prototype, "labelOutside", null);
        __decorate([
            core.HostBinding('attr.data-size')
        ], TuiTextAreaComponent.prototype, "size", null);
        __decorate([
            core.HostBinding('style.--border-end.rem')
        ], TuiTextAreaComponent.prototype, "border", null);
        __decorate([
            core.HostBinding('class._has-tooltip')
        ], TuiTextAreaComponent.prototype, "hasTooltip", null);
        __decorate([
            core.HostBinding('class._has-value')
        ], TuiTextAreaComponent.prototype, "hasValue", null);
        __decorate([
            core.HostBinding('class._has-counter')
        ], TuiTextAreaComponent.prototype, "hasCounter", null);
        __decorate([
            core.HostListener('focusin', ['true']),
            core.HostListener('focusout', ['false'])
        ], TuiTextAreaComponent.prototype, "onFocused", null);
        TuiTextAreaComponent = __decorate([
            core.Component({
                selector: 'tui-text-area',
                template: "<tui-wrapper\n    automation-id=\"tui-text-area__wrapper\"\n    class=\"t-outline\"\n    [appearance]=\"appearance\"\n    [disabled]=\"disabled\"\n    [focused]=\"computedFocused\"\n    [hovered]=\"computedHovered\"\n    [readOnly]=\"readOnly\"\n    [invalid]=\"computedInvalid\"\n>\n    <div\n        *ngIf=\"hasCounter\"\n        automation-id=\"tui-text-area__counter\"\n        class=\"t-counter\"\n    >\n        {{ value.length }}/{{ controller.maxLength }}\n    </div>\n\n    <label\n        class=\"t-content\"\n        (tuiHoveredChange)=\"onHovered($event)\"\n        (mousedown)=\"onMouseDown($event)\"\n    >\n        <div class=\"t-wrapper\">\n            <div\n                *ngIf=\"hasPlaceholder\"\n                automation-id=\"tui-text-area__placeholder\"\n                class=\"t-placeholder\"\n                [class.t-placeholder_raised]=\"placeholderRaised\"\n            >\n                <ng-content></ng-content>\n            </div>\n            <tui-scrollbar\n                automation-id=\"tui-text-area__scrollbar\"\n                class=\"t-box\"\n                [style.maxHeight.px]=\"computeMaxHeight\"\n            >\n                <div class=\"t-input-wrapper\">\n                    <div class=\"t-relative\">\n                        <div\n                            aria-hidden=\"true\"\n                            class=\"t-pseudo-content\"\n                        >\n                            <span [textContent]=\"fittedContent || controller.exampleText\"></span>\n                            <span\n                                class=\"t-pseudo-content__extra\"\n                                [textContent]=\"extraContent\"\n                            ></span>\n                            &nbsp;\n                        </div>\n                        <textarea\n                            #focusableElement\n                            automation-id=\"tui-text-area__native\"\n                            class=\"t-input\"\n                            [placeholder]=\"controller.exampleText\"\n                            [id]=\"id\"\n                            [tuiDescribedBy]=\"id\"\n                            [tuiFocusable]=\"computedFocusable\"\n                            [disabled]=\"computedDisabled\"\n                            [ngModel]=\"value\"\n                            [readOnly]=\"readOnly\"\n                            (ngModelChange)=\"onValueChange($event)\"\n                        ></textarea>\n                        <ng-content select=\"textarea\"></ng-content>\n                    </div>\n                </div>\n            </tui-scrollbar>\n        </div>\n        <div class=\"t-icons\">\n            <tui-svg\n                *ngIf=\"hasCleaner\"\n                src=\"tuiIconCloseLarge\"\n                class=\"t-cleaner\"\n                (click.stop)=\"onValueChange('')\"\n            ></tui-svg>\n            <tui-tooltip\n                *ngIf=\"hasTooltip\"\n                automation-id=\"tui-text-area__tooltip\"\n                class=\"t-tooltip\"\n                [describeId]=\"id\"\n                [direction]=\"hintController.direction\"\n                [mode]=\"hintController.mode\"\n                [content]=\"hintController.content || ''\"\n                [showDelay]=\"hintController.showDelay\"\n                [hideDelay]=\"hintController.hideDelay\"\n            ></tui-tooltip>\n        </div>\n    </label>\n</tui-wrapper>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: TUI_TEXT_AREA_PROVIDERS,
                host: {
                    '($.data-mode.attr)': 'mode$',
                    '[class._ios]': 'isIOS',
                },
                styles: [":host{position:relative;z-index:0;display:flex;font:var(--tui-font-text-m);flex-direction:column;min-height:6.75rem;border-radius:var(--tui-radius-m);color:var(--tui-text-01)}:host[data-mode=onDark]{color:var(--tui-text-01-night)}:host[data-size='m']{font:var(--tui-font-text-s);min-height:5.5rem}:host[data-size='m']._has-counter{min-height:6.625rem}:host[data-size='l']._has-counter{min-height:7.875rem}.t-outline{min-height:inherit}.t-content{display:block;margin-top:0;margin-bottom:0;min-height:inherit;box-sizing:border-box;overflow:hidden;cursor:text}:host._disabled .t-content{cursor:auto;opacity:var(--tui-disabled-opacity)}:host:not(._expandable) .t-content{position:absolute;top:0;left:0;bottom:1px;right:0;min-height:auto}:host._has-counter:not(._expandable) .t-content{bottom:1.6875rem}.t-wrapper{position:relative;width:100%;height:100%;min-height:inherit;padding-bottom:.4375rem;box-sizing:border-box}:host[data-size='m']._label-outside .t-wrapper{padding-top:.75rem}:host[data-size='l']._label-outside .t-wrapper{padding-top:1rem}:host[data-size='m']:not(._label-outside) .t-wrapper{padding-top:1.375rem}:host[data-size='l']:not(._label-outside) .t-wrapper{padding-top:1.625rem}:host-context(table)[data-size='m']._label-outside .t-wrapper{padding-bottom:.75rem}:host-context(table)[data-size='l']._label-outside .t-wrapper{padding-bottom:1rem}.t-input-wrapper{min-height:inherit;width:100%;flex:1}.t-relative{position:relative;min-height:inherit}.t-box{display:flex;min-height:100%;width:100%}:host:not(._expandable) .t-box{height:100%}:host._has-counter._expandable .t-box{margin-bottom:1.25rem}.t-pseudo-content{white-space:pre-wrap;word-wrap:break-word;pointer-events:none;color:transparent;overflow:hidden;border:solid transparent;border-width:0 var(--border-end,0) 0 var(--border-start,0);border-inline-start-width:var(--border-start,0);border-inline-end-width:var(--border-end,0)}:host[data-size='m'] .t-pseudo-content{padding:0 .75rem}:host[data-size='l'] .t-pseudo-content{padding:0 1rem}.t-pseudo-content__extra{background-color:var(--tui-error-bg-night)}.t-input{padding:0;margin:0;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;position:absolute;top:0;left:0;width:100%;height:100%;box-sizing:border-box;resize:none;overflow:hidden;outline:0;border:solid transparent;border-width:0 var(--border-end,0) 0 var(--border-start,0);border-inline-start-width:var(--border-start,0);border-inline-end-width:var(--border-end,0)}.t-input:-webkit-autofill,.t-input:-webkit-autofill:focus,.t-input:-webkit-autofill:hover{caret-color:var(--tui-base-09);border-radius:inherit;color:inherit!important;background-color:transparent!important;-webkit-text-fill-color:var(--tui-text-01)!important;border-color:var(--tui-autofill);-webkit-box-shadow:0 0 0 100rem var(--tui-autofill) inset!important}.t-input:not(:last-of-type){display:none}.t-input::-ms-input-placeholder{color:var(--tui-text-03);opacity:0}.t-input::placeholder{color:var(--tui-text-03);opacity:0}:host[data-mode=onDark] .t-input::-ms-input-placeholder{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-input::placeholder{color:var(--tui-text-03-night)}:host._focused .t-input:not(:read-only)::-ms-input-placeholder{opacity:1}:host._focused .t-input:not(:-moz-read-only)::placeholder{opacity:1}:host._focused .t-input:not(:read-only)::placeholder{opacity:1}:host[data-size='m'] .t-input{padding:0 .75rem}:host[data-size='l'] .t-input{padding:0 1rem}:host[data-mode=onDark]._disabled .t-input{color:var(--tui-text-03-night)}@supports (-webkit-marquee-repetition:infinite) and (object-fit:fill){:host._ios .t-input{padding-left:.8125rem}}.t-placeholder{transition-property:transform,font-size,color,letter-spacing;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;width:100%;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;color:var(--tui-text-02);pointer-events:none;will-change:transform;transform:translateY(0);position:absolute;left:0;max-width:100%;padding:0 1rem;border:solid transparent;border-width:0 var(--border-end,0) 0 var(--border-start,0);border-inline-start-width:var(--border-start,0);border-inline-end-width:var(--border-end,0);box-sizing:border-box}.t-placeholder_raised{transform:translateY(-.625rem)}:host[data-size='m'] .t-placeholder_raised{font:var(--tui-font-text-xs);transform:translateY(-.5rem);letter-spacing:.025rem}:host._invalid:not(._focused) .t-placeholder_raised,:host._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill)}:host[data-mode=onDark]._invalid:not(._focused) .t-placeholder_raised,:host[data-mode=onDark]._invalid:not(._focused):hover .t-placeholder_raised{color:var(--tui-error-fill-night)}:host._focused .t-placeholder,:host[data-size='l']._focused._label-outside .t-placeholder,:host[data-size='m']._focused._label-outside .t-placeholder{color:var(--tui-text-03)}:host[data-size='l'] .t-placeholder{font-size:.9375rem}:host[data-size='l'] .t-placeholder_raised{font-size:.8156rem}:host[data-size='l']._focused:not(._label-outside) .t-placeholder,:host[data-size='m']._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01)}:host[data-mode=onDark] .t-placeholder{color:var(--tui-text-02-night)}:host[data-size='l'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused:not(._label-outside) .t-placeholder{color:var(--tui-text-01-night)}:host[data-mode=onDark]._focused .t-placeholder,:host[data-size='l'][data-mode=onDark]._focused._label-outside .t-placeholder,:host[data-size='m'][data-mode=onDark]._focused._label-outside .t-placeholder{color:var(--tui-text-02-night)}@supports (-webkit-hyphens:none){.t-placeholder{will-change:unset;transition-property:transform,color,letter-spacing}}:host[data-size='m'] .t-placeholder{top:.75rem;padding:0 .75rem}:host[data-size='l'] .t-placeholder{top:1rem}:host._label-outside .t-placeholder{overflow:initial;height:auto;white-space:initial}.t-icons{position:absolute;top:0;left:0;bottom:0;right:0;display:flex;justify-content:flex-end;pointer-events:none;padding:1rem .75rem}:host[data-size='m'] .t-icons{padding:.75rem}.t-cleaner{transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:flex;width:1.5rem;height:1.5rem;align-items:center;justify-content:center;color:var(--tui-text-03);position:relative;box-sizing:border-box;cursor:pointer;transition-property:color,transform;pointer-events:auto}.t-cleaner:hover{color:var(--tui-text-02)}:host._disabled .t-cleaner,:host._readonly .t-cleaner{pointer-events:none}:host[data-mode=onDark] .t-cleaner{color:var(--tui-text-03-night)}:host[data-mode=onDark] .t-cleaner:hover{color:var(--tui-text-01-night)}.t-tooltip{pointer-events:auto}.t-counter{position:absolute;right:.75rem;bottom:.5rem;font:var(--tui-font-text-s);pointer-events:none;margin-top:auto;text-align:right;color:var(--tui-text-03)}:host[data-mode=onDark] .t-counter{color:var(--tui-text-03-night)}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core$1.TUI_TEXTFIELD_APPEARANCE)),
            __param(3, core.Inject(cdk.TUI_IS_IOS)),
            __param(4, core.Inject(core$1.TUI_MODE)),
            __param(5, core.Inject(core$1.TUI_TEXTFIELD_WATCHED_CONTROLLER)),
            __param(6, core.Inject(core$1.TUI_HINT_WATCHED_CONTROLLER))
        ], TuiTextAreaComponent);
        return TuiTextAreaComponent;
    }(cdk.AbstractTuiControl));

    var TuiTextAreaDirective = /** @class */ (function (_super) {
        __extends(TuiTextAreaDirective, _super);
        function TuiTextAreaDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TuiTextAreaDirective_1 = TuiTextAreaDirective;
        TuiTextAreaDirective.prototype.onValueChange = function (value) {
            this.host.onValueChange(value);
        };
        var TuiTextAreaDirective_1;
        TuiTextAreaDirective = TuiTextAreaDirective_1 = __decorate([
            core.Directive({
                selector: 'tui-text-area',
                providers: [
                    {
                        provide: core$1.TUI_TEXTFIELD_HOST,
                        useExisting: core.forwardRef(function () { return TuiTextAreaDirective_1; }),
                    },
                ],
            })
        ], TuiTextAreaDirective);
        return TuiTextAreaDirective;
    }(core$1.TuiAbstractTextfieldHost));

    var TuiTextAreaModule = /** @class */ (function () {
        function TuiTextAreaModule() {
        }
        TuiTextAreaModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiHoveredModule,
                    core$1.TuiScrollbarModule,
                    core$1.TuiTooltipModule,
                    core$1.TuiWrapperModule,
                    core$1.TuiDescribedByModule,
                    core$1.TuiSvgModule,
                    core$1.TuiPrimitiveTextfieldModule,
                ],
                declarations: [TuiTextAreaComponent, TuiTextAreaDirective],
                exports: [TuiTextAreaComponent, TuiTextAreaDirective, core$1.TuiTextfieldComponent],
            })
        ], TuiTextAreaModule);
        return TuiTextAreaModule;
    }());

    exports.DEFAULT_ROWS = DEFAULT_ROWS;
    exports.LINE_HEIGHT_L = LINE_HEIGHT_L;
    exports.LINE_HEIGHT_M = LINE_HEIGHT_M;
    exports.TUI_TEXT_AREA_PROVIDERS = TUI_TEXT_AREA_PROVIDERS;
    exports.TuiTextAreaComponent = TuiTextAreaComponent;
    exports.TuiTextAreaDirective = TuiTextAreaDirective;
    exports.TuiTextAreaModule = TuiTextAreaModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
