// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-portal.js
import { __decorate, __param } from 'tslib';
import { TemplateRef, Inject, Input, Directive, NgModule } from '@angular/core';
import { TuiPortalService } from '@taiga-ui/cdk/components/portal-host';

var TuiPortalDirective = /** @class */ (function () {
    function TuiPortalDirective(templateRef, portalService) {
        this.templateRef = templateRef;
        this.portalService = portalService;
    }
    Object.defineProperty(TuiPortalDirective.prototype, "tuiPortal", {
        set: function (show) {
            var _a;
            (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
            if (show) {
                this.viewRef = this.portalService.addTemplate(this.templateRef);
            }
        },
        enumerable: true,
        configurable: true
    });
    TuiPortalDirective.prototype.ngOnDestroy = function () {
        var _a;
        (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
    };
    TuiPortalDirective.ctorParameters = function () { return [
        { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }] },
        { type: TuiPortalService, decorators: [{ type: Inject, args: [TuiPortalService,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiPortalDirective.prototype, "tuiPortal", null);
    TuiPortalDirective = __decorate([
        Directive({
            selector: '[tuiPortal]',
        }),
        __param(0, Inject(TemplateRef)),
        __param(1, Inject(TuiPortalService))
    ], TuiPortalDirective);
    return TuiPortalDirective;
}());

var TuiPortalModule = /** @class */ (function () {
    function TuiPortalModule() {
    }
    TuiPortalModule = __decorate([
        NgModule({
            declarations: [TuiPortalDirective],
            exports: [TuiPortalDirective],
        })
    ], TuiPortalModule);
    return TuiPortalModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPortalDirective, TuiPortalModule };
