// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-pan.js
import { __decorate, __param } from 'tslib';
import { Inject, Directive, NgModule } from '@angular/core';
import { TuiPanService } from '@taiga-ui/cdk/services';
import { Observable } from 'rxjs';

// @dynamic
var TuiPanDirective = /** @class */ (function () {
    function TuiPanDirective(tuiPan) {
        this.tuiPan = tuiPan;
    }
    TuiPanDirective.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TuiPanService,] }] }
    ]; };
    TuiPanDirective = __decorate([
        Directive({
            selector: '[tuiPan]',
            outputs: ['tuiPan'],
            providers: [TuiPanService],
        }),
        __param(0, Inject(TuiPanService))
    ], TuiPanDirective);
    return TuiPanDirective;
}());

/**
 * @experimental
 */
var TuiPanModule = /** @class */ (function () {
    function TuiPanModule() {
    }
    TuiPanModule = __decorate([
        NgModule({
            declarations: [TuiPanDirective],
            exports: [TuiPanDirective],
        })
    ], TuiPanModule);
    return TuiPanModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPanDirective, TuiPanModule };
