// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-swipe.js
import { __decorate, __param } from 'tslib';
import { Inject, Directive, NgModule } from '@angular/core';
import { TuiSwipeService } from '@taiga-ui/cdk/services';
import { Observable } from 'rxjs';

// @dynamic
var TuiSwipeDirective = /** @class */ (function () {
    function TuiSwipeDirective(tuiSwipe) {
        this.tuiSwipe = tuiSwipe;
    }
    TuiSwipeDirective.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TuiSwipeService,] }] }
    ]; };
    TuiSwipeDirective = __decorate([
        Directive({
            selector: '[tuiSwipe]',
            outputs: ['tuiSwipe'],
            providers: [TuiSwipeService],
        }),
        __param(0, Inject(TuiSwipeService))
    ], TuiSwipeDirective);
    return TuiSwipeDirective;
}());

/**
 * @experimental
 */
var TuiSwipeModule = /** @class */ (function () {
    function TuiSwipeModule() {
    }
    TuiSwipeModule = __decorate([
        NgModule({
            declarations: [TuiSwipeDirective],
            exports: [TuiSwipeDirective],
        })
    ], TuiSwipeModule);
    return TuiSwipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiSwipeDirective, TuiSwipeModule };
