// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-autofilled.js
import { __decorate, __param } from 'tslib';
import { Component, ViewEncapsulation, EventEmitter, Inject, HostBinding, Output, HostListener, Directive, NgModule } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk/services';

var TuiAutofilledStyleComponent = /** @class */ (function () {
    function TuiAutofilledStyleComponent() {
    }
    TuiAutofilledStyleComponent = __decorate([
        Component({
            template: '',
            encapsulation: ViewEncapsulation.None,
            styles: [".tui-autofill input,input.tui-autofill{transition:box-shadow 10ms}"]
        })
    ], TuiAutofilledStyleComponent);
    return TuiAutofilledStyleComponent;
}());

var TuiAutofilledDirective = /** @class */ (function () {
    function TuiAutofilledDirective(directiveStyles) {
        this.autofilled = false;
        this.tuiAutofilledChange = new EventEmitter();
        directiveStyles.addComponent(TuiAutofilledStyleComponent);
    }
    TuiAutofilledDirective.prototype.transitionStartHandler = function (_a) {
        var propertyName = _a.propertyName, target = _a.target;
        var _b;
        var matchedAutofill = propertyName.includes('box-shadow') && ((_b = target) === null || _b === void 0 ? void 0 : _b.matches('input'));
        if (matchedAutofill) {
            this.autofilled = !this.autofilled;
            this.tuiAutofilledChange.emit(this.autofilled);
        }
    };
    TuiAutofilledDirective.ctorParameters = function () { return [
        { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
    ]; };
    __decorate([
        HostBinding('class._autofilled')
    ], TuiAutofilledDirective.prototype, "autofilled", void 0);
    __decorate([
        Output()
    ], TuiAutofilledDirective.prototype, "tuiAutofilledChange", void 0);
    __decorate([
        HostListener('transitionstart', ['$event'])
    ], TuiAutofilledDirective.prototype, "transitionStartHandler", null);
    TuiAutofilledDirective = __decorate([
        Directive({
            selector: '[tuiAutofilledChange]',
            host: { class: 'tui-autofill' },
        }),
        __param(0, Inject(TuiDirectiveStylesService))
    ], TuiAutofilledDirective);
    return TuiAutofilledDirective;
}());

var TuiAutofilledModule = /** @class */ (function () {
    function TuiAutofilledModule() {
    }
    TuiAutofilledModule = __decorate([
        NgModule({
            exports: [TuiAutofilledDirective],
            entryComponents: [TuiAutofilledStyleComponent],
            declarations: [TuiAutofilledDirective, TuiAutofilledStyleComponent],
        })
    ], TuiAutofilledModule);
    return TuiAutofilledModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAutofilledDirective, TuiAutofilledModule, TuiAutofilledStyleComponent };
