// @taiga-ui/kit/esm5/components/input-phone/input-phone.providers.js
import { DOCUMENT } from '@angular/common';
import { forwardRef, InjectionToken } from '@angular/core';
import { AbstractTuiControl, TUI_FOCUSABLE_ITEM_ACCESSOR, TuiDestroyService, } from '@taiga-ui/cdk';
import { TUI_DATA_LIST_HOST } from '@taiga-ui/core';
import { FIXED_DROPDOWN_CONTROLLER_PROVIDER } from '@taiga-ui/kit/providers';
import { fromEvent, merge } from 'rxjs';
import { flatMap, startWith, takeUntil, windowToggle } from 'rxjs/operators';
import { TuiInputPhoneComponent } from './input-phone.component';
export var SELECTION_STREAM = new InjectionToken("A stream of selection changes");
export var INPUT_PHONE_PROVIDERS = [
    TuiDestroyService,
    FIXED_DROPDOWN_CONTROLLER_PROVIDER,
    {
        provide: AbstractTuiControl,
        useExisting: forwardRef(function () { return TuiInputPhoneComponent; }),
    },
    {
        provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
        useExisting: forwardRef(function () { return TuiInputPhoneComponent; }),
    },
    {
        provide: TUI_DATA_LIST_HOST,
        useExisting: forwardRef(function () { return TuiInputPhoneComponent; }),
    },
    {
        provide: SELECTION_STREAM,
        deps: [TuiDestroyService, DOCUMENT],
        useFactory: selectionStreamFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
export function selectionStreamFactory(destroy$, documentRef) {
    return fromEvent(documentRef, "selectionchange").pipe(windowToggle(merge(fromEvent(documentRef, "mouseup"), fromEvent(documentRef, "keydown")), function () { return fromEvent(documentRef, "mousedown"); }), flatMap(function (windowed$) { return windowed$.pipe(startWith(null)); }), takeUntil(destroy$));
}
