// @taiga-ui/kit/esm5/components/input-phone/input-phone.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputPhoneDirective = /** @class */ (function (_super) {
    __extends(TuiInputPhoneDirective, _super);
    function TuiInputPhoneDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputPhoneDirective_1 = TuiInputPhoneDirective;
    Object.defineProperty(TuiInputPhoneDirective.prototype, "value", {
        get: function () {
            return this.host.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiInputPhoneDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    TuiInputPhoneDirective.prototype.process = function (input) {
        this.input = input;
    };
    TuiInputPhoneDirective.prototype.ngDoCheck = function () {
        if (!this.input) {
            return;
        }
        this.input.type = 'tel';
        this.input.inputMode = this.host.inputMode;
    };
    var TuiInputPhoneDirective_1;
    TuiInputPhoneDirective = TuiInputPhoneDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-phone',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputPhoneDirective_1; }),
                },
            ],
        })
    ], TuiInputPhoneDirective);
    return TuiInputPhoneDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputPhoneDirective };
