// @taiga-ui/core/esm5/pipes/calendar-sheet/calendar-sheet.pipe.js
import { __decorate, __param } from "tslib";
import { Inject, Pipe } from '@angular/core';
import { DAYS_IN_WEEK } from '@taiga-ui/cdk';
import { TUI_FIRST_DAY_OF_WEEK } from '@taiga-ui/core/tokens';
import { getDayFromMonthRowCol } from './utils';
var CALENDAR_ROWS_COUNT = 6;
var TuiCalendarSheetPipe = /** @class */ (function () {
    function TuiCalendarSheetPipe(firstDayOfWeek) {
        this.firstDayOfWeek = firstDayOfWeek;
        this.currentMonth = null;
        this.currentSheet = [];
    }
    TuiCalendarSheetPipe.prototype.transform = function (month, showAdjacentDays) {
        if (showAdjacentDays === void 0) { showAdjacentDays = false; }
        var _a;
        if ((_a = this.currentMonth) === null || _a === void 0 ? void 0 : _a.monthSame(month)) {
            return this.currentSheet;
        }
        var sheet = [];
        for (var rowIndex = 0; rowIndex < CALENDAR_ROWS_COUNT; rowIndex++) {
            var row = [];
            for (var colIndex = 0; colIndex < DAYS_IN_WEEK; colIndex++) {
                var day = getDayFromMonthRowCol({
                    month: month,
                    rowIndex: rowIndex,
                    colIndex: colIndex,
                    firstDayOfWeek: this.firstDayOfWeek,
                });
                var isPrevMonthDay = function (day, relativeToMonth) {
                    if (relativeToMonth === void 0) { relativeToMonth = month; }
                    return day.year < relativeToMonth.year || day.month < relativeToMonth.month;
                };
                var isNextMonthDay = function (day, relativeToMonth) {
                    if (relativeToMonth === void 0) { relativeToMonth = month; }
                    return day.year > relativeToMonth.year || day.month > relativeToMonth.month;
                };
                if (isPrevMonthDay(day) && !showAdjacentDays) {
                    continue;
                }
                if (isNextMonthDay(day) && !showAdjacentDays) {
                    break;
                }
                row.push(day);
            }
            sheet.push(row);
        }
        this.currentSheet = sheet.filter(function (row) { return row.length; });
        this.currentMonth = month;
        return this.currentSheet;
    };
    TuiCalendarSheetPipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_FIRST_DAY_OF_WEEK,] }] }
    ]; };
    TuiCalendarSheetPipe = __decorate([
        Pipe({
            name: "tuiCalendarSheet",
        }),
        __param(0, Inject(TUI_FIRST_DAY_OF_WEEK))
    ], TuiCalendarSheetPipe);
    return TuiCalendarSheetPipe;
}());
export { TuiCalendarSheetPipe };
