// @taiga-ui/kit/esm5/components/files/files.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, Inject, Input, TemplateRef, ViewEncapsulation, } from '@angular/core';
import { tuiDefaultProp, TuiItemDirective } from '@taiga-ui/cdk';
import { TUI_HIDE_TEXT, TUI_SHOW_ALL_TEXT } from '@taiga-ui/kit/tokens';
import { Observable } from 'rxjs';
var TuiFilesComponent = /** @class */ (function () {
    function TuiFilesComponent(hideText$, showAllText$) {
        this.hideText$ = hideText$;
        this.showAllText$ = showAllText$;
        this.items = null;
        this.max = 0;
        this.hidden = true;
    }
    Object.defineProperty(TuiFilesComponent.prototype, "hasExtraItems", {
        get: function () {
            var _a, _b;
            return !!this.max && ((_b = (_a = this.items) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) > this.max;
        },
        enumerable: true,
        configurable: true
    });
    TuiFilesComponent.prototype.toggle = function () {
        this.hidden = !this.hidden;
    };
    TuiFilesComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_HIDE_TEXT,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_SHOW_ALL_TEXT,] }] }
    ]; };
    __decorate([
        ContentChildren(TuiItemDirective, { read: TemplateRef })
    ], TuiFilesComponent.prototype, "items", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiFilesComponent.prototype, "max", void 0);
    TuiFilesComponent = __decorate([
        Component({
            selector: 'tui-files',
            template: "<ng-container *ngIf=\"items?.changes | async\"></ng-container>\n\n<section\n    tuiGroup\n    orientation=\"vertical\"\n    class=\"t-files\"\n    [collapsed]=\"true\"\n>\n    <ng-content></ng-content>\n    <ng-container *ngFor=\"let item of items; let index = index\">\n        <ng-container\n            *ngIf=\"!max || index < max\"\n            [ngTemplateOutlet]=\"item\"\n        ></ng-container>\n    </ng-container>\n    <tui-expand\n        *ngIf=\"hasExtraItems\"\n        [expanded]=\"!hidden\"\n    >\n        <div class=\"t-extra-items\">\n            <ng-container *ngFor=\"let item of items; let index = index\">\n                <ng-container\n                    *ngIf=\"max && index >= max\"\n                    [ngTemplateOutlet]=\"item\"\n                ></ng-container>\n            </ng-container>\n        </div>\n    </tui-expand>\n    <button\n        *ngIf=\"hasExtraItems\"\n        tuiButton\n        appearance=\"outline\"\n        size=\"m\"\n        class=\"t-button\"\n        [class.t-button_collapsed]=\"hidden\"\n        (click)=\"toggle()\"\n    >\n        {{ (hidden ? showAllText$ : hideText$) | async }}\n    </button>\n</section>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            encapsulation: ViewEncapsulation.None,
            styles: ["tui-files{display:block}tui-files .t-files{position:relative;display:block;width:100%;height:100%;border-radius:var(--tui-radius-m);overflow:hidden}tui-files .t-button{z-index:3;width:100%;background:var(--tui-base-01)}tui-files .t-button_collapsed{box-shadow:0 1.125rem 1.875rem rgba(0,0,0,.48);margin-top:-1.5rem}tui-files .t-extra-items tui-file{border-radius:0;margin-top:0;margin-bottom:-.0625rem}"]
        }),
        __param(0, Inject(TUI_HIDE_TEXT)),
        __param(1, Inject(TUI_SHOW_ALL_TEXT))
    ], TuiFilesComponent);
    return TuiFilesComponent;
}());
export { TuiFilesComponent };
