// @taiga-ui/cdk/esm5/abstract/portal-service.js
import { __decorate } from "tslib";
import { Injectable, } from '@angular/core';
import { TuiNoHostException } from '@taiga-ui/cdk/exceptions';
/**
 * Abstract service for displaying portals
 */
var AbstractTuiPortalService = /** @class */ (function () {
    function AbstractTuiPortalService() {
    }
    Object.defineProperty(AbstractTuiPortalService.prototype, "safeHost", {
        get: function () {
            if (!this.host) {
                throw new TuiNoHostException();
            }
            return this.host;
        },
        enumerable: true,
        configurable: true
    });
    AbstractTuiPortalService.prototype.attach = function (host) {
        this.host = host;
    };
    AbstractTuiPortalService.prototype.add = function (componentFactory, injector) {
        return this.safeHost.addComponentChild(componentFactory, injector);
    };
    AbstractTuiPortalService.prototype.remove = function (_a) {
        var hostView = _a.hostView;
        hostView.destroy();
    };
    AbstractTuiPortalService.prototype.addTemplate = function (templateRef, context) {
        return this.safeHost.addTemplateChild(templateRef, context);
    };
    AbstractTuiPortalService.prototype.removeTemplate = function (viewRef) {
        viewRef.destroy();
    };
    AbstractTuiPortalService = __decorate([
        Injectable()
    ], AbstractTuiPortalService);
    return AbstractTuiPortalService;
}());
export { AbstractTuiPortalService };
