// @taiga-ui/cdk/esm5/abstract/portal-host.js
import { __decorate, __param } from "tslib";
import { ComponentFactory, ComponentRef, Directive, ElementRef, EmbeddedViewRef, Inject, Injector, TemplateRef, ViewChild, ViewContainerRef, } from '@angular/core';
import { AbstractTuiPortalService } from './portal-service';
/**
 * Abstract class for host element for dynamically created portals.
 */
var AbstractTuiPortalHostComponent = /** @class */ (function () {
    function AbstractTuiPortalHostComponent(elementRef, portalService) {
        this.elementRef = elementRef;
        portalService.attach(this);
    }
    AbstractTuiPortalHostComponent_1 = AbstractTuiPortalHostComponent;
    Object.defineProperty(AbstractTuiPortalHostComponent.prototype, "clientRect", {
        get: function () {
            return this.elementRef.nativeElement.getBoundingClientRect();
        },
        enumerable: true,
        configurable: true
    });
    AbstractTuiPortalHostComponent.prototype.addComponentChild = function (componentFactory, injector) {
        return this.viewContainerRef.createComponent(componentFactory, undefined, Injector.create({
            parent: injector,
            providers: [
                {
                    provide: AbstractTuiPortalHostComponent_1,
                    useValue: this,
                },
            ],
        }));
    };
    AbstractTuiPortalHostComponent.prototype.addTemplateChild = function (templateRef, context) {
        return this.viewContainerRef.createEmbeddedView(templateRef, context);
    };
    var AbstractTuiPortalHostComponent_1;
    AbstractTuiPortalHostComponent.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: AbstractTuiPortalService, decorators: [{ type: Inject, args: [AbstractTuiPortalService,] }] }
    ]; };
    __decorate([
        ViewChild("viewContainer", { read: ViewContainerRef })
    ], AbstractTuiPortalHostComponent.prototype, "viewContainerRef", void 0);
    AbstractTuiPortalHostComponent = AbstractTuiPortalHostComponent_1 = __decorate([
        Directive(),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(AbstractTuiPortalService))
    ], AbstractTuiPortalHostComponent);
    return AbstractTuiPortalHostComponent;
}());
export { AbstractTuiPortalHostComponent };
