// @taiga-ui/cdk/esm5/abstract/theme-switcher.js
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, Inject } from '@angular/core';
/**
 * Use this abstract class to create your own toggleable themes.
 * A component extending this class must have CSS variables definitions
 * and have ViewEncapsulation set to NONE. A boolean input allows to
 * switch theme on or off.
 * @dynamic
 */
var AbstractTuiThemeSwitcher = /** @class */ (function () {
    function AbstractTuiThemeSwitcher(documentRef) {
        this.documentRef = documentRef;
        if (this.style !== null) {
            this.addTheme();
            return;
        }
        var styles = this.documentRef.head.querySelectorAll("style");
        this.constructor.style =
            styles[styles.length - 1];
    }
    AbstractTuiThemeSwitcher_1 = AbstractTuiThemeSwitcher;
    Object.defineProperty(AbstractTuiThemeSwitcher.prototype, "style", {
        get: function () {
            return this.constructor.style;
        },
        enumerable: true,
        configurable: true
    });
    AbstractTuiThemeSwitcher.prototype.ngOnDestroy = function () {
        this.removeTheme();
    };
    AbstractTuiThemeSwitcher.prototype.addTheme = function () {
        if (this.style) {
            this.documentRef.head.appendChild(this.style);
        }
    };
    AbstractTuiThemeSwitcher.prototype.removeTheme = function () {
        if (this.style && this.documentRef.head.contains(this.style)) {
            this.documentRef.head.removeChild(this.style);
        }
    };
    var AbstractTuiThemeSwitcher_1;
    AbstractTuiThemeSwitcher.style = null;
    AbstractTuiThemeSwitcher.ctorParameters = function () { return [
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    AbstractTuiThemeSwitcher = AbstractTuiThemeSwitcher_1 = __decorate([
        Directive(),
        __param(0, Inject(DOCUMENT))
    ], AbstractTuiThemeSwitcher);
    return AbstractTuiThemeSwitcher;
}());
export { AbstractTuiThemeSwitcher };
