// @taiga-ui/addon-doc/esm2015/components/navigation/navigation.providers.js
import { InjectionToken } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { isPresent, TuiDestroyService } from '@taiga-ui/cdk';
import { filter, map, mergeMap, takeUntil } from 'rxjs/operators';
import { TUI_DOC_PAGES } from '../../tokens/pages';
import { TUI_DOC_TITLE } from '../../tokens/title';
export const NAVIGATION_TITLE = new InjectionToken(`Page title`);
export const NAVIGATION_LABELS = new InjectionToken(`Navigation sections labels for search`);
export const NAVIGATION_ITEMS = new InjectionToken(`Navigation pages`);
export const NAVIGATION_PROVIDERS = [
    TuiDestroyService,
    {
        provide: NAVIGATION_TITLE,
        deps: [Router, ActivatedRoute, TUI_DOC_TITLE, TuiDestroyService],
        useFactory: titleProviderFactory,
    },
    {
        provide: NAVIGATION_LABELS,
        deps: [TUI_DOC_PAGES],
        useFactory: labelsProviderFactory,
    },
    {
        provide: NAVIGATION_ITEMS,
        deps: [TUI_DOC_PAGES],
        useFactory: itemsProviderFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
export function titleProviderFactory(router, activatedRoute, titlePrefix, destroy$) {
    return router.events.pipe(filter(event => event instanceof NavigationEnd), map(() => activatedRoute.firstChild), filter(isPresent), mergeMap(({ data }) => data), map(({ title }) => titlePrefix + title), takeUntil(destroy$));
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function labelsProviderFactory(pages) {
    return pages
        .map(({ section }) => section)
        .filter(isPresent)
        .filter((item, index, array) => array.indexOf(item) === index);
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function itemsProviderFactory(pages) {
    const labels = labelsProviderFactory(pages);
    return [
        ...labels.map(label => pages.filter(({ section }) => section === label)),
        pages.filter(page => !page.section),
    ];
}
