// @taiga-ui/cdk/esm2015/directives/autofilled/autofilled.directive.js
import { __decorate, __param } from "tslib";
import { Directive, EventEmitter, HostBinding, HostListener, Inject, Output, } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk/services';
import { TuiAutofilledStyleComponent } from './autofilled-style.component';
let TuiAutofilledDirective = class TuiAutofilledDirective {
    constructor(directiveStyles) {
        this.autofilled = false;
        this.tuiAutofilledChange = new EventEmitter();
        directiveStyles.addComponent(TuiAutofilledStyleComponent);
    }
    transitionStartHandler({ propertyName, target }) {
        var _a;
        const matchedAutofill = propertyName.includes('box-shadow') && ((_a = target) === null || _a === void 0 ? void 0 : _a.matches('input'));
        if (matchedAutofill) {
            this.autofilled = !this.autofilled;
            this.tuiAutofilledChange.emit(this.autofilled);
        }
    }
};
TuiAutofilledDirective.ctorParameters = () => [
    { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
];
__decorate([
    HostBinding('class._autofilled')
], TuiAutofilledDirective.prototype, "autofilled", void 0);
__decorate([
    Output()
], TuiAutofilledDirective.prototype, "tuiAutofilledChange", void 0);
__decorate([
    HostListener('transitionstart', ['$event'])
], TuiAutofilledDirective.prototype, "transitionStartHandler", null);
TuiAutofilledDirective = __decorate([
    Directive({
        selector: '[tuiAutofilledChange]',
        host: { class: 'tui-autofill' },
    }),
    __param(0, Inject(TuiDirectiveStylesService))
], TuiAutofilledDirective);
export { TuiAutofilledDirective };
