// @taiga-ui/core/esm2015/directives/manual-hint/manual-hint.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, Input, Optional } from '@angular/core';
import { TuiActiveZoneDirective, tuiDefaultProp, tuiRequiredSetter } from '@taiga-ui/cdk';
import { AbstractTuiHint } from '@taiga-ui/core/abstract';
import { TuiHintService } from '@taiga-ui/core/services';
import { TUI_MANUAL_HINT_OPTIONS } from './manual-hint-options';
let TuiManualHintDirective = class TuiManualHintDirective extends AbstractTuiHint {
    constructor(elementRef, hintService, activeZone, options) {
        super(elementRef, hintService, activeZone, options);
        this.options = options;
        this.content = '';
    }
    set tuiManualHintShow(show) {
        if (show) {
            this.showTooltip();
        }
        else {
            this.hideTooltip();
        }
    }
    getElementClientRect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
};
TuiManualHintDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] },
    { type: TuiActiveZoneDirective, decorators: [{ type: Optional }, { type: Inject, args: [TuiActiveZoneDirective,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_MANUAL_HINT_OPTIONS,] }] }
];
__decorate([
    Input('tuiManualHint'),
    tuiDefaultProp()
], TuiManualHintDirective.prototype, "content", void 0);
__decorate([
    Input(),
    tuiRequiredSetter()
], TuiManualHintDirective.prototype, "tuiManualHintShow", null);
TuiManualHintDirective = __decorate([
    Directive({
        selector: '[tuiManualHint]:not(ng-container)',
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(TuiHintService)),
    __param(2, Optional()),
    __param(2, Inject(TuiActiveZoneDirective)),
    __param(3, Inject(TUI_MANUAL_HINT_OPTIONS))
], TuiManualHintDirective);
export { TuiManualHintDirective };
