// @taiga-ui/core/esm2015/components/primitive-textfield/primitive-textfield.directive.js
var TuiPrimitiveTextfieldDirective_1;
import { __decorate, __param } from "tslib";
import { Directive, forwardRef, Inject } from '@angular/core';
import { TuiAbstractTextfieldHost } from '@taiga-ui/core/abstract';
import { TUI_TEXTFIELD_HOST } from '@taiga-ui/core/tokens';
import { TuiPrimitiveTextfieldComponent } from './primitive-textfield.component';
let TuiPrimitiveTextfieldDirective = TuiPrimitiveTextfieldDirective_1 = class TuiPrimitiveTextfieldDirective extends TuiAbstractTextfieldHost {
    constructor(textfield) {
        /**
         * TuiPrimitiveTextfieldComponent satisfies all required parts of
         * AbstractTuiControl interface. A new interface is not introduced
         * so AbstractTuiControl is automatically injected for all other controls.
         */
        super(textfield);
        this.textfield = textfield;
    }
    get readOnly() {
        return this.textfield.readOnly || !this.textfield.editable;
    }
    onValueChange(value) {
        this.textfield.onModelChange(value);
    }
};
TuiPrimitiveTextfieldDirective.ctorParameters = () => [
    { type: TuiPrimitiveTextfieldComponent, decorators: [{ type: Inject, args: [TuiPrimitiveTextfieldComponent,] }] }
];
TuiPrimitiveTextfieldDirective = TuiPrimitiveTextfieldDirective_1 = __decorate([
    Directive({
        selector: 'tui-primitive-textfield',
        providers: [
            {
                provide: TUI_TEXTFIELD_HOST,
                useExisting: forwardRef(() => TuiPrimitiveTextfieldDirective_1),
            },
        ],
    }),
    __param(0, Inject(TuiPrimitiveTextfieldComponent))
], TuiPrimitiveTextfieldDirective);
export { TuiPrimitiveTextfieldDirective };
