// @taiga-ui/kit/esm5/components/input-inline/input-inline.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, forwardRef, Inject, Input, Optional, Self, ViewChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiControl, isNativeFocused, TUI_FOCUSABLE_ITEM_ACCESSOR, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TUI_VALUE_ACCESSOR_PROVIDER } from '@taiga-ui/kit/providers';
var TuiInputInlineComponent = /** @class */ (function (_super) {
    __extends(TuiInputInlineComponent, _super);
    function TuiInputInlineComponent(control, changeDetectorRef) {
        var _this = _super.call(this, control, changeDetectorRef) || this;
        _this.maxLength = null;
        return _this;
    }
    TuiInputInlineComponent_1 = TuiInputInlineComponent;
    Object.defineProperty(TuiInputInlineComponent.prototype, "nativeFocusableElement", {
        get: function () {
            return !this.native ? null : this.native.nativeElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiInputInlineComponent.prototype, "focused", {
        get: function () {
            return isNativeFocused(this.nativeFocusableElement);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiInputInlineComponent.prototype, "hasValue", {
        get: function () {
            return this.value !== '';
        },
        enumerable: true,
        configurable: true
    });
    TuiInputInlineComponent.prototype.onValueChange = function (value) {
        this.updateValue(value);
    };
    TuiInputInlineComponent.prototype.onFocused = function (focused) {
        this.updateFocused(focused);
    };
    TuiInputInlineComponent.prototype.getFallbackValue = function () {
        return '';
    };
    var TuiInputInlineComponent_1;
    TuiInputInlineComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] }
    ]; };
    __decorate([
        ViewChild('native')
    ], TuiInputInlineComponent.prototype, "native", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiInputInlineComponent.prototype, "maxLength", void 0);
    TuiInputInlineComponent = TuiInputInlineComponent_1 = __decorate([
        Component({
            selector: 'tui-input-inline',
            template: "<span class=\"t-before\">{{ value }}</span>\n\n<span\n    *ngIf=\"!hasValue\"\n    automation-id=\"tui-input-inline__placeholder\"\n    class=\"t-placeholder\"\n>\n    <ng-content></ng-content>\n</span>\n\n<input\n    #native\n    automation-id=\"tui-input-inline__native\"\n    type=\"text\"\n    tuiMaskAccessor\n    class=\"t-native\"\n    [attr.maxLength]=\"maxLength\"\n    [id]=\"id\"\n    [disabled]=\"computedDisabled\"\n    [readOnly]=\"readOnly\"\n    [tuiFocusable]=\"focusable\"\n    [ngModel]=\"value\"\n    (ngModelChange)=\"onValueChange($event)\"\n    (tuiFocusedChange)=\"onFocused($event)\"\n/>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                TUI_VALUE_ACCESSOR_PROVIDER,
                {
                    provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
                    useExisting: forwardRef(function () { return TuiInputInlineComponent_1; }),
                },
            ],
            styles: [":host{position:relative;display:inline-block;overflow:hidden;white-space:nowrap;box-sizing:border-box}.t-before{padding-right:.02em;margin-left:1px;visibility:hidden;white-space:pre}.t-placeholder{margin-left:-1px}.t-native{position:absolute;top:0;left:0;background-color:transparent;padding:inherit;font:inherit;color:inherit;box-sizing:border-box;width:100%;height:100%;border-width:0;text-align:inherit;letter-spacing:inherit;text-indent:inherit;text-transform:inherit;outline:0}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(ChangeDetectorRef))
    ], TuiInputInlineComponent);
    return TuiInputInlineComponent;
}(AbstractTuiControl));
export { TuiInputInlineComponent };
