// @taiga-ui/addon-doc/esm5/components/page/page.component.js
import { __decorate, __param } from "tslib";
import { Attribute, ChangeDetectionStrategy, Component, ContentChildren, Inject, Input, } from '@angular/core';
import { EMPTY_QUERY } from '@taiga-ui/cdk';
import { TUI_DOC_DEFAULT_TABS } from '../../tokens/default-tabs';
import { PAGE_PROVIDERS, PAGE_SEE_ALSO } from './page.providers';
import { TuiDocPageTabConnectorDirective } from './page-tab.directive';
// Ambient type cannot be used without dynamic https://github.com/angular/angular/issues/23395
// @dynamic
var TuiDocPageComponent = /** @class */ (function () {
    function TuiDocPageComponent(deprecated, defaultTabs, seeAlso) {
        this.deprecated = deprecated;
        this.defaultTabs = defaultTabs;
        this.seeAlso = seeAlso;
        this.header = '';
        this.package = '';
        this.type = '';
        this.path = '';
        this.tabConnectors = EMPTY_QUERY;
        this.activeItemIndex = NaN;
    }
    Object.defineProperty(TuiDocPageComponent.prototype, "showSeeAlso", {
        get: function () {
            return !!this.seeAlso.length && this.activeItemIndex === 0;
        },
        enumerable: true,
        configurable: true
    });
    TuiDocPageComponent.prototype.getRouterLink = function (tab) {
        if (tab === void 0) { tab = ''; }
        return "./" + tab.replace(/ /g, '_');
    };
    TuiDocPageComponent.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Attribute, args: ['deprecated',] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_DOC_DEFAULT_TABS,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [PAGE_SEE_ALSO,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiDocPageComponent.prototype, "header", void 0);
    __decorate([
        Input()
    ], TuiDocPageComponent.prototype, "package", void 0);
    __decorate([
        Input()
    ], TuiDocPageComponent.prototype, "type", void 0);
    __decorate([
        Input()
    ], TuiDocPageComponent.prototype, "path", void 0);
    __decorate([
        ContentChildren(TuiDocPageTabConnectorDirective)
    ], TuiDocPageComponent.prototype, "tabConnectors", void 0);
    TuiDocPageComponent = __decorate([
        Component({
            selector: 'tui-doc-page',
            template: "<header class=\"t-header\">\n    <h1 class=\"t-title\">\n        {{ header }}\n        <tui-tag\n            *ngIf=\"deprecated !== null\"\n            value=\"deprecated\"\n            status=\"custom\"\n            class=\"t-tag t-tag_deprecated\"\n        ></tui-tag>\n        <tui-tag\n            *ngIf=\"package\"\n            status=\"custom\"\n            class=\"t-tag t-tag_package\"\n            [value]=\"package\"\n            [autoColor]=\"true\"\n        ></tui-tag>\n    </h1>\n    <tui-tabs-with-more\n        *ngIf=\"tabConnectors.length\"\n        class=\"t-tabs\"\n        [(activeItemIndex)]=\"activeItemIndex\"\n    >\n        <ng-container *ngFor=\"let tab of tabConnectors; first as first; index as index\">\n            <ng-container *ngIf=\"first; else dynamicTab\">\n                <a\n                    *tuiTab\n                    tuiTab\n                    routerLink=\"./\"\n                    routerLinkActive\n                    [routerLinkActiveOptions]=\"{exact: true}\"\n                >\n                    {{ tab.pageTab || defaultTabs[index] }}\n                </a>\n            </ng-container>\n            <ng-template #dynamicTab>\n                <a\n                    *tuiTab\n                    tuiTab\n                    routerLinkActive\n                    [routerLink]=\"getRouterLink(tab.pageTab || defaultTabs[index])\"\n                >\n                    {{ tab.pageTab || defaultTabs[index] }}\n                </a>\n            </ng-template>\n        </ng-container>\n    </tui-tabs-with-more>\n    <tui-doc-source-code\n        class=\"t-source-code\"\n        [header]=\"header\"\n        [package]=\"package\"\n        [type]=\"type\"\n        [path]=\"path\"\n    ></tui-doc-source-code>\n</header>\n<div class=\"t-content\">\n    <ng-content></ng-content>\n    <tui-doc-see-also\n        *ngIf=\"showSeeAlso\"\n        class=\"t-see-also\"\n        [seeAlso]=\"seeAlso\"\n    ></tui-doc-see-also>\n    <ng-container *ngFor=\"let tab of tabConnectors; index as index\">\n        <ng-container\n            *ngIf=\"index === activeItemIndex\"\n            [ngTemplateOutlet]=\"tab.template\"\n        ></ng-container>\n    </ng-container>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: PAGE_PROVIDERS,
            styles: [":host{display:flex;flex-direction:column;min-height:0;flex-basis:0;flex-grow:1}.t-header{display:flex;font:var(--tui-font-heading-3);flex-wrap:wrap;min-height:7.25rem;color:var(--tui-text-01);box-shadow:inset 0 -1px 0 0 var(--tui-base-03);padding:3.75rem 0 0;box-sizing:border-box;flex-shrink:0;margin:0 min(10vw,8.75rem)}:host-context(tui-root._mobile) .t-header{font:var(--tui-font-heading-4);min-height:4.5rem;padding:1.25rem 1.25rem 0;margin:0}.t-title{min-width:100%;font-size:inherit;margin:0}.t-tag{vertical-align:middle;text-transform:uppercase;margin-right:.5rem}.t-tag_deprecated{background-color:var(--tui-error-fill)}.t-tag.t-tag_package{color:#000}.t-tabs{flex:1;margin:1.125rem .3125rem 0 0}:host-context(tui-root._mobile) .t-tabs{margin-top:.25rem}.t-content{padding:2rem 0;margin:0 min(10vw,8.75rem)}:host-context(tui-root._mobile) .t-content{padding:2rem 1.25rem;margin:0}.t-see-also{min-width:18.75rem;width:30%;float:right;margin-left:1.5rem}:host-context(tui-root._mobile) .t-see-also{float:none;width:100%;margin:0 0 1.5rem}.t-source-code{align-self:flex-end;line-height:2.75rem}"]
        }),
        __param(0, Attribute('deprecated')),
        __param(1, Inject(TUI_DOC_DEFAULT_TABS)),
        __param(2, Inject(PAGE_SEE_ALSO))
    ], TuiDocPageComponent);
    return TuiDocPageComponent;
}());
export { TuiDocPageComponent };
