// @taiga-ui/kit/esm2015/components/breadcrumbs/breadcrumbs-wrapper.component.js
var TuiBreadcrumbsWrapperComponent_1;
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, forwardRef, HostBinding, Input, TemplateRef, } from '@angular/core';
import { EMPTY_QUERY, tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiModeDirective } from '@taiga-ui/core';
import { Subject } from 'rxjs';
import { TuiBreadcrumbDirective } from './breadcrumb.directive';
let TuiBreadcrumbsWrapperComponent = TuiBreadcrumbsWrapperComponent_1 = class TuiBreadcrumbsWrapperComponent {
    constructor() {
        this.size = 'm';
        this.items = EMPTY_QUERY;
        this.change$ = new Subject();
        this.mode = 'onLight';
    }
    ngOnChanges() { }
};
__decorate([
    Input(),
    HostBinding('attr.data-size'),
    tuiDefaultProp()
], TuiBreadcrumbsWrapperComponent.prototype, "size", void 0);
__decorate([
    ContentChildren(TuiBreadcrumbDirective, { read: TemplateRef })
], TuiBreadcrumbsWrapperComponent.prototype, "items", void 0);
TuiBreadcrumbsWrapperComponent = TuiBreadcrumbsWrapperComponent_1 = __decorate([
    Component({
        selector: 'tui-breadcrumbs:not([items])',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "<ng-container *ngIf=\"items.changes | async\"></ng-container>\n<ng-container *ngFor=\"let item of items; let last = last\">\n    <ng-container [ngTemplateOutlet]=\"item\"></ng-container>\n    <tui-svg\n        *ngIf=\"!last\"\n        src=\"tuiIconChevronRight\"\n        class=\"t-icon\"\n    ></tui-svg>\n</ng-container>\n",
        providers: [
            {
                provide: TuiModeDirective,
                useExisting: forwardRef(() => TuiBreadcrumbsWrapperComponent_1),
            },
        ],
        styles: [":host{display:flex;align-items:center;height:2.5rem;font:var(--tui-font-text-s)}:host[data-size='l']{font:var(--tui-font-text-m)}.t-icon{margin:0 .25rem;color:var(--tui-text-03)}"]
    })
], TuiBreadcrumbsWrapperComponent);
export { TuiBreadcrumbsWrapperComponent };
