// @taiga-ui/kit/esm2015/components/carousel/carousel.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { PAGE_VISIBILITY } from '@ng-web-apis/common';
import { typedFromEvent } from '@taiga-ui/cdk';
import { BehaviorSubject, combineLatest, EMPTY, interval, merge, Observable } from 'rxjs';
import { mapTo, switchMap } from 'rxjs/operators';
let TuiCarouselDirective = class TuiCarouselDirective extends Observable {
    constructor(elementRef, visible$) {
        super(subscriber => this.output$.subscribe(subscriber));
        this.elementRef = elementRef;
        this.visible$ = visible$;
        this.duration$ = new BehaviorSubject(0);
        this.running$ = merge(typedFromEvent(this.elementRef.nativeElement, 'mouseenter').pipe(mapTo(false)), typedFromEvent(this.elementRef.nativeElement, 'touchstart').pipe(mapTo(false)), typedFromEvent(this.elementRef.nativeElement, 'touchend').pipe(mapTo(true)), typedFromEvent(this.elementRef.nativeElement, 'mouseleave').pipe(mapTo(true)), this.visible$);
        this.output$ = combineLatest([this.duration$, this.running$]).pipe(switchMap(([duration, running]) => duration && running ? interval(duration) : EMPTY));
    }
    set duration(duration) {
        this.duration$.next(duration);
    }
    set index(_) {
        this.duration$.next(this.duration$.value);
    }
};
TuiCarouselDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [PAGE_VISIBILITY,] }] }
];
__decorate([
    Input()
], TuiCarouselDirective.prototype, "duration", null);
__decorate([
    Input()
], TuiCarouselDirective.prototype, "index", null);
TuiCarouselDirective = __decorate([
    Directive({
        selector: 'tui-carousel',
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(PAGE_VISIBILITY))
], TuiCarouselDirective);
export { TuiCarouselDirective };
