// @taiga-ui/kit/esm2015/components/data-list-wrapper/data-list-wrapper.component.js
var TuiDataListWrapperComponent_1;
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, forwardRef, Inject, Input, } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_DATA_LIST_ACCESSOR } from '@taiga-ui/core';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import { AbstractTuiDataListWrapper } from './data-list-wrapper';
let TuiDataListWrapperComponent = TuiDataListWrapperComponent_1 = class TuiDataListWrapperComponent extends AbstractTuiDataListWrapper {
    constructor(itemsHandlers) {
        super(itemsHandlers);
        this.items = [];
    }
};
TuiDataListWrapperComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_ITEMS_HANDLERS,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiDataListWrapperComponent.prototype, "items", void 0);
TuiDataListWrapperComponent = TuiDataListWrapperComponent_1 = __decorate([
    Component({
        selector: 'tui-data-list-wrapper:not([labels])',
        template: "<tui-data-list\n    *ngIf=\"items; else loading\"\n    [emptyContent]=\"emptyContent\"\n>\n    <button\n        *ngFor=\"let item of items\"\n        #elementRef=\"elementRef\"\n        tuiElement\n        tuiOption\n        automation-id=\"tui-data-list-wrapper__option\"\n        [size]=\"size\"\n        [value]=\"item\"\n        [disabled]=\"disabledItemHandler(item)\"\n    >\n        <span\n            polymorpheus-outlet\n            class=\"t-content\"\n            [content]=\"itemContent\"\n            [context]=\"getContext(item, elementRef)\"\n        ></span>\n    </button>\n</tui-data-list>\n<ng-template #loading>\n    <tui-loader\n        automation-id=\"tui-data-list-wrapper__loader\"\n        class=\"t-loader\"\n    ></tui-loader>\n</ng-template>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [
            {
                provide: TUI_DATA_LIST_ACCESSOR,
                useExisting: forwardRef(() => TuiDataListWrapperComponent_1),
            },
        ],
        styles: [":host{display:block}.t-content{flex:1}.t-loader{margin:.75rem 0}"]
    }),
    __param(0, Inject(TUI_ITEMS_HANDLERS))
], TuiDataListWrapperComponent);
export { TuiDataListWrapperComponent };
