// @taiga-ui/kit/esm2015/components/data-list-wrapper/data-list-wrapper.js
import { __decorate } from "tslib";
import { Directive, forwardRef, Input, ViewChildren, } from '@angular/core';
import { EMPTY_QUERY, isNativeFocused, isPresent, tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiOptionComponent, } from '@taiga-ui/core';
let AbstractTuiDataListWrapper = class AbstractTuiDataListWrapper {
    constructor(itemsHandlers) {
        this.itemsHandlers = itemsHandlers;
        this.optionsQuery = EMPTY_QUERY;
        this.disabledItemHandler = this.itemsHandlers.disabledItemHandler;
        this.emptyContent = ``;
        this.size = `m`;
        this.itemContent = ({ $implicit }) => this.itemsHandlers.stringify($implicit);
    }
    getContext($implicit, { nativeElement }) {
        return { $implicit, active: isNativeFocused(nativeElement) };
    }
    getOptions(includeDisabled = false) {
        return this.optionsQuery
            .toArray()
            .filter(({ disabled }) => includeDisabled || !disabled)
            .map(({ value }) => value)
            .filter(isPresent);
    }
};
AbstractTuiDataListWrapper.ctorParameters = () => [
    { type: undefined }
];
__decorate([
    ViewChildren(forwardRef(() => TuiOptionComponent))
], AbstractTuiDataListWrapper.prototype, "optionsQuery", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiDataListWrapper.prototype, "disabledItemHandler", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiDataListWrapper.prototype, "emptyContent", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiDataListWrapper.prototype, "size", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], AbstractTuiDataListWrapper.prototype, "itemContent", void 0);
AbstractTuiDataListWrapper = __decorate([
    Directive()
], AbstractTuiDataListWrapper);
export { AbstractTuiDataListWrapper };
