// @taiga-ui/kit/esm2015/components/input-month-range/input-month-range.directive.js
var TuiInputMonthRangeDirective_1;
import { __decorate, __param } from "tslib";
import { Directive, forwardRef, Inject, Self } from '@angular/core';
import { AbstractTuiControl, TuiDestroyService, } from '@taiga-ui/cdk';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
import { TUI_MONTH_FORMATTER } from '@taiga-ui/kit/tokens';
import { combineLatest, Observable, Subject } from 'rxjs';
import { distinctUntilChanged, switchMap, takeUntil } from 'rxjs/operators';
import { TuiInputMonthRangeComponent } from './input-month-range.component';
let TuiInputMonthRangeDirective = TuiInputMonthRangeDirective_1 = class TuiInputMonthRangeDirective extends TuiAbstractTextfieldHost {
    constructor(host, formatter, destroy$) {
        super(host);
        this.value$ = new Subject();
        this.localizedValue = ['', ''];
        this.value$
            .pipe(distinctUntilChanged(), switchMap((value) => combineLatest([
            formatter((value === null || value === void 0 ? void 0 : value.from) || null),
            formatter((value === null || value === void 0 ? void 0 : value.to) || null),
        ])), takeUntil(destroy$))
            .subscribe(localizedValue => {
            this.localizedValue = localizedValue;
        });
    }
    get readOnly() {
        return true;
    }
    get value() {
        return this.localizedValue[0]
            ? this.host.computeValue(...this.localizedValue)
            : '';
    }
    ngDoCheck() {
        this.value$.next(this.host.value);
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
};
TuiInputMonthRangeDirective.ctorParameters = () => [
    { type: TuiInputMonthRangeComponent, decorators: [{ type: Inject, args: [AbstractTuiControl,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_MONTH_FORMATTER,] }] },
    { type: Observable, decorators: [{ type: Self }, { type: Inject, args: [TuiDestroyService,] }] }
];
TuiInputMonthRangeDirective = TuiInputMonthRangeDirective_1 = __decorate([
    Directive({
        selector: 'tui-input-month-range',
        providers: [
            {
                provide: TUI_TEXTFIELD_HOST,
                useExisting: forwardRef(() => TuiInputMonthRangeDirective_1),
            },
            TuiDestroyService,
        ],
    }),
    __param(0, Inject(AbstractTuiControl)),
    __param(1, Inject(TUI_MONTH_FORMATTER)),
    __param(2, Self()), __param(2, Inject(TuiDestroyService))
], TuiInputMonthRangeDirective);
export { TuiInputMonthRangeDirective };
