// @taiga-ui/kit/esm5/components/select/select.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, Inject, Input, Optional, Self, TemplateRef, ViewChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiNullableControl, isNativeFocused, setNativeFocused, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TUI_TEXTFIELD_CLEANER, TuiDataListDirective, TuiDataListHost, TuiHostedDropdownComponent, TuiPrimitiveTextfieldComponent, TuiSizeL, TuiSizeM, TuiSizeS, TuiTextfieldCleanerDirective, TuiValueContentContext, } from '@taiga-ui/core';
import { TUI_ARROW_MODE } from '@taiga-ui/kit/components/arrow';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import { TUI_SELECT_PROVIDERS } from './select.providers';
import { TUI_SELECT_OPTIONS } from './select-options';
var TuiSelectComponent = /** @class */ (function (_super) {
    __extends(TuiSelectComponent, _super);
    function TuiSelectComponent(control, changeDetectorRef, textfieldCleaner, arrowMode, itemsHandlers, options) {
        var _this = _super.call(this, control, changeDetectorRef) || this;
        _this.textfieldCleaner = textfieldCleaner;
        _this.arrowMode = arrowMode;
        _this.itemsHandlers = itemsHandlers;
        _this.options = options;
        _this.stringify = _this.itemsHandlers.stringify;
        _this.identityMatcher = _this.itemsHandlers.identityMatcher;
        _this.valueContent = _this.options.valueContent;
        _this.datalist = '';
        return _this;
    }
    Object.defineProperty(TuiSelectComponent.prototype, "arrow", {
        get: function () {
            return !this.interactive ? this.arrowMode.disabled : this.arrowMode.interactive;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSelectComponent.prototype, "nativeFocusableElement", {
        get: function () {
            return this.textfield ? this.textfield.nativeFocusableElement : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSelectComponent.prototype, "focused", {
        get: function () {
            return (isNativeFocused(this.nativeFocusableElement) ||
                (!!this.hostedDropdown && this.hostedDropdown.focused));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSelectComponent.prototype, "computedValue", {
        get: function () {
            return this.value === null ? '' : this.stringify(this.value) || ' ';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSelectComponent.prototype, "computedContent", {
        get: function () {
            return this.valueContent || this.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiSelectComponent.prototype.onValueChange = function (value) {
        if (!value) {
            this.updateValue(null);
        }
    };
    TuiSelectComponent.prototype.onActiveZone = function (active) {
        this.updateFocused(active);
    };
    TuiSelectComponent.prototype.onHovered = function (hovered) {
        this.updateHovered(hovered);
    };
    TuiSelectComponent.prototype.onKeyDownDelete = function () {
        if (this.textfieldCleaner.cleaner) {
            this.updateValue(null);
        }
    };
    TuiSelectComponent.prototype.handleOption = function (option) {
        var _a;
        this.focusInput();
        this.updateValue(option);
        (_a = this.hostedDropdown) === null || _a === void 0 ? void 0 : _a.updateOpen(false);
    };
    TuiSelectComponent.prototype.focusInput = function (preventScroll) {
        if (preventScroll === void 0) { preventScroll = false; }
        if (this.nativeFocusableElement) {
            setNativeFocused(this.nativeFocusableElement, true, preventScroll);
        }
    };
    TuiSelectComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
        { type: TuiTextfieldCleanerDirective, decorators: [{ type: Inject, args: [TUI_TEXTFIELD_CLEANER,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ARROW_MODE,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ITEMS_HANDLERS,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_SELECT_OPTIONS,] }] }
    ]; };
    __decorate([
        ViewChild(TuiPrimitiveTextfieldComponent)
    ], TuiSelectComponent.prototype, "textfield", void 0);
    __decorate([
        ViewChild(TuiHostedDropdownComponent)
    ], TuiSelectComponent.prototype, "hostedDropdown", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiSelectComponent.prototype, "stringify", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiSelectComponent.prototype, "identityMatcher", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiSelectComponent.prototype, "valueContent", void 0);
    __decorate([
        ContentChild(TuiDataListDirective, { read: TemplateRef })
    ], TuiSelectComponent.prototype, "datalist", void 0);
    TuiSelectComponent = __decorate([
        Component({
            selector: 'tui-select',
            template: "<tui-hosted-dropdown\n    class=\"t-hosted\"\n    [canOpen]=\"interactive\"\n    [content]=\"datalist || ''\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-primitive-textfield\n        automation-id=\"tui-select__textfield\"\n        class=\"t-textfield\"\n        [editable]=\"false\"\n        [pseudoFocused]=\"computedFocused\"\n        [pseudoHovered]=\"pseudoHovered\"\n        [pseudoPressed]=\"pseudoPressed\"\n        [invalid]=\"computedInvalid\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [iconContent]=\"arrow\"\n        [disabled]=\"computedDisabled\"\n        [focusable]=\"computedFocusable\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($event)\"\n        (hoveredChange)=\"onHovered($event)\"\n        (keydown.delete.prevent)=\"onKeyDownDelete()\"\n        (keydown.backspace.prevent)=\"onKeyDownDelete()\"\n    >\n        <ng-content></ng-content>\n        <ng-content\n            select=\"input\"\n            ngProjectAs=\"input\"\n        ></ng-content>\n        <div\n            *ngIf=\"value !== null\"\n            polymorpheus-outlet\n            automation-id=\"tui-select__value\"\n            class=\"t-value\"\n            [content]=\"computedContent\"\n            [context]=\"{$implicit: value, active: computedFocused}\"\n        >\n            <ng-template let-primitiveValue>\n                <div class=\"t-primitive-value\">{{ primitiveValue }}</div>\n            </ng-template>\n        </div>\n    </tui-primitive-textfield>\n</tui-hosted-dropdown>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: TUI_SELECT_PROVIDERS,
            styles: [":host{display:block;border-radius:var(--tui-radius-m);text-align:left}:host:not(._readonly) ::ng-deep input:not(:disabled){cursor:pointer}:host._readonly ::ng-deep input{cursor:default}.t-hosted{display:block;border-radius:inherit}.t-textfield{border-radius:inherit;text-align:inherit}.t-value{display:flex;width:100%;align-items:center}.t-primitive-value{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(ChangeDetectorRef)),
        __param(2, Inject(TUI_TEXTFIELD_CLEANER)),
        __param(3, Inject(TUI_ARROW_MODE)),
        __param(4, Inject(TUI_ITEMS_HANDLERS)),
        __param(5, Inject(TUI_SELECT_OPTIONS))
    ], TuiSelectComponent);
    return TuiSelectComponent;
}(AbstractTuiNullableControl));
export { TuiSelectComponent };
