// @taiga-ui/kit/esm5/components/select/select.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiSelectDirective = /** @class */ (function (_super) {
    __extends(TuiSelectDirective, _super);
    function TuiSelectDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiSelectDirective_1 = TuiSelectDirective;
    Object.defineProperty(TuiSelectDirective.prototype, "readOnly", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiSelectDirective.prototype, "value", {
        get: function () {
            return this.host.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiSelectDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiSelectDirective_1;
    TuiSelectDirective = TuiSelectDirective_1 = __decorate([
        Directive({
            selector: 'tui-select',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiSelectDirective_1; }),
                },
            ],
        })
    ], TuiSelectDirective);
    return TuiSelectDirective;
}(TuiAbstractTextfieldHost));
export { TuiSelectDirective };
