// @taiga-ui/cdk/esm5/enums/autocomplete.js
/* eslint-disable @typescript-eslint/naming-convention */
export var TuiCreditCardAutofillName;
(function (TuiCreditCardAutofillName) {
    TuiCreditCardAutofillName["Off"] = "off";
    TuiCreditCardAutofillName["CcName"] = "cc-name";
    TuiCreditCardAutofillName["CcNumber"] = "cc-number";
    TuiCreditCardAutofillName["CcCsc"] = "cc-csc";
    TuiCreditCardAutofillName["CcExpMonth"] = "cc-exp-month";
    /**
     * @deprecated
     * TODO: 3.0 replace with {@link CcExpMonth}
     */
    TuiCreditCardAutofillName["CcExp_mounth"] = "cc-exp-month";
    TuiCreditCardAutofillName["CcExpYear"] = "cc-exp-year";
    /**
     * @deprecated
     * TODO: 3.0 replace with {@link CcExpYear}
     */
    TuiCreditCardAutofillName["CcExp_year"] = "cc-exp-year";
    TuiCreditCardAutofillName["CcExp"] = "cc-exp";
    TuiCreditCardAutofillName["CcType"] = "cc-type";
})(TuiCreditCardAutofillName || (TuiCreditCardAutofillName = {}));
export var TuiNameAutofillName;
(function (TuiNameAutofillName) {
    TuiNameAutofillName["Off"] = "off";
    TuiNameAutofillName["Name"] = "name";
    TuiNameAutofillName["GivenName"] = "given-name";
    TuiNameAutofillName["AdditionalName"] = "additional-name";
    TuiNameAutofillName["FamilyName"] = "family-name";
})(TuiNameAutofillName || (TuiNameAutofillName = {}));
export var TuiAccountAutofillName;
(function (TuiAccountAutofillName) {
    TuiAccountAutofillName["Off"] = "off";
    TuiAccountAutofillName["Username"] = "username";
    TuiAccountAutofillName["NewPassword"] = "new-password";
    TuiAccountAutofillName["CurrentPassword"] = "current-password";
})(TuiAccountAutofillName || (TuiAccountAutofillName = {}));
export var TuiEmailAutofillName;
(function (TuiEmailAutofillName) {
    TuiEmailAutofillName["Off"] = "off";
    TuiEmailAutofillName["Email"] = "email";
})(TuiEmailAutofillName || (TuiEmailAutofillName = {}));
export var TuiAddressAutofillName;
(function (TuiAddressAutofillName) {
    TuiAddressAutofillName["Off"] = "off";
    TuiAddressAutofillName["StreetAddress"] = "street-address";
    TuiAddressAutofillName["PostalCode"] = "postal-code";
    TuiAddressAutofillName["CountryName"] = "country-name";
})(TuiAddressAutofillName || (TuiAddressAutofillName = {}));
export var TuiPhoneAutofillName;
(function (TuiPhoneAutofillName) {
    TuiPhoneAutofillName["Off"] = "off";
    TuiPhoneAutofillName["Tel"] = "tel";
})(TuiPhoneAutofillName || (TuiPhoneAutofillName = {}));
export var TuiDateAutofillName;
(function (TuiDateAutofillName) {
    TuiDateAutofillName["Off"] = "off";
    TuiDateAutofillName["Bday"] = "bday";
})(TuiDateAutofillName || (TuiDateAutofillName = {}));
export var TuiTransactionAutofillName;
(function (TuiTransactionAutofillName) {
    TuiTransactionAutofillName["Off"] = "off";
    TuiTransactionAutofillName["TransactionCurrency"] = "transaction-currency";
    TuiTransactionAutofillName["TransactionAmount"] = "transaction-amount";
})(TuiTransactionAutofillName || (TuiTransactionAutofillName = {}));
