// @taiga-ui/addon-doc/esm5/components/navigation/navigation.providers.js
import { __read, __spread } from "tslib";
import { InjectionToken } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { isPresent, TuiDestroyService } from '@taiga-ui/cdk';
import { filter, map, mergeMap, takeUntil } from 'rxjs/operators';
import { TUI_DOC_PAGES } from '../../tokens/pages';
import { TUI_DOC_TITLE } from '../../tokens/title';
export var NAVIGATION_TITLE = new InjectionToken("Page title");
export var NAVIGATION_LABELS = new InjectionToken("Navigation sections labels for search");
export var NAVIGATION_ITEMS = new InjectionToken("Navigation pages");
export var NAVIGATION_PROVIDERS = [
    TuiDestroyService,
    {
        provide: NAVIGATION_TITLE,
        deps: [Router, ActivatedRoute, TUI_DOC_TITLE, TuiDestroyService],
        useFactory: titleProviderFactory,
    },
    {
        provide: NAVIGATION_LABELS,
        deps: [TUI_DOC_PAGES],
        useFactory: labelsProviderFactory,
    },
    {
        provide: NAVIGATION_ITEMS,
        deps: [TUI_DOC_PAGES],
        useFactory: itemsProviderFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
export function titleProviderFactory(router, activatedRoute, titlePrefix, destroy$) {
    return router.events.pipe(filter(function (event) { return event instanceof NavigationEnd; }), map(function () { return activatedRoute.firstChild; }), filter(isPresent), mergeMap(function (_a) {
        var data = _a.data;
        return data;
    }), map(function (_a) {
        var title = _a.title;
        return titlePrefix + title;
    }), takeUntil(destroy$));
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function labelsProviderFactory(pages) {
    return pages
        .map(function (_a) {
        var section = _a.section;
        return section;
    })
        .filter(isPresent)
        .filter(function (item, index, array) { return array.indexOf(item) === index; });
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function itemsProviderFactory(pages) {
    var labels = labelsProviderFactory(pages);
    return __spread(labels.map(function (label) { return pages.filter(function (_a) {
        var section = _a.section;
        return section === label;
    }); }), [
        pages.filter(function (page) { return !page.section; }),
    ]);
}
