// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-input-mode.directive.js
var TuiTextfieldInputModeDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_INPUT_MODE = new InjectionToken('tuiTextfieldInputMode', {
    factory: inputModeDirectiveFactory,
});
let TuiTextfieldInputModeDirective = TuiTextfieldInputModeDirective_1 = class TuiTextfieldInputModeDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.inputMode = 'text';
    }
};
__decorate([
    Input('tuiTextfieldInputMode')
], TuiTextfieldInputModeDirective.prototype, "inputMode", void 0);
TuiTextfieldInputModeDirective = TuiTextfieldInputModeDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldInputMode]',
        providers: [
            {
                provide: TUI_TEXTFIELD_INPUT_MODE,
                useExisting: forwardRef(() => TuiTextfieldInputModeDirective_1),
            },
        ],
    })
], TuiTextfieldInputModeDirective);
export { TuiTextfieldInputModeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function inputModeDirectiveFactory() {
    return new TuiTextfieldInputModeDirective();
}
