// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-size.directive.js
var TuiTextfieldSizeDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_SIZE = new InjectionToken('tuiTextfieldSize', {
    factory: sizeDirectiveFactory,
});
let TuiTextfieldSizeDirective = TuiTextfieldSizeDirective_1 = class TuiTextfieldSizeDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.size = 'l';
    }
};
__decorate([
    Input('tuiTextfieldSize')
], TuiTextfieldSizeDirective.prototype, "size", void 0);
TuiTextfieldSizeDirective = TuiTextfieldSizeDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldSize]',
        providers: [
            {
                provide: TUI_TEXTFIELD_SIZE,
                useExisting: forwardRef(() => TuiTextfieldSizeDirective_1),
            },
        ],
    })
], TuiTextfieldSizeDirective);
export { TuiTextfieldSizeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function sizeDirectiveFactory() {
    return new TuiTextfieldSizeDirective();
}
