// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-type.directive.js
var TuiTextfieldTypeDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_TYPE = new InjectionToken('tuiTextfieldType', {
    factory: typeDirectiveFactory,
});
let TuiTextfieldTypeDirective = TuiTextfieldTypeDirective_1 = class TuiTextfieldTypeDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.type = 'text';
    }
};
__decorate([
    Input('tuiTextfieldType')
], TuiTextfieldTypeDirective.prototype, "type", void 0);
TuiTextfieldTypeDirective = TuiTextfieldTypeDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldType]',
        providers: [
            {
                provide: TUI_TEXTFIELD_TYPE,
                useExisting: forwardRef(() => TuiTextfieldTypeDirective_1),
            },
        ],
    })
], TuiTextfieldTypeDirective);
export { TuiTextfieldTypeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function typeDirectiveFactory() {
    return new TuiTextfieldTypeDirective();
}
