// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-autocomplete.directive.js
var TuiTextfieldAutocompleteDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_AUTOCOMPLETE = new InjectionToken('tuiTextfieldAutocomplete', {
    factory: autocompleteDirectiveFactory,
});
let TuiTextfieldAutocompleteDirective = TuiTextfieldAutocompleteDirective_1 = class TuiTextfieldAutocompleteDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.autocomplete = '';
    }
};
__decorate([
    Input('tuiTextfieldAutocomplete')
], TuiTextfieldAutocompleteDirective.prototype, "autocomplete", void 0);
TuiTextfieldAutocompleteDirective = TuiTextfieldAutocompleteDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldAutocomplete]',
        providers: [
            {
                provide: TUI_TEXTFIELD_AUTOCOMPLETE,
                useExisting: forwardRef(() => TuiTextfieldAutocompleteDirective_1),
            },
        ],
    })
], TuiTextfieldAutocompleteDirective);
export { TuiTextfieldAutocompleteDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function autocompleteDirectiveFactory() {
    return new TuiTextfieldAutocompleteDirective();
}
