// @taiga-ui/core/esm2015/directives/textfield-controller/textfield.controller.js
export class TuiTextfieldController {
    constructor(change$, autocompleteDirective, cleanerDirective, customContentDirective, exampleTextDirective, inputModeDirective, labelOutsideDirective, maxLengthDirective, sizeDirective, typeDirective) {
        this.change$ = change$;
        this.autocompleteDirective = autocompleteDirective;
        this.cleanerDirective = cleanerDirective;
        this.customContentDirective = customContentDirective;
        this.exampleTextDirective = exampleTextDirective;
        this.inputModeDirective = inputModeDirective;
        this.labelOutsideDirective = labelOutsideDirective;
        this.maxLengthDirective = maxLengthDirective;
        this.sizeDirective = sizeDirective;
        this.typeDirective = typeDirective;
    }
    get autocomplete() {
        return this.autocompleteDirective.autocomplete;
    }
    get cleaner() {
        return this.cleanerDirective.cleaner;
    }
    get customContent() {
        return this.customContentDirective.customContent || ``;
    }
    get exampleText() {
        return this.exampleTextDirective.exampleText;
    }
    get inputMode() {
        return this.inputModeDirective.inputMode;
    }
    get labelOutside() {
        return this.labelOutsideDirective.labelOutside;
    }
    get maxLength() {
        return this.maxLengthDirective.maxLength;
    }
    get size() {
        return this.sizeDirective.size;
    }
    get type() {
        return this.typeDirective.type;
    }
}
