// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-cleaner.directive.js
var TuiTextfieldCleanerDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_CLEANER = new InjectionToken('tuiTextfieldCleaner', {
    factory: cleanerDirectiveFactory,
});
let TuiTextfieldCleanerDirective = TuiTextfieldCleanerDirective_1 = class TuiTextfieldCleanerDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.cleaner = false;
    }
};
__decorate([
    Input('tuiTextfieldCleaner')
], TuiTextfieldCleanerDirective.prototype, "cleaner", void 0);
TuiTextfieldCleanerDirective = TuiTextfieldCleanerDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldCleaner]',
        providers: [
            {
                provide: TUI_TEXTFIELD_CLEANER,
                useExisting: forwardRef(() => TuiTextfieldCleanerDirective_1),
            },
        ],
    })
], TuiTextfieldCleanerDirective);
export { TuiTextfieldCleanerDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function cleanerDirectiveFactory() {
    return new TuiTextfieldCleanerDirective();
}
