// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-label-outside.directive.js
var TuiTextfieldLabelOutsideDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_LABEL_OUTSIDE = new InjectionToken('tuiTextfieldLabelOutside', {
    factory: labelOutsideDirectiveFactory,
});
let TuiTextfieldLabelOutsideDirective = TuiTextfieldLabelOutsideDirective_1 = class TuiTextfieldLabelOutsideDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.labelOutside = false;
    }
};
__decorate([
    Input('tuiTextfieldLabelOutside')
], TuiTextfieldLabelOutsideDirective.prototype, "labelOutside", void 0);
TuiTextfieldLabelOutsideDirective = TuiTextfieldLabelOutsideDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldLabelOutside]',
        providers: [
            {
                provide: TUI_TEXTFIELD_LABEL_OUTSIDE,
                useExisting: forwardRef(() => TuiTextfieldLabelOutsideDirective_1),
            },
        ],
    })
], TuiTextfieldLabelOutsideDirective);
export { TuiTextfieldLabelOutsideDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function labelOutsideDirectiveFactory() {
    return new TuiTextfieldLabelOutsideDirective();
}
