// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-max-length.directive.js
var TuiTextfieldMaxLengthDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_MAX_LENGTH = new InjectionToken('tuiTextfieldMaxLength', {
    factory: maxLengthDirectiveFactory,
});
let TuiTextfieldMaxLengthDirective = TuiTextfieldMaxLengthDirective_1 = class TuiTextfieldMaxLengthDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.maxLength = null;
    }
};
__decorate([
    Input('tuiTextfieldMaxLength')
], TuiTextfieldMaxLengthDirective.prototype, "maxLength", void 0);
TuiTextfieldMaxLengthDirective = TuiTextfieldMaxLengthDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldMaxLength]',
        providers: [
            {
                provide: TUI_TEXTFIELD_MAX_LENGTH,
                useExisting: forwardRef(() => TuiTextfieldMaxLengthDirective_1),
            },
        ],
    })
], TuiTextfieldMaxLengthDirective);
export { TuiTextfieldMaxLengthDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function maxLengthDirectiveFactory() {
    return new TuiTextfieldMaxLengthDirective();
}
