// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-custom-content.directive.js
var TuiTextfieldCustomContentDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_CUSTOM_CONTENT = new InjectionToken('tuiTextfieldCustomContent', {
    factory: customContentDirectiveFactory,
});
let TuiTextfieldCustomContentDirective = TuiTextfieldCustomContentDirective_1 = class TuiTextfieldCustomContentDirective extends TuiController {
    constructor() {
        super(...arguments);
        // TODO: 3.0 Remove null
        this.customContent = '';
    }
};
__decorate([
    Input('tuiTextfieldCustomContent')
], TuiTextfieldCustomContentDirective.prototype, "customContent", void 0);
TuiTextfieldCustomContentDirective = TuiTextfieldCustomContentDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldCustomContent]',
        providers: [
            {
                provide: TUI_TEXTFIELD_CUSTOM_CONTENT,
                useExisting: forwardRef(() => TuiTextfieldCustomContentDirective_1),
            },
        ],
    })
], TuiTextfieldCustomContentDirective);
export { TuiTextfieldCustomContentDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function customContentDirectiveFactory() {
    return new TuiTextfieldCustomContentDirective();
}
