// @taiga-ui/core/esm2015/directives/textfield-controller/textfield-example-text.directive.js
var TuiTextfieldExampleTextDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export const TUI_TEXTFIELD_EXAMPLE_TEXT = new InjectionToken('tuiTextfieldExampleText', {
    factory: exampleTextDirectiveFactory,
});
let TuiTextfieldExampleTextDirective = TuiTextfieldExampleTextDirective_1 = class TuiTextfieldExampleTextDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.exampleText = '';
    }
};
__decorate([
    Input('tuiTextfieldExampleText')
], TuiTextfieldExampleTextDirective.prototype, "exampleText", void 0);
TuiTextfieldExampleTextDirective = TuiTextfieldExampleTextDirective_1 = __decorate([
    Directive({
        selector: '[tuiTextfieldExampleText]',
        providers: [
            {
                provide: TUI_TEXTFIELD_EXAMPLE_TEXT,
                useExisting: forwardRef(() => TuiTextfieldExampleTextDirective_1),
            },
        ],
    })
], TuiTextfieldExampleTextDirective);
export { TuiTextfieldExampleTextDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function exampleTextDirectiveFactory() {
    return new TuiTextfieldExampleTextDirective();
}
