// @taiga-ui/kit/esm2015/components/arrow/arrow.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { sizeBigger, TUI_TEXTFIELD_SIZE, TuiHostedDropdownComponent, TuiTextfieldSizeDirective, } from '@taiga-ui/core';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TUI_ARROW_OPTIONS } from './arrow-options';
let TuiArrowComponent = class TuiArrowComponent {
    constructor(dropdown, textfieldSize, options) {
        this.dropdown = dropdown;
        this.textfieldSize = textfieldSize;
        this.options = options;
    }
    get arrowIcon() {
        return sizeBigger(this.textfieldSize.size)
            ? this.options.iconLarge
            : this.options.iconSmall;
    }
};
TuiArrowComponent.ctorParameters = () => [
    { type: TuiHostedDropdownComponent, decorators: [{ type: Inject, args: [TuiHostedDropdownComponent,] }] },
    { type: TuiTextfieldSizeDirective, decorators: [{ type: Inject, args: [TUI_TEXTFIELD_SIZE,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_ARROW_OPTIONS,] }] }
];
TuiArrowComponent = __decorate([
    Component({
        selector: 'tui-arrow',
        template: "<span\n    polymorpheus-outlet\n    class=\"t-icon\"\n    [class.t-icon_rotated]=\"dropdown.openChange | async\"\n    [content]=\"arrowIcon\"\n>\n    <ng-template let-icon>\n        <tui-svg [src]=\"icon\"></tui-svg>\n    </ng-template>\n</span>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".t-icon{transition-property:transform;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;display:block;cursor:pointer;pointer-events:none}.t-icon_rotated{transform:rotate(180deg)}"]
    }),
    __param(0, Inject(TuiHostedDropdownComponent)),
    __param(1, Inject(TUI_TEXTFIELD_SIZE)),
    __param(2, Inject(TUI_ARROW_OPTIONS))
], TuiArrowComponent);
export { TuiArrowComponent };
export const TUI_ARROW = new PolymorpheusComponent(TuiArrowComponent);
