// @taiga-ui/kit/esm2015/components/push/push.directive.js
var TuiPushDirective_1;
import { __decorate, __param } from "tslib";
import { Directive, Inject, Optional, SkipSelf } from '@angular/core';
import { TUI_BUTTON_OPTIONS, TuiButtonOptions, TuiModeDirective } from '@taiga-ui/core';
import { Subject } from 'rxjs';
let TuiPushDirective = TuiPushDirective_1 = class TuiPushDirective extends TuiModeDirective {
    constructor(modeDirective) {
        var _a;
        super();
        this.modeDirective = modeDirective;
        this.size = 's';
        this.shape = null;
        this.change$ = ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.change$) || new Subject();
    }
    get appearance() {
        var _a;
        return ((_a = this.modeDirective) === null || _a === void 0 ? void 0 : _a.mode) === 'onDark' ? 'accent' : 'secondary';
    }
};
TuiPushDirective.ctorParameters = () => [
    { type: TuiModeDirective, decorators: [{ type: Optional }, { type: SkipSelf }, { type: Inject, args: [TuiModeDirective,] }] }
];
TuiPushDirective = TuiPushDirective_1 = __decorate([
    Directive({
        selector: 'tui-push',
        providers: [
            {
                provide: TuiModeDirective,
                useExisting: TuiPushDirective_1,
            },
            {
                provide: TUI_BUTTON_OPTIONS,
                useExisting: TuiPushDirective_1,
            },
        ],
    }),
    __param(0, Optional()),
    __param(0, SkipSelf()),
    __param(0, Inject(TuiModeDirective))
], TuiPushDirective);
export { TuiPushDirective };
