// @taiga-ui/kit/esm5/components/push/push-alert.component.js
import { __assign, __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Inject } from '@angular/core';
import { TUI_ANIMATION_OPTIONS, tuiFadeIn, tuiHeightCollapse, tuiSlideInRight, } from '@taiga-ui/core';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import { TuiPushAlertDirective } from './push-alert.directive';
var TuiPushAlertComponent = /** @class */ (function () {
    function TuiPushAlertComponent(options, context) {
        this.options = options;
        this.context = context;
        this.animation = __assign({ value: '' }, this.options);
    }
    Object.defineProperty(TuiPushAlertComponent.prototype, "isDirective", {
        get: function () {
            return this.context.content instanceof TuiPushAlertDirective;
        },
        enumerable: true,
        configurable: true
    });
    TuiPushAlertComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ANIMATION_OPTIONS,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [POLYMORPHEUS_CONTEXT,] }] }
    ]; };
    __decorate([
        HostBinding('@tuiFadeIn'),
        HostBinding('@tuiSlideInRight'),
        HostBinding('@tuiHeightCollapse')
    ], TuiPushAlertComponent.prototype, "animation", void 0);
    TuiPushAlertComponent = __decorate([
        Component({
            selector: 'tui-push-alert',
            template: "<tui-push\n    *ngIf=\"!isDirective; else content\"\n    [heading]=\"context.heading\"\n    [type]=\"context.type\"\n    [timestamp]=\"context.timestamp\"\n    (close)=\"context.$implicit.complete()\"\n>\n    <img\n        *ngIf=\"context.image\"\n        alt=\"\"\n        [src]=\"context.image\"\n    />\n    <tui-svg\n        *ngIf=\"context.icon\"\n        [style.color]=\"context.iconColor\"\n        [src]=\"context.icon\"\n    ></tui-svg>\n    <div\n        polymorpheus-outlet\n        [content]=\"context.content\"\n    ></div>\n    <button\n        *ngIf=\"context.buttons.length > 1\"\n        tuiButton\n        (click)=\"context.$implicit.next(context.buttons[0])\"\n    >\n        {{ context.buttons[0] }}\n    </button>\n    <button\n        *ngIf=\"context.buttons.length\"\n        tuiLink\n        (click)=\"context.$implicit.next(context.buttons[context.buttons.length - 1])\"\n    >\n        {{ context.buttons[context.buttons.length - 1] }}\n    </button>\n</tui-push>\n<ng-template #content>\n    <div\n        polymorpheus-outlet\n        [content]=\"context.content\"\n    ></div>\n</ng-template>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiFadeIn, tuiSlideInRight, tuiHeightCollapse],
            host: { role: 'alert' },
            styles: [":host{max-width:calc(100% - 2rem);margin:0 1rem 1rem auto}:host:first-child{margin-top:auto}"]
        }),
        __param(0, Inject(TUI_ANIMATION_OPTIONS)),
        __param(1, Inject(POLYMORPHEUS_CONTEXT))
    ], TuiPushAlertComponent);
    return TuiPushAlertComponent;
}());
export { TuiPushAlertComponent };
