// @taiga-ui/kit/esm5/components/combo-box/combo-box.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiComboBoxDirective = /** @class */ (function (_super) {
    __extends(TuiComboBoxDirective, _super);
    function TuiComboBoxDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiComboBoxDirective_1 = TuiComboBoxDirective;
    Object.defineProperty(TuiComboBoxDirective.prototype, "value", {
        get: function () {
            return this.host.nativeValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiComboBoxDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiComboBoxDirective_1;
    TuiComboBoxDirective = TuiComboBoxDirective_1 = __decorate([
        Directive({
            selector: 'tui-combo-box',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiComboBoxDirective_1; }),
                },
            ],
        })
    ], TuiComboBoxDirective);
    return TuiComboBoxDirective;
}(TuiAbstractTextfieldHost));
export { TuiComboBoxDirective };
