// @taiga-ui/kit/esm2015/directives/lazy-loading/lazy-loading.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, HostBinding, HostListener, Inject, Input, } from '@angular/core';
import { IntersectionObserverService } from '@ng-web-apis/intersection-observer';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { TuiLazyLoadingService } from './lazy-loading.service';
let TuiLazyLoadingDirective = class TuiLazyLoadingDirective {
    constructor(src$, elementRef) {
        this.src$ = src$;
        this.elementRef = elementRef;
        this.animation = 'tuiSkeletonVibe ease-in-out 1s infinite alternate';
        this.background = 'var(--tui-clear-hover)';
        this.src = null;
        if (!this.supported) {
            this.src$.subscribe(src => {
                this.src = src;
            });
        }
    }
    set srcSetter(src) {
        this.src = this.supported ? src : null;
        this.src$.next(src);
    }
    get supported() {
        return 'loading' in this.elementRef.nativeElement;
    }
    onLoad() {
        this.background = '';
        this.animation = '';
    }
};
TuiLazyLoadingDirective.ctorParameters = () => [
    { type: TuiLazyLoadingService, decorators: [{ type: Inject, args: [TuiLazyLoadingService,] }] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
];
__decorate([
    Input('src')
], TuiLazyLoadingDirective.prototype, "srcSetter", null);
__decorate([
    HostBinding('style.animation')
], TuiLazyLoadingDirective.prototype, "animation", void 0);
__decorate([
    HostBinding('style.background')
], TuiLazyLoadingDirective.prototype, "background", void 0);
__decorate([
    HostBinding('attr.src')
], TuiLazyLoadingDirective.prototype, "src", void 0);
__decorate([
    HostListener('load')
], TuiLazyLoadingDirective.prototype, "onLoad", null);
TuiLazyLoadingDirective = __decorate([
    Directive({
        selector: 'img[loading="lazy"]',
        providers: [TuiLazyLoadingService, IntersectionObserverService, TuiDestroyService],
    }),
    __param(0, Inject(TuiLazyLoadingService)),
    __param(1, Inject(ElementRef))
], TuiLazyLoadingDirective);
export { TuiLazyLoadingDirective };
