// @taiga-ui/kit/fesm5/taiga-ui-kit-pipes-to-country-code.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { CHAR_PLUS } from '@taiga-ui/cdk';
import { TuiCountryIsoCode } from '@taiga-ui/i18n';
import { TUI_COUNTRIES_MASKS } from '@taiga-ui/kit/tokens';
import { tuiGetMaxAllowedPhoneLength, tuiIsoToCountryCode, tuiNotKzRegion } from '@taiga-ui/kit/utils';

// @dynamic
var TuiToCountryCodePipe = /** @class */ (function () {
    function TuiToCountryCodePipe(countriesMasks) {
        this.countriesMasks = countriesMasks;
    }
    TuiToCountryCodePipe.prototype.transform = function (value, countries) {
        var _this = this;
        return countries.find(function (countryIsoCode) {
            var ruCodeTest = countryIsoCode === TuiCountryIsoCode.RU &&
                /^[78]/.test(value) &&
                /^(?!880[1-9 ])/.test(value) &&
                value.length + 1 ===
                    tuiGetMaxAllowedPhoneLength(_this.countriesMasks, TuiCountryIsoCode.RU);
            var matched = ruCodeTest ||
                (value.startsWith(tuiIsoToCountryCode(_this.countriesMasks, countryIsoCode).replace(CHAR_PLUS, "")) &&
                    value.length + 1 ===
                        tuiGetMaxAllowedPhoneLength(_this.countriesMasks, countryIsoCode));
            if (matched) {
                switch (countryIsoCode) {
                    case TuiCountryIsoCode.RU:
                        return tuiNotKzRegion(value);
                    case TuiCountryIsoCode.KZ:
                        return !tuiNotKzRegion(value);
                    default:
                        return true;
                }
            }
            return false;
        });
    };
    TuiToCountryCodePipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_COUNTRIES_MASKS,] }] }
    ]; };
    TuiToCountryCodePipe = __decorate([
        Pipe({
            name: "tuiToCountryCode",
        }),
        __param(0, Inject(TUI_COUNTRIES_MASKS))
    ], TuiToCountryCodePipe);
    return TuiToCountryCodePipe;
}());

var TuiExtractCountryCodeModule = /** @class */ (function () {
    function TuiExtractCountryCodeModule() {
    }
    TuiExtractCountryCodeModule = __decorate([
        NgModule({
            declarations: [TuiToCountryCodePipe],
            exports: [TuiToCountryCodePipe],
        })
    ], TuiExtractCountryCodeModule);
    return TuiExtractCountryCodeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiExtractCountryCodeModule, TuiToCountryCodePipe };
