// @taiga-ui/kit/fesm5/taiga-ui-kit-constants.js
import { __read, __spread } from 'tslib';
import { TuiCountryIsoCode } from '@taiga-ui/i18n';
import { CHAR_PLUS, CHAR_HYPHEN, CHAR_NO_BREAK_SPACE, CHAR_EN_DASH } from '@taiga-ui/cdk';
import { TUI_DIGIT_REGEXP } from '@taiga-ui/core';

var _a;
/**
 * @deprecated Dont use it (will be deleted soon).
 * - Use token `TUI_COUNTRIES_MASKS` to get phone mask by country iso code
 * - Use token `TUI_COUNTRIES` to get localized country name by its iso code
 * @example
 * import {TUI_COUNTRIES, TUI_COUNTRIES_MASKS} from '@taiga-ui/kit';
 *
 * @todo TODO: remove it in 3.0
 */
var COUNTRIES = (_a = {},
    _a[TuiCountryIsoCode.AD] = { mask: "+376###-###", name: "\u0410\u043D\u0434\u043E\u0440\u0440\u0430" },
    _a[TuiCountryIsoCode.AE] = {
        mask: "+971-##-###-####",
        name: "\u041E\u0431\u044A\u0435\u0434\u0438\u043D\u0435\u043D\u043D\u044B\u0435 \u0410\u0440\u0430\u0431\u0441\u043A\u0438\u0435 \u042D\u043C\u0438\u0440\u0430\u0442\u044B",
    },
    _a[TuiCountryIsoCode.AF] = { mask: "+93##-###-####", name: "\u0410\u0444\u0433\u0430\u043D\u0438\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.AG] = { mask: "+1(268) ###-####", name: "\u0410\u043D\u0442\u0438\u0433\u0443\u0430 \u0438 \u0411\u0430\u0440\u0431\u0443\u0434\u0430" },
    _a[TuiCountryIsoCode.AI] = { mask: "+1(264) ###-####", name: "\u0410\u043D\u0433\u0438\u043B\u044C\u044F" },
    _a[TuiCountryIsoCode.AL] = { mask: "+355(###) ###-###", name: "\u0410\u043B\u0431\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.AM] = { mask: "+374##-###-###", name: "\u0410\u0440\u043C\u0435\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.AO] = { mask: "+244(###) ###-###", name: "\u0410\u043D\u0433\u043E\u043B\u0430" },
    _a[TuiCountryIsoCode.AR] = { mask: "+54(###) ###-####", name: "\u0410\u0440\u0433\u0435\u043D\u0442\u0438\u043D\u0430" },
    _a[TuiCountryIsoCode.AT] = { mask: "+43(###) ###-####", name: "\u0410\u0432\u0441\u0442\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.AU] = { mask: "+61#-####-####", name: "\u0410\u0432\u0441\u0442\u0440\u0430\u043B\u0438\u044F" },
    _a[TuiCountryIsoCode.AW] = { mask: "+297###-####", name: "\u0410\u0440\u0443\u0431\u0430" },
    _a[TuiCountryIsoCode.AZ] = { mask: "+994##-###-##-##", name: "\u0410\u0437\u0435\u0440\u0431\u0430\u0439\u0434\u0436\u0430\u043D" },
    _a[TuiCountryIsoCode.BA] = { mask: "+387 ###-####-#", name: "\u0411\u043E\u0441\u043D\u0438\u044F \u0438 \u0413\u0435\u0440\u0446\u0435\u0433\u043E\u0432\u0438\u043D\u0430" },
    _a[TuiCountryIsoCode.BB] = { mask: "+1(246) ###-####", name: "\u0411\u0430\u0440\u0431\u0430\u0434\u043E\u0441" },
    _a[TuiCountryIsoCode.BD] = { mask: "+880##-###-###", name: "\u0411\u0430\u043D\u0433\u043B\u0430\u0434\u0435\u0448" },
    _a[TuiCountryIsoCode.BE] = { mask: "+32(###) ###-###", name: "\u0411\u0435\u043B\u044C\u0433\u0438\u044F" },
    _a[TuiCountryIsoCode.BF] = { mask: "+226##-##-####", name: "\u0411\u0443\u0440\u043A\u0438\u043D\u0430 \u0424\u0430\u0441\u043E" },
    _a[TuiCountryIsoCode.BG] = { mask: "+359(###) ###-###", name: "\u0411\u043E\u043B\u0433\u0430\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.BH] = { mask: "+973####-####", name: "\u0411\u0430\u0445\u0440\u0435\u0439\u043D" },
    _a[TuiCountryIsoCode.BI] = { mask: "+257##-##-####", name: "\u0411\u0443\u0440\u0443\u043D\u0434\u0438" },
    _a[TuiCountryIsoCode.BJ] = { mask: "+229##-##-####", name: "\u0411\u0435\u043D\u0438\u043D" },
    _a[TuiCountryIsoCode.BL] = { mask: "+590 ## ## ## ## ##", name: "\u0421\u0435\u043D-\u0411\u0430\u0440\u0442\u0435\u043B\u044C\u043C\u0438" },
    _a[TuiCountryIsoCode.BM] = { mask: "+1(441) ###-####", name: "\u0411\u0435\u0440\u043C\u0443\u0434\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.BN] = { mask: "+673###-####", name: "\u0411\u0440\u0443\u043D\u0435\u0439-\u0414\u0430\u0440\u0443\u0441\u0441\u0430\u043B\u0430\u043C" },
    _a[TuiCountryIsoCode.BO] = { mask: "+591#-###-####", name: "\u0411\u043E\u043B\u0438\u0432\u0438\u044F" },
    _a[TuiCountryIsoCode.BQ] = {
        mask: "+599-###-####",
        name: "\u0411\u043E\u043D\u0430\u0439\u0440\u0435, \u0421\u0438\u043D\u0442-\u042D\u0441\u0442\u0430\u0442\u0438\u0443\u0441 \u0438 \u0421\u0430\u0431\u0430",
    },
    _a[TuiCountryIsoCode.BR] = { mask: "+55(##) ####-####", name: "\u0411\u0440\u0430\u0437\u0438\u043B\u0438\u044F" },
    _a[TuiCountryIsoCode.BS] = { mask: "+1(242) ###-####", name: "\u0411\u0430\u0433\u0430\u043C\u0441\u043A\u0438\u0435 \u041E\u0441\u0442\u0440\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.BT] = { mask: "+975#-###-###", name: "\u0411\u0443\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.BW] = { mask: "+267##-###-###", name: "\u0411\u043E\u0442\u0441\u0432\u0430\u043D\u0430" },
    _a[TuiCountryIsoCode.BY] = { mask: "+375(##) ###-##-##", name: "\u0411\u0435\u043B\u0430\u0440\u0443\u0441\u044C (\u0411\u0435\u043B\u043E\u0440\u0443\u0441\u0441\u0438\u044F)" },
    _a[TuiCountryIsoCode.BZ] = { mask: "+501###-####", name: "\u0411\u0435\u043B\u0438\u0437" },
    _a[TuiCountryIsoCode.CA] = { mask: "+1(###) ###-####", name: "\u041A\u0430\u043D\u0430\u0434\u0430" },
    _a[TuiCountryIsoCode.CD] = {
        mask: "+243(###) ###-###",
        name: "\u0414\u0435\u043C. \u0420\u0435\u0441\u043F. \u041A\u043E\u043D\u0433\u043E (\u041A\u0438\u043D\u0448\u0430\u0441\u0430)",
    },
    _a[TuiCountryIsoCode.CF] = {
        mask: "+236-##-##-####",
        name: "\u0426\u0435\u043D\u0442\u0440\u0430\u043B\u044C\u043D\u043E\u0430\u0444\u0440\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430",
    },
    _a[TuiCountryIsoCode.CG] = { mask: "+242##-###-####", name: "\u041A\u043E\u043D\u0433\u043E (\u0411\u0440\u0430\u0437\u0437\u0430\u0432\u0438\u043B\u044C)" },
    _a[TuiCountryIsoCode.CH] = { mask: "+41##-###-####", name: "\u0428\u0432\u0435\u0439\u0446\u0430\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.CI] = { mask: "+225##-###-###", name: "\u041A\u043E\u0442-\u0434\u2019\u0418\u0432\u0443\u0430\u0440" },
    _a[TuiCountryIsoCode.CL] = { mask: "+56#-####-####", name: "\u0427\u0438\u043B\u0438" },
    _a[TuiCountryIsoCode.CM] = { mask: "+237####-####", name: "\u041A\u0430\u043C\u0435\u0440\u0443\u043D" },
    _a[TuiCountryIsoCode.CN] = { mask: "+86(###) ####-####", name: "\u041A\u0438\u0442\u0430\u0439\u0441\u043A\u0430\u044F \u041D.\u0420." },
    _a[TuiCountryIsoCode.CO] = { mask: "+57(###) ###-####", name: "\u041A\u043E\u043B\u0443\u043C\u0431\u0438\u044F" },
    _a[TuiCountryIsoCode.CR] = { mask: "+506####-####", name: "\u041A\u043E\u0441\u0442\u0430-\u0420\u0438\u043A\u0430" },
    _a[TuiCountryIsoCode.CU] = { mask: "+53#-###-####", name: "\u041A\u0443\u0431\u0430" },
    _a[TuiCountryIsoCode.CV] = { mask: "+238(###) ##-##", name: "\u041A\u0430\u0431\u043E-\u0412\u0435\u0440\u0434\u0435" },
    _a[TuiCountryIsoCode.CW] = { mask: "+5999-###-####", name: "\u041A\u044E\u0440\u0430\u0441\u0430\u043E" },
    _a[TuiCountryIsoCode.CY] = { mask: "+357##-###-###", name: "\u041A\u0438\u043F\u0440" },
    _a[TuiCountryIsoCode.CZ] = { mask: "+420(###) ###-###", name: "\u0427\u0435\u0445\u0438\u044F" },
    _a[TuiCountryIsoCode.DE] = { mask: "+49(###) ###-###-##", name: "\u0413\u0435\u0440\u043C\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.DJ] = { mask: "+253##-##-##-##", name: "\u0414\u0436\u0438\u0431\u0443\u0442\u0438" },
    _a[TuiCountryIsoCode.DK] = { mask: "+45##-##-##-##", name: "\u0414\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.DM] = { mask: "+1(767) ###-####", name: "\u0414\u043E\u043C\u0438\u043D\u0438\u043A\u0430" },
    _a[TuiCountryIsoCode.DO] = { mask: "+1(###) ###-####", name: "\u0414\u043E\u043C\u0438\u043D\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430" },
    _a[TuiCountryIsoCode.DZ] = { mask: "+213##-###-####", name: "\u0410\u043B\u0436\u0438\u0440" },
    _a[TuiCountryIsoCode.EC] = { mask: "+593##-###-####", name: "\u042D\u043A\u0432\u0430\u0434\u043E\u0440 " },
    _a[TuiCountryIsoCode.EE] = { mask: "+372####-####", name: "\u042D\u0441\u0442\u043E\u043D\u0438\u044F " },
    _a[TuiCountryIsoCode.EG] = { mask: "+20(###) ###-####", name: "\u0415\u0433\u0438\u043F\u0435\u0442" },
    _a[TuiCountryIsoCode.ER] = { mask: "+291#-###-###", name: "\u042D\u0440\u0438\u0442\u0440\u0435\u044F" },
    _a[TuiCountryIsoCode.ES] = { mask: "+34(###) ###-###", name: "\u0418\u0441\u043F\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.ET] = { mask: "+251##-###-####", name: "\u042D\u0444\u0438\u043E\u043F\u0438\u044F" },
    _a[TuiCountryIsoCode.FI] = { mask: "+358(###) ###-##-##", name: "\u0424\u0438\u043D\u043B\u044F\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.FJ] = { mask: "+679##-#####", name: "\u0424\u0438\u0434\u0436\u0438" },
    _a[TuiCountryIsoCode.FK] = { mask: "+500#####", name: "\u0424\u043E\u043B\u043A\u043B\u0435\u043D\u0434\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.FM] = { mask: "+691###-####", name: "\u0424.\u0428. \u041C\u0438\u043A\u0440\u043E\u043D\u0435\u0437\u0438\u0438" },
    _a[TuiCountryIsoCode.FR] = { mask: "+33 ## ## ## ## ##", name: "\u0424\u0440\u0430\u043D\u0446\u0438\u044F" },
    _a[TuiCountryIsoCode.GA] = { mask: "+241#-##-##-##", name: "\u0413\u0430\u0431\u043E\u043D" },
    _a[TuiCountryIsoCode.GB] = { mask: "+44##-####-####", name: "\u0412\u0435\u043B\u0438\u043A\u043E\u0431\u0440\u0438\u0442\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.GD] = { mask: "+1(473) ###-####", name: "\u0413\u0440\u0435\u043D\u0430\u0434\u0430" },
    _a[TuiCountryIsoCode.GE] = { mask: "+995(###) ###-###", name: "\u0413\u0440\u0443\u0437\u0438\u044F" },
    _a[TuiCountryIsoCode.GF] = { mask: "+594 ## ## ## ## ##", name: "\u0413\u0430\u0439\u0430\u043D\u0430" },
    _a[TuiCountryIsoCode.GH] = { mask: "+233(###) ###-###", name: "\u0413\u0430\u043D\u0430" },
    _a[TuiCountryIsoCode.GI] = { mask: "+350###-#####", name: "\u0413\u0438\u0431\u0440\u0430\u043B\u0442\u0430\u0440" },
    _a[TuiCountryIsoCode.GL] = { mask: "+299##-##-##", name: "\u0413\u0440\u0435\u043D\u043B\u0430\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.GM] = { mask: "+220(###) ##-##", name: "\u0413\u0430\u043C\u0431\u0438\u044F" },
    _a[TuiCountryIsoCode.GN] = { mask: "+224##-###-###", name: "\u0413\u0432\u0438\u043D\u0435\u044F" },
    _a[TuiCountryIsoCode.GP] = { mask: "+590 ## ## ## ## ##", name: "\u0413\u0432\u0430\u0434\u0435\u043B\u0443\u043F\u0430" },
    _a[TuiCountryIsoCode.GQ] = { mask: "+240##-###-####", name: "\u042D\u043A\u0432\u0430\u0442\u043E\u0440\u0438\u0430\u043B\u044C\u043D\u0430\u044F \u0413\u0432\u0438\u043D\u0435\u044F" },
    _a[TuiCountryIsoCode.GR] = { mask: "+30(###) ###-####", name: "\u0413\u0440\u0435\u0446\u0438\u044F" },
    _a[TuiCountryIsoCode.GT] = { mask: "+502#-###-####", name: "\u0413\u0432\u0430\u0442\u0435\u043C\u0430\u043B\u0430" },
    _a[TuiCountryIsoCode.GW] = { mask: "+245#-######", name: "\u0413\u0432\u0438\u043D\u0435\u044F-\u0411\u0438\u0441\u0430\u0443" },
    _a[TuiCountryIsoCode.GY] = { mask: "+592###-####", name: "\u0413\u0430\u0439\u0430\u043D\u0430" },
    _a[TuiCountryIsoCode.HK] = { mask: "+852####-####", name: "\u0413\u043E\u043D\u043A\u043E\u043D\u0433" },
    _a[TuiCountryIsoCode.HN] = { mask: "+504####-####", name: "\u0413\u043E\u043D\u0434\u0443\u0440\u0430\u0441" },
    _a[TuiCountryIsoCode.HR] = { mask: "+385##-###-###", name: "\u0425\u043E\u0440\u0432\u0430\u0442\u0438\u044F" },
    _a[TuiCountryIsoCode.HT] = { mask: "+509##-##-####", name: "\u0413\u0430\u0438\u0442\u0438" },
    _a[TuiCountryIsoCode.HU] = { mask: "+36(###) ###-###", name: "\u0412\u0435\u043D\u0433\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.ID] = { mask: "+62(###) ###-##-###", name: "\u0418\u043D\u0434\u043E\u043D\u0435\u0437\u0438\u044F " },
    _a[TuiCountryIsoCode.IE] = { mask: "+353(###) ###-###", name: "\u0418\u0440\u043B\u0430\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.IL] = { mask: "+972##-###-####", name: "\u0418\u0437\u0440\u0430\u0438\u043B\u044C" },
    _a[TuiCountryIsoCode.IN] = { mask: "+91(####) ###-###", name: "\u0418\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.IQ] = { mask: "+964(###) ###-####", name: "\u0418\u0440\u0430\u043A" },
    _a[TuiCountryIsoCode.IR] = { mask: "+98(###) ###-####", name: "\u0418\u0440\u0430\u043D" },
    _a[TuiCountryIsoCode.IS] = { mask: "+354###-####", name: "\u0418\u0441\u043B\u0430\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.IT] = { mask: "+39(###) ####-###", name: "\u0418\u0442\u0430\u043B\u0438\u044F" },
    _a[TuiCountryIsoCode.JM] = { mask: "+1(876) ###-####", name: "\u042F\u043C\u0430\u0439\u043A\u0430" },
    _a[TuiCountryIsoCode.JO] = { mask: "+962#-####-####", name: "\u0418\u043E\u0440\u0434\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.JP] = { mask: "+81-##-####-####", name: "\u042F\u043F\u043E\u043D\u0438\u044F " },
    _a[TuiCountryIsoCode.KE] = { mask: "+254###-######", name: "\u041A\u0435\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.KG] = { mask: "+996(###) ###-###", name: "\u041A\u0438\u0440\u0433\u0438\u0437\u0438\u044F" },
    _a[TuiCountryIsoCode.KH] = { mask: "+855##-###-###", name: "\u041A\u0430\u043C\u0431\u043E\u0434\u0436\u0430" },
    _a[TuiCountryIsoCode.KM] = { mask: "+269##-#####", name: "\u041A\u043E\u043C\u043E\u0440\u044B" },
    _a[TuiCountryIsoCode.KN] = { mask: "+1(869) ###-####", name: "\u0421\u0435\u043D\u0442-\u041A\u0438\u0442\u0441 \u0438 \u041D\u0435\u0432\u0438\u0441" },
    _a[TuiCountryIsoCode.KP] = { mask: "+850####-#############", name: "\u041A\u043E\u0440\u0435\u0439\u0441\u043A\u0430\u044F \u041D\u0414\u0420" },
    _a[TuiCountryIsoCode.KR] = { mask: "+82##-###-####", name: "\u0420\u0435\u0441\u043F. \u041A\u043E\u0440\u0435\u044F" },
    _a[TuiCountryIsoCode.KW] = { mask: "+965####-####", name: "\u041A\u0443\u0432\u0435\u0439\u0442" },
    _a[TuiCountryIsoCode.KY] = { mask: "+1(345) ###-####", name: "\u041A\u0430\u0439\u043C\u0430\u043D\u043E\u0432\u044B \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.KZ] = { mask: "+7(###) ###-##-##", name: "\u041A\u0430\u0437\u0430\u0445\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.LA] = { mask: "+856##-##-###-###", name: "\u041B\u0430\u043E\u0441" },
    _a[TuiCountryIsoCode.LB] = { mask: "+961##-###-###", name: "\u041B\u0438\u0432\u0430\u043D " },
    _a[TuiCountryIsoCode.LC] = { mask: "+1(758) ###-####", name: "\u0421\u0435\u043D\u0442-\u041B\u044E\u0441\u0438\u044F" },
    _a[TuiCountryIsoCode.LI] = { mask: "+423(###) ###-####", name: "\u041B\u0438\u0445\u0442\u0435\u043D\u0448\u0442\u0435\u0439\u043D" },
    _a[TuiCountryIsoCode.LK] = { mask: "+94##-###-####", name: "\u0428\u0440\u0438-\u041B\u0430\u043D\u043A\u0430" },
    _a[TuiCountryIsoCode.LR] = { mask: "+231##-###-###", name: "\u041B\u0438\u0431\u0435\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.LS] = { mask: "+266#-###-####", name: "\u041B\u0435\u0441\u043E\u0442\u043E" },
    _a[TuiCountryIsoCode.LT] = { mask: "+370(###) ##-###", name: "\u041B\u0438\u0442\u0432\u0430" },
    _a[TuiCountryIsoCode.LU] = { mask: "+352(###) ###-###", name: "\u041B\u044E\u043A\u0441\u0435\u043C\u0431\u0443\u0440\u0433" },
    _a[TuiCountryIsoCode.LV] = { mask: "+371##-###-###", name: "\u041B\u0430\u0442\u0432\u0438\u044F" },
    _a[TuiCountryIsoCode.LY] = { mask: "+218##-###-####", name: "\u041B\u0438\u0432\u0438\u044F" },
    _a[TuiCountryIsoCode.MA] = { mask: "+212##-####-###", name: "\u041C\u0430\u0440\u043E\u043A\u043A\u043E" },
    _a[TuiCountryIsoCode.MC] = { mask: "+377###-###-###", name: "\u041C\u043E\u043D\u0430\u043A\u043E" },
    _a[TuiCountryIsoCode.MD] = { mask: "+373####-####", name: "\u041C\u043E\u043B\u0434\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.ME] = { mask: "+382##-###-###", name: "\u0427\u0435\u0440\u043D\u043E\u0433\u043E\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.MF] = { mask: "+590 ## ## ## ## ##", name: "\u0421\u0435\u043D-\u041C\u0430\u0440\u0442\u0435\u043D" },
    _a[TuiCountryIsoCode.MG] = { mask: "+261##-##-#####", name: "\u041C\u0430\u0434\u0430\u0433\u0430\u0441\u043A\u0430\u0440" },
    _a[TuiCountryIsoCode.MK] = { mask: "+389##-###-###", name: "\u0420\u0435\u0441\u043F. \u041C\u0430\u043A\u0435\u0434\u043E\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.ML] = { mask: "+223##-##-####", name: "\u041C\u0430\u043B\u0438" },
    _a[TuiCountryIsoCode.MM] = { mask: "+95##-###-###", name: "\u0411\u0438\u0440\u043C\u0430 (\u041C\u044C\u044F\u043D\u043C\u0430)" },
    _a[TuiCountryIsoCode.MN] = { mask: "+976##-##-####", name: "\u041C\u043E\u043D\u0433\u043E\u043B\u0438\u044F" },
    _a[TuiCountryIsoCode.MO] = { mask: "+853####-####", name: "\u041C\u0430\u043A\u0430\u043E" },
    _a[TuiCountryIsoCode.MQ] = { mask: "+596 ## ## ## ## ##", name: "\u041C\u0430\u0440\u0442\u0438\u043D\u0438\u043A\u0430" },
    _a[TuiCountryIsoCode.MR] = { mask: "+222##-##-####", name: "\u041C\u0430\u0432\u0440\u0438\u0442\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.MS] = { mask: "+1(664) ###-####", name: "\u041C\u043E\u043D\u0442\u0441\u0435\u0440\u0440\u0430\u0442" },
    _a[TuiCountryIsoCode.MT] = { mask: "+356####-####", name: "\u041C\u0430\u043B\u044C\u0442\u0430" },
    _a[TuiCountryIsoCode.MU] = { mask: "+230###-####", name: "\u041C\u0430\u0432\u0440\u0438\u043A\u0438\u0439" },
    _a[TuiCountryIsoCode.MV] = { mask: "+960###-####", name: "\u041C\u0430\u043B\u044C\u0434\u0438\u0432\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430" },
    _a[TuiCountryIsoCode.MW] = { mask: "+265#-####-####", name: "\u041C\u0430\u043B\u0430\u0432\u0438" },
    _a[TuiCountryIsoCode.MX] = { mask: "+52(###) ###-####", name: "\u041C\u0435\u043A\u0441\u0438\u043A\u0430" },
    _a[TuiCountryIsoCode.MY] = { mask: "+60(###) ###-###", name: "\u041C\u0430\u043B\u0430\u0439\u0437\u0438\u044F" },
    _a[TuiCountryIsoCode.MZ] = { mask: "+258##-###-###", name: "\u041C\u043E\u0437\u0430\u043C\u0431\u0438\u043A" },
    _a[TuiCountryIsoCode.NA] = { mask: "+264##-###-####", name: "\u041D\u0430\u043C\u0438\u0431\u0438\u044F" },
    _a[TuiCountryIsoCode.NC] = { mask: "+687 ### ###", name: "\u041A\u0430\u043B\u0435\u0434\u043E\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.NE] = { mask: "+227##-##-####", name: "\u041D\u0438\u0433\u0435\u0440" },
    _a[TuiCountryIsoCode.NG] = { mask: "+234(###) ###-####", name: "\u041D\u0438\u0433\u0435\u0440\u0438\u044F" },
    _a[TuiCountryIsoCode.NI] = { mask: "+505####-####", name: "\u041D\u0438\u043A\u0430\u0440\u0430\u0433\u0443\u0430" },
    _a[TuiCountryIsoCode.NL] = { mask: "+31##-###-####", name: "\u041D\u0438\u0434\u0435\u0440\u043B\u0430\u043D\u0434\u044B" },
    _a[TuiCountryIsoCode.NO] = { mask: "+47(###) ##-###", name: "\u041D\u043E\u0440\u0432\u0435\u0433\u0438\u044F" },
    _a[TuiCountryIsoCode.NP] = { mask: "+977##-###-###", name: "\u041D\u0435\u043F\u0430\u043B" },
    _a[TuiCountryIsoCode.NZ] = { mask: "+64(###) ###-####", name: "\u041D\u043E\u0432\u0430\u044F \u0417\u0435\u043B\u0430\u043D\u0434\u0438\u044F" },
    _a[TuiCountryIsoCode.OM] = { mask: "+968##-###-###", name: "\u041E\u043C\u0430\u043D" },
    _a[TuiCountryIsoCode.PA] = { mask: "+507###-####", name: "\u041F\u0430\u043D\u0430\u043C\u0430" },
    _a[TuiCountryIsoCode.PE] = { mask: "+51(###) ###-###", name: "\u041F\u0435\u0440\u0443" },
    _a[TuiCountryIsoCode.PF] = {
        mask: "+689##-##-##",
        name: "\u0424\u0440\u0430\u043D\u0446\u0443\u0437\u0441\u043A\u0430\u044F \u041F\u043E\u043B\u0438\u043D\u0435\u0437\u0438\u044F (\u0422\u0430\u0438\u0442\u0438)",
    },
    _a[TuiCountryIsoCode.PG] = { mask: "+675(###) ##-###", name: "\u041F\u0430\u043F\u0443\u0430-\u041D\u043E\u0432\u0430\u044F \u0413\u0432\u0438\u043D\u0435\u044F" },
    _a[TuiCountryIsoCode.PH] = { mask: "+63(###) ###-####", name: "\u0424\u0438\u043B\u0438\u043F\u043F\u0438\u043D\u044B" },
    _a[TuiCountryIsoCode.PK] = { mask: "+92(###) ###-####", name: "\u041F\u0430\u043A\u0438\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.PL] = { mask: "+48(###) ###-###", name: "\u041F\u043E\u043B\u044C\u0448\u0430" },
    _a[TuiCountryIsoCode.PT] = { mask: "+351##-###-####", name: "\u041F\u043E\u0440\u0442\u0443\u0433\u0430\u043B\u0438\u044F" },
    _a[TuiCountryIsoCode.PW] = { mask: "+680###-####", name: "\u041F\u0430\u043B\u0430\u0443" },
    _a[TuiCountryIsoCode.PY] = { mask: "+595(###) ###-###", name: "\u041F\u0430\u0440\u0430\u0433\u0432\u0430\u0439" },
    _a[TuiCountryIsoCode.QA] = { mask: "+974####-####", name: "\u041A\u0430\u0442\u0430\u0440" },
    _a[TuiCountryIsoCode.RE] = { mask: "+262 ## ## ## ## ##", name: "\u0420\u0435\u044E\u043D\u044C\u043E\u043D" },
    _a[TuiCountryIsoCode.RO] = { mask: "+40##-###-####", name: "\u0420\u0443\u043C\u044B\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.RS] = { mask: "+381##-###-####", name: "\u0421\u0435\u0440\u0431\u0438\u044F" },
    _a[TuiCountryIsoCode.RU] = { mask: "+7### ###-##-##", name: "\u0420\u043E\u0441\u0441\u0438\u044F" },
    _a[TuiCountryIsoCode.RW] = { mask: "+250(###) ###-###", name: "\u0420\u0443\u0430\u043D\u0434\u0430" },
    _a[TuiCountryIsoCode.SA] = { mask: "+966#-####-####", name: "\u0421\u0430\u0443\u0434\u043E\u0432\u0441\u043A\u0430\u044F \u0410\u0440\u0430\u0432\u0438\u044F " },
    _a[TuiCountryIsoCode.SB] = { mask: "+677###-####", name: "\u0421\u043E\u043B\u043E\u043C\u043E\u043D\u043E\u0432\u044B \u041E\u0441\u0442\u0440\u043E\u0432\u0430 " },
    _a[TuiCountryIsoCode.SC] = { mask: "+248#-###-###", name: "\u0421\u0435\u0439\u0448\u0435\u043B\u044B" },
    _a[TuiCountryIsoCode.SD] = { mask: "+249##-###-####", name: "\u0421\u0443\u0434\u0430\u043D" },
    _a[TuiCountryIsoCode.SE] = { mask: "+46##-###-####", name: "\u0428\u0432\u0435\u0446\u0438\u044F" },
    _a[TuiCountryIsoCode.SG] = { mask: "+65####-####", name: "\u0421\u0438\u043D\u0433\u0430\u043F\u0443\u0440" },
    _a[TuiCountryIsoCode.SH] = { mask: "+290####", name: "\u041E\u0441\u0442\u0440\u043E\u0432 \u0421\u0432\u044F\u0442\u043E\u0439 \u0415\u043B\u0435\u043D\u044B" },
    _a[TuiCountryIsoCode.SI] = { mask: "+386##-###-###", name: "\u0421\u043B\u043E\u0432\u0435\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.SK] = { mask: "+421(###) ###-###", name: "\u0421\u043B\u043E\u0432\u0430\u043A\u0438\u044F" },
    _a[TuiCountryIsoCode.SL] = { mask: "+232##-######", name: "\u0421\u044C\u0435\u0440\u0440\u0430-\u041B\u0435\u043E\u043D\u0435" },
    _a[TuiCountryIsoCode.SM] = { mask: "+378####-######", name: "\u0421\u0430\u043D-\u041C\u0430\u0440\u0438\u043D\u043E" },
    _a[TuiCountryIsoCode.SN] = { mask: "+221##-###-####", name: "\u0421\u0435\u043D\u0435\u0433\u0430\u043B" },
    _a[TuiCountryIsoCode.SO] = { mask: "+252##-###-###", name: "\u0421\u043E\u043C\u0430\u043B\u0438" },
    _a[TuiCountryIsoCode.SR] = { mask: "+597###-####", name: "\u0421\u0443\u0440\u0438\u043D\u0430\u043C " },
    _a[TuiCountryIsoCode.ST] = { mask: "+239##-#####", name: "\u0421\u0430\u043D-\u0422\u043E\u043C\u0435 \u0438 \u041F\u0440\u0438\u043D\u0441\u0438\u043F\u0438" },
    _a[TuiCountryIsoCode.SV] = { mask: "+503##-##-####", name: "\u0421\u0430\u043B\u044C\u0432\u0430\u0434\u043E\u0440" },
    _a[TuiCountryIsoCode.SX] = { mask: "+1(721) ###-####", name: "\u0421\u0438\u043D\u0442-\u041C\u0430\u0440\u0442\u0435\u043D" },
    _a[TuiCountryIsoCode.SY] = {
        mask: "+963##-####-###",
        name: "\u0421\u0438\u0440\u0438\u0439\u0441\u043A\u0430\u044F \u0430\u0440\u0430\u0431\u0441\u043A\u0430\u044F \u0440\u0435\u0441\u043F\u0443\u0431\u043B\u0438\u043A\u0430",
    },
    _a[TuiCountryIsoCode.SZ] = { mask: "+268##-##-####", name: "\u0421\u0432\u0430\u0437\u0438\u043B\u0435\u043D\u0434" },
    _a[TuiCountryIsoCode.TC] = { mask: "+1(649) ###-####", name: "\u0422\u0451\u0440\u043A\u0441 \u0438 \u041A\u0430\u0439\u043A\u043E\u0441" },
    _a[TuiCountryIsoCode.TD] = { mask: "+235##-##-##-##", name: "\u0427\u0430\u0434" },
    _a[TuiCountryIsoCode.TG] = { mask: "+228##-###-###", name: "\u0422\u043E\u0433\u043E" },
    _a[TuiCountryIsoCode.TH] = { mask: "+66##-###-####", name: "\u0422\u0430\u0438\u043B\u0430\u043D\u0434 " },
    _a[TuiCountryIsoCode.TJ] = { mask: "+992##-###-####", name: "\u0422\u0430\u0434\u0436\u0438\u043A\u0438\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.TL] = { mask: "+670###-#####", name: "\u0412\u043E\u0441\u0442\u043E\u0447\u043D\u044B\u0439 \u0422\u0438\u043C\u043E\u0440" },
    _a[TuiCountryIsoCode.TM] = { mask: "+993#-###-####", name: "\u0422\u0443\u0440\u043A\u043C\u0435\u043D\u0438\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.TN] = { mask: "+216##-###-###", name: "\u0422\u0443\u043D\u0438\u0441" },
    _a[TuiCountryIsoCode.TO] = { mask: "+676#####", name: "\u0422\u043E\u043D\u0433\u0430" },
    _a[TuiCountryIsoCode.TR] = { mask: "+90(###) ###-####", name: "\u0422\u0443\u0440\u0446\u0438\u044F" },
    _a[TuiCountryIsoCode.TT] = { mask: "+1(868) ###-####", name: "\u0422\u0440\u0438\u043D\u0438\u0434\u0430\u0434 \u0438 \u0422\u043E\u0431\u0430\u0433\u043E" },
    _a[TuiCountryIsoCode.TW] = { mask: "+886#-####-####", name: "\u0422\u0430\u0439\u0432\u0430\u043D\u044C" },
    _a[TuiCountryIsoCode.TZ] = { mask: "+255##-###-####", name: "\u0422\u0430\u043D\u0437\u0430\u043D\u0438\u044F" },
    _a[TuiCountryIsoCode.UA] = { mask: "+380(##) ###-##-##", name: "\u0423\u043A\u0440\u0430\u0438\u043D\u0430" },
    _a[TuiCountryIsoCode.UG] = { mask: "+256(###) ###-###", name: "\u0423\u0433\u0430\u043D\u0434\u0430" },
    _a[TuiCountryIsoCode.US] = { mask: "+1(###) ###-####", name: "\u0421\u0428\u0410" },
    _a[TuiCountryIsoCode.UY] = { mask: "+598#-###-##-##", name: "\u0423\u0440\u0443\u0433\u0432\u0430\u0439" },
    _a[TuiCountryIsoCode.UZ] = { mask: "+998##-###-####", name: "\u0423\u0437\u0431\u0435\u043A\u0438\u0441\u0442\u0430\u043D" },
    _a[TuiCountryIsoCode.VC] = { mask: "+1(784) ###-####", name: "\u0421\u0435\u043D\u0442-\u0412\u0438\u043D\u0441\u0435\u043D\u0442 \u0438 \u0413\u0440\u0435\u043D\u0430\u0434\u0438\u043D\u044B" },
    _a[TuiCountryIsoCode.VE] = { mask: "+58(###) ###-####", name: "\u0412\u0435\u043D\u0435\u0441\u0443\u044D\u043B\u0430" },
    _a[TuiCountryIsoCode.VG] = {
        mask: "+1(284)###-####",
        name: "\u0411\u0440\u0438\u0442\u0430\u043D\u0441\u043A\u0438\u0435 \u0412\u0438\u0440\u0433\u0438\u043D\u0441\u043A\u0438\u0435 \u043E\u0441\u0442\u0440\u043E\u0432\u0430",
    },
    _a[TuiCountryIsoCode.VN] = { mask: "+84(###) ####-###", name: "\u0412\u044C\u0435\u0442\u043D\u0430\u043C" },
    _a[TuiCountryIsoCode.VU] = { mask: "+678##-#####", name: "\u0412\u0430\u043D\u0443\u0430\u0442\u0443 " },
    _a[TuiCountryIsoCode.WS] = { mask: "+685##-####", name: "\u0421\u0430\u043C\u043E\u0430" },
    _a[TuiCountryIsoCode.XK] = { mask: "+383##-###-###", name: "\u041A\u043E\u0441\u043E\u0432\u043E" },
    _a[TuiCountryIsoCode.YE] = { mask: "+967###-###-###", name: "\u0419\u0435\u043C\u0435\u043D " },
    _a[TuiCountryIsoCode.YT] = { mask: "+262 ## ## ## ## ##", name: "\u041C\u0430\u0439\u043E\u0442\u0442\u0430" },
    _a[TuiCountryIsoCode.ZA] = { mask: "+27##-###-####", name: "\u042E\u0436\u043D\u043E-\u0410\u0444\u0440\u0438\u043A\u0430\u043D\u0441\u043A\u0430\u044F \u0420\u0435\u0441\u043F." },
    _a[TuiCountryIsoCode.ZM] = { mask: "+260##-###-####", name: "\u0417\u0430\u043C\u0431\u0438\u044F" },
    _a[TuiCountryIsoCode.ZW] = { mask: "+263#-######", name: "\u0417\u0438\u043C\u0431\u0430\u0431\u0432\u0435" },
    _a);
/**
 * @deprecated will be deleted soon
 * TODO: remove in 3.0 (in ivy compilation)
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function countriesMasksReducer(acc, _a) {
    var _b = __read(_a, 2), countryIsoCode = _b[0], mask = _b[1].mask;
    acc[countryIsoCode] = mask;
    return acc;
}
/**
 * @deprecated will be deleted soon
 * Use token `TUI_COUNTRIES_MASKS`
 * TODO: remove in 3.0
 */
var COUNTRIES_MASKS = Object.entries(COUNTRIES).reduce(countriesMasksReducer, {});

var DATE_TIME_SEPARATOR = ", ";

var EMPTY_MASK = {
    mask: function () {
        return false;
    },
};

var GROUP_CLASS_NAMES = [
    "_active",
    "_disabled",
    "_focused",
    "_focus-visible",
    "_hosted_dropdown_focused",
    "ng-invalid",
    "ng-touched",
    "_pressed",
    "_readonly",
];

var iconBlank = "<svg></svg>";

var MASK_AFTER_CODE_REGEXP = /\([#]+\)|[#\- ]/g;

var TUI_PHONE_MASK = [
    CHAR_PLUS,
    "7",
    " ",
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    " ",
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    CHAR_HYPHEN,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    CHAR_HYPHEN,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
];
var TUI_TIME_MASK = [
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    ":",
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
];
/**
 * @deprecated use {@link tuiCreateDateMask} (from '@taiga-ui/kit') instead
 */
var TUI_DATE_MASK = [
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    ".",
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    ".",
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
    TUI_DIGIT_REGEXP,
];
/**
 * @deprecated use {@link tuiCreateDateRangeMask} (from '@taiga-ui/kit') instead
 */
var TUI_DATE_RANGE_MASK = __spread(TUI_DATE_MASK, [
    CHAR_NO_BREAK_SPACE,
    CHAR_EN_DASH,
    CHAR_NO_BREAK_SPACE
], TUI_DATE_MASK);

/**
 * Used as a limit for eliminating JS issues with floating point math
 */
var TUI_FLOATING_PRECISION = 7;

var MAX_DAY_RANGE_LENGTH_MAPPER = function (min, value, maxLength, backwards) {
    if (!(value === null || value === void 0 ? void 0 : value.isSingleDay) || !maxLength) {
        return min;
    }
    var dateShift = value.from
        .append(maxLength, backwards)
        .append({ day: 1 }, !backwards);
    if (backwards) {
        return dateShift.dayBefore(min) ? min : dateShift;
    }
    if (!min) {
        return dateShift;
    }
    return dateShift.dayAfter(min) ? min : dateShift;
};

var MAX_TIME_VALUES = {
    HH: 23,
    MM: 59,
    SS: 59,
    MS: 999,
};

var DEFAULT_ROUTER_LINK_OPTIONS = {
    fragment: "",
    preserveFragment: true,
    replaceUrl: false,
    skipLocationChange: false,
    queryParams: null,
    useHref: false,
    target: null,
};

/**
 * Generated bundle index. Do not edit.
 */

export { COUNTRIES, COUNTRIES_MASKS, DATE_TIME_SEPARATOR, DEFAULT_ROUTER_LINK_OPTIONS, EMPTY_MASK, GROUP_CLASS_NAMES, MASK_AFTER_CODE_REGEXP, MAX_DAY_RANGE_LENGTH_MAPPER, MAX_TIME_VALUES, TUI_DATE_MASK, TUI_DATE_RANGE_MASK, TUI_FLOATING_PRECISION, TUI_PHONE_MASK, TUI_TIME_MASK, countriesMasksReducer, iconBlank };
