// @taiga-ui/kit/fesm5/taiga-ui-kit-components-badged-content.js
import { __decorate } from 'tslib';
import { Input, HostBinding, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { px, isNumber, tuiDefaultProp } from '@taiga-ui/cdk';
import { sizeBigger, TuiSvgModule } from '@taiga-ui/core';
import { CommonModule } from '@angular/common';
import { TuiBadgeModule } from '@taiga-ui/kit/components/badge';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';

var BADGE_SIZE = {
    xs: 'xs',
    s: 's',
    m: 's',
    l: 'm',
    xl: 'm',
    xxl: 'l',
};
var TuiBadgedContentComponent = /** @class */ (function () {
    function TuiBadgedContentComponent() {
        this.contentTop = '';
        this.contentBottom = '';
        this.size = 'm';
        this.colorTop = '';
        this.colorBottom = '';
        this.rounded = false;
    }
    Object.defineProperty(TuiBadgedContentComponent.prototype, "topNotification", {
        get: function () {
            return !this.contentTop && this.colorTop ? this.colorTop : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiBadgedContentComponent.prototype, "bottomNotification", {
        get: function () {
            return !this.contentBottom && this.colorBottom ? this.colorBottom : '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiBadgedContentComponent.prototype, "badgeSize", {
        get: function () {
            return BADGE_SIZE[this.size];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiBadgedContentComponent.prototype, "sizeBig", {
        get: function () {
            return sizeBigger(this.size);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiBadgedContentComponent.prototype, "boxShadow", {
        get: function () {
            var borderWidth = this.sizeBig ? 3 : 2;
            return "0 0 0 " + px(borderWidth);
        },
        enumerable: true,
        configurable: true
    });
    TuiBadgedContentComponent.prototype.contentIsNumber = function (content) {
        return isNumber(content.valueOf());
    };
    TuiBadgedContentComponent.prototype.contentIsString = function (content) {
        return typeof content.valueOf() === 'string';
    };
    TuiBadgedContentComponent.prototype.getStatus = function (color) {
        return color ? 'custom' : 'primary';
    };
    __decorate([
        Input(),
        HostBinding('attr.data-tui-host-top'),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "contentTop", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-tui-host-bottom'),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "contentBottom", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size'),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "size", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "colorTop", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "colorBottom", void 0);
    __decorate([
        Input(),
        HostBinding('class._rounded'),
        tuiDefaultProp()
    ], TuiBadgedContentComponent.prototype, "rounded", void 0);
    TuiBadgedContentComponent = __decorate([
        Component({
            selector: 'tui-badged-content',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "<ng-content></ng-content>\n\n<div\n    *ngIf=\"topNotification\"\n    class=\"t-notification t-notification_top\"\n    [style.background]=\"topNotification\"\n    [style.boxShadow]=\"boxShadow\"\n></div>\n\n<div\n    *ngIf=\"bottomNotification\"\n    class=\"t-notification t-notification_bottom\"\n    [style.background]=\"bottomNotification\"\n    [style.boxShadow]=\"boxShadow\"\n></div>\n\n<div\n    *ngIf=\"contentTop\"\n    polymorpheus-outlet\n    class=\"t-content t-content_top\"\n    [style.boxShadow]=\"boxShadow\"\n    [content]=\"contentTop\"\n>\n    <ng-template let-content>\n        <tui-badge\n            *ngIf=\"contentIsNumber(content)\"\n            [status]=\"getStatus(colorTop)\"\n            [style.backgroundColor]=\"colorTop\"\n            [size]=\"badgeSize\"\n            [value]=\"contentTop\"\n        ></tui-badge>\n        <tui-svg\n            *ngIf=\"contentIsString(content)\"\n            class=\"t-icon\"\n            [style.color]=\"colorTop\"\n            [src]=\"content\"\n        ></tui-svg>\n    </ng-template>\n</div>\n\n<div\n    *ngIf=\"contentBottom && sizeBig\"\n    polymorpheus-outlet\n    class=\"t-content t-content_bottom\"\n    [style.boxShadow]=\"boxShadow\"\n    [content]=\"contentBottom\"\n>\n    <ng-template let-content>\n        <tui-badge\n            *ngIf=\"contentIsNumber(content)\"\n            [status]=\"getStatus(colorBottom)\"\n            [style.backgroundColor]=\"colorBottom\"\n            [size]=\"badgeSize\"\n            [value]=\"contentBottom\"\n        ></tui-badge>\n        <tui-svg\n            *ngIf=\"contentIsString(content)\"\n            class=\"t-icon\"\n            [style.color]=\"colorBottom\"\n            [src]=\"content\"\n        ></tui-svg>\n    </ng-template>\n</div>\n",
            styles: [":host{position:relative;display:inline-block;color:var(--tui-base-01)}.t-notification{position:absolute;border-radius:100%;box-shadow:0 0 0 2px currentColor}.t-notification:not([data-tui-background]){background-color:currentColor}:host[data-size='m']:not(._rounded) .t-notification_top,:host[data-size='s']:not(._rounded) .t-notification_top,:host[data-size=xs]:not(._rounded) .t-notification_top{top:-.125rem;right:-.125rem}:host[data-size='l']:not(._rounded) .t-notification_top,:host[data-size=xl]:not(._rounded) .t-notification_top{top:-.25rem;right:-.25rem}:host[data-size=xxl]:not(._rounded) .t-notification_top{top:-.5rem;right:-.5rem}:host[data-size=xs]._rounded .t-notification_top{top:0;right:0}:host[data-size='s']._rounded .t-notification_top{top:.0625rem;right:.0625rem}:host[data-size='m']._rounded .t-notification_top{top:.1875rem;right:.1875rem}:host[data-size='l']._rounded .t-notification_top{top:.25rem;right:.25rem}:host[data-size=xl]._rounded .t-notification_top{top:.375rem;right:.375rem}:host[data-size=xxl]._rounded .t-notification_top{top:.5rem;right:.5rem}:host[data-size='m']:not(._rounded) .t-notification_bottom{bottom:-.125rem;right:-.125rem}:host[data-size='l']:not(._rounded) .t-notification_bottom,:host[data-size=xl]:not(._rounded) .t-notification_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl]:not(._rounded) .t-notification_bottom{bottom:-.5rem;right:-.5rem}:host[data-size='m']._rounded .t-notification_bottom{bottom:.1875rem;right:.1875rem}:host[data-size='l']._rounded .t-notification_bottom{bottom:.25rem;right:.25rem}:host[data-size=xl]._rounded .t-notification_bottom{bottom:.375rem;right:.375rem}:host[data-size=xxl]._rounded .t-notification_bottom{bottom:.5rem;right:.5rem}:host[data-size='m'] .t-notification,:host[data-size='s'] .t-notification,:host[data-size=xs] .t-notification{width:.5rem;height:.5rem}:host[data-size='l'] .t-notification{width:.75rem;height:.75rem}:host[data-size=xl] .t-notification{width:1rem;height:1rem}:host[data-size=xxl] .t-notification{width:1.5rem;height:1.5rem}.t-icon[data-tui-color=normal],.t-notification[data-tui-color=normal]{color:var(--tui-base-06)}.t-icon[data-tui-color=onDark],.t-notification[data-tui-color=onDark]{color:var(--tui-secondary)}.t-icon[data-tui-color=gray],.t-notification[data-tui-color=gray]{color:var(--tui-base-03)}.t-icon[data-tui-color=warning],.t-notification[data-tui-color=warning]{color:var(--tui-primary)}.t-icon[data-tui-color=secondary],.t-notification[data-tui-color=secondary]{color:var(--tui-info-fill)}.t-icon[data-tui-color=highlight],.t-icon[data-tui-color=white],.t-notification[data-tui-color=highlight],.t-notification[data-tui-color=white]{color:var(--tui-base-01)}.t-icon[data-tui-color=transparent-dark],.t-notification[data-tui-color=transparent-dark]{color:var(--tui-clear)}.t-icon[data-tui-color=transparent-light],.t-notification[data-tui-color=transparent-light]{color:var(--tui-clear-inverse)}.t-icon{color:var(--tui-success-fill)}:host[data-size=xs] .t-icon{transform:scale(.66666667);margin:-.25rem}:host[data-size='m'] .t-icon,:host[data-size='s'] .t-icon{width:1.5rem;height:1.5rem;margin:-.125rem}:host[data-size='l'] .t-icon,:host[data-size=xl] .t-icon{transform:scale(1.2)}:host[data-size=xxl] .t-icon{transform:scale(1.6)}.t-content{position:absolute;display:flex;align-items:center;justify-content:center;border-radius:6.25rem;overflow:hidden;background-color:currentColor;box-shadow:0 0 0 2px currentColor}:host[data-size='l'][data-tui-host-top]:not(._rounded) .t-content_top,:host[data-size=xxl][data-tui-host-top]:not(._rounded) .t-content_top{right:-.5rem;top:-.5rem}:host[data-size=xl][data-tui-host-top]:not(._rounded) .t-content_top{right:-.375rem;top:-.375rem}:host[data-size='m'][data-tui-host-top]:not(._rounded) .t-content_top,:host[data-size='s'][data-tui-host-top]:not(._rounded) .t-content_top,:host[data-size=xs][data-tui-host-top]:not(._rounded) .t-content_top{right:-.25rem;top:-.25rem}:host[data-size=xxl][data-tui-host-top]._rounded .t-content_top{right:.25rem;top:.25rem}:host[data-size=xl][data-tui-host-top]._rounded .t-content_top{right:.125rem;top:.125rem}:host[data-size='l'][data-tui-host-top]._rounded .t-content_top,:host[data-size='m'][data-tui-host-top]._rounded .t-content_top,:host[data-size='s'][data-tui-host-top]._rounded .t-content_top,:host[data-size=xs][data-tui-host-top]._rounded .t-content_top{right:-.25rem;top:-.25rem}:host[data-size='l'][data-tui-host-bottom]:not(._rounded) .t-content_bottom,:host[data-size=xxl][data-tui-host-bottom]:not(._rounded) .t-content_bottom{bottom:-.5rem;right:-.5rem}:host[data-size=xl][data-tui-host-bottom]:not(._rounded) .t-content_bottom{bottom:-.375rem;right:-.375rem}:host[data-size='m'][data-tui-host-bottom]:not(._rounded) .t-content_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl][data-tui-host-bottom]._rounded .t-content_bottom{bottom:.25rem;right:.25rem}:host[data-size=xl][data-tui-host-bottom]._rounded .t-content_bottom{bottom:.125rem;right:.125rem}:host[data-size='l'][data-tui-host-bottom]._rounded .t-content_bottom,:host[data-size='m'][data-tui-host-bottom]._rounded .t-content_bottom{bottom:-.25rem;right:-.25rem}:host[data-size=xxl] .t-content{height:2rem;min-width:2rem}:host[data-size='l'] .t-content,:host[data-size=xl] .t-content{height:1.5rem;min-width:1.5rem}:host[data-size='m'] .t-content,:host[data-size='s'] .t-content{height:1.25rem;min-width:1.25rem}:host[data-size=xs] .t-content{height:1rem;min-width:1rem}"]
        })
    ], TuiBadgedContentComponent);
    return TuiBadgedContentComponent;
}());

var TuiBadgedContentModule = /** @class */ (function () {
    function TuiBadgedContentModule() {
    }
    TuiBadgedContentModule = __decorate([
        NgModule({
            imports: [CommonModule, PolymorpheusModule, TuiSvgModule, TuiBadgeModule],
            declarations: [TuiBadgedContentComponent],
            exports: [TuiBadgedContentComponent],
        })
    ], TuiBadgedContentModule);
    return TuiBadgedContentModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBadgedContentComponent, TuiBadgedContentModule };
