// @taiga-ui/kit/fesm5/taiga-ui-kit-pipes-iso-to-country-code.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TUI_COUNTRIES_MASKS } from '@taiga-ui/kit/tokens';
import { tuiIsoToCountryCode } from '@taiga-ui/kit/utils';

// @dynamic
var TuiIsoToCountryCodePipe = /** @class */ (function () {
    function TuiIsoToCountryCodePipe(countriesMasks) {
        this.countriesMasks = countriesMasks;
    }
    TuiIsoToCountryCodePipe.prototype.transform = function (isoCode) {
        return tuiIsoToCountryCode(this.countriesMasks, isoCode);
    };
    TuiIsoToCountryCodePipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_COUNTRIES_MASKS,] }] }
    ]; };
    TuiIsoToCountryCodePipe = __decorate([
        Pipe({
            name: "tuiIsoToCountryCode",
        }),
        __param(0, Inject(TUI_COUNTRIES_MASKS))
    ], TuiIsoToCountryCodePipe);
    return TuiIsoToCountryCodePipe;
}());

var TuiIsoToCountryCodeModule = /** @class */ (function () {
    function TuiIsoToCountryCodeModule() {
    }
    TuiIsoToCountryCodeModule = __decorate([
        NgModule({
            declarations: [TuiIsoToCountryCodePipe],
            exports: [TuiIsoToCountryCodePipe],
        })
    ], TuiIsoToCountryCodeModule);
    return TuiIsoToCountryCodeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiIsoToCountryCodeModule, TuiIsoToCountryCodePipe };
