// @taiga-ui/kit/fesm5/taiga-ui-kit-components-multi-select-option.js
import { __extends, __decorate } from 'tslib';
import { Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { isPresent } from '@taiga-ui/cdk';
import { sizeBigger, TuiPrimitiveCheckboxModule } from '@taiga-ui/core';
import { TuiSelectOptionComponent } from '@taiga-ui/kit/components/select-option';
import { CommonModule } from '@angular/common';

var TuiMultiSelectOptionComponent = /** @class */ (function (_super) {
    __extends(TuiMultiSelectOptionComponent, _super);
    function TuiMultiSelectOptionComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(TuiMultiSelectOptionComponent.prototype, "size", {
        get: function () {
            return sizeBigger(this.option.size) ? 'l' : 'm';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiMultiSelectOptionComponent.prototype, "selected", {
        get: function () {
            var _this = this;
            return (isPresent(this.option.value) &&
                isPresent(this.control.value) &&
                !!this.control.value.find(function (item) {
                    return _this.matcher(item, _this.option.value);
                }));
        },
        enumerable: true,
        configurable: true
    });
    TuiMultiSelectOptionComponent = __decorate([
        Component({
            selector: 'tui-multi-select-option',
            template: "<tui-primitive-checkbox\n    class=\"t-checkbox\"\n    [size]=\"size\"\n    [value]=\"selected$ | async\"\n></tui-primitive-checkbox>\n<ng-container [ngTemplateOutlet]=\"context.$implicit\"></ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: ["@keyframes retrigger{from{left:1px}to{left:2px}}:host{display:flex;flex:1;align-items:center;overflow:hidden;animation:1s retrigger}.t-checkbox{margin-right:.75rem}"]
        })
    ], TuiMultiSelectOptionComponent);
    return TuiMultiSelectOptionComponent;
}(TuiSelectOptionComponent));

var TuiMultiSelectOptionModule = /** @class */ (function () {
    function TuiMultiSelectOptionModule() {
    }
    TuiMultiSelectOptionModule = __decorate([
        NgModule({
            imports: [CommonModule, TuiPrimitiveCheckboxModule],
            declarations: [TuiMultiSelectOptionComponent],
            exports: [TuiMultiSelectOptionComponent],
            entryComponents: [TuiMultiSelectOptionComponent],
        })
    ], TuiMultiSelectOptionModule);
    return TuiMultiSelectOptionModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMultiSelectOptionComponent, TuiMultiSelectOptionModule };
