// @taiga-ui/kit/fesm5/taiga-ui-kit-providers.js
import { TuiMonthPipe, TuiDropdownControllerDirective, TUI_DROPDOWN_CONTROLLER, TUI_VALUE_ACCESSOR } from '@taiga-ui/core';
import { TUI_MONTH_FORMATTER } from '@taiga-ui/kit/tokens';
import { of, identity } from 'rxjs';
import { map } from 'rxjs/operators';
import { Optional, Self } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

var TUI_MONTH_FORMATTER_PROVIDER = {
    provide: TUI_MONTH_FORMATTER,
    deps: [TuiMonthPipe],
    useFactory: monthFormatterFactory,
};
/**
 * Range.setStart/set-end, except it uses offset in characters only
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
function monthFormatterFactory(pipe) {
    return function (month) {
        return month
            ? pipe
                .transform(month)
                .pipe(map(function (formatted) { return formatted + " " + month.formattedYear; }))
            : of("");
    };
}

// eslint-disable-next-line @typescript-eslint/naming-convention
function fixedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.limitWidth = "fixed";
    return directive;
}
var FIXED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: fixedDropdownControllerFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
function leftAlignedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.align = "left";
    return directive;
}
var LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: leftAlignedDropdownControllerFactory,
    },
];

// TODO: 3.0 remove in ivy compilation
var IDENTITY = identity;
var TUI_VALUE_ACCESSOR_PROVIDER = {
    provide: TUI_VALUE_ACCESSOR,
    deps: [[new Optional(), new Self(), NG_VALUE_ACCESSOR]],
    useFactory: IDENTITY,
};

/**
 * Generated bundle index. Do not edit.
 */

export { FIXED_DROPDOWN_CONTROLLER_PROVIDER, IDENTITY, LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER, TUI_MONTH_FORMATTER_PROVIDER, TUI_VALUE_ACCESSOR_PROVIDER, fixedDropdownControllerFactory, leftAlignedDropdownControllerFactory, monthFormatterFactory };
