// @taiga-ui/kit/fesm5/taiga-ui-kit-pipes-stringify-content.js
import { __decorate } from 'tslib';
import { Pipe, NgModule } from '@angular/core';

var TuiStringifyContentPipe = /** @class */ (function () {
    function TuiStringifyContentPipe() {
    }
    TuiStringifyContentPipe.prototype.transform = function (stringify) {
        return function (_a) {
            var $implicit = _a.$implicit;
            return stringify($implicit);
        };
    };
    TuiStringifyContentPipe = __decorate([
        Pipe({
            name: "tuiStringifyContent",
        })
    ], TuiStringifyContentPipe);
    return TuiStringifyContentPipe;
}());

var TuiStringifyContentPipeModule = /** @class */ (function () {
    function TuiStringifyContentPipeModule() {
    }
    TuiStringifyContentPipeModule = __decorate([
        NgModule({
            declarations: [TuiStringifyContentPipe],
            exports: [TuiStringifyContentPipe],
        })
    ], TuiStringifyContentPipeModule);
    return TuiStringifyContentPipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiStringifyContentPipe, TuiStringifyContentPipeModule };
