// @taiga-ui/kit/fesm5/taiga-ui-kit-enums.js
/**
 * @deprecated use string values
 */
var TuiBorders;
(function (TuiBorders) {
    TuiBorders["All"] = "all";
    TuiBorders["TopBottom"] = "top-bottom";
})(TuiBorders || (TuiBorders = {}));

/**
 * @deprecated use {@link TuiDropdownPositionT}
 */
var TuiDropdownPosition;
(function (TuiDropdownPosition) {
    TuiDropdownPosition["Selection"] = "selection";
    TuiDropdownPosition["Word"] = "word";
    TuiDropdownPosition["Tag"] = "tag";
})(TuiDropdownPosition || (TuiDropdownPosition = {}));

/**
 * @internal
 */
var TuiFileState;
(function (TuiFileState) {
    TuiFileState["Error"] = "error";
    TuiFileState["Deleted"] = "deleted";
    TuiFileState["Loading"] = "loading";
    TuiFileState["Normal"] = "normal";
})(TuiFileState || (TuiFileState = {}));

/**
 * @deprecated use join type {@link TuiMarkerIconModeT}
 */
var TuiMarkerIconMode;
(function (TuiMarkerIconMode) {
    TuiMarkerIconMode["Link"] = "link";
    TuiMarkerIconMode["Primary"] = "primary";
    TuiMarkerIconMode["Warning"] = "warning";
    TuiMarkerIconMode["White"] = "white";
    TuiMarkerIconMode["Secondary"] = "secondary";
    TuiMarkerIconMode["Success"] = "success";
    TuiMarkerIconMode["Error"] = "error";
    TuiMarkerIconMode["OnDark"] = "onDark";
})(TuiMarkerIconMode || (TuiMarkerIconMode = {}));

/**
 * @deprecated use join type {@link TuiStatusT}
 */
var TuiStatus;
(function (TuiStatus) {
    TuiStatus["Default"] = "default";
    TuiStatus["Primary"] = "primary";
    TuiStatus["Custom"] = "custom";
    TuiStatus["Success"] = "success";
    TuiStatus["Error"] = "error";
    TuiStatus["Warning"] = "warning";
})(TuiStatus || (TuiStatus = {}));

/**
 * @deprecated use {@link TuiStepStateT}
 */
var TuiStepState;
(function (TuiStepState) {
    TuiStepState["Normal"] = "normal";
    TuiStepState["Error"] = "error";
    TuiStepState["Pass"] = "pass";
})(TuiStepState || (TuiStepState = {}));

/**
 * @deprecated use string values
 */
var TuiTextAlign;
(function (TuiTextAlign) {
    TuiTextAlign["Left"] = "left";
    TuiTextAlign["Center"] = "center";
    TuiTextAlign["Right"] = "right";
})(TuiTextAlign || (TuiTextAlign = {}));

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBorders, TuiDropdownPosition, TuiFileState, TuiMarkerIconMode, TuiStatus, TuiStepState, TuiTextAlign };
