// @taiga-ui/kit/esm2015/components/input-number/input-number.directive.js
var TuiInputNumberDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
let TuiInputNumberDirective = TuiInputNumberDirective_1 = class TuiInputNumberDirective extends TuiAbstractTextfieldHost {
    get value() {
        return this.host.computedValue;
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
    process(input) {
        this.input = input;
    }
    ngDoCheck() {
        if (!this.input) {
            return;
        }
        this.input.maxLength = this.host.calculatedMaxLength;
        this.input.inputMode = 'decimal';
    }
};
TuiInputNumberDirective = TuiInputNumberDirective_1 = __decorate([
    Directive({
        selector: 'tui-input-number',
        providers: [
            {
                provide: TUI_TEXTFIELD_HOST,
                useExisting: forwardRef(() => TuiInputNumberDirective_1),
            },
        ],
    })
], TuiInputNumberDirective);
export { TuiInputNumberDirective };
