// @taiga-ui/kit/esm2015/components/pdf-viewer/pdf-viewer.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiPdfViewerComponent } from './pdf-viewer.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
const DIALOG = new PolymorpheusComponent(TuiPdfViewerComponent);
const DEFAULT_OPTIONS = { label: ``, actions: `` };
let TuiPdfViewerService = class TuiPdfViewerService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.component = DIALOG;
        this.defaultOptions = DEFAULT_OPTIONS;
    }
    open(content, options = {}) {
        return super.open(content, options);
    }
};
TuiPdfViewerService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiPdfViewerService_Factory() { return new TuiPdfViewerService(i0.ɵɵinject(i1.TuiIdService)); }, token: TuiPdfViewerService, providedIn: "root" });
TuiPdfViewerService = __decorate([
    Injectable({
        providedIn: `root`,
    })
], TuiPdfViewerService);
export { TuiPdfViewerService };
