// @taiga-ui/kit/esm5/components/radio-labeled/radio-labeled.component.js
import { __decorate, __extends, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, forwardRef, HostBinding, Inject, Input, Optional, Self, ViewChild, } from '@angular/core';
import { NgControl } from '@angular/forms';
import { AbstractTuiNullableControl, TUI_DEFAULT_IDENTITY_MATCHER, TUI_FOCUSABLE_ITEM_ACCESSOR, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TuiBrightness, TuiModeDirective, TuiSizeL } from '@taiga-ui/core';
import { TuiRadioComponent } from '@taiga-ui/kit/components/radio';
var TuiRadioLabeledComponent = /** @class */ (function (_super) {
    __extends(TuiRadioLabeledComponent, _super);
    function TuiRadioLabeledComponent(control, changeDetectorRef, modeDirective) {
        var _this = _super.call(this, control, changeDetectorRef) || this;
        _this.modeDirective = modeDirective;
        _this.size = 'm';
        _this.identityMatcher = TUI_DEFAULT_IDENTITY_MATCHER;
        _this.pseudoDisabled = false;
        return _this;
    }
    TuiRadioLabeledComponent_1 = TuiRadioLabeledComponent;
    Object.defineProperty(TuiRadioLabeledComponent.prototype, "nativeFocusableElement", {
        get: function () {
            return this.radio ? this.radio.nativeFocusableElement : null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRadioLabeledComponent.prototype, "focused", {
        get: function () {
            return !!this.radio && this.radio.focused;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRadioLabeledComponent.prototype, "computedDisabled", {
        get: function () {
            return this.disabled || this.pseudoDisabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiRadioLabeledComponent.prototype, "mode", {
        get: function () {
            return this.modeDirective ? this.modeDirective.mode : null;
        },
        enumerable: true,
        configurable: true
    });
    TuiRadioLabeledComponent.prototype.onFocused = function (focused) {
        this.updateFocused(focused);
    };
    TuiRadioLabeledComponent.prototype.onHovered = function (hovered) {
        this.updateHovered(hovered);
    };
    TuiRadioLabeledComponent.prototype.onPressed = function (pressed) {
        this.updatePressed(pressed);
    };
    TuiRadioLabeledComponent.prototype.onModelChange = function (value) {
        this.updateValue(value);
    };
    var TuiRadioLabeledComponent_1;
    TuiRadioLabeledComponent.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }, { type: Inject, args: [NgControl,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
        { type: TuiModeDirective, decorators: [{ type: Optional }, { type: Inject, args: [TuiModeDirective,] }] }
    ]; };
    __decorate([
        ViewChild(TuiRadioComponent)
    ], TuiRadioLabeledComponent.prototype, "radio", void 0);
    __decorate([
        Input()
    ], TuiRadioLabeledComponent.prototype, "item", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size'),
        tuiDefaultProp()
    ], TuiRadioLabeledComponent.prototype, "size", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRadioLabeledComponent.prototype, "identityMatcher", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRadioLabeledComponent.prototype, "pseudoDisabled", void 0);
    __decorate([
        HostBinding('class._disabled')
    ], TuiRadioLabeledComponent.prototype, "computedDisabled", null);
    __decorate([
        HostBinding('attr.data-mode')
    ], TuiRadioLabeledComponent.prototype, "mode", null);
    TuiRadioLabeledComponent = TuiRadioLabeledComponent_1 = __decorate([
        Component({
            selector: 'tui-radio-labeled',
            template: "<label\n    class=\"t-wrapper\"\n    (tuiHoveredChange)=\"onHovered($event)\"\n    (tuiPressedChange)=\"onPressed($event)\"\n>\n    <tui-radio\n        [focusable]=\"focusable\"\n        [identityMatcher]=\"identityMatcher\"\n        [name]=\"computedName || ''\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [pseudoDisabled]=\"computedDisabled\"\n        [pseudoInvalid]=\"computedInvalid\"\n        [pseudoFocused]=\"pseudoFocused\"\n        [pseudoHovered]=\"computedHovered\"\n        [pseudoPressed]=\"computedPressed\"\n        [size]=\"size\"\n        [item]=\"item\"\n        [ngModel]=\"value\"\n        (ngModelChange)=\"onModelChange($event)\"\n        (focusedChange)=\"onFocused($event)\"\n    ></tui-radio>\n    <div class=\"t-content\">\n        <ng-content></ng-content>\n    </div>\n</label>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                {
                    provide: TUI_FOCUSABLE_ITEM_ACCESSOR,
                    useExisting: forwardRef(function () { return TuiRadioLabeledComponent_1; }),
                },
            ],
            styles: [":host{position:relative;display:block;font:var(--tui-font-text-s);line-height:1rem}:host._readonly{pointer-events:none}.t-wrapper{display:inline-flex;max-width:100%;cursor:pointer;vertical-align:top}:host._disabled .t-wrapper{cursor:default}:host[data-mode=onDark] .t-wrapper{color:var(--tui-text-01-night)}.t-content{word-wrap:break-word;min-width:0}:host[data-size='m'] .t-content{margin-left:.5rem}:host[data-size='l'] .t-content{font:var(--tui-font-text-m);margin-left:.75rem}"]
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __param(0, Inject(NgControl)),
        __param(1, Inject(ChangeDetectorRef)),
        __param(2, Optional()),
        __param(2, Inject(TuiModeDirective))
    ], TuiRadioLabeledComponent);
    return TuiRadioLabeledComponent;
}(AbstractTuiNullableControl));
export { TuiRadioLabeledComponent };
