// @taiga-ui/addon-preview/fesm5/taiga-ui-addon-preview-components-preview-dialog.js
import { __decorate, __param, __extends } from 'tslib';
import { Inject, HostListener, Component, ChangeDetectionStrategy, ViewEncapsulation, NgModule, ɵɵdefineInjectable, ɵɵinject, Injectable } from '@angular/core';
import { tuiSlideInTop } from '@taiga-ui/core';
import { POLYMORPHEUS_CONTEXT, PolymorpheusModule, PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { AbstractTuiDialogService, TUI_DIALOGS } from '@taiga-ui/cdk';
import { TuiIdService } from '@taiga-ui/cdk/services';

var TuiPreviewDialogComponent = /** @class */ (function () {
    function TuiPreviewDialogComponent(context) {
        this.context = context;
    }
    TuiPreviewDialogComponent.prototype.onKeyDownEsc = function () {
        this.context.$implicit.complete();
    };
    TuiPreviewDialogComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [POLYMORPHEUS_CONTEXT,] }] }
    ]; };
    __decorate([
        HostListener('document:keydown.esc')
    ], TuiPreviewDialogComponent.prototype, "onKeyDownEsc", null);
    TuiPreviewDialogComponent = __decorate([
        Component({
            selector: 'tui-preview-dialog',
            template: "<div\n    polymorpheus-outlet\n    class=\"t-dialog-content\"\n    @tuiSlideInTop\n    [content]=\"context.content\"\n    [context]=\"context\"\n></div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiSlideInTop],
            encapsulation: ViewEncapsulation.None,
            styles: [".t-dialog-content{position:fixed;width:100%;height:100%}tui-wrapper[data-appearance=preview-action]{background:rgba(104,104,104,.96);color:var(--tui-text-01-night)}tui-wrapper[data-appearance=preview-action][data-state=hovered]{background:rgba(159,159,159,.86)}tui-wrapper[data-appearance=preview-action][data-state=pressed]{background:rgba(159,159,159,.75)}tui-wrapper[data-appearance=preview-action][data-state=disabled]{opacity:var(--tui-disabled-opacity)}"]
        }),
        __param(0, Inject(POLYMORPHEUS_CONTEXT))
    ], TuiPreviewDialogComponent);
    return TuiPreviewDialogComponent;
}());

var TuiPreviewDialogModule = /** @class */ (function () {
    function TuiPreviewDialogModule() {
    }
    TuiPreviewDialogModule = __decorate([
        NgModule({
            imports: [PolymorpheusModule],
            declarations: [TuiPreviewDialogComponent],
            exports: [TuiPreviewDialogComponent],
        })
    ], TuiPreviewDialogModule);
    return TuiPreviewDialogModule;
}());

/**
 * @deprecated: use {@link TuiPreviewDialogService}
 * TODO: remove in v3.0
 */
var PreviewDialogService = /** @class */ (function (_super) {
    __extends(PreviewDialogService, _super);
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function PreviewDialogService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultOptions = {};
        _this.component = new PolymorpheusComponent(TuiPreviewDialogComponent);
        return _this;
    }
    PreviewDialogService.ɵprov = ɵɵdefineInjectable({ factory: function PreviewDialogService_Factory() { return new PreviewDialogService(ɵɵinject(TuiIdService)); }, token: PreviewDialogService, providedIn: "root" });
    PreviewDialogService = __decorate([
        Injectable({ providedIn: "root" })
        // eslint-disable-next-line @typescript-eslint/naming-convention
    ], PreviewDialogService);
    return PreviewDialogService;
}(AbstractTuiDialogService));
var TuiPreviewDialogService = /** @class */ (function (_super) {
    __extends(TuiPreviewDialogService, _super);
    function TuiPreviewDialogService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiPreviewDialogService.ɵprov = ɵɵdefineInjectable({ factory: function TuiPreviewDialogService_Factory() { return new TuiPreviewDialogService(ɵɵinject(TuiIdService)); }, token: TuiPreviewDialogService, providedIn: "root" });
    TuiPreviewDialogService = __decorate([
        Injectable({ providedIn: "root" })
    ], TuiPreviewDialogService);
    return TuiPreviewDialogService;
}(PreviewDialogService));
var PREVIEW_DIALOG_PROVIDER = [
    {
        provide: TUI_DIALOGS,
        useExisting: TuiPreviewDialogService,
        multi: true,
    },
    {
        provide: TUI_DIALOGS,
        useExisting: PreviewDialogService,
        multi: true,
    },
];

/**
 * Generated bundle index. Do not edit.
 */

export { PREVIEW_DIALOG_PROVIDER, PreviewDialogService, TuiPreviewDialogComponent, TuiPreviewDialogModule, TuiPreviewDialogService };
