// @taiga-ui/core/esm5/components/alert/alert.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TUI_ALERTS } from '@taiga-ui/cdk';
import { TuiNotificationModule } from '@taiga-ui/core/components/notification';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiAlertComponent } from './alert.component';
import { TuiAlertService } from './alert.service';
var TuiAlertModule = /** @class */ (function () {
    function TuiAlertModule() {
    }
    TuiAlertModule = __decorate([
        NgModule({
            imports: [CommonModule, PolymorpheusModule, TuiNotificationModule],
            declarations: [TuiAlertComponent],
            exports: [TuiAlertComponent],
            entryComponents: [TuiAlertComponent],
            providers: [
                {
                    provide: TUI_ALERTS,
                    useExisting: TuiAlertService,
                    multi: true,
                },
            ],
        })
    ], TuiAlertModule);
    return TuiAlertModule;
}());
export { TuiAlertModule };
/** @deprecated use {@link TuiAlertModule} */
var TuiNotificationsModule = /** @class */ (function () {
    function TuiNotificationsModule() {
    }
    TuiNotificationsModule = __decorate([
        NgModule({ imports: [TuiAlertModule] })
    ], TuiNotificationsModule);
    return TuiNotificationsModule;
}());
export { TuiNotificationsModule };
