// @taiga-ui/core/esm5/services/format-date.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable, LOCALE_ID } from '@angular/core';
import { of } from 'rxjs';
import * as i0 from "@angular/core";
var TuiFormatDateService = /** @class */ (function () {
    function TuiFormatDateService(locale) {
        this.locale = locale;
    }
    TuiFormatDateService.prototype.format = function (timestamp) {
        return of(new Date(timestamp).toLocaleTimeString(this.locale, {
            hour: "numeric",
            minute: "2-digit",
        }));
    };
    TuiFormatDateService.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
    ]; };
    TuiFormatDateService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiFormatDateService_Factory() { return new TuiFormatDateService(i0.ɵɵinject(i0.LOCALE_ID)); }, token: TuiFormatDateService, providedIn: "root" });
    TuiFormatDateService = __decorate([
        Injectable({
            providedIn: "root",
        }),
        __param(0, Inject(LOCALE_ID))
    ], TuiFormatDateService);
    return TuiFormatDateService;
}());
export { TuiFormatDateService };
