// @taiga-ui/core/esm5/services/hint.service.js
import { __decorate, __extends, __read, __spread } from "tslib";
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Service for displaying hints/tooltips
 */
var TuiHintService = /** @class */ (function (_super) {
    __extends(TuiHintService, _super);
    function TuiHintService() {
        var _this = _super.call(this, []) || this;
        /**
         * TODO:
         * We need the following logic for desribedBy
         * move it into another service that can register hints and
         * manage it using TuiHintService inside
         */
        _this.directives = [];
        return _this;
    }
    TuiHintService.prototype.add = function (directive) {
        this.next(this.value.concat(directive));
    };
    TuiHintService.prototype.remove = function (directive) {
        if (this.value.includes(directive)) {
            this.next(this.value.filter(function (hint) { return hint !== directive; }));
        }
    };
    TuiHintService.prototype.register = function (directive) {
        this.directives = __spread(this.directives, [directive]);
    };
    TuiHintService.prototype.unregister = function (directive) {
        this.remove(directive);
        this.directives = this.directives.filter(function (dir) { return dir !== directive; });
    };
    TuiHintService.prototype.showHintForId = function (id) {
        var directive = this.findDirectiveWithHintId(id);
        if (directive) {
            this.add(directive);
        }
    };
    TuiHintService.prototype.hideHintForId = function (id) {
        var directive = this.findDirectiveWithHintId(id);
        if (directive) {
            this.remove(directive);
        }
    };
    TuiHintService.prototype.findDirectiveWithHintId = function (id) {
        return this.directives.find(function (directive) { return directive.tuiHintId === id; });
    };
    TuiHintService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiHintService_Factory() { return new TuiHintService(); }, token: TuiHintService, providedIn: "root" });
    TuiHintService = __decorate([
        Injectable({
            providedIn: "root",
        })
    ], TuiHintService);
    return TuiHintService;
}(BehaviorSubject));
export { TuiHintService };
