// @taiga-ui/core/esm5/components/root/root.component.js
import { __decorate, __param, __read, __spread } from "tslib";
import { DOCUMENT } from '@angular/common';
import { ChangeDetectionStrategy, Component, ElementRef, Inject } from '@angular/core';
import { TUI_DIALOGS, TUI_IS_MOBILE, TUI_VERSION, tuiAssert } from '@taiga-ui/cdk';
import { TUI_IS_MOBILE_RES_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_ANIMATIONS_DURATION, TUI_ASSERT_ENABLED, TUI_IS_MOBILE_RES, TUI_THEME, } from '@taiga-ui/core/tokens';
import { combineLatest, Observable, of } from 'rxjs';
import { map } from 'rxjs/operators';
// @dynamic
var TuiRootComponent = /** @class */ (function () {
    function TuiRootComponent(duration, elementRef, dialogs, isMobile, enabled, isMobileRes$, _a, theme) {
        var body = _a.body;
        this.duration = duration;
        this.elementRef = elementRef;
        this.dialogs = dialogs;
        this.isMobile = isMobile;
        this.isMobileRes$ = isMobileRes$;
        this.scrollbars$ = this.dialogs.length && !this.isMobile
            ? combineLatest(__spread(this.dialogs)).pipe(map(function (dialogs) { return !dialogs.some(function (_a) {
                var length = _a.length;
                return length;
            }); }))
            : of(!this.isMobile);
        tuiAssert.enabled = enabled;
        body.setAttribute('data-tui-theme', theme.toLowerCase());
    }
    TuiRootComponent.ctorParameters = function () { return [
        { type: Number, decorators: [{ type: Inject, args: [TUI_ANIMATIONS_DURATION,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_DIALOGS,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_MOBILE,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_ASSERT_ENABLED,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_IS_MOBILE_RES,] }] },
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: String, decorators: [{ type: Inject, args: [TUI_THEME,] }] }
    ]; };
    TuiRootComponent = __decorate([
        Component({
            selector: 'tui-root',
            template: "<tui-scroll-controls\n    *ngIf=\"scrollbars$ | async\"\n    class=\"t-scrollbar\"\n></tui-scroll-controls>\n<tui-svg-defs-host></tui-svg-defs-host>\n<tui-dropdown-host>\n    <div class=\"t-content\">\n        <ng-content></ng-content>\n    </div>\n    <ng-content select=\"tuiOverContent\"></ng-content>\n    <tui-dialog-host></tui-dialog-host>\n    <ng-content select=\"tuiOverDialogs\"></ng-content>\n    <tui-alert-host></tui-alert-host>\n    <ng-content select=\"tuiOverAlerts\"></ng-content>\n    <!-- TODO: 3.0 Remove-->\n    <ng-content select=\"tuiOverNotifications\"></ng-content>\n</tui-dropdown-host>\n<ng-content select=\"tuiOverPortals\"></ng-content>\n<tui-hints-host></tui-hints-host>\n<ng-content select=\"tuiOverHints\"></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [TUI_IS_MOBILE_RES_PROVIDER],
            host: {
                'data-tui-version': TUI_VERSION,
                '[style.--tui-duration]': 'duration + "ms"',
                '($.class._mobile)': 'isMobileRes$',
            },
            styles: ["@keyframes tuiPresent{to{content:'1'}}:host{position:relative;display:block;flex:1;-webkit-tap-highlight-color:transparent}.t-scrollbar{position:fixed;top:0;left:0;bottom:0;right:0;margin:0}.t-content{position:relative;z-index:0;height:100%}"]
        }),
        __param(0, Inject(TUI_ANIMATIONS_DURATION)),
        __param(1, Inject(ElementRef)),
        __param(2, Inject(TUI_DIALOGS)),
        __param(3, Inject(TUI_IS_MOBILE)),
        __param(4, Inject(TUI_ASSERT_ENABLED)),
        __param(5, Inject(TUI_IS_MOBILE_RES)),
        __param(6, Inject(DOCUMENT)),
        __param(7, Inject(TUI_THEME))
    ], TuiRootComponent);
    return TuiRootComponent;
}());
export { TuiRootComponent };
