// @taiga-ui/core/esm2015/pipes/calendar-sheet/calendar-sheet.pipe.js
import { __decorate, __param } from "tslib";
import { Inject, Pipe } from '@angular/core';
import { DAYS_IN_WEEK } from '@taiga-ui/cdk';
import { TUI_FIRST_DAY_OF_WEEK } from '@taiga-ui/core/tokens';
import { getDayFromMonthRowCol } from './utils';
const CALENDAR_ROWS_COUNT = 6;
let TuiCalendarSheetPipe = class TuiCalendarSheetPipe {
    constructor(firstDayOfWeek) {
        this.firstDayOfWeek = firstDayOfWeek;
        this.currentMonth = null;
        this.currentSheet = [];
    }
    transform(month, showAdjacentDays = false) {
        var _a;
        if ((_a = this.currentMonth) === null || _a === void 0 ? void 0 : _a.monthSame(month)) {
            return this.currentSheet;
        }
        const sheet = [];
        for (let rowIndex = 0; rowIndex < CALENDAR_ROWS_COUNT; rowIndex++) {
            const row = [];
            for (let colIndex = 0; colIndex < DAYS_IN_WEEK; colIndex++) {
                const day = getDayFromMonthRowCol({
                    month,
                    rowIndex,
                    colIndex,
                    firstDayOfWeek: this.firstDayOfWeek,
                });
                const isPrevMonthDay = (day, relativeToMonth = month) => day.year < relativeToMonth.year || day.month < relativeToMonth.month;
                const isNextMonthDay = (day, relativeToMonth = month) => day.year > relativeToMonth.year || day.month > relativeToMonth.month;
                if (isPrevMonthDay(day) && !showAdjacentDays) {
                    continue;
                }
                if (isNextMonthDay(day) && !showAdjacentDays) {
                    break;
                }
                row.push(day);
            }
            sheet.push(row);
        }
        this.currentSheet = sheet.filter(row => row.length);
        this.currentMonth = month;
        return this.currentSheet;
    }
};
TuiCalendarSheetPipe.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_FIRST_DAY_OF_WEEK,] }] }
];
TuiCalendarSheetPipe = __decorate([
    Pipe({
        name: `tuiCalendarSheet`,
    }),
    __param(0, Inject(TUI_FIRST_DAY_OF_WEEK))
], TuiCalendarSheetPipe);
export { TuiCalendarSheetPipe };
