// @taiga-ui/addon-preview/esm5/components/preview-dialog/preview-dialog.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, HostListener, Inject, ViewEncapsulation, } from '@angular/core';
import { tuiSlideInTop } from '@taiga-ui/core';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
var TuiPreviewDialogComponent = /** @class */ (function () {
    function TuiPreviewDialogComponent(context) {
        this.context = context;
    }
    TuiPreviewDialogComponent.prototype.onKeyDownEsc = function () {
        this.context.$implicit.complete();
    };
    TuiPreviewDialogComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [POLYMORPHEUS_CONTEXT,] }] }
    ]; };
    __decorate([
        HostListener('document:keydown.esc')
    ], TuiPreviewDialogComponent.prototype, "onKeyDownEsc", null);
    TuiPreviewDialogComponent = __decorate([
        Component({
            selector: 'tui-preview-dialog',
            template: "<div\n    polymorpheus-outlet\n    class=\"t-dialog-content\"\n    @tuiSlideInTop\n    [content]=\"context.content\"\n    [context]=\"context\"\n></div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiSlideInTop],
            encapsulation: ViewEncapsulation.None,
            styles: [".t-dialog-content{position:fixed;width:100%;height:100%}tui-wrapper[data-appearance=preview-action]{background:rgba(104,104,104,.96);color:var(--tui-text-01-night)}tui-wrapper[data-appearance=preview-action][data-state=hovered]{background:rgba(159,159,159,.86)}tui-wrapper[data-appearance=preview-action][data-state=pressed]{background:rgba(159,159,159,.75)}tui-wrapper[data-appearance=preview-action][data-state=disabled]{opacity:var(--tui-disabled-opacity)}"]
        }),
        __param(0, Inject(POLYMORPHEUS_CONTEXT))
    ], TuiPreviewDialogComponent);
    return TuiPreviewDialogComponent;
}());
export { TuiPreviewDialogComponent };
