// @taiga-ui/addon-preview/esm5/components/preview-dialog/preview-dialog.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractTuiDialogService, TUI_DIALOGS } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiPreviewDialogComponent } from './preview-dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
/**
 * @deprecated: use {@link TuiPreviewDialogService}
 * TODO: remove in v3.0
 */
var PreviewDialogService = /** @class */ (function (_super) {
    __extends(PreviewDialogService, _super);
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function PreviewDialogService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultOptions = {};
        _this.component = new PolymorpheusComponent(TuiPreviewDialogComponent);
        return _this;
    }
    PreviewDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PreviewDialogService_Factory() { return new PreviewDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: PreviewDialogService, providedIn: "root" });
    PreviewDialogService = __decorate([
        Injectable({ providedIn: "root" })
        // eslint-disable-next-line @typescript-eslint/naming-convention
    ], PreviewDialogService);
    return PreviewDialogService;
}(AbstractTuiDialogService));
export { PreviewDialogService };
var TuiPreviewDialogService = /** @class */ (function (_super) {
    __extends(TuiPreviewDialogService, _super);
    function TuiPreviewDialogService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiPreviewDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiPreviewDialogService_Factory() { return new TuiPreviewDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: TuiPreviewDialogService, providedIn: "root" });
    TuiPreviewDialogService = __decorate([
        Injectable({ providedIn: "root" })
    ], TuiPreviewDialogService);
    return TuiPreviewDialogService;
}(PreviewDialogService));
export { TuiPreviewDialogService };
export var PREVIEW_DIALOG_PROVIDER = [
    {
        provide: TUI_DIALOGS,
        useExisting: TuiPreviewDialogService,
        multi: true,
    },
    {
        provide: TUI_DIALOGS,
        useExisting: PreviewDialogService,
        multi: true,
    },
];
