// @taiga-ui/addon-doc/esm5/internal/header/header.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { Router } from '@angular/router';
import { TuiSwipeService } from '@taiga-ui/cdk';
import { merge, Subject } from 'rxjs';
import { distinctUntilChanged, filter, map, mapTo, startWith } from 'rxjs/operators';
import { TUI_DOC_MENU_TEXT } from '../../tokens/i18n';
import { TUI_DOC_LOGO } from '../../tokens/logo';
var TuiDocHeaderComponent = /** @class */ (function () {
    function TuiDocHeaderComponent(logo, menu, router, swipes$) {
        this.logo = logo;
        this.menu = menu;
        this.router = router;
        this.swipes$ = swipes$;
        this.stream$ = new Subject();
        this.open$ = merge(this.router.events.pipe(mapTo(false)), this.stream$, this.swipes$.pipe(filter(function (swipe) { return swipe.direction === 'left' || swipe.direction === 'right'; }), map(function (swipe) { return swipe.direction === 'right'; }))).pipe(startWith(false), distinctUntilChanged());
    }
    TuiDocHeaderComponent.prototype.onClick = function () {
        this.stream$.next(true);
    };
    TuiDocHeaderComponent.prototype.onActiveZone = function (active) {
        if (!active) {
            this.stream$.next(false);
        }
    };
    TuiDocHeaderComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_DOC_LOGO,] }] },
        { type: String, decorators: [{ type: Inject, args: [TUI_DOC_MENU_TEXT,] }] },
        { type: Router, decorators: [{ type: Inject, args: [Router,] }] },
        { type: TuiSwipeService, decorators: [{ type: Inject, args: [TuiSwipeService,] }] }
    ]; };
    TuiDocHeaderComponent = __decorate([
        Component({
            selector: 'header[tuiDocHeader]',
            template: "<button\n    tuiIconButton\n    icon=\"tuiIconMenuLarge\"\n    appearance=\"icon\"\n    class=\"t-menu\"\n    [title]=\"menu\"\n    (click)=\"onClick()\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-doc-navigation\n        *tuiSidebar=\"!!(open$ | async)\"\n        class=\"t-navigation\"\n    ></tui-doc-navigation>\n</button>\n<div\n    polymorpheus-outlet\n    class=\"t-logo\"\n    [content]=\"logo\"\n>\n    <ng-template let-src>\n        <img\n            alt=\"Logo\"\n            class=\"t-img\"\n            [src]=\"src\"\n        />\n    </ng-template>\n</div>\n<ng-content></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{box-shadow:0 .125rem 1rem rgba(0,0,0,.08);position:fixed;top:0;left:0;right:0;z-index:1;display:flex;height:3.9375rem;align-items:center;padding:0 1.25rem;background:var(--tui-base-01);border-bottom:1px solid var(--tui-base-03)}.t-logo{margin-right:auto;font:var(--tui-font-heading-5)}.t-img{display:block}.t-menu{margin-left:-1.25rem}@media screen and (min-width:48em){.t-menu{display:none}}.t-navigation{min-height:100%}"]
        }),
        __param(0, Inject(TUI_DOC_LOGO)),
        __param(1, Inject(TUI_DOC_MENU_TEXT)),
        __param(2, Inject(Router)),
        __param(3, Inject(TuiSwipeService))
    ], TuiDocHeaderComponent);
    return TuiDocHeaderComponent;
}());
export { TuiDocHeaderComponent };
