// @taiga-ui/core/fesm2015/taiga-ui-core-directives-pointer-hint.js
import { __decorate, __param } from 'tslib';
import { InjectionToken, ElementRef, Inject, Input, Directive, NgModule } from '@angular/core';
import { typedFromEvent, TuiDestroyService, TuiHoveredService, tuiDefaultProp, tuiRequiredSetter } from '@taiga-ui/cdk';
import { TUI_ABSTRACT_HINT_DEFAULT_OPTIONS, AbstractTuiHint } from '@taiga-ui/core/abstract';
import { TuiHintService } from '@taiga-ui/core/services';
import { of } from 'rxjs';
import { filter, startWith, distinctUntilChanged, switchMap, delay, takeUntil } from 'rxjs/operators';

/** Default values for pointer hint options */
const TUI_POINTER_HINT_DEFAULT_OPTIONS = Object.assign(Object.assign({}, TUI_ABSTRACT_HINT_DEFAULT_OPTIONS), { tuiHintShowDelay: 0, tuiHintHideDelay: 0 });
const TUI_POINTER_HINT_OPTIONS = new InjectionToken(`Default parameters for pointer hint directive`, {
    factory: () => TUI_POINTER_HINT_DEFAULT_OPTIONS,
});
const tuiPointerHintOptionsProvider = (options) => ({
    provide: TUI_POINTER_HINT_OPTIONS,
    useValue: Object.assign(Object.assign({}, TUI_POINTER_HINT_DEFAULT_OPTIONS), options),
});

let TuiPointerHintDirective = class TuiPointerHintDirective extends AbstractTuiHint {
    constructor(elementRef, hintService, destroy$, hoveredService, options) {
        super(elementRef, hintService, null, options);
        this.destroy$ = destroy$;
        this.options = options;
        this.currentMouseRect = this.mousePositionToClientRect();
        this.tuiHintShowDelay = this.options.tuiHintShowDelay;
        this.tuiHintHideDelay = this.options.tuiHintHideDelay;
        this.content = '';
        const hint$ = hoveredService.createHovered$(this.elementRef.nativeElement).pipe(filter(() => !!this.content), startWith(false), distinctUntilChanged());
        hint$
            .pipe(switchMap(visible => of(visible).pipe(delay(visible ? this.tuiHintShowDelay : this.tuiHintHideDelay))), takeUntil(destroy$))
            .subscribe({
            next: visible => {
                if (visible) {
                    this.showTooltip();
                }
                else {
                    this.hideTooltip();
                }
            },
            complete: () => {
                this.hideTooltip();
            },
        });
        this.initMouseMoveSubscription();
    }
    // TODO: 3.0 Remove null
    set tuiPointerHint(value) {
        if (!value) {
            this.hideTooltip();
            this.content = '';
            return;
        }
        this.content = value;
    }
    getElementClientRect() {
        return this.currentMouseRect;
    }
    initMouseMoveSubscription() {
        const mouseMove$ = typedFromEvent(this.elementRef.nativeElement, 'mousemove');
        mouseMove$.pipe(takeUntil(this.destroy$)).subscribe(({ clientX, clientY }) => {
            this.currentMouseRect = this.mousePositionToClientRect(clientX, clientY);
        });
    }
    mousePositionToClientRect(x = 0, y = 0) {
        return {
            left: x,
            right: x,
            top: y,
            bottom: y,
            width: 0,
            height: 0,
        };
    }
};
TuiPointerHintDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] },
    { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] },
    { type: TuiHoveredService, decorators: [{ type: Inject, args: [TuiHoveredService,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [TUI_POINTER_HINT_OPTIONS,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiPointerHintDirective.prototype, "tuiHintShowDelay", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiPointerHintDirective.prototype, "tuiHintHideDelay", void 0);
__decorate([
    Input(),
    tuiRequiredSetter()
], TuiPointerHintDirective.prototype, "tuiPointerHint", null);
TuiPointerHintDirective = __decorate([
    Directive({
        selector: '[tuiPointerHint]:not(ng-container)',
        providers: [TuiDestroyService],
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(TuiHintService)),
    __param(2, Inject(TuiDestroyService)),
    __param(3, Inject(TuiHoveredService)),
    __param(4, Inject(TUI_POINTER_HINT_OPTIONS))
], TuiPointerHintDirective);

let TuiPointerHintModule = class TuiPointerHintModule {
};
TuiPointerHintModule = __decorate([
    NgModule({
        declarations: [TuiPointerHintDirective],
        exports: [TuiPointerHintDirective],
    })
], TuiPointerHintModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_POINTER_HINT_DEFAULT_OPTIONS, TUI_POINTER_HINT_OPTIONS, TuiPointerHintDirective, TuiPointerHintModule, tuiPointerHintOptionsProvider };
