// @taiga-ui/core/fesm2015/taiga-ui-core-components-error.js
import { __decorate, __param } from 'tslib';
import { Inject, Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { TuiValidationError, tuiDefaultProp } from '@taiga-ui/cdk';
import { tuiHeightCollapse, tuiFadeIn } from '@taiga-ui/core/animations';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_ANIMATION_OPTIONS, TUI_MODE, TUI_DEFAULT_ERROR_MESSAGE } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
import { CommonModule } from '@angular/common';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';

let TuiErrorComponent = class TuiErrorComponent {
    constructor(options, mode$, defaultErrorMessage$) {
        this.options = options;
        this.mode$ = mode$;
        this.defaultErrorMessage$ = defaultErrorMessage$;
        this.error = null;
        this.animation = Object.assign({ value: '' }, this.options);
    }
    set errorSetter(error) {
        this.error = typeof error === 'string' ? new TuiValidationError(error) : error;
    }
};
TuiErrorComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_ANIMATION_OPTIONS,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_DEFAULT_ERROR_MESSAGE,] }] }
];
__decorate([
    Input('error'),
    tuiDefaultProp()
], TuiErrorComponent.prototype, "errorSetter", null);
TuiErrorComponent = __decorate([
    Component({
        selector: 'tui-error',
        template: "<div\n    *ngIf=\"error\"\n    polymorpheus-outlet\n    automation-id=\"tui-error__text\"\n    class=\"t-message-text\"\n    [@tuiHeightCollapse]=\"animation\"\n    [@tuiFadeIn]=\"animation\"\n    [class.t-message-text_light]=\"(mode$ | async) === 'onDark'\"\n    [content]=\"error.message || (defaultErrorMessage$ | async)\"\n    [context]=\"error.context || {}\"\n></div>\n",
        providers: [MODE_PROVIDER],
        changeDetection: ChangeDetectionStrategy.OnPush,
        animations: [tuiHeightCollapse, tuiFadeIn],
        styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);display:block;color:var(--tui-negative);word-wrap:break-word}.t-message-text{margin-top:.25rem;white-space:pre-wrap}.t-message-text_light{color:var(--tui-negative-night)}"]
    }),
    __param(0, Inject(TUI_ANIMATION_OPTIONS)),
    __param(1, Inject(TUI_MODE)),
    __param(2, Inject(TUI_DEFAULT_ERROR_MESSAGE))
], TuiErrorComponent);

let TuiErrorModule = class TuiErrorModule {
};
TuiErrorModule = __decorate([
    NgModule({
        imports: [CommonModule, PolymorpheusModule],
        declarations: [TuiErrorComponent],
        exports: [TuiErrorComponent],
    })
], TuiErrorModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiErrorComponent, TuiErrorModule };
