// @taiga-ui/core/fesm2015/taiga-ui-core-pipes-format-date.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TuiFormatDateService } from '@taiga-ui/core/services';

let TuiFormatDatePipe = class TuiFormatDatePipe {
    constructor(service) {
        this.service = service;
    }
    transform(timestampOrDate) {
        return this.service.format(timestampOrDate.valueOf());
    }
};
TuiFormatDatePipe.ctorParameters = () => [
    { type: TuiFormatDateService, decorators: [{ type: Inject, args: [TuiFormatDateService,] }] }
];
TuiFormatDatePipe = __decorate([
    Pipe({
        name: `tuiFormatDate`,
    }),
    __param(0, Inject(TuiFormatDateService))
], TuiFormatDatePipe);

let TuiFormatDatePipeModule = class TuiFormatDatePipeModule {
};
TuiFormatDatePipeModule = __decorate([
    NgModule({
        declarations: [TuiFormatDatePipe],
        exports: [TuiFormatDatePipe],
    })
], TuiFormatDatePipeModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFormatDatePipe, TuiFormatDatePipeModule };
