// @taiga-ui/core/fesm2015/taiga-ui-core-enums.js
var TuiAppearance;
(function (TuiAppearance) {
    TuiAppearance["Icon"] = "icon";
    TuiAppearance["Primary"] = "primary";
    TuiAppearance["Accent"] = "accent";
    TuiAppearance["Secondary"] = "secondary";
    TuiAppearance["SecondaryDestructive"] = "secondary-destructive";
    TuiAppearance["Table"] = "table";
    TuiAppearance["Textfield"] = "textfield";
    TuiAppearance["Flat"] = "flat";
    TuiAppearance["Outline"] = "outline";
    TuiAppearance["Whiteblock"] = "whiteblock";
    TuiAppearance["WhiteblockActive"] = "whiteblock-active";
})(TuiAppearance || (TuiAppearance = {}));

/**
 * @deprecated
 */
var TuiBaseColor;
(function (TuiBaseColor) {
    TuiBaseColor["Primary"] = "primary";
    TuiBaseColor["Secondary"] = "secondary";
    TuiBaseColor["Success"] = "success";
    TuiBaseColor["Error"] = "error";
})(TuiBaseColor || (TuiBaseColor = {}));

/**
 * @deprecated use string values
 */
var TuiButtonShape;
(function (TuiButtonShape) {
    TuiButtonShape["Square"] = "square";
    TuiButtonShape["Rounded"] = "rounded";
})(TuiButtonShape || (TuiButtonShape = {}));

/**
 * @deprecated use join type {@link TuiDecimalT}
 */
var TuiDecimal;
(function (TuiDecimal) {
    TuiDecimal["NotZero"] = "not-zero";
    TuiDecimal["Always"] = "always";
    TuiDecimal["Never"] = "never";
})(TuiDecimal || (TuiDecimal = {}));

var TuiDropdownAnimation;
(function (TuiDropdownAnimation) {
    TuiDropdownAnimation["FadeInBottom"] = "fadeInBottom";
    TuiDropdownAnimation["FadeInTop"] = "fadeInTop";
})(TuiDropdownAnimation || (TuiDropdownAnimation = {}));

/**
 * @deprecated use join type {@link TuiDropdownWidthT}
 */
var TuiDropdownWidth;
(function (TuiDropdownWidth) {
    TuiDropdownWidth["Fixed"] = "fixed";
    TuiDropdownWidth["Min"] = "min";
    TuiDropdownWidth["Auto"] = "auto";
})(TuiDropdownWidth || (TuiDropdownWidth = {}));

/**
 * @deprecated use join type {@link TuiHintModeT}
 */
var TuiHintMode;
(function (TuiHintMode) {
    TuiHintMode["Error"] = "error";
    TuiHintMode["OnDark"] = "onDark";
    TuiHintMode["Overflow"] = "overflow";
})(TuiHintMode || (TuiHintMode = {}));

/**
 * @internal used in calendar, year picker and wrapper
 */
var TuiInteractiveState;
(function (TuiInteractiveState) {
    TuiInteractiveState["Disabled"] = "disabled";
    TuiInteractiveState["Readonly"] = "readonly";
    // TODO: 3.0 Rename to `active`
    TuiInteractiveState["Pressed"] = "pressed";
    // TODO: 3.0 Rename to `hover`
    TuiInteractiveState["Hovered"] = "hovered";
})(TuiInteractiveState || (TuiInteractiveState = {}));

/**
 * @deprecated use string values
 */
var TuiLinkMode;
(function (TuiLinkMode) {
    TuiLinkMode["Positive"] = "positive";
    TuiLinkMode["Negative"] = "negative";
})(TuiLinkMode || (TuiLinkMode = {}));

var TuiNotification;
(function (TuiNotification) {
    TuiNotification["Info"] = "info";
    TuiNotification["Warning"] = "warning";
    TuiNotification["Success"] = "success";
    TuiNotification["Error"] = "error";
})(TuiNotification || (TuiNotification = {}));

/**
 * @deprecated use join type {@link TuiOrientationT}
 */
var TuiOrientation;
(function (TuiOrientation) {
    TuiOrientation["Vertical"] = "vertical";
    TuiOrientation["Horizontal"] = "horizontal";
})(TuiOrientation || (TuiOrientation = {}));

/**
 * @internal used in calendar and year picker
 */
var TuiRangeState;
(function (TuiRangeState) {
    TuiRangeState["Start"] = "start";
    TuiRangeState["End"] = "end";
    TuiRangeState["Single"] = "single";
})(TuiRangeState || (TuiRangeState = {}));

/**
 * @deprecated
 */
var TuiSupportColor;
(function (TuiSupportColor) {
    TuiSupportColor["Mustard"] = "support-01";
    TuiSupportColor["Texas"] = "support-02";
    TuiSupportColor["Tan"] = "support-03";
    TuiSupportColor["Salmon"] = "support-04";
    TuiSupportColor["Sienna"] = "support-05";
    TuiSupportColor["Bittersweet"] = "support-06";
    TuiSupportColor["Pinkie"] = "support-07";
    TuiSupportColor["Charm"] = "support-08";
    TuiSupportColor["Amethist"] = "support-09";
    TuiSupportColor["Helio"] = "support-10";
    TuiSupportColor["Lilac"] = "support-11";
    TuiSupportColor["Malibu"] = "support-12";
    TuiSupportColor["Havelock"] = "support-13";
    TuiSupportColor["Picton"] = "support-14";
    TuiSupportColor["Mint"] = "support-15";
    TuiSupportColor["Fountain"] = "support-16";
    TuiSupportColor["Puertorico"] = "support-17";
    TuiSupportColor["Bay"] = "support-18";
    TuiSupportColor["Forest"] = "support-19";
    TuiSupportColor["York"] = "support-20";
    TuiSupportColor["Feijoa"] = "support-21";
})(TuiSupportColor || (TuiSupportColor = {}));

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAppearance, TuiBaseColor, TuiButtonShape, TuiDecimal, TuiDropdownAnimation, TuiDropdownWidth, TuiHintMode, TuiInteractiveState, TuiLinkMode, TuiNotification, TuiOrientation, TuiRangeState, TuiSupportColor };
