// @taiga-ui/core/esm5/enums/support-color.js
/**
 * @deprecated
 */
export var TuiSupportColor;
(function (TuiSupportColor) {
    TuiSupportColor["Mustard"] = "support-01";
    TuiSupportColor["Texas"] = "support-02";
    TuiSupportColor["Tan"] = "support-03";
    TuiSupportColor["Salmon"] = "support-04";
    TuiSupportColor["Sienna"] = "support-05";
    TuiSupportColor["Bittersweet"] = "support-06";
    TuiSupportColor["Pinkie"] = "support-07";
    TuiSupportColor["Charm"] = "support-08";
    TuiSupportColor["Amethist"] = "support-09";
    TuiSupportColor["Helio"] = "support-10";
    TuiSupportColor["Lilac"] = "support-11";
    TuiSupportColor["Malibu"] = "support-12";
    TuiSupportColor["Havelock"] = "support-13";
    TuiSupportColor["Picton"] = "support-14";
    TuiSupportColor["Mint"] = "support-15";
    TuiSupportColor["Fountain"] = "support-16";
    TuiSupportColor["Puertorico"] = "support-17";
    TuiSupportColor["Bay"] = "support-18";
    TuiSupportColor["Forest"] = "support-19";
    TuiSupportColor["York"] = "support-20";
    TuiSupportColor["Feijoa"] = "support-21";
})(TuiSupportColor || (TuiSupportColor = {}));
