// @taiga-ui/cdk/esm2015/directives/portal/portal.directive.js
import { __decorate, __param } from "tslib";
import { Directive, EmbeddedViewRef, Inject, Input, OnDestroy, TemplateRef, } from '@angular/core';
import { TuiPortalService } from '@taiga-ui/cdk/components/portal-host';
let TuiPortalDirective = class TuiPortalDirective {
    constructor(templateRef, portalService) {
        this.templateRef = templateRef;
        this.portalService = portalService;
    }
    set tuiPortal(show) {
        var _a;
        (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
        if (show) {
            this.viewRef = this.portalService.addTemplate(this.templateRef);
        }
    }
    ngOnDestroy() {
        var _a;
        (_a = this.viewRef) === null || _a === void 0 ? void 0 : _a.destroy();
    }
};
TuiPortalDirective.ctorParameters = () => [
    { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }] },
    { type: TuiPortalService, decorators: [{ type: Inject, args: [TuiPortalService,] }] }
];
__decorate([
    Input()
], TuiPortalDirective.prototype, "tuiPortal", null);
TuiPortalDirective = __decorate([
    Directive({
        selector: '[tuiPortal]',
    }),
    __param(0, Inject(TemplateRef)),
    __param(1, Inject(TuiPortalService))
], TuiPortalDirective);
export { TuiPortalDirective };
