// @taiga-ui/addon-doc/esm5/internal/see-also/see-also.component.js
import { __decorate, __param, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { TUI_DOC_SEE_ALSO_TEXT } from '../../tokens/i18n';
import { TUI_DOC_PAGES } from '../../tokens/pages';
// @dynamic
var TuiDocSeeAlsoComponent = /** @class */ (function () {
    function TuiDocSeeAlsoComponent(text, pages) {
        this.text = text;
        this.pages = pages;
        this.seeAlso = [];
    }
    TuiDocSeeAlsoComponent.prototype.getRouterLink = function (pageTitle) {
        for (var i = 0; i < this.pages.length; i++) {
            var page = this.pages
                .map(function (page) { return ('subPages' in page ? page.subPages : [page]); })
                .reduce(function (pages, subPages) { return __spread(pages, subPages); }, [])
                .find(function (page) { return page.title === pageTitle; });
            if (page === null || page === void 0 ? void 0 : page.route) {
                return page.route;
            }
        }
        return '';
    };
    TuiDocSeeAlsoComponent.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Inject, args: [TUI_DOC_SEE_ALSO_TEXT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [TUI_DOC_PAGES,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiDocSeeAlsoComponent.prototype, "seeAlso", void 0);
    TuiDocSeeAlsoComponent = __decorate([
        Component({
            selector: 'tui-doc-see-also',
            template: "<h5 class=\"t-header\">{{ text }}</h5>\n<ng-container *ngFor=\"let item of seeAlso; last as last\">\n    <a\n        tuiLink\n        [routerLink]=\"getRouterLink(item)\"\n    >\n        {{ item }}\n    </a>\n    <ng-container *ngIf=\"!last\">{{ ', ' }}</ng-container>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{display:block}.t-header{font-size:.6875rem;line-height:1rem;text-transform:uppercase;letter-spacing:.075em;color:var(--tui-text-02);margin:0 0 .5rem;white-space:nowrap}"]
        }),
        __param(0, Inject(TUI_DOC_SEE_ALSO_TEXT)),
        __param(1, Inject(TUI_DOC_PAGES))
    ], TuiDocSeeAlsoComponent);
    return TuiDocSeeAlsoComponent;
}());
export { TuiDocSeeAlsoComponent };
