// @taiga-ui/kit/esm2015/directives/present/present.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, HostListener, Inject, OnDestroy, Output, } from '@angular/core';
import { USER_AGENT } from '@ng-web-apis/common';
import { isCurrentTarget, isFirefox } from '@taiga-ui/cdk';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged, skip } from 'rxjs/operators';
let TuiPresentDirective = class TuiPresentDirective {
    constructor({ nativeElement }, userAgent) {
        this.visibility$ = new BehaviorSubject(false);
        this.tuiPresentChange = this.visibility$.pipe(distinctUntilChanged(), skip(1));
        if (isFirefox(userAgent)) {
            return;
        }
        const observer = new MutationObserver(() => {
            if (!nativeElement.offsetParent &&
                nativeElement.offsetWidth === 0 &&
                nativeElement.offsetHeight === 0) {
                this.visibility$.next(false);
            }
        });
        observer.observe(nativeElement, {
            attributes: true,
            attributeFilter: ['style', 'class'],
        });
    }
    /**
     * Someday animationcancel would work and mutation observer would not be needed:
     * https://drafts.csswg.org/css-animations/#eventdef-animationevent-animationcancel
     * It would also trigger on CSS like display: none on parent nodes which is awesome
     * but currently only works in Firefox
     * ___
     * TODO: remove MutationObserver when we bump versions of supported browsers:
     *** Safari 12+
     *** Chrome 83+
     * See: {@link https://caniuse.com/mdn-api_window_animationcancel_event}
     */
    onAnimation(event, visibility) {
        if (isCurrentTarget(event)) {
            this.visibility$.next(visibility);
        }
    }
    ngOnDestroy() {
        this.visibility$.next(false);
    }
};
TuiPresentDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: String, decorators: [{ type: Inject, args: [USER_AGENT,] }] }
];
__decorate([
    Output()
], TuiPresentDirective.prototype, "tuiPresentChange", void 0);
__decorate([
    HostListener('animationcancel', ['$event', 'false']),
    HostListener('animationstart', ['$event', 'true'])
], TuiPresentDirective.prototype, "onAnimation", null);
TuiPresentDirective = __decorate([
    Directive({
        selector: '[tuiPresentChange]',
        host: {
            '[style.animation]': '"tuiPresent 1s infinite"',
        },
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(USER_AGENT))
], TuiPresentDirective);
export { TuiPresentDirective };
