// @taiga-ui/cdk/esm5/directives/focus-trap/focus-trap.directive.js
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, HostListener, Inject, OnDestroy, Renderer2, } from '@angular/core';
import { containsOrAfter } from '@taiga-ui/cdk/utils/dom';
import { blurNativeFocused, getClosestFocusable, getNativeFocused, setNativeFocused, } from '@taiga-ui/cdk/utils/focus';
// @dynamic
var TuiFocusTrapDirective = /** @class */ (function () {
    function TuiFocusTrapDirective(documentRef, elementRef, renderer) {
        var _this = this;
        this.documentRef = documentRef;
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.activeElement = getNativeFocused(this.documentRef);
        /**
         * This would cause currently focused element to lose focus,
         * but it might cause ExpressionChanged error due to potential HostBinding.
         * Microtask keeps it in the same frame but allows change detection to run
         */
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        Promise.resolve().then(function () {
            setNativeFocused(_this.elementRef.nativeElement);
        });
    }
    TuiFocusTrapDirective.prototype.onBlur = function () {
        this.renderer.removeAttribute(this.elementRef.nativeElement, 'tabIndex');
    };
    TuiFocusTrapDirective.prototype.onFocusIn = function (node) {
        if (containsOrAfter(this.elementRef.nativeElement, node)) {
            return;
        }
        var focusable = getClosestFocusable(this.elementRef.nativeElement, false, this.elementRef.nativeElement);
        if (focusable) {
            setNativeFocused(focusable);
        }
    };
    TuiFocusTrapDirective.prototype.ngOnDestroy = function () {
        var _this = this;
        blurNativeFocused(this.documentRef);
        /**
         * HostListeners are triggered even after ngOnDestroy
         * {@link https://github.com/angular/angular/issues/38100}
         * so we need to delay it but stay in the same sync cycle,
         * therefore using Promise instead of setTimeout
         */
        // eslint-disable-next-line
        Promise.resolve().then(function () {
            // TODO: iframe warning
            if (_this.activeElement instanceof HTMLElement) {
                setNativeFocused(_this.activeElement);
            }
        });
    };
    TuiFocusTrapDirective.ctorParameters = function () { return [
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Renderer2, decorators: [{ type: Inject, args: [Renderer2,] }] }
    ]; };
    __decorate([
        HostListener('blur')
    ], TuiFocusTrapDirective.prototype, "onBlur", null);
    __decorate([
        HostListener('window:focusin.silent', ['$event.target'])
    ], TuiFocusTrapDirective.prototype, "onFocusIn", null);
    TuiFocusTrapDirective = __decorate([
        Directive({
            selector: '[tuiFocusTrap]',
            host: {
                tabIndex: '0',
            },
        }),
        __param(0, Inject(DOCUMENT)),
        __param(1, Inject(ElementRef)),
        __param(2, Inject(Renderer2))
    ], TuiFocusTrapDirective);
    return TuiFocusTrapDirective;
}());
export { TuiFocusTrapDirective };
