// @taiga-ui/addon-preview/esm5/components/preview/pagination/preview-pagination.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, HostListener, Inject, Input, Output, } from '@angular/core';
import { clamp, tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_PAGINATION_TEXTS } from '@taiga-ui/kit';
import { Observable } from 'rxjs';
var TuiPreviewPaginationComponent = /** @class */ (function () {
    function TuiPreviewPaginationComponent(texts$) {
        this.texts$ = texts$;
        this.length = 1;
        this.index = 0;
        this.indexChange = new EventEmitter();
    }
    Object.defineProperty(TuiPreviewPaginationComponent.prototype, "leftButtonDisabled", {
        get: function () {
            return this.index === 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPreviewPaginationComponent.prototype, "rightButtonDisabled", {
        get: function () {
            return this.index === this.length - 1;
        },
        enumerable: true,
        configurable: true
    });
    TuiPreviewPaginationComponent.prototype.onArrowClick = function (step) {
        this.updateIndex(clamp(this.index + step, 0, this.length - 1));
    };
    TuiPreviewPaginationComponent.prototype.updateIndex = function (index) {
        if (this.index === index) {
            return;
        }
        this.index = index;
        this.indexChange.emit(index);
    };
    TuiPreviewPaginationComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_PAGINATION_TEXTS,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPreviewPaginationComponent.prototype, "length", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPreviewPaginationComponent.prototype, "index", void 0);
    __decorate([
        Output()
    ], TuiPreviewPaginationComponent.prototype, "indexChange", void 0);
    __decorate([
        HostListener('document:keydown.arrowRight.prevent', ['1']),
        HostListener('document:keydown.arrowLeft.prevent', ['-1'])
    ], TuiPreviewPaginationComponent.prototype, "onArrowClick", null);
    TuiPreviewPaginationComponent = __decorate([
        Component({
            selector: 'tui-preview-pagination',
            template: "<ng-container *ngIf=\"texts$ | async as texts\">\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconArrowLeft\"\n        class=\"t-arrow_left\"\n        [title]=\"texts[0]\"\n        [disabled]=\"leftButtonDisabled\"\n        (click)=\"onArrowClick(-1)\"\n    ></button>\n    {{ index + 1 }}/{{ length }}\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconArrowRight\"\n        class=\"t-arrow_right\"\n        [title]=\"texts[1]\"\n        [disabled]=\"rightButtonDisabled\"\n        (click)=\"onArrowClick(1)\"\n    ></button>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:6.25rem}.t-arrow_left{border-top-right-radius:0;border-bottom-right-radius:0}.t-arrow_right{border-top-left-radius:0;border-bottom-left-radius:0}"]
        }),
        __param(0, Inject(TUI_PAGINATION_TEXTS))
    ], TuiPreviewPaginationComponent);
    return TuiPreviewPaginationComponent;
}());
export { TuiPreviewPaginationComponent };
