// @taiga-ui/kit/esm5/pipes/filter-by-input/filter-by-input.base.js
import { __decorate, __extends } from "tslib";
/* eslint-disable @typescript-eslint/naming-convention */
import { isPresent, tuiPure, } from '@taiga-ui/cdk';
import { isFlat } from '@taiga-ui/kit/utils';
/**
 * @deprecated
 * TODO: 3.0 replace with {@link AbstractTuiFilterByInput}
 */
var TuiFilterByInputBase = /** @class */ (function () {
    function TuiFilterByInputBase() {
    }
    Object.defineProperty(TuiFilterByInputBase.prototype, "query", {
        get: function () {
            return this.accessor.nativeFocusableElement
                ? this.accessor.nativeFocusableElement.value || ""
                : "";
        },
        enumerable: true,
        configurable: true
    });
    TuiFilterByInputBase.prototype.filter = function (items, matcher, stringify, query) {
        if (!items) {
            return null;
        }
        return isFlat(items)
            ? this.filterFlat(items, matcher, stringify, query)
            : this.filter2d(items, matcher, stringify, query);
    };
    TuiFilterByInputBase.prototype.filterFlat = function (items, matcher, stringify, query) {
        var match = this.getMatch(items, stringify, query);
        return isPresent(match)
            ? items
            : items.filter(function (item) { return matcher(item, query, stringify); });
    };
    TuiFilterByInputBase.prototype.filter2d = function (items, matcher, stringify, query) {
        var _this = this;
        var match = items.find(function (item) {
            return isPresent(_this.getMatch(item, stringify, query));
        });
        return isPresent(match)
            ? items
            : items.map(function (inner) { return _this.filterFlat(inner, matcher, stringify, query); });
    };
    TuiFilterByInputBase.prototype.getMatch = function (items, stringify, query) {
        return items.find(function (item) { return stringify(item).toLocaleLowerCase() === query.toLocaleLowerCase(); });
    };
    __decorate([
        tuiPure
    ], TuiFilterByInputBase.prototype, "filter", null);
    return TuiFilterByInputBase;
}());
export { TuiFilterByInputBase };
var AbstractTuiFilterByInput = /** @class */ (function (_super) {
    __extends(AbstractTuiFilterByInput, _super);
    function AbstractTuiFilterByInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return AbstractTuiFilterByInput;
}(TuiFilterByInputBase));
export { AbstractTuiFilterByInput };
