// @taiga-ui/kit/esm5/components/input-copy/input-copy.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputCopyDirective = /** @class */ (function (_super) {
    __extends(TuiInputCopyDirective, _super);
    function TuiInputCopyDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputCopyDirective_1 = TuiInputCopyDirective;
    TuiInputCopyDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    var TuiInputCopyDirective_1;
    TuiInputCopyDirective = TuiInputCopyDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-copy',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputCopyDirective_1; }),
                },
            ],
        })
    ], TuiInputCopyDirective);
    return TuiInputCopyDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputCopyDirective };
