// @taiga-ui/core/esm5/internal/primitive-year-month-pagination/primitive-year-month-pagination.component.js
import { __decorate } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TUI_FIRST_DAY, TUI_LAST_DAY, tuiDefaultProp, TuiMonth, } from '@taiga-ui/cdk';
var TuiPrimitiveYearMonthPaginationComponent = /** @class */ (function () {
    function TuiPrimitiveYearMonthPaginationComponent() {
        this.value = TuiMonth.currentLocal();
        this.min = TUI_FIRST_DAY;
        this.max = TUI_LAST_DAY;
        this.valueChange = new EventEmitter();
        this.yearClick = new EventEmitter();
    }
    Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "prevMonthDisabled", {
        get: function () {
            return this.value.monthSameOrBefore(this.min);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "nextMonthDisabled", {
        get: function () {
            return this.value.monthSameOrAfter(this.max);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPrimitiveYearMonthPaginationComponent.prototype, "oneYear", {
        get: function () {
            return this.min.year === this.max.year;
        },
        enumerable: true,
        configurable: true
    });
    TuiPrimitiveYearMonthPaginationComponent.prototype.onYearClick = function () {
        this.yearClick.next(this.value);
    };
    TuiPrimitiveYearMonthPaginationComponent.prototype.onPrevMonthClick = function () {
        this.appendValueWithLimit({ month: -1 });
    };
    TuiPrimitiveYearMonthPaginationComponent.prototype.onNextMonthClick = function () {
        this.appendValueWithLimit({ month: 1 });
    };
    TuiPrimitiveYearMonthPaginationComponent.prototype.appendValueWithLimit = function (date) {
        var newMonth = this.value.append(date);
        if (this.min.monthSameOrAfter(newMonth)) {
            this.updateValue(this.min);
            return;
        }
        if (this.max.monthSameOrBefore(newMonth)) {
            this.updateValue(this.max);
            return;
        }
        this.updateValue(newMonth);
    };
    TuiPrimitiveYearMonthPaginationComponent.prototype.updateValue = function (value) {
        if (this.value.monthSame(value)) {
            return;
        }
        this.value = value;
        this.valueChange.emit(value);
    };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveYearMonthPaginationComponent.prototype, "value", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveYearMonthPaginationComponent.prototype, "min", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPrimitiveYearMonthPaginationComponent.prototype, "max", void 0);
    __decorate([
        Output()
    ], TuiPrimitiveYearMonthPaginationComponent.prototype, "valueChange", void 0);
    __decorate([
        Output()
    ], TuiPrimitiveYearMonthPaginationComponent.prototype, "yearClick", void 0);
    TuiPrimitiveYearMonthPaginationComponent = __decorate([
        Component({
            selector: 'tui-primitive-year-month-pagination',
            template: "<tui-primitive-spin-button\n    [focusable]=\"false\"\n    [leftDisabled]=\"prevMonthDisabled\"\n    [rightDisabled]=\"nextMonthDisabled\"\n    (leftClick)=\"onPrevMonthClick()\"\n    (rightClick)=\"onNextMonthClick()\"\n>\n    {{ value | tuiMonth | async }}\n    <ng-container *ngIf=\"oneYear; else button\">\n        {{ value.formattedYear }}\n    </ng-container>\n    <ng-template #button>\n        <button\n            automation-id=\"tui-primitive-year-month-pagination__year-button\"\n            tuiLink\n            type=\"button\"\n            [tuiFocusable]=\"false\"\n            (click)=\"onYearClick()\"\n        >\n            {{ value.formattedYear }}\n        </button>\n    </ng-template>\n</tui-primitive-spin-button>\n",
            styles: [":host{display:block}"]
        })
    ], TuiPrimitiveYearMonthPaginationComponent);
    return TuiPrimitiveYearMonthPaginationComponent;
}());
export { TuiPrimitiveYearMonthPaginationComponent };
