// @taiga-ui/kit/esm2015/components/multi-select-option/multi-select-option.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { isPresent } from '@taiga-ui/cdk';
import { sizeBigger } from '@taiga-ui/core';
import { TuiSelectOptionComponent } from '@taiga-ui/kit/components/select-option';
let TuiMultiSelectOptionComponent = class TuiMultiSelectOptionComponent extends TuiSelectOptionComponent {
    get size() {
        return sizeBigger(this.option.size) ? 'l' : 'm';
    }
    get selected() {
        return (isPresent(this.option.value) &&
            isPresent(this.control.value) &&
            !!this.control.value.find((item) => this.matcher(item, this.option.value)));
    }
};
TuiMultiSelectOptionComponent = __decorate([
    Component({
        selector: 'tui-multi-select-option',
        template: "<tui-primitive-checkbox\n    class=\"t-checkbox\"\n    [size]=\"size\"\n    [value]=\"selected$ | async\"\n></tui-primitive-checkbox>\n<ng-container [ngTemplateOutlet]=\"context.$implicit\"></ng-container>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: ["@keyframes retrigger{from{left:1px}to{left:2px}}:host{display:flex;flex:1;align-items:center;overflow:hidden;animation:1s retrigger}.t-checkbox{margin-right:.75rem}"]
    })
], TuiMultiSelectOptionComponent);
export { TuiMultiSelectOptionComponent };
