// @taiga-ui/addon-doc/esm5/directives/scroll-into-view/scroll-into-view.directive.js
import { __decorate, __extends, __param } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { getElementObscurers, TuiDestroyService } from '@taiga-ui/cdk';
import { Observable, ReplaySubject } from 'rxjs';
import { debounceTime, filter, switchMapTo, takeUntil } from 'rxjs/operators';
import { TUI_DOC_PAGE_LOADED } from '../../tokens/page-loaded';
/**
 * @deprecated: use {@link TuiScrollIntoViewLinkDirective}
 * TODO: remove in v3.0
 */
var ScrollIntoViewDirective = /** @class */ (function () {
    function ScrollIntoViewDirective(destroy$, _a, readyToScroll$) {
        var nativeElement = _a.nativeElement;
        this.readyToScroll$ = readyToScroll$;
        this.scroll$ = new ReplaySubject(1);
        this.readyToScroll$
            .pipe(filter(Boolean), switchMapTo(this.scroll$), debounceTime(750), filter(function (shallWe) { return shallWe && !!getElementObscurers(nativeElement); }), takeUntil(destroy$))
            .subscribe(function () {
            nativeElement.scrollIntoView();
        });
    }
    Object.defineProperty(ScrollIntoViewDirective.prototype, "scrollIntoView", {
        set: function (shallWe) {
            this.scroll$.next(shallWe);
        },
        enumerable: true,
        configurable: true
    });
    ScrollIntoViewDirective.ctorParameters = function () { return [
        { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_DOC_PAGE_LOADED,] }] }
    ]; };
    __decorate([
        Input()
    ], ScrollIntoViewDirective.prototype, "scrollIntoView", null);
    ScrollIntoViewDirective = __decorate([
        Directive({
            selector: '[scrollIntoView]',
            providers: [TuiDestroyService],
        })
        // eslint-disable-next-line @typescript-eslint/naming-convention
        ,
        __param(0, Inject(TuiDestroyService)),
        __param(1, Inject(ElementRef)),
        __param(2, Inject(TUI_DOC_PAGE_LOADED))
    ], ScrollIntoViewDirective);
    return ScrollIntoViewDirective;
}());
export { ScrollIntoViewDirective };
var TuiScrollIntoViewLinkDirective = /** @class */ (function (_super) {
    __extends(TuiScrollIntoViewLinkDirective, _super);
    function TuiScrollIntoViewLinkDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiScrollIntoViewLinkDirective = __decorate([
        Directive({
            selector: '[tuiScrollIntoViewLink]',
            providers: [TuiDestroyService],
        })
    ], TuiScrollIntoViewLinkDirective);
    return TuiScrollIntoViewLinkDirective;
}(ScrollIntoViewDirective));
export { TuiScrollIntoViewLinkDirective };
