// @taiga-ui/core/fesm5/taiga-ui-core-directives-table-mode.js
import { __decorate } from 'tslib';
import { Directive, NgModule } from '@angular/core';
import { TUI_TEXTFIELD_APPEARANCE } from '@taiga-ui/core/tokens';

/**
 * @deprecated use @taiga-ui/addon-table
 */
var TuiTableModeDirective = /** @class */ (function () {
    function TuiTableModeDirective() {
    }
    TuiTableModeDirective = __decorate([
        Directive({
            selector: '[tuiTableMode]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_APPEARANCE,
                    useValue: 'table',
                },
            ],
        })
    ], TuiTableModeDirective);
    return TuiTableModeDirective;
}());

/** @deprecated */
var TuiTableModeModule = /** @class */ (function () {
    function TuiTableModeModule() {
    }
    TuiTableModeModule = __decorate([
        NgModule({
            declarations: [TuiTableModeDirective],
            exports: [TuiTableModeDirective],
        })
    ], TuiTableModeModule);
    return TuiTableModeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiTableModeDirective, TuiTableModeModule };
