// @taiga-ui/core/fesm5/taiga-ui-core-pipes-format-date.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TuiFormatDateService } from '@taiga-ui/core/services';

var TuiFormatDatePipe = /** @class */ (function () {
    function TuiFormatDatePipe(service) {
        this.service = service;
    }
    TuiFormatDatePipe.prototype.transform = function (timestampOrDate) {
        return this.service.format(timestampOrDate.valueOf());
    };
    TuiFormatDatePipe.ctorParameters = function () { return [
        { type: TuiFormatDateService, decorators: [{ type: Inject, args: [TuiFormatDateService,] }] }
    ]; };
    TuiFormatDatePipe = __decorate([
        Pipe({
            name: "tuiFormatDate",
        }),
        __param(0, Inject(TuiFormatDateService))
    ], TuiFormatDatePipe);
    return TuiFormatDatePipe;
}());

var TuiFormatDatePipeModule = /** @class */ (function () {
    function TuiFormatDatePipeModule() {
    }
    TuiFormatDatePipeModule = __decorate([
        NgModule({
            declarations: [TuiFormatDatePipe],
            exports: [TuiFormatDatePipe],
        })
    ], TuiFormatDatePipeModule);
    return TuiFormatDatePipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFormatDatePipe, TuiFormatDatePipeModule };
