// @taiga-ui/core/fesm5/taiga-ui-core-components-group.js
import { __decorate, __param, __extends } from 'tslib';
import { Component, ViewEncapsulation, Inject, Input, HostBinding, Directive, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { TuiDirectiveStylesService, tuiDefaultProp } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

var TuiGroupStylesComponent = /** @class */ (function () {
    function TuiGroupStylesComponent() {
    }
    TuiGroupStylesComponent = __decorate([
        Component({
            template: '',
            encapsulation: ViewEncapsulation.None,
            styles: [".tui-group{position:relative;z-index:0;display:flex}.tui-group>*{flex:1 1 0;min-width:0}.tui-group>:nth-child(n){border-radius:0}.tui-group>:not(:last-child){margin-right:.125rem}.tui-group_collapsed>*{z-index:1}.tui-group_collapsed>:not(:last-child){margin-right:-1px}.tui-group_collapsed>._disabled,.tui-group_collapsed>._disabled:hover,.tui-group_collapsed>._readonly,.tui-group_collapsed>._readonly:hover{z-index:0}.tui-group_collapsed>._hovered,.tui-group_collapsed>._pressed,.tui-group_collapsed>:hover{z-index:3}.tui-group_collapsed>._invalid{z-index:2}.tui-group_collapsed>._invalid._hovered,.tui-group_collapsed>._invalid._pressed,.tui-group_collapsed>._invalid:hover{z-index:4}.tui-group_collapsed>._focus-visible,.tui-group_collapsed>._focused.ng-touched,.tui-group_collapsed>._focused.ng-untouched,.tui-group_collapsed>._hosted_dropdown_focused{z-index:5}.tui-group_collapsed>._active,.tui-group_collapsed>[data-appearance=whiteblock-active]{z-index:6}.tui-group_collapsed>._focus-visible._active,.tui-group_collapsed>._focus-visible._focused,.tui-group_collapsed>._focus-visible[data-appearance=whiteblock-active]{z-index:7}.tui-group_orientation_vertical{display:inline-flex;flex-direction:column}.tui-group_orientation_vertical>*{min-height:auto;flex:0 0 auto}.tui-group_orientation_vertical>:not(:last-child){margin-right:0;margin-bottom:.125rem}.tui-group_orientation_vertical.tui-group_collapsed>:not(:last-child){margin-right:0;margin-bottom:-1px}.tui-group_rounded.tui-group_orientation_horizontal>:first-child{border-top-left-radius:var(--tui-radius-m);border-bottom-left-radius:var(--tui-radius-m)}.tui-group_rounded.tui-group_orientation_horizontal>:last-child{border-top-right-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.tui-group_rounded.tui-group_orientation_horizontal.tui-group_radius_large>:first-child{border-top-left-radius:var(--tui-radius-l);border-bottom-left-radius:var(--tui-radius-l)}.tui-group_rounded.tui-group_orientation_horizontal.tui-group_radius_large>:last-child{border-top-right-radius:var(--tui-radius-l);border-bottom-right-radius:var(--tui-radius-l)}.tui-group_rounded.tui-group_orientation_vertical>:first-child{border-top-left-radius:var(--tui-radius-m);border-top-right-radius:var(--tui-radius-m)}.tui-group_rounded.tui-group_orientation_vertical>:last-child{border-bottom-left-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.tui-group_rounded.tui-group_orientation_vertical.tui-group_radius_large>:first-child{border-top-left-radius:var(--tui-radius-l);border-top-right-radius:var(--tui-radius-l)}.tui-group_rounded.tui-group_orientation_vertical.tui-group_radius_large>:last-child{border-bottom-left-radius:var(--tui-radius-l);border-bottom-right-radius:var(--tui-radius-l)}@media screen and (max-width:47.9625em){.tui-group_rounded.tui-group_adaptive.tui-group_orientation_horizontal>:first-child:not(:last-child),.tui-group_rounded.tui-group_adaptive.tui-group_orientation_horizontal>:last-child:not(:first-child),.tui-group_rounded.tui-group_adaptive.tui-group_orientation_vertical:not(.tui-group_collapsed)>:first-child:not(:last-child),.tui-group_rounded.tui-group_adaptive.tui-group_orientation_vertical:not(.tui-group_collapsed)>:last-child:not(:first-child){border-radius:var(--tui-radius-m)}.tui-group_rounded.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal>:first-child:not(:last-child),.tui-group_rounded.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal>:last-child:not(:first-child),.tui-group_rounded.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical:not(.tui-group_collapsed)>:first-child:not(:last-child),.tui-group_rounded.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical:not(.tui-group_collapsed)>:last-child:not(:first-child){border-radius:var(--tui-radius-l)}.tui-group_adaptive.tui-group_orientation_vertical{width:100%}.tui-group_adaptive.tui-group_orientation_vertical>:nth-child(n):not(:only-child){border-radius:var(--tui-radius-m)}.tui-group_adaptive.tui-group_orientation_vertical>:not(:last-child){margin-bottom:1rem}.tui-group_adaptive.tui-group_orientation_vertical.tui-group_collapsed>:nth-child(n):not(:only-child){border-radius:0}.tui-group_adaptive.tui-group_orientation_vertical.tui-group_collapsed>:first-child:not(:last-child){border-radius:var(--tui-radius-m) var(--tui-radius-m) 0 0}.tui-group_adaptive.tui-group_orientation_vertical.tui-group_collapsed>:last-child:not(:first-child){border-radius:0 0 var(--tui-radius-m) var(--tui-radius-m)}.tui-group_adaptive.tui-group_orientation_vertical.tui-group_collapsed>:not(:last-child){margin-bottom:-1px}.tui-group_adaptive.tui-group_orientation_horizontal{flex-direction:column;width:100%}.tui-group_adaptive.tui-group_orientation_horizontal>:nth-child(n):not(:only-child){border-radius:var(--tui-radius-m)}.tui-group_adaptive.tui-group_orientation_horizontal>:not(:last-child){margin-right:0;margin-bottom:1rem}.tui-group_adaptive.tui-group_orientation_horizontal.tui-group_collapsed>:nth-child(n):not(:only-child){border-radius:0}.tui-group_adaptive.tui-group_orientation_horizontal.tui-group_collapsed>:first-child:not(:last-child){border-radius:var(--tui-radius-m) var(--tui-radius-m) 0 0}.tui-group_adaptive.tui-group_orientation_horizontal.tui-group_collapsed>:last-child:not(:first-child){border-radius:0 0 var(--tui-radius-m) var(--tui-radius-m)}.tui-group_adaptive.tui-group_orientation_horizontal.tui-group_collapsed>:not(:last-child){margin-right:0;margin-bottom:-1px}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical>:nth-child(n):not(:only-child){border-radius:var(--tui-radius-l)}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical.tui-group_collapsed>:nth-child(n):not(:only-child){border-radius:0}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical.tui-group_collapsed>:first-child:not(:last-child){border-radius:var(--tui-radius-l) var(--tui-radius-l) 0 0}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_vertical.tui-group_collapsed>:last-child:not(:first-child){border-radius:0 0 var(--tui-radius-l) var(--tui-radius-l)}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal>:nth-child(n):not(:only-child){border-radius:var(--tui-radius-l)}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal.tui-group_collapsed>:nth-child(n):not(:only-child){border-radius:0}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal.tui-group_collapsed>:first-child:not(:last-child){border-radius:var(--tui-radius-l) var(--tui-radius-l) 0 0}.tui-group_adaptive.tui-group_radius_large.tui-group_orientation_horizontal.tui-group_collapsed>:last-child:not(:first-child){border-radius:0 0 var(--tui-radius-l) var(--tui-radius-l)}}.tui-group__auto-width-item{min-width:auto;flex:0 0 auto}.tui-group .tui-group__inherit-item{border-radius:inherit}"]
        })
    ], TuiGroupStylesComponent);
    return TuiGroupStylesComponent;
}());

var TuiGroupDirective = /** @class */ (function () {
    function TuiGroupDirective(directiveStyles) {
        this.orientation = 'horizontal';
        this.adaptive = false;
        this.collapsed = false;
        this.rounded = true;
        this.size = 'm';
        directiveStyles.addComponent(TuiGroupStylesComponent);
    }
    Object.defineProperty(TuiGroupDirective.prototype, "orientationHorizontal", {
        get: function () {
            return this.orientation === 'horizontal';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "orientationVertical", {
        get: function () {
            return this.orientation === 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "sizeLarge", {
        get: function () {
            return this.size === 'l';
        },
        enumerable: true,
        configurable: true
    });
    TuiGroupDirective.ctorParameters = function () { return [
        { type: TuiDirectiveStylesService, decorators: [{ type: Inject, args: [TuiDirectiveStylesService,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "orientation", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_adaptive'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "adaptive", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_collapsed'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "collapsed", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_rounded'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "rounded", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "size", void 0);
    __decorate([
        HostBinding('class.tui-group_orientation_horizontal')
    ], TuiGroupDirective.prototype, "orientationHorizontal", null);
    __decorate([
        HostBinding('class.tui-group_orientation_vertical')
    ], TuiGroupDirective.prototype, "orientationVertical", null);
    __decorate([
        HostBinding('class.tui-group_radius_large')
    ], TuiGroupDirective.prototype, "sizeLarge", null);
    TuiGroupDirective = __decorate([
        Directive({
            selector: '[tuiGroup]:not(ng-container)',
            host: {
                class: 'tui-group',
                role: 'group',
            },
        }),
        __param(0, Inject(TuiDirectiveStylesService))
    ], TuiGroupDirective);
    return TuiGroupDirective;
}());

/** @deprecated */
var TuiGroupComponent = /** @class */ (function (_super) {
    __extends(TuiGroupComponent, _super);
    function TuiGroupComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiGroupComponent = __decorate([
        Component({
            selector: 'tui-group',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "\n        <ng-content></ng-content>\n    ",
            host: {
                class: 'tui-group',
                role: 'group',
            }
        })
    ], TuiGroupComponent);
    return TuiGroupComponent;
}(TuiGroupDirective));

var TuiGroupModule = /** @class */ (function () {
    function TuiGroupModule() {
    }
    TuiGroupModule = __decorate([
        NgModule({
            imports: [CommonModule],
            entryComponents: [TuiGroupStylesComponent],
            declarations: [TuiGroupComponent, TuiGroupDirective, TuiGroupStylesComponent],
            exports: [TuiGroupComponent, TuiGroupDirective, TuiGroupStylesComponent],
        })
    ], TuiGroupModule);
    return TuiGroupModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiGroupComponent, TuiGroupDirective, TuiGroupModule, TuiGroupStylesComponent };
