// @taiga-ui/core/fesm5/taiga-ui-core-pipes-flag.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TUI_ICONS_PATH } from '@taiga-ui/core/tokens';
import { TuiCountryIsoCode } from '@taiga-ui/i18n';

var TuiFlagPipe = /** @class */ (function () {
    function TuiFlagPipe(iconsPath) {
        this.iconsPath = iconsPath;
        this.staticPath = this.iconsPath("tuiIcon").replace("tuiIcon.svg#tuiIcon", "");
    }
    TuiFlagPipe.prototype.transform = function (countryIsoCode) {
        if (!countryIsoCode) {
            return null;
        }
        switch (countryIsoCode) {
            case TuiCountryIsoCode.BL:
            case TuiCountryIsoCode.BQ:
            case TuiCountryIsoCode.CW:
            case TuiCountryIsoCode.GF:
            case TuiCountryIsoCode.GP:
            case TuiCountryIsoCode.MF:
            case TuiCountryIsoCode.MQ:
            case TuiCountryIsoCode.NC:
            case TuiCountryIsoCode.RE:
            case TuiCountryIsoCode.YT:
                return "" + this.staticPath + TuiCountryIsoCode.FR + ".png";
            case TuiCountryIsoCode.SX:
                return "" + this.staticPath + TuiCountryIsoCode.NL + ".png";
            default:
                return "" + this.staticPath + countryIsoCode + ".png";
        }
    };
    TuiFlagPipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_ICONS_PATH,] }] }
    ]; };
    TuiFlagPipe = __decorate([
        Pipe({ name: "tuiFlag" }),
        __param(0, Inject(TUI_ICONS_PATH))
    ], TuiFlagPipe);
    return TuiFlagPipe;
}());

var TuiFlagPipeModule = /** @class */ (function () {
    function TuiFlagPipeModule() {
    }
    TuiFlagPipeModule = __decorate([
        NgModule({
            declarations: [TuiFlagPipe],
            exports: [TuiFlagPipe],
        })
    ], TuiFlagPipeModule);
    return TuiFlagPipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFlagPipe, TuiFlagPipeModule };
