// @taiga-ui/core/fesm5/taiga-ui-core-pipes-format-number.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TUI_NUMBER_FORMAT } from '@taiga-ui/core/tokens';
import { formatNumber } from '@taiga-ui/core/utils/format';

var TuiFormatNumberPipe = /** @class */ (function () {
    function TuiFormatNumberPipe(numberFormat) {
        this.numberFormat = numberFormat;
    }
    /**
     * Formats number adding thousand separators and correct decimal separator
     * padding decimal part with zeroes to given length
     * @param value number
     * @param decimalSeparator
     * @param thousandSeparator
     * @param decimalLimit number of digits of decimal part, null to keep untouched
     * @param zeroPadding enable zeros at the end of decimal part
     */
    TuiFormatNumberPipe.prototype.transform = function (value, decimalLimit, decimalSeparator, thousandSeparator, zeroPadding) {
        if (decimalLimit === void 0) { decimalLimit = null; }
        if (decimalSeparator === void 0) { decimalSeparator = this.numberFormat.decimalSeparator; }
        if (thousandSeparator === void 0) { thousandSeparator = this.numberFormat.thousandSeparator; }
        if (zeroPadding === void 0) { zeroPadding = this.numberFormat.zeroPadding; }
        return formatNumber(value, decimalLimit, decimalSeparator, thousandSeparator, zeroPadding);
    };
    TuiFormatNumberPipe.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_NUMBER_FORMAT,] }] }
    ]; };
    TuiFormatNumberPipe = __decorate([
        Pipe({ name: "tuiFormatNumber" }),
        __param(0, Inject(TUI_NUMBER_FORMAT))
    ], TuiFormatNumberPipe);
    return TuiFormatNumberPipe;
}());

var TuiFormatNumberPipeModule = /** @class */ (function () {
    function TuiFormatNumberPipeModule() {
    }
    TuiFormatNumberPipeModule = __decorate([
        NgModule({
            exports: [TuiFormatNumberPipe],
            declarations: [TuiFormatNumberPipe],
        })
    ], TuiFormatNumberPipeModule);
    return TuiFormatNumberPipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFormatNumberPipe, TuiFormatNumberPipeModule };
