// @taiga-ui/core/fesm5/taiga-ui-core-directives-hint-controller.js
import { __extends, __decorate, __param } from 'tslib';
import { InjectionToken, inject, Inject, Input, Directive, forwardRef, NgModule, ChangeDetectorRef } from '@angular/core';
import { tuiDefaultProp, TuiController, TuiDestroyService } from '@taiga-ui/cdk';
import { TUI_HINT_OPTIONS } from '@taiga-ui/core/directives/hint';
import { watchedControllerFactory } from '@taiga-ui/core/providers';

var TUI_HINT_CONTROLLER = new InjectionToken("Controls configuration of hints", {
    factory: function () { return new TuiHintControllerDirective(inject(TUI_HINT_OPTIONS)); },
});

// TODO: v2.0 use in Charts
var TuiHintControllerDirective = /** @class */ (function (_super) {
    __extends(TuiHintControllerDirective, _super);
    function TuiHintControllerDirective(options) {
        var _this = _super.call(this) || this;
        _this.options = options;
        // TODO: 3.0 Remove null
        _this.content = null;
        _this.direction = _this.options.direction;
        _this.mode = _this.options.mode;
        _this.showDelay = _this.options.tuiHintShowDelay;
        _this.hideDelay = _this.options.tuiHintHideDelay;
        return _this;
    }
    TuiHintControllerDirective_1 = TuiHintControllerDirective;
    var TuiHintControllerDirective_1;
    TuiHintControllerDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_HINT_OPTIONS,] }] }
    ]; };
    __decorate([
        Input('tuiHintContent'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "content", void 0);
    __decorate([
        Input('tuiHintDirection'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "direction", void 0);
    __decorate([
        Input('tuiHintMode'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "mode", void 0);
    __decorate([
        Input('tuiHintShowDelay'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "showDelay", void 0);
    __decorate([
        Input('tuiHintHideDelay'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "hideDelay", void 0);
    TuiHintControllerDirective = TuiHintControllerDirective_1 = __decorate([
        Directive({
            selector: '[tuiHintContent]',
            providers: [
                {
                    provide: TUI_HINT_CONTROLLER,
                    useExisting: forwardRef(function () { return TuiHintControllerDirective_1; }),
                },
            ],
        }),
        __param(0, Inject(TUI_HINT_OPTIONS))
    ], TuiHintControllerDirective);
    return TuiHintControllerDirective;
}(TuiController));

var TuiHintControllerModule = /** @class */ (function () {
    function TuiHintControllerModule() {
    }
    TuiHintControllerModule = __decorate([
        NgModule({
            declarations: [TuiHintControllerDirective],
            exports: [TuiHintControllerDirective],
        })
    ], TuiHintControllerModule);
    return TuiHintControllerModule;
}());

var TUI_HINT_WATCHED_CONTROLLER = new InjectionToken("watched hint controller");
// TODO: 3.0 remove in ivy compilation
var HINT_CONTROLLER_FACTORY = watchedControllerFactory;
var HINT_CONTROLLER_PROVIDER = [
    TuiDestroyService,
    {
        provide: TUI_HINT_WATCHED_CONTROLLER,
        deps: [TUI_HINT_CONTROLLER, ChangeDetectorRef, TuiDestroyService],
        useFactory: HINT_CONTROLLER_FACTORY,
    },
];

/**
 * Generated bundle index. Do not edit.
 */

export { HINT_CONTROLLER_FACTORY, HINT_CONTROLLER_PROVIDER, TUI_HINT_CONTROLLER, TUI_HINT_WATCHED_CONTROLLER, TuiHintControllerDirective, TuiHintControllerModule };
