// @taiga-ui/core/esm2015/components/scrollbar/scrollbar.component.js
var TuiScrollbarComponent_1;
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, HostListener, Inject, Input, } from '@angular/core';
import { CSS as CSS_TOKEN, USER_AGENT } from '@ng-web-apis/common';
import { getElementOffset, isFirefox, TUI_IS_IOS, tuiDefaultProp, } from '@taiga-ui/cdk';
import { TUI_SCROLL_INTO_VIEW, TUI_SCROLLABLE } from '@taiga-ui/core/constants';
import { TUI_SCROLL_REF } from '@taiga-ui/core/tokens';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function scrollRefFactory({ browserScrollRef, }) {
    return browserScrollRef;
}
// @dynamic
let TuiScrollbarComponent = TuiScrollbarComponent_1 = class TuiScrollbarComponent {
    constructor(cssRef, elementRef, userAgent, isIos) {
        this.cssRef = cssRef;
        this.elementRef = elementRef;
        this.userAgent = userAgent;
        this.isIos = isIos;
        this.delegated = false;
        this.isLegacy = !this.cssRef.supports('position', 'sticky') ||
            (isFirefox(this.userAgent) && !this.cssRef.supports('scrollbar-width', 'none'));
        this.hidden = false;
        this.browserScrollRef = new ElementRef(this.elementRef.nativeElement);
    }
    get showScrollbars() {
        return !this.hidden && !this.isIos && (!this.isLegacy || this.delegated);
    }
    get showNative() {
        return this.isLegacy && !this.hidden && !this.delegated;
    }
    onScrollable(element) {
        this.delegated = true;
        this.browserScrollRef.nativeElement = element;
    }
    scrollIntoView(detail) {
        if (this.delegated) {
            return;
        }
        const { nativeElement } = this.browserScrollRef;
        const { offsetTop, offsetLeft } = getElementOffset(nativeElement, detail);
        nativeElement.scrollTop =
            offsetTop + detail.offsetHeight / 2 - nativeElement.clientHeight / 2;
        nativeElement.scrollLeft =
            offsetLeft + detail.offsetWidth / 2 - nativeElement.clientWidth / 2;
    }
};
TuiScrollbarComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CSS_TOKEN,] }] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: String, decorators: [{ type: Inject, args: [USER_AGENT,] }] },
    { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_IOS,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiScrollbarComponent.prototype, "hidden", void 0);
__decorate([
    HostBinding('class._legacy')
], TuiScrollbarComponent.prototype, "showNative", null);
__decorate([
    HostListener(`${TUI_SCROLLABLE}.stop`, ['$event.detail'])
], TuiScrollbarComponent.prototype, "onScrollable", null);
__decorate([
    HostListener(`${TUI_SCROLL_INTO_VIEW}.stop`, ['$event.detail'])
], TuiScrollbarComponent.prototype, "scrollIntoView", null);
TuiScrollbarComponent = TuiScrollbarComponent_1 = __decorate([
    Component({
        selector: 'tui-scrollbar',
        template: "<tui-scroll-controls *ngIf=\"showScrollbars\"></tui-scroll-controls>\n<div class=\"t-content\">\n    <ng-content></ng-content>\n</div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [
            {
                provide: TUI_SCROLL_REF,
                deps: [TuiScrollbarComponent_1],
                useFactory: scrollRefFactory,
            },
        ],
        styles: [":host{z-index:0;position:relative;display:flex;overflow:auto}:host:not(._legacy){scrollbar-width:none;-ms-overflow-style:none}:host:not(._legacy)::-webkit-scrollbar,:host:not(._legacy)::-webkit-scrollbar-thumb{background:0 0;width:0;height:0}:host._legacy{overflow:overlay}@media all and (-webkit-min-device-pixel-ratio:0) and (min-resolution:0.001dpcm){:host._legacy::-webkit-scrollbar,:host._legacy::-webkit-scrollbar-thumb{width:1rem;height:1rem;border-radius:6.25rem;background-clip:padding-box;border:2.667rem solid transparent}:host._legacy::-webkit-scrollbar{background-color:transparent}:host._legacy::-webkit-scrollbar-thumb{background-color:var(--tui-clear-hover)}:host._legacy::-webkit-scrollbar-thumb:hover{background-color:var(--tui-clear-active)}:host._legacy::-webkit-scrollbar-thumb:active{background-color:var(--tui-text-03)}}.t-content{position:relative;z-index:0;flex:1;flex-basis:auto;width:100%;height:-webkit-max-content;height:-moz-max-content;height:max-content}"]
    }),
    __param(0, Inject(CSS_TOKEN)),
    __param(1, Inject(ElementRef)),
    __param(2, Inject(USER_AGENT)),
    __param(3, Inject(TUI_IS_IOS))
], TuiScrollbarComponent);
export { TuiScrollbarComponent };
