// @taiga-ui/addon-preview/esm2015/components/preview-dialog/preview-dialog.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractTuiDialogService, TUI_DIALOGS } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiPreviewDialogComponent } from './preview-dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
/**
 * @deprecated: use {@link TuiPreviewDialogService}
 * TODO: remove in v3.0
 */
let PreviewDialogService = 
// eslint-disable-next-line @typescript-eslint/naming-convention
class PreviewDialogService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.defaultOptions = {};
        this.component = new PolymorpheusComponent(TuiPreviewDialogComponent);
    }
};
PreviewDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PreviewDialogService_Factory() { return new PreviewDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: PreviewDialogService, providedIn: "root" });
PreviewDialogService = __decorate([
    Injectable({ providedIn: `root` })
    // eslint-disable-next-line @typescript-eslint/naming-convention
], PreviewDialogService);
export { PreviewDialogService };
let TuiPreviewDialogService = class TuiPreviewDialogService extends PreviewDialogService {
};
TuiPreviewDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiPreviewDialogService_Factory() { return new TuiPreviewDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: TuiPreviewDialogService, providedIn: "root" });
TuiPreviewDialogService = __decorate([
    Injectable({ providedIn: `root` })
], TuiPreviewDialogService);
export { TuiPreviewDialogService };
export const PREVIEW_DIALOG_PROVIDER = [
    {
        provide: TUI_DIALOGS,
        useExisting: TuiPreviewDialogService,
        multi: true,
    },
    {
        provide: TUI_DIALOGS,
        useExisting: PreviewDialogService,
        multi: true,
    },
];
