// @taiga-ui/core/esm5/directives/described-by/described-by.providers.js
import { ElementRef, InjectionToken } from '@angular/core';
import { stopPropagation, TuiDestroyService, TuiFocusVisibleService, typedFromEvent, } from '@taiga-ui/cdk';
import { merge, timer } from 'rxjs';
import { distinctUntilChanged, filter, mapTo, startWith, switchMapTo, take, takeUntil, } from 'rxjs/operators';
var DELAY = 1000;
export var TUI_DESCRIBED_BY_SHOW = new InjectionToken("Accessible tooltip visibility stream");
export var TUI_DESCRIBED_BY_PROVIDERS = [
    TuiDestroyService,
    TuiFocusVisibleService,
    {
        provide: TUI_DESCRIBED_BY_SHOW,
        deps: [TuiFocusVisibleService, ElementRef],
        useFactory: describedByFactory,
    },
];
// eslint-disable-next-line @typescript-eslint/naming-convention
export function describedByFactory(focusVisible$, _a) {
    var nativeElement = _a.nativeElement;
    return focusVisible$
        .pipe(filter(Boolean), switchMapTo(timer(DELAY).pipe(mapTo(true), takeUntil(merge(typedFromEvent(nativeElement, "keydown"), typedFromEvent(nativeElement, "blur"))))), switchMapTo(merge(typedFromEvent(nativeElement, "keydown").pipe(filter(function (_a) {
        var key = _a.key;
        return key === "Escape";
    }), take(1), stopPropagation(), mapTo(false), startWith(true)), typedFromEvent(nativeElement, "blur").pipe(mapTo(false)))))
        .pipe(distinctUntilChanged());
}
