// @taiga-ui/kit/esm5/components/input-date-time/input-date-time.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputDateTimeDirective = /** @class */ (function (_super) {
    __extends(TuiInputDateTimeDirective, _super);
    function TuiInputDateTimeDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputDateTimeDirective_1 = TuiInputDateTimeDirective;
    Object.defineProperty(TuiInputDateTimeDirective.prototype, "value", {
        get: function () {
            return this.host.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiInputDateTimeDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    TuiInputDateTimeDirective.prototype.process = function (input) {
        input.inputMode = 'numeric';
    };
    var TuiInputDateTimeDirective_1;
    TuiInputDateTimeDirective = TuiInputDateTimeDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-date-time',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputDateTimeDirective_1; }),
                },
            ],
        })
    ], TuiInputDateTimeDirective);
    return TuiInputDateTimeDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputDateTimeDirective };
