// @taiga-ui/addon-doc/esm2015/directives/scroll-into-view/scroll-into-view.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, Input } from '@angular/core';
import { getElementObscurers, TuiDestroyService } from '@taiga-ui/cdk';
import { Observable, ReplaySubject } from 'rxjs';
import { debounceTime, filter, switchMapTo, takeUntil } from 'rxjs/operators';
import { TUI_DOC_PAGE_LOADED } from '../../tokens/page-loaded';
/**
 * @deprecated: use {@link TuiScrollIntoViewLinkDirective}
 * TODO: remove in v3.0
 */
let ScrollIntoViewDirective = 
// eslint-disable-next-line @typescript-eslint/naming-convention
class ScrollIntoViewDirective {
    constructor(destroy$, { nativeElement }, readyToScroll$) {
        this.readyToScroll$ = readyToScroll$;
        this.scroll$ = new ReplaySubject(1);
        this.readyToScroll$
            .pipe(filter(Boolean), switchMapTo(this.scroll$), debounceTime(750), filter(shallWe => shallWe && !!getElementObscurers(nativeElement)), takeUntil(destroy$))
            .subscribe(() => {
            nativeElement.scrollIntoView();
        });
    }
    set scrollIntoView(shallWe) {
        this.scroll$.next(shallWe);
    }
};
ScrollIntoViewDirective.ctorParameters = () => [
    { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_DOC_PAGE_LOADED,] }] }
];
__decorate([
    Input()
], ScrollIntoViewDirective.prototype, "scrollIntoView", null);
ScrollIntoViewDirective = __decorate([
    Directive({
        selector: '[scrollIntoView]',
        providers: [TuiDestroyService],
    })
    // eslint-disable-next-line @typescript-eslint/naming-convention
    ,
    __param(0, Inject(TuiDestroyService)),
    __param(1, Inject(ElementRef)),
    __param(2, Inject(TUI_DOC_PAGE_LOADED))
], ScrollIntoViewDirective);
export { ScrollIntoViewDirective };
let TuiScrollIntoViewLinkDirective = class TuiScrollIntoViewLinkDirective extends ScrollIntoViewDirective {
};
TuiScrollIntoViewLinkDirective = __decorate([
    Directive({
        selector: '[tuiScrollIntoViewLink]',
        providers: [TuiDestroyService],
    })
], TuiScrollIntoViewLinkDirective);
export { TuiScrollIntoViewLinkDirective };
