// @taiga-ui/kit/esm5/components/slider/helpers/slider-readonly.directive.js
import { __decorate, __param, __read } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, HostListener, Inject, Input } from '@angular/core';
import { tuiCoerceBooleanProperty, tuiDefaultProp, TuiDestroyService, typedFromEvent, } from '@taiga-ui/cdk';
import { combineLatest, merge, Observable } from 'rxjs';
import { filter, mapTo, takeUntil, tap } from 'rxjs/operators';
var SLIDER_INTERACTION_KEYS = new Set([
    'ArrowLeft',
    'ArrowRight',
    'ArrowUp',
    'ArrowDown',
    'Home',
    'End',
    'PageUp',
    'PageDown',
]);
/**
 * Native <input type='range' readonly> doesn't work.
 * This directive imitates this native behaviour.
 */
// @dynamic
var TuiSliderReadonlyDirective = /** @class */ (function () {
    function TuiSliderReadonlyDirective(elementRef, documentRef, destroy$) {
        var _this = this;
        this.readonly = true;
        var touchStart$ = typedFromEvent(elementRef.nativeElement, 'touchstart', {
            passive: false,
        });
        var touchMove$ = typedFromEvent(documentRef, 'touchmove', {
            passive: false,
        });
        var touchEnd$ = typedFromEvent(documentRef, 'touchend', {
            passive: true,
        });
        var shouldPreventMove$ = merge(touchStart$.pipe(tap(function (e) { return _this.preventEvent(e); }), mapTo(true)), touchEnd$.pipe(mapTo(false)));
        /**
         * @bad TODO think about another solution.
         * Keep in mind that preventing touch event (on slider) inside `@HostListener('touchstart')` doesn't work for mobile chrome.
         */
        combineLatest([touchMove$, shouldPreventMove$])
            .pipe(filter(function (_a) {
            var _b = __read(_a, 2), _ = _b[0], shouldPreventMove = _b[1];
            return shouldPreventMove;
        }), takeUntil(destroy$))
            .subscribe(function (_a) {
            var _b = __read(_a, 1), moveEvent = _b[0];
            return _this.preventEvent(moveEvent);
        });
    }
    TuiSliderReadonlyDirective.prototype.preventEvent = function (event) {
        if (event.cancelable && tuiCoerceBooleanProperty(this.readonly)) {
            event.preventDefault();
        }
    };
    TuiSliderReadonlyDirective.prototype.preventKeyboardInteraction = function (event) {
        if (SLIDER_INTERACTION_KEYS.has(event.key)) {
            this.preventEvent(event);
        }
    };
    TuiSliderReadonlyDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiSliderReadonlyDirective.prototype, "readonly", void 0);
    __decorate([
        HostListener('mousedown', ['$event'])
    ], TuiSliderReadonlyDirective.prototype, "preventEvent", null);
    __decorate([
        HostListener('keydown', ['$event'])
    ], TuiSliderReadonlyDirective.prototype, "preventKeyboardInteraction", null);
    TuiSliderReadonlyDirective = __decorate([
        Directive({
            selector: 'input[tuiSlider][readonly]',
            providers: [TuiDestroyService],
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(DOCUMENT)),
        __param(2, Inject(TuiDestroyService))
    ], TuiSliderReadonlyDirective);
    return TuiSliderReadonlyDirective;
}());
export { TuiSliderReadonlyDirective };
