// @taiga-ui/cdk/esm5/utils/math/round.js
import { tuiAssert } from '@taiga-ui/cdk/classes';
var MAX_PRECISION = 292;
/**
 * @deprecated: use {@link tuiRound} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function round(value, precision) {
    if (precision === void 0) { precision = 0; }
    return calculate(value, precision, Math.round);
}
/**
 * @deprecated: use {@link tuiCeil} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function ceil(value, precision) {
    if (precision === void 0) { precision = 0; }
    return calculate(value, precision, Math.ceil);
}
/**
 * @deprecated: use {@link tuiFloor} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function floor(value, precision) {
    if (precision === void 0) { precision = 0; }
    return calculate(value, precision, Math.floor);
}
export var tuiRound = round;
export var tuiCeil = ceil;
export var tuiFloor = floor;
/**
 * Rounding number to the set precision
 *
 * @param value
 * @param precision number of digits in a float part
 * @param func rounding function (round, floor, ceil)
 */
function calculate(value, precision, func) {
    if (value === Infinity) {
        return value;
    }
    tuiAssert.assert(!Number.isNaN(value), "Value must be number");
    tuiAssert.assert(Number.isInteger(precision), "Precision must be integer");
    precision = Math.min(precision, MAX_PRECISION);
    var pair = (value + "e").split("e");
    var tempValue = func(Number(pair[0] + "e" + (Number(pair[1]) + precision)));
    var processedPair = (tempValue + "e").split("e");
    return Number(processedPair[0] + "e" + (Number(processedPair[1]) - precision));
}
