// @taiga-ui/cdk/esm2015/directives/for-async/for-async.directive.js
import { __decorate, __param } from "tslib";
import { Directive, Inject, Input, OnChanges, OnDestroy, TemplateRef, ViewContainerRef, } from '@angular/core';
import { from, of, Subject } from 'rxjs';
import { concatMap, delay, takeUntil } from 'rxjs/operators';
let TuiForAsyncDirective = class TuiForAsyncDirective {
    constructor(view, template) {
        this.view = view;
        this.template = template;
        this.destroy$ = new Subject();
        this.tuiForAsyncTimeout = 10;
    }
    ngOnChanges() {
        this.clearViewForOldNodes();
        this.createAsyncViewForNewNodes();
    }
    ngOnDestroy() {
        this.clearViewForOldNodes();
        this.destroy$.complete();
    }
    createAsyncViewForNewNodes() {
        from((this.tuiForAsyncOf || []).entries())
            .pipe(concatMap(entry => of(entry).pipe(delay(this.tuiForAsyncTimeout))), takeUntil(this.destroy$))
            .subscribe(([index, item]) => this.view
            .createEmbeddedView(this.template, { $implicit: item, index }, index)
            .detectChanges());
    }
    clearViewForOldNodes() {
        this.destroy$.next();
        this.view.clear();
    }
};
TuiForAsyncDirective.ctorParameters = () => [
    { type: ViewContainerRef, decorators: [{ type: Inject, args: [ViewContainerRef,] }] },
    { type: TemplateRef, decorators: [{ type: Inject, args: [TemplateRef,] }] }
];
__decorate([
    Input()
], TuiForAsyncDirective.prototype, "tuiForAsyncOf", void 0);
__decorate([
    Input()
], TuiForAsyncDirective.prototype, "tuiForAsyncTimeout", void 0);
TuiForAsyncDirective = __decorate([
    Directive({ selector: '[tuiForAsync][tuiForAsyncOf]' }),
    __param(0, Inject(ViewContainerRef)),
    __param(1, Inject(TemplateRef))
], TuiForAsyncDirective);
export { TuiForAsyncDirective };
