// @taiga-ui/core/esm5/directives/textfield-controller/textfield-input-mode.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_INPUT_MODE = new InjectionToken('tuiTextfieldInputMode', {
    factory: inputModeDirectiveFactory,
});
var TuiTextfieldInputModeDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldInputModeDirective, _super);
    function TuiTextfieldInputModeDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.inputMode = 'text';
        return _this;
    }
    TuiTextfieldInputModeDirective_1 = TuiTextfieldInputModeDirective;
    var TuiTextfieldInputModeDirective_1;
    __decorate([
        Input('tuiTextfieldInputMode')
    ], TuiTextfieldInputModeDirective.prototype, "inputMode", void 0);
    TuiTextfieldInputModeDirective = TuiTextfieldInputModeDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldInputMode]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_INPUT_MODE,
                    useExisting: forwardRef(function () { return TuiTextfieldInputModeDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldInputModeDirective);
    return TuiTextfieldInputModeDirective;
}(TuiController));
export { TuiTextfieldInputModeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function inputModeDirectiveFactory() {
    return new TuiTextfieldInputModeDirective();
}
