// @taiga-ui/core/esm5/directives/textfield-controller/textfield-custom-content.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_CUSTOM_CONTENT = new InjectionToken('tuiTextfieldCustomContent', {
    factory: customContentDirectiveFactory,
});
var TuiTextfieldCustomContentDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldCustomContentDirective, _super);
    function TuiTextfieldCustomContentDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        // TODO: 3.0 Remove null
        _this.customContent = '';
        return _this;
    }
    TuiTextfieldCustomContentDirective_1 = TuiTextfieldCustomContentDirective;
    var TuiTextfieldCustomContentDirective_1;
    __decorate([
        Input('tuiTextfieldCustomContent')
    ], TuiTextfieldCustomContentDirective.prototype, "customContent", void 0);
    TuiTextfieldCustomContentDirective = TuiTextfieldCustomContentDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldCustomContent]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_CUSTOM_CONTENT,
                    useExisting: forwardRef(function () { return TuiTextfieldCustomContentDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldCustomContentDirective);
    return TuiTextfieldCustomContentDirective;
}(TuiController));
export { TuiTextfieldCustomContentDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function customContentDirectiveFactory() {
    return new TuiTextfieldCustomContentDirective();
}
