// @taiga-ui/core/esm5/directives/textfield-controller/textfield-max-length.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_MAX_LENGTH = new InjectionToken('tuiTextfieldMaxLength', {
    factory: maxLengthDirectiveFactory,
});
var TuiTextfieldMaxLengthDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldMaxLengthDirective, _super);
    function TuiTextfieldMaxLengthDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.maxLength = null;
        return _this;
    }
    TuiTextfieldMaxLengthDirective_1 = TuiTextfieldMaxLengthDirective;
    var TuiTextfieldMaxLengthDirective_1;
    __decorate([
        Input('tuiTextfieldMaxLength')
    ], TuiTextfieldMaxLengthDirective.prototype, "maxLength", void 0);
    TuiTextfieldMaxLengthDirective = TuiTextfieldMaxLengthDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldMaxLength]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_MAX_LENGTH,
                    useExisting: forwardRef(function () { return TuiTextfieldMaxLengthDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldMaxLengthDirective);
    return TuiTextfieldMaxLengthDirective;
}(TuiController));
export { TuiTextfieldMaxLengthDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function maxLengthDirectiveFactory() {
    return new TuiTextfieldMaxLengthDirective();
}
