// @taiga-ui/core/esm5/directives/textfield-controller/textfield-type.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_TYPE = new InjectionToken('tuiTextfieldType', {
    factory: typeDirectiveFactory,
});
var TuiTextfieldTypeDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldTypeDirective, _super);
    function TuiTextfieldTypeDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'text';
        return _this;
    }
    TuiTextfieldTypeDirective_1 = TuiTextfieldTypeDirective;
    var TuiTextfieldTypeDirective_1;
    __decorate([
        Input('tuiTextfieldType')
    ], TuiTextfieldTypeDirective.prototype, "type", void 0);
    TuiTextfieldTypeDirective = TuiTextfieldTypeDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldType]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_TYPE,
                    useExisting: forwardRef(function () { return TuiTextfieldTypeDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldTypeDirective);
    return TuiTextfieldTypeDirective;
}(TuiController));
export { TuiTextfieldTypeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function typeDirectiveFactory() {
    return new TuiTextfieldTypeDirective();
}
