// @taiga-ui/core/esm5/directives/textfield-controller/textfield-size.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_SIZE = new InjectionToken('tuiTextfieldSize', {
    factory: sizeDirectiveFactory,
});
var TuiTextfieldSizeDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldSizeDirective, _super);
    function TuiTextfieldSizeDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.size = 'l';
        return _this;
    }
    TuiTextfieldSizeDirective_1 = TuiTextfieldSizeDirective;
    var TuiTextfieldSizeDirective_1;
    __decorate([
        Input('tuiTextfieldSize')
    ], TuiTextfieldSizeDirective.prototype, "size", void 0);
    TuiTextfieldSizeDirective = TuiTextfieldSizeDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldSize]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_SIZE,
                    useExisting: forwardRef(function () { return TuiTextfieldSizeDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldSizeDirective);
    return TuiTextfieldSizeDirective;
}(TuiController));
export { TuiTextfieldSizeDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function sizeDirectiveFactory() {
    return new TuiTextfieldSizeDirective();
}
