// @taiga-ui/core/esm5/directives/textfield-controller/textfield.controller.js
var TuiTextfieldController = /** @class */ (function () {
    function TuiTextfieldController(change$, autocompleteDirective, cleanerDirective, customContentDirective, exampleTextDirective, inputModeDirective, labelOutsideDirective, maxLengthDirective, sizeDirective, typeDirective) {
        this.change$ = change$;
        this.autocompleteDirective = autocompleteDirective;
        this.cleanerDirective = cleanerDirective;
        this.customContentDirective = customContentDirective;
        this.exampleTextDirective = exampleTextDirective;
        this.inputModeDirective = inputModeDirective;
        this.labelOutsideDirective = labelOutsideDirective;
        this.maxLengthDirective = maxLengthDirective;
        this.sizeDirective = sizeDirective;
        this.typeDirective = typeDirective;
    }
    Object.defineProperty(TuiTextfieldController.prototype, "autocomplete", {
        get: function () {
            return this.autocompleteDirective.autocomplete;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "cleaner", {
        get: function () {
            return this.cleanerDirective.cleaner;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "customContent", {
        get: function () {
            return this.customContentDirective.customContent || "";
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "exampleText", {
        get: function () {
            return this.exampleTextDirective.exampleText;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "inputMode", {
        get: function () {
            return this.inputModeDirective.inputMode;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "labelOutside", {
        get: function () {
            return this.labelOutsideDirective.labelOutside;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "maxLength", {
        get: function () {
            return this.maxLengthDirective.maxLength;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "size", {
        get: function () {
            return this.sizeDirective.size;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiTextfieldController.prototype, "type", {
        get: function () {
            return this.typeDirective.type;
        },
        enumerable: true,
        configurable: true
    });
    return TuiTextfieldController;
}());
export { TuiTextfieldController };
