// @taiga-ui/core/esm5/directives/textfield-controller/textfield-cleaner.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_CLEANER = new InjectionToken('tuiTextfieldCleaner', {
    factory: cleanerDirectiveFactory,
});
var TuiTextfieldCleanerDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldCleanerDirective, _super);
    function TuiTextfieldCleanerDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.cleaner = false;
        return _this;
    }
    TuiTextfieldCleanerDirective_1 = TuiTextfieldCleanerDirective;
    var TuiTextfieldCleanerDirective_1;
    __decorate([
        Input('tuiTextfieldCleaner')
    ], TuiTextfieldCleanerDirective.prototype, "cleaner", void 0);
    TuiTextfieldCleanerDirective = TuiTextfieldCleanerDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldCleaner]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_CLEANER,
                    useExisting: forwardRef(function () { return TuiTextfieldCleanerDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldCleanerDirective);
    return TuiTextfieldCleanerDirective;
}(TuiController));
export { TuiTextfieldCleanerDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function cleanerDirectiveFactory() {
    return new TuiTextfieldCleanerDirective();
}
