// @taiga-ui/core/esm5/directives/textfield-controller/textfield-label-outside.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_LABEL_OUTSIDE = new InjectionToken('tuiTextfieldLabelOutside', {
    factory: labelOutsideDirectiveFactory,
});
var TuiTextfieldLabelOutsideDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldLabelOutsideDirective, _super);
    function TuiTextfieldLabelOutsideDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.labelOutside = false;
        return _this;
    }
    TuiTextfieldLabelOutsideDirective_1 = TuiTextfieldLabelOutsideDirective;
    var TuiTextfieldLabelOutsideDirective_1;
    __decorate([
        Input('tuiTextfieldLabelOutside')
    ], TuiTextfieldLabelOutsideDirective.prototype, "labelOutside", void 0);
    TuiTextfieldLabelOutsideDirective = TuiTextfieldLabelOutsideDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldLabelOutside]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_LABEL_OUTSIDE,
                    useExisting: forwardRef(function () { return TuiTextfieldLabelOutsideDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldLabelOutsideDirective);
    return TuiTextfieldLabelOutsideDirective;
}(TuiController));
export { TuiTextfieldLabelOutsideDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function labelOutsideDirectiveFactory() {
    return new TuiTextfieldLabelOutsideDirective();
}
