// @taiga-ui/core/esm5/directives/textfield-controller/textfield-autocomplete.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_AUTOCOMPLETE = new InjectionToken('tuiTextfieldAutocomplete', {
    factory: autocompleteDirectiveFactory,
});
var TuiTextfieldAutocompleteDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldAutocompleteDirective, _super);
    function TuiTextfieldAutocompleteDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.autocomplete = '';
        return _this;
    }
    TuiTextfieldAutocompleteDirective_1 = TuiTextfieldAutocompleteDirective;
    var TuiTextfieldAutocompleteDirective_1;
    __decorate([
        Input('tuiTextfieldAutocomplete')
    ], TuiTextfieldAutocompleteDirective.prototype, "autocomplete", void 0);
    TuiTextfieldAutocompleteDirective = TuiTextfieldAutocompleteDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldAutocomplete]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_AUTOCOMPLETE,
                    useExisting: forwardRef(function () { return TuiTextfieldAutocompleteDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldAutocompleteDirective);
    return TuiTextfieldAutocompleteDirective;
}(TuiController));
export { TuiTextfieldAutocompleteDirective };
// eslint-disable-next-line @typescript-eslint/naming-convention
export function autocompleteDirectiveFactory() {
    return new TuiTextfieldAutocompleteDirective();
}
