// @taiga-ui/core/esm5/directives/textfield-controller/textfield-controller.provider.js
import { __read, __spread } from "tslib";
import { ChangeDetectorRef, InjectionToken } from '@angular/core';
import { TuiDestroyService, watch } from '@taiga-ui/cdk';
import { merge, NEVER } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TuiTextfieldController } from './textfield.controller';
import { TUI_TEXTFIELD_AUTOCOMPLETE, } from './textfield-autocomplete.directive';
import { TUI_TEXTFIELD_CLEANER, } from './textfield-cleaner.directive';
import { TUI_TEXTFIELD_CUSTOM_CONTENT, } from './textfield-custom-content.directive';
import { TUI_TEXTFIELD_EXAMPLE_TEXT, } from './textfield-example-text.directive';
import { TUI_TEXTFIELD_INPUT_MODE, } from './textfield-input-mode.directive';
import { TUI_TEXTFIELD_LABEL_OUTSIDE, } from './textfield-label-outside.directive';
import { TUI_TEXTFIELD_MAX_LENGTH, } from './textfield-max-length.directive';
import { TUI_TEXTFIELD_SIZE } from './textfield-size.directive';
import { TUI_TEXTFIELD_TYPE } from './textfield-type.directive';
export var TUI_TEXTFIELD_WATCHED_CONTROLLER = new InjectionToken("watched textfield controller");
/**
 * @deprecated: use {@link tuiTextfieldWatchedControllerFactory} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function textfieldWatchedControllerFactory(changeDetectorRef, destroy$) {
    var controllers = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        controllers[_i - 2] = arguments[_i];
    }
    var change$ = merge.apply(void 0, __spread(controllers.map(function (_a) {
        var change$ = _a.change$;
        return change$ || NEVER;
    }))).pipe(watch(changeDetectorRef), takeUntil(destroy$));
    change$.subscribe();
    return new (TuiTextfieldController.bind.apply(TuiTextfieldController, __spread([void 0, change$], controllers)))();
}
export var TEXTFIELD_CONTROLLER_PROVIDER = [
    TuiDestroyService,
    {
        provide: TUI_TEXTFIELD_WATCHED_CONTROLLER,
        deps: [
            ChangeDetectorRef,
            TuiDestroyService,
            TUI_TEXTFIELD_AUTOCOMPLETE,
            TUI_TEXTFIELD_CLEANER,
            TUI_TEXTFIELD_CUSTOM_CONTENT,
            TUI_TEXTFIELD_EXAMPLE_TEXT,
            TUI_TEXTFIELD_INPUT_MODE,
            TUI_TEXTFIELD_LABEL_OUTSIDE,
            TUI_TEXTFIELD_MAX_LENGTH,
            TUI_TEXTFIELD_SIZE,
            TUI_TEXTFIELD_TYPE,
        ],
        useFactory: textfieldWatchedControllerFactory,
    },
];
export var tuiTextfieldWatchedControllerFactory = textfieldWatchedControllerFactory;
