// @taiga-ui/addon-preview/esm5/components/preview/zoom/preview-zoom.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Inject, Input, Output, } from '@angular/core';
import { TUI_PREVIEW_ZOOM_TEXTS } from '@taiga-ui/addon-preview/tokens';
import { clamp, tuiDefaultProp } from '@taiga-ui/cdk';
import { merge, Observable, of, timer } from 'rxjs';
import { mapTo, startWith, switchMap } from 'rxjs/operators';
var STEP = 0.5;
var TuiPreviewZoomComponent = /** @class */ (function () {
    function TuiPreviewZoomComponent(zoomTexts$) {
        this.zoomTexts$ = zoomTexts$;
        this.min = 0.5;
        this.max = 2;
        this.value = 1;
        this.valueChange = new EventEmitter();
        this.reset = new EventEmitter();
        this.hintShow$ = this.valueChange.pipe(switchMap(function () { return merge(of(true), timer(1000).pipe(mapTo(false))); }), startWith(false));
    }
    Object.defineProperty(TuiPreviewZoomComponent.prototype, "leftButtonDisabled", {
        get: function () {
            return this.value === this.min;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPreviewZoomComponent.prototype, "rightButtonDisabled", {
        get: function () {
            return this.value === this.max;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiPreviewZoomComponent.prototype, "collapseVisible", {
        get: function () {
            return this.value > this.min;
        },
        enumerable: true,
        configurable: true
    });
    TuiPreviewZoomComponent.prototype.onModelChange = function (value) {
        var clamped = clamp(value, this.min, this.max);
        if (clamped === this.value) {
            return;
        }
        this.value = clamped;
        this.valueChange.emit(clamped);
    };
    TuiPreviewZoomComponent.prototype.onReset = function () {
        this.reset.emit();
    };
    TuiPreviewZoomComponent.prototype.onMinus = function () {
        this.onModelChange(this.value - STEP);
    };
    TuiPreviewZoomComponent.prototype.onPlus = function () {
        this.onModelChange(this.value < 1 ? 1 : this.value + STEP);
    };
    TuiPreviewZoomComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_PREVIEW_ZOOM_TEXTS,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPreviewZoomComponent.prototype, "min", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPreviewZoomComponent.prototype, "max", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiPreviewZoomComponent.prototype, "value", void 0);
    __decorate([
        Output()
    ], TuiPreviewZoomComponent.prototype, "valueChange", void 0);
    __decorate([
        Output()
    ], TuiPreviewZoomComponent.prototype, "reset", void 0);
    TuiPreviewZoomComponent = __decorate([
        Component({
            selector: 'tui-preview-zoom',
            template: "<ng-container *ngIf=\"zoomTexts$ | async as texts\">\n    <section class=\"t-zoom\">\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            icon=\"tuiIconMinus\"\n            class=\"t-sign_minus\"\n            [title]=\"texts.zoomOut\"\n            [disabled]=\"leftButtonDisabled\"\n            (click)=\"onMinus()\"\n        ></button>\n        <label tuiSliderThumbLabel>\n            <div\n                tuiHintMode=\"onDark\"\n                tuiHintDirection=\"top-right\"\n                [tuiManualHint]=\"hint\"\n                [tuiManualHintShow]=\"!!(hintShow$ | async)\"\n            ></div>\n\n            <ng-template #hint>\n                {{ (valueChange | async) || value | percent }}\n            </ng-template>\n\n            <input\n                tuiSlider\n                type=\"range\"\n                step=\"any\"\n                class=\"t-slider\"\n                [min]=\"min\"\n                [max]=\"max\"\n                [ngModel]=\"value\"\n                (ngModelChange)=\"onModelChange($event)\"\n            />\n        </label>\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            icon=\"tuiIconPlus\"\n            class=\"t-sign_plus\"\n            [title]=\"texts.zoomIn\"\n            [disabled]=\"rightButtonDisabled\"\n            (click)=\"onPlus()\"\n        ></button>\n    </section>\n\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconCollapse\"\n        tuiHintId=\"Collapse\"\n        tuiHintDirection=\"top-right\"\n        tuiHintMode=\"onDark\"\n        class=\"t-reset-button\"\n        [tuiHint]=\"texts.reset\"\n        [class.t-invisible]=\"!collapseVisible\"\n        (click)=\"onReset()\"\n    ></button>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{position:relative;display:flex}.t-zoom{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:12rem}.t-slider{width:7.5rem}.t-sign_minus{border-top-right-radius:0;border-bottom-right-radius:0}.t-sign_plus{border-top-left-radius:0;border-bottom-left-radius:0}.t-invisible{visibility:hidden}.t-reset-button{margin-left:.3125rem}"]
        }),
        __param(0, Inject(TUI_PREVIEW_ZOOM_TEXTS))
    ], TuiPreviewZoomComponent);
    return TuiPreviewZoomComponent;
}());
export { TuiPreviewZoomComponent };
