// @taiga-ui/cdk/esm2015/directives/validator/validator.directive.js
var TuiValidatorDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { NG_VALIDATORS, Validators, } from '@angular/forms';
import { EMPTY_FUNCTION } from '@taiga-ui/cdk/constants';
import { tuiDefaultProp } from '@taiga-ui/cdk/decorators';
let TuiValidatorDirective = TuiValidatorDirective_1 = class TuiValidatorDirective {
    constructor() {
        this.onChange = EMPTY_FUNCTION;
        this.tuiValidator = Validators.nullValidator;
    }
    validate(control) {
        return this.tuiValidator(control);
    }
    registerOnValidatorChange(onChange) {
        this.onChange = onChange;
    }
    ngOnChanges() {
        this.onChange();
    }
    ngOnDestroy() {
        this.tuiValidator = Validators.nullValidator;
        this.onChange();
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], TuiValidatorDirective.prototype, "tuiValidator", void 0);
TuiValidatorDirective = TuiValidatorDirective_1 = __decorate([
    Directive({
        selector: '[tuiValidator]',
        providers: [
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => TuiValidatorDirective_1),
                multi: true,
            },
        ],
    })
], TuiValidatorDirective);
export { TuiValidatorDirective };
