// @taiga-ui/kit/esm5/components/progress/progress-segmented/progress-segmented.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function nonNegativeInt(value) {
    return Number.isInteger(value) && value >= 0;
}
// eslint-disable-next-line @typescript-eslint/naming-convention
export function positiveInt(value) {
    return Number.isInteger(value) && value > 0;
}
var TuiProgressSegmentedComponent = /** @class */ (function () {
    function TuiProgressSegmentedComponent() {
        this.value = 0;
        this.max = 1;
        this.size = 'm';
        this.colors = 'var(--tui-primary)';
    }
    TuiProgressSegmentedComponent.prototype.getActiveColor = function (index) {
        if (index === void 0) { index = 0; }
        return typeof this.colors === 'string'
            ? this.colors
            : this.colors[index] || this.colors[this.colors.length - 1];
    };
    __decorate([
        Input(),
        tuiDefaultProp(nonNegativeInt, 'Must be non-negative integer between 0 and max')
    ], TuiProgressSegmentedComponent.prototype, "value", void 0);
    __decorate([
        Input(),
        tuiDefaultProp(positiveInt, 'Must be positive integer')
    ], TuiProgressSegmentedComponent.prototype, "max", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-size')
    ], TuiProgressSegmentedComponent.prototype, "size", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiProgressSegmentedComponent.prototype, "colors", void 0);
    TuiProgressSegmentedComponent = __decorate([
        Component({
            selector: 'tui-progress-segmented',
            template: "<progress\n    class=\"t-hidden-progress\"\n    [value]=\"value\"\n    [max]=\"max\"\n></progress>\n\n<span\n    *tuiRepeatTimes=\"let index of max\"\n    aria-hidden=\"true\"\n    class=\"t-segment\"\n    [style.background]=\"index < value ? getActiveColor(index) : null\"\n></span>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{display:flex}.t-segment{transition-property:background-color;transition-duration:var(--tui-duration,300ms);transition-timing-function:ease-in-out;flex:1 1 100%;border-radius:var(--tui-radius-xs);background-color:var(--tui-base-03)}.t-segment:not(:last-child){margin-right:.25rem}:host[data-size='s'] .t-segment{height:.25rem}:host[data-size='m'] .t-segment{height:.5rem}.t-hidden-progress{position:absolute;clip:rect(1px,1px,1px,1px);-webkit-clip-path:inset(50%);clip-path:inset(50%);height:1px;width:1px;margin:-1px;overflow:hidden;padding:0}"]
        })
    ], TuiProgressSegmentedComponent);
    return TuiProgressSegmentedComponent;
}());
export { TuiProgressSegmentedComponent };
