// @taiga-ui/kit/esm2015/components/input-phone/input-phone.directive.js
var TuiInputPhoneDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
let TuiInputPhoneDirective = TuiInputPhoneDirective_1 = class TuiInputPhoneDirective extends TuiAbstractTextfieldHost {
    get value() {
        return this.host.computedValue;
    }
    onValueChange(value) {
        this.host.onValueChange(value);
    }
    process(input) {
        this.input = input;
    }
    ngDoCheck() {
        if (!this.input) {
            return;
        }
        this.input.type = 'tel';
        this.input.inputMode = this.host.inputMode;
    }
};
TuiInputPhoneDirective = TuiInputPhoneDirective_1 = __decorate([
    Directive({
        selector: 'tui-input-phone',
        providers: [
            {
                provide: TUI_TEXTFIELD_HOST,
                useExisting: forwardRef(() => TuiInputPhoneDirective_1),
            },
        ],
    })
], TuiInputPhoneDirective);
export { TuiInputPhoneDirective };
