// @taiga-ui/kit/esm5/components/tree/components/tree/tree.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input, Optional, ViewChild, } from '@angular/core';
import { Subject } from 'rxjs';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
import { TuiTreeChildrenDirective } from '../../directives/tree-children.directive';
import { TUI_TREE_NODE } from '../../misc/tree.tokens';
import { TuiTreeItemComponent } from '../tree-item/tree-item.component';
var TuiTreeComponent = /** @class */ (function () {
    function TuiTreeComponent(directive) {
        var _this = this;
        this.directive = directive;
        this.check$ = new Subject();
        this.children$ = this.check$.pipe(startWith(null), map(function () { return _this.handler(_this.value); }), distinctUntilChanged());
        this.content = function (_a) {
            var $implicit = _a.$implicit;
            return String($implicit);
        };
    }
    TuiTreeComponent_1 = TuiTreeComponent;
    TuiTreeComponent.prototype.ngDoCheck = function () {
        var _a, _b;
        this.check$.next();
        (_a = this.item) === null || _a === void 0 ? void 0 : _a.ngDoCheck();
        (_b = this.child) === null || _b === void 0 ? void 0 : _b.ngDoCheck();
    };
    Object.defineProperty(TuiTreeComponent.prototype, "handler", {
        get: function () {
            var _a, _b;
            return (_b = (_a = this.directive) === null || _a === void 0 ? void 0 : _a.childrenHandler) !== null && _b !== void 0 ? _b : TuiTreeChildrenDirective.defaultHandler;
        },
        enumerable: true,
        configurable: true
    });
    var TuiTreeComponent_1;
    TuiTreeComponent.ctorParameters = function () { return [
        { type: TuiTreeChildrenDirective, decorators: [{ type: Optional }, { type: Inject, args: [TuiTreeChildrenDirective,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiTreeComponent.prototype, "value", void 0);
    __decorate([
        ViewChild(TuiTreeItemComponent)
    ], TuiTreeComponent.prototype, "item", void 0);
    __decorate([
        ViewChild(TuiTreeComponent_1)
    ], TuiTreeComponent.prototype, "child", void 0);
    __decorate([
        Input()
    ], TuiTreeComponent.prototype, "content", void 0);
    TuiTreeComponent = TuiTreeComponent_1 = __decorate([
        Component({
            selector: 'tui-tree[value]',
            template: "<tui-tree-item\n    *tuiLet=\"children$ | async as children\"\n    #view\n    [tuiTreeNode]=\"value\"\n>\n    <div\n        *ngIf=\"$any(value) !== children\"\n        polymorpheus-outlet\n        [content]=\"content\"\n        [context]=\"{$implicit: value, node: view}\"\n    ></div>\n    <tui-tree\n        *ngFor=\"let child of children\"\n        [value]=\"child\"\n        [content]=\"content\"\n    ></tui-tree>\n</tui-tree-item>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [
                {
                    provide: TUI_TREE_NODE,
                    useExisting: TuiTreeComponent_1,
                },
            ],
            host: {
                role: 'tree',
            },
            styles: [":host{position:relative;display:block}"]
        }),
        __param(0, Optional()),
        __param(0, Inject(TuiTreeChildrenDirective))
    ], TuiTreeComponent);
    return TuiTreeComponent;
}());
export { TuiTreeComponent };
