// @taiga-ui/kit/esm5/components/input-count/input-count.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputCountDirective = /** @class */ (function (_super) {
    __extends(TuiInputCountDirective, _super);
    function TuiInputCountDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputCountDirective_1 = TuiInputCountDirective;
    TuiInputCountDirective.prototype.onValueChange = function () {
        this.host.onValueChange();
    };
    TuiInputCountDirective.prototype.process = function (input) {
        input.autocomplete = 'off';
        input.inputMode = 'numeric';
    };
    var TuiInputCountDirective_1;
    TuiInputCountDirective = TuiInputCountDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-count',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputCountDirective_1; }),
                },
            ],
        })
    ], TuiInputCountDirective);
    return TuiInputCountDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputCountDirective };
