// @taiga-ui/cdk/esm5/observables/drag-and-drop-from.js
import { TuiOwnerDocumentException } from '@taiga-ui/cdk/exceptions';
import { isPresent } from '@taiga-ui/cdk/utils/miscellaneous';
import { concat, merge } from 'rxjs';
import { endWith, map, repeat, take, takeWhile } from 'rxjs/operators';
import { mouseDragFinishFrom } from './mouse-drag-finish-from';
import { typedFromEvent } from './typed-from-event';
export var TuiDragStage;
(function (TuiDragStage) {
    TuiDragStage[TuiDragStage["Start"] = 0] = "Start";
    TuiDragStage[TuiDragStage["Continues"] = 1] = "Continues";
    TuiDragStage[TuiDragStage["End"] = 2] = "End";
})(TuiDragStage || (TuiDragStage = {}));
var TuiDragState = /** @class */ (function () {
    function TuiDragState(stage, event) {
        this.stage = stage;
        this.event = event;
    }
    return TuiDragState;
}());
export { TuiDragState };
/**
 * @deprecated: use {@link tuiDragAndDropFrom} instead
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export function dragAndDropFrom(element) {
    var ownerDocument = element.ownerDocument;
    if (!ownerDocument) {
        throw new TuiOwnerDocumentException();
    }
    return concat(typedFromEvent(element, "mousedown").pipe(take(1), map(function (event) { return new TuiDragState(0 /* Start */, event); })), merge(typedFromEvent(ownerDocument, "mousemove").pipe(map(function (event) { return new TuiDragState(1 /* Continues */, event); })), mouseDragFinishFrom(ownerDocument).pipe(take(1), map(function (event) { return new TuiDragState(2 /* End */, event); }), endWith(null))).pipe(takeWhile(isPresent))).pipe(repeat());
}
export var tuiDragAndDropFrom = dragAndDropFrom;
