// @taiga-ui/kit/esm5/components/input-date-range/input-date-range.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef } from '@angular/core';
import { TUI_TEXTFIELD_HOST, TuiAbstractTextfieldHost } from '@taiga-ui/core';
var TuiInputDateRangeDirective = /** @class */ (function (_super) {
    __extends(TuiInputDateRangeDirective, _super);
    function TuiInputDateRangeDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiInputDateRangeDirective_1 = TuiInputDateRangeDirective;
    Object.defineProperty(TuiInputDateRangeDirective.prototype, "value", {
        get: function () {
            return this.host.computedValue;
        },
        enumerable: true,
        configurable: true
    });
    TuiInputDateRangeDirective.prototype.onValueChange = function (value) {
        this.host.onValueChange(value);
    };
    TuiInputDateRangeDirective.prototype.process = function (input) {
        input.inputMode = 'numeric';
    };
    var TuiInputDateRangeDirective_1;
    TuiInputDateRangeDirective = TuiInputDateRangeDirective_1 = __decorate([
        Directive({
            selector: 'tui-input-date-range',
            providers: [
                {
                    provide: TUI_TEXTFIELD_HOST,
                    useExisting: forwardRef(function () { return TuiInputDateRangeDirective_1; }),
                },
            ],
        })
    ], TuiInputDateRangeDirective);
    return TuiInputDateRangeDirective;
}(TuiAbstractTextfieldHost));
export { TuiInputDateRangeDirective };
