// @taiga-ui/cdk/esm2015/services/hovered.service.js
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable, NgZone } from '@angular/core';
import { tuiZoneOptimized, typedFromEvent } from '@taiga-ui/cdk/observables';
import { getActualTarget } from '@taiga-ui/cdk/utils/dom';
import { merge } from 'rxjs';
import { distinctUntilChanged, filter, mapTo, startWith, switchMap, take, } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
/** @deprecated TODO remove in v3.0 */
// @dynamic
let TuiHoveredService = class TuiHoveredService {
    constructor(documentRef, ngZone) {
        this.ngZone = ngZone;
        this.documentEvents$ = merge(typedFromEvent(documentRef, `mousemove`), typedFromEvent(documentRef, `touchstart`, { capture: true }));
    }
    createHovered$(target, options = { passive: true }) {
        return merge(typedFromEvent(target, `mouseenter`, options), typedFromEvent(target, `touchstart`, options)).pipe(switchMap(() => merge(typedFromEvent(target, `mouseleave`, options), this.documentEvents$.pipe(filter(event => !target.contains(getActualTarget(event))), tuiZoneOptimized(this.ngZone), take(1))).pipe(mapTo(false), startWith(true))), distinctUntilChanged());
    }
};
TuiHoveredService.ctorParameters = () => [
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] }
];
TuiHoveredService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiHoveredService_Factory() { return new TuiHoveredService(i0.ɵɵinject(i1.DOCUMENT), i0.ɵɵinject(i0.NgZone)); }, token: TuiHoveredService, providedIn: "root" });
TuiHoveredService = __decorate([
    Injectable({
        providedIn: `root`,
    }),
    __param(0, Inject(DOCUMENT)),
    __param(1, Inject(NgZone))
], TuiHoveredService);
export { TuiHoveredService };
