// @taiga-ui/cdk/esm2015/services/scroll.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { ANIMATION_FRAME, PERFORMANCE } from '@ng-web-apis/common';
import { tuiAssert } from '@taiga-ui/cdk/classes';
import { tuiClamp } from '@taiga-ui/cdk/utils/math';
import { tuiEaseInOutQuad } from '@taiga-ui/cdk/utils/miscellaneous';
import { defer, Observable, of, timer } from 'rxjs';
import { map, switchMap, takeUntil, tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@ng-web-apis/common";
const SCROLL_TIME = 300;
// @dynamic
let TuiScrollService = class TuiScrollService {
    constructor(performanceRef, animationFrame$) {
        this.performanceRef = performanceRef;
        this.animationFrame$ = animationFrame$;
    }
    scroll$(elementOrWindow, scrollTop, scrollLeft = getX(elementOrWindow), duration = SCROLL_TIME) {
        tuiAssert.assert(duration >= 0, `Duration cannot be negative`);
        tuiAssert.assert(scrollTop >= 0, `scrollTop cannot be negative`);
        tuiAssert.assert(scrollLeft >= 0, `scrollLeft cannot be negative`);
        const initialTop = getY(elementOrWindow);
        const initialLeft = getX(elementOrWindow);
        const deltaTop = scrollTop - initialTop;
        const deltaLeft = scrollLeft - initialLeft;
        const observable = !duration
            ? of([scrollTop, scrollLeft])
            : defer(() => of(this.performanceRef.now())).pipe(switchMap(start => this.animationFrame$.pipe(map(now => now - start))), map(elapsed => tuiEaseInOutQuad(tuiClamp(elapsed / duration, 0, 1))), map(percent => [
                initialTop + deltaTop * percent,
                initialLeft + deltaLeft * percent,
            ]), takeUntil(timer(duration)));
        return observable.pipe(tap(([scrollTop, scrollLeft]) => {
            var _a;
            (_a = elementOrWindow.scrollTo) === null || _a === void 0 ? void 0 : _a.call(elementOrWindow, scrollLeft, scrollTop);
        }));
    }
};
TuiScrollService.ctorParameters = () => [
    { type: Performance, decorators: [{ type: Inject, args: [PERFORMANCE,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [ANIMATION_FRAME,] }] }
];
TuiScrollService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiScrollService_Factory() { return new TuiScrollService(i0.ɵɵinject(i1.PERFORMANCE), i0.ɵɵinject(i1.ANIMATION_FRAME)); }, token: TuiScrollService, providedIn: "root" });
TuiScrollService = __decorate([
    Injectable({
        providedIn: `root`,
    }),
    __param(0, Inject(PERFORMANCE)),
    __param(1, Inject(ANIMATION_FRAME))
], TuiScrollService);
export { TuiScrollService };
function getX(elementOrWindow) {
    return `scrollX` in elementOrWindow
        ? elementOrWindow.scrollX
        : elementOrWindow.scrollLeft;
}
function getY(elementOrWindow) {
    return `scrollY` in elementOrWindow
        ? elementOrWindow.scrollY
        : elementOrWindow.scrollTop;
}
