// @taiga-ui/addon-preview/bundles/taiga-ui-addon-preview-components-preview.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/kit'), require('rxjs'), require('@taiga-ui/addon-preview/tokens'), require('@taiga-ui/core'), require('rxjs/operators'), require('@angular/common'), require('@angular/forms'), require('@ng-web-apis/mutation-observer'), require('@taiga-ui/addon-preview/components/preview-dialog'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-preview/components/preview', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/kit', 'rxjs', '@taiga-ui/addon-preview/tokens', '@taiga-ui/core', 'rxjs/operators', '@angular/common', '@angular/forms', '@ng-web-apis/mutation-observer', '@taiga-ui/addon-preview/components/preview-dialog', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui']['addon-preview'] = global['taiga-ui']['addon-preview'] || {}, global['taiga-ui']['addon-preview'].components = global['taiga-ui']['addon-preview'].components || {}, global['taiga-ui']['addon-preview'].components.preview = {}), global.ng.core, global.cdk, global.kit, global.rxjs, global['taiga-ui']['addon-preview'].tokens, global.core$1, global.rxjs.operators, global.ng.common, global.ng.forms, global.mutationObserver, global['taiga-ui']['addon-preview'].components['preview-dialog'], global.ngPolymorpheus));
}(this, (function (exports, core, cdk, kit, rxjs, tokens, core$1, operators, common, forms, mutationObserver, previewDialog, ngPolymorpheus) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __esDecorate(ctor, descriptorIn, decorators, contextIn, initializers, extraInitializers) {
        function accept(f) { if (f !== void 0 && typeof f !== "function") throw new TypeError("Function expected"); return f; }
        var kind = contextIn.kind, key = kind === "getter" ? "get" : kind === "setter" ? "set" : "value";
        var target = !descriptorIn && ctor ? contextIn["static"] ? ctor : ctor.prototype : null;
        var descriptor = descriptorIn || (target ? Object.getOwnPropertyDescriptor(target, contextIn.name) : {});
        var _, done = false;
        for (var i = decorators.length - 1; i >= 0; i--) {
            var context = {};
            for (var p in contextIn) context[p] = p === "access" ? {} : contextIn[p];
            for (var p in contextIn.access) context.access[p] = contextIn.access[p];
            context.addInitializer = function (f) { if (done) throw new TypeError("Cannot add initializers after decoration has completed"); extraInitializers.push(accept(f || null)); };
            var result = (0, decorators[i])(kind === "accessor" ? { get: descriptor.get, set: descriptor.set } : descriptor[key], context);
            if (kind === "accessor") {
                if (result === void 0) continue;
                if (result === null || typeof result !== "object") throw new TypeError("Object expected");
                if (_ = accept(result.get)) descriptor.get = _;
                if (_ = accept(result.set)) descriptor.set = _;
                if (_ = accept(result.init)) initializers.push(_);
            }
            else if (_ = accept(result)) {
                if (kind === "field") initializers.push(_);
                else descriptor[key] = _;
            }
        }
        if (target) Object.defineProperty(target, contextIn.name, descriptor);
        done = true;
    };

    function __runInitializers(thisArg, initializers, value) {
        var useValue = arguments.length > 2;
        for (var i = 0; i < initializers.length; i++) {
            value = useValue ? initializers[i].call(thisArg, value) : initializers[i].call(thisArg);
        }
        return useValue ? value : void 0;
    };

    function __propKey(x) {
        return typeof x === "symbol" ? x : "".concat(x);
    };

    function __setFunctionName(f, name, prefix) {
        if (typeof name === "symbol") name = name.description ? "[".concat(name.description, "]") : "";
        return Object.defineProperty(f, "name", { configurable: true, value: prefix ? "".concat(prefix, " ", name) : name });
    };

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (g && (g = 0, op[0] && (_ = 0)), _) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var __createBinding = Object.create ? (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        var desc = Object.getOwnPropertyDescriptor(m, k);
        if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
            desc = { enumerable: true, get: function() { return m[k]; } };
        }
        Object.defineProperty(o, k2, desc);
    }) : (function(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    });

    function __exportStar(m, o) {
        for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p)) __createBinding(o, m, p);
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }

    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
            if (ar || !(i in from)) {
                if (!ar) ar = Array.prototype.slice.call(from, 0, i);
                ar[i] = from[i];
            }
        }
        return to.concat(ar || Array.prototype.slice.call(from));
    }

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: false } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    var __setModuleDefault = Object.create ? (function(o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function(o, v) {
        o["default"] = v;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }

    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m") throw new TypeError("Private method is not writable");
        if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    function __classPrivateFieldIn(state, receiver) {
        if (receiver === null || (typeof receiver !== "object" && typeof receiver !== "function")) throw new TypeError("Cannot use 'in' operator on non-object");
        return typeof state === "function" ? receiver === state : state.has(receiver);
    }

    var TuiPreviewPaginationComponent = /** @class */ (function () {
        function TuiPreviewPaginationComponent(texts$) {
            this.texts$ = texts$;
            this.length = 1;
            this.index = 0;
            this.indexChange = new core.EventEmitter();
        }
        Object.defineProperty(TuiPreviewPaginationComponent.prototype, "leftButtonDisabled", {
            get: function () {
                return this.index === 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPreviewPaginationComponent.prototype, "rightButtonDisabled", {
            get: function () {
                return this.index === this.length - 1;
            },
            enumerable: true,
            configurable: true
        });
        TuiPreviewPaginationComponent.prototype.onArrowClick = function (step) {
            this.updateIndex(cdk.clamp(this.index + step, 0, this.length - 1));
        };
        TuiPreviewPaginationComponent.prototype.updateIndex = function (index) {
            if (this.index === index) {
                return;
            }
            this.index = index;
            this.indexChange.emit(index);
        };
        TuiPreviewPaginationComponent.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [kit.TUI_PAGINATION_TEXTS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewPaginationComponent.prototype, "length", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewPaginationComponent.prototype, "index", void 0);
        __decorate([
            core.Output()
        ], TuiPreviewPaginationComponent.prototype, "indexChange", void 0);
        __decorate([
            core.HostListener('document:keydown.arrowRight.prevent', ['1']),
            core.HostListener('document:keydown.arrowLeft.prevent', ['-1'])
        ], TuiPreviewPaginationComponent.prototype, "onArrowClick", null);
        TuiPreviewPaginationComponent = __decorate([
            core.Component({
                selector: 'tui-preview-pagination',
                template: "<ng-container *ngIf=\"texts$ | async as texts\">\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconArrowLeft\"\n        class=\"t-arrow_left\"\n        [title]=\"texts[0]\"\n        [disabled]=\"leftButtonDisabled\"\n        (click)=\"onArrowClick(-1)\"\n    ></button>\n    {{ index + 1 }}/{{ length }}\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconArrowRight\"\n        class=\"t-arrow_right\"\n        [title]=\"texts[1]\"\n        [disabled]=\"rightButtonDisabled\"\n        (click)=\"onArrowClick(1)\"\n    ></button>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:6.25rem}.t-arrow_left{border-top-right-radius:0;border-bottom-right-radius:0}.t-arrow_right{border-top-left-radius:0;border-bottom-left-radius:0}"]
            }),
            __param(0, core.Inject(kit.TUI_PAGINATION_TEXTS))
        ], TuiPreviewPaginationComponent);
        return TuiPreviewPaginationComponent;
    }());

    var INITIAL_SCALE_COEF = 0.8;
    var EMPTY_COORDINATES = [0, 0];
    var ROTATION_ANGLE = 90;
    var TuiPreviewComponent = /** @class */ (function () {
        function TuiPreviewComponent(elementRef, destroy$, texts$) {
            this.elementRef = elementRef;
            this.destroy$ = destroy$;
            this.texts$ = texts$;
            this.zoomable = true;
            this.rotatable = false;
            this.minZoom = 1;
            this.width = 0;
            this.height = 0;
            this.zoom$ = new rxjs.BehaviorSubject(this.minZoom);
            this.rotation$ = new rxjs.BehaviorSubject(0);
            this.coordinates$ = new rxjs.BehaviorSubject(EMPTY_COORDINATES);
            this.transitioned$ = rxjs.merge(cdk.dragAndDropFrom(this.elementRef.nativeElement).pipe(operators.map(function (_a) {
                var stage = _a.stage;
                return stage !== 1 /* Continues */;
            })), cdk.typedFromEvent(this.elementRef.nativeElement, 'touchmove', { passive: true }).pipe(operators.mapTo(false)), cdk.typedFromEvent(this.elementRef.nativeElement, 'wheel', { passive: true }).pipe(operators.mapTo(false)));
            this.cursor$ = cdk.dragAndDropFrom(this.elementRef.nativeElement).pipe(operators.map(function (_a) {
                var stage = _a.stage;
                return (stage === 1 /* Continues */ ? 'grabbing' : 'initial');
            }), operators.startWith('initial'));
            this.wrapperTransform$ = rxjs.combineLatest([
                this.coordinates$.pipe(operators.map(function (_a) {
                    var _b = __read(_a, 2), x = _b[0], y = _b[1];
                    return cdk.px(x) + ", " + cdk.px(y);
                })),
                this.zoom$,
                this.rotation$,
            ]).pipe(operators.map(function (_a) {
                var _b = __read(_a, 3), translate = _b[0], zoom = _b[1], rotation = _b[2];
                return "translate(" + translate + ") scale(" + zoom + ") rotate(" + rotation + "deg)";
            }));
        }
        TuiPreviewComponent.prototype.rotate = function () {
            this.rotation$.next(this.rotation$.value - ROTATION_ANGLE);
        };
        TuiPreviewComponent.prototype.onPan = function (delta) {
            this.coordinates$.next(this.getGuardedCoordinates(this.coordinates$.value[0] + delta[0], this.coordinates$.value[1] + delta[1]));
        };
        TuiPreviewComponent.prototype.onMutation = function (contentWrapper) {
            var clientWidth = contentWrapper.clientWidth, clientHeight = contentWrapper.clientHeight;
            this.refresh(clientWidth, clientHeight);
        };
        TuiPreviewComponent.prototype.onZoom = function (_a) {
            var clientX = _a.clientX, clientY = _a.clientY, delta = _a.delta;
            if (this.zoomable) {
                this.processZoom(clientX, clientY, delta);
            }
        };
        TuiPreviewComponent.prototype.onResize = function (contentResizeEntries) {
            if (contentResizeEntries.length === 0) {
                return;
            }
            var _a = contentResizeEntries[0].contentRect, width = _a.width, height = _a.height;
            this.refresh(width, height);
        };
        TuiPreviewComponent.prototype.reset = function () {
            this.zoom$.next(this.minZoom);
            this.coordinates$.next(EMPTY_COORDINATES);
        };
        Object.defineProperty(TuiPreviewComponent.prototype, "offsets", {
            get: function () {
                var offsetX = ((this.zoom$.value - this.minZoom) * this.width) / 2;
                var offsetY = ((this.zoom$.value - this.minZoom) * this.height) / 2;
                return { offsetX: offsetX, offsetY: offsetY };
            },
            enumerable: true,
            configurable: true
        });
        TuiPreviewComponent.prototype.calculateMinZoom = function (contentHeight, contentWidth, boxHeight, boxWidth) {
            var bigSize = contentHeight > boxHeight * INITIAL_SCALE_COEF ||
                contentWidth > boxWidth * INITIAL_SCALE_COEF;
            var _a = this.elementRef.nativeElement, clientHeight = _a.clientHeight, clientWidth = _a.clientWidth;
            return bigSize
                ? cdk.round(Math.min((clientHeight * INITIAL_SCALE_COEF) / contentHeight, (clientWidth * INITIAL_SCALE_COEF) / contentWidth), 2)
                : 1;
        };
        TuiPreviewComponent.prototype.refresh = function (width, height) {
            this.width = width;
            this.height = height;
            this.minZoom = this.calculateMinZoom(height, width, this.elementRef.nativeElement.clientHeight, this.elementRef.nativeElement.clientWidth);
            this.zoom$.next(this.minZoom);
            this.coordinates$.next(EMPTY_COORDINATES);
            this.rotation$.next(0);
        };
        TuiPreviewComponent.prototype.processZoom = function (clientX, clientY, delta) {
            var oldScale = this.zoom$.value;
            var newScale = cdk.clamp(oldScale + delta, this.minZoom, 2);
            var center = this.getScaleCenter({ clientX: clientX, clientY: clientY }, this.coordinates$.value, this.zoom$.value);
            var moveX = center[0] * oldScale - center[0] * newScale;
            var moveY = center[1] * oldScale - center[1] * newScale;
            this.zoom$.next(newScale);
            this.coordinates$.next(this.getGuardedCoordinates(this.coordinates$.value[0] + moveX, this.coordinates$.value[1] + moveY));
        };
        TuiPreviewComponent.prototype.getGuardedCoordinates = function (x, y) {
            var _a = this.offsets, offsetX = _a.offsetX, offsetY = _a.offsetY;
            return [cdk.clamp(x, -offsetX, offsetX), cdk.clamp(y, -offsetY, offsetY)];
        };
        TuiPreviewComponent.prototype.getScaleCenter = function (_a, _b, scale) {
            var clientX = _a.clientX, clientY = _a.clientY;
            var _c = __read(_b, 2), x = _c[0], y = _c[1];
            return [
                (clientX - x - this.elementRef.nativeElement.offsetWidth / 2) / scale,
                (clientY - y - this.elementRef.nativeElement.offsetHeight / 2) / scale,
            ];
        };
        TuiPreviewComponent.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_PREVIEW_TEXTS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewComponent.prototype, "zoomable", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewComponent.prototype, "rotatable", void 0);
        TuiPreviewComponent = __decorate([
            core.Component({
                selector: 'tui-preview',
                template: "<ng-container *ngIf=\"texts$ | async as texts\">\n    <section\n        #contentWrapper\n        subtree\n        childList\n        characterData\n        attributeFilter=\"src\"\n        class=\"t-wrapper\"\n        [class.t-not-interactive-content]=\"zoomable\"\n        [class.t-transitive]=\"transitioned$ | async\"\n        [style.transform]=\"wrapperTransform$ | async\"\n        [style.cursor]=\"cursor$ | async\"\n        (tuiPan)=\"onPan($event)\"\n        (tuiZoom)=\"onZoom($event)\"\n        (tuiResize)=\"onResize($event)\"\n        (waMutationObserver)=\"onMutation(contentWrapper)\"\n    >\n        <ng-content></ng-content>\n    </section>\n\n    <header class=\"t-header\">\n        <div class=\"t-title\">\n            <ng-content select=\"tui-preview-title\"></ng-content>\n        </div>\n\n        <ng-content select=\"tui-preview-pagination\"></ng-content>\n\n        <div class=\"t-actions\">\n            <ng-content select=\"[tuiPreviewAction]\"></ng-content>\n        </div>\n    </header>\n\n    <footer class=\"t-footer\">\n        <button\n            *ngIf=\"rotatable\"\n            tuiIconButton\n            tuiPreviewAction\n            icon=\"tuiIconRotate\"\n            tuiHintId=\"Rotate\"\n            tuiHintDirection=\"top-right\"\n            tuiHintMode=\"onDark\"\n            class=\"t-rotate-button\"\n            [tuiHint]=\"texts.rotate\"\n            (click)=\"rotate()\"\n        ></button>\n\n        <tui-preview-zoom\n            *ngIf=\"zoomable\"\n            [min]=\"minZoom\"\n            [value]=\"(zoom$ | async) || 1\"\n            (valueChange)=\"zoom$.next($event)\"\n            (reset)=\"reset()\"\n        ></tui-preview-zoom>\n    </footer>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                animations: [core$1.tuiSlideInTop],
                providers: [cdk.TuiDestroyService],
                styles: [":host{position:relative;display:flex;justify-content:center;align-items:center;width:100%;height:100%;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.t-header{position:fixed;top:1rem;display:flex;width:100%;padding:0 1rem;box-sizing:border-box}.t-footer{position:absolute;bottom:1rem;display:flex;width:100%;padding:0 1rem;box-sizing:border-box;justify-content:center}.t-actions{display:flex;flex:1;justify-content:flex-end}.t-actions ::ng-deep>*{margin-left:.625rem}.t-rotate-button{margin-right:.3125rem}.t-title{flex:1}:host-context(tui-root._mobile) .t-title{display:none}.t-not-interactive-content ::ng-deep>*{pointer-events:none}.t-wrapper{will-change:transform}.t-transitive{transition-duration:.3s}"]
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(cdk.TuiDestroyService)),
            __param(2, core.Inject(tokens.TUI_PREVIEW_TEXTS))
        ], TuiPreviewComponent);
        return TuiPreviewComponent;
    }());

    // TODO: 3.0 remove in ivy compilation
    var PREVIEW_ACTION_OPTIONS = {
        appearance: 'preview-action',
        shape: 'rounded',
        size: 's',
    };
    var TuiPreviewActionDirective = /** @class */ (function () {
        function TuiPreviewActionDirective() {
        }
        TuiPreviewActionDirective = __decorate([
            core.Directive({
                selector: '[tuiPreviewAction]',
                providers: [
                    {
                        provide: core$1.TUI_BUTTON_OPTIONS,
                        useValue: PREVIEW_ACTION_OPTIONS,
                    },
                ],
            })
        ], TuiPreviewActionDirective);
        return TuiPreviewActionDirective;
    }());

    var TuiPreviewActionModule = /** @class */ (function () {
        function TuiPreviewActionModule() {
        }
        TuiPreviewActionModule = __decorate([
            core.NgModule({
                declarations: [TuiPreviewActionDirective],
                exports: [TuiPreviewActionDirective],
            })
        ], TuiPreviewActionModule);
        return TuiPreviewActionModule;
    }());

    var TuiPreviewTitleComponent = /** @class */ (function () {
        function TuiPreviewTitleComponent() {
        }
        TuiPreviewTitleComponent = __decorate([
            core.Component({
                selector: 'tui-preview-title',
                template: "\n        <ng-content></ng-content>\n    ",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:inline-block;padding:.375rem .75rem}"]
            })
        ], TuiPreviewTitleComponent);
        return TuiPreviewTitleComponent;
    }());

    var STEP = 0.5;
    var TuiPreviewZoomComponent = /** @class */ (function () {
        function TuiPreviewZoomComponent(zoomTexts$) {
            this.zoomTexts$ = zoomTexts$;
            this.min = 0.5;
            this.max = 2;
            this.value = 1;
            this.valueChange = new core.EventEmitter();
            this.reset = new core.EventEmitter();
            this.hintShow$ = this.valueChange.pipe(operators.switchMap(function () { return rxjs.merge(rxjs.of(true), rxjs.timer(1000).pipe(operators.mapTo(false))); }), operators.startWith(false));
        }
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "leftButtonDisabled", {
            get: function () {
                return this.value === this.min;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "rightButtonDisabled", {
            get: function () {
                return this.value === this.max;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPreviewZoomComponent.prototype, "collapseVisible", {
            get: function () {
                return this.value > this.min;
            },
            enumerable: true,
            configurable: true
        });
        TuiPreviewZoomComponent.prototype.onModelChange = function (value) {
            var clamped = cdk.clamp(value, this.min, this.max);
            if (clamped === this.value) {
                return;
            }
            this.value = clamped;
            this.valueChange.emit(clamped);
        };
        TuiPreviewZoomComponent.prototype.onReset = function () {
            this.reset.emit();
        };
        TuiPreviewZoomComponent.prototype.onMinus = function () {
            this.onModelChange(this.value - STEP);
        };
        TuiPreviewZoomComponent.prototype.onPlus = function () {
            this.onModelChange(this.value < 1 ? 1 : this.value + STEP);
        };
        TuiPreviewZoomComponent.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_PREVIEW_ZOOM_TEXTS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewZoomComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewZoomComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPreviewZoomComponent.prototype, "value", void 0);
        __decorate([
            core.Output()
        ], TuiPreviewZoomComponent.prototype, "valueChange", void 0);
        __decorate([
            core.Output()
        ], TuiPreviewZoomComponent.prototype, "reset", void 0);
        TuiPreviewZoomComponent = __decorate([
            core.Component({
                selector: 'tui-preview-zoom',
                template: "<ng-container *ngIf=\"zoomTexts$ | async as texts\">\n    <section class=\"t-zoom\">\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            icon=\"tuiIconMinus\"\n            class=\"t-sign_minus\"\n            [title]=\"texts.zoomOut\"\n            [disabled]=\"leftButtonDisabled\"\n            (click)=\"onMinus()\"\n        ></button>\n        <label tuiSliderThumbLabel>\n            <div\n                tuiHintMode=\"onDark\"\n                tuiHintDirection=\"top-right\"\n                [tuiManualHint]=\"hint\"\n                [tuiManualHintShow]=\"!!(hintShow$ | async)\"\n            ></div>\n\n            <ng-template #hint>\n                {{ (valueChange | async) || value | percent }}\n            </ng-template>\n\n            <input\n                tuiSlider\n                type=\"range\"\n                step=\"any\"\n                class=\"t-slider\"\n                [min]=\"min\"\n                [max]=\"max\"\n                [ngModel]=\"value\"\n                (ngModelChange)=\"onModelChange($event)\"\n            />\n        </label>\n        <button\n            tuiIconButton\n            tuiPreviewAction\n            icon=\"tuiIconPlus\"\n            class=\"t-sign_plus\"\n            [title]=\"texts.zoomIn\"\n            [disabled]=\"rightButtonDisabled\"\n            (click)=\"onPlus()\"\n        ></button>\n    </section>\n\n    <button\n        tuiIconButton\n        tuiPreviewAction\n        icon=\"tuiIconCollapse\"\n        tuiHintId=\"Collapse\"\n        tuiHintDirection=\"top-right\"\n        tuiHintMode=\"onDark\"\n        class=\"t-reset-button\"\n        [tuiHint]=\"texts.reset\"\n        [class.t-invisible]=\"!collapseVisible\"\n        (click)=\"onReset()\"\n    ></button>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{position:relative;display:flex}.t-zoom{border-radius:1rem;background:rgba(104,104,104,.96);color:var(--tui-text-01-night);font:var(--tui-font-text-s);display:flex;justify-content:space-between;align-items:center;width:12rem}.t-slider{width:7.5rem}.t-sign_minus{border-top-right-radius:0;border-bottom-right-radius:0}.t-sign_plus{border-top-left-radius:0;border-bottom-left-radius:0}.t-invisible{visibility:hidden}.t-reset-button{margin-left:.3125rem}"]
            }),
            __param(0, core.Inject(tokens.TUI_PREVIEW_ZOOM_TEXTS))
        ], TuiPreviewZoomComponent);
        return TuiPreviewZoomComponent;
    }());

    var TuiPreviewModule = /** @class */ (function () {
        function TuiPreviewModule() {
        }
        TuiPreviewModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    ngPolymorpheus.PolymorpheusModule,
                    mutationObserver.MutationObserverModule,
                    core$1.TuiButtonModule,
                    forms.FormsModule,
                    kit.TuiSliderModule,
                    core$1.TuiHintModule,
                    cdk.TuiDragModule,
                    core$1.TuiManualHintModule,
                    cdk.TuiResizeModule,
                    core$1.TuiSvgModule,
                    core$1.TuiLoaderModule,
                    cdk.TuiPanModule,
                    cdk.TuiZoomModule,
                    TuiPreviewActionModule,
                ],
                declarations: [
                    TuiPreviewComponent,
                    TuiPreviewTitleComponent,
                    TuiPreviewPaginationComponent,
                    TuiPreviewZoomComponent,
                ],
                exports: [
                    TuiPreviewComponent,
                    TuiPreviewActionDirective,
                    TuiPreviewTitleComponent,
                    TuiPreviewPaginationComponent,
                    TuiPreviewZoomComponent,
                ],
                providers: [previewDialog.PREVIEW_DIALOG_PROVIDER],
            })
        ], TuiPreviewModule);
        return TuiPreviewModule;
    }());

    exports.PREVIEW_ACTION_OPTIONS = PREVIEW_ACTION_OPTIONS;
    exports.TuiPreviewActionDirective = TuiPreviewActionDirective;
    exports.TuiPreviewActionModule = TuiPreviewActionModule;
    exports.TuiPreviewComponent = TuiPreviewComponent;
    exports.TuiPreviewModule = TuiPreviewModule;
    exports.TuiPreviewPaginationComponent = TuiPreviewPaginationComponent;
    exports.TuiPreviewTitleComponent = TuiPreviewTitleComponent;
    exports.TuiPreviewZoomComponent = TuiPreviewZoomComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
