// @taiga-ui/kit/esm2015/components/tree/directives/tree-item-controller.directive.js
var TuiTreeItemControllerDirective_1;
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_TREE_CONTROLLER } from '../misc/tree.tokens';
let TuiTreeItemControllerDirective = TuiTreeItemControllerDirective_1 = class TuiTreeItemControllerDirective {
    constructor() {
        this.map = new WeakMap();
        this.fallback = true;
    }
    isExpanded(item) {
        var _a;
        return (_a = this.map.get(item)) !== null && _a !== void 0 ? _a : this.fallback;
    }
    toggle(item) {
        this.map.set(item, !this.isExpanded(item));
    }
};
__decorate([
    Input('tuiTreeController'),
    tuiDefaultProp()
], TuiTreeItemControllerDirective.prototype, "fallback", void 0);
TuiTreeItemControllerDirective = TuiTreeItemControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiTreeController]:not([map])',
        exportAs: 'tuiTreeController',
        providers: [
            {
                provide: TUI_TREE_CONTROLLER,
                useExisting: TuiTreeItemControllerDirective_1,
            },
        ],
    })
], TuiTreeItemControllerDirective);
export { TuiTreeItemControllerDirective };
