// @taiga-ui/kit/esm2015/components/tree/directives/tree-controller.directive.js
var TuiTreeControllerDirective_1;
import { __decorate } from "tslib";
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { isPresent, tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_TREE_ACCESSOR, TUI_TREE_CONTROLLER } from '../misc/tree.tokens';
let TuiTreeControllerDirective = TuiTreeControllerDirective_1 = class TuiTreeControllerDirective {
    constructor() {
        this.fallback = true;
        this.map = new Map();
        this.toggled = new EventEmitter();
        this.items = new Map();
    }
    register(item, value) {
        this.items.set(item, value);
    }
    unregister(item) {
        this.items.delete(item);
    }
    isExpanded(item) {
        var _a;
        const value = this.items.get(item);
        return (_a = (value && this.map.get(value))) !== null && _a !== void 0 ? _a : this.fallback;
    }
    toggle(item) {
        const value = this.items.get(item);
        const expanded = this.isExpanded(item);
        if (!isPresent(value)) {
            return;
        }
        this.toggled.emit(value);
        this.map.set(value, !expanded);
    }
};
__decorate([
    Input('tuiTreeController'),
    tuiDefaultProp()
], TuiTreeControllerDirective.prototype, "fallback", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiTreeControllerDirective.prototype, "map", void 0);
__decorate([
    Output()
], TuiTreeControllerDirective.prototype, "toggled", void 0);
TuiTreeControllerDirective = TuiTreeControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiTreeController][map]',
        exportAs: 'tuiTreeController',
        providers: [
            {
                provide: TUI_TREE_ACCESSOR,
                useExisting: TuiTreeControllerDirective_1,
            },
            {
                provide: TUI_TREE_CONTROLLER,
                useExisting: TuiTreeControllerDirective_1,
            },
        ],
    })
], TuiTreeControllerDirective);
export { TuiTreeControllerDirective };
