// @taiga-ui/kit/esm5/components/tabs/tabs-with-more/tabs-with-more.providers.js
import { DOCUMENT } from '@angular/common';
import { ChangeDetectorRef, ElementRef, InjectionToken } from '@angular/core';
import { MUTATION_OBSERVER_INIT, MutationObserverService, } from '@ng-web-apis/mutation-observer';
import { TuiDestroyService, TuiResizeService } from '@taiga-ui/cdk';
import { merge } from 'rxjs';
import { debounceTime, filter, startWith, takeUntil, tap } from 'rxjs/operators';
// eslint-disable-next-line @typescript-eslint/naming-convention
export function tabsRefreshFactory(resize$, mutations$, destroy$, _a, _b, changeDetectorRef) {
    var body = _a.body;
    var nativeElement = _b.nativeElement;
    return merge(resize$, mutations$.pipe(tap(function () { return changeDetectorRef.detectChanges(); }))).pipe(
    // Ignoring cases when host is detached from DOM
    filter(function () { return body.contains(nativeElement); }), debounceTime(0), startWith(null), takeUntil(destroy$));
}
// TODO: 3.0 remove in ivy compilation
export var MUTATION_CONFIG = {
    childList: true,
    subtree: true,
    characterData: true,
};
export var TABS_REFRESH = new InjectionToken("Refresh stream");
export var TABS_PROVIDERS = [
    TuiResizeService,
    TuiDestroyService,
    MutationObserverService,
    {
        provide: MUTATION_OBSERVER_INIT,
        useValue: MUTATION_CONFIG,
    },
    {
        provide: TABS_REFRESH,
        deps: [
            TuiResizeService,
            MutationObserverService,
            TuiDestroyService,
            DOCUMENT,
            ElementRef,
            ChangeDetectorRef,
        ],
        useFactory: tabsRefreshFactory,
    },
];
