// @taiga-ui/kit/esm2015/pipes/field-error/field-error-content-pipe.js
import { __decorate, __param } from "tslib";
import { AsyncPipe } from '@angular/common';
import { Inject, Injector, OnDestroy, Pipe, PipeTransform } from '@angular/core';
import { TuiFieldErrorPipe } from './field-error-pipe';
// @dynamic
let TuiFieldErrorContentPipe = class TuiFieldErrorContentPipe {
    constructor(injector) {
        this.injector = injector;
        this.localInjector = Injector.create({
            providers: [{ provide: AsyncPipe }, { provide: TuiFieldErrorPipe }],
            parent: this.injector,
        });
        this.asyncPipe = this.localInjector.get(AsyncPipe);
        this.fieldErrorPipe = this.localInjector.get(TuiFieldErrorPipe);
    }
    transform(order) {
        return this.getErrorContent(order);
    }
    ngOnDestroy() {
        this.asyncPipe.ngOnDestroy();
    }
    getErrorContent(order) {
        const error = this.asyncPipe.transform(this.fieldErrorPipe.transform(order));
        if (!error) {
            return ``;
        }
        return typeof error.message === `function`
            ? error.message(error.context)
            : error.message;
    }
};
TuiFieldErrorContentPipe.ctorParameters = () => [
    { type: Injector, decorators: [{ type: Inject, args: [Injector,] }] }
];
TuiFieldErrorContentPipe = __decorate([
    Pipe({
        name: `tuiFieldErrorContent`,
        pure: false,
    }),
    __param(0, Inject(Injector))
], TuiFieldErrorContentPipe);
export { TuiFieldErrorContentPipe };
