// @taiga-ui/kit/esm5/components/tree/directives/tree-item-controller.directive.js
import { __decorate } from "tslib";
import { Directive, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_TREE_CONTROLLER } from '../misc/tree.tokens';
var TuiTreeItemControllerDirective = /** @class */ (function () {
    function TuiTreeItemControllerDirective() {
        this.map = new WeakMap();
        this.fallback = true;
    }
    TuiTreeItemControllerDirective_1 = TuiTreeItemControllerDirective;
    TuiTreeItemControllerDirective.prototype.isExpanded = function (item) {
        var _a;
        return (_a = this.map.get(item)) !== null && _a !== void 0 ? _a : this.fallback;
    };
    TuiTreeItemControllerDirective.prototype.toggle = function (item) {
        this.map.set(item, !this.isExpanded(item));
    };
    var TuiTreeItemControllerDirective_1;
    __decorate([
        Input('tuiTreeController'),
        tuiDefaultProp()
    ], TuiTreeItemControllerDirective.prototype, "fallback", void 0);
    TuiTreeItemControllerDirective = TuiTreeItemControllerDirective_1 = __decorate([
        Directive({
            selector: '[tuiTreeController]:not([map])',
            exportAs: 'tuiTreeController',
            providers: [
                {
                    provide: TUI_TREE_CONTROLLER,
                    useExisting: TuiTreeItemControllerDirective_1,
                },
            ],
        })
    ], TuiTreeItemControllerDirective);
    return TuiTreeItemControllerDirective;
}());
export { TuiTreeItemControllerDirective };
