// @taiga-ui/kit/esm5/components/tree/directives/tree-controller.directive.js
import { __decorate } from "tslib";
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { isPresent, tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_TREE_ACCESSOR, TUI_TREE_CONTROLLER } from '../misc/tree.tokens';
var TuiTreeControllerDirective = /** @class */ (function () {
    function TuiTreeControllerDirective() {
        this.fallback = true;
        this.map = new Map();
        this.toggled = new EventEmitter();
        this.items = new Map();
    }
    TuiTreeControllerDirective_1 = TuiTreeControllerDirective;
    TuiTreeControllerDirective.prototype.register = function (item, value) {
        this.items.set(item, value);
    };
    TuiTreeControllerDirective.prototype.unregister = function (item) {
        this.items.delete(item);
    };
    TuiTreeControllerDirective.prototype.isExpanded = function (item) {
        var _a;
        var value = this.items.get(item);
        return (_a = (value && this.map.get(value))) !== null && _a !== void 0 ? _a : this.fallback;
    };
    TuiTreeControllerDirective.prototype.toggle = function (item) {
        var value = this.items.get(item);
        var expanded = this.isExpanded(item);
        if (!isPresent(value)) {
            return;
        }
        this.toggled.emit(value);
        this.map.set(value, !expanded);
    };
    var TuiTreeControllerDirective_1;
    __decorate([
        Input('tuiTreeController'),
        tuiDefaultProp()
    ], TuiTreeControllerDirective.prototype, "fallback", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiTreeControllerDirective.prototype, "map", void 0);
    __decorate([
        Output()
    ], TuiTreeControllerDirective.prototype, "toggled", void 0);
    TuiTreeControllerDirective = TuiTreeControllerDirective_1 = __decorate([
        Directive({
            selector: '[tuiTreeController][map]',
            exportAs: 'tuiTreeController',
            providers: [
                {
                    provide: TUI_TREE_ACCESSOR,
                    useExisting: TuiTreeControllerDirective_1,
                },
                {
                    provide: TUI_TREE_CONTROLLER,
                    useExisting: TuiTreeControllerDirective_1,
                },
            ],
        })
    ], TuiTreeControllerDirective);
    return TuiTreeControllerDirective;
}());
export { TuiTreeControllerDirective };
